diff --git a/c/parallel/src/merge_sort.cu b/c/parallel/src/merge_sort.cu
index ec11d4eccda..5aab61d713d 100644
--- a/c/parallel/src/merge_sort.cu
+++ b/c/parallel/src/merge_sort.cu
@@ -10,10 +10,11 @@
 
 #include <cub/detail/choose_offset.cuh>
 #include <cub/detail/launcher/cuda_driver.cuh>
-#include <cub/detail/ptx-json-parser.cuh>
 #include <cub/device/device_merge_sort.cuh>
+#include <cub/device/dispatch/tuning/tuning_merge_sort.cuh>
 
 #include <format>
+#include <sstream>
 #include <vector>
 
 #include "kernels/iterators.h"
@@ -39,25 +40,6 @@ struct output_items_iterator_t;
 
 namespace merge_sort
 {
-struct merge_sort_runtime_tuning_policy
-{
-  cub::detail::RuntimeMergeSortAgentPolicy merge_sort;
-
-  auto MergeSort() const
-  {
-    return merge_sort;
-  }
-
-  using MergeSortPolicy = cub::detail::RuntimeMergeSortAgentPolicy;
-  using MaxPolicy       = merge_sort_runtime_tuning_policy;
-
-  template <typename F>
-  cudaError_t Invoke(int, F& op)
-  {
-    return op.template Invoke<merge_sort_runtime_tuning_policy>(*this);
-  }
-};
-
 enum class merge_sort_iterator_t
 {
   input_keys   = 0,
@@ -137,7 +119,7 @@ std::string get_merge_sort_kernel_name(
       : cccl_type_enum_to_name<items_storage_t>(output_items_it.value_type.type);
 
   return std::format(
-    "cub::detail::merge_sort::{0}<{1}, {2}, {3}, {4}, {5}, {6}, {7}, {8}, {9}, device_merge_sort_vsmem_helper>",
+    "cub::detail::merge_sort::{0}<{1}, {2}, {3}, {4}, {5}, {6}, {7}, {8}, {9}>",
     kernel_name,
     chained_policy_t,
     input_keys_iterator_t,
@@ -199,33 +181,6 @@ struct merge_sort_kernel_source
     return build.item_type.size;
   }
 };
-
-struct dynamic_vsmem_helper_t
-{
-  template <typename PolicyT, typename... Ts>
-  static ::cuda::std::size_t BlockSortVSMemPerBlock(PolicyT /*policy*/)
-  {
-    return 0;
-  }
-
-  template <typename PolicyT, typename... Ts>
-  static ::cuda::std::size_t MergeVSMemPerBlock(PolicyT /*policy*/)
-  {
-    return 0;
-  }
-
-  template <typename PolicyT, typename... Ts>
-  static int BlockThreads(PolicyT policy)
-  {
-    return policy.BlockThreads();
-  }
-
-  template <typename PolicyT, typename... Ts>
-  static int ItemsPerTile(PolicyT policy)
-  {
-    return policy.ItemsPerTile();
-  }
-};
 } // namespace merge_sort
 
 CUresult cccl_device_merge_sort_build_ex(
@@ -277,8 +232,14 @@ try
 
   const std::string op_src = make_kernel_user_comparison_operator(input_keys_it_value_t, op);
 
-  std::string policy_hub_expr =
-    std::format("cub::detail::merge_sort::policy_hub<{}>",
+  const auto policy_sel = cub::detail::merge_sort::policy_selector{static_cast<int>(input_keys_it.value_type.size)};
+
+  // TODO(bgruber): drop this if tuning policies become formattable
+  std::stringstream policy_sel_str;
+  policy_sel_str << policy_sel(cuda::to_arch_id(cuda::compute_capability{cc_major, cc_minor}));
+
+  auto policy_selector_expr =
+    std::format("cub::detail::merge_sort::policy_selector_from_types<{}>",
                 get_iterator_name(input_keys_it, merge_sort::merge_sort_iterator_t::input_keys));
 
   std::string final_src = std::format(
@@ -297,36 +258,10 @@ struct __align__({3}) items_storage_t {{
 {6}
 {7}
 {8}
-using device_merge_sort_policy = {9}::MaxPolicy;
-
-struct device_merge_sort_vsmem_helper {{
-  template<typename ActivePolicyT, typename KeyInputIteratorT, typename ValueInputIteratorT, typename... Ts>
-  struct MergeSortVSMemHelperT {{
-    using policy_t = device_merge_sort_policy::ActivePolicy::MergeSortPolicy;
-    using block_sort_agent_t = cub::detail::merge_sort::AgentBlockSort<policy_t, KeyInputIteratorT, ValueInputIteratorT, Ts...>;
-    using merge_agent_t = cub::detail::merge_sort::AgentMerge<policy_t, Ts...>;
-  }};
-  template <typename AgentT>
-  struct VSmemHelperT {{
-    using static_temp_storage_t = typename AgentT::TempStorage;
-    static _CCCL_DEVICE _CCCL_FORCEINLINE static_temp_storage_t& get_temp_storage(
-      static_temp_storage_t& static_temp_storage, cub::detail::vsmem_t& vsmem)
-    {{
-        return static_temp_storage;
-    }}
-    template <bool needs_vsmem_ = false, ::cuda::std::enable_if_t<!needs_vsmem_, int> = 0>
-    static _CCCL_DEVICE _CCCL_FORCEINLINE bool discard_temp_storage(static_temp_storage_t& temp_storage)
-    {{
-      return false;
-    }}
-  }};
-}};
-
-#include <cub/detail/ptx-json/json.cuh>
-__device__ consteval auto& policy_generator() {{
-  return ptx_json::id<ptx_json::string("device_merge_sort_policy")>()
-    = cub::detail::merge_sort::MergeSortPolicyWrapper<device_merge_sort_policy::ActivePolicy>::EncodedPolicy();
-}}
+using device_merge_sort_policy = {9};
+using namespace cub;
+using namespace cub::detail::merge_sort;
+static_assert(device_merge_sort_policy()(::cuda::arch_id{{CUB_PTX_ARCH / 10}}) == {10}, "Host generated and JIT compiled policy mismatch");
 )XXX",
     input_keys_it.value_type.size, // 0
     input_keys_it.value_type.alignment, // 1
@@ -337,7 +272,8 @@ __device__ consteval auto& policy_generator() {{
     output_keys_iterator_src, // 6
     output_items_iterator_src, // 7
     op_src, // 8
-    policy_hub_expr); // 9
+    policy_selector_expr, // 9
+    policy_sel_str.view()); // 10
 
 #if false // CCCL_DEBUGGING_SWITCH
   fflush(stderr);
@@ -365,7 +301,6 @@ __device__ consteval auto& policy_generator() {{
     "-rdc=true",
     "-dlto",
     "-DCUB_DISABLE_CDP",
-    "-DCUB_ENABLE_POLICY_PTX_JSON",
     "-std=c++20"};
 
   cccl::detail::extend_args_with_build_config(args, config);
@@ -402,18 +337,12 @@ __device__ consteval auto& policy_generator() {{
   check(cuLibraryGetKernel(&build_ptr->partition_kernel, build_ptr->library, partition_kernel_lowered_name.c_str()));
   check(cuLibraryGetKernel(&build_ptr->merge_kernel, build_ptr->library, merge_kernel_lowered_name.c_str()));
 
-  nlohmann::json runtime_policy =
-    cub::detail::ptx_json::parse("device_merge_sort_policy", {result.data.get(), result.size});
-
-  using cub::detail::RuntimeMergeSortAgentPolicy;
-  auto ms_policy = RuntimeMergeSortAgentPolicy::from_json(runtime_policy, "MergeSortPolicy");
-
   build_ptr->cc             = cc;
   build_ptr->cubin          = (void*) result.data.release();
   build_ptr->cubin_size     = result.size;
   build_ptr->key_type       = input_keys_it.value_type;
   build_ptr->item_type      = input_items_it.value_type;
-  build_ptr->runtime_policy = new merge_sort::merge_sort_runtime_tuning_policy{ms_policy};
+  build_ptr->runtime_policy = new cub::detail::merge_sort::policy_selector{policy_sel};
 
   return CUDA_SUCCESS;
 }
@@ -456,30 +385,21 @@ CUresult cccl_device_merge_sort(
     CUdevice cu_device;
     check(cuCtxGetDevice(&cu_device));
 
-    auto exec_status = cub::DispatchMergeSort<
-      indirect_arg_t,
-      indirect_arg_t,
-      indirect_arg_t,
-      indirect_arg_t,
-      OffsetT,
-      indirect_arg_t,
-      merge_sort::merge_sort_runtime_tuning_policy,
-      merge_sort::merge_sort_kernel_source,
-      cub::detail::CudaDriverLauncherFactory,
-      merge_sort::dynamic_vsmem_helper_t,
-      indirect_arg_t,
-      indirect_arg_t>::Dispatch(d_temp_storage,
-                                *temp_storage_bytes,
-                                d_in_keys,
-                                d_in_items,
-                                d_out_keys,
-                                d_out_items,
-                                num_items,
-                                op,
-                                stream,
-                                {build},
-                                cub::detail::CudaDriverLauncherFactory{cu_device, build.cc},
-                                *reinterpret_cast<merge_sort::merge_sort_runtime_tuning_policy*>(build.runtime_policy));
+    auto exec_status = cub::detail::merge_sort::dispatch(
+      d_temp_storage,
+      *temp_storage_bytes,
+      indirect_iterator_t{d_in_keys},
+      indirect_iterator_t{d_in_items},
+      indirect_iterator_t{d_out_keys},
+      indirect_iterator_t{d_out_items},
+      static_cast<OffsetT>(num_items),
+      indirect_arg_t{op},
+      stream,
+      *static_cast<cub::detail::merge_sort::policy_selector*>(build.runtime_policy),
+      merge_sort::merge_sort_kernel_source{build},
+      cub::detail::CudaDriverLauncherFactory{cu_device, build.cc},
+      static_cast<indirect_arg_t*>(nullptr),
+      static_cast<indirect_arg_t*>(nullptr));
 
     error = static_cast<CUresult>(exec_status);
   }
@@ -539,7 +459,8 @@ try
   }
 
   std::unique_ptr<char[]> cubin(reinterpret_cast<char*>(build_ptr->cubin));
-  std::unique_ptr<char[]> policy(reinterpret_cast<char*>(build_ptr->runtime_policy));
+  std::unique_ptr<cub::detail::merge_sort::policy_selector> policy(
+    static_cast<cub::detail::merge_sort::policy_selector*>(build_ptr->runtime_policy));
   check(cuLibraryUnload(build_ptr->library));
 
   return CUDA_SUCCESS;
diff --git a/cub/benchmarks/bench/merge_sort/keys.cu b/cub/benchmarks/bench/merge_sort/keys.cu
index 64b6db4fa9d..22afc7be022 100644
--- a/cub/benchmarks/bench/merge_sort/keys.cu
+++ b/cub/benchmarks/bench/merge_sort/keys.cu
@@ -18,60 +18,28 @@
 using value_t = cub::NullType;
 
 #if !TUNE_BASE
-
-#  if TUNE_TRANSPOSE == 0
-#    define TUNE_LOAD_ALGORITHM  cub::BLOCK_LOAD_DIRECT
-#    define TUNE_STORE_ALGORITHM cub::BLOCK_STORE_DIRECT
-#  else // TUNE_TRANSPOSE == 1
-#    define TUNE_LOAD_ALGORITHM  cub::BLOCK_LOAD_WARP_TRANSPOSE
-#    define TUNE_STORE_ALGORITHM cub::BLOCK_STORE_WARP_TRANSPOSE
-#  endif // TUNE_TRANSPOSE
-
-#  if TUNE_LOAD == 0
-#    define TUNE_LOAD_MODIFIER cub::LOAD_DEFAULT
-#  elif TUNE_LOAD == 1
-#    define TUNE_LOAD_MODIFIER cub::LOAD_LDG
-#  else // TUNE_LOAD == 2
-#    define TUNE_LOAD_MODIFIER cub::LOAD_CA
-#  endif // TUNE_LOAD
-
 template <typename KeyT>
-struct policy_hub_t
+struct policy_selector
 {
-  struct policy_t : cub::ChainedPolicy<300, policy_t, policy_t>
+  _CCCL_API constexpr auto operator()(::cuda::arch_id /*arch*/) const -> cub::detail::merge_sort::merge_sort_policy
   {
-    using MergeSortPolicy =
-      cub::AgentMergeSortPolicy<TUNE_THREADS_PER_BLOCK,
-                                cub::Nominal4BItemsToItems<KeyT>(TUNE_ITEMS_PER_THREAD),
-                                TUNE_LOAD_ALGORITHM,
-                                TUNE_LOAD_MODIFIER,
-                                TUNE_STORE_ALGORITHM>;
-  };
-
-  using MaxPolicy = policy_t;
+    return cub::detail::merge_sort::merge_sort_policy{
+      TUNE_THREADS_PER_BLOCK,
+      cub::Nominal4BItemsToItems<KeyT>(TUNE_ITEMS_PER_THREAD),
+      (TUNE_TRANSPOSE == 0 ? cub::BLOCK_LOAD_DIRECT : cub::BLOCK_LOAD_WARP_TRANSPOSE),
+      (TUNE_LOAD == 0 ? cub::LOAD_DEFAULT : (TUNE_LOAD == 1 ? cub::LOAD_LDG : cub::LOAD_CA)),
+      (TUNE_TRANSPOSE == 0 ? cub::BLOCK_STORE_DIRECT : cub::BLOCK_STORE_WARP_TRANSPOSE)};
+  }
 };
 #endif // !TUNE_BASE
 
 template <typename T, typename OffsetT>
 void keys(nvbench::state& state, nvbench::type_list<T, OffsetT>)
 {
-  using key_t            = T;
-  using value_t          = cub::NullType;
-  using key_input_it_t   = key_t*;
-  using value_input_it_t = value_t*;
-  using key_it_t         = key_t*;
-  using value_it_t       = value_t*;
-  using offset_t         = cub::detail::choose_offset_t<OffsetT>;
-  using compare_op_t     = less_t;
-
-#if !TUNE_BASE
-  using policy_t = policy_hub_t<key_t>;
-  using dispatch_t =
-    cub::DispatchMergeSort<key_input_it_t, value_input_it_t, key_it_t, value_it_t, offset_t, compare_op_t, policy_t>;
-#else // TUNE_BASE
-  using dispatch_t =
-    cub::DispatchMergeSort<key_input_it_t, value_input_it_t, key_it_t, value_it_t, offset_t, compare_op_t>;
-#endif // TUNE_BASE
+  using key_t        = T;
+  using value_t      = cub::NullType;
+  using offset_t     = cub::detail::choose_offset_t<OffsetT>;
+  using compare_op_t = less_t;
 
   // Retrieve axis parameters
   const auto elements       = static_cast<std::size_t>(state.get_int64("Elements{io}"));
@@ -90,31 +58,41 @@ void keys(nvbench::state& state, nvbench::type_list<T, OffsetT>)
 
   // Allocate temporary storage:
   std::size_t temp_size{};
-  dispatch_t::Dispatch(
+  cub::detail::merge_sort::dispatch(
     nullptr,
     temp_size,
     d_buffer_1,
-    nullptr,
+    static_cast<value_t*>(nullptr),
     d_buffer_2,
-    nullptr,
+    static_cast<value_t*>(nullptr),
     static_cast<offset_t>(elements),
     compare_op_t{},
-    0 /* stream */);
+    0 /* stream */
+#if !TUNE_BASE
+    ,
+    policy_selector<key_t>{}
+#endif // !TUNE_BASE
+  );
 
-  thrust::device_vector<nvbench::uint8_t> temp(temp_size);
+  thrust::device_vector<nvbench::uint8_t> temp(temp_size, thrust::no_init);
   auto* temp_storage = thrust::raw_pointer_cast(temp.data());
 
   state.exec(nvbench::exec_tag::gpu | nvbench::exec_tag::no_batch, [&](nvbench::launch& launch) {
-    dispatch_t::Dispatch(
+    cub::detail::merge_sort::dispatch(
       temp_storage,
       temp_size,
       d_buffer_1,
-      nullptr,
+      static_cast<value_t*>(nullptr),
       d_buffer_2,
-      nullptr,
+      static_cast<value_t*>(nullptr),
       static_cast<offset_t>(elements),
       compare_op_t{},
-      launch.get_stream());
+      launch.get_stream()
+#if !TUNE_BASE
+        ,
+      policy_selector<key_t>{}
+#endif // !TUNE_BASE
+    );
   });
 }
 
diff --git a/cub/benchmarks/bench/merge_sort/pairs.cu b/cub/benchmarks/bench/merge_sort/pairs.cu
index 8dab8340dc7..e091b4a1d08 100644
--- a/cub/benchmarks/bench/merge_sort/pairs.cu
+++ b/cub/benchmarks/bench/merge_sort/pairs.cu
@@ -16,59 +16,28 @@
 #endif
 
 #if !TUNE_BASE
-#  if TUNE_TRANSPOSE == 0
-#    define TUNE_LOAD_ALGORITHM  cub::BLOCK_LOAD_DIRECT
-#    define TUNE_STORE_ALGORITHM cub::BLOCK_STORE_DIRECT
-#  else // TUNE_TRANSPOSE == 1
-#    define TUNE_LOAD_ALGORITHM  cub::BLOCK_LOAD_WARP_TRANSPOSE
-#    define TUNE_STORE_ALGORITHM cub::BLOCK_STORE_WARP_TRANSPOSE
-#  endif // TUNE_TRANSPOSE
-
-#  if TUNE_LOAD == 0
-#    define TUNE_LOAD_MODIFIER cub::LOAD_DEFAULT
-#  elif TUNE_LOAD == 1
-#    define TUNE_LOAD_MODIFIER cub::LOAD_LDG
-#  else // TUNE_LOAD == 2
-#    define TUNE_LOAD_MODIFIER cub::LOAD_CA
-#  endif // TUNE_LOAD
-
 template <typename KeyT>
-struct policy_hub_t
+struct policy_selector
 {
-  struct policy_t : cub::ChainedPolicy<300, policy_t, policy_t>
+  _CCCL_API constexpr auto operator()(::cuda::arch_id /*arch*/) const -> cub::detail::merge_sort::merge_sort_policy
   {
-    using MergeSortPolicy =
-      cub::AgentMergeSortPolicy<TUNE_THREADS_PER_BLOCK,
-                                cub::Nominal4BItemsToItems<KeyT>(TUNE_ITEMS_PER_THREAD),
-                                TUNE_LOAD_ALGORITHM,
-                                TUNE_LOAD_MODIFIER,
-                                TUNE_STORE_ALGORITHM>;
-  };
-
-  using MaxPolicy = policy_t;
+    return cub::detail::merge_sort::merge_sort_policy{
+      TUNE_THREADS_PER_BLOCK,
+      cub::Nominal4BItemsToItems<KeyT>(TUNE_ITEMS_PER_THREAD),
+      (TUNE_TRANSPOSE == 0 ? cub::BLOCK_LOAD_DIRECT : cub::BLOCK_LOAD_WARP_TRANSPOSE),
+      (TUNE_LOAD == 0 ? cub::LOAD_DEFAULT : (TUNE_LOAD == 1 ? cub::LOAD_LDG : cub::LOAD_CA)),
+      (TUNE_TRANSPOSE == 0 ? cub::BLOCK_STORE_DIRECT : cub::BLOCK_STORE_WARP_TRANSPOSE)};
+  }
 };
 #endif // TUNE_BASE
 
 template <typename KeyT, typename ValueT, typename OffsetT>
 void pairs(nvbench::state& state, nvbench::type_list<KeyT, ValueT, OffsetT>)
 {
-  using key_t            = KeyT;
-  using value_t          = ValueT;
-  using key_input_it_t   = key_t*;
-  using value_input_it_t = value_t*;
-  using key_it_t         = key_t*;
-  using value_it_t       = value_t*;
-  using offset_t         = cub::detail::choose_offset_t<OffsetT>;
-  using compare_op_t     = less_t;
-
-#if !TUNE_BASE
-  using policy_t = policy_hub_t<key_t>;
-  using dispatch_t =
-    cub::DispatchMergeSort<key_input_it_t, value_input_it_t, key_it_t, value_it_t, offset_t, compare_op_t, policy_t>;
-#else // TUNE_BASE
-  using dispatch_t =
-    cub::DispatchMergeSort<key_input_it_t, value_input_it_t, key_it_t, value_it_t, offset_t, compare_op_t>;
-#endif // TUNE_BASE
+  using key_t        = KeyT;
+  using value_t      = ValueT;
+  using offset_t     = cub::detail::choose_offset_t<OffsetT>;
+  using compare_op_t = less_t;
 
   // Retrieve axis parameters
   const auto elements       = static_cast<std::size_t>(state.get_int64("Elements{io}"));
@@ -93,7 +62,7 @@ void pairs(nvbench::state& state, nvbench::type_list<KeyT, ValueT, OffsetT>)
 
   // Allocate temporary storage:
   std::size_t temp_size{};
-  dispatch_t::Dispatch(
+  cub::detail::merge_sort::dispatch(
     nullptr,
     temp_size,
     d_keys_buffer_1,
@@ -102,13 +71,18 @@ void pairs(nvbench::state& state, nvbench::type_list<KeyT, ValueT, OffsetT>)
     d_values_buffer_2,
     static_cast<offset_t>(elements),
     compare_op_t{},
-    0 /* stream */);
+    0 /* stream */
+#if !TUNE_BASE
+    ,
+    policy_selector<key_t>{}
+#endif // !TUNE_BASE
+  );
 
-  thrust::device_vector<nvbench::uint8_t> temp(temp_size);
+  thrust::device_vector<nvbench::uint8_t> temp(temp_size, thrust::no_init);
   auto* temp_storage = thrust::raw_pointer_cast(temp.data());
 
   state.exec(nvbench::exec_tag::gpu | nvbench::exec_tag::no_batch, [&](nvbench::launch& launch) {
-    dispatch_t::Dispatch(
+    cub::detail::merge_sort::dispatch(
       temp_storage,
       temp_size,
       d_keys_buffer_1,
@@ -117,7 +91,12 @@ void pairs(nvbench::state& state, nvbench::type_list<KeyT, ValueT, OffsetT>)
       d_values_buffer_2,
       static_cast<offset_t>(elements),
       compare_op_t{},
-      launch.get_stream());
+      launch.get_stream()
+#if !TUNE_BASE
+        ,
+      policy_selector<key_t>{}
+#endif // !TUNE_BASE
+    );
   });
 }
 
diff --git a/cub/cub/agent/agent_merge_sort.cuh b/cub/cub/agent/agent_merge_sort.cuh
index 4b0e3fec532..d3812169e12 100644
--- a/cub/cub/agent/agent_merge_sort.cuh
+++ b/cub/cub/agent/agent_merge_sort.cuh
@@ -16,55 +16,17 @@
 #include <cub/block/block_load.cuh>
 #include <cub/block/block_merge_sort.cuh>
 #include <cub/block/block_store.cuh>
+#include <cub/device/dispatch/tuning/tuning_merge_sort.cuh>
 #include <cub/iterator/cache_modified_input_iterator.cuh>
 #include <cub/util_namespace.cuh>
 #include <cub/util_type.cuh>
 
-#include <cuda/std/__algorithm/max.h>
 #include <cuda/std/__algorithm/min.h>
 
 CUB_NAMESPACE_BEGIN
-
-template <int BlockThreads,
-          int ItemsPerThread                      = 1,
-          cub::BlockLoadAlgorithm LoadAlgorithm   = cub::BLOCK_LOAD_DIRECT,
-          cub::CacheLoadModifier LoadModifier     = cub::LOAD_LDG,
-          cub::BlockStoreAlgorithm StoreAlgorithm = cub::BLOCK_STORE_DIRECT>
-struct AgentMergeSortPolicy
-{
-  static constexpr int BLOCK_THREADS    = BlockThreads;
-  static constexpr int ITEMS_PER_THREAD = ItemsPerThread;
-  static constexpr int ITEMS_PER_TILE   = BLOCK_THREADS * ITEMS_PER_THREAD;
-
-  static constexpr cub::BlockLoadAlgorithm LOAD_ALGORITHM   = LoadAlgorithm;
-  static constexpr cub::CacheLoadModifier LOAD_MODIFIER     = LoadModifier;
-  static constexpr cub::BlockStoreAlgorithm STORE_ALGORITHM = StoreAlgorithm;
-};
-
-#if defined(CUB_DEFINE_RUNTIME_POLICIES) || defined(CUB_ENABLE_POLICY_PTX_JSON)
-namespace detail
-{
-// Only define this when needed.
-// Because of overload woes, this depends on C++20 concepts. util_device.h checks that concepts are available when
-// either runtime policies or PTX JSON information are enabled, so if they are, this is always valid. The generic
-// version is always defined, and that's the only one needed for regular CUB operations.
-//
-// TODO: enable this unconditionally once concepts are always available
-CUB_DETAIL_POLICY_WRAPPER_DEFINE(
-  MergeSortAgentPolicy,
-  (GenericAgentPolicy),
-  (BLOCK_THREADS, BlockThreads, int),
-  (ITEMS_PER_THREAD, ItemsPerThread, int),
-  (ITEMS_PER_TILE, ItemsPerTile, int),
-  (LOAD_ALGORITHM, LoadAlgorithm, cub::BlockLoadAlgorithm),
-  (LOAD_MODIFIER, LoadModifier, cub::CacheLoadModifier),
-  (STORE_ALGORITHM, StoreAlgorithm, cub::BlockStoreAlgorithm))
-} // namespace detail
-#endif // defined(CUB_DEFINE_RUNTIME_POLICIES
-
 namespace detail::merge_sort
 {
-template <typename Policy,
+template <typename PolicyGetter,
           typename KeyInputIteratorT,
           typename ValueInputIteratorT,
           typename KeyIteratorT,
@@ -81,18 +43,25 @@ struct AgentBlockSort
 
   static constexpr bool KEYS_ONLY = ::cuda::std::is_same_v<ValueT, NullType>;
 
-  using BlockMergeSortT = BlockMergeSort<KeyT, Policy::BLOCK_THREADS, Policy::ITEMS_PER_THREAD, ValueT>;
+  static constexpr merge_sort_policy policy = PolicyGetter{}();
+  static constexpr int BLOCK_THREADS        = policy.block_threads;
+  static constexpr int ITEMS_PER_THREAD     = policy.items_per_thread;
+  static constexpr int ITEMS_PER_TILE       = policy.items_per_tile();
+
+  using BlockMergeSortT = BlockMergeSort<KeyT, BLOCK_THREADS, ITEMS_PER_THREAD, ValueT>;
 
-  using KeysLoadIt  = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, KeyInputIteratorT>;
-  using ItemsLoadIt = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, ValueInputIteratorT>;
+  using KeysLoadIt  = try_make_cache_modified_iterator_t<policy.load_modifier, KeyInputIteratorT>;
+  using ItemsLoadIt = try_make_cache_modified_iterator_t<policy.load_modifier, ValueInputIteratorT>;
 
-  using BlockLoadKeys  = typename cub::BlockLoadType<Policy, KeysLoadIt>::type;
-  using BlockLoadItems = typename cub::BlockLoadType<Policy, ItemsLoadIt>::type;
+  using BlockLoadKeys  = BlockLoad<it_value_t<KeysLoadIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.load_algorithm>;
+  using BlockLoadItems = BlockLoad<it_value_t<ItemsLoadIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.load_algorithm>;
 
-  using BlockStoreKeysIt   = typename cub::BlockStoreType<Policy, KeyIteratorT>::type;
-  using BlockStoreItemsIt  = typename cub::BlockStoreType<Policy, ValueIteratorT>::type;
-  using BlockStoreKeysRaw  = typename cub::BlockStoreType<Policy, KeyT*>::type;
-  using BlockStoreItemsRaw = typename cub::BlockStoreType<Policy, ValueT*>::type;
+  using BlockStoreKeysIt =
+    BlockStore<it_value_t<KeyIteratorT>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+  using BlockStoreItemsIt =
+    BlockStore<it_value_t<ValueIteratorT>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+  using BlockStoreKeysRaw  = BlockStore<KeyT, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+  using BlockStoreItemsRaw = BlockStore<ValueT, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
 
   union _TempStorage
   {
@@ -106,12 +75,7 @@ struct AgentBlockSort
   };
 
   /// Alias wrapper allowing storage to be unioned
-  struct TempStorage : Uninitialized<_TempStorage>
-  {};
-
-  static constexpr int BLOCK_THREADS    = Policy::BLOCK_THREADS;
-  static constexpr int ITEMS_PER_THREAD = Policy::ITEMS_PER_THREAD;
-  static constexpr int ITEMS_PER_TILE   = Policy::ITEMS_PER_TILE;
+  using TempStorage = Uninitialized<_TempStorage>;
 
   //---------------------------------------------------------------------
   // Per thread data
@@ -399,7 +363,7 @@ _CCCL_DEVICE _CCCL_FORCEINLINE void reg_to_shared(It output, T (&input)[ITEMS_PE
 }
 
 /// \brief The agent is responsible for merging N consecutive sorted arrays into N/2 sorted arrays.
-template <typename Policy,
+template <typename PolicyGetter, // TODO(bgruber): pass policy as NTTP in C++20
           typename KeyIteratorT,
           typename ValueIteratorT,
           typename OffsetT,
@@ -411,20 +375,33 @@ struct AgentMerge
   //---------------------------------------------------------------------
   // Types and constants
   //---------------------------------------------------------------------
-  using KeysLoadPingIt  = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, KeyIteratorT>;
-  using ItemsLoadPingIt = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, ValueIteratorT>;
-  using KeysLoadPongIt  = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, KeyT*>;
-  using ItemsLoadPongIt = try_make_cache_modified_iterator_t<Policy::LOAD_MODIFIER, ValueT*>;
+
+  static constexpr bool KEYS_ONLY = ::cuda::std::is_same_v<ValueT, NullType>;
+
+  static constexpr merge_sort_policy policy = PolicyGetter{}();
+  static constexpr int BLOCK_THREADS        = policy.block_threads;
+  static constexpr int ITEMS_PER_THREAD     = policy.items_per_thread;
+  static constexpr int ITEMS_PER_TILE       = policy.items_per_tile();
+
+  using KeysLoadPingIt  = try_make_cache_modified_iterator_t<policy.load_modifier, KeyIteratorT>;
+  using ItemsLoadPingIt = try_make_cache_modified_iterator_t<policy.load_modifier, ValueIteratorT>;
+  using KeysLoadPongIt  = try_make_cache_modified_iterator_t<policy.load_modifier, KeyT*>;
+  using ItemsLoadPongIt = try_make_cache_modified_iterator_t<policy.load_modifier, ValueT*>;
 
   using KeysOutputPongIt  = KeyIteratorT;
   using ItemsOutputPongIt = ValueIteratorT;
   using KeysOutputPingIt  = KeyT*;
   using ItemsOutputPingIt = ValueT*;
 
-  using BlockStoreKeysPong  = typename BlockStoreType<Policy, KeysOutputPongIt>::type;
-  using BlockStoreItemsPong = typename BlockStoreType<Policy, ItemsOutputPongIt>::type;
-  using BlockStoreKeysPing  = typename BlockStoreType<Policy, KeysOutputPingIt>::type;
-  using BlockStoreItemsPing = typename BlockStoreType<Policy, ItemsOutputPingIt>::type;
+  using BlockStoreKeysPong =
+    BlockStore<it_value_t<KeysOutputPongIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+  using BlockStoreItemsPong =
+    BlockStore<it_value_t<ItemsOutputPongIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+
+  using BlockStoreKeysPing =
+    BlockStore<it_value_t<KeysOutputPingIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
+  using BlockStoreItemsPing =
+    BlockStore<it_value_t<ItemsOutputPingIt>, BLOCK_THREADS, ITEMS_PER_THREAD, policy.store_algorithm>;
 
   /// Parameterized BlockReduce primitive
 
@@ -435,18 +412,12 @@ struct AgentMerge
     typename BlockStoreKeysPong::TempStorage store_keys_pong;
     typename BlockStoreItemsPong::TempStorage store_items_pong;
 
-    KeyT keys_shared[Policy::ITEMS_PER_TILE + 1];
-    ValueT items_shared[Policy::ITEMS_PER_TILE + 1];
+    KeyT keys_shared[policy.items_per_tile() + 1];
+    ValueT items_shared[policy.items_per_tile() + 1];
   };
 
   /// Alias wrapper allowing storage to be unioned
-  struct TempStorage : Uninitialized<_TempStorage>
-  {};
-
-  static constexpr bool KEYS_ONLY       = ::cuda::std::is_same_v<ValueT, NullType>;
-  static constexpr int BLOCK_THREADS    = Policy::BLOCK_THREADS;
-  static constexpr int ITEMS_PER_THREAD = Policy::ITEMS_PER_THREAD;
-  static constexpr int ITEMS_PER_TILE   = Policy::ITEMS_PER_TILE;
+  using TempStorage = Uninitialized<_TempStorage>;
 
   //---------------------------------------------------------------------
   // Per thread data
diff --git a/cub/cub/device/device_merge_sort.cuh b/cub/cub/device/device_merge_sort.cuh
index c242f2d2839..4a99ff97952 100644
--- a/cub/cub/device/device_merge_sort.cuh
+++ b/cub/cub/device/device_merge_sort.cuh
@@ -110,11 +110,16 @@ private:
   {
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
-    using DispatchMergeSortT =
-      DispatchMergeSort<KeyIteratorT, ValueIteratorT, KeyIteratorT, ValueIteratorT, ChooseOffsetT, CompareOpT>;
-
-    return DispatchMergeSortT::Dispatch(
-      d_temp_storage, temp_storage_bytes, d_keys, d_items, d_keys, d_items, num_items, compare_op, stream);
+    return detail::merge_sort::dispatch(
+      d_temp_storage,
+      temp_storage_bytes,
+      d_keys,
+      d_items,
+      d_keys,
+      d_items,
+      static_cast<ChooseOffsetT>(num_items),
+      compare_op,
+      stream);
   }
 
 public:
@@ -297,11 +302,8 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyIteratorT, ValueIteratorT, KeyIteratorT, ValueIteratorT, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
-        storage, bytes, d_keys, d_items, d_keys, d_items, num_items, compare_op, stream);
+      return detail::merge_sort::dispatch(
+        storage, bytes, d_keys, d_items, d_keys, d_items, static_cast<ChooseOffsetT>(num_items), compare_op, stream);
     });
   }
 
@@ -441,17 +443,14 @@ public:
     _CCCL_NVTX_RANGE_SCOPE_IF(d_temp_storage, GetName());
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
-    using DispatchMergeSortT =
-      DispatchMergeSort<KeyInputIteratorT, ValueInputIteratorT, KeyIteratorT, ValueIteratorT, ChooseOffsetT, CompareOpT>;
-
-    return DispatchMergeSortT::Dispatch(
+    return detail::merge_sort::dispatch(
       d_temp_storage,
       temp_storage_bytes,
       d_input_keys,
       d_input_items,
       d_output_keys,
       d_output_items,
-      num_items,
+      static_cast<ChooseOffsetT>(num_items),
       compare_op,
       stream);
   }
@@ -546,11 +545,16 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyInputIteratorT, ValueInputIteratorT, KeyIteratorT, ValueIteratorT, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
-        storage, bytes, d_input_keys, d_input_items, d_output_keys, d_output_items, num_items, compare_op, stream);
+      return detail::merge_sort::dispatch(
+        storage,
+        bytes,
+        d_input_keys,
+        d_input_items,
+        d_output_keys,
+        d_output_items,
+        static_cast<ChooseOffsetT>(num_items),
+        compare_op,
+        stream);
     });
   }
 
@@ -567,17 +571,14 @@ private:
   {
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
-    using DispatchMergeSortT =
-      DispatchMergeSort<KeyIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-    return DispatchMergeSortT::Dispatch(
+    return detail::merge_sort::dispatch(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
       static_cast<NullType*>(nullptr),
       d_keys,
       static_cast<NullType*>(nullptr),
-      num_items,
+      static_cast<ChooseOffsetT>(num_items),
       compare_op,
       stream);
   }
@@ -742,17 +743,14 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
+      return detail::merge_sort::dispatch(
         storage,
         bytes,
         d_keys,
         static_cast<NullType*>(nullptr),
         d_keys,
         static_cast<NullType*>(nullptr),
-        num_items,
+        static_cast<ChooseOffsetT>(num_items),
         compare_op,
         stream);
     });
@@ -772,17 +770,14 @@ private:
   {
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
-    using DispatchMergeSortT =
-      DispatchMergeSort<KeyInputIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-    return DispatchMergeSortT::Dispatch(
+    return detail::merge_sort::dispatch(
       d_temp_storage,
       temp_storage_bytes,
       d_input_keys,
       static_cast<NullType*>(nullptr),
       d_output_keys,
       static_cast<NullType*>(nullptr),
-      num_items,
+      static_cast<ChooseOffsetT>(num_items),
       compare_op,
       stream);
   }
@@ -974,17 +969,14 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyInputIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
+      return detail::merge_sort::dispatch(
         storage,
         bytes,
         d_input_keys,
         static_cast<NullType*>(nullptr),
         d_output_keys,
         static_cast<NullType*>(nullptr),
-        num_items,
+        static_cast<ChooseOffsetT>(num_items),
         compare_op,
         stream);
     });
@@ -1171,11 +1163,8 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyIteratorT, ValueIteratorT, KeyIteratorT, ValueIteratorT, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
-        storage, bytes, d_keys, d_items, d_keys, d_items, num_items, compare_op, stream);
+      return detail::merge_sort::dispatch(
+        storage, bytes, d_keys, d_items, d_keys, d_items, static_cast<ChooseOffsetT>(num_items), compare_op, stream);
     });
   }
 
@@ -1341,17 +1330,14 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
+      return detail::merge_sort::dispatch(
         storage,
         bytes,
         d_keys,
         static_cast<NullType*>(nullptr),
         d_keys,
         static_cast<NullType*>(nullptr),
-        num_items,
+        static_cast<ChooseOffsetT>(num_items),
         compare_op,
         stream);
     });
@@ -1544,17 +1530,14 @@ public:
     using ChooseOffsetT = detail::choose_offset_t<OffsetT>;
 
     return detail::dispatch_with_env(env, [&]([[maybe_unused]] auto tuning, void* storage, size_t& bytes, auto stream) {
-      using DispatchMergeSortT =
-        DispatchMergeSort<KeyInputIteratorT, NullType*, KeyIteratorT, NullType*, ChooseOffsetT, CompareOpT>;
-
-      return DispatchMergeSortT::Dispatch(
+      return detail::merge_sort::dispatch(
         storage,
         bytes,
         d_input_keys,
         static_cast<NullType*>(nullptr),
         d_output_keys,
         static_cast<NullType*>(nullptr),
-        num_items,
+        static_cast<ChooseOffsetT>(num_items),
         compare_op,
         stream);
     });
diff --git a/cub/cub/device/dispatch/dispatch_merge_sort.cuh b/cub/cub/device/dispatch/dispatch_merge_sort.cuh
index 0d01f8d89bc..7225e44ba55 100644
--- a/cub/cub/device/dispatch/dispatch_merge_sort.cuh
+++ b/cub/cub/device/dispatch/dispatch_merge_sort.cuh
@@ -1,4 +1,4 @@
-// SPDX-FileCopyrightText: Copyright (c) 2011-2021, NVIDIA CORPORATION. All rights reserved.
+// SPDX-FileCopyrightText: Copyright (c) 2011-2026, NVIDIA CORPORATION. All rights reserved.
 // SPDX-License-Identifier: BSD-3
 
 #pragma once
@@ -14,10 +14,10 @@
 #endif // no system header
 
 #include <cub/agent/agent_merge_sort.cuh>
+#include <cub/detail/arch_dispatch.cuh>
 #include <cub/device/dispatch/kernels/kernel_merge_sort.cuh>
 #include <cub/device/dispatch/tuning/tuning_merge_sort.cuh>
 #include <cub/util_device.cuh>
-#include <cub/util_math.cuh>
 #include <cub/util_namespace.cuh>
 #include <cub/util_vsmem.cuh>
 
@@ -27,11 +27,15 @@
 #include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/cstdint>
 
+#if !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+#  include <sstream>
+#endif // !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+
 CUB_NAMESPACE_BEGIN
 
 namespace detail::merge_sort
 {
-template <typename MaxPolicyT,
+template <typename PolicySelectorT,
           typename KeyInputIteratorT,
           typename ValueInputIteratorT,
           typename KeyIteratorT,
@@ -40,13 +44,17 @@ template <typename MaxPolicyT,
           typename CompareOpT>
 struct DeviceMergeSortKernelSource
 {
+#if _CCCL_HAS_CONCEPTS()
+  static_assert(detail::merge_sort::merge_sort_policy_selector<PolicySelectorT>);
+#endif // _CCCL_HAS_CONCEPTS()
+
   using KeyT   = cub::detail::it_value_t<KeyIteratorT>;
   using ValueT = cub::detail::it_value_t<ValueIteratorT>;
 
   CUB_DEFINE_KERNEL_GETTER(
     MergeSortBlockSortKernel,
     DeviceMergeSortBlockSortKernel<
-      MaxPolicyT,
+      PolicySelectorT,
       KeyInputIteratorT,
       ValueInputIteratorT,
       KeyIteratorT,
@@ -61,7 +69,7 @@ struct DeviceMergeSortKernelSource
 
   CUB_DEFINE_KERNEL_GETTER(
     MergeSortMergeKernel,
-    DeviceMergeSortMergeKernel<MaxPolicyT,
+    DeviceMergeSortMergeKernel<PolicySelectorT,
                                KeyInputIteratorT,
                                ValueInputIteratorT,
                                KeyIteratorT,
@@ -87,6 +95,7 @@ struct DeviceMergeSortKernelSource
  * Policy
  ******************************************************************************/
 
+// TODO(bgruber): deprecate this when we make the tuning API public and remove in CCCL 4.0
 template <typename KeyInputIteratorT,
           typename ValueInputIteratorT,
           typename KeyIteratorT,
@@ -95,7 +104,7 @@ template <typename KeyInputIteratorT,
           typename CompareOpT,
           typename PolicyHub    = detail::merge_sort::policy_hub<KeyIteratorT>,
           typename KernelSource = detail::merge_sort::DeviceMergeSortKernelSource<
-            typename PolicyHub::MaxPolicy,
+            detail::merge_sort::policy_selector_from_hub<PolicyHub>,
             KeyInputIteratorT,
             ValueInputIteratorT,
             KeyIteratorT,
@@ -103,7 +112,6 @@ template <typename KeyInputIteratorT,
             OffsetT,
             CompareOpT>,
           typename KernelLauncherFactory = CUB_DETAIL_DEFAULT_KERNEL_LAUNCHER_FACTORY,
-          typename VSMemHelperT          = detail::merge_sort::VSMemHelper,
           typename KeyT                  = cub::detail::it_value_t<KeyIteratorT>,
           typename ValueT                = cub::detail::it_value_t<ValueIteratorT>>
 struct DispatchMergeSort
@@ -176,9 +184,21 @@ struct DispatchMergeSort
       , launcher_factory(launcher_factory)
   {}
 
+private:
+  template <typename ActivePolicyT>
+  struct policy_getter
+  {
+    _CCCL_API constexpr auto operator()() -> detail::merge_sort::merge_sort_policy
+    {
+      using mp = typename ActivePolicyT::MergeSortPolicy;
+      return {mp::BLOCK_THREADS, mp::ITEMS_PER_THREAD, mp::LOAD_ALGORITHM, mp::LOAD_MODIFIER, mp::STORE_ALGORITHM};
+    }
+  };
+
+public:
   // Invocation
   template <typename ActivePolicyT>
-  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t Invoke(ActivePolicyT policy = {})
+  CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t Invoke([[maybe_unused]] ActivePolicyT policy = {})
   {
     if (num_items == 0)
     {
@@ -189,17 +209,17 @@ struct DispatchMergeSort
       return cudaSuccess;
     }
 
-    auto wrapped_policy  = detail::merge_sort::MakeMergeSortPolicyWrapper(policy);
-    const auto tile_size = VSMemHelperT::template ItemsPerTile<
-      typename ActivePolicyT::MergeSortPolicy,
-      KeyInputIteratorT,
-      ValueInputIteratorT,
-      KeyIteratorT,
-      ValueIteratorT,
-      OffsetT,
-      CompareOpT,
-      KeyT,
-      ValueT>(wrapped_policy.MergeSort());
+    constexpr auto tile_size =
+      detail::merge_sort::merge_sort_vsmem_helper_t<
+        policy_getter<ActivePolicyT>,
+        KeyInputIteratorT,
+        ValueInputIteratorT,
+        KeyIteratorT,
+        ValueIteratorT,
+        OffsetT,
+        CompareOpT,
+        KeyT,
+        ValueT>::policy.items_per_tile();
     const auto num_tiles = ::cuda::ceil_div(num_items, tile_size);
 
     const auto merge_partitions_size         = static_cast<size_t>(1 + num_tiles) * sizeof(OffsetT);
@@ -210,30 +230,21 @@ struct DispatchMergeSort
      * Merge sort supports large types, which can lead to excessive shared memory size requirements. In these cases,
      * merge sort allocates virtual shared memory that resides in global memory.
      */
+
+    using merge_sort_vsmem_t = detail::merge_sort::merge_sort_vsmem_helper_t<
+      policy_getter<ActivePolicyT>,
+      KeyInputIteratorT,
+      ValueInputIteratorT,
+      KeyIteratorT,
+      ValueIteratorT,
+      OffsetT,
+      CompareOpT,
+      KeyT,
+      ValueT>;
     const ::cuda::std::size_t block_sort_smem_size =
-      num_tiles
-      * VSMemHelperT::template BlockSortVSMemPerBlock<
-        typename ActivePolicyT::MergeSortPolicy,
-        KeyInputIteratorT,
-        ValueInputIteratorT,
-        KeyIteratorT,
-        ValueIteratorT,
-        OffsetT,
-        CompareOpT,
-        KeyT,
-        ValueT>(wrapped_policy.MergeSort());
+      num_tiles * detail::vsmem_helper_impl<typename merge_sort_vsmem_t::block_sort_agent_t>::vsmem_per_block;
     const ::cuda::std::size_t merge_smem_size =
-      num_tiles
-      * VSMemHelperT::template MergeVSMemPerBlock<
-        typename ActivePolicyT::MergeSortPolicy,
-        KeyInputIteratorT,
-        ValueInputIteratorT,
-        KeyIteratorT,
-        ValueIteratorT,
-        OffsetT,
-        CompareOpT,
-        KeyT,
-        ValueT>(wrapped_policy.MergeSort());
+      num_tiles * detail::vsmem_helper_impl<typename merge_sort_vsmem_t::merge_agent_t>::vsmem_per_block;
     const ::cuda::std::size_t virtual_shared_memory_size = (::cuda::std::max) (block_sort_smem_size, merge_smem_size);
 
     void* allocations[4]       = {nullptr, nullptr, nullptr, nullptr};
@@ -271,16 +282,17 @@ struct DispatchMergeSort
     auto keys_buffer      = static_cast<KeyT*>(allocations[1]);
     auto items_buffer     = static_cast<ValueT*>(allocations[2]);
 
-    const int block_threads = VSMemHelperT::template BlockThreads<
-      typename ActivePolicyT::MergeSortPolicy,
-      KeyInputIteratorT,
-      ValueInputIteratorT,
-      KeyIteratorT,
-      ValueIteratorT,
-      OffsetT,
-      CompareOpT,
-      KeyT,
-      ValueT>(wrapped_policy.MergeSort());
+    const int block_threads =
+      detail::merge_sort::merge_sort_vsmem_helper_t<
+        policy_getter<ActivePolicyT>,
+        KeyInputIteratorT,
+        ValueInputIteratorT,
+        KeyIteratorT,
+        ValueIteratorT,
+        OffsetT,
+        CompareOpT,
+        KeyT,
+        ValueT>::policy.block_threads;
 
     // Invoke DeviceMergeSortBlockSortKernel
     launcher_factory(static_cast<int>(num_tiles), block_threads, 0, stream, true)
@@ -426,4 +438,218 @@ struct DispatchMergeSort
   }
 };
 
+namespace detail::merge_sort
+{
+template <typename KeyInputIteratorT,
+          typename ValueInputIteratorT,
+          typename KeyIteratorT,
+          typename ValueIteratorT,
+          typename OffsetT,
+          typename CompareOpT,
+          typename PolicySelector        = policy_selector_from_types<KeyIteratorT>,
+          typename KernelSource          = DeviceMergeSortKernelSource<PolicySelector,
+                                                                       KeyInputIteratorT,
+                                                                       ValueInputIteratorT,
+                                                                       KeyIteratorT,
+                                                                       ValueIteratorT,
+                                                                       OffsetT,
+                                                                       CompareOpT>,
+          typename KernelLauncherFactory = CUB_DETAIL_DEFAULT_KERNEL_LAUNCHER_FACTORY,
+          typename KeyT                  = it_value_t<KeyIteratorT>,
+          typename ValueT                = it_value_t<ValueIteratorT>>
+#if _CCCL_HAS_CONCEPTS()
+  requires merge_sort_policy_selector<PolicySelector>
+#endif // _CCCL_HAS_CONCEPTS()
+CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE auto dispatch(
+  void* d_temp_storage,
+  size_t& temp_storage_bytes,
+  KeyInputIteratorT d_input_keys,
+  ValueInputIteratorT d_input_items,
+  KeyIteratorT d_output_keys,
+  ValueIteratorT d_output_items,
+  OffsetT num_items,
+  CompareOpT compare_op,
+  cudaStream_t stream,
+  PolicySelector policy_selector         = {},
+  KernelSource kernel_source             = {},
+  KernelLauncherFactory launcher_factory = {},
+  KeyT*                                  = nullptr /* for CCCL.C */,
+  ValueT*                                = nullptr /* for CCCL.C */) -> cudaError_t
+{
+  [[maybe_unused]] constexpr bool keys_only = ::cuda::std::is_same_v<ValueT, NullType>;
+
+  if (num_items == 0)
+  {
+    if (d_temp_storage == nullptr)
+    {
+      temp_storage_bytes = 0;
+    }
+    return cudaSuccess;
+  }
+
+  ::cuda::arch_id arch_id{};
+  if (const auto error = CubDebug(launcher_factory.PtxArchId(arch_id)))
+  {
+    return error;
+  }
+
+  return detail::dispatch_arch(policy_selector, arch_id, [&](auto policy_getter) -> cudaError_t {
+#ifdef CUB_DEFINE_RUNTIME_POLICIES
+    const merge_sort_policy active_policy = policy_getter();
+#else // CUB_DEFINE_RUNTIME_POLICIES
+    using vsmem_adapted_agents = merge_sort_vsmem_helper_t<
+      decltype(policy_getter),
+      KeyInputIteratorT,
+      ValueInputIteratorT,
+      KeyIteratorT,
+      ValueIteratorT,
+      OffsetT,
+      CompareOpT,
+      KeyT,
+      ValueT>;
+  constexpr merge_sort_policy active_policy = vsmem_adapted_agents::policy;
+#endif // CUB_DEFINE_RUNTIME_POLICIES
+
+#if !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+    NV_IF_TARGET(
+      NV_IS_HOST,
+      (std::stringstream ss; ss << active_policy;
+       _CubLog("Dispatching DeviceMergeSort to arch %d with tuning: %s\n", (int) arch_id, ss.str().c_str());))
+#endif
+
+    const auto tile_size = active_policy.items_per_tile();
+    const auto num_tiles = ::cuda::ceil_div(num_items, tile_size);
+
+    const auto merge_partitions_size         = static_cast<size_t>(1 + num_tiles) * sizeof(OffsetT);
+    const auto temporary_keys_storage_size   = static_cast<size_t>(num_items * kernel_source.KeySize());
+    const auto temporary_values_storage_size = static_cast<size_t>(num_items * kernel_source.ValueSize()) * !keys_only;
+
+#ifdef CUB_DEFINE_RUNTIME_POLICIES
+    const ::cuda::std::size_t block_sort_smem_size = 0;
+    const ::cuda::std::size_t merge_smem_size      = 0;
+#else // CUB_DEFINE_RUNTIME_POLICIES
+    const ::cuda::std::size_t block_sort_smem_size =
+      num_tiles * vsmem_helper_impl<typename vsmem_adapted_agents::block_sort_agent_t>::vsmem_per_block;
+    const ::cuda::std::size_t merge_smem_size = num_tiles * vsmem_helper_impl<typename vsmem_adapted_agents::merge_agent_t>::vsmem_per_block;
+#endif // CUB_DEFINE_RUNTIME_POLICIES
+    const ::cuda::std::size_t virtual_shared_memory_size = (::cuda::std::max) (block_sort_smem_size, merge_smem_size);
+
+    void* allocations[4]       = {nullptr, nullptr, nullptr, nullptr};
+    size_t allocation_sizes[4] = {
+      merge_partitions_size, temporary_keys_storage_size, temporary_values_storage_size, virtual_shared_memory_size};
+
+    if (const auto error =
+          CubDebug(detail::alias_temporaries(d_temp_storage, temp_storage_bytes, allocations, allocation_sizes)))
+    {
+      return error;
+    }
+
+    if (d_temp_storage == nullptr)
+    {
+      return cudaSuccess;
+    }
+
+    const int num_passes = ::cuda::ceil_ilog2(num_tiles);
+    bool ping            = num_passes % 2 == 0;
+
+    auto merge_partitions = static_cast<OffsetT*>(allocations[0]);
+    auto keys_buffer      = static_cast<KeyT*>(allocations[1]);
+    auto items_buffer     = static_cast<ValueT*>(allocations[2]);
+
+    if (const auto error = CubDebug(
+          launcher_factory(static_cast<int>(num_tiles), active_policy.block_threads, 0, stream, true)
+            .doit(kernel_source.MergeSortBlockSortKernel(),
+                  ping,
+                  d_input_keys,
+                  d_input_items,
+                  d_output_keys,
+                  d_output_items,
+                  num_items,
+                  keys_buffer,
+                  items_buffer,
+                  compare_op,
+                  cub::detail::vsmem_t{allocations[3]})))
+    {
+      return error;
+    }
+    if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+    {
+      return error;
+    }
+    if (const auto error = CubDebug(cudaPeekAtLastError()))
+    {
+      return error;
+    }
+
+    const OffsetT num_partitions              = num_tiles + 1;
+    constexpr int threads_per_partition_block = 256;
+    const int partition_grid_size = static_cast<int>(::cuda::ceil_div(num_partitions, threads_per_partition_block));
+
+    if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+    {
+      return error;
+    }
+    if (const auto error = CubDebug(cudaPeekAtLastError()))
+    {
+      return error;
+    }
+
+    for (int pass = 0; pass < num_passes; ++pass, ping = !ping)
+    {
+      const OffsetT target_merged_tiles_number = OffsetT(2) << pass;
+
+      if (const auto error = CubDebug(
+            launcher_factory(partition_grid_size, threads_per_partition_block, 0, stream, true)
+              .doit(kernel_source.MergeSortPartitionKernel(),
+                    ping,
+                    d_output_keys,
+                    keys_buffer,
+                    num_items,
+                    num_partitions,
+                    merge_partitions,
+                    compare_op,
+                    target_merged_tiles_number,
+                    tile_size)))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(cudaPeekAtLastError()))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(
+            launcher_factory(static_cast<int>(num_tiles), active_policy.block_threads, 0, stream, true)
+              .doit(kernel_source.MergeSortMergeKernel(),
+                    ping,
+                    d_output_keys,
+                    d_output_items,
+                    num_items,
+                    keys_buffer,
+                    items_buffer,
+                    compare_op,
+                    merge_partitions,
+                    target_merged_tiles_number,
+                    cub::detail::vsmem_t{allocations[3]})))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(cudaPeekAtLastError()))
+      {
+        return error;
+      }
+    }
+
+    return cudaSuccess;
+  });
+}
+} // namespace detail::merge_sort
+
 CUB_NAMESPACE_END
diff --git a/cub/cub/device/dispatch/kernels/kernel_merge_sort.cuh b/cub/cub/device/dispatch/kernels/kernel_merge_sort.cuh
index a678ee22ff0..35685ff3990 100644
--- a/cub/cub/device/dispatch/kernels/kernel_merge_sort.cuh
+++ b/cub/cub/device/dispatch/kernels/kernel_merge_sort.cuh
@@ -14,138 +14,69 @@
 #endif // no system header
 
 #include <cub/agent/agent_merge_sort.cuh>
+#include <cub/detail/arch_dispatch.cuh>
+#include <cub/device/dispatch/tuning/tuning_merge_sort.cuh>
 #include <cub/iterator/cache_modified_input_iterator.cuh>
-#include <cub/util_policy_wrapper_t.cuh>
 #include <cub/util_vsmem.cuh>
 
 CUB_NAMESPACE_BEGIN
 
 namespace detail::merge_sort
 {
-/**
- * @brief Helper class template that provides two agent template instantiations: one instantiated with the default
- * policy and one with the fallback policy. This helps to avoid having to enlist all the agent's template parameters
- * twice: once for the default agent and once for the fallback agent
- */
-template <typename DefaultPolicyT, typename FallbackPolicyT, template <typename...> class AgentT, typename... AgentParamsT>
-struct dual_policy_agent_helper_t
-{
-  using default_agent_t  = AgentT<DefaultPolicyT, AgentParamsT...>;
-  using fallback_agent_t = AgentT<FallbackPolicyT, AgentParamsT...>;
-
-  static constexpr auto default_size  = sizeof(typename default_agent_t::TempStorage);
-  static constexpr auto fallback_size = sizeof(typename fallback_agent_t::TempStorage);
-};
-
-/**
- * @brief Helper class template for merge sort-specific virtual shared memory handling. The merge sort algorithm in its
- * current implementation relies on the fact that both the sorting as well as the merging kernels use the same tile
- * size. This circumstance needs to be respected when determining whether the fallback policy for large user types is
- * applicable: we must either use the fallback for both or for none of the two agents.
- */
-template <typename DefaultPolicyT,
-          typename KeyInputIteratorT,
-          typename ValueInputIteratorT,
-          typename KeyIteratorT,
-          typename ValueIteratorT,
+//! @brief Helper class template for merge sort-specific virtual shared memory handling. The merge sort algorithm in
+//! its current implementation relies on the fact that both the sorting as well as the merging kernels use the same tile
+//! size. This circumstance needs to be respected when determining whether the fallback policy for large user types is
+//! applicable: we must either use the fallback for both or for none of the two agents.
+template <typename DefaultPolicyGetter,
+          typename KeyInIt,
+          typename ValInIt,
+          typename KeyOutIt,
+          typename ValOutIt,
           typename OffsetT,
           typename CompareOpT,
           typename KeyT,
           typename ValueT>
 class merge_sort_vsmem_helper_t
 {
-private:
-  // Default fallback policy with a smaller tile size
-  using fallback_policy_t = cub::detail::policy_wrapper_t<DefaultPolicyT, 64, 1>;
-
-  // Helper for the `AgentBlockSort` template with one member type alias for the agent template instantiated with the
-  // default policy and one instantiated with the fallback policy
-  using block_sort_helper_t = dual_policy_agent_helper_t<
-    DefaultPolicyT,
-    fallback_policy_t,
-    merge_sort::AgentBlockSort,
-    KeyInputIteratorT,
-    ValueInputIteratorT,
-    KeyIteratorT,
-    ValueIteratorT,
-    OffsetT,
-    CompareOpT,
-    KeyT,
-    ValueT>;
-  using default_block_sort_agent_t  = typename block_sort_helper_t::default_agent_t;
-  using fallback_block_sort_agent_t = typename block_sort_helper_t::fallback_agent_t;
-
-  // Helper for the `AgentMerge` template with one member type alias for the agent template instantiated with the
-  // default policy and one instantiated with the fallback policy
-  using merge_helper_t = dual_policy_agent_helper_t<
-    DefaultPolicyT,
-    fallback_policy_t,
-    merge_sort::AgentMerge,
-    KeyIteratorT,
-    ValueIteratorT,
-    OffsetT,
-    CompareOpT,
-    KeyT,
-    ValueT>;
-  using default_merge_agent_t  = typename merge_helper_t::default_agent_t;
-  using fallback_merge_agent_t = typename merge_helper_t::fallback_agent_t;
+  struct fallback_pol_getter
+  {
+    _CCCL_API _CCCL_FORCEINLINE constexpr auto operator()() const
+    {
+      merge_sort_policy policy = DefaultPolicyGetter{}();
+      policy.block_threads     = 64;
+      policy.items_per_thread  = 1;
+      return policy;
+    }
+  };
+
+  using default_block_sort_agent_t =
+    AgentBlockSort<DefaultPolicyGetter, KeyInIt, ValInIt, KeyOutIt, ValOutIt, OffsetT, CompareOpT, KeyT, ValueT>;
+  using fallback_block_sort_agent_t =
+    AgentBlockSort<fallback_pol_getter, KeyInIt, ValInIt, KeyOutIt, ValOutIt, OffsetT, CompareOpT, KeyT, ValueT>;
+
+  using default_merge_agent_t  = AgentMerge<DefaultPolicyGetter, KeyOutIt, ValOutIt, OffsetT, CompareOpT, KeyT, ValueT>;
+  using fallback_merge_agent_t = AgentMerge<fallback_pol_getter, KeyOutIt, ValOutIt, OffsetT, CompareOpT, KeyT, ValueT>;
 
   // Use fallback if either (a) the default block sort or (b) the block merge agent exceed the maximum shared memory
   // available per block and both (1) the fallback block sort and (2) the fallback merge agent would not exceed the
   // available shared memory
   static constexpr auto max_default_size =
-    (::cuda::std::max) (block_sort_helper_t::default_size, merge_helper_t::default_size);
+    (::cuda::std::max) (sizeof(typename default_block_sort_agent_t::TempStorage),
+                        sizeof(typename default_merge_agent_t::TempStorage));
   static constexpr auto max_fallback_size =
-    (::cuda::std::max) (block_sort_helper_t::fallback_size, merge_helper_t::fallback_size);
+    (::cuda::std::max) (sizeof(typename fallback_block_sort_agent_t::TempStorage),
+                        sizeof(typename fallback_merge_agent_t::TempStorage));
   static constexpr bool uses_fallback_policy =
     (max_default_size > max_smem_per_block) && (max_fallback_size <= max_smem_per_block);
 
 public:
-  using policy_t = ::cuda::std::_If<uses_fallback_policy, fallback_policy_t, DefaultPolicyT>;
+  static constexpr merge_sort_policy policy = uses_fallback_policy ? fallback_pol_getter{}() : DefaultPolicyGetter{}();
   using block_sort_agent_t =
     ::cuda::std::_If<uses_fallback_policy, fallback_block_sort_agent_t, default_block_sort_agent_t>;
   using merge_agent_t = ::cuda::std::_If<uses_fallback_policy, fallback_merge_agent_t, default_merge_agent_t>;
 };
 
-// TODO: this class should be templated on `typename... Ts` to avoid repetition,
-// but due to an issue with NVCC 12.0 we currently template each member function
-// individually instead.
-struct VSMemHelper
-{
-  template <typename ActivePolicyT, typename... Ts>
-  using MergeSortVSMemHelperT = merge_sort_vsmem_helper_t<ActivePolicyT, Ts...>;
-
-  template <typename ActivePolicyT, typename... Ts>
-  _CCCL_HOST_DEVICE static constexpr int BlockThreads(ActivePolicyT /*policy*/)
-  {
-    return MergeSortVSMemHelperT<ActivePolicyT, Ts...>::policy_t::BLOCK_THREADS;
-  }
-
-  template <typename ActivePolicyT, typename... Ts>
-  _CCCL_HOST_DEVICE static constexpr int ItemsPerTile(ActivePolicyT /*policy*/)
-  {
-    return MergeSortVSMemHelperT<ActivePolicyT, Ts...>::policy_t::ITEMS_PER_TILE;
-  }
-
-  template <typename ActivePolicyT, typename... Ts>
-  _CCCL_HOST_DEVICE static constexpr ::cuda::std::size_t BlockSortVSMemPerBlock(ActivePolicyT /*policy*/)
-  {
-    return detail::vsmem_helper_impl<
-      typename MergeSortVSMemHelperT<ActivePolicyT, Ts...>::block_sort_agent_t>::vsmem_per_block;
-  }
-
-  template <typename ActivePolicyT, typename... Ts>
-  _CCCL_HOST_DEVICE static constexpr ::cuda::std::size_t MergeVSMemPerBlock(ActivePolicyT /*policy*/)
-  {
-    return detail::vsmem_helper_impl<
-      typename MergeSortVSMemHelperT<ActivePolicyT, Ts...>::merge_agent_t>::vsmem_per_block;
-  }
-
-  template <typename AgentT>
-  using VSmemHelperT = vsmem_helper_impl<AgentT>;
-};
-
-template <typename ChainedPolicyT,
+template <typename PolicySelectorT,
           typename KeyInputIteratorT,
           typename ValueInputIteratorT,
           typename KeyIteratorT,
@@ -153,19 +84,17 @@ template <typename ChainedPolicyT,
           typename OffsetT,
           typename CompareOpT,
           typename KeyT,
-          typename ValueT,
-          typename VSMemHelperT = VSMemHelper>
+          typename ValueT>
 __launch_bounds__(
-  VSMemHelperT::template MergeSortVSMemHelperT<
-    typename ChainedPolicyT::ActivePolicy::MergeSortPolicy,
-    KeyInputIteratorT,
-    ValueInputIteratorT,
-    KeyIteratorT,
-    ValueIteratorT,
-    OffsetT,
-    CompareOpT,
-    KeyT,
-    ValueT>::policy_t::BLOCK_THREADS)
+  merge_sort_vsmem_helper_t<policy_getter<PolicySelectorT, cuda::arch_id{CUB_PTX_ARCH / 10}>,
+                            KeyInputIteratorT,
+                            ValueInputIteratorT,
+                            KeyIteratorT,
+                            ValueIteratorT,
+                            OffsetT,
+                            CompareOpT,
+                            KeyT,
+                            ValueT>::policy.block_threads)
   CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceMergeSortBlockSortKernel(
     bool ping,
     KeyInputIteratorT keys_in,
@@ -178,8 +107,8 @@ __launch_bounds__(
     CompareOpT compare_op,
     vsmem_t vsmem)
 {
-  using MergeSortHelperT = typename VSMemHelperT::template MergeSortVSMemHelperT<
-    typename ChainedPolicyT::ActivePolicy::MergeSortPolicy,
+  using vsmem_adapted_agents = merge_sort_vsmem_helper_t<
+    policy_getter<PolicySelectorT, cuda::arch_id{CUB_PTX_ARCH / 10}>,
     KeyInputIteratorT,
     ValueInputIteratorT,
     KeyIteratorT,
@@ -189,23 +118,21 @@ __launch_bounds__(
     KeyT,
     ValueT>;
 
-  using ActivePolicyT = typename MergeSortHelperT::policy_t;
-
-  using AgentBlockSortT = typename MergeSortHelperT::block_sort_agent_t;
-
-  using VSmemHelperT = typename VSMemHelperT::template VSmemHelperT<AgentBlockSortT>;
+  static constexpr merge_sort_policy active_policy = vsmem_adapted_agents::policy;
+  using agent_block_sort_t                         = typename vsmem_adapted_agents::block_sort_agent_t;
+  using vsmem_helper_t                             = vsmem_helper_impl<agent_block_sort_t>;
 
   // Static shared memory allocation
-  __shared__ typename VSmemHelperT::static_temp_storage_t static_temp_storage;
+  __shared__ typename vsmem_helper_t::static_temp_storage_t static_temp_storage;
 
   // Get temporary storage
-  typename AgentBlockSortT::TempStorage& temp_storage = VSmemHelperT::get_temp_storage(static_temp_storage, vsmem);
+  typename agent_block_sort_t::TempStorage& temp_storage = vsmem_helper_t::get_temp_storage(static_temp_storage, vsmem);
 
-  AgentBlockSortT agent(
+  agent_block_sort_t agent(
     ping,
     temp_storage,
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(keys_in),
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(items_in),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(keys_in),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(items_in),
     keys_count,
     keys_out,
     items_out,
@@ -216,7 +143,7 @@ __launch_bounds__(
   agent.Process();
 
   // If applicable, hints to discard modified cache lines for vsmem
-  VSmemHelperT::discard_temp_storage(temp_storage);
+  vsmem_helper_t::discard_temp_storage(temp_storage);
 }
 
 template <typename KeyIteratorT, typename OffsetT, typename CompareOpT, typename KeyT>
@@ -249,7 +176,7 @@ CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceMergeSortPartitionKernel(
   }
 }
 
-template <typename ChainedPolicyT,
+template <typename PolicySelectorT,
           typename KeyInputIteratorT,
           typename ValueInputIteratorT,
           typename KeyIteratorT,
@@ -257,19 +184,17 @@ template <typename ChainedPolicyT,
           typename OffsetT,
           typename CompareOpT,
           typename KeyT,
-          typename ValueT,
-          typename VSMemHelperT = VSMemHelper>
+          typename ValueT>
 __launch_bounds__(
-  VSMemHelperT::template MergeSortVSMemHelperT<
-    typename ChainedPolicyT::ActivePolicy::MergeSortPolicy,
-    KeyInputIteratorT,
-    ValueInputIteratorT,
-    KeyIteratorT,
-    ValueIteratorT,
-    OffsetT,
-    CompareOpT,
-    KeyT,
-    ValueT>::policy_t::BLOCK_THREADS)
+  merge_sort_vsmem_helper_t<policy_getter<PolicySelectorT, cuda::arch_id{CUB_PTX_ARCH / 10}>,
+                            KeyInputIteratorT,
+                            ValueInputIteratorT,
+                            KeyIteratorT,
+                            ValueIteratorT,
+                            OffsetT,
+                            CompareOpT,
+                            KeyT,
+                            ValueT>::policy.block_threads)
   CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceMergeSortMergeKernel(
     bool ping,
     KeyIteratorT keys_ping,
@@ -282,8 +207,8 @@ __launch_bounds__(
     OffsetT target_merged_tiles_number,
     vsmem_t vsmem)
 {
-  using MergeSortHelperT = typename VSMemHelperT::template MergeSortVSMemHelperT<
-    typename ChainedPolicyT::ActivePolicy::MergeSortPolicy,
+  using vsmem_adapted_agents = merge_sort_vsmem_helper_t<
+    policy_getter<PolicySelectorT, cuda::arch_id{CUB_PTX_ARCH / 10}>,
     KeyInputIteratorT,
     ValueInputIteratorT,
     KeyIteratorT,
@@ -293,25 +218,23 @@ __launch_bounds__(
     KeyT,
     ValueT>;
 
-  using ActivePolicyT = typename MergeSortHelperT::policy_t;
-
-  using AgentMergeT = typename MergeSortHelperT::merge_agent_t;
-
-  using VSmemHelperT = typename VSMemHelperT::template VSmemHelperT<AgentMergeT>;
+  static constexpr merge_sort_policy active_policy = vsmem_adapted_agents::policy;
+  using agent_merge_t                              = typename vsmem_adapted_agents::merge_agent_t;
+  using vsmem_helper_t                             = vsmem_helper_impl<agent_merge_t>;
 
   // Static shared memory allocation
-  __shared__ typename VSmemHelperT::static_temp_storage_t static_temp_storage;
+  __shared__ typename vsmem_helper_t::static_temp_storage_t static_temp_storage;
 
   // Get temporary storage
-  typename AgentMergeT::TempStorage& temp_storage = VSmemHelperT::get_temp_storage(static_temp_storage, vsmem);
+  typename agent_merge_t::TempStorage& temp_storage = vsmem_helper_t::get_temp_storage(static_temp_storage, vsmem);
 
-  AgentMergeT agent(
+  agent_merge_t agent(
     ping,
     temp_storage,
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(keys_ping),
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(items_ping),
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(keys_pong),
-    try_make_cache_modified_iterator<ActivePolicyT::LOAD_MODIFIER>(items_pong),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(keys_ping),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(items_ping),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(keys_pong),
+    try_make_cache_modified_iterator<active_policy.load_modifier>(items_pong),
     keys_count,
     keys_pong,
     items_pong,
@@ -324,7 +247,7 @@ __launch_bounds__(
   agent.Process();
 
   // If applicable, hints to discard modified cache lines for vsmem
-  VSmemHelperT::discard_temp_storage(temp_storage);
+  vsmem_helper_t::discard_temp_storage(temp_storage);
 }
 } // namespace detail::merge_sort
 
diff --git a/cub/cub/device/dispatch/tuning/tuning_merge_sort.cuh b/cub/cub/device/dispatch/tuning/tuning_merge_sort.cuh
index 80db5aae12c..4345e76ed43 100644
--- a/cub/cub/device/dispatch/tuning/tuning_merge_sort.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_merge_sort.cuh
@@ -13,46 +13,40 @@
 #  pragma system_header
 #endif // no system header
 
-#include <cub/agent/agent_merge_sort.cuh>
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_store.cuh>
+#include <cub/device/dispatch/tuning/common.cuh>
+#include <cub/thread/thread_load.cuh>
 #include <cub/util_device.cuh>
+#include <cub/util_math.cuh>
 
-CUB_NAMESPACE_BEGIN
+#include <cuda/__device/arch_id.h>
 
-namespace detail::merge_sort
-{
-template <typename PolicyT, typename = void>
-struct MergeSortPolicyWrapper : PolicyT
-{
-  _CCCL_HOST_DEVICE MergeSortPolicyWrapper(PolicyT base)
-      : PolicyT(base)
-  {}
-};
+#if !_CCCL_COMPILER(NVRTC)
+#  include <ostream>
+#endif
 
-template <typename StaticPolicyT>
-struct MergeSortPolicyWrapper<StaticPolicyT, ::cuda::std::void_t<decltype(StaticPolicyT::MergeSortPolicy::LOAD_MODIFIER)>>
-    : StaticPolicyT
-{
-  _CCCL_HOST_DEVICE MergeSortPolicyWrapper(StaticPolicyT base)
-      : StaticPolicyT(base)
-  {}
+CUB_NAMESPACE_BEGIN
 
-  CUB_DEFINE_SUB_POLICY_GETTER(MergeSort);
+template <int BlockThreads,
+          int ItemsPerThread                      = 1,
+          cub::BlockLoadAlgorithm LoadAlgorithm   = cub::BLOCK_LOAD_DIRECT,
+          cub::CacheLoadModifier LoadModifier     = cub::LOAD_LDG,
+          cub::BlockStoreAlgorithm StoreAlgorithm = cub::BLOCK_STORE_DIRECT>
+struct AgentMergeSortPolicy
+{
+  static constexpr int BLOCK_THREADS    = BlockThreads;
+  static constexpr int ITEMS_PER_THREAD = ItemsPerThread;
+  static constexpr int ITEMS_PER_TILE   = BLOCK_THREADS * ITEMS_PER_THREAD;
 
-#if defined(CUB_ENABLE_POLICY_PTX_JSON)
-  _CCCL_DEVICE static constexpr auto EncodedPolicy()
-  {
-    using namespace ptx_json;
-    return object<key<"MergeSortPolicy">() = MergeSort().EncodedPolicy()>();
-  }
-#endif
+  static constexpr cub::BlockLoadAlgorithm LOAD_ALGORITHM   = LoadAlgorithm;
+  static constexpr cub::CacheLoadModifier LOAD_MODIFIER     = LoadModifier;
+  static constexpr cub::BlockStoreAlgorithm STORE_ALGORITHM = StoreAlgorithm;
 };
 
-template <typename PolicyT>
-_CCCL_HOST_DEVICE MergeSortPolicyWrapper<PolicyT> MakeMergeSortPolicyWrapper(PolicyT policy)
+namespace detail::merge_sort
 {
-  return MergeSortPolicyWrapper<PolicyT>{policy};
-}
-
+// TODO(bgruber): drop in CCCL 4.0 when we remove all public CUB dispatchers
 template <typename KeyIteratorT>
 struct policy_hub
 {
@@ -95,6 +89,89 @@ struct policy_hub
 
   using MaxPolicy = Policy600;
 };
+
+struct merge_sort_policy
+{
+  int block_threads;
+  int items_per_thread;
+  BlockLoadAlgorithm load_algorithm;
+  CacheLoadModifier load_modifier;
+  BlockStoreAlgorithm store_algorithm;
+
+  [[nodiscard]] _CCCL_API constexpr int items_per_tile() const
+  {
+    return block_threads * items_per_thread;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool operator==(const merge_sort_policy& lhs, const merge_sort_policy& rhs)
+  {
+    return lhs.block_threads == rhs.block_threads && lhs.items_per_thread == rhs.items_per_thread
+        && lhs.load_algorithm == rhs.load_algorithm && lhs.load_modifier == rhs.load_modifier
+        && lhs.store_algorithm == rhs.store_algorithm;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool operator!=(const merge_sort_policy& lhs, const merge_sort_policy& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+#if !_CCCL_COMPILER(NVRTC)
+  friend ::std::ostream& operator<<(::std::ostream& os, const merge_sort_policy& p)
+  {
+    return os << "merge_sort_policy { .block_threads = " << p.block_threads
+              << ", .items_per_thread = " << p.items_per_thread << ", .load_algorithm = " << p.load_algorithm
+              << ", .load_modifier = " << p.load_modifier << ", .store_algorithm = " << p.store_algorithm << " }";
+  }
+#endif
+};
+
+#if _CCCL_HAS_CONCEPTS()
+template <typename T>
+concept merge_sort_policy_selector = policy_selector<T, merge_sort_policy>;
+#endif // _CCCL_HAS_CONCEPTS()
+
+struct policy_selector
+{
+  int key_size;
+
+  [[nodiscard]] _CCCL_API constexpr auto operator()(::cuda::arch_id /*arch*/) const -> merge_sort_policy
+  {
+    // from SM60
+    return merge_sort_policy{
+      256,
+      detail::nominal_4B_items_to_items(17, key_size),
+      BLOCK_LOAD_WARP_TRANSPOSE,
+      LOAD_DEFAULT,
+      BLOCK_STORE_WARP_TRANSPOSE};
+  }
+};
+
+#if _CCCL_HAS_CONCEPTS()
+static_assert(merge_sort_policy_selector<policy_selector>);
+#endif // _CCCL_HAS_CONCEPTS()
+
+template <typename KeyIteratorT>
+struct policy_selector_from_types
+{
+  [[nodiscard]] _CCCL_API constexpr auto operator()(::cuda::arch_id arch) const -> merge_sort_policy
+  {
+    return policy_selector{int{sizeof(it_value_t<KeyIteratorT>)}}(arch);
+  }
+};
+
+// TODO(bgruber): remove in CCCL 4.0
+template <typename PolicyHub>
+struct policy_selector_from_hub
+{
+  // this is only called in device code, so we can ignore the arch parameter
+  _CCCL_HOST_DEVICE constexpr auto operator()(::cuda::arch_id /*arch*/) const -> merge_sort_policy
+  {
+    using ap = typename PolicyHub::MaxPolicy::ActivePolicy;
+    using mp = typename ap::MergeSortPolicy;
+    return merge_sort_policy{
+      mp::BLOCK_THREADS, mp::ITEMS_PER_THREAD, mp::LOAD_ALGORITHM, mp::LOAD_MODIFIER, mp::STORE_ALGORITHM};
+  }
+};
 } // namespace detail::merge_sort
 
 CUB_NAMESPACE_END
