diff --git a/flash_attn/cute/logging.py b/flash_attn/cute/fa_logging.py
similarity index 100%
rename from flash_attn/cute/logging.py
rename to flash_attn/cute/fa_logging.py
diff --git a/flash_attn/cute/interface.py b/flash_attn/cute/interface.py
index 0a15aa65b93..4cc7e103f46 100644
--- a/flash_attn/cute/interface.py
+++ b/flash_attn/cute/interface.py
@@ -46,7 +46,7 @@
 
 
 from flash_attn.cute import utils
-from flash_attn.cute import logging as fa_logging
+from flash_attn.cute import fa_logging
 from flash_attn.cute.cute_dsl_utils import (
     to_cute_tensor, to_cute_aux_tensor, get_aux_tensor_metadata, get_broadcast_dims,
 )
