diff --git a/libcudacxx/include/cuda/std/__internal/thread_api.h b/libcudacxx/include/cuda/std/__internal/thread_api.h
index 25199ba6b06..1731bb0dfc2 100644
--- a/libcudacxx/include/cuda/std/__internal/thread_api.h
+++ b/libcudacxx/include/cuda/std/__internal/thread_api.h
@@ -28,7 +28,7 @@
 #endif // _CCCL_HAS_THREAD_API_EXTERNAL
 
 #ifndef _CCCL_HAS_THREAD_API_CUDA
-#  if ((_CCCL_DEVICE_COMPILATION() && !_CCCL_CUDA_COMPILER(NVHPC)) || defined(__EMSCRIPTEN__))
+#  if ((_CCCL_DEVICE_COMPILATION() && !_CCCL_CUDA_COMPILER(NVHPC)) || defined(__EMSCRIPTEN__) || _CCCL_HOSTJIT())
 #    define _CCCL_HAS_THREAD_API_CUDA
 #  endif // ((_CCCL_DEVICE_COMPILATION() && !_CCCL_CUDA_COMPILER(NVHPC)) || defined(__EMSCRIPTEN__))
 #endif // _CCCL_HAS_THREAD_API_CUDA
