diff --git a/csrc/trtllm_fused_moe_routing_renormalize.cu b/csrc/trtllm_fused_moe_routing_renormalize.cu
index 1a4823d481..153d412387 100644
--- a/csrc/trtllm_fused_moe_routing_renormalize.cu
+++ b/csrc/trtllm_fused_moe_routing_renormalize.cu
@@ -143,6 +143,14 @@ __global__ void __launch_bounds__(KernelParams::MaxNumExperts)
         }
       }
     }  // end if (validToken)
+  } else if (params.mPtrTopKPacked != nullptr) {
+    if (validToken) {
+      if (laneIdx < params.mTopK) {
+        int offset =
+            warpIdx * MaxNumExperts + params.mPtrTopKPacked[warpIdx * params.mTopK + laneIdx].idx;
+        smemKIdx[offset] = static_cast<int8_t>(laneIdx);
+      }
+    }
   }
   __syncthreads();
 
