diff --git a/cudax/include/cuda/experimental/__hierarchy/fwd.cuh b/cudax/include/cuda/experimental/__hierarchy/fwd.cuh
new file mode 100644
index 00000000000..7a2ab917e98
--- /dev/null
+++ b/cudax/include/cuda/experimental/__hierarchy/fwd.cuh
@@ -0,0 +1,56 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_EXPERIMENTAL___HIERARCHY_FWD_CUH
+#define _CUDA_EXPERIMENTAL___HIERARCHY_FWD_CUH
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__cccl/prologue.h>
+
+namespace cuda::experimental
+{
+// hierarchy group kinds
+
+class __this_hierarchy_group_kind
+{};
+
+// hierarchy group base
+
+template <class _Level, class _Hierarchy, class _Kind>
+class __hierarchy_group_base;
+template <class _Level, class _Hierarchy>
+using __this_hierarchy_group_base = __hierarchy_group_base<_Level, _Hierarchy, __this_hierarchy_group_kind>;
+
+// hierarchy groups
+
+template <class _Hierarchy, class _Kind>
+class thread_group;
+template <class _Hierarchy, class _Kind>
+class warp_group;
+template <class _Hierarchy, class _Kind>
+class block_group;
+template <class _Hierarchy, class _Kind>
+class cluster_group;
+template <class _Hierarchy, class _Kind>
+class grid_group;
+} // namespace cuda::experimental
+
+#include <cuda/std/__cccl/epilogue.h>
+
+#endif // _CUDA_EXPERIMENTAL___HIERARCHY_FWD_CUH
diff --git a/cudax/include/cuda/experimental/__hierarchy/grid_sync.cuh b/cudax/include/cuda/experimental/__hierarchy/grid_sync.cuh
new file mode 100644
index 00000000000..49a112b1ee1
--- /dev/null
+++ b/cudax/include/cuda/experimental/__hierarchy/grid_sync.cuh
@@ -0,0 +1,107 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_EXPERIMENTAL___HIERARCHY_GRID_SYNC_CUH
+#define _CUDA_EXPERIMENTAL___HIERARCHY_GRID_SYNC_CUH
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__fwd/hierarchy.h>
+#include <cuda/std/__type_traits/always_false.h>
+#include <cuda/std/__type_traits/is_same.h>
+
+#include <cuda/std/__cccl/prologue.h>
+
+namespace cuda::experimental
+{
+// Synchronizing whole grid requires driver support and the kernel must be launched using the cooperative launch API.
+// This part is extracted from grid synchronization implementation in cooperative groups.
+namespace __cg_imported
+{
+struct __grid_workspace
+{
+  unsigned int __size_;
+  unsigned int __barrier_;
+};
+
+[[nodiscard]] _CCCL_DEVICE_API inline __grid_workspace* __grid_workspace_ptr() noexcept
+{
+  unsigned long long __ret;
+  asm("mov.b64 %0, {%%envreg2, %%envreg1};" : "=l"(__ret));
+  _CCCL_ASSERT(__ret != 0, "Synchronizing grid requires the kernel to be launched using the cooperative launch.");
+  return reinterpret_cast<__grid_workspace*>(__ret);
+}
+
+[[nodiscard]] _CCCL_DEVICE_API inline bool __grid_barrier_has_flipped(unsigned __old, unsigned __curr) noexcept
+{
+  return (((__old ^ __curr) & 0x80000000) != 0);
+}
+
+_CCCL_DEVICE_API inline void __grid_sync()
+{
+  const auto __bar_ptr = &::cuda::experimental::__cg_imported::__grid_workspace_ptr()->__barrier_;
+
+  // Synchronize the block before synchronizing with the other blocks.
+  ::__barrier_sync(0);
+
+  // Synchronize with other blocks using the thread 0 in block.
+  if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0)
+  {
+    const unsigned __expected = gridDim.x * gridDim.y * gridDim.z;
+    unsigned __nblocks        = 1;
+
+    if (blockIdx.x == 0 && blockIdx.y == 0 && blockIdx.z == 0)
+    {
+      __nblocks = 0x80000000 - (__expected - 1);
+    }
+
+    unsigned __old_barrier_value;
+
+    NV_IF_ELSE_TARGET(
+      NV_PROVIDES_SM_70,
+      ({
+        asm volatile("atom.add.release.gpu.u32 %0,[%1],%2;"
+                     : "=r"(__old_barrier_value)
+                     : "l"(__bar_ptr), "r"(__nblocks)
+                     : "memory");
+        unsigned __curr_barrier_value;
+        do
+        {
+          asm volatile("ld.acquire.gpu.u32 %0,[%1];" : "=r"(__curr_barrier_value) : "l"(__bar_ptr) : "memory");
+        } while (
+          !::cuda::experimental::__cg_imported::__grid_barrier_has_flipped(__old_barrier_value, __curr_barrier_value));
+      }),
+      ({
+        ::__threadfence();
+        __old_barrier_value = ::atomicAdd(__bar_ptr, __nblocks);
+        while (!::cuda::experimental::__cg_imported::__grid_barrier_has_flipped(__old_barrier_value, *__bar_ptr))
+        {
+        }
+        ::__threadfence();
+      }))
+  }
+
+  // Wait for the thread 0 to finish the inter block synchronization.
+  ::__barrier_sync(0);
+}
+} // namespace __cg_imported
+} // namespace cuda::experimental
+
+#include <cuda/std/__cccl/epilogue.h>
+
+#endif // _CUDA_EXPERIMENTAL___HIERARCHY_GRID_SYNC_CUH
diff --git a/cudax/include/cuda/experimental/__hierarchy/group.cuh b/cudax/include/cuda/experimental/__hierarchy/group.cuh
new file mode 100644
index 00000000000..63283086501
--- /dev/null
+++ b/cudax/include/cuda/experimental/__hierarchy/group.cuh
@@ -0,0 +1,264 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_EXPERIMENTAL___HIERARCHY_GROUP_CUH
+#define _CUDA_EXPERIMENTAL___HIERARCHY_GROUP_CUH
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/hierarchy>
+#include <cuda/std/__concepts/concept_macros.h>
+#include <cuda/std/__type_traits/is_integer.h>
+#include <cuda/std/__type_traits/is_same.h>
+
+#include <cuda/experimental/__hierarchy/fwd.cuh>
+#include <cuda/experimental/__hierarchy/grid_sync.cuh>
+
+#include <cuda/std/__cccl/prologue.h>
+
+namespace cuda::experimental
+{
+template <class _HierarchyLike>
+using __hierarchy_type_of =
+  ::cuda::std::remove_cvref_t<decltype(::cuda::__unpack_hierarchy_if_needed(::cuda::std::declval<_HierarchyLike>()))>;
+
+// todo: use __hier_ in queries
+template <class _Level, class _Hierarchy>
+class __hierarchy_group_base<_Level, _Hierarchy, __this_hierarchy_group_kind>
+{
+  static_assert(__is_hierarchy_level_v<_Level>);
+  static_assert(__is_hierarchy_v<_Hierarchy>);
+
+  const _Hierarchy& __hier_;
+
+public:
+  using hierarchy_type = _Hierarchy;
+
+  _CCCL_TEMPLATE(class _HierarchyLike)
+  _CCCL_REQUIRES(::cuda::std::is_same_v<_Hierarchy, __hierarchy_type_of<_HierarchyLike>>)
+  _CCCL_DEVICE_API __hierarchy_group_base(const _HierarchyLike& __hier_like) noexcept
+      : __hier_{::cuda::__unpack_hierarchy_if_needed(__hier_like)}
+  {}
+
+  [[nodiscard]] _CCCL_API const _Hierarchy& hierarchy() const noexcept
+  {
+    return __hier_;
+  }
+
+  _CCCL_TEMPLATE(class _Tp, class _InLevel, class _Level2 = _Level)
+  _CCCL_REQUIRES(::cuda::std::__cccl_is_integer_v<_Tp> _CCCL_AND __is_hierarchy_level_v<_InLevel> _CCCL_AND(
+    !::cuda::std::is_same_v<_Level2, grid_level>))
+  [[nodiscard]] _CCCL_API constexpr _Tp count_as(const _InLevel& __in_level) const noexcept
+  {
+    return _Level{}.template count_as<_Tp>(__in_level);
+  }
+
+  _CCCL_TEMPLATE(class _InLevel, class _Level2 = _Level)
+  _CCCL_REQUIRES(__is_hierarchy_level_v<_InLevel> _CCCL_AND(!::cuda::std::is_same_v<_Level2, grid_level>))
+  [[nodiscard]] _CCCL_API constexpr auto count(const _InLevel& __in_level) const noexcept
+  {
+    return _Level{}.count(__in_level);
+  }
+
+#if _CCCL_CUDA_COMPILATION()
+  _CCCL_TEMPLATE(class _Tp, class _InLevel, class _Level2 = _Level)
+  _CCCL_REQUIRES(::cuda::std::__cccl_is_integer_v<_Tp> _CCCL_AND __is_hierarchy_level_v<_InLevel> _CCCL_AND(
+    !::cuda::std::is_same_v<_Level2, grid_level>))
+  [[nodiscard]] _CCCL_DEVICE_API _Tp rank_as(const _InLevel& __in_level) const noexcept
+  {
+    return _Level{}.template rank_as<_Tp>(__in_level);
+  }
+
+  _CCCL_TEMPLATE(class _InLevel, class _Level2 = _Level)
+  _CCCL_REQUIRES(__is_hierarchy_level_v<_InLevel> _CCCL_AND(!::cuda::std::is_same_v<_Level2, grid_level>))
+  [[nodiscard]] _CCCL_DEVICE_API auto rank(const _InLevel& __in_level) const noexcept
+  {
+    return _Level{}.rank(__in_level);
+  }
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+template <class _Hierarchy>
+class thread_group<_Hierarchy, __this_hierarchy_group_kind> : __this_hierarchy_group_base<thread_level, _Hierarchy>
+{
+  using __base_type = __this_hierarchy_group_base<thread_level, _Hierarchy>;
+
+public:
+  using __base_type::__base_type;
+  using __base_type::count;
+  using __base_type::count_as;
+  using __base_type::hierarchy;
+#if _CCCL_CUDA_COMPILATION()
+  using __base_type::rank;
+  using __base_type::rank_as;
+
+  _CCCL_DEVICE_API void sync() noexcept {}
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+_CCCL_TEMPLATE(class _Hierarchy)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_Hierarchy>)
+_CCCL_HOST_DEVICE thread_group(const _Hierarchy&)
+  -> thread_group<__hierarchy_type_of<_Hierarchy>, __this_hierarchy_group_kind>;
+
+template <class _Hierarchy>
+class warp_group<_Hierarchy, __this_hierarchy_group_kind> : __this_hierarchy_group_base<warp_level, _Hierarchy>
+{
+  using __base_type = __this_hierarchy_group_base<warp_level, _Hierarchy>;
+
+public:
+  using __base_type::__base_type;
+  using __base_type::count;
+  using __base_type::count_as;
+  using __base_type::hierarchy;
+#if _CCCL_CUDA_COMPILATION()
+  using __base_type::rank;
+  using __base_type::rank_as;
+
+  _CCCL_DEVICE_API void sync() noexcept
+  {
+    ::__syncwarp();
+  }
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+_CCCL_TEMPLATE(class _Hierarchy)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_Hierarchy>)
+_CCCL_HOST_DEVICE warp_group(const _Hierarchy&)
+  -> warp_group<__hierarchy_type_of<_Hierarchy>, __this_hierarchy_group_kind>;
+
+template <class _Hierarchy>
+class block_group<_Hierarchy, __this_hierarchy_group_kind> : __this_hierarchy_group_base<block_level, _Hierarchy>
+{
+  using __base_type = __this_hierarchy_group_base<block_level, _Hierarchy>;
+
+public:
+  using __base_type::__base_type;
+  using __base_type::count;
+  using __base_type::count_as;
+  using __base_type::hierarchy;
+
+#if _CCCL_CUDA_COMPILATION()
+  using __base_type::rank;
+  using __base_type::rank_as;
+
+  _CCCL_DEVICE_API void sync() noexcept
+  {
+    ::__syncthreads();
+  }
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+_CCCL_TEMPLATE(class _Hierarchy)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_Hierarchy>)
+_CCCL_HOST_DEVICE block_group(const _Hierarchy&)
+  -> block_group<__hierarchy_type_of<_Hierarchy>, __this_hierarchy_group_kind>;
+
+template <class _Hierarchy>
+class cluster_group<_Hierarchy, __this_hierarchy_group_kind> : __this_hierarchy_group_base<cluster_level, _Hierarchy>
+{
+  using __base_type = __this_hierarchy_group_base<cluster_level, _Hierarchy>;
+
+public:
+  using __base_type::__base_type;
+  using __base_type::count;
+  using __base_type::count_as;
+  using __base_type::hierarchy;
+
+#if _CCCL_CUDA_COMPILATION()
+  using __base_type::rank;
+  using __base_type::rank_as;
+
+  _CCCL_DEVICE_API void sync() noexcept
+  {
+    NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,
+                      ({
+                        ::__cluster_barrier_arrive();
+                        ::__cluster_barrier_wait();
+                      }),
+                      ({ ::__syncthreads(); }))
+  }
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+_CCCL_TEMPLATE(class _Hierarchy)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_Hierarchy>)
+_CCCL_HOST_DEVICE cluster_group(const _Hierarchy&)
+  -> cluster_group<__hierarchy_type_of<_Hierarchy>, __this_hierarchy_group_kind>;
+
+template <class _Hierarchy>
+class grid_group<_Hierarchy, __this_hierarchy_group_kind> : __this_hierarchy_group_base<grid_level, _Hierarchy>
+{
+  using __base_type = __this_hierarchy_group_base<grid_level, _Hierarchy>;
+
+public:
+  using __base_type::__base_type;
+  using __base_type::hierarchy;
+
+#if _CCCL_CUDA_COMPILATION()
+  _CCCL_DEVICE_API void sync() noexcept
+  {
+    ::cuda::experimental::__cg_imported::__grid_sync();
+  }
+#endif // _CCCL_CUDA_COMPILATION()
+};
+
+_CCCL_TEMPLATE(class _Hierarchy)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_Hierarchy>)
+_CCCL_HOST_DEVICE grid_group(const _Hierarchy&)
+  -> grid_group<__hierarchy_type_of<_Hierarchy>, __this_hierarchy_group_kind>;
+
+_CCCL_TEMPLATE(class _HierarchyLike)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_HierarchyLike>)
+[[nodiscard]] _CCCL_DEVICE_API auto this_thread(const _HierarchyLike& __hier_like) noexcept
+{
+  return thread_group{__hier_like};
+}
+
+_CCCL_TEMPLATE(class _HierarchyLike)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_HierarchyLike>)
+[[nodiscard]] _CCCL_DEVICE_API auto this_warp(const _HierarchyLike& __hier_like) noexcept
+{
+  return warp_group{__hier_like};
+}
+
+_CCCL_TEMPLATE(class _HierarchyLike)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_HierarchyLike>)
+[[nodiscard]] _CCCL_DEVICE_API auto this_block(const _HierarchyLike& __hier_like) noexcept
+{
+  return block_group{__hier_like};
+}
+
+_CCCL_TEMPLATE(class _HierarchyLike)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_HierarchyLike>)
+[[nodiscard]] _CCCL_DEVICE_API auto this_cluster(const _HierarchyLike& __hier_like) noexcept
+{
+  return cluster_group{__hier_like};
+}
+
+_CCCL_TEMPLATE(class _HierarchyLike)
+_CCCL_REQUIRES(__is_or_has_hierarchy_member_v<_HierarchyLike>)
+[[nodiscard]] _CCCL_DEVICE_API auto this_grid(const _HierarchyLike& __hier_like) noexcept
+{
+  return grid_group{__hier_like};
+}
+} // namespace cuda::experimental
+
+#include <cuda/std/__cccl/epilogue.h>
+
+#endif // _CUDA_EXPERIMENTAL___HIERARCHY_GROUP_CUH
diff --git a/cudax/include/cuda/experimental/hierarchy.cuh b/cudax/include/cuda/experimental/hierarchy.cuh
new file mode 100644
index 00000000000..474d251b1c0
--- /dev/null
+++ b/cudax/include/cuda/experimental/hierarchy.cuh
@@ -0,0 +1,28 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_EXPERIMENTAL_HIERARCHY
+#define _CUDA_EXPERIMENTAL_HIERARCHY
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/experimental/__hierarchy/fwd.cuh>
+#include <cuda/experimental/__hierarchy/grid_sync.cuh>
+#include <cuda/experimental/__hierarchy/group.cuh>
+
+#endif // _CUDA_EXPERIMENTAL_HIERARCHY
diff --git a/cudax/test/CMakeLists.txt b/cudax/test/CMakeLists.txt
index 9bfdd20031e..4626d575cf9 100644
--- a/cudax/test/CMakeLists.txt
+++ b/cudax/test/CMakeLists.txt
@@ -110,6 +110,10 @@ cudax_add_catch2_test(test_target algorithm
     algorithm/copy.cu
 )
 
+cudax_add_catch2_test(test_target hierarchy_groups
+    hierarchy/group.cu
+)
+
 if (cudax_ENABLE_CUFILE)
   cudax_add_catch2_test(test_target cufile.driver_attributes
       cufile/driver_attributes.cu
diff --git a/cudax/test/hierarchy/group.cu b/cudax/test/hierarchy/group.cu
new file mode 100644
index 00000000000..77f354fddd7
--- /dev/null
+++ b/cudax/test/hierarchy/group.cu
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#include <cub/block/block_reduce.cuh>
+#include <cub/thread/thread_reduce.cuh>
+#include <cub/warp/warp_reduce.cuh>
+
+#include <cuda/devices>
+#include <cuda/hierarchy>
+#include <cuda/launch>
+#include <cuda/stream>
+
+#include <cuda/experimental/hierarchy.cuh>
+
+#include "testing.cuh"
+
+template <class... GArgs, class T, cuda::std::size_t N>
+[[nodiscard]] __device__ T sum(cudax::thread_group<GArgs...> group, T (&array)[N])
+{
+  return cub::ThreadReduce(array, cuda::std::plus<T>{});
+}
+
+template <class... GArgs, class T, cuda::std::size_t N>
+[[nodiscard]] __device__ T sum(cudax::warp_group<GArgs...> group, T (&array)[N])
+{
+  using WarpReduce = cub::WarpReduce<T>;
+
+  __shared__ typename WarpReduce::TempStorage temp_storage;
+
+  const auto partial = cub::ThreadReduce(array, cuda::std::plus<T>{});
+  return WarpReduce{temp_storage}.Sum(partial);
+}
+
+template <class... GArgs, class T, cuda::std::size_t N>
+[[nodiscard]] __device__ T sum(cudax::block_group<GArgs...> group, T (&array)[N])
+{
+  // todo: Replace 32 with value from group.
+  using BlockReduce = cub::BlockReduce<T, 32>;
+
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  return BlockReduce{temp_storage}.Sum(array);
+}
+
+struct TestKernel
+{
+  template <class Config>
+  __device__ void operator()(const Config& config)
+  {
+    {
+      unsigned array[]{1, 2, 3};
+
+      auto this_thread = cudax::this_thread(config);
+      CUDAX_REQUIRE(this_thread.count(cuda::grid) == cuda::gpu_thread.count(cuda::grid));
+      CUDAX_REQUIRE(this_thread.rank(cuda::grid) == cuda::gpu_thread.rank(cuda::grid));
+      this_thread.sync();
+
+      const auto result = sum(this_thread, array);
+      CUDAX_REQUIRE(result == 6);
+    }
+    {
+      unsigned array[]{1, 2, 3};
+
+      auto this_warp = cudax::this_warp(config);
+      CUDAX_REQUIRE(this_warp.count(cuda::grid) == cuda::warp.count(cuda::grid));
+      CUDAX_REQUIRE(this_warp.rank(cuda::grid) == cuda::warp.rank(cuda::grid));
+      this_warp.sync();
+
+      const auto result = sum(this_warp, array);
+      if (cuda::gpu_thread.rank(cuda::warp) == 0)
+      {
+        CUDAX_REQUIRE(result == 6 * cuda::gpu_thread.count(cuda::warp));
+      }
+    }
+    {
+      unsigned array[]{1, 2, 3};
+
+      auto this_block = cudax::this_block(config);
+      CUDAX_REQUIRE(this_block.count(cuda::grid) == cuda::block.count(cuda::grid));
+      CUDAX_REQUIRE(this_block.rank(cuda::grid) == cuda::block.rank(cuda::grid));
+      this_block.sync();
+
+      const auto result = sum(this_block, array);
+      if (cuda::gpu_thread.rank(cuda::block) == 0)
+      {
+        CUDAX_REQUIRE(result == 6 * cuda::gpu_thread.count(cuda::block));
+      }
+    }
+    {
+      auto this_cluster = cudax::this_cluster(config);
+      CUDAX_REQUIRE(this_cluster.count(cuda::grid) == cuda::cluster.count(cuda::grid));
+      CUDAX_REQUIRE(this_cluster.rank(cuda::grid) == cuda::cluster.rank(cuda::grid));
+      this_cluster.sync();
+    }
+    {
+      auto this_grid = cudax::this_grid(config);
+      this_grid.sync();
+    }
+  }
+};
+
+C2H_TEST("Hierarchy groups", "[hierarchy]")
+{
+  const auto device = cuda::devices[0];
+
+  const cuda::stream stream{device};
+
+  const auto config = cuda::make_config(cuda::grid_dims<2>(), cuda::block_dims<32>(), cuda::cooperative_launch{});
+
+  cuda::launch(stream, config, TestKernel{});
+
+  stream.sync();
+}
