diff --git a/flashinfer/autotuner.py b/flashinfer/autotuner.py
index 2e153ad26c..32bf52d113 100644
--- a/flashinfer/autotuner.py
+++ b/flashinfer/autotuner.py
@@ -61,8 +61,8 @@ def __post_init__(self):
         # Set default tensor_initializers if not provided
         if self.tensor_initializers is None:
             self.tensor_initializers = [
-                lambda shapes, dtype, device: torch.randn(
-                    shapes, device=device, dtype=dtype
+                lambda shapes, dtype, device: torch.randn(shapes, device=device).to(
+                    dtype
                 )
                 for _ in range(len(self.input_idx))
             ]
