diff --git a/c/parallel/src/radix_sort.cu b/c/parallel/src/radix_sort.cu
index 4f4ca7275b2..7cecab03c79 100644
--- a/c/parallel/src/radix_sort.cu
+++ b/c/parallel/src/radix_sort.cu
@@ -229,6 +229,7 @@ try
   const auto policy_sel = cub::detail::radix_sort::policy_selector{
     static_cast<int>(input_keys_it.value_type.size),
     // FIXME(bgruber): input_values_it.value_type.size is 4 when it represents cub::NullType, which is very odd
+    // TODO(bgruber): instead of 0 we should probably use int{sizeof(cub::NullType)}
     keys_only ? 0 : static_cast<int>(input_values_it.value_type.size),
     int{sizeof(OffsetT)},
     key_type};
diff --git a/c/parallel/src/segmented_sort.cu b/c/parallel/src/segmented_sort.cu
index eefc530faf3..634410632d2 100644
--- a/c/parallel/src/segmented_sort.cu
+++ b/c/parallel/src/segmented_sort.cu
@@ -37,7 +37,7 @@
 #include <nvrtc/ltoir_list_appender.h>
 #include <util/build_utils.h>
 
-struct device_segmented_sort_policy;
+struct device_segmented_sort_policy_selector;
 struct device_three_way_partition_policy;
 using OffsetT = ptrdiff_t;
 static_assert(std::is_same_v<cub::detail::choose_signed_offset_t<OffsetT>, OffsetT>, "OffsetT must be long");
@@ -55,15 +55,15 @@ std::string get_device_segmented_sort_fallback_kernel_name(
   std::string_view value_t,
   cccl_sort_order_t sort_order)
 {
-  std::string chained_policy_t;
-  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy>(&chained_policy_t));
+  std::string policy_selector_t;
+  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy_selector>(&policy_selector_t));
 
   std::string offset_t;
   check(cccl_type_name_from_nvrtc<OffsetT>(&offset_t));
 
   /*
   template <SortOrder Order,             // 0 (ascending)
-            typename ChainedPolicyT,     // 1
+            typename PolicySelector,     // 1
             typename KeyT,               // 2
             typename ValueT,             // 3
             typename BeginOffsetIteratorT, // 4
@@ -74,7 +74,7 @@ std::string get_device_segmented_sort_fallback_kernel_name(
   return std::format(
     "cub::detail::segmented_sort::DeviceSegmentedSortFallbackKernel<{0}, {1}, {2}, {3}, {4}, {5}, {6}>",
     (sort_order == CCCL_ASCENDING) ? "cub::SortOrder::Ascending" : "cub::SortOrder::Descending",
-    chained_policy_t, // 0
+    policy_selector_t, // 0
     key_t, // 1
     value_t, // 2
     start_offset_iterator_t, // 3
@@ -90,7 +90,7 @@ std::string get_device_segmented_sort_kernel_small_name(
   cccl_sort_order_t sort_order)
 {
   std::string chained_policy_t;
-  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy>(&chained_policy_t));
+  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy_selector>(&chained_policy_t));
 
   std::string offset_t;
   check(cccl_type_name_from_nvrtc<OffsetT>(&offset_t));
@@ -124,7 +124,7 @@ std::string get_device_segmented_sort_kernel_large_name(
   cccl_sort_order_t sort_order)
 {
   std::string chained_policy_t;
-  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy>(&chained_policy_t));
+  check(cccl_type_name_from_nvrtc<device_segmented_sort_policy_selector>(&chained_policy_t));
 
   std::string offset_t;
   check(cccl_type_name_from_nvrtc<OffsetT>(&offset_t));
@@ -349,97 +349,6 @@ struct partition_kernel_source
   }
 };
 
-struct segmented_sort_runtime_tuning_policy
-{
-  cub::detail::RuntimeRadixSortDownsweepAgentPolicy large_segment;
-  cub::detail::RuntimeSubWarpMergeSortAgentPolicy small_segment;
-  cub::detail::RuntimeSubWarpMergeSortAgentPolicy medium_segment;
-  int partitioning_threshold;
-
-  auto LargeSegment() const
-  {
-    return large_segment;
-  }
-
-  auto SmallSegment() const
-  {
-    return small_segment;
-  }
-
-  auto MediumSegment() const
-  {
-    return medium_segment;
-  }
-
-  int PartitioningThreshold() const
-  {
-    return partitioning_threshold;
-  }
-
-  int LargeSegmentRadixBits() const
-  {
-    return large_segment.RadixBits();
-  }
-
-  int SegmentsPerSmallBlock() const
-  {
-    return small_segment.SegmentsPerBlock();
-  }
-
-  int SegmentsPerMediumBlock() const
-  {
-    return medium_segment.SegmentsPerBlock();
-  }
-
-  int SmallPolicyItemsPerTile() const
-  {
-    return small_segment.ItemsPerTile();
-  }
-
-  int MediumPolicyItemsPerTile() const
-  {
-    return medium_segment.ItemsPerTile();
-  }
-
-  cub::CacheLoadModifier LargeSegmentLoadModifier() const
-  {
-    return large_segment.LoadModifier();
-  }
-
-  cub::BlockLoadAlgorithm LargeSegmentLoadAlgorithm() const
-  {
-    return large_segment.LoadAlgorithm();
-  }
-
-  cub::WarpLoadAlgorithm MediumSegmentLoadAlgorithm() const
-  {
-    return medium_segment.LoadAlgorithm();
-  }
-
-  cub::WarpLoadAlgorithm SmallSegmentLoadAlgorithm() const
-  {
-    return small_segment.LoadAlgorithm();
-  }
-
-  cub::WarpStoreAlgorithm MediumSegmentStoreAlgorithm() const
-  {
-    return medium_segment.StoreAlgorithm();
-  }
-
-  cub::WarpStoreAlgorithm SmallSegmentStoreAlgorithm() const
-  {
-    return small_segment.StoreAlgorithm();
-  }
-
-  using MaxPolicy = segmented_sort_runtime_tuning_policy;
-
-  template <typename F>
-  cudaError_t Invoke(int, F& op)
-  {
-    return op.template Invoke<segmented_sort_runtime_tuning_policy>(*this);
-  }
-};
-
 struct partition_runtime_tuning_policy
 {
   cub::detail::RuntimeThreeWayPartitionAgentPolicy three_way_partition;
@@ -534,26 +443,22 @@ try
 
   const char* name = "device_segmented_sort";
 
-  const int cc = cc_major * 10 + cc_minor;
-
   const auto [keys_in_iterator_name, keys_in_iterator_src] =
     get_specialization<segmented_sort_keys_input_iterator_tag>(template_id<input_iterator_traits>(), keys_in_it);
 
   const bool keys_only = values_in_it.type == cccl_iterator_kind_t::CCCL_POINTER && values_in_it.state == nullptr;
 
-  std::string values_in_iterator_name, values_in_iterator_src;
+  std::string values_in_iterator_src;
 
   if (!keys_only)
   {
     const auto [vi_name, vi_src] =
       get_specialization<segmented_sort_values_input_iterator_tag>(template_id<input_iterator_traits>(), values_in_it);
-    values_in_iterator_name = vi_name;
-    values_in_iterator_src  = vi_src;
+    values_in_iterator_src = vi_src;
   }
   else
   {
-    values_in_iterator_name = "cub::NullType*";
-    values_in_iterator_src  = "";
+    values_in_iterator_src = "";
   }
 
   const auto [start_offset_iterator_name, start_offset_iterator_src] =
@@ -624,8 +529,17 @@ try
   const auto [small_selector_name, small_selector_src] = get_specialization<segmented_sort_small_selector_tag>(
     template_id<user_operation_traits>(), small_selector_op, selector_result_t, selector_input_t);
 
-  const auto segmented_sort_policy_hub_expr = std::format(
-    "cub::detail::segmented_sort::policy_hub<{0}, {1}>",
+  const auto policy_sel = cub::detail::segmented_sort::policy_selector{
+    static_cast<int>(keys_in_it.value_type.size),
+    keys_only ? int{sizeof(cub::NullType)} : static_cast<int>(values_in_it.value_type.size),
+    keys_only};
+
+  // TODO(bgruber): drop this if tuning policies become formattable
+  std::stringstream segmented_sort_policy_sel_str;
+  segmented_sort_policy_sel_str << policy_sel(cuda::to_arch_id(cuda::compute_capability{cc_major, cc_minor}));
+
+  const auto segmented_sort_policy_expr = std::format(
+    "cub::detail::segmented_sort::policy_selector_from_types<{}, {}>",
     key_t, // 0
     value_t); // 1
 
@@ -658,14 +572,15 @@ struct __align__({4}) items_storage_t {{
 {8}
 {9}
 {10}
-using device_segmented_sort_policy = {11}::MaxPolicy;
-using device_three_way_partition_policy = {12}::MaxPolicy;
+using device_segmented_sort_policy_selector = {11};
+using namespace cub;
+using namespace cub::detail::segmented_sort;
+static_assert(
+  device_segmented_sort_policy_selector()(::cuda::arch_id{{CUB_PTX_ARCH / 10}}) == {12},
+  "Host generated and JIT compiled policy mismatch");
+using device_three_way_partition_policy = {13}::MaxPolicy;
 
 #include <cub/detail/ptx-json/json.cuh>
-__device__ consteval auto& segmented_sort_policy_generator() {{
-  return ptx_json::id<ptx_json::string("device_segmented_sort_policy")>()
-    = cub::detail::segmented_sort::SegmentedSortPolicyWrapper<device_segmented_sort_policy::ActivePolicy>::EncodedPolicy();
-}}
 __device__ consteval auto& three_way_partition_policy_generator() {{
   return ptx_json::id<ptx_json::string("device_three_way_partition_policy")>()
     = cub::detail::three_way_partition::ThreeWayPartitionPolicyWrapper<device_three_way_partition_policy::ActivePolicy>::EncodedPolicy();
@@ -682,8 +597,15 @@ __device__ consteval auto& three_way_partition_policy_generator() {{
     end_offset_iterator_src, // 8
     large_selector_src, // 9
     small_selector_src, // 10
-    segmented_sort_policy_hub_expr, // 11
-    three_way_partition_policy_hub_expr); // 12
+    segmented_sort_policy_expr, // 11
+    segmented_sort_policy_sel_str.view(), // 12
+    three_way_partition_policy_hub_expr); // 13
+
+#if false // CCCL_DEBUGGING_SWITCH
+  fflush(stderr);
+  printf("\nCODE4NVRTC BEGIN\n%sCODE4NVRTC END\n", final_src.c_str());
+  fflush(stdout);
+#endif
 
   std::vector<const char*> args = {
     arch.c_str(),
@@ -695,7 +617,7 @@ __device__ consteval auto& three_way_partition_policy_generator() {{
     "-dlto",
     "-default-device",
     "-DCUB_DISABLE_CDP",
-    "-DCUB_ENABLE_POLICY_PTX_JSON",
+    "-DCUB_ENABLE_POLICY_PTX_JSON", // TODO(bgruber): remove after we ported three way partition to the new tuning API
     "-std=c++20"};
 
   cccl::detail::extend_args_with_build_config(args, config);
@@ -769,18 +691,7 @@ __device__ consteval auto& three_way_partition_policy_generator() {{
   check(cuLibraryGetKernel(
     &build_ptr->three_way_partition_kernel, build_ptr->library, three_way_partition_kernel_lowered_name.c_str()));
 
-  nlohmann::json runtime_policy =
-    cub::detail::ptx_json::parse("device_segmented_sort_policy", {result.data.get(), result.size});
-
-  using cub::detail::RuntimeRadixSortDownsweepAgentPolicy;
-  auto large_segment_policy = RuntimeRadixSortDownsweepAgentPolicy::from_json(runtime_policy, "LargeSegmentPolicy");
-
-  using cub::detail::RuntimeSubWarpMergeSortAgentPolicy;
-  auto small_segment_policy = RuntimeSubWarpMergeSortAgentPolicy::from_json(runtime_policy, "SmallSegmentPolicy");
-
-  auto medium_segment_policy = RuntimeSubWarpMergeSortAgentPolicy::from_json(runtime_policy, "MediumSegmentPolicy");
-
-  int partitioning_threshold = runtime_policy["PartitioningThreshold"].get<int>();
+  // TODO(bgruber): convert to the new tuning API
   nlohmann::json partition_policy =
     cub::detail::ptx_json::parse("device_three_way_partition_policy", {result.data.get(), result.size});
 
@@ -788,7 +699,7 @@ __device__ consteval auto& three_way_partition_policy_generator() {{
   auto three_way_partition_policy =
     RuntimeThreeWayPartitionAgentPolicy::from_json(partition_policy, "ThreeWayPartitionPolicy");
 
-  build_ptr->cc                         = cc;
+  build_ptr->cc                         = cc_major * 10 + cc_minor;
   build_ptr->large_segments_selector_op = large_selector_op;
   build_ptr->small_segments_selector_op = small_selector_op;
   build_ptr->cubin                      = (void*) result.data.release();
@@ -796,8 +707,7 @@ __device__ consteval auto& three_way_partition_policy_generator() {{
   build_ptr->key_type                   = keys_in_it.value_type;
   build_ptr->offset_type                = cccl_type_info{sizeof(OffsetT), alignof(OffsetT), cccl_type_enum::CCCL_INT64};
   // Use the runtime policy extracted via from_json
-  build_ptr->runtime_policy = new segmented_sort::segmented_sort_runtime_tuning_policy{
-    large_segment_policy, small_segment_policy, medium_segment_policy, partitioning_threshold};
+  build_ptr->runtime_policy           = new cub::detail::segmented_sort::policy_selector{policy_sel};
   build_ptr->partition_runtime_policy = new segmented_sort::partition_runtime_tuning_policy{three_way_partition_policy};
   build_ptr->order                    = sort_order;
 
@@ -888,35 +798,34 @@ CUresult cccl_device_segmented_sort_impl(
     cub::DoubleBuffer<indirect_arg_t> d_values_double_buffer(
       *static_cast<indirect_arg_t**>(&val_arg_in), *static_cast<indirect_arg_t**>(&val_arg_out));
 
-    auto exec_status = cub::DispatchSegmentedSort<
+    // TODO(bgruber): remove all template arguments except the first two (the others can be deduced)
+    auto exec_status = cub::detail::segmented_sort::dispatch<
       Order,
+      OffsetT, // OffsetT
       indirect_arg_t, // KeyT
       indirect_arg_t, // ValueT
-      OffsetT, // OffsetT
       indirect_iterator_t, // BeginOffsetIteratorT
       indirect_iterator_t, // EndOffsetIteratorT
-      segmented_sort::segmented_sort_runtime_tuning_policy, // PolicyHub
+      cub::detail::segmented_sort::policy_selector, // PolicySelector
       segmented_sort::segmented_sort_kernel_source, // KernelSource
-      segmented_sort::partition_runtime_tuning_policy, // PartitionPolicyHub
-      segmented_sort::partition_kernel_source, // PartitionKernelSource
-      cub::detail::CudaDriverLauncherFactory>:: // KernelLaunchFactory
-      Dispatch(
-        d_temp_storage,
+      segmented_sort::partition_runtime_tuning_policy // PartitionPolicyHub
+      >(d_temp_storage,
         *temp_storage_bytes,
         d_keys_double_buffer,
         d_values_double_buffer,
         num_items,
         num_segments,
-        start_offset_in,
-        end_offset_in,
+        indirect_iterator_t{start_offset_in},
+        indirect_iterator_t{end_offset_in},
         is_overwrite_okay,
         stream,
-        /* kernel_source */ {build},
-        /* partition_kernel_source */ {build},
-        /* launcher_factory */ cub::detail::CudaDriverLauncherFactory{cu_device, build.cc},
-        /* policy */ *reinterpret_cast<segmented_sort::segmented_sort_runtime_tuning_policy*>(build.runtime_policy),
-        /* partition_policy */
-        *reinterpret_cast<segmented_sort::partition_runtime_tuning_policy*>(build.partition_runtime_policy));
+        /* policy_selector */
+        *static_cast<cub::detail::segmented_sort::policy_selector*>(build.runtime_policy),
+        /* partition_max_policy */
+        *static_cast<segmented_sort::partition_runtime_tuning_policy*>(build.partition_runtime_policy),
+        /* kernel_source */ segmented_sort::segmented_sort_kernel_source{build},
+        /* partition_kernel_source */ segmented_sort::partition_kernel_source{build},
+        /* launcher_factory */ cub::detail::CudaDriverLauncherFactory{cu_device, build.cc});
 
     *selector = d_keys_double_buffer.selector;
     error     = static_cast<CUresult>(exec_status);
@@ -998,8 +907,8 @@ try
   std::unique_ptr<char[]> small_code(const_cast<char*>(build_ptr->small_segments_selector_op.code));
 
   // Clean up the runtime policies
-  std::unique_ptr<segmented_sort::segmented_sort_runtime_tuning_policy> rtp(
-    static_cast<segmented_sort::segmented_sort_runtime_tuning_policy*>(build_ptr->runtime_policy));
+  std::unique_ptr<cub::detail::segmented_sort::policy_selector> rtp(
+    static_cast<cub::detail::segmented_sort::policy_selector*>(build_ptr->runtime_policy));
   std::unique_ptr<segmented_sort::partition_runtime_tuning_policy> prtp(
     static_cast<segmented_sort::partition_runtime_tuning_policy*>(build_ptr->partition_runtime_policy));
   check(cuLibraryUnload(build_ptr->library));
diff --git a/cub/benchmarks/bench/segmented_sort/keys.cu b/cub/benchmarks/bench/segmented_sort/keys.cu
index 9d2bb53b59a..03b1ec54d65 100644
--- a/cub/benchmarks/bench/segmented_sort/keys.cu
+++ b/cub/benchmarks/bench/segmented_sort/keys.cu
@@ -1,7 +1,7 @@
 // SPDX-FileCopyrightText: Copyright (c) 2011-2023, NVIDIA CORPORATION. All rights reserved.
 // SPDX-License-Identifier: BSD-3
 
-#include <cub/device/device_segmented_sort.cuh>
+#include <cub/device/dispatch/dispatch_segmented_sort.cuh>
 
 #include <nvbench_helper.cuh>
 
@@ -22,95 +22,58 @@
 // %RANGE% TUNE_M_TRANSPOSE mtrp 0:1:1
 
 #if !TUNE_BASE
-
-#  define TUNE_SW_THREADS (1 << TUNE_SW_THREADS_POW2)
-#  define TUNE_MW_THREADS (1 << TUNE_MW_THREADS_POW2)
-
-#  define SMALL_SEGMENT_SIZE  TUNE_S_ITEMS* TUNE_SW_THREADS
-#  define MEDIUM_SEGMENT_SIZE TUNE_M_ITEMS* TUNE_MW_THREADS
-#  define LARGE_SEGMENT_SIZE  TUNE_L_ITEMS* TUNE_THREADS
-
-#  if (LARGE_SEGMENT_SIZE <= SMALL_SEGMENT_SIZE) || (LARGE_SEGMENT_SIZE <= MEDIUM_SEGMENT_SIZE)
-#    error Large segment size must be larger than small and medium segment sizes
-#  endif
-
-#  if (MEDIUM_SEGMENT_SIZE <= SMALL_SEGMENT_SIZE)
-#    error Medium segment size must be larger than small one
-#  endif
-
-#  if TUNE_LOAD == 0
-#    define TUNE_LOAD_MODIFIER cub::LOAD_DEFAULT
-#  elif TUNE_LOAD == 1
-#    define TUNE_LOAD_MODIFIER cub::LOAD_LDG
-#  else // TUNE_LOAD == 2
-#    define TUNE_LOAD_MODIFIER cub::LOAD_CA
-#  endif // TUNE_LOAD
-
-#  if TUNE_S_LOAD == 0
-#    define TUNE_S_LOAD_MODIFIER cub::LOAD_DEFAULT
-#  elif TUNE_S_LOAD == 1
-#    define TUNE_S_LOAD_MODIFIER cub::LOAD_LDG
-#  else // TUNE_S_LOAD == 2
-#    define TUNE_S_LOAD_MODIFIER cub::LOAD_CA
-#  endif // TUNE_S_LOAD
-
-#  if TUNE_M_LOAD == 0
-#    define TUNE_M_LOAD_MODIFIER cub::LOAD_DEFAULT
-#  elif TUNE_M_LOAD == 1
-#    define TUNE_M_LOAD_MODIFIER cub::LOAD_LDG
-#  else // TUNE_M_LOAD == 2
-#    define TUNE_M_LOAD_MODIFIER cub::LOAD_CA
-#  endif // TUNE_M_LOAD
-
-#  if TUNE_TRANSPOSE == 0
-#    define TUNE_LOAD_ALGORITHM cub::BLOCK_LOAD_DIRECT
-#  else // TUNE_TRANSPOSE == 1
-#    define TUNE_LOAD_ALGORITHM cub::BLOCK_LOAD_WARP_TRANSPOSE
-#  endif // TUNE_TRANSPOSE
-
-#  if TUNE_S_TRANSPOSE == 0
-#    define TUNE_S_LOAD_ALGORITHM cub::WarpLoadAlgorithm::WARP_LOAD_DIRECT
-#  else // TUNE_S_TRANSPOSE == 1
-#    define TUNE_S_LOAD_ALGORITHM cub::WarpLoadAlgorithm::WARP_LOAD_TRANSPOSE
-#  endif // TUNE_S_TRANSPOSE
-
-#  if TUNE_M_TRANSPOSE == 0
-#    define TUNE_M_LOAD_ALGORITHM cub::WarpLoadAlgorithm::WARP_LOAD_DIRECT
-#  else // TUNE_M_TRANSPOSE == 1
-#    define TUNE_M_LOAD_ALGORITHM cub::WarpLoadAlgorithm::WARP_LOAD_TRANSPOSE
-#  endif // TUNE_M_TRANSPOSE
-
 template <class KeyT>
-struct device_seg_sort_policy_hub
+struct device_seg_sort_policy_selector
 {
-  using DominantT = KeyT;
-
-  struct Policy500 : cub::ChainedPolicy<500, Policy500, Policy500>
+  _CCCL_API constexpr auto operator()(::cuda::arch_id /*arch*/) const
   {
-    static constexpr int BLOCK_THREADS          = TUNE_THREADS;
-    static constexpr int RADIX_BITS             = TUNE_RADIX_BITS;
-    static constexpr int PARTITIONING_THRESHOLD = TUNE_PARTITIONING_THRESHOLD;
-
-    using LargeSegmentPolicy = cub::AgentRadixSortDownsweepPolicy<
-      BLOCK_THREADS,
-      TUNE_L_ITEMS,
-      DominantT,
-      TUNE_LOAD_ALGORITHM,
-      TUNE_LOAD_MODIFIER,
-      static_cast<cub::RadixRankAlgorithm>(TUNE_RANK_ALGORITHM),
-      cub::BLOCK_SCAN_WARP_SCANS,
-      RADIX_BITS>;
-
-    static constexpr int ITEMS_PER_SMALL_THREAD  = TUNE_S_ITEMS;
-    static constexpr int ITEMS_PER_MEDIUM_THREAD = TUNE_M_ITEMS;
-
-    using SmallSegmentPolicy = cub::
-      AgentSubWarpMergeSortPolicy<TUNE_SW_THREADS, ITEMS_PER_SMALL_THREAD, TUNE_S_LOAD_ALGORITHM, TUNE_S_LOAD_MODIFIER>;
-    using MediumSegmentPolicy = cub::
-      AgentSubWarpMergeSortPolicy<TUNE_MW_THREADS, ITEMS_PER_MEDIUM_THREAD, TUNE_M_LOAD_ALGORITHM, TUNE_M_LOAD_MODIFIER>;
-  };
-
-  using MaxPolicy = Policy500;
+    constexpr int tune_sw_threads     = 1 << TUNE_SW_THREADS_POW2;
+    constexpr int tune_mw_threads     = 1 << TUNE_MW_THREADS_POW2;
+    constexpr int small_segment_size  = TUNE_S_ITEMS * tune_sw_threads;
+    constexpr int medium_segment_size = TUNE_M_ITEMS * tune_mw_threads;
+    constexpr int large_segment_size  = TUNE_L_ITEMS * TUNE_THREADS;
+
+    static_assert((large_segment_size > small_segment_size) && (large_segment_size > medium_segment_size),
+                  "Large segment size must be larger than small and medium segment sizes");
+    static_assert(medium_segment_size > small_segment_size, "Medium segment size must be larger than small one");
+
+    using namespace cub::detail::segmented_sort;
+
+    return segmented_sort_policy{
+      segmented_radix_sort_policy{
+        TUNE_THREADS,
+        TUNE_L_ITEMS,
+        (TUNE_TRANSPOSE == 0) ? cub::BLOCK_LOAD_DIRECT : cub::BLOCK_LOAD_WARP_TRANSPOSE,
+        (TUNE_LOAD == 0)   ? cub::LOAD_DEFAULT
+        : (TUNE_LOAD == 1) ? cub::LOAD_LDG
+                           : cub::LOAD_CA,
+        static_cast<cub::RadixRankAlgorithm>(TUNE_RANK_ALGORITHM),
+        cub::BLOCK_SCAN_WARP_SCANS,
+        TUNE_RADIX_BITS,
+      },
+      sub_warp_merge_sort_policy{
+        TUNE_THREADS,
+        tune_sw_threads,
+        TUNE_S_ITEMS,
+        (TUNE_S_TRANSPOSE == 0) ? cub::WarpLoadAlgorithm::WARP_LOAD_DIRECT : cub::WarpLoadAlgorithm::WARP_LOAD_TRANSPOSE,
+        cub::WARP_STORE_DIRECT,
+        (TUNE_S_LOAD == 0)   ? cub::LOAD_DEFAULT
+        : (TUNE_S_LOAD == 1) ? cub::LOAD_LDG
+                             : cub::LOAD_CA,
+      },
+      sub_warp_merge_sort_policy{
+        TUNE_THREADS,
+        tune_mw_threads,
+        TUNE_M_ITEMS,
+        (TUNE_M_TRANSPOSE == 0) ? cub::WarpLoadAlgorithm::WARP_LOAD_DIRECT : cub::WarpLoadAlgorithm::WARP_LOAD_TRANSPOSE,
+        cub::WARP_STORE_DIRECT,
+        (TUNE_M_LOAD == 0)   ? cub::LOAD_DEFAULT
+        : (TUNE_M_LOAD == 1) ? cub::LOAD_LDG
+                             : cub::LOAD_CA,
+      },
+      TUNE_PARTITIONING_THRESHOLD,
+    };
+  }
 };
 #endif // !TUNE_BASE
 
@@ -129,15 +92,6 @@ void seg_sort(nvbench::state& state,
   using key_t             = T;
   using value_t           = cub::NullType;
 
-#if !TUNE_BASE
-  using policy_t   = device_seg_sort_policy_hub<key_t>;
-  using dispatch_t = //
-    cub::DispatchSegmentedSort<sort_order, key_t, value_t, offset_t, begin_offset_it_t, end_offset_it_t, policy_t>;
-#else
-  using dispatch_t = //
-    cub::DispatchSegmentedSort<sort_order, key_t, value_t, offset_t, begin_offset_it_t, end_offset_it_t>;
-#endif
-
   const auto elements = static_cast<std::size_t>(state.get_int64("Elements{io}"));
   const auto segments = offsets.size() - 1;
 
@@ -160,7 +114,7 @@ void seg_sort(nvbench::state& state,
 
   std::size_t temp_storage_bytes{};
   std::uint8_t* d_temp_storage{};
-  dispatch_t::Dispatch(
+  cub::detail::segmented_sort::dispatch<sort_order, offset_t>(
     d_temp_storage,
     temp_storage_bytes,
     d_keys,
@@ -170,9 +124,14 @@ void seg_sort(nvbench::state& state,
     d_begin_offsets,
     d_end_offsets,
     is_overwrite_ok,
-    0);
+    0
+#if !TUNE_BASE
+    ,
+    device_seg_sort_policy_selector<key_t>{}
+#endif // !TUNE_BASE
+  );
 
-  thrust::device_vector<nvbench::uint8_t> temp_storage(temp_storage_bytes);
+  thrust::device_vector<nvbench::uint8_t> temp_storage(temp_storage_bytes, thrust::no_init);
   d_temp_storage = thrust::raw_pointer_cast(temp_storage.data());
 
   state.exec(nvbench::exec_tag::gpu | nvbench::exec_tag::no_batch | nvbench::exec_tag::sync,
@@ -180,7 +139,7 @@ void seg_sort(nvbench::state& state,
                cub::DoubleBuffer<key_t> keys     = d_keys;
                cub::DoubleBuffer<value_t> values = d_values;
 
-               dispatch_t::Dispatch(
+               cub::detail::segmented_sort::dispatch<sort_order, offset_t>(
                  d_temp_storage,
                  temp_storage_bytes,
                  keys,
@@ -190,7 +149,12 @@ void seg_sort(nvbench::state& state,
                  d_begin_offsets,
                  d_end_offsets,
                  is_overwrite_ok,
-                 launch.get_stream());
+                 launch.get_stream()
+#if !TUNE_BASE
+                   ,
+                 device_seg_sort_policy_selector<key_t>{}
+#endif // !TUNE_BASE
+               );
              });
 }
 
diff --git a/cub/cub/device/device_segmented_sort.cuh b/cub/cub/device/device_segmented_sort.cuh
index 6553bf9491e..8de12cf33c0 100644
--- a/cub/cub/device/device_segmented_sort.cuh
+++ b/cub/cub/device/device_segmented_sort.cuh
@@ -128,13 +128,11 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Ascending, KeyT, cub::NullType, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<KeyT> d_keys(const_cast<KeyT*>(d_keys_in), d_keys_out);
     DoubleBuffer<NullType> d_values;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Ascending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -306,13 +304,11 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Descending, KeyT, cub::NullType, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<KeyT> d_keys(const_cast<KeyT*>(d_keys_in), d_keys_out);
     DoubleBuffer<NullType> d_values;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Descending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -477,12 +473,10 @@ private:
     constexpr bool is_overwrite_okay = true;
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Ascending, KeyT, cub::NullType, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<NullType> d_values;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Ascending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -650,12 +644,10 @@ private:
     constexpr bool is_overwrite_okay = true;
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Descending, KeyT, cub::NullType, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<NullType> d_values;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Descending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -1387,13 +1379,11 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Ascending, KeyT, ValueT, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<KeyT> d_keys(const_cast<KeyT*>(d_keys_in), d_keys_out);
     DoubleBuffer<ValueT> d_values(const_cast<ValueT*>(d_values_in), d_values_out);
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Ascending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -1592,13 +1582,11 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Descending, KeyT, ValueT, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
     DoubleBuffer<KeyT> d_keys(const_cast<KeyT*>(d_keys_in), d_keys_out);
     DoubleBuffer<ValueT> d_values(const_cast<ValueT*>(d_values_in), d_values_out);
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Descending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -1791,10 +1779,8 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Ascending, KeyT, ValueT, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Ascending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
@@ -1992,10 +1978,8 @@ private:
 
     using OffsetT =
       detail::choose_signed_offset_t<detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>>;
-    using DispatchT =
-      DispatchSegmentedSort<SortOrder::Descending, KeyT, ValueT, OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>;
 
-    return DispatchT::Dispatch(
+    return detail::segmented_sort::dispatch<SortOrder::Descending, OffsetT>(
       d_temp_storage,
       temp_storage_bytes,
       d_keys,
diff --git a/cub/cub/device/dispatch/dispatch_segmented_sort.cuh b/cub/cub/device/dispatch/dispatch_segmented_sort.cuh
index a4c523211db..728ec456f88 100644
--- a/cub/cub/device/dispatch/dispatch_segmented_sort.cuh
+++ b/cub/cub/device/dispatch/dispatch_segmented_sort.cuh
@@ -13,6 +13,7 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cub/detail/arch_dispatch.cuh>
 #include <cub/detail/device_double_buffer.cuh>
 #include <cub/detail/temporary_storage.cuh>
 #include <cub/device/device_partition.cuh>
@@ -35,25 +36,24 @@
 
 #include <nv/target>
 
+#if !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+#  include <sstream>
+#endif // !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+
 CUB_NAMESPACE_BEGIN
 
 namespace detail::segmented_sort
 {
-/*
- * Continuation is called after the partitioning stage. It launches kernels
- * to sort large and small segments using the partitioning results. Separation
- * of this stage is required to eliminate device-side synchronization in
- * the CDP mode.
- */
-template <typename WrappedPolicyT,
-          typename LargeKernelT,
+// Continuation is called after the partitioning stage. It launches kernels to sort large and small segments using the
+// partitioning results. Separation of this stage is required to eliminate device-side synchronization in the CDP mode.
+template <typename LargeKernelT,
           typename SmallKernelT,
           typename KeyT,
           typename ValueT,
           typename BeginOffsetIteratorT,
           typename EndOffsetIteratorT,
           typename KernelLauncherFactory = CUB_DETAIL_DEFAULT_KERNEL_LAUNCHER_FACTORY>
-CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortContinuation(
+CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t device_segmented_sort_continuation(
   LargeKernelT large_kernel,
   SmallKernelT small_kernel,
   int num_segments,
@@ -70,7 +70,10 @@ CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortCont
   local_segment_index_t* small_segments_indices,
   cudaStream_t stream,
   KernelLauncherFactory launcher_factory,
-  WrappedPolicyT wrapped_policy)
+  int large_block_threads,
+  int small_block_threads,
+  int medium_segments_per_block,
+  int small_segments_per_block)
 {
   using local_segment_index_t                = local_segment_index_t;
   const local_segment_index_t large_segments = group_sizes[0];
@@ -84,21 +87,25 @@ CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortCont
     _CubLog("Invoking "
             "DeviceSegmentedSortKernelLarge<<<%d, %d, 0, %lld>>>()\n",
             static_cast<int>(blocks_in_grid),
-            wrapped_policy.LargeSegment().BlockThreads(),
+            large_block_threads,
             (long long) stream);
 #endif // CUB_DEBUG_LOG
 
-    launcher_factory(blocks_in_grid, wrapped_policy.LargeSegment().BlockThreads(), 0, stream)
-      .doit(large_kernel,
-            large_and_medium_segments_indices,
-            d_current_keys,
-            d_final_keys,
-            d_keys_double_buffer,
-            d_current_values,
-            d_final_values,
-            d_values_double_buffer,
-            d_begin_offsets,
-            d_end_offsets);
+    if (const auto error = CubDebug(
+          launcher_factory(blocks_in_grid, large_block_threads, 0, stream)
+            .doit(large_kernel,
+                  large_and_medium_segments_indices,
+                  d_current_keys,
+                  d_final_keys,
+                  d_keys_double_buffer,
+                  d_current_values,
+                  d_final_values,
+                  d_values_double_buffer,
+                  d_begin_offsets,
+                  d_end_offsets)))
+    {
+      return error;
+    }
 
     // Check for failure to launch
     if (const auto error = CubDebug(cudaPeekAtLastError()))
@@ -117,10 +124,9 @@ CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortCont
   const local_segment_index_t medium_segments =
     static_cast<local_segment_index_t>(num_segments) - (large_segments + small_segments);
 
-  const local_segment_index_t small_blocks = ::cuda::ceil_div(small_segments, wrapped_policy.SegmentsPerSmallBlock());
+  const local_segment_index_t small_blocks = ::cuda::ceil_div(small_segments, small_segments_per_block);
 
-  const local_segment_index_t medium_blocks =
-    ::cuda::ceil_div(medium_segments, wrapped_policy.SegmentsPerMediumBlock());
+  const local_segment_index_t medium_blocks = ::cuda::ceil_div(medium_segments, medium_segments_per_block);
 
   const local_segment_index_t small_and_medium_blocks_in_grid = small_blocks + medium_blocks;
 
@@ -130,11 +136,11 @@ CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortCont
     _CubLog("Invoking "
             "DeviceSegmentedSortKernelSmall<<<%d, %d, 0, %lld>>>()\n",
             static_cast<int>(small_and_medium_blocks_in_grid),
-            wrapped_policy.SmallSegment().BlockThreads(),
+            small_block_threads,
             (long long) stream);
 #endif // CUB_DEBUG_LOG
 
-    launcher_factory(small_and_medium_blocks_in_grid, wrapped_policy.SmallSegment().BlockThreads(), 0, stream)
+    launcher_factory(small_and_medium_blocks_in_grid, small_block_threads, 0, stream)
       .doit(small_kernel,
             small_segments,
             medium_segments,
@@ -167,10 +173,9 @@ CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN cudaError_t DeviceSegmentedSortCont
 #ifdef CUB_RDC_ENABLED
 /*
  * Continuation kernel is used only in the CDP mode. It's used to
- * launch DeviceSegmentedSortContinuation as a separate kernel.
+ * launch device_segmented_sort_continuation as a separate kernel.
  */
-template <typename WrappedPolicyT,
-          typename LargeKernelT,
+template <typename LargeKernelT,
           typename SmallKernelT,
           typename KeyT,
           typename ValueT,
@@ -193,7 +198,10 @@ __launch_bounds__(1) CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortContin
   local_segment_index_t* large_and_medium_segments_indices,
   local_segment_index_t* small_segments_indices,
   KernelLauncherFactory launcher_factory,
-  WrappedPolicyT wrapped_policy)
+  int large_block_threads,
+  int small_block_threads,
+  int medium_segments_per_block,
+  int small_segments_per_block)
 {
   // In case of CDP:
   // 1. each CTA has a different main stream
@@ -204,7 +212,7 @@ __launch_bounds__(1) CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortContin
   //
   // Due to (4, 5), we can't pass the user-provided stream in the continuation.
   // Due to (1, 2, 3) it's safe to pass the main stream.
-  [[maybe_unused]] const auto error = CubDebug(detail::segmented_sort::DeviceSegmentedSortContinuation<WrappedPolicyT>(
+  [[maybe_unused]] const auto error = CubDebug(detail::segmented_sort::device_segmented_sort_continuation(
     large_kernel,
     small_kernel,
     num_segments,
@@ -221,10 +229,14 @@ __launch_bounds__(1) CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortContin
     small_segments_indices,
     0, // always launching on the main stream (see motivation above)
     launcher_factory,
-    wrapped_policy));
+    large_block_threads,
+    small_block_threads,
+    medium_segments_per_block,
+    small_segments_per_block));
 }
 #endif // CUB_RDC_ENABLED
-template <typename MaxPolicyT,
+
+template <typename PolicySelector,
           SortOrder Order,
           typename KeyT,
           typename ValueT,
@@ -235,15 +247,21 @@ struct DeviceSegmentedSortKernelSource
 {
   CUB_DEFINE_KERNEL_GETTER(
     SegmentedSortFallbackKernel,
-    DeviceSegmentedSortFallbackKernel<Order, MaxPolicyT, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>);
+    DeviceSegmentedSortFallbackKernel<Order,
+                                      PolicySelector,
+                                      KeyT,
+                                      ValueT,
+                                      BeginOffsetIteratorT,
+                                      EndOffsetIteratorT,
+                                      OffsetT>);
 
   CUB_DEFINE_KERNEL_GETTER(
     SegmentedSortKernelSmall,
-    DeviceSegmentedSortKernelSmall<Order, MaxPolicyT, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>);
+    DeviceSegmentedSortKernelSmall<Order, PolicySelector, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>);
 
   CUB_DEFINE_KERNEL_GETTER(
     SegmentedSortKernelLarge,
-    DeviceSegmentedSortKernelLarge<Order, MaxPolicyT, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>);
+    DeviceSegmentedSortKernelLarge<Order, PolicySelector, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>);
 
   CUB_RUNTIME_FUNCTION static constexpr size_t KeySize()
   {
@@ -274,8 +292,53 @@ struct DeviceSegmentedSortKernelSource
     selector.base_segment_offset = base_segment_offset;
   }
 };
+
+// TODO(bgruber): remove in CCCL 4.0
+template <typename PolicyHub>
+struct policy_selector_from_hub
+{
+  using max_policy = typename PolicyHub::MaxPolicy;
+
+  // this is only called in device code, so we can ignore the arch parameter
+  _CCCL_DEVICE_API constexpr auto operator()(::cuda::arch_id /*arch*/) const -> segmented_sort_policy
+  {
+    using ap = typename PolicyHub::MaxPolicy::ActivePolicy;
+    using lp = typename ap::LargeSegmentPolicy;
+    using sp = typename ap::SmallSegmentPolicy;
+    using mp = typename ap::MediumSegmentPolicy;
+
+    return segmented_sort_policy{
+      segmented_radix_sort_policy{
+        lp::BLOCK_THREADS,
+        lp::ITEMS_PER_THREAD,
+        lp::LOAD_ALGORITHM,
+        lp::LOAD_MODIFIER,
+        lp::RANK_ALGORITHM,
+        lp::SCAN_ALGORITHM,
+        lp::RADIX_BITS},
+      sub_warp_merge_sort_policy{
+        sp::BLOCK_THREADS,
+        sp::WARP_THREADS,
+        sp::ITEMS_PER_THREAD,
+        sp::LOAD_ALGORITHM,
+        sp::STORE_ALGORITHM,
+        sp::LOAD_MODIFIER},
+      sub_warp_merge_sort_policy{
+        mp::BLOCK_THREADS,
+        mp::WARP_THREADS,
+        mp::ITEMS_PER_THREAD,
+        mp::LOAD_ALGORITHM,
+        mp::STORE_ALGORITHM,
+        mp::LOAD_MODIFIER},
+      ap::PARTITIONING_THRESHOLD};
+  }
+};
+
+// Partition selects large and small groups. The middle group is not selected.
+static constexpr size_t num_selected_groups = 2;
 } // namespace detail::segmented_sort
 
+// TODO(bgruber): deprecate when we make the tuning API public and remove in CCCL 4.0
 template <
   SortOrder Order,
   typename KeyT,
@@ -285,7 +348,7 @@ template <
   typename EndOffsetIteratorT,
   typename PolicyHub    = detail::segmented_sort::policy_hub<KeyT, ValueT>,
   typename KernelSource = detail::segmented_sort::DeviceSegmentedSortKernelSource<
-    typename PolicyHub::MaxPolicy,
+    detail::segmented_sort::policy_selector_from_hub<PolicyHub>,
     Order,
     KeyT,
     ValueT,
@@ -317,7 +380,7 @@ struct DispatchSegmentedSort
   static constexpr int KEYS_ONLY = ::cuda::std::is_same_v<ValueT, NullType>;
 
   // Partition selects large and small groups. The middle group is not selected.
-  static constexpr size_t num_selected_groups = 2;
+  static constexpr size_t num_selected_groups = detail::segmented_sort::num_selected_groups;
 
   /**
    * Device-accessible allocation of temporary storage. When `nullptr`, the
@@ -779,7 +842,6 @@ private:
           launcher_factory(1, 1, 0, stream)                                             \
             .doit(                                                                      \
               detail::segmented_sort::DeviceSegmentedSortContinuationKernel<            \
-                WrappedPolicyT,                                                         \
                 LargeKernelT,                                                           \
                 SmallKernelT,                                                           \
                 KeyT,                                                                   \
@@ -802,7 +864,10 @@ private:
               large_and_medium_segments_indices.get(),                                  \
               small_segments_indices.get(),                                             \
               launcher_factory,                                                         \
-              wrapped_policy)))                                                         \
+              wrapped_policy.large_segment.BlockThreads(),                              \
+              wrapped_policy.small_segment.BlockThreads(),                              \
+              wrapped_policy.medium_segment.SegmentsPerBlock(),                         \
+              wrapped_policy.small_segment.SegmentsPerBlock())))                        \
     {                                                                                   \
       return error;                                                                     \
     }                                                                                   \
@@ -813,20 +878,26 @@ private:
     }
 #endif // CUB_RDC_ENABLED
 
-      NV_IF_TARGET(
+      NV_IF_ELSE_TARGET(
         NV_IS_HOST,
-        (
+        ({
           local_segment_index_t h_group_sizes[num_selected_groups];
           if (const auto error = CubDebug(launcher_factory.MemcpyAsync(
                 h_group_sizes,
                 group_sizes.get(),
                 num_selected_groups * sizeof(local_segment_index_t),
                 cudaMemcpyDeviceToHost,
-                stream))) { return error; }
+                stream)))
+          {
+            return error;
+          }
 
-          if (const auto error = CubDebug(SyncStream(stream))) { return error; }
+          if (const auto error = CubDebug(SyncStream(stream)))
+          {
+            return error;
+          }
 
-          if (const auto error = detail::segmented_sort::DeviceSegmentedSortContinuation(
+          if (const auto error = detail::segmented_sort::device_segmented_sort_continuation(
                 large_kernel,
                 small_kernel,
                 current_num_segments,
@@ -843,7 +914,14 @@ private:
                 small_segments_indices.get(),
                 stream,
                 launcher_factory,
-                wrapped_policy)) { return error; }),
+                wrapped_policy.LargeSegment().BlockThreads(),
+                wrapped_policy.SmallSegment().BlockThreads(),
+                wrapped_policy.SegmentsPerMediumBlock(),
+                wrapped_policy.SegmentsPerSmallBlock()))
+          {
+            return error;
+          }
+        }),
         // NV_IS_DEVICE:
         (CUB_TEMP_DEVICE_CODE));
     }
@@ -896,4 +974,594 @@ private:
   }
 };
 
+namespace detail::segmented_sort
+{
+template <typename PartitionPolicyHub,
+          typename LargeKernelT,
+          typename SmallKernelT,
+          typename KeyT,
+          typename ValueT,
+          typename BeginOffsetIteratorT,
+          typename EndOffsetIteratorT,
+          typename KernelSource,
+          typename PartitionKernelSource,
+          typename KernelLauncherFactory,
+          typename GetFinalOutputOp>
+CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN _CCCL_FORCEINLINE cudaError_t sort_with_partitioning(
+  LargeKernelT large_kernel,
+  SmallKernelT small_kernel,
+  global_segment_offset_t num_segments,
+  ::cuda::std::int64_t num_items,
+  BeginOffsetIteratorT d_begin_offsets,
+  EndOffsetIteratorT d_end_offsets,
+  cudaStream_t stream,
+  size_t three_way_partition_temp_storage_bytes,
+  DoubleBuffer<KeyT>& d_keys,
+  DoubleBuffer<ValueT>& d_values,
+  device_double_buffer<KeyT>& d_keys_double_buffer,
+  device_double_buffer<ValueT>& d_values_double_buffer,
+  typename KernelSource::LargeSegmentsSelectorT& large_segments_selector,
+  typename KernelSource::SmallSegmentsSelectorT& small_segments_selector,
+  temporary_storage::alias<uint8_t>& device_partition_temp_storage,
+  temporary_storage::alias<local_segment_index_t>& large_and_medium_segments_indices,
+  temporary_storage::alias<local_segment_index_t>& small_segments_indices,
+  temporary_storage::alias<local_segment_index_t>& group_sizes,
+  KernelSource& kernel_source,
+  PartitionKernelSource& partition_kernel_source,
+  KernelLauncherFactory& launcher_factory,
+  typename PartitionPolicyHub::MaxPolicy partition_max_policy,
+  const segmented_sort_policy& active_policy,
+  GetFinalOutputOp&& get_final_output)
+{
+  constexpr auto num_segments_per_invocation_limit =
+    static_cast<global_segment_offset_t>(::cuda::std::numeric_limits<int>::max());
+
+  const global_segment_offset_t num_invocations =
+    ::cuda::ceil_div(static_cast<global_segment_offset_t>(num_segments), num_segments_per_invocation_limit);
+  for (global_segment_offset_t invocation_index = 0; invocation_index < num_invocations; invocation_index++)
+  {
+    const global_segment_offset_t current_seg_offset = invocation_index * num_segments_per_invocation_limit;
+    const local_segment_index_t current_num_segments =
+      (invocation_index == (num_invocations - 1))
+        ? static_cast<local_segment_index_t>(num_segments - current_seg_offset)
+        : num_segments_per_invocation_limit;
+
+    kernel_source.SetSegmentOffset(large_segments_selector, current_seg_offset);
+    kernel_source.SetSegmentOffset(small_segments_selector, current_seg_offset);
+
+    BeginOffsetIteratorT current_begin_offset = d_begin_offsets;
+    EndOffsetIteratorT current_end_offset     = d_end_offsets;
+    current_begin_offset += current_seg_offset;
+    current_end_offset += current_seg_offset;
+
+    auto medium_indices_iterator =
+      ::cuda::std::make_reverse_iterator(large_and_medium_segments_indices.get() + current_num_segments);
+
+    using ChooseOffsetT                = choose_signed_offset<global_segment_offset_t>;
+    using PartitionOffsetT             = typename ChooseOffsetT::type;
+    using DispatchThreeWayPartitionIfT = cub::DispatchThreeWayPartitionIf<
+      THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>,
+      decltype(large_and_medium_segments_indices.get()),
+      decltype(small_segments_indices.get()),
+      decltype(medium_indices_iterator),
+      decltype(group_sizes.get()),
+      decltype(large_segments_selector),
+      decltype(small_segments_selector),
+      PartitionOffsetT,
+      PartitionPolicyHub,
+      PartitionKernelSource,
+      KernelLauncherFactory>;
+
+    if (const auto error = ChooseOffsetT::is_exceeding_offset_type(num_items))
+    {
+      return error;
+    }
+
+    if (const auto error = DispatchThreeWayPartitionIfT::Dispatch(
+          device_partition_temp_storage.get(),
+          three_way_partition_temp_storage_bytes,
+          THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>(0),
+          large_and_medium_segments_indices.get(),
+          small_segments_indices.get(),
+          medium_indices_iterator,
+          group_sizes.get(),
+          large_segments_selector,
+          small_segments_selector,
+          current_num_segments,
+          stream,
+          partition_kernel_source,
+          launcher_factory,
+          partition_max_policy))
+    {
+      return error;
+    }
+
+    [[maybe_unused]] auto device_path = [&] {
+#ifdef CUB_RDC_ENABLED
+      if (const auto error = CubDebug(
+            launcher_factory(1, 1, 0, stream)
+              .doit(
+                detail::segmented_sort::DeviceSegmentedSortContinuationKernel<
+                  LargeKernelT,
+                  SmallKernelT,
+                  KeyT,
+                  ValueT,
+                  BeginOffsetIteratorT,
+                  EndOffsetIteratorT,
+                  KernelLauncherFactory>,
+                large_kernel,
+                small_kernel,
+                current_num_segments,
+                d_keys.Current(),
+                get_final_output(d_keys, active_policy.large_segment.radix_bits),
+                d_keys_double_buffer,
+                d_values.Current(),
+                get_final_output(d_values, active_policy.large_segment.radix_bits),
+                d_values_double_buffer,
+                current_begin_offset,
+                current_end_offset,
+                group_sizes.get(),
+                large_and_medium_segments_indices.get(),
+                small_segments_indices.get(),
+                launcher_factory,
+                active_policy.large_segment.block_threads,
+                active_policy.small_segment.block_threads,
+                active_policy.medium_segment.segments_per_block(),
+                active_policy.small_segment.segments_per_block())))
+      {
+        return error;
+      }
+      if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+      {
+        return error;
+      }
+#endif // CUB_RDC_ENABLED
+      return cudaSuccess;
+    };
+
+    NV_IF_TARGET(NV_IS_HOST,
+                 ({
+                   local_segment_index_t h_group_sizes[num_selected_groups];
+                   if (const auto error = CubDebug(launcher_factory.MemcpyAsync(
+                         h_group_sizes,
+                         group_sizes.get(),
+                         num_selected_groups * sizeof(local_segment_index_t),
+                         cudaMemcpyDeviceToHost,
+                         stream)))
+                   {
+                     return error;
+                   }
+
+                   if (const auto error = CubDebug(SyncStream(stream)))
+                   {
+                     return error;
+                   }
+
+                   if (const auto error = detail::segmented_sort::device_segmented_sort_continuation(
+                         large_kernel,
+                         small_kernel,
+                         current_num_segments,
+                         d_keys.Current(),
+                         get_final_output(d_keys, active_policy.large_segment.radix_bits),
+                         d_keys_double_buffer,
+                         d_values.Current(),
+                         get_final_output(d_values, active_policy.large_segment.radix_bits),
+                         d_values_double_buffer,
+                         current_begin_offset,
+                         current_end_offset,
+                         h_group_sizes,
+                         large_and_medium_segments_indices.get(),
+                         small_segments_indices.get(),
+                         stream,
+                         launcher_factory,
+                         active_policy.large_segment.block_threads,
+                         active_policy.small_segment.block_threads,
+                         active_policy.medium_segment.segments_per_block(),
+                         active_policy.small_segment.segments_per_block()))
+                   {
+                     return error;
+                   }
+                 }),
+                 ({
+                   if (const auto error = device_path())
+                   {
+                     return error;
+                   }
+                 }));
+  }
+
+  return cudaSuccess;
+}
+
+template <bool KeysOnly, typename PolicyGetter>
+CUB_RUNTIME_FUNCTION void check_policy(PolicyGetter policy_getter)
+{
+  [[maybe_unused]] CUB_DETAIL_CONSTEXPR_ISH segmented_sort_policy active_policy = policy_getter();
+
+  CUB_DETAIL_STATIC_ISH_ASSERT(active_policy.large_segment.load_modifier != CacheLoadModifier::LOAD_LDG,
+                               "The memory consistency model does not apply to texture accesses");
+  CUB_DETAIL_STATIC_ISH_ASSERT(
+    KeysOnly || active_policy.large_segment.load_algorithm != BLOCK_LOAD_STRIPED
+      || active_policy.medium_segment.load_algorithm != WARP_LOAD_STRIPED
+      || active_policy.small_segment.load_algorithm != WARP_LOAD_STRIPED,
+    "Striped load will make this algorithm unstable");
+  CUB_DETAIL_STATIC_ISH_ASSERT(active_policy.medium_segment.store_algorithm != WARP_STORE_STRIPED
+                                 || active_policy.small_segment.store_algorithm != WARP_STORE_STRIPED,
+                               "Striped stores will produce unsorted results");
+}
+
+template <typename KeyT,
+          typename ValueT,
+          typename FallbackKernelT,
+          typename BeginOffsetIteratorT,
+          typename EndOffsetIteratorT,
+          typename KernelLauncherFactory,
+          typename GetFinalOutputOp>
+CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN _CCCL_FORCEINLINE cudaError_t sort_without_partitioning(
+  FallbackKernelT fallback_kernel,
+  global_segment_offset_t num_segments,
+  BeginOffsetIteratorT d_begin_offsets,
+  EndOffsetIteratorT d_end_offsets,
+  cudaStream_t stream,
+  DoubleBuffer<KeyT>& d_keys,
+  DoubleBuffer<ValueT>& d_values,
+  device_double_buffer<KeyT>& d_keys_double_buffer,
+  device_double_buffer<ValueT>& d_values_double_buffer,
+  KernelLauncherFactory& launcher_factory,
+  const segmented_sort_policy& active_policy,
+  GetFinalOutputOp&& get_final_output)
+{
+  const auto blocks_in_grid   = static_cast<local_segment_index_t>(num_segments);
+  const auto threads_in_block = static_cast<unsigned int>(active_policy.large_segment.block_threads);
+#ifdef CUB_DEBUG_LOG
+  _CubLog("Invoking DeviceSegmentedSortFallbackKernel<<<%d, %d, 0, %lld>>>(), %d items per thread, bit_grain %d\n",
+          blocks_in_grid,
+          threads_in_block,
+          (long long) stream,
+          active_policy.large_segment.items_per_thread,
+          active_policy.large_segment.radix_bits);
+#endif // CUB_DEBUG_LOG
+
+  if (const auto error = CubDebug(
+        launcher_factory(blocks_in_grid, threads_in_block, 0, stream)
+          .doit(fallback_kernel,
+                d_keys.Current(),
+                get_final_output(d_keys, active_policy.large_segment.radix_bits),
+                d_keys_double_buffer,
+                d_values.Current(),
+                get_final_output(d_values, active_policy.large_segment.radix_bits),
+                d_values_double_buffer,
+                d_begin_offsets,
+                d_end_offsets)))
+  {
+    return error;
+  }
+
+  if (const auto error = CubDebug(cudaPeekAtLastError()))
+  {
+    return error;
+  }
+  if (const auto error = CubDebug(detail::DebugSyncStream(stream)))
+  {
+    return error;
+  }
+
+  return cudaSuccess;
+}
+
+template <
+  SortOrder Order,
+  typename OffsetT,
+  typename KeyT,
+  typename ValueT,
+  typename BeginOffsetIteratorT,
+  typename EndOffsetIteratorT,
+  typename PolicySelector = policy_selector_from_types<KeyT, ValueT>,
+  typename KernelSource =
+    DeviceSegmentedSortKernelSource<PolicySelector, Order, KeyT, ValueT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>,
+  typename PartitionPolicyHub = detail::three_way_partition::policy_hub<
+    cub::detail::it_value_t<THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>>,
+    three_way_partition::per_partition_offset_t>,
+  typename PartitionKernelSource = detail::three_way_partition::DeviceThreeWayPartitionKernelSource<
+    typename PartitionPolicyHub::MaxPolicy,
+    THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>,
+    local_segment_index_t*,
+    local_segment_index_t*,
+    ::cuda::std::reverse_iterator<local_segment_index_t*>,
+    local_segment_index_t*,
+    three_way_partition::ScanTileStateT,
+    LargeSegmentsSelectorT<OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>,
+    SmallSegmentsSelectorT<OffsetT, BeginOffsetIteratorT, EndOffsetIteratorT>,
+    three_way_partition::per_partition_offset_t,
+    three_way_partition::streaming_context_t<global_segment_offset_t>,
+    choose_signed_offset<global_segment_offset_t>::type>,
+  typename KernelLauncherFactory = CUB_DETAIL_DEFAULT_KERNEL_LAUNCHER_FACTORY>
+#if _CCCL_HAS_CONCEPTS()
+  requires segmented_sort_policy_selector<PolicySelector>
+#endif // _CCCL_HAS_CONCEPTS()
+CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE auto dispatch(
+  void* d_temp_storage,
+  size_t& temp_storage_bytes,
+  DoubleBuffer<KeyT>& d_keys,
+  DoubleBuffer<ValueT>& d_values,
+  ::cuda::std::int64_t num_items,
+  global_segment_offset_t num_segments,
+  BeginOffsetIteratorT d_begin_offsets,
+  EndOffsetIteratorT d_end_offsets,
+  bool is_overwrite_okay,
+  cudaStream_t stream,
+  PolicySelector policy_selector                              = {},
+  typename PartitionPolicyHub::MaxPolicy partition_max_policy = {},
+  KernelSource kernel_source                                  = {},
+  PartitionKernelSource partition_kernel_source               = {},
+  KernelLauncherFactory launcher_factory                      = {}) -> cudaError_t
+{
+  [[maybe_unused]] static constexpr bool keys_only = ::cuda::std::is_same_v<ValueT, NullType>;
+
+  const auto get_num_passes = [&](int radix_bits) {
+    const int num_bits   = static_cast<int>(kernel_source.KeySize()) * CHAR_BIT;
+    const int num_passes = ::cuda::ceil_div(num_bits, radix_bits);
+    return num_passes;
+  };
+
+  const auto get_final_selector = [&](int selector, int radix_bits) {
+    if (!is_overwrite_okay)
+    {
+      return (selector + 1) & 1;
+    }
+    return (selector + get_num_passes(radix_bits)) & 1;
+  };
+
+  const auto get_final_output = [&](auto& buffer, int radix_bits) {
+    const int final_selector = get_final_selector(buffer.selector, radix_bits);
+    return buffer.d_buffers[final_selector];
+  };
+
+  ::cuda::arch_id arch_id{};
+  if (const auto error = CubDebug(launcher_factory.PtxArchId(arch_id)))
+  {
+    return error;
+  }
+
+  return detail::dispatch_arch(policy_selector, arch_id, [&](auto policy_getter) -> cudaError_t {
+    check_policy<keys_only>(policy_getter); // MSVC fails to evaluate static_asserts inside this lambda, so move them to
+                                            // a function
+    CUB_DETAIL_CONSTEXPR_ISH const segmented_sort_policy active_policy = policy_getter();
+
+#if !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+    NV_IF_TARGET(
+      NV_IS_HOST,
+      (::std::stringstream ss; ss << active_policy;
+       _CubLog("Dispatching DeviceSegmentedSort to arch %d with tuning: %s\n", (int) arch_id, ss.str().c_str());))
+#endif // !_CCCL_COMPILER(NVRTC) && defined(CUB_DEBUG_LOG)
+
+    const int radix_bits = active_policy.large_segment.radix_bits;
+
+    //------------------------------------------------------------------------
+    // Prepare temporary storage layout
+    //------------------------------------------------------------------------
+
+    const bool partition_segments = num_segments > active_policy.partitioning_threshold;
+
+    temporary_storage::layout<5> temporary_storage_layout;
+    auto keys_slot                          = temporary_storage_layout.get_slot(0);
+    auto values_slot                        = temporary_storage_layout.get_slot(1);
+    auto large_and_medium_partitioning_slot = temporary_storage_layout.get_slot(2);
+    auto small_partitioning_slot            = temporary_storage_layout.get_slot(3);
+    auto group_sizes_slot                   = temporary_storage_layout.get_slot(4);
+
+    auto keys_allocation   = keys_slot->create_alias<KeyT>();
+    auto values_allocation = values_slot->create_alias<ValueT>();
+
+    if (!is_overwrite_okay)
+    {
+      keys_allocation.grow(num_items);
+      if (!keys_only)
+      {
+        values_allocation.grow(num_items);
+      }
+    }
+
+    auto large_and_medium_segments_indices = large_and_medium_partitioning_slot->create_alias<local_segment_index_t>();
+    auto small_segments_indices            = small_partitioning_slot->create_alias<local_segment_index_t>();
+    auto group_sizes                       = group_sizes_slot->create_alias<local_segment_index_t>();
+
+    size_t three_way_partition_temp_storage_bytes{};
+
+    auto large_segments_selector = kernel_source.LargeSegmentsSelector(
+      active_policy.medium_segment.items_per_tile(), d_begin_offsets, d_end_offsets);
+    auto small_segments_selector = kernel_source.SmallSegmentsSelector(
+      active_policy.small_segment.items_per_tile() + 1, d_begin_offsets, d_end_offsets);
+
+    auto device_partition_temp_storage = keys_slot->create_alias<uint8_t>();
+    if (partition_segments)
+    {
+      constexpr auto num_segments_per_invocation_limit =
+        static_cast<global_segment_offset_t>(::cuda::std::numeric_limits<int>::max());
+      auto const max_num_segments_per_invocation = static_cast<global_segment_offset_t>(
+        (::cuda::std::min) (static_cast<global_segment_offset_t>(num_segments), num_segments_per_invocation_limit));
+
+      large_and_medium_segments_indices.grow(max_num_segments_per_invocation);
+      small_segments_indices.grow(max_num_segments_per_invocation);
+      group_sizes.grow(num_selected_groups);
+
+      auto medium_indices_iterator = ::cuda::std::make_reverse_iterator(large_and_medium_segments_indices.get());
+
+      // We call partition through dispatch instead of device because c.parallel needs to be able to call the kernel.
+      // This approach propagates the type erasure to partition.
+      using ChooseOffsetT                = choose_signed_offset<global_segment_offset_t>;
+      using PartitionOffsetT             = typename ChooseOffsetT::type;
+      using DispatchThreeWayPartitionIfT = cub::DispatchThreeWayPartitionIf<
+        THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>,
+        decltype(large_and_medium_segments_indices.get()),
+        decltype(small_segments_indices.get()),
+        decltype(medium_indices_iterator),
+        decltype(group_sizes.get()),
+        decltype(large_segments_selector),
+        decltype(small_segments_selector),
+        PartitionOffsetT,
+        PartitionPolicyHub,
+        PartitionKernelSource,
+        KernelLauncherFactory>;
+
+      // Signed integer type for global offsets
+      // Check if the number of items exceeds the range covered by the selected signed offset type
+      if (const auto error = ChooseOffsetT::is_exceeding_offset_type(num_items))
+      {
+        return error;
+      }
+
+      DispatchThreeWayPartitionIfT::Dispatch(
+        nullptr,
+        three_way_partition_temp_storage_bytes,
+        THRUST_NS_QUALIFIER::counting_iterator<local_segment_index_t>(0),
+        large_and_medium_segments_indices.get(),
+        small_segments_indices.get(),
+        medium_indices_iterator,
+        group_sizes.get(),
+        large_segments_selector,
+        small_segments_selector,
+        max_num_segments_per_invocation,
+        stream,
+        partition_kernel_source,
+        launcher_factory,
+        partition_max_policy);
+
+      device_partition_temp_storage.grow(three_way_partition_temp_storage_bytes);
+    }
+
+    if (d_temp_storage == nullptr)
+    {
+      temp_storage_bytes = temporary_storage_layout.get_size();
+
+      // Return if the caller is simply requesting the size of the storage allocation
+      return cudaSuccess;
+    }
+
+    if (num_items == 0 || num_segments == 0)
+    {
+      return cudaSuccess;
+    }
+
+    if (const auto error = CubDebug(temporary_storage_layout.map_to_buffer(d_temp_storage, temp_storage_bytes)))
+    {
+      return error;
+    }
+
+    //------------------------------------------------------------------------
+    // Sort
+    //------------------------------------------------------------------------
+
+    const bool is_num_passes_odd = get_num_passes(radix_bits) & 1;
+
+    // This algorithm sorts segments that don't fit into shared memory with the in-global-memory radix sort. Radix sort
+    // splits key representation into multiple "digits". Each digit is RADIX_BITS wide. The algorithm iterates over
+    // these digits. Each of these iterations consists of a couple of stages. The first stage computes a histogram for a
+    // current digit in each segment key. This histogram helps to determine the starting position of the keys group with
+    // a similar digit.
+    // For example:
+    // keys_digits  = [ 1, 0, 0, 1 ]
+    // digit_prefix = [ 0, 2 ]
+    // The second stage checks the keys again and increments the prefix to determine the final position of the key:
+    //
+    //               expression            |  key  |   idx   |     result
+    // ----------------------------------- | ----- | ------- | --------------
+    // result[prefix[keys[0]]++] = keys[0] |   1   |    2    | [ ?, ?, 1, ? ]
+    // result[prefix[keys[1]]++] = keys[0] |   0   |    0    | [ 0, ?, 1, ? ]
+    // result[prefix[keys[2]]++] = keys[0] |   0   |    1    | [ 0, 0, 1, ? ]
+    // result[prefix[keys[3]]++] = keys[0] |   1   |    3    | [ 0, 0, 1, 1 ]
+    //
+    // If the resulting memory is aliased to the input one, we'll face the following issues:
+    //
+    //     input      |  key  |   idx   |   result/input   |      issue
+    // -------------- | ----- | ------- | ---------------- | ----------------
+    // [ 1, 0, 0, 1 ] |   1   |    2    | [ 1, 0, 1, 1 ]   | overwrite keys[2]
+    // [ 1, 0, 1, 1 ] |   0   |    0    | [ 0, 0, 1, 1 ]   |
+    // [ 0, 0, 1, 1 ] |   1   |    3    | [ 0, 0, 1, 1 ]   | extra key
+    // [ 0, 0, 1, 1 ] |   1   |    4    | [ 0, 0, 1, 1 ] 1 | OOB access
+    //
+    // To avoid these issues, we have to use extra memory. The extra memory holds temporary storage for writing
+    // intermediate results of each stage. Since we iterate over digits in keys, we potentially need: `sizeof(KeyT) *
+    // num_items * cuda::ceil_div(sizeof(KeyT),RADIX_BITS)` auxiliary memory bytes. To reduce the auxiliary memory
+    // storage requirements, the algorithm relies on a double buffer facility. The idea behind it is in swapping
+    // destination and source buffers at each iteration. This way, we can use only two buffers. One of these buffers can
+    // be the final algorithm output destination. Therefore, only one auxiliary array is needed. Depending on the number
+    // of iterations, we can initialize the double buffer so that the algorithm output array will match the double
+    // buffer result one at the final iteration. A user can provide this algorithm with a double buffer straightaway to
+    // further reduce the auxiliary memory requirements. `is_overwrite_okay`
+    // indicates this use case.
+
+    device_double_buffer<KeyT> d_keys_double_buffer(
+      (is_overwrite_okay || is_num_passes_odd) ? d_keys.Alternate() : keys_allocation.get(),
+      (is_overwrite_okay)   ? d_keys.Current()
+      : (is_num_passes_odd) ? keys_allocation.get()
+                            : d_keys.Alternate());
+    device_double_buffer<ValueT> d_values_double_buffer(
+      (is_overwrite_okay || is_num_passes_odd) ? d_values.Alternate() : values_allocation.get(),
+      (is_overwrite_okay)   ? d_values.Current()
+      : (is_num_passes_odd) ? values_allocation.get()
+                            : d_values.Alternate());
+
+    const auto segmented_sort_fallback_kernel = kernel_source.SegmentedSortFallbackKernel();
+    const auto segmented_sort_kernel_small    = kernel_source.SegmentedSortKernelSmall();
+    const auto segmented_sort_kernel_large    = kernel_source.SegmentedSortKernelLarge();
+
+    if (partition_segments)
+    {
+      if (const auto error = sort_with_partitioning<PartitionPolicyHub>(
+            segmented_sort_kernel_large,
+            segmented_sort_kernel_small,
+            num_segments,
+            num_items,
+            d_begin_offsets,
+            d_end_offsets,
+            stream,
+            three_way_partition_temp_storage_bytes,
+            d_keys,
+            d_values,
+            d_keys_double_buffer,
+            d_values_double_buffer,
+            large_segments_selector,
+            small_segments_selector,
+            device_partition_temp_storage,
+            large_and_medium_segments_indices,
+            small_segments_indices,
+            group_sizes,
+            kernel_source,
+            partition_kernel_source,
+            launcher_factory,
+            partition_max_policy,
+            active_policy,
+            get_final_output))
+      {
+        return error;
+      }
+    }
+    else
+    {
+      if (const auto error = sort_without_partitioning(
+            segmented_sort_fallback_kernel,
+            num_segments,
+            d_begin_offsets,
+            d_end_offsets,
+            stream,
+            d_keys,
+            d_values,
+            d_keys_double_buffer,
+            d_values_double_buffer,
+            launcher_factory,
+            active_policy,
+            get_final_output))
+      {
+        return error;
+      }
+    }
+
+    d_keys.selector   = get_final_selector(d_keys.selector, radix_bits);
+    d_values.selector = get_final_selector(d_values.selector, radix_bits);
+    return cudaSuccess;
+  });
+}
+} // namespace detail::segmented_sort
+
 CUB_NAMESPACE_END
diff --git a/cub/cub/device/dispatch/kernels/kernel_segmented_sort.cuh b/cub/cub/device/dispatch/kernels/kernel_segmented_sort.cuh
index 4133c7448b2..7c8fc153698 100644
--- a/cub/cub/device/dispatch/kernels/kernel_segmented_sort.cuh
+++ b/cub/cub/device/dispatch/kernels/kernel_segmented_sort.cuh
@@ -17,6 +17,7 @@
 #include <cub/agent/agent_sub_warp_merge_sort.cuh>
 #include <cub/detail/device_double_buffer.cuh>
 #include <cub/device/dispatch/dispatch_common.cuh>
+#include <cub/device/dispatch/tuning/tuning_segmented_sort.cuh>
 #include <cub/util_device.cuh>
 #include <cub/warp/warp_reduce.cuh>
 
@@ -119,13 +120,16 @@ struct SmallSegmentsSelectorT
  *   considered empty.
  */
 template <SortOrder Order,
-          typename ChainedPolicyT,
+          typename PolicySelector,
           typename KeyT,
           typename ValueT,
           typename BeginOffsetIteratorT,
           typename EndOffsetIteratorT,
           typename OffsetT>
-__launch_bounds__(ChainedPolicyT::ActivePolicy::LargeSegmentPolicy::BLOCK_THREADS)
+#if _CCCL_HAS_CONCEPTS()
+  requires segmented_sort_policy_selector<PolicySelector>
+#endif // _CCCL_HAS_CONCEPTS()
+__launch_bounds__(PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10}).large_segment.block_threads)
   CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortFallbackKernel(
     const KeyT* d_keys_in_orig,
     KeyT* d_keys_out_orig,
@@ -136,9 +140,26 @@ __launch_bounds__(ChainedPolicyT::ActivePolicy::LargeSegmentPolicy::BLOCK_THREAD
     BeginOffsetIteratorT d_begin_offsets,
     EndOffsetIteratorT d_end_offsets)
 {
-  using ActivePolicyT       = typename ChainedPolicyT::ActivePolicy;
-  using LargeSegmentPolicyT = typename ActivePolicyT::LargeSegmentPolicy;
-  using MediumPolicyT       = typename ActivePolicyT::MediumSegmentPolicy;
+  static constexpr segmented_sort_policy active_policy = PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10});
+  static constexpr auto large_policy                   = active_policy.large_segment;
+  using LargeSegmentPolicyT                            = AgentRadixSortDownsweepPolicy<
+                               0,
+                               0,
+                               void,
+                               large_policy.load_algorithm,
+                               large_policy.load_modifier,
+                               large_policy.rank_algorithm,
+                               large_policy.scan_algorithm,
+                               large_policy.radix_bits,
+                               NoScaling<large_policy.block_threads, large_policy.items_per_thread>>;
+  static constexpr auto medium_policy = active_policy.medium_segment;
+  using MediumPolicyT =
+    AgentSubWarpMergeSortPolicy<medium_policy.block_threads,
+                                medium_policy.warp_threads,
+                                medium_policy.items_per_thread,
+                                medium_policy.load_algorithm,
+                                medium_policy.load_modifier,
+                                medium_policy.store_algorithm>;
 
   const auto segment_id = static_cast<local_segment_index_t>(blockIdx.x);
   OffsetT segment_begin = d_begin_offsets[segment_id];
@@ -296,13 +317,16 @@ __launch_bounds__(ChainedPolicyT::ActivePolicy::LargeSegmentPolicy::BLOCK_THREAD
  *   considered empty.
  */
 template <SortOrder Order,
-          typename ChainedPolicyT,
+          typename PolicySelector,
           typename KeyT,
           typename ValueT,
           typename BeginOffsetIteratorT,
           typename EndOffsetIteratorT,
           typename OffsetT>
-__launch_bounds__(ChainedPolicyT::ActivePolicy::SmallSegmentPolicy::BLOCK_THREADS)
+#if _CCCL_HAS_CONCEPTS()
+  requires segmented_sort_policy_selector<PolicySelector>
+#endif // _CCCL_HAS_CONCEPTS()
+__launch_bounds__(PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10}).small_segment.block_threads)
   CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortKernelSmall(
     local_segment_index_t small_segments,
     local_segment_index_t medium_segments,
@@ -321,9 +345,23 @@ __launch_bounds__(ChainedPolicyT::ActivePolicy::SmallSegmentPolicy::BLOCK_THREAD
   const local_segment_index_t tid = threadIdx.x;
   const local_segment_index_t bid = blockIdx.x;
 
-  using ActivePolicyT = typename ChainedPolicyT::ActivePolicy;
-  using SmallPolicyT  = typename ActivePolicyT::SmallSegmentPolicy;
-  using MediumPolicyT = typename ActivePolicyT::MediumSegmentPolicy;
+  static constexpr segmented_sort_policy active_policy = PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10});
+  static constexpr auto small_policy                   = active_policy.small_segment;
+  using SmallPolicyT =
+    AgentSubWarpMergeSortPolicy<small_policy.block_threads,
+                                small_policy.warp_threads,
+                                small_policy.items_per_thread,
+                                small_policy.load_algorithm,
+                                small_policy.load_modifier,
+                                small_policy.store_algorithm>;
+  static constexpr auto medium_policy = active_policy.medium_segment;
+  using MediumPolicyT =
+    AgentSubWarpMergeSortPolicy<medium_policy.block_threads,
+                                medium_policy.warp_threads,
+                                medium_policy.items_per_thread,
+                                medium_policy.load_algorithm,
+                                medium_policy.load_modifier,
+                                medium_policy.store_algorithm>;
 
   constexpr auto threads_per_medium_segment = static_cast<local_segment_index_t>(MediumPolicyT::WARP_THREADS);
   constexpr auto threads_per_small_segment  = static_cast<local_segment_index_t>(SmallPolicyT::WARP_THREADS);
@@ -417,13 +455,16 @@ __launch_bounds__(ChainedPolicyT::ActivePolicy::SmallSegmentPolicy::BLOCK_THREAD
  *   considered empty.
  */
 template <SortOrder Order,
-          typename ChainedPolicyT,
+          typename PolicySelector,
           typename KeyT,
           typename ValueT,
           typename BeginOffsetIteratorT,
           typename EndOffsetIteratorT,
           typename OffsetT>
-__launch_bounds__(ChainedPolicyT::ActivePolicy::LargeSegmentPolicy::BLOCK_THREADS)
+#if _CCCL_HAS_CONCEPTS()
+  requires segmented_sort_policy_selector<PolicySelector>
+#endif // _CCCL_HAS_CONCEPTS()
+__launch_bounds__(PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10}).large_segment.block_threads)
   CUB_DETAIL_KERNEL_ATTRIBUTES void DeviceSegmentedSortKernelLarge(
     const local_segment_index_t* d_segments_indices,
     const KeyT* d_keys_in_orig,
@@ -435,8 +476,19 @@ __launch_bounds__(ChainedPolicyT::ActivePolicy::LargeSegmentPolicy::BLOCK_THREAD
     BeginOffsetIteratorT d_begin_offsets,
     EndOffsetIteratorT d_end_offsets)
 {
-  using ActivePolicyT         = typename ChainedPolicyT::ActivePolicy;
-  using LargeSegmentPolicyT   = typename ActivePolicyT::LargeSegmentPolicy;
+  static constexpr segmented_radix_sort_policy large_policy =
+    PolicySelector{}(::cuda::arch_id{CUB_PTX_ARCH / 10}).large_segment;
+  using LargeSegmentPolicyT = AgentRadixSortDownsweepPolicy<
+    0,
+    0,
+    void,
+    large_policy.load_algorithm,
+    large_policy.load_modifier,
+    large_policy.rank_algorithm,
+    large_policy.scan_algorithm,
+    large_policy.radix_bits,
+    NoScaling<large_policy.block_threads, large_policy.items_per_thread>>;
+
   using local_segment_index_t = local_segment_index_t;
 
   constexpr int small_tile_size = LargeSegmentPolicyT::BLOCK_THREADS * LargeSegmentPolicyT::ITEMS_PER_THREAD;
diff --git a/cub/cub/device/dispatch/tuning/tuning_segmented_sort.cuh b/cub/cub/device/dispatch/tuning/tuning_segmented_sort.cuh
index 31bd67d8795..8f7a73bbb40 100644
--- a/cub/cub/device/dispatch/tuning/tuning_segmented_sort.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_segmented_sort.cuh
@@ -15,12 +15,279 @@
 
 #include <cub/agent/agent_segmented_radix_sort.cuh>
 #include <cub/agent/agent_sub_warp_merge_sort.cuh>
+#include <cub/device/dispatch/tuning/common.cuh>
 #include <cub/util_device.cuh>
 
+#include <cuda/__device/arch_id.h>
+#include <cuda/std/__algorithm/max.h>
+#include <cuda/std/__type_traits/is_same.h>
+#include <cuda/std/concepts>
+
+#if !_CCCL_COMPILER(NVRTC)
+#  include <ostream>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 CUB_NAMESPACE_BEGIN
 
 namespace detail::segmented_sort
 {
+struct segmented_radix_sort_policy
+{
+  int block_threads;
+  int items_per_thread;
+  BlockLoadAlgorithm load_algorithm;
+  CacheLoadModifier load_modifier;
+  RadixRankAlgorithm rank_algorithm;
+  BlockScanAlgorithm scan_algorithm;
+  int radix_bits;
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator==(const segmented_radix_sort_policy& lhs, const segmented_radix_sort_policy& rhs)
+  {
+    return lhs.block_threads == rhs.block_threads && lhs.items_per_thread == rhs.items_per_thread
+        && lhs.load_algorithm == rhs.load_algorithm && lhs.load_modifier == rhs.load_modifier
+        && lhs.rank_algorithm == rhs.rank_algorithm && lhs.scan_algorithm == rhs.scan_algorithm
+        && lhs.radix_bits == rhs.radix_bits;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator!=(const segmented_radix_sort_policy& lhs, const segmented_radix_sort_policy& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+#if !_CCCL_COMPILER(NVRTC)
+  friend ::std::ostream& operator<<(::std::ostream& os, const segmented_radix_sort_policy& p)
+  {
+    return os
+        << "segmented_radix_sort_policy { .block_threads = " << p.block_threads
+        << ", .items_per_thread = " << p.items_per_thread << ", .load_algorithm = " << p.load_algorithm
+        << ", .load_modifier = " << p.load_modifier << ", .rank_algorithm = " << p.rank_algorithm
+        << ", .scan_algorithm = " << p.scan_algorithm << ", .radix_bits = " << p.radix_bits << " }";
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+};
+
+struct sub_warp_merge_sort_policy
+{
+  int block_threads;
+  int warp_threads;
+  int items_per_thread;
+  WarpLoadAlgorithm load_algorithm;
+  WarpStoreAlgorithm store_algorithm;
+  CacheLoadModifier load_modifier;
+
+  [[nodiscard]] _CCCL_API constexpr int segments_per_block() const
+  {
+    return block_threads / warp_threads;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr int items_per_tile() const
+  {
+    return warp_threads * items_per_thread;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator==(const sub_warp_merge_sort_policy& lhs, const sub_warp_merge_sort_policy& rhs)
+  {
+    return lhs.block_threads == rhs.block_threads && lhs.warp_threads == rhs.warp_threads
+        && lhs.items_per_thread == rhs.items_per_thread && lhs.load_algorithm == rhs.load_algorithm
+        && lhs.store_algorithm == rhs.store_algorithm && lhs.load_modifier == rhs.load_modifier;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator!=(const sub_warp_merge_sort_policy& lhs, const sub_warp_merge_sort_policy& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+#if !_CCCL_COMPILER(NVRTC)
+  friend ::std::ostream& operator<<(::std::ostream& os, const sub_warp_merge_sort_policy& p)
+  {
+    return os
+        << "sub_warp_merge_sort_policy { .block_threads = " << p.block_threads << ", .warp_threads = " << p.warp_threads
+        << ", .items_per_thread = " << p.items_per_thread << ", .load_algorithm = " << p.load_algorithm
+        << ", .store_algorithm = " << p.store_algorithm << ", .load_modifier = " << p.load_modifier << " }";
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+};
+
+struct segmented_sort_policy
+{
+  segmented_radix_sort_policy large_segment;
+  sub_warp_merge_sort_policy small_segment;
+  sub_warp_merge_sort_policy medium_segment;
+  int partitioning_threshold;
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator==(const segmented_sort_policy& lhs, const segmented_sort_policy& rhs)
+  {
+    return lhs.large_segment == rhs.large_segment && lhs.small_segment == rhs.small_segment
+        && lhs.medium_segment == rhs.medium_segment && lhs.partitioning_threshold == rhs.partitioning_threshold;
+  }
+
+  [[nodiscard]] _CCCL_API constexpr friend bool
+  operator!=(const segmented_sort_policy& lhs, const segmented_sort_policy& rhs)
+  {
+    return !(lhs == rhs);
+  }
+
+#if !_CCCL_COMPILER(NVRTC)
+  friend ::std::ostream& operator<<(::std::ostream& os, const segmented_sort_policy& p)
+  {
+    return os << "segmented_sort_policy { .large_segment = " << p.large_segment
+              << ", .small_segment = " << p.small_segment << ", .medium_segment = " << p.medium_segment
+              << ", .partitioning_threshold = " << p.partitioning_threshold << " }";
+  }
+#endif // !_CCCL_COMPILER(NVRTC)
+};
+
+#if _CCCL_HAS_CONCEPTS()
+template <typename T>
+concept segmented_sort_policy_selector = policy_selector<T, segmented_sort_policy>;
+#endif // _CCCL_HAS_CONCEPTS()
+
+struct policy_selector
+{
+  int key_size;
+  int value_size;
+  bool keys_only;
+
+  _CCCL_API constexpr auto __dominant_size() const
+  {
+    return ::cuda::std::max(key_size, value_size);
+  }
+
+  _CCCL_API constexpr auto __make_scaled_segmented_radix_sort_policy(
+    int nominal_4B_block_threads,
+    int nominal_4B_items_per_thread,
+    BlockLoadAlgorithm load_algorithm,
+    CacheLoadModifier load_modifier,
+    RadixRankAlgorithm rank_algorithm,
+    BlockScanAlgorithm scan_algorithm,
+    int radix_bits) const
+  {
+    const auto scaled = scale_reg_bound(nominal_4B_block_threads, nominal_4B_items_per_thread, __dominant_size());
+    return segmented_radix_sort_policy{
+      scaled.block_threads,
+      scaled.items_per_thread,
+      load_algorithm,
+      load_modifier,
+      rank_algorithm,
+      scan_algorithm,
+      radix_bits};
+  }
+
+  [[nodiscard]] _CCCL_API constexpr auto operator()(::cuda::arch_id arch) const -> segmented_sort_policy
+  {
+    const auto scale_items = [&](int nominal_4b_items_per_thread) {
+      return nominal_4B_items_to_items(nominal_4b_items_per_thread, __dominant_size());
+    };
+
+    if (arch >= ::cuda::arch_id::sm_86)
+    {
+      const bool large_items = __dominant_size() > 4;
+      const int radix_bits   = key_size > 1 ? 6 : 4;
+      const int small_itp    = scale_items(large_items ? 7 : 9);
+      const int medium_itp   = scale_items(large_items ? 9 : 7);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 23, BLOCK_LOAD_TRANSPOSE, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_WARP_SCANS, radix_bits),
+        sub_warp_merge_sort_policy{
+          256, large_items ? 8 : 2, small_itp, WARP_LOAD_TRANSPOSE, WARP_STORE_DIRECT, LOAD_LDG},
+        sub_warp_merge_sort_policy{256, 16, medium_itp, WARP_LOAD_TRANSPOSE, WARP_STORE_DIRECT, LOAD_LDG},
+        500};
+    }
+
+    if (arch >= ::cuda::arch_id::sm_80)
+    {
+      const int radix_bits = key_size > 1 ? 6 : 4;
+      const int small_itp  = scale_items(9);
+      const int medium_itp = scale_items(keys_only ? 7 : 11);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 23, BLOCK_LOAD_TRANSPOSE, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_WARP_SCANS, radix_bits),
+        sub_warp_merge_sort_policy{
+          256, keys_only ? 4 : 2, small_itp, WARP_LOAD_TRANSPOSE, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        sub_warp_merge_sort_policy{256, 32, medium_itp, WARP_LOAD_TRANSPOSE, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        500};
+    }
+
+    if (arch >= ::cuda::arch_id::sm_70)
+    {
+      const int radix_bits = key_size > 1 ? 6 : 4;
+      const int small_itp  = scale_items(7);
+      const int medium_itp = scale_items(keys_only ? 11 : 7);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 19, BLOCK_LOAD_DIRECT, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_WARP_SCANS, radix_bits),
+        sub_warp_merge_sort_policy{256, keys_only ? 4 : 8, small_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        sub_warp_merge_sort_policy{256, 32, medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        500};
+    }
+
+    if (arch >= ::cuda::arch_id::sm_62)
+    {
+      const int radix_bits       = key_size > 1 ? 5 : 4;
+      const int small_medium_itp = scale_items(9);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 16, BLOCK_LOAD_TRANSPOSE, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_RAKING_MEMOIZE, radix_bits),
+        sub_warp_merge_sort_policy{256, 4, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        sub_warp_merge_sort_policy{256, 32, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        500};
+    }
+
+    if (arch >= ::cuda::arch_id::sm_61)
+    {
+      const int radix_bits       = key_size > 1 ? 6 : 4;
+      const int small_medium_itp = scale_items(9);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 19, BLOCK_LOAD_DIRECT, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_WARP_SCANS, radix_bits),
+        sub_warp_merge_sort_policy{256, 4, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        sub_warp_merge_sort_policy{256, 32, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        500};
+    }
+
+    if (arch >= ::cuda::arch_id::sm_60)
+    {
+      const int radix_bits       = key_size > 1 ? 6 : 4;
+      const int small_medium_itp = scale_items(9);
+      return segmented_sort_policy{
+        __make_scaled_segmented_radix_sort_policy(
+          256, 19, BLOCK_LOAD_TRANSPOSE, LOAD_DEFAULT, RADIX_RANK_MATCH, BLOCK_SCAN_WARP_SCANS, radix_bits),
+        sub_warp_merge_sort_policy{256, 4, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        sub_warp_merge_sort_policy{256, 32, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+        500};
+    }
+
+    // default for SM50
+    const int radix_bits       = key_size > 1 ? 6 : 4;
+    const int small_medium_itp = scale_items(7);
+    return segmented_sort_policy{
+      __make_scaled_segmented_radix_sort_policy(
+        256, 16, BLOCK_LOAD_DIRECT, LOAD_DEFAULT, RADIX_RANK_MEMOIZE, BLOCK_SCAN_RAKING_MEMOIZE, radix_bits),
+      sub_warp_merge_sort_policy{256, 4, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+      sub_warp_merge_sort_policy{256, 32, small_medium_itp, WARP_LOAD_DIRECT, WARP_STORE_DIRECT, LOAD_DEFAULT},
+      300};
+  }
+};
+
+#if _CCCL_HAS_CONCEPTS()
+static_assert(segmented_sort_policy_selector<policy_selector>);
+#endif // _CCCL_HAS_CONCEPTS()
+
+template <typename KeyT, typename ValueT>
+struct policy_selector_from_types
+{
+  [[nodiscard]] _CCCL_API constexpr auto operator()(::cuda::arch_id arch) const -> segmented_sort_policy
+  {
+    return policy_selector{int{sizeof(KeyT)}, int{sizeof(ValueT)}, ::cuda::std::is_same_v<ValueT, NullType>}(arch);
+  }
+};
+
+// TODO(bgruber): remove when we drop the CUB dispatchers in CCCL 4.0
 template <typename PolicyT, typename = void>
 struct SegmentedSortPolicyWrapper : PolicyT
 {
@@ -29,6 +296,7 @@ struct SegmentedSortPolicyWrapper : PolicyT
   {}
 };
 
+// TODO(bgruber): remove when we drop the CUB dispatchers in CCCL 4.0
 template <typename StaticPolicyT>
 struct SegmentedSortPolicyWrapper<StaticPolicyT,
                                   ::cuda::std::void_t<typename StaticPolicyT::LargeSegmentPolicy,
@@ -126,12 +394,14 @@ struct SegmentedSortPolicyWrapper<StaticPolicyT,
 #endif
 };
 
+// TODO(bgruber): remove when we drop the CUB dispatchers in CCCL 4.0
 template <typename PolicyT>
 _CCCL_HOST_DEVICE SegmentedSortPolicyWrapper<PolicyT> MakeSegmentedSortPolicyWrapper(PolicyT policy)
 {
   return SegmentedSortPolicyWrapper<PolicyT>{policy};
 }
 
+// TODO(bgruber): remove when we drop the CUB dispatchers in CCCL 4.0
 template <typename KeyT, typename ValueT>
 struct policy_hub
 {
diff --git a/cub/cub/util_math.cuh b/cub/cub/util_math.cuh
index 2bab1d862f1..888a52e9993 100644
--- a/cub/cub/util_math.cuh
+++ b/cub/cub/util_math.cuh
@@ -19,6 +19,7 @@
 #endif // no system header
 
 #include <cuda/cmath>
+#include <cuda/std/__algorithm/clamp.h>
 #include <cuda/std/__algorithm/max.h>
 #include <cuda/std/__algorithm/min.h>
 #include <cuda/std/__type_traits/integral_constant.h>
@@ -50,12 +51,19 @@ _CCCL_HOST_DEVICE _CCCL_FORCEINLINE OffsetT safe_add_bound_to_max(OffsetT lhs, O
   return lhs + capped_operand_rhs;
 }
 
+constexpr _CCCL_HOST_DEVICE int nominal_4B_items_to_items(int nominal_4b_items_per_thread, int item_size)
+{
+  return ::cuda::std::clamp(nominal_4b_items_per_thread * 4 / item_size, 1, nominal_4b_items_per_thread);
+}
+
 constexpr _CCCL_HOST_DEVICE int nominal_8B_items_to_items(int nominal_8b_items_per_thread, int item_size)
 {
-  return item_size <= 8
-         ? nominal_8b_items_per_thread
-         : (::cuda::std::min) (nominal_8b_items_per_thread,
-                               (::cuda::std::max) (1, ((nominal_8b_items_per_thread * 8) + item_size - 1) / item_size));
+  if (item_size <= 8)
+  {
+    return nominal_8b_items_per_thread;
+  }
+  return ::cuda::std::clamp(
+    ::cuda::ceil_div(nominal_8b_items_per_thread * 8, item_size), 1, nominal_8b_items_per_thread);
 }
 } // namespace detail
 
@@ -68,8 +76,7 @@ constexpr _CCCL_HOST_DEVICE int Nominal4BItemsToItemsCombined(int nominal_4b_ite
 template <typename T>
 constexpr _CCCL_HOST_DEVICE int Nominal4BItemsToItems(int nominal_4b_items_per_thread)
 {
-  return (::cuda::std::min) (nominal_4b_items_per_thread,
-                             (::cuda::std::max) (1, nominal_4b_items_per_thread * 4 / static_cast<int>(sizeof(T))));
+  return detail::nominal_4B_items_to_items(nominal_4b_items_per_thread, int{sizeof(T)});
 }
 
 template <typename ItemT>
diff --git a/cub/cub/warp/warp_load.cuh b/cub/cub/warp/warp_load.cuh
index ed1943e2038..74bc34dcb74 100644
--- a/cub/cub/warp/warp_load.cuh
+++ b/cub/cub/warp/warp_load.cuh
@@ -25,6 +25,10 @@
 #include <cuda/__cmath/pow2.h>
 #include <cuda/__ptx/instructions/get_sreg.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <ostream>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 CUB_NAMESPACE_BEGIN
 
 //! @rst
@@ -106,6 +110,25 @@ enum WarpLoadAlgorithm
   WARP_LOAD_TRANSPOSE
 };
 
+#if !_CCCL_COMPILER(NVRTC)
+inline ::std::ostream& operator<<(::std::ostream& os, WarpLoadAlgorithm algorithm)
+{
+  switch (algorithm)
+  {
+    case WARP_LOAD_DIRECT:
+      return os << "WARP_LOAD_DIRECT";
+    case WARP_LOAD_STRIPED:
+      return os << "WARP_LOAD_STRIPED";
+    case WARP_LOAD_VECTORIZE:
+      return os << "WARP_LOAD_VECTORIZE";
+    case WARP_LOAD_TRANSPOSE:
+      return os << "WARP_LOAD_TRANSPOSE";
+    default:
+      return os << "<unknown WarpLoadAlgorithm: " << static_cast<int>(algorithm) << ">";
+  }
+}
+#endif // !_CCCL_COMPILER(NVRTC) && !_CCCL_DOXYGEN_INVOKED
+
 //! @rst
 //! The WarpLoad class provides :ref:`collective <collective-primitives>` data movement methods for
 //! loading a linear segment of items from memory into a
diff --git a/cub/cub/warp/warp_store.cuh b/cub/cub/warp/warp_store.cuh
index 8f28d970e3a..64ac40a6370 100644
--- a/cub/cub/warp/warp_store.cuh
+++ b/cub/cub/warp/warp_store.cuh
@@ -24,6 +24,10 @@
 #include <cuda/__cmath/pow2.h>
 #include <cuda/__ptx/instructions/get_sreg.h>
 
+#if !_CCCL_COMPILER(NVRTC)
+#  include <ostream>
+#endif // !_CCCL_COMPILER(NVRTC)
+
 CUB_NAMESPACE_BEGIN
 
 //! @rst
@@ -110,6 +114,25 @@ enum WarpStoreAlgorithm
   WARP_STORE_TRANSPOSE
 };
 
+#if !_CCCL_COMPILER(NVRTC)
+inline ::std::ostream& operator<<(::std::ostream& os, WarpStoreAlgorithm algorithm)
+{
+  switch (algorithm)
+  {
+    case WARP_STORE_DIRECT:
+      return os << "WARP_STORE_DIRECT";
+    case WARP_STORE_STRIPED:
+      return os << "WARP_STORE_STRIPED";
+    case WARP_STORE_VECTORIZE:
+      return os << "WARP_STORE_VECTORIZE";
+    case WARP_STORE_TRANSPOSE:
+      return os << "WARP_STORE_TRANSPOSE";
+    default:
+      return os << "<unknown WarpStoreAlgorithm: " << static_cast<int>(algorithm) << ">";
+  }
+}
+#endif // !_CCCL_COMPILER(NVRTC) && !_CCCL_DOXYGEN_INVOKED
+
 //! @rst
 //! The WarpStore class provides :ref:`collective <collective-primitives>`
 //! data movement methods for writing a :ref:`blocked arrangement <flexible-data-arrangement>`
