diff --git a/hopper/flash_attn_interface.py b/hopper/flash_attn_interface.py
index d0f20020b69..a107b665f14 100644
--- a/hopper/flash_attn_interface.py
+++ b/hopper/flash_attn_interface.py
@@ -165,6 +165,7 @@ def forward(
         softcap=0.0,
         deterministic=False,
         num_heads_q=None,
+        sm_margin=0,
     ):
         if softmax_scale is None:
             softmax_scale = qkv.shape[-1] ** (-0.5)
@@ -195,6 +196,7 @@ def forward(
             window_size=window_size,
             attention_chunk=attention_chunk,
             softcap=softcap,
+            sm_margin=sm_margin,
         )
         # ctx.save_for_backward(q, k, v, out_padded, softmax_lse)
         ctx.save_for_backward(q, k, v, out, softmax_lse)
@@ -205,6 +207,7 @@ def forward(
         ctx.softcap = softcap
         ctx.deterministic = deterministic
         ctx.ndim = qkv.dim()
+        ctx.sm_margin = sm_margin
         # return out, softmax_lse
         return out
 
@@ -240,6 +243,7 @@ def backward(ctx, dout, *args):
             ctx.window_size,
             ctx.softcap,
             ctx.deterministic,
+            ctx.sm_margin,
         )
         dqkv = dqkv[..., : dout.shape[-1]]  # We could have padded the head dimension
         return dqkv, None, None, None, None, None, None, None, None, None, None, None
@@ -444,6 +448,7 @@ def flash_attn_qkvpacked_func(
     softcap=0.0,
     deterministic=False,
     num_heads_q=None,
+    sm_margin=0,
 ):
     """dropout_p should be set to 0.0 during evaluation
     If Q, K, V are already stacked into 1 tensor, this function will be faster than
@@ -489,6 +494,7 @@ def flash_attn_qkvpacked_func(
         softcap,
         deterministic,
         num_heads_q,
+        sm_margin,
     )
 
 
