diff --git a/include/flashinfer/comm/trtllm_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
index 1ce278c0bc..d83354dcf5 100644
--- a/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
+++ b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
@@ -8,6 +8,7 @@
 #endif
 
 #include <cuda/std/optional>
+#include <optional>
 #include <tuple>
 #include <type_traits>
 
diff --git a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
index 5d731bbbfc..a75237d7db 100644
--- a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
+++ b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
@@ -8,6 +8,7 @@
 #endif
 
 #include <cuda/std/optional>
+#include <optional>
 #include <tuple>
 #include <type_traits>
 
