diff --git a/libcudacxx/include/cuda/__functional/for_each_canceled.h b/libcudacxx/include/cuda/__functional/for_each_canceled.h
index 773cfccb705..724935cdb65 100644
--- a/libcudacxx/include/cuda/__functional/for_each_canceled.h
+++ b/libcudacxx/include/cuda/__functional/for_each_canceled.h
@@ -34,37 +34,57 @@
 
 _CCCL_BEGIN_NAMESPACE_CUDA_DEVICE
 
-#  if _CCCL_HAS_INT128() && __cccl_ptx_isa >= 870
+#  if __cccl_ptx_isa >= 870
+
+#    if _CCCL_HAS_INT128()
+using _QueryCancelResult = __uint128_t;
+#    else // ^^^ _CCCL_HAS_INT128() ^^^ / vvv !_CCCL_HAS_INT128() vvv
+struct alignas(16) _QueryCancelResult
+{
+  ::cuda::std::uint64_t __lo_;
+  ::cuda::std::uint64_t __hi_;
+};
+#    endif // ^^^ !_CCCL_HAS_INT128() ^^^
 
 template <int _Index>
-[[nodiscard]] _CCCL_DEVICE_API int __cluster_get_dim(__int128 __result) noexcept
+[[nodiscard]] _CCCL_DEVICE_API int __cluster_get_dim(_QueryCancelResult __result) noexcept
 {
-  int __r;
+  unsigned __r;
+
+  asm volatile("{\n\t"
+               ".reg .b128 query_result;");
+#    if _CCCL_HAS_INT128()
+  asm volatile("mov.b128 query_result, %0;" : : "q"(__result));
+#    else // ^^^ _CCCL_HAS_INT128() ^^^ / vvv !_CCCL_HAS_INT128() vvv
+  asm volatile("mov.b128 query_result, {%0, %1};" : : "l"(__result.__lo_), "l"(__result.__hi_));
+#    endif // ^^^ !_CCCL_HAS_INT128() ^^^
+
   if constexpr (_Index == 0)
   {
-    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::x.b32.b128 %0, %1;"
+    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::x.b32.b128 %0, query_result;"
                  : "=r"(__r)
-                 : "q"(__result)
+                 :
                  : "memory");
   }
   else if constexpr (_Index == 1)
   {
-    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::y.b32.b128 %0, %1;"
+    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::y.b32.b128 %0, query_result;"
                  : "=r"(__r)
-                 : "q"(__result)
+                 :
                  : "memory");
   }
   else if constexpr (_Index == 2)
   {
-    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::z.b32.b128 %0, %1;"
+    asm volatile("clusterlaunchcontrol.query_cancel.get_first_ctaid::z.b32.b128 %0, query_result;"
                  : "=r"(__r)
-                 : "q"(__result)
+                 :
                  : "memory");
   }
   else
   {
     _CCCL_UNREACHABLE();
   }
+  asm volatile("}");
   return __r;
 }
 
@@ -83,7 +103,7 @@ template <int __ThreadBlockRank = 3, typename __UnaryFunction = void>
 _CCCL_DEVICE_API void __for_each_canceled_block_sm100(::dim3 __block_idx, bool __is_leader, __UnaryFunction __uf)
 {
   __shared__ ::cuda::std::uint64_t __barrier; // TODO: use 2 barriers and 2 results to avoid last sync threads
-  __shared__ __int128 __result;
+  __shared__ _QueryCancelResult __result;
   bool __phase = false;
 
   // Initialize barrier and kick-start try_cancel pipeline:
@@ -130,14 +150,18 @@ _CCCL_DEVICE_API void __for_each_canceled_block_sm100(::dim3 __block_idx, bool _
     // the next query operations from being re-ordered above the poll loop.
     {
       int __success = 0;
-      asm volatile(
-        "{\n\t"
-        ".reg .pred p;\n\t"
-        "clusterlaunchcontrol.query_cancel.is_canceled.pred.b128 p, %1;\n\t"
-        "selp.b32 %0, 1, 0, p;\n\t"
-        "}\n\t"
-        : "=r"(__success)
-        : "q"(__result));
+      asm volatile("{\n\t"
+                   ".reg .pred p;\t\n"
+                   ".reg .b128 query_result;");
+#    if _CCCL_HAS_INT128()
+      asm volatile("mov.b128 query_result, %0;" : : "q"(__result));
+#    else // ^^^ _CCCL_HAS_INT128() ^^^ / vvv !_CCCL_HAS_INT128() vvv
+      asm volatile("mov.b128 query_result, {%0, %1};" : : "l"(__result.__lo_), "l"(__result.__hi_));
+#    endif // ^^^ !_CCCL_HAS_INT128() ^^^
+      asm volatile("clusterlaunchcontrol.query_cancel.is_canceled.pred.b128 p, query_result;\n\t"
+                   "selp.b32 %0, 1, 0, p;\n\t"
+                   "}\n\t"
+                   : "=r"(__success));
       if (__success != 1)
       {
         // Invalidating mbarrier and synchronizing before exiting not
@@ -188,14 +212,14 @@ _CCCL_DEVICE_API void __for_each_canceled_block_sm100(::dim3 __block_idx, bool _
   } while (true);
 }
 
-#  else // ^^^ _CCCL_HAS_INT128() && __cccl_ptx_isa >= 870 ^^^ / vvv !_CCCL_HAS_INT128() || __cccl_ptx_isa < 870 vvv
+#  else // ^^^ __cccl_ptx_isa >= 870 ^^^ / vvv __cccl_ptx_isa < 870 vvv
 template <int __ThreadBlockRank = 3, typename __UnaryFunction = void>
 _CCCL_DEVICE_API void __for_each_canceled_block_sm100(::dim3 __block_idx, bool __is_leader, __UnaryFunction __uf)
 {
   // We are compiling for SM100 but PTX 8.7 is not supported, so fall back to just calling the function
   ::cuda::std::invoke(::cuda::std::move(__uf), __block_idx);
 }
-#  endif // _CCCL_HAS_INT128() && __cccl_ptx_isa >= 870
+#  endif // ^^^ __cccl_ptx_isa < 870 ^^^
 
 //! This API for implementing work-stealing, repeatedly attempts to cancel the launch of a thread block
 //! from the current grid. On success, it invokes the unary function `__uf` before trying again.
diff --git a/libcudacxx/test/libcudacxx/cuda/work_stealing/for_each_canceled/for_each_canceled.pass.cpp b/libcudacxx/test/libcudacxx/cuda/work_stealing/for_each_canceled/for_each_canceled.pass.cpp
index 05cdd162d60..0ecc4343e0b 100644
--- a/libcudacxx/test/libcudacxx/cuda/work_stealing/for_each_canceled/for_each_canceled.pass.cpp
+++ b/libcudacxx/test/libcudacxx/cuda/work_stealing/for_each_canceled/for_each_canceled.pass.cpp
@@ -8,7 +8,6 @@
 //
 //===----------------------------------------------------------------------===//
 
-// UNSUPPORTED: !nvcc
 // UNSUPPORTED: nvrtc
 
 // ADDITIONAL_COMPILE_DEFINITIONS: CCCL_IGNORE_DEPRECATED_API
@@ -16,8 +15,6 @@
 #include <cuda/std/cmath>
 #include <cuda/work_stealing>
 
-#if _CCCL_HAS_INT128()
-
 __device__ void vec_add_impl1(int* a, int* b, int* c, int n, dim3 block_idx)
 {
   int idx = threadIdx.x + block_idx.x * blockDim.x;
@@ -52,7 +49,7 @@ __device__ void vec_add_impl3(int* a, int* b, int* c, int n, dim3 block_idx)
 
 __global__ void vec_add_det1(int* a, int* b, int* c, int n, int leader_tidx = 0)
 {
-  ::cuda::device::__for_each_canceled_block<1>(threadIdx.x == leader_tidx, [=](dim3 block_idx) {
+  ::cuda::device::__for_each_canceled_block<1>(threadIdx.x == static_cast<unsigned>(leader_tidx), [=](dim3 block_idx) {
     vec_add_impl1(a, b, c, n, block_idx);
   });
 }
@@ -60,7 +57,8 @@ __global__ void vec_add_det1(int* a, int* b, int* c, int n, int leader_tidx = 0)
 __global__ void vec_add_det2(int* a, int* b, int* c, int n, int leader_tidx = 0)
 {
   ::cuda::device::__for_each_canceled_block<2>(
-    threadIdx.x == leader_tidx && threadIdx.y == leader_tidx, [=](dim3 block_idx) {
+    threadIdx.x == static_cast<unsigned>(leader_tidx) && threadIdx.y == static_cast<unsigned>(leader_tidx),
+    [=](dim3 block_idx) {
       vec_add_impl2(a, b, c, n, block_idx);
     });
 }
@@ -68,7 +66,9 @@ __global__ void vec_add_det2(int* a, int* b, int* c, int n, int leader_tidx = 0)
 __global__ void vec_add_det3(int* a, int* b, int* c, int n, int leader_tidx = 0)
 {
   ::cuda::device::__for_each_canceled_block<3>(
-    threadIdx.x == leader_tidx && threadIdx.y == leader_tidx && threadIdx.z == leader_tidx, [=](dim3 block_idx) {
+    threadIdx.x == static_cast<unsigned>(leader_tidx) && threadIdx.y == static_cast<unsigned>(leader_tidx)
+      && threadIdx.z == static_cast<unsigned>(leader_tidx),
+    [=](dim3 block_idx) {
       vec_add_impl3(a, b, c, n, block_idx);
     });
 }
@@ -218,13 +218,8 @@ void test()
   }
 }
 
-#endif // _CCCL_HAS_INT128()
-
 int main(int argc, char** argv)
 {
-#if _CCCL_HAS_INT128()
   NV_IF_TARGET(NV_IS_HOST, (test();))
-#endif // _CCCL_HAS_INT128()
-
   return 0;
 }
