diff --git a/src/backend/cuda/op/gemm.cc b/src/backend/cuda/op/gemm.cc
index abe33333f..5f80d3e19 100644
--- a/src/backend/cuda/op/gemm.cc
+++ b/src/backend/cuda/op/gemm.cc
@@ -297,7 +297,8 @@ struct Gemm {
     if (gemm_inst == kCudaWGMMA) {
       return ComputeWgmmaWarpPartition(policy, M, N, num_warps);
     }
-    int k_n_per_warp = TargetIsVolta(target) ? 16 : 8;
+    int k_n_per_warp =
+        (TargetIsVolta(target) || TargetIsTuring(target)) ? 16 : 8;
     return ComputeDefaultWarpPartition(policy, M, N, num_warps, k_n_per_warp);
   }
 
diff --git a/src/tl_templates/cuda/common.h b/src/tl_templates/cuda/common.h
index eba257309..77ead126d 100644
--- a/src/tl_templates/cuda/common.h
+++ b/src/tl_templates/cuda/common.h
@@ -790,7 +790,11 @@ TL_DEVICE __nv_bfloat162 fma2(__nv_bfloat162 a, __nv_bfloat162 b,
 #if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 800)
   return __hfma2(a, b, c);
 #else
-  return __nv_bfloat162{__hfma(a.x, b.x, c.x), __hfma(a.y, b.y, c.y)};
+  float a_x = __bfloat162float(a.x), a_y = __bfloat162float(a.y);
+  float b_x = __bfloat162float(b.x), b_y = __bfloat162float(b.y);
+  float c_x = __bfloat162float(c.x), c_y = __bfloat162float(c.y);
+  return __nv_bfloat162{__float2bfloat16(a_x * b_x + c_x),
+                        __float2bfloat16(a_y * b_y + c_y)};
 #endif
 }
 
diff --git a/src/tl_templates/cuda/gemm_mma.h b/src/tl_templates/cuda/gemm_mma.h
index b73369ecf..1b8bfdc25 100644
--- a/src/tl_templates/cuda/gemm_mma.h
+++ b/src/tl_templates/cuda/gemm_mma.h
@@ -105,7 +105,6 @@ TL_DISPATCH_MMA(int8_t, int8_t, int, SM80_16x8x32_S32S8S8S32_TN)
 TL_DISPATCH_MMA(double, double, double, SM80_8x8x4_F64F64F64F64_TN)
 #elif __CUDA_ARCH_LIST__ >= 750
 #include <cute/arch/mma_sm75.hpp>
-TL_DISPATCH_MMA(half_t, half_t, half_t, SM75_16x8x8_F16F16F16F16_TN)
 TL_DISPATCH_MMA(half_t, half_t, float, SM75_16x8x8_F32F16F16F32_TN)
 #endif
 #endif
diff --git a/tilelang/cuda/op/gemm/__init__.py b/tilelang/cuda/op/gemm/__init__.py
index f78878b54..8cd0da401 100644
--- a/tilelang/cuda/op/gemm/__init__.py
+++ b/tilelang/cuda/op/gemm/__init__.py
@@ -7,15 +7,15 @@
 from .gemm_mma_sm70 import GemmMMASm70
 from .gemm_tcgen05 import GEMM_INST_TCGEN05, GemmTCGEN5
 from .gemm_wgmma import GEMM_INST_WGMMA, GemmWGMMA
-from tilelang.utils.target import target_is_cuda, target_is_volta
+from tilelang.utils.target import target_is_cuda, target_is_turing, target_is_volta
 
 
 def _match_mma(target) -> bool:
-    return target_is_cuda(target) and not target_is_volta(target)
+    return target_is_cuda(target) and not (target_is_volta(target) or target_is_turing(target))
 
 
 def _match_mma_sm70(target) -> bool:
-    return target_is_volta(target)
+    return target_is_volta(target) or target_is_turing(target)
 
 
 def _match_wgmma(target) -> bool:
