diff --git a/csrc/trtllm_alltoall.cu b/csrc/trtllm_alltoall.cu
index 9ad4fff110..005fe10f1e 100644
--- a/csrc/trtllm_alltoall.cu
+++ b/csrc/trtllm_alltoall.cu
@@ -296,7 +296,7 @@ void moePrepareOp(TensorView expertsIds, Optional<TensorView> scales,
     CHECK_INPUT_TYPE(scales.value(), dl_float32);
     scalesPtr = static_cast<float*>(scales.value().data_ptr());
     CHECK_DEVICE(preparedLocalScales.value(), expertsIds);
-    CHECK_INPUT_TYPE(preparedLocalScales.value(), dl_int32);
+    CHECK_INPUT_TYPE(preparedLocalScales.value(), dl_float32);
     TVM_FFI_ICHECK_EQ(preparedLocalScales.value().ndim(), 2);
     TVM_FFI_ICHECK_EQ(preparedLocalScales.value().size(0), maxTokenCountPerRank * epSize);
     TVM_FFI_ICHECK_EQ(preparedLocalScales.value().size(1), topK);
diff --git a/flashinfer/comm/trtllm_alltoall.py b/flashinfer/comm/trtllm_alltoall.py
index 114b85b30c..b6a57f41c2 100644
--- a/flashinfer/comm/trtllm_alltoall.py
+++ b/flashinfer/comm/trtllm_alltoall.py
@@ -236,7 +236,9 @@ def moe_prepare(
         )
         if scales is not None:
             prepared_local_scales = torch.empty(
-                (max_token_count_per_rank * ep_size, top_k), **attrs
+                (max_token_count_per_rank * ep_size, top_k),
+                dtype=torch.float32,
+                device=attrs["device"],
             )
         else:
             prepared_local_scales = None
