diff --git a/cub/benchmarks/bench/transform/babelstream.cu b/cub/benchmarks/bench/transform/babelstream.cu
index 36fd0b9696f..bff37527b39 100644
--- a/cub/benchmarks/bench/transform/babelstream.cu
+++ b/cub/benchmarks/bench/transform/babelstream.cu
@@ -32,7 +32,7 @@ using element_types =
                      std::int16_t,
                      float,
                      double
-#  ifdef _CCCL_HAS_INT128()
+#  if _CCCL_HAS_INT128()
                      ,
                      __int128
 #  endif
diff --git a/cub/benchmarks/bench/transform/complex_cmp.cu b/cub/benchmarks/bench/transform/complex_cmp.cu
index fb647dfa028..df724a0a18c 100644
--- a/cub/benchmarks/bench/transform/complex_cmp.cu
+++ b/cub/benchmarks/bench/transform/complex_cmp.cu
@@ -37,11 +37,11 @@ try
     return;
   }
 
-  thrust::device_vector<complex> in = generate(n);
+  thrust::device_vector<complex32> in = generate(n);
   thrust::device_vector<bool> out(n - 1);
 
   state.add_element_count(n);
-  state.add_global_memory_reads<complex>(n);
+  state.add_global_memory_reads<complex32>(n);
   state.add_global_memory_writes<bool>(n);
 
   // the complex comparison needs lots of compute and transform reads from overlapping input
diff --git a/nvbench_helper/nvbench_helper/nvbench_helper.cu b/nvbench_helper/nvbench_helper/nvbench_helper.cu
index 66b739ca220..b9a7a6b9343 100644
--- a/nvbench_helper/nvbench_helper/nvbench_helper.cu
+++ b/nvbench_helper/nvbench_helper/nvbench_helper.cu
@@ -191,30 +191,29 @@ struct and_t
     return cuda::std::bit_cast<double>(result);
   }
 
-  __host__ __device__ complex operator()(complex a, complex b) const
+  template <typename T>
+  __host__ __device__ cuda::std::complex<T> operator()(cuda::std::complex<T> a, cuda::std::complex<T> b) const
   {
-    double a_real = a.real();
-    double a_imag = a.imag();
-
-    double b_real = b.real();
-    double b_imag = b.imag();
+    const T a_real = a.real();
+    const T a_imag = a.imag();
 
-    const std::uint64_t result_real =
-      cuda::std::bit_cast<std::uint64_t>(a_real) & cuda::std::bit_cast<std::uint64_t>(b_real);
+    const T b_real = b.real();
+    const T b_imag = b.imag();
 
-    const std::uint64_t result_imag =
-      cuda::std::bit_cast<std::uint64_t>(a_imag) & cuda::std::bit_cast<std::uint64_t>(b_imag);
+    using uint_t           = std::conditional_t<sizeof(T) == 4, std::uint32_t, std::uint64_t>;
+    const auto result_real = cuda::std::bit_cast<uint_t>(a_real) & cuda::std::bit_cast<uint_t>(b_real);
+    const auto result_imag = cuda::std::bit_cast<uint_t>(a_imag) & cuda::std::bit_cast<uint_t>(b_imag);
 
-    return {static_cast<float>(cuda::std::bit_cast<double>(result_real)),
-            static_cast<float>(cuda::std::bit_cast<double>(result_imag))};
+    return {cuda::std::bit_cast<T>(result_real), cuda::std::bit_cast<T>(result_imag)};
   }
 };
 
+template <typename T>
 struct set_real_t
 {
-  complex m_min{};
-  complex m_max{};
-  complex* m_d_in{};
+  cuda::std::complex<T> m_min{};
+  cuda::std::complex<T> m_max{};
+  cuda::std::complex<T>* m_d_in{};
   const double* m_d_tmp{};
 
   __host__ __device__ void operator()(std::size_t i) const
@@ -223,11 +222,12 @@ struct set_real_t
   }
 };
 
+template <typename T>
 struct set_imag_t
 {
-  complex m_min{};
-  complex m_max{};
-  complex* m_d_in{};
+  cuda::std::complex<T> m_min{};
+  cuda::std::complex<T> m_max{};
+  cuda::std::complex<T>* m_d_in{};
   const double* m_d_tmp{};
 
   __host__ __device__ void operator()(std::size_t i) const
@@ -303,14 +303,14 @@ private:
   template <typename ExecT, typename DistT, typename T>
   void generate(const ExecT& exec, DistT& dist, seed_t seed, cuda::std::span<T> span, bit_entropy entropy, T min, T max);
 
-  template <typename ExecT, typename DistT>
+  template <typename ExecT, typename DistT, typename T>
   void generate(const ExecT& exec,
                 DistT& dist,
                 seed_t seed,
-                cuda::std::span<complex> span,
+                cuda::std::span<cuda::std::complex<T>> span,
                 bit_entropy entropy,
-                complex min,
-                complex max);
+                cuda::std::complex<T> min,
+                cuda::std::complex<T> max);
 
   template <typename ExecT, typename DistT>
   void generate(
@@ -384,15 +384,15 @@ void generator_t::generate(
   };
 }
 
-template <typename ExecT, typename DistT>
+template <typename ExecT, typename DistT, typename T>
 void generator_t::generate(
   const ExecT& exec,
   DistT& dist,
   seed_t seed,
-  cuda::std::span<complex> span,
+  cuda::std::span<cuda::std::complex<T>> span,
   bit_entropy entropy,
-  complex min,
-  complex max)
+  cuda::std::complex<T> min,
+  cuda::std::complex<T> max)
 {
   switch (entropy)
   {
@@ -401,14 +401,14 @@ void generator_t::generate(
       thrust::for_each_n(exec,
                          thrust::make_counting_iterator(std::size_t{0}),
                          span.size(),
-                         set_real_t{min, max, span.data(), uniform_distribution});
+                         set_real_t<T>{min, max, span.data(), uniform_distribution});
       ++seed;
 
       uniform_distribution = dist.new_uniform_distribution(seed, span.size());
       thrust::for_each_n(exec,
                          thrust::make_counting_iterator(std::size_t{0}),
                          span.size(),
-                         set_imag_t{min, max, span.data(), uniform_distribution});
+                         set_imag_t<T>{min, max, span.data(), uniform_distribution});
       ++seed;
       return;
     }
@@ -418,7 +418,7 @@ void generator_t::generate(
       std::uniform_real_distribution<double> dist(0.0f, 1.0f);
       const float random_imag = random_to_item_t<double>(min.imag(), max.imag())(dist(rng));
       const float random_real = random_to_item_t<double>(min.imag(), max.imag())(dist(rng));
-      thrust::fill(exec, span.data(), span.data() + span.size(), complex{random_real, random_imag});
+      thrust::fill(exec, span.data(), span.data() + span.size(), cuda::std::complex<T>{random_real, random_imag});
       return;
     }
     default: {
@@ -426,23 +426,25 @@ void generator_t::generate(
       thrust::for_each_n(exec,
                          thrust::make_counting_iterator(std::size_t{0}),
                          span.size(),
-                         set_real_t{min, max, span.data(), uniform_distribution});
+                         set_real_t<T>{min, max, span.data(), uniform_distribution});
       ++seed;
 
       uniform_distribution = dist.new_uniform_distribution(seed, span.size());
       thrust::for_each_n(exec,
                          thrust::make_counting_iterator(std::size_t{0}),
                          span.size(),
-                         set_imag_t{min, max, span.data(), uniform_distribution});
+                         set_imag_t<T>{min, max, span.data(), uniform_distribution});
       ++seed;
 
       const int number_of_steps = static_cast<int>(entropy);
 
       constexpr bool is_device = std::is_same_v<DistT, device_generator_t>;
-      using vec_t = std::conditional_t<is_device, thrust::device_vector<complex>, thrust::host_vector<complex>>;
+      using vec_t              = std::conditional_t<is_device,
+                                                    thrust::device_vector<cuda::std::complex<T>>,
+                                                    thrust::host_vector<cuda::std::complex<T>>>;
 
       vec_t tmp_vec(span.size());
-      cuda::std::span<complex> tmp(thrust::raw_pointer_cast(tmp_vec.data()), tmp_vec.size());
+      cuda::std::span<cuda::std::complex<T>> tmp(thrust::raw_pointer_cast(tmp_vec.data()), tmp_vec.size());
 
       for (int i = 0; i < number_of_steps; i++, ++seed)
       {
@@ -807,5 +809,6 @@ INSTANTIATE(uint128_t);
 
 INSTANTIATE(float);
 INSTANTIATE(double);
-INSTANTIATE(complex);
+INSTANTIATE(complex32);
+INSTANTIATE(complex64);
 #undef INSTANTIATE
diff --git a/nvbench_helper/nvbench_helper/nvbench_helper.cuh b/nvbench_helper/nvbench_helper/nvbench_helper.cuh
index 1feda8f16ff..2446c366476 100644
--- a/nvbench_helper/nvbench_helper/nvbench_helper.cuh
+++ b/nvbench_helper/nvbench_helper/nvbench_helper.cuh
@@ -31,7 +31,8 @@ NVBENCH_DECLARE_TYPE_STRINGS(int128_t, "I128", "int128_t");
 NVBENCH_DECLARE_TYPE_STRINGS(uint128_t, "U128", "uint128_t");
 #endif
 
-using complex = cuda::std::complex<float>;
+using complex32 = cuda::std::complex<float>;
+using complex64 = cuda::std::complex<double>;
 
 #if _CCCL_HAS_NVFP16()
 NVBENCH_DECLARE_TYPE_STRINGS(__half, "F16", "half");
@@ -41,8 +42,8 @@ NVBENCH_DECLARE_TYPE_STRINGS(cuda::std::complex<__half>, "C16", "complex_half");
 NVBENCH_DECLARE_TYPE_STRINGS(__nv_bfloat16, "BF16", "bfloat16");
 NVBENCH_DECLARE_TYPE_STRINGS(cuda::std::complex<__nv_bfloat16>, "CB16", "complex_bfloat16");
 #endif
-NVBENCH_DECLARE_TYPE_STRINGS(complex, "C32", "complex32");
-NVBENCH_DECLARE_TYPE_STRINGS(cuda::std::complex<double>, "C64", "complex64");
+NVBENCH_DECLARE_TYPE_STRINGS(complex32, "C32", "complex32");
+NVBENCH_DECLARE_TYPE_STRINGS(complex64, "C64", "complex64");
 
 NVBENCH_DECLARE_TYPE_STRINGS(::cuda::std::false_type, "false", "false_type");
 NVBENCH_DECLARE_TYPE_STRINGS(::cuda::std::true_type, "true", "true_type");
@@ -113,7 +114,7 @@ using all_types =
 #  endif
                      float,
                      double,
-                     complex>;
+                     complex32>;
 #endif
 
 template <class T>
@@ -302,12 +303,13 @@ struct vector_generator_t<void> : generator_base_t
 
   // This overload is needed because numeric limits is not specialized for complex, making
   // the min and max values for complex equal zero.
-  operator thrust::device_vector<complex>()
+  template <typename T>
+  operator thrust::device_vector<::cuda::std::complex<T>>()
   {
-    const complex min = complex{
-      ::cuda::std::numeric_limits<complex::value_type>::min(), ::cuda::std::numeric_limits<complex::value_type>::min()};
-    const complex max = complex{
-      ::cuda::std::numeric_limits<complex::value_type>::max(), ::cuda::std::numeric_limits<complex::value_type>::max()};
+    const auto min =
+      ::cuda::std::complex<T>{::cuda::std::numeric_limits<T>::min(), ::cuda::std::numeric_limits<T>::min()};
+    const auto max =
+      ::cuda::std::complex<T>{::cuda::std::numeric_limits<T>::max(), ::cuda::std::numeric_limits<T>::max()};
 
     return generator_base_t::generate(min, max);
   }
@@ -457,7 +459,9 @@ struct less_t
     return lhs < rhs;
   }
 
-  __host__ __device__ inline bool operator()(const complex& lhs, const complex& rhs) const
+  template <typename T>
+  __host__ __device__ inline bool
+  operator()(const ::cuda::std::complex<T>& lhs, const ::cuda::std::complex<T>& rhs) const
   {
     double magnitude_0 = cuda::std::abs(lhs);
     double magnitude_1 = cuda::std::abs(rhs);
@@ -478,20 +482,20 @@ struct less_t
       // std::abs(z) == inf;
       // ```
       // Dividing both components by a constant before computing the magnitude prevents overflow.
-      const complex::value_type scaler = 0.5;
+      const T scaler = 0.5;
 
       magnitude_0 = cuda::std::abs(lhs * scaler);
       magnitude_1 = cuda::std::abs(rhs * scaler);
     }
 
-    const complex::value_type difference = cuda::std::abs(magnitude_0 - magnitude_1);
-    const complex::value_type threshold  = ::cuda::std::numeric_limits<complex::value_type>::epsilon() * 2;
+    const T difference = cuda::std::abs(magnitude_0 - magnitude_1);
+    const T threshold  = ::cuda::std::numeric_limits<T>::epsilon() * 2;
 
     if (difference < threshold)
     {
       // Triangles with the same magnitude are sorted by their phase angle.
-      const complex::value_type phase_angle_0 = cuda::std::arg(lhs);
-      const complex::value_type phase_angle_1 = cuda::std::arg(rhs);
+      const T phase_angle_0 = cuda::std::arg(lhs);
+      const T phase_angle_1 = cuda::std::arg(rhs);
 
       return phase_angle_0 < phase_angle_1;
     }
diff --git a/nvbench_helper/test/gen_seed.cu b/nvbench_helper/test/gen_seed.cu
index 29822a87ae4..b5c0474d3b2 100644
--- a/nvbench_helper/test/gen_seed.cu
+++ b/nvbench_helper/test/gen_seed.cu
@@ -8,18 +8,19 @@
 
 #include <catch2/catch_template_test_macros.hpp>
 
-using types =
-  nvbench::type_list<bool,
-                     int8_t,
-                     int16_t,
-                     int32_t,
-                     int64_t,
+using types = nvbench::type_list<
+  bool,
+  int8_t,
+  int16_t,
+  int32_t,
+  int64_t,
 #if _CCCL_HAS_INT128()
-                     int128_t,
+  int128_t,
 #endif
-                     float,
-                     double,
-                     complex>;
+  float,
+  double,
+  complex32,
+  complex64>;
 
 TEMPLATE_LIST_TEST_CASE("Generator seeds the data", "[gen]", types)
 {
diff --git a/nvbench_helper/test/gen_uniform_distribution.cu b/nvbench_helper/test/gen_uniform_distribution.cu
index 287db0d1496..59185256eb3 100644
--- a/nvbench_helper/test/gen_uniform_distribution.cu
+++ b/nvbench_helper/test/gen_uniform_distribution.cu
@@ -75,7 +75,8 @@ TEMPLATE_LIST_TEST_CASE("Generators produce uniformly distributed data", "[gen][
 
 struct complex_to_real_t
 {
-  __host__ __device__ float operator()(const complex& c) const
+  template <typename T>
+  __host__ __device__ T operator()(const cuda::std::complex<T>& c) const
   {
     return c.real();
   }
@@ -83,25 +84,29 @@ struct complex_to_real_t
 
 struct complex_to_imag_t
 {
-  __host__ __device__ float operator()(const complex& c) const
+  template <typename T>
+  __host__ __device__ T operator()(const cuda::std::complex<T>& c) const
   {
     return c.imag();
   }
 };
 
-TEST_CASE("Generators produce uniformly distributed complex", "[gen]")
+using complex_value_types = nvbench::type_list<float, double>;
+
+TEMPLATE_LIST_TEST_CASE("Generators produce uniformly distributed complex", "[gen]", complex_value_types)
 {
-  const float min = ::cuda::std::numeric_limits<float>::min();
-  const float max = ::cuda::std::numeric_limits<float>::max();
+  using value_type = TestType;
+  const auto min   = ::cuda::std::numeric_limits<value_type>::min();
+  const auto max   = ::cuda::std::numeric_limits<value_type>::max();
 
-  const thrust::device_vector<complex> data = generate(1 << 16);
+  const thrust::device_vector<cuda::std::complex<value_type>> data = generate(1 << 16);
 
-  thrust::device_vector<float> component(data.size());
+  thrust::device_vector<value_type> component(data.size());
   thrust::transform(data.begin(), data.end(), component.begin(), complex_to_real_t());
-  REQUIRE(is_uniform<float>(component, min, max));
+  REQUIRE(is_uniform<value_type>(component, min, max));
 
   thrust::transform(data.begin(), data.end(), component.begin(), complex_to_imag_t());
-  REQUIRE(is_uniform<float>(component, min, max));
+  REQUIRE(is_uniform<value_type>(component, min, max));
 }
 
 TEST_CASE("Generators produce uniformly distributed bools", "[gen]")
