diff --git a/cub/benchmarks/bench/transform/babelstream.cu b/cub/benchmarks/bench/transform/babelstream.cu
index ee0783c1c7f..700b3dd34b3 100644
--- a/cub/benchmarks/bench/transform/babelstream.cu
+++ b/cub/benchmarks/bench/transform/babelstream.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/common.h b/cub/benchmarks/bench/transform/common.h
index a1c3e2fc0d0..0260aa884be 100644
--- a/cub/benchmarks/bench/transform/common.h
+++ b/cub/benchmarks/bench/transform/common.h
@@ -32,30 +32,41 @@ struct policy_selector
   {
     const int min_bytes_in_flight =
       cub::detail::transform::arch_to_min_bytes_in_flight(::cuda::arch_id{__CUDA_ARCH_LIST__ / 10}) + TUNE_BIF_BIAS;
-#  if TUNE_ALGORITHM == 0
-    constexpr auto algorithm = cub::detail::transform::Algorithm::prefetch;
-    constexpr auto policy    = cub::detail::transform::prefetch_policy{
-      TUNE_THREADS
+#  if TUNE_ALGORITHM == 0 || TUNE_ALGORITHM == 1
+    // setup prefetch, since it's either used directly or the fallback to vectorized
+    auto algorithm            = cub::detail::transform::Algorithm::prefetch;
+    auto pref_policy          = cub::detail::transform::prefetch_policy{};
+    pref_policy.block_threads = TUNE_THREADS;
+    pref_policy.unroll_factor = TUNE_UNROLL_FACTOR;
+#    ifdef TUNE_PREFETCH_MULT
+    pref_policy.prefetch_byte_stride = 32 * TUNE_PREFETCH_MULT;
+#    endif //  TUNE_PREFETCH_MULT
 #    ifdef TUNE_ITEMS_PER_THREAD_NO_INPUT
-      ,
-      TUNE_ITEMS_PER_THREAD_NO_INPUT
+    pref_policy.items_per_thread_no_input = TUNE_ITEMS_PER_THREAD_NO_INPUT;
 #    endif // TUNE_ITEMS_PER_THREAD_NO_INPUT
-    };
-    return {min_bytes_in_flight, algorithm, policy, {}, {}};
-#  elif TUNE_ALGORITHM == 1
-    constexpr auto algorithm = cub::detail::transform::Algorithm::vectorized;
-    constexpr auto policy    = cub::detail::transform::vectorized_policy{
-      TUNE_THREADS, (1 << TUNE_VEC_SIZE_POW2) * TUNE_VECTORS_PER_THREAD, (1 << TUNE_VEC_SIZE_POW2)};
-    return {min_bytes_in_flight, algorithm, {}, policy, {}};
+
+    // setup vectorized if requested
+    auto vec_policy = cub::detail::transform::vectorized_policy{};
+#    if TUNE_ALGORITHM == 1
+    algorithm                   = cub::detail::transform::Algorithm::vectorized;
+    vec_policy.block_threads    = TUNE_THREADS;
+    vec_policy.vec_size         = (1 << TUNE_VEC_SIZE_POW2);
+    vec_policy.items_per_thread = policy.vec_size * TUNE_UNROLL_FACTOR;
+#    endif
+    return {min_bytes_in_flight, algorithm, pref_policy, vec_policy, {}};
 #  elif TUNE_ALGORITHM == 2
-    constexpr auto algorithm = cub::detail::transform::Algorithm::memcpy_async;
-    constexpr auto policy =
-      cub::detail::transform::async_copy_policy{TUNE_THREADS, cub::detail::transform::ldgsts_size_and_align};
+    constexpr auto algorithm   = cub::detail::transform::Algorithm::memcpy_async;
+    auto policy                = cub::detail::transform::async_copy_policy{};
+    policy.block_threads       = TUNE_THREADS;
+    policy.bulk_copy_alignment = cub::detail::transform::ldgsts_size_and_align;
+    policy.unroll_factor       = TUNE_UNROLL_FACTOR;
     return {min_bytes_in_flight, algorithm, {}, {}, policy};
 #  elif TUNE_ALGORITHM == 3
-    constexpr auto algorithm = cub::detail::transform::Algorithm::ublkcp;
-    constexpr auto policy    = cub::detail::transform::async_copy_policy{
-      TUNE_THREADS, cub::detail::transform::bulk_copy_alignment(::cuda::arch_id{__CUDA_ARCH_LIST__ / 10})};
+    constexpr auto algorithm   = cub::detail::transform::Algorithm::ublkcp;
+    auto policy                = cub::detail::transform::async_copy_policy{};
+    policy.block_threads       = TUNE_THREADS;
+    policy.bulk_copy_alignment = cub::detail::transform::bulk_copy_alignment(::cuda::arch_id{__CUDA_ARCH_LIST__ / 10});
+    policy.unroll_factor       = TUNE_UNROLL_FACTOR;
     return {min_bytes_in_flight, algorithm, {}, {}, policy};
 #  else // TUNE_ALGORITHM
 #    error Policy hub does not yet implement the specified value for algorithm
diff --git a/cub/benchmarks/bench/transform/complex_cmp.cu b/cub/benchmarks/bench/transform/complex_cmp.cu
index 4ce389763b2..fb647dfa028 100644
--- a/cub/benchmarks/bench/transform/complex_cmp.cu
+++ b/cub/benchmarks/bench/transform/complex_cmp.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/fib.cu b/cub/benchmarks/bench/transform/fib.cu
index 27e32b1bb32..b698e874550 100644
--- a/cub/benchmarks/bench/transform/fib.cu
+++ b/cub/benchmarks/bench/transform/fib.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/fill.cu b/cub/benchmarks/bench/transform/fill.cu
index a43010144a3..408178a4b3d 100644
--- a/cub/benchmarks/bench/transform/fill.cu
+++ b/cub/benchmarks/bench/transform/fill.cu
@@ -6,20 +6,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:2:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
 // those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
 // %RANGE% TUNE_ITEMS_PER_THREAD_NO_INPUT ipt 1:32:1
 
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
 #if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_ITEMS_PER_THREAD_NO_INPUT != 1)
 #  error "Non-prefetch algorithms require the no input items per thread to be 1 since they ignore the parameters"
 #endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/grayscale.cu b/cub/benchmarks/bench/transform/grayscale.cu
index 62cb186346e..d90a6c69819 100644
--- a/cub/benchmarks/bench/transform/grayscale.cu
+++ b/cub/benchmarks/bench/transform/grayscale.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/heavy.cu b/cub/benchmarks/bench/transform/heavy.cu
index f86070e7e6e..9d96c94a611 100644
--- a/cub/benchmarks/bench/transform/heavy.cu
+++ b/cub/benchmarks/bench/transform/heavy.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/benchmarks/bench/transform/pytorch.cu b/cub/benchmarks/bench/transform/pytorch.cu
index 6285a5a4a70..109cbe59ed0 100644
--- a/cub/benchmarks/bench/transform/pytorch.cu
+++ b/cub/benchmarks/bench/transform/pytorch.cu
@@ -5,13 +5,22 @@
 // %RANGE% TUNE_ALGORITHM alg 0:4:1
 // %RANGE% TUNE_THREADS tpb 128:1024:128
 
+// for TUNE_ALGORITHM == 1 (vectorized), this is the number of vectors per thread, which is similar in spirit
+// %RANGE% TUNE_UNROLL_FACTOR unrl 1:4:1
+
+// those parameters only apply if TUNE_ALGORITHM == 0 (prefetch)
+// %RANGE% TUNE_PREFETCH_MULT pref 1:3:1
+
 // those parameters only apply if TUNE_ALGORITHM == 1 (vectorized)
 // %RANGE% TUNE_VEC_SIZE_POW2 vsp2 1:6:1
-// %RANGE% TUNE_VECTORS_PER_THREAD vpt 1:4:1
 
-#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
-#  error "Non-vectorized algorithms require vector size and vectors per thread to be 1 since they ignore the parameters"
-#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1 || TUNE_VECTORS_PER_THREAD != 1)
+#if !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+#  error "Non-prefetch algorithms require prefetch multiple to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 0 && (TUNE_PREFETCH_MULT != 1)
+
+#if !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
+#  error "Non-vectorized algorithms require vector size to be 1 since they ignore the parameters"
+#endif // !TUNE_BASE && TUNE_ALGORITHM != 1 && (TUNE_VEC_SIZE_POW2 != 1)
 
 #include "common.h"
 
diff --git a/cub/cub/device/dispatch/kernels/kernel_transform.cuh b/cub/cub/device/dispatch/kernels/kernel_transform.cuh
index 7db552cbfee..c2f46a2bff3 100644
--- a/cub/cub/device/dispatch/kernels/kernel_transform.cuh
+++ b/cub/cub/device/dispatch/kernels/kernel_transform.cuh
@@ -78,29 +78,52 @@ _CCCL_DEVICE _CCCL_FORCEINLINE void prefetch(const T* addr)
   asm volatile("prefetch.global.L2 [%0];" : : "l"(__cvta_generic_to_global(addr)) : "memory");
 }
 
-template <int BlockDim, typename It>
+template <int BlockDim, int PrefetchStride, typename It>
 _CCCL_DEVICE _CCCL_FORCEINLINE void prefetch_tile(It begin, int items)
 {
   if constexpr (THRUST_NS_QUALIFIER::is_contiguous_iterator_v<It>)
   {
-    constexpr int prefetch_byte_stride = 128; // TODO(bgruber): should correspond to cache line size. Does this need to
-                                              // be architecture dependent?
     const int items_bytes = items * sizeof(it_value_t<It>);
 
     // prefetch does not stall and unrolling just generates a lot of unnecessary computations and predicate handling
     _CCCL_PRAGMA_NOUNROLL()
-    for (int offset = threadIdx.x * prefetch_byte_stride; offset < items_bytes;
-         offset += BlockDim * prefetch_byte_stride)
+    for (int offset = threadIdx.x * PrefetchStride; offset < items_bytes; offset += BlockDim * PrefetchStride)
     {
       prefetch(reinterpret_cast<const char*>(::cuda::std::to_address(begin)) + offset);
     }
   }
 }
 
+// TODO(bgruber): I would love if we could make _CCCL_PRAGMA_UNROLL(UnrollFactor) portably accept some sentinel value
+// that would disable the pragma and leave it to the compiler to choose whether to unroll or not. nvcc supports
+// _CCCL_PRAGMA_UNROLL(-1), but issues a warning though. Trying to suppress this warning inside the _CCCL_PRAGMA_UNROLL
+// macro kills MSVC. Also clang in CUDA mode does not support a non-positive unroll factor.
+template <int UnrollFactor, typename F>
+_CCCL_DEVICE _CCCL_FORCEINLINE void unrolled_for(int count, F&& body)
+{
+  if constexpr (UnrollFactor > 0)
+  {
+    _CCCL_PRAGMA_UNROLL(UnrollFactor)
+    for (int i = 0; i < count; ++i)
+    {
+      body(i);
+    }
+  }
+  else
+  {
+    for (int i = 0; i < count; ++i)
+    {
+      body(i);
+    }
+  }
+}
+
 // This kernel guarantees that objects passed as arguments to the user-provided transformation function f reside in
 // global memory. No intermediate copies are taken. If the parameter type of f is a reference, taking the address of the
 // parameter yields a global memory address.
 template <int BlockThreads,
+          int PrefetchByteStride,
+          int UnrollFactor,
           typename Offset,
           typename Predicate,
           typename F,
@@ -126,14 +149,10 @@ _CCCL_DEVICE void transform_kernel_prefetch(
   }
 
   _CCCL_PDL_GRID_DEPENDENCY_SYNC();
-  (..., prefetch_tile<block_threads>(ins, valid_items));
+  (..., prefetch_tile<block_threads, PrefetchByteStride>(ins, valid_items));
 
   auto process_tile = [&](auto full_tile, auto... ins2 /* nvcc fails to compile when just using the captured ins */) {
-    // ahendriksen: various unrolling yields less <1% gains at much higher compile-time cost
-    // bgruber: but A6000 and H100 show small gains without pragma
-    // _CCCL_PRAGMA_NOUNROLL()
-    for (int j = 0; j < num_elem_per_thread; ++j)
-    {
+    unrolled_for<UnrollFactor>(num_elem_per_thread, [&](int j) {
       const int idx = j * block_threads + threadIdx.x;
       if (full_tile || idx < valid_items)
       {
@@ -143,7 +162,7 @@ _CCCL_DEVICE void transform_kernel_prefetch(
           out[idx] = f(THRUST_NS_QUALIFIER::raw_reference_cast(ins2[idx])...);
         }
       }
-    }
+    });
   };
   if (tile_size == valid_items)
   {
@@ -213,6 +232,8 @@ template < // const transform_policy& Policy,
   int block_threads,
   int items_per_thread,
   int vec_size,
+  int PrefetchByteStride,
+  int PrefetchUnrollFactor,
   typename Offset,
   typename F,
   typename RandomAccessIteratorOut,
@@ -236,7 +257,7 @@ _CCCL_DEVICE void transform_kernel_vectorized(
   // if we cannot vectorize or don't have a full tile, fall back to prefetch kernel
   if (!can_vectorize || valid_items != tile_size)
   {
-    transform_kernel_prefetch<block_threads>(
+    transform_kernel_prefetch<block_threads, PrefetchByteStride, PrefetchUnrollFactor>(
       num_items,
       num_elem_per_thread_prefetch,
       always_true_predicate{},
@@ -561,6 +582,7 @@ _CCCL_DEVICE auto copy_and_return_smem_dst_fallback(
 // note: there is no PDL in this kernel since PDL is not supported below Hopper and this kernel is intended for Ampere
 template < // const transform_policy& Policy,
   int block_threads,
+  int UnrollFactor,
   typename Offset,
   typename Predicate,
   typename F,
@@ -603,10 +625,7 @@ _CCCL_DEVICE void transform_kernel_ldgsts(
   // TODO(bgruber): fbusato suggests to move the valid_items and smem_base_ptrs by threadIdx.x before the loop below
 
   auto process_tile = [&](auto full_tile) {
-    // Unroll 1 tends to improve performance, especially for smaller data types (confirmed by benchmark)
-    _CCCL_PRAGMA_NOUNROLL()
-    for (int j = 0; j < num_elem_per_thread; ++j)
-    {
+    unrolled_for<UnrollFactor>(num_elem_per_thread, [&](int j) {
       const int idx = j * block_threads + threadIdx.x;
       if (full_tile || idx < valid_items)
       {
@@ -619,7 +638,7 @@ _CCCL_DEVICE void transform_kernel_ldgsts(
           },
           smem_ptrs);
       }
-    }
+    });
   };
 
   // explicitly calling the lambda on literal true/false lets the compiler emit the lambda twice
@@ -715,6 +734,7 @@ _CCCL_DEVICE void bulk_copy_maybe_unaligned(
 template < // const transform_policy& Policy,
   int block_threads,
   int bulk_copy_alignment,
+  int UnrollFactor,
   typename Offset,
   typename Predicate,
   typename F,
@@ -899,10 +919,7 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
   out += offset;
 
   auto process_tile = [&](auto full_tile) {
-    // Unroll 1 tends to improve performance, especially for smaller data types (confirmed by benchmark)
-    _CCCL_PRAGMA_NOUNROLL()
-    for (int j = 0; j < num_elem_per_thread; ++j)
-    {
+    unrolled_for<UnrollFactor>(num_elem_per_thread, [&](int j) {
       // TODO(bgruber): fbusato suggests to hoist threadIdx.x out of the loop below
       const int idx = j * block_threads + threadIdx.x;
       if (full_tile || idx < valid_items)
@@ -926,7 +943,7 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
           },
           ::cuda::std::tuple<InTs...>{fetch_operand(aligned_ptrs)...});
       }
-    }
+    });
   };
   // explicitly calling the lambda on literal true/false lets the compiler emit the lambda twice
   if (tile_size == valid_items)
@@ -1030,7 +1047,9 @@ __launch_bounds__(get_block_threads<PolicySelector>) CUB_DETAIL_KERNEL_ATTRIBUTE
 
   if constexpr (policy.algorithm == Algorithm::prefetch)
   {
-    transform_kernel_prefetch<policy.prefetch.block_threads>(
+    transform_kernel_prefetch<policy.prefetch.block_threads,
+                              policy.prefetch.prefetch_byte_stride,
+                              policy.prefetch.unroll_factor>(
       num_items,
       num_elem_per_thread,
       ::cuda::std::move(pred),
@@ -1045,7 +1064,9 @@ __launch_bounds__(get_block_threads<PolicySelector>) CUB_DETAIL_KERNEL_ATTRIBUTE
 
     transform_kernel_vectorized</*policy*/ policy.vectorized.block_threads,
                                 policy.vectorized.items_per_thread,
-                                policy.vectorized.vec_size>(
+                                policy.vectorized.vec_size,
+                                policy.prefetch.prefetch_byte_stride,
+                                policy.prefetch.unroll_factor>(
       num_items,
       num_elem_per_thread,
       can_vectorize,
@@ -1057,7 +1078,7 @@ __launch_bounds__(get_block_threads<PolicySelector>) CUB_DETAIL_KERNEL_ATTRIBUTE
   {
     NV_IF_TARGET(
       NV_PROVIDES_SM_80,
-      (transform_kernel_ldgsts</*policy*/ policy.async_copy.block_threads>(
+      (transform_kernel_ldgsts</*policy*/ policy.async_copy.block_threads, policy.async_copy.unroll_factor>(
          num_items,
          num_elem_per_thread,
          ::cuda::std::move(pred),
@@ -1069,7 +1090,9 @@ __launch_bounds__(get_block_threads<PolicySelector>) CUB_DETAIL_KERNEL_ATTRIBUTE
   {
     NV_IF_TARGET(
       NV_PROVIDES_SM_90,
-      (transform_kernel_ublkcp</*policy*/ policy.async_copy.block_threads, policy.async_copy.bulk_copy_alignment>(
+      (transform_kernel_ublkcp</*policy*/ policy.async_copy.block_threads,
+                               policy.async_copy.bulk_copy_alignment,
+                               policy.async_copy.unroll_factor>(
          num_items,
          num_elem_per_thread,
          ::cuda::std::move(pred),
diff --git a/cub/cub/device/dispatch/tuning/tuning_transform.cuh b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
index c369eb4ccc8..1530529cf9c 100644
--- a/cub/cub/device/dispatch/tuning/tuning_transform.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
@@ -88,11 +88,16 @@ struct prefetch_policy
   int items_per_thread_no_input = 2; // when there are no input iterators, the kernel is just filling
   int min_items_per_thread      = 1;
   int max_items_per_thread      = 32;
+  int prefetch_byte_stride      = 128; // somewhat of a cache line size
+  // ahendriksen: various unrolling yields less <1% gains at much higher compile-time cost, so prevent unrolling
+  // bgruber: but A6000 and H100 show small gains without pragma, so omitting pragma
+  int unroll_factor = -1; // -1 means we leave it to the compiler
 
   [[nodiscard]] _CCCL_API constexpr friend bool operator==(const prefetch_policy& lhs, const prefetch_policy& rhs)
   {
     return lhs.block_threads == rhs.block_threads && lhs.items_per_thread_no_input == rhs.items_per_thread_no_input
-        && lhs.min_items_per_thread == rhs.min_items_per_thread && lhs.max_items_per_thread == rhs.max_items_per_thread;
+        && lhs.min_items_per_thread == rhs.min_items_per_thread && lhs.max_items_per_thread == rhs.max_items_per_thread
+        && lhs.prefetch_byte_stride == rhs.prefetch_byte_stride && lhs.unroll_factor == rhs.unroll_factor;
   }
 
   [[nodiscard]] _CCCL_API constexpr friend bool operator!=(const prefetch_policy& lhs, const prefetch_policy& rhs)
@@ -103,9 +108,11 @@ struct prefetch_policy
 #if !_CCCL_COMPILER(NVRTC)
   friend ::std::ostream& operator<<(::std::ostream& os, const prefetch_policy& policy)
   {
-    return os << "prefetch_policy { .block_threads = " << policy.block_threads << ", .items_per_thread_no_input = "
-              << policy.items_per_thread_no_input << ", .min_items_per_thread = " << policy.min_items_per_thread
-              << ", .max_items_per_thread = " << policy.max_items_per_thread << " }";
+    return os
+        << "prefetch_policy { .block_threads = " << policy.block_threads << ", .items_per_thread_no_input = "
+        << policy.items_per_thread_no_input << ", .min_items_per_thread = " << policy.min_items_per_thread
+        << ", .max_items_per_thread = " << policy.max_items_per_thread << ", .prefetch_byte_stride = "
+        << policy.prefetch_byte_stride << ", .unroll_factor = " << policy.unroll_factor << " }";
   }
 #endif // !_CCCL_COMPILER(NVRTC)
 };
@@ -143,11 +150,14 @@ struct async_copy_policy
   // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
   int min_items_per_thread = 1;
   int max_items_per_thread = 32;
+  // Unroll 1 tends to improve performance, especially for smaller data types (confirmed by benchmark)
+  int unroll_factor = 1;
 
   [[nodiscard]] _CCCL_API constexpr friend bool operator==(const async_copy_policy& lhs, const async_copy_policy& rhs)
   {
     return lhs.block_threads == rhs.block_threads && lhs.bulk_copy_alignment == rhs.bulk_copy_alignment
-        && lhs.min_items_per_thread == rhs.min_items_per_thread && lhs.max_items_per_thread == rhs.max_items_per_thread;
+        && lhs.min_items_per_thread == rhs.min_items_per_thread && lhs.max_items_per_thread == rhs.max_items_per_thread
+        && lhs.unroll_factor == rhs.unroll_factor;
   }
 
   [[nodiscard]] _CCCL_API constexpr friend bool operator!=(const async_copy_policy& lhs, const async_copy_policy& rhs)
@@ -158,9 +168,11 @@ struct async_copy_policy
 #if !_CCCL_COMPILER(NVRTC)
   friend ::std::ostream& operator<<(::std::ostream& os, const async_copy_policy& policy)
   {
-    return os << "async_copy_policy { .block_threads = " << policy.block_threads << ", .bulk_copy_alignment = "
-              << policy.bulk_copy_alignment << ", .min_items_per_thread = " << policy.min_items_per_thread
-              << ", .max_items_per_thread = " << policy.max_items_per_thread << " }";
+    return os
+        << "async_copy_policy { .block_threads = " << policy.block_threads << ", .bulk_copy_alignment = "
+        << policy.bulk_copy_alignment << ", .min_items_per_thread = " << policy.min_items_per_thread
+        << ", .max_items_per_thread = " << policy.max_items_per_thread << ", .unroll_factor = " << policy.unroll_factor
+        << " }";
   }
 #endif // !_CCCL_COMPILER(NVRTC)
 };
