diff --git a/flashinfer/jit/core.py b/flashinfer/jit/core.py
index f188bda802..7797ad2aaa 100644
--- a/flashinfer/jit/core.py
+++ b/flashinfer/jit/core.py
@@ -114,6 +114,9 @@ def load_cuda_ops(
             "--ptxas-options=-v",
             "--ptxas-options=--verbose,--register-usage-level=10,--warn-on-local-memory-usage",
         ]
+    else:
+        # non debug mode
+        cuda_cflags += ["-DNDEBUG"]
 
     cflags += extra_cflags
     cuda_cflags += extra_cuda_cflags
diff --git a/setup.py b/setup.py
index 6aa2bc2d25..cb3478d194 100644
--- a/setup.py
+++ b/setup.py
@@ -225,6 +225,7 @@ def __init__(self, *args, **kwargs) -> None:
         "-Xfatbin",
         "-compress-all",
         "-use_fast_math",
+        "-DNDEBUG",
         "-DPy_LIMITED_API=0x03080000",
     ]
     libraries = [
