diff --git a/kernels/matmul/educational/level_04.cu b/kernels/matmul/educational/level_04.cu
index d207f5f5b..361d9d2fc 100644
--- a/kernels/matmul/educational/level_04.cu
+++ b/kernels/matmul/educational/level_04.cu
@@ -38,6 +38,7 @@ __global__ void kernel(const __grid_constant__ matmul_globals g) {
     int col = blockIdx.x; 
     int row = blockIdx.y; 
 
+    zero(C_accum);
     int num_tiles = (g.N + BLOCK_SIZE - 1) / BLOCK_SIZE;
     for (int tile = 0; tile < num_tiles; ++tile) {
         load(As, g.A, {0, 0, row, tile});
