diff --git a/flashinfer/gemm/gemm_base.py b/flashinfer/gemm/gemm_base.py
index 60bc5eb76f..f7cd63cbd4 100644
--- a/flashinfer/gemm/gemm_base.py
+++ b/flashinfer/gemm/gemm_base.py
@@ -1720,18 +1720,22 @@ def _is_cublas_fp4_available_in_cudnn():
     )
 
 
-# Global cudnn handle. need to make it per device in future
-_cudnn_handle = None
+# One cudnn handle per each GPU
+_cudnn_handles: dict[int, int] = {}
 
 
-def _get_cudnn_handle(stream: torch.cuda.Stream):
+def _get_cudnn_handle(device, stream: torch.cuda.Stream):
     """Create and return a cached cuDNN handle."""
-    global _cudnn_handle
-    if _cudnn_handle is None:
+    global _cudnn_handles
+    device_id = device.index
+
+    if _cudnn_handles.get(device_id) is None:
         _check_cudnn_availability()
-        _cudnn_handle = cudnn.create_handle()
-    cudnn.set_stream(_cudnn_handle, stream.cuda_stream)
-    return _cudnn_handle
+        _cudnn_handles[device_id] = cudnn.create_handle()
+        print("cudnn_handle created for device_id = {}\n".format(device_id))
+    cudnn.set_stream(_cudnn_handles[device_id], stream.cuda_stream)
+
+    return _cudnn_handles[device_id]
 
 
 def _validate_fp8_output_dtype(dtype: torch.dtype):
@@ -1752,7 +1756,7 @@ def _validate_bf16_output_dtype(dtype: torch.dtype):
         )
 
 
-@functools.cache
+@functools.lru_cache(maxsize=1024)
 def create_cudnn_execution_plans_fp4_gemm(
     a_shape,
     a_stride,
@@ -1770,7 +1774,7 @@ def create_cudnn_execution_plans_fp4_gemm(
     use_nvfp4,
 ):
     stream = torch.cuda.current_stream(device)
-    with cudnn.graph(_get_cudnn_handle(stream)) as (graph, _):
+    with cudnn.graph(_get_cudnn_handle(device, stream)) as (graph, _):
         scale_type = cudnn.data_type.FP8_E4M3 if use_nvfp4 else cudnn.data_type.FP8_E8M0
 
         a_cudnn_tensor = graph.tensor(
@@ -1853,7 +1857,7 @@ def create_cudnn_execution_plans_fp4_gemm(
         return graph
 
 
-@functools.cache
+@functools.lru_cache(maxsize=1024)
 def build_plans_cudnn_fp4_gemm_graph(
     a_shape,
     a_stride,
@@ -1927,10 +1931,15 @@ def execute_cudnn_gemm_fp4_graph(
     stream = torch.cuda.current_stream(a.device)
 
     if tactic == -1:
-        graph.execute(variant_pack, workspace_buffer, handle=_get_cudnn_handle(stream))
+        graph.execute(
+            variant_pack, workspace_buffer, handle=_get_cudnn_handle(a.device, stream)
+        )
     else:
         graph.execute_plan_at_index(
-            variant_pack, workspace_buffer, tactic, handle=_get_cudnn_handle(stream)
+            variant_pack,
+            workspace_buffer,
+            tactic,
+            handle=_get_cudnn_handle(a.device, stream),
         )
 
 
@@ -1962,14 +1971,19 @@ def execute_cudnn_gemm_mxfp8_graph(
     stream = torch.cuda.current_stream(a.device)
 
     if tactic == -1:
-        graph.execute(variant_pack, workspace_buffer, handle=_get_cudnn_handle(stream))
+        graph.execute(
+            variant_pack, workspace_buffer, handle=_get_cudnn_handle(a.device, stream)
+        )
     else:
         graph.execute_plan_at_index(
-            variant_pack, workspace_buffer, tactic, handle=_get_cudnn_handle(stream)
+            variant_pack,
+            workspace_buffer,
+            tactic,
+            handle=_get_cudnn_handle(a.device, stream),
         )
 
 
-@functools.cache
+@functools.lru_cache(maxsize=1024)
 def build_cudnn_gemm_with_per_tensor_q_graph(
     a_shape, a_stride, b_shape, b_stride, a_type, b_type, o_type, device
 ):
@@ -1992,7 +2006,7 @@ def build_cudnn_gemm_with_per_tensor_q_graph(
     _check_cudnn_availability()
 
     stream = torch.cuda.current_stream(device)
-    with cudnn.graph(_get_cudnn_handle(stream)) as (graph, _):
+    with cudnn.graph(_get_cudnn_handle(device, stream)) as (graph, _):
         a_cudnn_tensor = graph.tensor(
             name="a", dim=a_shape, stride=a_stride, data_type=a_type
         )
@@ -2062,7 +2076,7 @@ def execute_cudnn_gemm_with_per_tensor_q_graph(
     }
 
     stream = torch.cuda.current_stream(a.device)
-    cudnn_handle = _get_cudnn_handle(stream)
+    cudnn_handle = _get_cudnn_handle(a.device, stream)
 
     if workspace.numel() < graph.get_workspace_size():
         workspace = torch.empty(
@@ -2154,7 +2168,7 @@ def build_cudnn_gemm_bf16_graph(a_shape, a_stride, b_shape, b_stride, o_type, de
     _check_cudnn_availability()
 
     stream = torch.cuda.current_stream(device)
-    with cudnn.graph(_get_cudnn_handle(stream)) as (graph, _):
+    with cudnn.graph(_get_cudnn_handle(device, stream)) as (graph, _):
         a_cudnn_tensor = graph.tensor(
             name="a", dim=a_shape, stride=a_stride, data_type=cudnn.data_type.BFLOAT16
         )
@@ -2190,7 +2204,7 @@ def execute_cudnn_gemm_bf16_graph(graph, a, b, c_final, workspace, tactic: int =
     }
 
     stream = torch.cuda.current_stream(a.device)
-    cudnn_handle = _get_cudnn_handle(stream)
+    cudnn_handle = _get_cudnn_handle(a.device, stream)
 
     if workspace.numel() < graph.get_workspace_size():
         workspace = torch.empty(
@@ -5696,7 +5710,7 @@ def div_up(a, b):
     return block_scale_dim_m, block_scale_dim_n, block_scale_dim_k
 
 
-@functools.cache
+@functools.lru_cache(maxsize=1024)
 def create_cudnn_execution_plans_mxfp8_gemm(
     a_shape,
     a_stride,
@@ -5755,7 +5769,7 @@ def create_cudnn_execution_plans_mxfp8_gemm(
     scale_type = cudnn.data_type.FP8_E8M0
 
     stream = torch.cuda.current_stream(device)
-    with cudnn.graph(_get_cudnn_handle(stream)) as (graph, _):
+    with cudnn.graph(_get_cudnn_handle(device, stream)) as (graph, _):
         a_cudnn_tensor = graph.tensor(
             name="a",
             dim=tuple(a_shape),  # [b, m, k]
