diff --git a/libcudacxx/include/cuda/std/__cccl/assert.h b/libcudacxx/include/cuda/std/__cccl/assert.h
index b5e3afb7f52..3d94a117510 100644
--- a/libcudacxx/include/cuda/std/__cccl/assert.h
+++ b/libcudacxx/include/cuda/std/__cccl/assert.h
@@ -89,7 +89,7 @@ void __assert_fail(const char* __assertion, const char* __file, unsigned int __l
 #  if _CCCL_OS(APPLE)
 #    define _CCCL_ASSERT_IMPL_HOST(expression, message)      \
       _CCCL_BUILTIN_EXPECT(static_cast<bool>(expression), 1) \
-      ? (void) 0 : __assert_rtn(__func__, __FILE__, __LINE__, __message__)
+      ? (void) 0 : __assert_rtn(__func__, __FILE__, __LINE__, message)
 #  elif _CCCL_OS(ANDROID)
 #    define _CCCL_ASSERT_IMPL_HOST(expression, message)      \
       _CCCL_BUILTIN_EXPECT(static_cast<bool>(expression), 1) \
diff --git a/thrust/testing/sort_by_key.cu b/thrust/testing/sort_by_key.cu
index 999163e1624..6e9fa17aed9 100644
--- a/thrust/testing/sort_by_key.cu
+++ b/thrust/testing/sort_by_key.cu
@@ -137,3 +137,16 @@ void TestSortByKeyBoolDescending()
   ASSERT_EQUAL(h_values, d_values);
 }
 DECLARE_UNITTEST(TestSortByKeyBoolDescending);
+
+void TestSortByKeyLongDouble()
+{
+  thrust::host_vector<long double> h_keys    = {10.0L, 9.0L, 8.0L, 7.0L, 6.0L, 5.0L, 4.0L, 3.0L, 2.0L, 1.0L};
+  thrust::host_vector<int> h_values          = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+  thrust::host_vector<int> h_values_expected = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
+
+  thrust::sort_by_key(h_keys.begin(), h_keys.end(), h_values.begin());
+
+  ASSERT_EQUAL(thrust::is_sorted(h_keys.begin(), h_keys.end()), true);
+  ASSERT_EQUAL(h_values, h_values_expected);
+}
+DECLARE_UNITTEST(TestSortByKeyLongDouble);
diff --git a/thrust/thrust/system/detail/sequential/sort.h b/thrust/thrust/system/detail/sequential/sort.h
index 7bca6f3bac6..08a55d4bbed 100644
--- a/thrust/thrust/system/detail/sequential/sort.h
+++ b/thrust/thrust/system/detail/sequential/sort.h
@@ -34,6 +34,10 @@ namespace sort_detail
 template <typename KeyType, typename Compare>
 inline constexpr bool use_primitive_sort =
   ::cuda::std::is_arithmetic_v<KeyType>
+  // RadixEncoder is not specialized for long double
+  && !::cuda::std::is_same_v<KeyType, long double>
+  // radix_sort_dispatcher only supports key sizes of 1, 2, 4, or 8 bytes
+  && (sizeof(KeyType) == 1 || sizeof(KeyType) == 2 || sizeof(KeyType) == 4 || sizeof(KeyType) == 8)
   && (::cuda::std::is_same_v<Compare, ::cuda::std::less<KeyType>>
       || ::cuda::std::is_same_v<Compare, ::cuda::std::greater<KeyType>>);
 
