diff --git a/cub/cub/block/block_reduce.cuh b/cub/cub/block/block_reduce.cuh
index e11ef3cbd14..c1472ea6744 100644
--- a/cub/cub/block/block_reduce.cuh
+++ b/cub/cub/block/block_reduce.cuh
@@ -45,7 +45,6 @@
 #include <cub/block/specializations/block_reduce_raking.cuh>
 #include <cub/block/specializations/block_reduce_raking_commutative_only.cuh>
 #include <cub/block/specializations/block_reduce_warp_reductions.cuh>
-#include <cub/block/specializations/block_reduce_warp_reductions_nondeterministic.cuh>
 #include <cub/thread/thread_operators.cuh>
 #include <cub/util_ptx.cuh>
 #include <cub/util_type.cuh>
@@ -282,7 +281,7 @@ private:
 
   using WarpReductions = detail::BlockReduceWarpReductions<T, BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z>;
   using WarpReductionsNondeterministic =
-    detail::BlockReduceWarpReductionsNondeterministic<T, BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z>;
+    detail::BlockReduceWarpReductions<T, BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z, false>;
   using RakingCommutativeOnly = detail::BlockReduceRakingCommutativeOnly<T, BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z>;
   using Raking                = detail::BlockReduceRaking<T, BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z>;
 
diff --git a/cub/cub/block/specializations/block_reduce_warp_reductions.cuh b/cub/cub/block/specializations/block_reduce_warp_reductions.cuh
index d6658a5fa24..b9be948800a 100644
--- a/cub/cub/block/specializations/block_reduce_warp_reductions.cuh
+++ b/cub/cub/block/specializations/block_reduce_warp_reductions.cuh
@@ -26,11 +26,11 @@
  *
  ******************************************************************************/
 
-/**
- * @file
- * cub::BlockReduceWarpReductions provides variants of warp-reduction-based parallel reduction
- * across a CUDA thread block. Supports non-commutative reduction operators.
- */
+//! @rst
+//! @file
+//! cub::BlockReduceWarpReductions provides variants of warp-reduction-based parallel reduction
+//! across a CUDA thread block. Supports non-commutative reduction operators.
+//! @endrst
 
 #pragma once
 
@@ -48,67 +48,64 @@
 #include <cub/util_ptx.cuh>
 #include <cub/warp/warp_reduce.cuh>
 
+#include <cuda/__cmath/ceil_div.h>
+#include <cuda/atomic>
 #include <cuda/ptx>
+#include <cuda/std/__algorithm/min.h>
 
 CUB_NAMESPACE_BEGIN
 namespace detail
 {
-/**
- * @brief BlockReduceWarpReductions provides variants of warp-reduction-based parallel reduction
- *        across a CUDA thread block. Supports non-commutative reduction operators.
- * @tparam T
- *   Data type being reduced
- *
- * @tparam BLOCK_DIM_X
- *   The thread block length in threads along the X dimension
- *
- * @tparam BLOCK_DIM_Y
- *   The thread block length in threads along the Y dimension
- *
- * @tparam BLOCK_DIM_Z
- *   The thread block length in threads along the Z dimension
- */
-template <typename T, int BLOCK_DIM_X, int BLOCK_DIM_Y, int BLOCK_DIM_Z>
+//! @rst
+//! BlockReduceWarpReductions provides variants of warp-reduction-based parallel reduction
+//! across a CUDA thread block. Supports non-commutative reduction operators.
+//! @endrst
+//!
+//! @tparam T
+//!   Data type being reduced
+//!
+//! @tparam BlockDimX
+//!   The thread block length in threads along the X dimension
+//!
+//! @tparam BlockDimY
+//!   The thread block length in threads along the Y dimension
+//!
+//! @tparam BlockDimZ
+//!   The thread block length in threads along the Z dimension
+//!
+//! @tparam IsDeterministic
+//!   Whether the reduction is deterministic
+template <typename T, int BlockDimX, int BlockDimY, int BlockDimZ, bool IsDeterministic = true>
 struct BlockReduceWarpReductions
 {
-  /// Constants
-  enum
-  {
-    /// The thread block size in threads
-    BLOCK_THREADS = BLOCK_DIM_X * BLOCK_DIM_Y * BLOCK_DIM_Z,
+  /// The thread block size in threads
+  static constexpr int block_threads = BlockDimX * BlockDimY * BlockDimZ;
 
-    /// Number of warp threads
-    WARP_THREADS = warp_threads,
+  /// Number of active warps
+  static constexpr int warps = ::cuda::ceil_div(block_threads, warp_threads);
 
-    /// Number of active warps
-    WARPS = (BLOCK_THREADS + WARP_THREADS - 1) / WARP_THREADS,
+  /// The logical warp size for warp reductions
+  static constexpr int logical_warp_size = _CUDA_VSTD::min(block_threads, warp_threads);
 
-    /// The logical warp size for warp reductions
-    LOGICAL_WARP_SIZE = (BLOCK_THREADS < WARP_THREADS ? BLOCK_THREADS : WARP_THREADS), // MSVC bug with cuda::std::min
+  /// Whether or not the logical warp size evenly divides the thread block size
+  static constexpr bool even_warp_multiple = (block_threads % logical_warp_size == 0);
 
-    /// Whether or not the logical warp size evenly divides the thread block size
-    EVEN_WARP_MULTIPLE = (BLOCK_THREADS % LOGICAL_WARP_SIZE == 0)
-  };
-
-  ///  WarpReduce utility type
-  using WarpReduceInternal = typename WarpReduce<T, LOGICAL_WARP_SIZE>::InternalWarpReduce;
+  using WarpReduceInternal = typename WarpReduce<T, logical_warp_size>::InternalWarpReduce;
 
   /// Shared memory storage layout type
   struct _TempStorage
   {
     /// Buffer for warp-synchronous reduction
-    typename WarpReduceInternal::TempStorage warp_reduce[WARPS];
+    typename WarpReduceInternal::TempStorage warp_reduce[warps];
 
     /// Shared totals from each warp-synchronous reduction
-    T warp_aggregates[WARPS];
+    T warp_aggregates[warps];
 
     /// Shared prefix for the entire thread block
     T block_prefix;
   };
 
-  /// Alias wrapper allowing storage to be unioned
-  struct TempStorage : Uninitialized<_TempStorage>
-  {};
+  using TempStorage = Uninitialized<_TempStorage>;
 
   // Thread fields
   _TempStorage& temp_storage;
@@ -116,66 +113,59 @@ struct BlockReduceWarpReductions
   int warp_id;
   int lane_id;
 
-  /// Constructor
   _CCCL_DEVICE _CCCL_FORCEINLINE BlockReduceWarpReductions(TempStorage& temp_storage)
       : temp_storage(temp_storage.Alias())
-      , linear_tid(RowMajorTid(BLOCK_DIM_X, BLOCK_DIM_Y, BLOCK_DIM_Z))
-      , warp_id((WARPS == 1) ? 0 : linear_tid / WARP_THREADS)
+      , linear_tid(RowMajorTid(BlockDimX, BlockDimY, BlockDimZ))
+      , warp_id((warps == 1) ? 0 : linear_tid / warp_threads)
       , lane_id(::cuda::ptx::get_sreg_laneid())
   {}
 
-  /**
-   * @param[in] reduction_op
-   *   Binary reduction operator
-   *
-   * @param[in] warp_aggregate
-   *   <b>[<em>lane</em><sub>0</sub> only]</b> Warp-wide aggregate reduction of input items
-   *
-   * @param[in] num_valid
-   *   Number of valid elements (may be less than BLOCK_THREADS)
-   */
-  template <bool FULL_TILE, typename ReductionOp, int SUCCESSOR_WARP>
-  _CCCL_DEVICE _CCCL_FORCEINLINE T ApplyWarpAggregates(
-    ReductionOp reduction_op, T warp_aggregate, int num_valid, constant_t<SUCCESSOR_WARP> /*successor_warp*/)
+  //! @rst
+  //! Returns block-wide aggregate in *thread*\ :sub:`0`.
+  //! @endrst
+  //!
+  //! @tparam ReductionOp
+  //!   **[inferred]** Binary reduction operator type
+  //!
+  //! @param[in] reduction_op
+  //!   Binary reduction operator
+  //!
+  //! @param[in] warp_aggregate
+  //!   **[**\ *lane*\ :sub:`0` **only]** Warp-wide aggregate reduction of input items
+  template <typename ReductionOp>
+  _CCCL_DEVICE _CCCL_FORCEINLINE T ApplyWarpAggregatesNonDeterministic(ReductionOp reduction_op, T warp_aggregate)
   {
-    if (FULL_TILE || (SUCCESSOR_WARP * LOGICAL_WARP_SIZE < num_valid))
+    if (linear_tid == 0)
     {
-      T addend       = temp_storage.warp_aggregates[SUCCESSOR_WARP];
-      warp_aggregate = reduction_op(warp_aggregate, addend);
+      detail::uninitialized_copy_single(temp_storage.warp_aggregates, warp_aggregate);
     }
-    return ApplyWarpAggregates<FULL_TILE>(reduction_op, warp_aggregate, num_valid, constant_v<SUCCESSOR_WARP + 1>);
-  }
 
-  /**
-   * @param[in] reduction_op
-   *   Binary reduction operator
-   *
-   * @param[in] warp_aggregate
-   *   <b>[<em>lane</em><sub>0</sub> only]</b> Warp-wide aggregate reduction of input items
-   *
-   * @param[in] num_valid
-   *   Number of valid elements (may be less than BLOCK_THREADS)
-   */
-  template <bool FULL_TILE, typename ReductionOp>
-  _CCCL_DEVICE _CCCL_FORCEINLINE T ApplyWarpAggregates(
-    ReductionOp /*reduction_op*/, T warp_aggregate, int /*num_valid*/, constant_t<int{WARPS}> /*successor_warp*/)
-  {
-    return warp_aggregate;
+    __syncthreads();
+
+    // Warp 0 already contributed its aggregate above since its also linear_tid == 0
+    if (lane_id == 0 && warp_id != 0)
+    {
+      // TODO: replace this with other atomic operations when specified
+      NV_IF_TARGET(NV_PROVIDES_SM_60,
+                   (::cuda::atomic_ref<T, ::cuda::thread_scope_block> atomic_target(temp_storage.warp_aggregates[0]);
+                    atomic_target.fetch_add(warp_aggregate, ::cuda::memory_order_relaxed);),
+                   (atomicAdd(&temp_storage.warp_aggregates[0], warp_aggregate);));
+    }
+
+    __syncthreads();
+    return temp_storage.warp_aggregates[0];
   }
 
-  /**
-   * @brief Returns block-wide aggregate in <em>thread</em><sub>0</sub>.
-   *
-   * @param[in] reduction_op
-   *   Binary reduction operator
-   *
-   * @param[in] warp_aggregate
-   *   <b>[<em>lane</em><sub>0</sub> only]</b> Warp-wide aggregate reduction of input items
-   *
-   * @param[in] num_valid
-   *   Number of valid elements (may be less than BLOCK_THREADS)
-   */
-  template <bool FULL_TILE, typename ReductionOp>
+  //! @rst
+  //! Recursively applies warp aggregates using template unrolling for deterministic reduction.
+  //! @endrst
+  //!
+  //! @tparam FullTile
+  //!   **[inferred]** Whether this is a full tile
+  //!
+  //! @tparam ReductionOp
+  //!   **[inferred]** Binary reduction operator type
+  template <bool FullTile, typename ReductionOp>
   _CCCL_DEVICE _CCCL_FORCEINLINE T ApplyWarpAggregates(ReductionOp reduction_op, T warp_aggregate, int num_valid)
   {
     // Share lane aggregates
@@ -189,69 +179,101 @@ struct BlockReduceWarpReductions
     // Update total aggregate in warp 0, lane 0
     if (linear_tid == 0)
     {
-      warp_aggregate = ApplyWarpAggregates<FULL_TILE>(reduction_op, warp_aggregate, num_valid, constant_v<1>);
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int warp_idx = 1; warp_idx < warps; ++warp_idx)
+      {
+        if (FullTile || (warp_idx * logical_warp_size < num_valid))
+        {
+          T addend       = temp_storage.warp_aggregates[warp_idx];
+          warp_aggregate = reduction_op(warp_aggregate, addend);
+        }
+      }
     }
 
     return warp_aggregate;
   }
 
-  /**
-   * @brief Computes a thread block-wide reduction using addition (+) as the reduction operator.
-   *        The first num_valid threads each contribute one reduction partial. The return value is
-   *        only valid for thread<sub>0</sub>.
-   *
-   * @param[in] input
-   *   Calling thread's input partial reductions
-   *
-   * @param[in] num_valid
-   *   Number of valid elements (may be less than BLOCK_THREADS)
-   */
-  template <bool FULL_TILE>
+  //! @rst
+  //! Computes a thread block-wide reduction using addition (+/cuda::std::plus<>) as the reduction operator.
+  //! The first num_valid threads each contribute one reduction partial. The return value is
+  //! only valid for *thread*\ :sub:`0`.
+  //! @endrst
+  //!
+  //! @tparam FullTile
+  //!   **[inferred]** Whether this is a full tile
+  //!
+  //! @param[in] input
+  //!   Calling thread's input partial reductions
+  //!
+  //! @param[in] num_valid
+  //!   Number of valid elements (may be less than block_threads)
+  template <bool FullTile>
   _CCCL_DEVICE _CCCL_FORCEINLINE T Sum(T input, int num_valid)
   {
     ::cuda::std::plus<> reduction_op;
-    int warp_offset    = (warp_id * LOGICAL_WARP_SIZE);
-    int warp_num_valid = ((FULL_TILE && EVEN_WARP_MULTIPLE) || (warp_offset + LOGICAL_WARP_SIZE <= num_valid))
-                         ? LOGICAL_WARP_SIZE
-                         : num_valid - warp_offset;
+    const int warp_offset = (warp_id * logical_warp_size);
+    const int warp_num_valid =
+      ((FullTile && even_warp_multiple) || (warp_offset + logical_warp_size <= num_valid))
+        ? logical_warp_size
+        : num_valid - warp_offset;
 
     // Warp reduction in every warp
-    T warp_aggregate =
-      WarpReduceInternal(temp_storage.warp_reduce[warp_id])
-        .template Reduce<(FULL_TILE && EVEN_WARP_MULTIPLE)>(input, warp_num_valid, ::cuda::std::plus<>{});
+    T warp_aggregate = WarpReduceInternal(temp_storage.warp_reduce[warp_id])
+                         .template Reduce<(FullTile && even_warp_multiple)>(input, warp_num_valid, reduction_op);
 
     // Update outputs and block_aggregate with warp-wide aggregates from lane-0s
-    return ApplyWarpAggregates<FULL_TILE>(reduction_op, warp_aggregate, num_valid);
+    if constexpr (IsDeterministic)
+    {
+      return ApplyWarpAggregates<FullTile>(reduction_op, warp_aggregate, num_valid);
+    }
+    else
+    {
+      return ApplyWarpAggregatesNonDeterministic(reduction_op, warp_aggregate);
+    }
   }
 
-  /**
-   * @brief Computes a thread block-wide reduction using the specified reduction operator.
-   *        The first num_valid threads each contribute one reduction partial.
-   *        The return value is only valid for thread<sub>0</sub>.
-   *
-   * @param[in] input
-   *   Calling thread's input partial reductions
-   *
-   * @param[in] num_valid
-   *   Number of valid elements (may be less than BLOCK_THREADS)
-   *
-   * @param[in] reduction_op
-   *   Binary reduction operator
-   */
-  template <bool FULL_TILE, typename ReductionOp>
+  //! @rst
+  //! Computes a thread block-wide reduction using the specified reduction operator.
+  //! The first num_valid threads each contribute one reduction partial.
+  //! The return value is only valid for *thread*\ :sub:`0`.
+  //! @endrst
+  //!
+  //! @tparam FullTile
+  //!   **[inferred]** Whether this is a full tile
+  //!
+  //! @tparam ReductionOp
+  //!   **[inferred]** Binary reduction operator type
+  //!
+  //! @param[in] input
+  //!   Calling thread's input partial reductions
+  //!
+  //! @param[in] num_valid
+  //!   Number of valid elements (may be less than block_threads)
+  //!
+  //! @param[in] reduction_op
+  //!   Binary reduction operator
+  template <bool FullTile, typename ReductionOp>
   _CCCL_DEVICE _CCCL_FORCEINLINE T Reduce(T input, int num_valid, ReductionOp reduction_op)
   {
-    int warp_offset    = warp_id * LOGICAL_WARP_SIZE;
-    int warp_num_valid = ((FULL_TILE && EVEN_WARP_MULTIPLE) || (warp_offset + LOGICAL_WARP_SIZE <= num_valid))
-                         ? LOGICAL_WARP_SIZE
-                         : num_valid - warp_offset;
+    const int warp_offset = warp_id * logical_warp_size;
+    const int warp_num_valid =
+      ((FullTile && even_warp_multiple) || (warp_offset + logical_warp_size <= num_valid))
+        ? logical_warp_size
+        : num_valid - warp_offset;
 
     // Warp reduction in every warp
-    T warp_aggregate = WarpReduceInternal(temp_storage.warp_reduce[warp_id])
-                         .template Reduce<(FULL_TILE && EVEN_WARP_MULTIPLE)>(input, warp_num_valid, reduction_op);
+    const T warp_aggregate = WarpReduceInternal(temp_storage.warp_reduce[warp_id])
+                               .template Reduce<(FullTile && even_warp_multiple)>(input, warp_num_valid, reduction_op);
 
     // Update outputs and block_aggregate with warp-wide aggregates from lane-0s
-    return ApplyWarpAggregates<FULL_TILE>(reduction_op, warp_aggregate, num_valid);
+    if constexpr (IsDeterministic)
+    {
+      return ApplyWarpAggregates<FullTile>(reduction_op, warp_aggregate, num_valid);
+    }
+    else
+    {
+      return ApplyWarpAggregatesNonDeterministic(reduction_op, warp_aggregate);
+    }
   }
 };
 } // namespace detail
diff --git a/cub/cub/block/specializations/block_reduce_warp_reductions_nondeterministic.cuh b/cub/cub/block/specializations/block_reduce_warp_reductions_nondeterministic.cuh
deleted file mode 100644
index 58d5688abbf..00000000000
--- a/cub/cub/block/specializations/block_reduce_warp_reductions_nondeterministic.cuh
+++ /dev/null
@@ -1,205 +0,0 @@
-// SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
-// SPDX-License-Identifier: BSD-3-Clause
-
-//! @rst
-//! @file
-//! cub::BlockReduceWarpReductionsNondeterministic provides variants of warp-reduction-based parallel reduction
-//! across a CUDA thread block. Supports non-commutative reduction operators.
-//! @endrst
-
-#pragma once
-
-#include <cub/config.cuh>
-
-#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
-#  pragma GCC system_header
-#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
-#  pragma clang system_header
-#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
-#  pragma system_header
-#endif // no system header
-
-#include <cub/detail/uninitialized_copy.cuh>
-#include <cub/util_ptx.cuh>
-#include <cub/warp/warp_reduce.cuh>
-
-#include <cuda/atomic>
-#include <cuda/ptx>
-#include <cuda/std/__algorithm_>
-#include <cuda/std/cmath>
-
-CUB_NAMESPACE_BEGIN
-namespace detail
-{
-//! @rst
-//! BlockReduceWarpReductionsNondeterministic provides variants of warp-reduction-based parallel reduction
-//! across a CUDA thread block. Supports non-commutative reduction operators.
-//! @endrst
-//!
-//! @tparam T
-//!   Data type being reduced
-//!
-//! @tparam BlockDimX
-//!   The thread block length in threads along the X dimension
-//!
-//! @tparam BlockDimY
-//!   The thread block length in threads along the Y dimension
-//!
-//! @tparam BlockDimZ
-//!   The thread block length in threads along the Z dimension
-template <typename T, int BlockDimX, int BlockDimY, int BlockDimZ>
-struct BlockReduceWarpReductionsNondeterministic
-{
-  /// The thread block size in threads
-  static constexpr int block_threads = BlockDimX * BlockDimY * BlockDimZ;
-
-  /// Number of active warps
-  static constexpr int warps = ::cuda::ceil_div(block_threads, warp_threads);
-
-  /// The logical warp size for warp reductions
-  static constexpr int logical_warp_size = _CUDA_VSTD::min(block_threads, warp_threads);
-
-  /// Whether or not the logical warp size evenly divides the thread block size
-  static constexpr bool even_warp_multiple = (block_threads % logical_warp_size == 0);
-
-  using WarpReduceInternal = typename WarpReduce<T, logical_warp_size>::InternalWarpReduce;
-
-  /// Shared memory storage layout type
-  struct _TempStorage
-  {
-    /// Buffer for warp-synchronous reduction
-    typename WarpReduceInternal::TempStorage warp_reduce[warps];
-
-    /// Shared totals from each warp-synchronous reduction
-    T warp_aggregates[warps];
-
-    /// Shared prefix for the entire thread block
-    T block_prefix;
-  };
-
-  using TempStorage = Uninitialized<_TempStorage>;
-
-  // Thread fields
-  _TempStorage& temp_storage;
-  int linear_tid;
-  int warp_id;
-  int lane_id;
-
-  _CCCL_DEVICE _CCCL_FORCEINLINE BlockReduceWarpReductionsNondeterministic(TempStorage& temp_storage)
-      : temp_storage(temp_storage.Alias())
-      , linear_tid(RowMajorTid(BlockDimX, BlockDimY, BlockDimZ))
-      , warp_id((warps == 1) ? 0 : linear_tid / warp_threads)
-      , lane_id(::cuda::ptx::get_sreg_laneid())
-  {}
-
-  //! @rst
-  //! Returns block-wide aggregate in *thread*\ :sub:`0`.
-  //! @endrst
-  //!
-  //! @tparam ReductionOp
-  //!   **[inferred]** Binary reduction operator type
-  //!
-  //! @param[in] reduction_op
-  //!   Binary reduction operator
-  //!
-  //! @param[in] warp_aggregate
-  //!   **[**\ *lane*\ :sub:`0` **only]** Warp-wide aggregate reduction of input items
-  //!
-  //! @param[in] num_valid
-  //!   Number of valid elements (may be less than block_threads)
-  template <typename ReductionOp>
-  _CCCL_DEVICE _CCCL_FORCEINLINE T ApplyWarpAggregates(ReductionOp reduction_op, T warp_aggregate, int num_valid)
-  {
-    if (linear_tid == 0)
-    {
-      detail::uninitialized_copy_single(temp_storage.warp_aggregates, warp_aggregate);
-    }
-
-    __syncthreads();
-
-    // Warp 0 already contributed its aggregate above since its also linear_tid == 0
-    if (lane_id == 0 && warp_id != 0)
-    {
-      // TODO: replace this with other atomic operations when specified
-      NV_IF_TARGET(NV_PROVIDES_SM_60,
-                   (::cuda::atomic_ref<T, ::cuda::thread_scope_block> atomic_target(temp_storage.warp_aggregates[0]);
-                    atomic_target.fetch_add(warp_aggregate, ::cuda::memory_order_relaxed);),
-                   (atomicAdd(&temp_storage.warp_aggregates[0], warp_aggregate);));
-    }
-
-    __syncthreads();
-    return temp_storage.warp_aggregates[0];
-  }
-
-  //! @rst
-  //! Computes a thread block-wide reduction using addition (+/cuda::std::plus<>) as the reduction operator.
-  //! The first num_valid threads each contribute one reduction partial. The return value is
-  //! only valid for *thread*\ :sub:`0`.
-  //! @endrst
-  //!
-  //! @tparam FullTile
-  //!   **[inferred]** Whether this is a full tile
-  //!
-  //! @param[in] input
-  //!   Calling thread's input partial reductions
-  //!
-  //! @param[in] num_valid
-  //!   Number of valid elements (may be less than block_threads)
-  template <bool FullTile>
-  _CCCL_DEVICE _CCCL_FORCEINLINE T Sum(T input, int num_valid)
-  {
-    ::cuda::std::plus<> reduction_op;
-    const int warp_offset = (warp_id * logical_warp_size);
-    const int warp_num_valid =
-      ((FullTile && even_warp_multiple) || (warp_offset + logical_warp_size <= num_valid))
-        ? logical_warp_size
-        : num_valid - warp_offset;
-
-    // Warp reduction in every warp
-    T warp_aggregate = WarpReduceInternal(temp_storage.warp_reduce[warp_id])
-                         .template Reduce<(FullTile && even_warp_multiple)>(input, warp_num_valid, reduction_op);
-
-    // Update outputs and block_aggregate with warp-wide aggregates from lane-0s
-    return ApplyWarpAggregates(reduction_op, warp_aggregate, num_valid);
-  }
-
-  //! @rst
-  //! Computes a thread block-wide reduction using the specified reduction operator.
-  //! The first num_valid threads each contribute one reduction partial.
-  //! The return value is only valid for *thread*\ :sub:`0`.
-  //! @endrst
-  //!
-  //! @tparam FullTile
-  //!   **[inferred]** Whether this is a full tile
-  //!
-  //! @tparam ReductionOp
-  //!   **[inferred]** Binary reduction operator type
-  //!
-  //! @param[in] input
-  //!   Calling thread's input partial reductions
-  //!
-  //! @param[in] num_valid
-  //!   Number of valid elements (may be less than block_threads)
-  //!
-  //! @param[in] reduction_op
-  //!   Binary reduction operator
-  template <bool FullTile, typename ReductionOp>
-  _CCCL_DEVICE _CCCL_FORCEINLINE T Reduce(T input, int num_valid, ReductionOp reduction_op)
-  {
-    const int warp_offset = warp_id * logical_warp_size;
-    const int warp_num_valid =
-      ((FullTile && even_warp_multiple) || (warp_offset + logical_warp_size <= num_valid))
-        ? logical_warp_size
-        : num_valid - warp_offset;
-
-    // Warp reduction in every warp
-    const T warp_aggregate = WarpReduceInternal(temp_storage.warp_reduce[warp_id])
-                               .template Reduce<(FullTile && even_warp_multiple)>(input, warp_num_valid, reduction_op);
-
-    // Update outputs and block_aggregate with warp-wide aggregates from lane-0s
-    return ApplyWarpAggregates(reduction_op, warp_aggregate, num_valid);
-  }
-};
-} // namespace detail
-
-CUB_NAMESPACE_END
