diff --git a/benchmarks/bench_rope.py b/benchmarks/bench_rope.py
index d87af5f1ad..a77b9b5dd7 100644
--- a/benchmarks/bench_rope.py
+++ b/benchmarks/bench_rope.py
@@ -1,93 +1,141 @@
-import argparse
-from typing import cast
+"""
+Benchmark RoPE for flashinfer and vLLM. vLLM installation is required to run this benchmark.
+
+Usage:
+$ pip install vllm
+$ python bench_rope.py
+"""
+
+import math
+from typing import Any, Dict, List, Optional, Tuple, Union
 
 import torch
-from triton.testing import do_bench
-
-import flashinfer
-
-
-def generate_cos_sin_f32_cache(max_seq_len, head_dim, theta=1e4):
-    position = torch.arange(max_seq_len).float().unsqueeze(1)
-    freqs = 1.0 / (theta ** (torch.arange(0, head_dim, 2).float() / head_dim))
-    freqs = torch.cat([freqs, freqs], dim=-1).contiguous()
-    args = position * freqs
-    sin_cache = torch.sin(args)
-    cos_cache = torch.cos(args)
-    return cos_cache, sin_cache
-
-
-@torch.inference_mode()
-def main():
-    parser = argparse.ArgumentParser()
-    parser.add_argument("--batch-sizes", nargs="+", type=int, default=[1, 19, 99, 128])
-    parser.add_argument("--append-len", nargs="+", type=int, default=[1, 128, 1024])
-    parser.add_argument("--num-qo-heads", type=int, default=32)
-    parser.add_argument("--num-kv-heads", type=int, default=8)
-    parser.add_argument("--head-dim", type=int, default=128)
-    args = parser.parse_args()
-
-    eps = 1e-6
-    dtype = torch.float16
-    num_qo_heads = args.num_qo_heads
-    num_kv_heads = args.num_kv_heads
-    head_dim = args.head_dim
-
-    # Loop over each combination of batch_size, hidden_size, and dtype
-    for batch_size in args.batch_sizes:
-        for append_len in args.append_len:
-            for use_cos_sin_cache in [False, True]:
-                # Define tensors with the correct dtype
-
-                q = torch.randn(
-                    (batch_size * append_len, num_qo_heads, args.head_dim),
-                    dtype=dtype,
-                    device="cuda",
-                )
-                k = torch.randn(
-                    (batch_size * append_len, num_kv_heads, args.head_dim),
-                    dtype=dtype,
-                    device="cuda",
-                )
-                pos_ids = torch.repeat_interleave(
-                    torch.arange(append_len, dtype=torch.int32, device=q.device),
-                    batch_size,
-                )
-                cos_cache, sin_cache = generate_cos_sin_f32_cache(4096, head_dim)
-                cos_cache = cos_cache.to(q.device)
-                sin_cache = sin_cache.to(q.device)
-
-                @torch.cuda.nvtx.range(
-                    f"apply_rope batch_size={batch_size}, append_len={append_len}, num_qo_heads={num_qo_heads}, num_kv_heads={num_kv_heads}, head_dim={head_dim}"
-                )
-                def fn() -> None:
-                    if use_cos_sin_cache:
-                        flashinfer.apply_rope_with_cos_sin_cache(
-                            q, k, cos_cache, sin_cache, pos_ids
-                        )
-                    else:
-                        flashinfer.apply_rope_pos_ids(q, k, pos_ids)
-
-                # Run benchmarking
-                latency_ms = cast(float, do_bench(fn))
-                throughput = (
-                    q.numel() * q.element_size() * 2 + k.numel() * k.element_size() * 2
-                ) / (latency_ms * 1e-3)
-                print(
-                    f"batch_size: {batch_size:3},",
-                    f"append_len: {append_len:5},",
-                    f"num_qo_heads: {num_qo_heads:5},",
-                    f"num_kv_heads: {num_kv_heads:5},",
-                    f"head_dim: {head_dim:5},",
-                    f"use_cos_sin_cache: {use_cos_sin_cache},",
-                    f"latency: {latency_ms*1e3:2.0f}us,",
-                    f"throughput: {throughput*1e-9:7.3f}GB/s",
-                )
-
-        print("---")
-
-    torch.cuda.profiler.stop()
+import torch.nn as nn
+from flashinfer.rope import apply_rope_with_cos_sin_cache_inplace
+from vllm.model_executor.layers.rotary_embedding import RotaryEmbedding as vLLMRotaryEmbedding
+import triton
+
+class FlashInferRotaryEmbedding(nn.Module):
+
+    def __init__(
+        self,
+        head_size: int,
+        rotary_dim: int,
+        max_position_embeddings: int,
+        base: int,
+        is_neox_style: bool,
+        dtype: torch.dtype,
+    ) -> None:
+        super().__init__()
+        self.head_size = head_size
+        self.rotary_dim = rotary_dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        self.is_neox_style = is_neox_style
+        self.dtype = dtype
+
+        cache = self._compute_cos_sin_cache()
+        self.cos_sin_cache: torch.Tensor
+        self.register_buffer("cos_sin_cache", cache, persistent=False)
+
+    def _compute_inv_freq(self, base: Union[int, float]) -> torch.Tensor:
+        inv_freq = 1.0 / (base**(torch.arange(
+            0, self.rotary_dim, 2, dtype=torch.float) / self.rotary_dim))
+        return inv_freq
+
+    def _compute_cos_sin_cache(self) -> torch.Tensor:
+        """Compute the cos and sin cache."""
+        inv_freq = self._compute_inv_freq(self.base)
+        t = torch.arange(self.max_position_embeddings, dtype=torch.float)
+
+        freqs = torch.einsum("i,j -> ij", t, inv_freq)
+        cos = freqs.cos()
+        sin = freqs.sin()
+        cache = torch.cat((cos, sin), dim=-1)
+        return cache
+
+    def _apply_rotary_emb(
+        self,
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+        is_neox_style: bool,
+    ) -> torch.Tensor:
+        """
+        Args:
+            x: [num_tokens, num_heads, head_size]
+            cos: [num_tokens, head_size // 2]
+            sin: [num_tokens, head_size // 2]
+            is_neox_style: Whether to use the Neox-style or GPT-J-style rotary
+                positional embeddings.
+        """
+        cos = cos.unsqueeze(-2).to(x.dtype)
+        sin = sin.unsqueeze(-2).to(x.dtype)
+        if is_neox_style:
+            x1, x2 = torch.chunk(x, 2, dim=-1)
+        else:
+            x1 = x[..., ::2]
+            x2 = x[..., 1::2]
+        o1 = x1 * cos - x2 * sin
+        o2 = x2 * cos + x1 * sin
+        if is_neox_style:
+            return torch.cat((o1, o2), dim=-1)
+        else:
+            return torch.stack((o1, o2), dim=-1).flatten(-2)
+        
+    def forward_cuda(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        offsets: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        apply_rope_with_cos_sin_cache_inplace(
+            positions=positions,
+            query=query,
+            key=key,
+            head_size=self.head_size,
+            cos_sin_cache=self.cos_sin_cache,
+            is_neox=self.is_neox_style,
+        )
+        return query, key
+
+@triton.testing.perf_report(
+    triton.testing.Benchmark(
+        x_names=["seq_len"],
+        x_vals=[2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536],
+        line_arg="provider",
+        line_vals=["flashinfer", "native", "vllm"],
+        line_names=["FlashInfer", "Native", "vLLM"],
+        styles=[("blue", "-"), ("red", "-"), ("green", "-")],
+        ylabel="Latency (ms)",
+        plot_name="rope-latency",
+        args={"head_size": 4096//32, "rotary_dim": 4096//32, "max_position_embeddings": 65536, "base": 500000, "is_neox_style": True, "dtype": torch.bfloat16, "device": "cuda", "batch_size": 2, "num_q_heads": 32, "num_kv_heads": 8},
+    )
+)
+def benchmark(provider, head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype, device, batch_size, seq_len, num_q_heads, num_kv_heads):
+    print(f"provider: {provider}, head_size: {head_size}, rotary_dim: {rotary_dim}, max_position_embeddings: {max_position_embeddings}, base: {base}, is_neox_style: {is_neox_style}, dtype: {dtype}, device: {device}, batch_size: {batch_size}, seq_len: {seq_len}, num_q_heads: {num_q_heads}, num_kv_heads: {num_kv_heads}")
+    
+    rope_forward = None
+
+    if provider == "vllm":
+        rope = vLLMRotaryEmbedding(head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype).to(device)
+        rope_forward = rope.forward_cuda
+    elif provider == "flashinfer":
+        rope = FlashInferRotaryEmbedding(head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype).to(device)
+        rope_forward = rope.forward_cuda
+    elif provider == "native":
+        rope = vLLMRotaryEmbedding(head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype).to(device)
+        rope_forward = rope.forward_native
+
+    pos_ids = torch.arange(seq_len, device=device).repeat(batch_size)
+    query = torch.randn(batch_size * seq_len, num_q_heads * head_size, dtype=dtype, device=device)
+    key = torch.randn(batch_size * seq_len, num_kv_heads * head_size, dtype=dtype, device=device)
+
+    quantiles = [0.5, 0.2, 0.8]
+    ms, min_ms, max_ms = triton.testing.do_bench(lambda: rope_forward(pos_ids, query, key), quantiles=quantiles)
 
+    return ms, min_ms, max_ms
 
 if __name__ == "__main__":
-    main()
+    benchmark.run(print_data=True, show_plots=True, save_path="rope_benchmark.png")
diff --git a/csrc/flashinfer_ops.cu b/csrc/flashinfer_ops.cu
index e7bdcd5a7a..59234182e1 100644
--- a/csrc/flashinfer_ops.cu
+++ b/csrc/flashinfer_ops.cu
@@ -153,7 +153,7 @@ void apply_llama31_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, a
                                 int64_t cuda_stream);
 
 void apply_rope_pos_ids_cos_sin_cache(at::Tensor q, at::Tensor k, at::Tensor q_rope,
-                                      at::Tensor k_rope, at::Tensor cos_cache, at::Tensor sin_cache,
+                                      at::Tensor k_rope, at::Tensor cos_sin_cache,
                                       at::Tensor pos_ids, bool interleave, int64_t cuda_stream);
 
 //========== sampling ==========
diff --git a/csrc/flashinfer_rope_ops.cu b/csrc/flashinfer_rope_ops.cu
index 41fb284a22..f447aba719 100644
--- a/csrc/flashinfer_rope_ops.cu
+++ b/csrc/flashinfer_rope_ops.cu
@@ -37,7 +37,7 @@ void apply_llama31_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, a
                                 int64_t cuda_stream);
 
 void apply_rope_pos_ids_cos_sin_cache(at::Tensor q, at::Tensor k, at::Tensor q_rope,
-                                      at::Tensor k_rope, at::Tensor cos_cache, at::Tensor sin_cache,
+                                      at::Tensor k_rope, at::Tensor cos_sin_cache,
                                       at::Tensor pos_ids, bool interleave, int64_t cuda_stream);
 
 PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
diff --git a/csrc/rope.cu b/csrc/rope.cu
index 3f10357bfe..b4eefcc2c9 100644
--- a/csrc/rope.cu
+++ b/csrc/rope.cu
@@ -105,26 +105,24 @@ void apply_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, at::Tenso
 }
 
 void apply_rope_pos_ids_cos_sin_cache(at::Tensor q, at::Tensor k, at::Tensor q_rope,
-                                      at::Tensor k_rope, at::Tensor cos_cache, at::Tensor sin_cache,
+                                      at::Tensor k_rope, at::Tensor cos_sin_cache,
                                       at::Tensor pos_ids, bool interleave, int64_t cuda_stream) {
   CHECK_LAST_DIM_CONTIGUOUS(q);
   CHECK_LAST_DIM_CONTIGUOUS(k);
-  CHECK_INPUT(cos_cache);
-  CHECK_INPUT(sin_cache);
+  CHECK_INPUT(cos_sin_cache);
   CHECK_INPUT(pos_ids);
   auto device = q.device();
   CHECK_EQ(k.device(), device);
-  CHECK_EQ(cos_cache.device(), device);
-  CHECK_EQ(sin_cache.device(), device);
+  CHECK_EQ(cos_sin_cache.device(), device);
   CHECK_EQ(pos_ids.device(), device);
   CHECK_DIM(3, q);          // q: (nnz, H_Q, D)
   CHECK_DIM(3, k);          // k: (nnz, H_K, D)
-  CHECK_DIM(2, cos_cache);  // cos_cache: (max_seq_len, D)
-  CHECK_DIM(2, sin_cache);  // sin_cache: (max_seq_len, D)
+  // cos_sin_cache: (max_seq_len, R)
+  // First half of R is cos, second half is sin
+  CHECK_DIM(2, cos_sin_cache);
   CHECK_EQ(q.size(0), k.size(0));
   CHECK_EQ(q.size(2), k.size(2));
-  unsigned int rotary_dim = cos_cache.size(1);
-  CHECK_EQ(sin_cache.size(1), rotary_dim);
+  unsigned int rotary_dim = cos_sin_cache.size(1);
   unsigned int num_qo_heads = q.size(1);
   unsigned int num_kv_heads = k.size(1);
   unsigned int head_dim = q.size(2);
@@ -143,7 +141,7 @@ void apply_rope_pos_ids_cos_sin_cache(at::Tensor q, at::Tensor k, at::Tensor q_r
     cudaError_t status = BatchQKApplyRotaryPosIdsCosSinCache(
         static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
         static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<float*>(cos_cache.data_ptr()), static_cast<float*>(sin_cache.data_ptr()),
+        static_cast<float*>(cos_sin_cache.data_ptr()),
         static_cast<int32_t*>(pos_ids.data_ptr()), nnz, num_qo_heads, num_kv_heads, rotary_dim,
         head_dim, q_stride_n, q_stride_h, k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h,
         k_rope_stride_n, k_rope_stride_h, interleave, stream);
diff --git a/flashinfer/rope.py b/flashinfer/rope.py
index 62d9c39b92..e587f0b603 100644
--- a/flashinfer/rope.py
+++ b/flashinfer/rope.py
@@ -196,8 +196,7 @@ def _apply_rope_pos_ids_cos_sin_cache(
     k: torch.Tensor,
     q_rope: torch.Tensor,
     k_rope: torch.Tensor,
-    cos_cache: torch.Tensor,
-    sin_cache: torch.Tensor,
+    cos_sin_cache: torch.Tensor,
     pos_ids: torch.Tensor,
     interleave: bool,
 ) -> None:
@@ -208,8 +207,7 @@ def _apply_rope_pos_ids_cos_sin_cache(
             k,
             q_rope,
             k_rope,
-            cos_cache,
-            sin_cache,
+            cos_sin_cache,
             pos_ids,
             interleave,
             get_cuda_stream(device),
@@ -283,7 +281,6 @@ def _fake_apply_llama31_rope_pos_ids(
 ) -> None:
     pass
 
-
 def apply_rope_inplace(
     q: torch.Tensor,
     k: torch.Tensor,
@@ -1003,99 +1000,115 @@ def apply_llama31_rope_pos_ids(
 
 
 def apply_rope_with_cos_sin_cache(
-    q: torch.Tensor,
-    k: torch.Tensor,
-    cos_cache: torch.Tensor,
-    sin_cache: torch.Tensor,
-    pos_ids: torch.Tensor,
-    interleave: bool = False,
+    positions: torch.Tensor,
+    query: torch.Tensor, 
+    key: torch.Tensor,
+    head_size: int,
+    cos_sin_cache: torch.Tensor,
+    is_neox: bool = True,
 ) -> Tuple[torch.Tensor, torch.Tensor]:
-    r"""Apply rotary embedding to keys and queries with precomputed cos/sin values.
+    r"""
+    Apply rotary embedding to keys and queries with precomputed cos/sin values.
+    This is designed to be compatible with the SGL/vLLM implementation.
 
     Parameters
     ----------
-    q : torch.Tensor
-        Query tensor, shape: ``(nnz, num_q_heads, head_dim)``.
-    k : torch.Tensor
-        Key tensor, shape: ``(nnz, num_k_heads, head_dim)``.
-    cos_cache : torch.Tensor
-        Cosine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
-    sin_cache : torch.Tensor
-        Sine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
-    pos_ids : torch.Tensor
+    positions : torch.Tensor
         Position indices, shape: ``(nnz)``.
-    interleave : bool
-        Whether to use interleaved layout in the last dimension, default: ``False``.
+    query : torch.Tensor
+        Query tensor, shape: ``(nnz, num_q_heads * head_size)``.
+    key : torch.Tensor
+        Key tensor, shape: ``(nnz, num_k_heads * head_size)``.
+    cos_sin_cache : torch.Tensor
+        Cosine and Sine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
+        Cosine is the first half and Sine is the second half on rotary_dim.
+    is_neox : bool
+        Whether to use Neox style RoPE, default: ``True``.
+
+        * If ``True``, the last dimension of the query/key tensor is not interleaved, i.e.,
+          we rorate the first half dimensions ``([..., :head_dim//2])`` and the second half
+          dimensions ``([..., head_dim//2:])``.
 
-        * If ``True``, the last dimension of the query/key tensor is interleaved, i.e.,
+        * If ``False``, the last dimension of the query/key tensor is interleaved, i.e.,
           we rotate the even dimensions ``([..., ::2])`` and odd dimensions ``([..., 1::2])``.
 
-        * If ``False``, the last dimension of the query/key tensor is not interleaved, i.e.,
-          we rotate the first half dimensions ``([..., :head_dim//2])`` and the second half
-          dimensions ``([..., head_dim//2:])``.
-
     Returns
     -------
-    q_rope : torch.Tensor
-        The rotated query tensor, shape: ``(nnz, num_q_heads, head_dim)``.
-    k_rope : torch.Tensor
-        The rotated key tensor, shape: ``(nnz, num_k_heads, head_dim)``.
+    query_out : torch.Tensor
+        The rotated query tensor, shape: ``(nnz, num_q_heads * head_size)``.
+    key_out : torch.Tensor
+        The rotated key tensor, shape: ``(nnz, num_k_heads * head_size)``.
 
     Note
     ----
     The rotary dimension is determined by the cosine cache and sine cache.
     """
-    if cos_cache.dtype != torch.float32 or sin_cache.dtype != torch.float32:
-        raise ValueError("cos_cache and sin_cache should be float32")
-    q_rope = torch.empty_like(q)
-    k_rope = torch.empty_like(k)
+    if cos_sin_cache.dtype != torch.float32:
+        raise ValueError("cos_sin_cache should be float32")
+    
+    query_out = torch.empty_like(query)
+    key_out = torch.empty_like(key)
+
     _apply_rope_pos_ids_cos_sin_cache(
-        q, k, q_rope, k_rope, cos_cache, sin_cache, pos_ids, interleave
+        q=query.view(query.shape[0], -1, head_size),
+        k=key.view(key.shape[0], -1, head_size),
+        q_rope=query_out.view(query_out.shape[0], -1, head_size),
+        k_rope=key_out.view(key_out.shape[0], -1, head_size),
+        cos_sin_cache=cos_sin_cache,
+        pos_ids=positions,
+        interleave=(not is_neox)
     )
-    return q_rope, k_rope
 
+    return query_out, key_out
 
 def apply_rope_with_cos_sin_cache_inplace(
-    q: torch.Tensor,
-    k: torch.Tensor,
-    cos_cache: torch.Tensor,
-    sin_cache: torch.Tensor,
-    pos_ids: torch.Tensor,
-    interleave: bool = False,
+    positions: torch.Tensor,
+    query: torch.Tensor, 
+    key: torch.Tensor,
+    head_size: int,
+    cos_sin_cache: torch.Tensor,
+    is_neox: bool = True,
 ) -> None:
-    r"""Apply rotary embedding to keys and queries with precomputed cos/sin values.
-    The result is stored in the input tensors inplace.
+    r"""
+    Apply rotary embedding to keys and queries with precomputed cos/sin values.
+    This is designed to be compatible with the SGL/vLLM implementation.
+    The result is inplace applied to the input tensors.
 
     Parameters
     ----------
-    q : torch.Tensor
-        Query tensor, shape: ``(nnz, num_q_heads, head_dim)``.
-    k : torch.Tensor
-        Key tensor, shape: ``(nnz, num_k_heads, head_dim)``.
-    cos_cache : torch.Tensor
-        Cosine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
-        Expect float32 data type.
-    sin_cache : torch.Tensor
-        Sine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
-        Expect float32 data type.
-    pos_ids : torch.Tensor
+    positions : torch.Tensor
         Position indices, shape: ``(nnz)``.
-    interleave : bool
-        Whether to use interleaved layout in the last dimension, default: ``False``.
-
-        * If ``True``, the last dimension of the query/key tensor is interleaved, i.e.,
-          we rotate the even dimensions ``([..., ::2])`` and odd dimensions ``([..., 1::2])``.
-
-        * If ``False``, the last dimension of the query/key tensor is not interleaved, i.e.,
-          we rotate the first half dimensions ``([..., :head_dim//2])`` and the second half
+    query : torch.Tensor
+        Query tensor, shape: ``(nnz, num_q_heads * head_size)``.
+    key : torch.Tensor
+        Key tensor, shape: ``(nnz, num_k_heads * head_size)``.
+    cos_sin_cache : torch.Tensor
+        Cosine and Sine cache tensor, shape: ``(max_seq_len, rotary_dim)``.
+        Cosine is the first half and Sine is the second half on rotary_dim.
+    is_neox : bool
+        Whether to use Neox style RoPE, default: ``True``.
+
+        * If ``True``, the last dimension of the query/key tensor is not interleaved, i.e.,
+          we rorate the first half dimensions ``([..., :head_dim//2])`` and the second half
           dimensions ``([..., head_dim//2:])``.
 
+        * If ``False``, the last dimension of the query/key tensor is interleaved, i.e.,
+          we rotate the even dimensions ``([..., ::2])`` and odd dimensions ``([..., 1::2])``.
     Note
     ----
     The rotary dimension is determined by the cosine cache and sine cache.
     """
-    if cos_cache.dtype != torch.float32 or sin_cache.dtype != torch.float32:
-        raise ValueError("cos_cache and sin_cache should be float32")
+    if cos_sin_cache.dtype != torch.float32:
+        raise ValueError("cos_sin_cache should be float32")
+    
+    # pass q_rope and k_rope as q and k to perform inplace operation
     _apply_rope_pos_ids_cos_sin_cache(
-        q, k, q, k, cos_cache, sin_cache, pos_ids, interleave
+        q=query.view(query.shape[0], -1, head_size),
+        k=key.view(key.shape[0], -1, head_size),
+        q_rope=query.view(query.shape[0], -1, head_size),
+        k_rope=key.view(key.shape[0], -1, head_size),
+        cos_sin_cache=cos_sin_cache,
+        pos_ids=positions,
+        interleave=(not is_neox)
     )
+
diff --git a/include/flashinfer/pos_enc.cuh b/include/flashinfer/pos_enc.cuh
index 6c41849668..196b5f33f4 100644
--- a/include/flashinfer/pos_enc.cuh
+++ b/include/flashinfer/pos_enc.cuh
@@ -168,11 +168,48 @@ __device__ __forceinline__ vec_t<float, vec_size> vec_apply_llama_rope_cos_sin_i
   return vec;
 }
 
+
+/*
+HACK (ByronHsu): in the interleave mode with cos_sin_cache, we actually only use the first half of cos and sin
+
+For example,
+In the below example, the vec_size is 4
+the computation in the kernel is:
+    [x1, x2, x3, x4...] * [cos1, cos1, cos2, cos2] + [-x2, x1, -x4, x3...] * [sin1, sin1, sin2, sin2]
+the data we loaded are:
+    - loaded vec = [x1, x2, x3, x4]
+    - loaded cos = [cos1, cos2, cos3, cos4]
+    - loaded sin = [sin1, sin2, sin3, sin4]
+But only the first half of cos and sin is used in the computation.
+
+However, we argue the additional overhead is acceptable:
+    1. loading additional elements of cos and sin is not adding much overhead. The arithmetic intensity 
+       is the same as non-interleave mode. Each elements of cos and sin is load twice
+    2. we don't want two code paths of cos and sin vector for interleave and non-interleave mode.
+*/
+template <uint32_t vec_size, uint32_t bdx, typename T>
+__device__ __forceinline__ vec_t<float, vec_size> vec_apply_llama_rope_cos_sin_interleave_reuse_half(
+    const T* x, const vec_t<float, vec_size>& cos, const vec_t<float, vec_size>& sin,
+    const uint32_t rotary_dim = vec_size * bdx) {
+  vec_t<float, vec_size> vec, vec_before;
+  vec.cast_load(x + threadIdx.x * vec_size);
+
+  if (threadIdx.x * vec_size < rotary_dim) {
+    vec_before = vec;
+#pragma unroll
+    for (uint32_t i = 0; i < vec_size; ++i) {
+      // i / 2 is to get the index of the first half of cos and sin
+      vec[i] = vec[i] * cos[i / 2] + ((i % 2 == 0) ? -vec_before[i ^ 1] : vec_before[i ^ 1]) * sin[i / 2];
+    }
+  }
+  return vec;
+}
+
 template <bool interleave, uint32_t head_dim, uint32_t vec_size, uint32_t bdx, typename DType,
           typename IdType>
 __global__ void BatchQKApplyRotaryPosIdsCosSinCacheHeadParallelismKernel(
-    DType* q, DType* k, DType* q_rope, DType* k_rope, float* __restrict__ cos_cache,
-    float* __restrict__ sin_cache, IdType* __restrict__ pos_ids, uint32_t nnz,
+    DType* q, DType* k, DType* q_rope, DType* k_rope, float* __restrict__ cos_sin_cache, 
+    IdType* __restrict__ pos_ids, uint32_t nnz,
     uint32_t num_qo_heads, uint32_t num_kv_heads, uint32_t rotary_dim, size_t q_stride_n,
     size_t q_stride_h, size_t k_stride_n, size_t k_stride_h, size_t q_rope_stride_n,
     size_t q_rope_stride_h, size_t k_rope_stride_n, size_t k_rope_stride_h) {
@@ -185,9 +222,24 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheHeadParallelismKernel(
     const uint32_t idx = bx * bdy + ty;
     const IdType pos = pos_ids[idx];
 
+    const int half_rotary_dim = rotary_dim / 2;
+
+    // 1. if interleave: 
+    //  - cos = cos_sin_cache[pos_id][tx * vec_size // 2]
+    //  - sin = cos_sin_cache[pos_id][(rot_dim // 2) + tx * vec_size // 2]
+    // 2. if not interleave
+    //  - cos = cos_cache[pos_id][(tx * vec_size) % (rot_dim // 2)]
+    //  - sin = sin_cache[pos_id][(rot_dim // 2) + (tx * vec_size) % (rot_dim // 2)]
     if (tx * vec_size < rotary_dim) {
-      cos.load(cos_cache + pos * rotary_dim + tx * vec_size);
-      sin.load(sin_cache + pos * rotary_dim + tx * vec_size);
+      int sin_offset = rotary_dim / 2;
+      int vec_idx;
+      if constexpr (interleave) {
+        vec_idx = (tx * vec_size) / 2;  // Force integer division
+      } else {
+        vec_idx = (tx * vec_size) % half_rotary_dim;  // Use half_rotary_dim
+      }
+      cos.load(cos_sin_cache + (pos * rotary_dim) + vec_idx);
+      sin.load(cos_sin_cache + (pos * rotary_dim) + (sin_offset + vec_idx));
     }
 
     if (by < num_qo_heads) {
@@ -197,7 +249,7 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheHeadParallelismKernel(
           q_rope + get_elem_offset_impl(idx, qo_head_idx, 0, q_rope_stride_n, q_rope_stride_h);
       vec_t<float, vec_size> q_vec;
       if constexpr (interleave) {
-        q_vec = vec_apply_llama_rope_cos_sin_interleave<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
+        q_vec = vec_apply_llama_rope_cos_sin_interleave_reuse_half<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
       } else {
         q_vec = vec_apply_llama_rope_cos_sin<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
       }
@@ -209,7 +261,7 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheHeadParallelismKernel(
           k_rope + get_elem_offset_impl(idx, kv_head_idx, 0, k_rope_stride_n, k_rope_stride_h);
       vec_t<float, vec_size> k_vec;
       if constexpr (interleave) {
-        k_vec = vec_apply_llama_rope_cos_sin_interleave<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
+        k_vec = vec_apply_llama_rope_cos_sin_interleave_reuse_half<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
       } else {
         k_vec = vec_apply_llama_rope_cos_sin<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
       }
@@ -221,8 +273,8 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheHeadParallelismKernel(
 template <bool interleave, uint32_t head_dim, uint32_t vec_size, uint32_t bdx, typename DType,
           typename IdType>
 __global__ void BatchQKApplyRotaryPosIdsCosSinCacheKernel(
-    DType* q, DType* k, DType* q_rope, DType* k_rope, float* __restrict__ cos_cache,
-    float* __restrict__ sin_cache, IdType* __restrict__ pos_ids, uint32_t nnz,
+    DType* q, DType* k, DType* q_rope, DType* k_rope, float* __restrict__ cos_sin_cache,
+    IdType* __restrict__ pos_ids, uint32_t nnz,
     uint32_t num_qo_heads, uint32_t num_kv_heads, uint32_t rotary_dim, size_t q_stride_n,
     size_t q_stride_h, size_t k_stride_n, size_t k_stride_h, size_t q_rope_stride_n,
     size_t q_rope_stride_h, size_t k_rope_stride_n, size_t k_rope_stride_h) {
@@ -233,12 +285,27 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheKernel(
   if (bx * bdy + ty < nnz) {
     const uint32_t idx = bx * bdy + ty;
     const IdType pos = pos_ids[idx];
-
+    const int half_rotary_dim = rotary_dim / 2;
+
+    // 1. if interleave: 
+    //  - cos = cos_sin_cache[pos_id][tx * vec_size // 2]
+    //  - sin = cos_sin_cache[pos_id][(rot_dim // 2) + tx * vec_size // 2]
+    // 2. if not interleave
+    //  - cos = cos_cache[pos_id][(tx * vec_size) % (rot_dim // 2)]
+    //  - sin = sin_cache[pos_id][(rot_dim // 2) + (tx * vec_size) % (rot_dim // 2)]
     if (tx * vec_size < rotary_dim) {
-      cos.load(cos_cache + pos * rotary_dim + tx * vec_size);
-      sin.load(sin_cache + pos * rotary_dim + tx * vec_size);
+      int sin_offset = rotary_dim / 2;
+      int vec_idx;
+      if constexpr (interleave) {
+        vec_idx = (tx * vec_size) / 2;  // Force integer division
+      } else {
+        vec_idx = (tx * vec_size) % half_rotary_dim;  // Use half_rotary_dim
+      }
+      cos.load(cos_sin_cache + (pos * rotary_dim) + vec_idx);
+      sin.load(cos_sin_cache + (pos * rotary_dim) + (sin_offset + vec_idx));
     }
-
+    
+  // not to unroll the loop, because num head might be large and might lead to worse performance
 #pragma unroll 1
     for (uint32_t qo_head_idx = 0; qo_head_idx < num_qo_heads; ++qo_head_idx) {
       DType* q_ptr = q + get_elem_offset_impl(idx, qo_head_idx, 0, q_stride_n, q_stride_h);
@@ -246,7 +313,7 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheKernel(
           q_rope + get_elem_offset_impl(idx, qo_head_idx, 0, q_rope_stride_n, q_rope_stride_h);
       vec_t<float, vec_size> q_vec;
       if constexpr (interleave) {
-        q_vec = vec_apply_llama_rope_cos_sin_interleave<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
+        q_vec = vec_apply_llama_rope_cos_sin_interleave_reuse_half<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
       } else {
         q_vec = vec_apply_llama_rope_cos_sin<vec_size, bdx>(q_ptr, cos, sin, rotary_dim);
       }
@@ -260,7 +327,7 @@ __global__ void BatchQKApplyRotaryPosIdsCosSinCacheKernel(
           k_rope + get_elem_offset_impl(idx, kv_head_idx, 0, k_rope_stride_n, k_rope_stride_h);
       vec_t<float, vec_size> k_vec;
       if constexpr (interleave) {
-        k_vec = vec_apply_llama_rope_cos_sin_interleave<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
+        k_vec = vec_apply_llama_rope_cos_sin_interleave_reuse_half<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
       } else {
         k_vec = vec_apply_llama_rope_cos_sin<vec_size, bdx>(k_ptr, cos, sin, rotary_dim);
       }
@@ -503,7 +570,7 @@ __global__ void BatchQKApplyRotaryKernel(
 
 template <typename DType, typename IdType>
 cudaError_t BatchQKApplyRotaryPosIdsCosSinCache(
-    DType* q, DType* k, DType* q_rope, DType* k_rope, float* cos_cache, float* sin_cache,
+    DType* q, DType* k, DType* q_rope, DType* k_rope, float* cos_sin_cache,
     IdType* pos_ids, uint32_t nnz, uint32_t num_qo_heads, uint32_t num_kv_heads,
     uint32_t rotary_dim, uint32_t head_dim, size_t q_stride_n, size_t q_stride_h, size_t k_stride_n,
     size_t k_stride_h, size_t q_rope_stride_n, size_t q_rope_stride_h, size_t k_rope_stride_n,
@@ -515,17 +582,21 @@ cudaError_t BatchQKApplyRotaryPosIdsCosSinCache(
 
   DISPATCH_INTERLEAVE(interleave, INTERLEAVE, {
     DISPATCH_HEAD_DIM(head_dim, HEAD_DIM, {
+      // operate on 16 Bytes at a time
       constexpr uint32_t vec_size = std::max(16 / sizeof(DType), HEAD_DIM / 32);
+      // how many threads needed per head_dim
       constexpr uint32_t bdx = HEAD_DIM / vec_size;
+      // how many threads needed per block
       uint32_t num_threads = std::max(128U, bdx);
+      // how many tokens can we process in a block
       uint32_t bdy = num_threads / bdx;
+      // how many blocks needed to process all tokens
       uint32_t nblks_x = (nnz + bdy - 1) / bdy;
       void* args[] = {(void*)&q,
                       (void*)&k,
                       (void*)&q_rope,
                       (void*)&k_rope,
-                      (void*)&cos_cache,
-                      (void*)&sin_cache,
+                      (void*)&cos_sin_cache,
                       (void*)&pos_ids,
                       (void*)&nnz,
                       (void*)&num_qo_heads,
diff --git a/include/flashinfer/utils.cuh b/include/flashinfer/utils.cuh
index da3b1b0a2e..ee7d2cb10c 100644
--- a/include/flashinfer/utils.cuh
+++ b/include/flashinfer/utils.cuh
@@ -166,6 +166,7 @@
     }                                                         \
   }
 
+// convert head_dim to compile-time constant
 #define DISPATCH_HEAD_DIM(head_dim, HEAD_DIM, ...)     \
   switch (head_dim) {                                  \
     case 64: {                                         \
diff --git a/tests/rope_reference.py b/tests/rope_reference.py
index 7a176692aa..7f4e81a790 100644
--- a/tests/rope_reference.py
+++ b/tests/rope_reference.py
@@ -6,13 +6,12 @@
 # the top-level of this source tree.
 
 import math
-from typing import Tuple
+from typing import Tuple, Optional, Union
 
 # Copyright (c) Meta Platforms, Inc. and affiliates.
 # This software may be used and distributed in accordance with the terms of the Llama 3 Community License Agreement.
 import torch
 
-
 def apply_scaling(freqs: torch.Tensor):
     # Values obtained from grid search
     scale_factor = 8
@@ -97,3 +96,114 @@ def generate_cos_sin_f32_cache(
     sin_cache = torch.sin(args)
     cos_cache = torch.cos(args)
     return cos_cache, sin_cache
+
+# The following code is from the vLLM's implementation of RoPE.
+# https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/rotary_embedding.py
+    
+class RotaryEmbedding(torch.nn.Module):
+    def __init__(
+        self,
+        head_size: int,
+        rotary_dim: int,
+        max_position_embeddings: int,
+        base: int,
+        is_neox_style: bool,
+        dtype: torch.dtype,
+    ) -> None:
+        super().__init__()
+        self.head_size = head_size
+        self.rotary_dim = rotary_dim
+        self.max_position_embeddings = max_position_embeddings
+        self.base = base
+        self.is_neox_style = is_neox_style
+        self.dtype = dtype
+
+        cache = self._compute_cos_sin_cache()
+        self.cos_sin_cache: torch.Tensor
+        self.register_buffer("cos_sin_cache", cache, persistent=False)
+
+    def _compute_inv_freq(self, base: Union[int, float]) -> torch.Tensor:
+        inv_freq = 1.0 / (base**(torch.arange(
+            0, self.rotary_dim, 2, dtype=torch.float) / self.rotary_dim))
+        return inv_freq
+
+    def _compute_cos_sin_cache(self) -> torch.Tensor:
+        """Compute the cos and sin cache."""
+        inv_freq = self._compute_inv_freq(self.base)
+        t = torch.arange(self.max_position_embeddings, dtype=torch.float)
+
+        freqs = torch.einsum("i,j -> ij", t, inv_freq)
+        cos = freqs.cos()
+        sin = freqs.sin()
+        cache = torch.cat((cos, sin), dim=-1)
+        return cache
+
+    def _apply_rotary_emb(
+        self,
+        x: torch.Tensor,
+        cos: torch.Tensor,
+        sin: torch.Tensor,
+        is_neox_style: bool,
+    ) -> torch.Tensor:
+        """
+        Args:
+            x: [num_tokens, num_heads, head_size]
+            cos: [num_tokens, head_size // 2]
+            sin: [num_tokens, head_size // 2]
+            is_neox_style: Whether to use the Neox-style or GPT-J-style rotary
+                positional embeddings.
+        """
+        cos = cos.unsqueeze(-2).to(x.dtype)
+        sin = sin.unsqueeze(-2).to(x.dtype)
+        if is_neox_style:
+            x1, x2 = torch.chunk(x, 2, dim=-1)
+        else:
+            x1 = x[..., ::2]
+            x2 = x[..., 1::2]
+        o1 = x1 * cos - x2 * sin
+        o2 = x2 * cos + x1 * sin
+        if is_neox_style:
+            return torch.cat((o1, o2), dim=-1)
+        else:
+            return torch.stack((o1, o2), dim=-1).flatten(-2)
+
+    def forward_native(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        offsets: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """A PyTorch-native implementation of forward()."""
+        if offsets is not None:
+            positions = positions + offsets
+
+        positions = positions.flatten()
+        num_tokens = positions.shape[0]
+        cos_sin = self.cos_sin_cache.index_select(0, positions)
+
+        # Note: the is different from the vLLM's implementation,
+        # We added float32 conversion because float32 is required for the rotary embedding to work correctly for long contexts
+        query = query.to(torch.float32)
+        key = key.to(torch.float32)
+        
+        cos, sin = cos_sin.chunk(2, dim=-1)
+
+        query_shape = query.shape
+        query = query.view(num_tokens, -1, self.head_size)
+        query_rot = query[..., :self.rotary_dim]
+        query_pass = query[..., self.rotary_dim:]
+        query_rot = self._apply_rotary_emb(query_rot, cos, sin, self.is_neox_style)
+        query = torch.cat((query_rot, query_pass), dim=-1).reshape(query_shape)
+
+        key_shape = key.shape
+        key = key.view(num_tokens, -1, self.head_size)
+        key_rot = key[..., :self.rotary_dim]
+        key_pass = key[..., self.rotary_dim:]
+        key_rot = self._apply_rotary_emb(key_rot, cos, sin, self.is_neox_style)
+        key = torch.cat((key_rot, key_pass), dim=-1).reshape(key_shape)
+
+        query = query.to(self.dtype)
+        key = key.to(self.dtype)
+        return query, key
+    
diff --git a/tests/test_rope.py b/tests/test_rope.py
index 7fab236177..d60c3dee41 100644
--- a/tests/test_rope.py
+++ b/tests/test_rope.py
@@ -145,6 +145,7 @@ def test_rope(
 @pytest.mark.parametrize("llama_version", ["llama", "llama31"])
 @pytest.mark.parametrize("partial_rotary_factor", [0.25, 0.5, 0.75, 1.0])
 @pytest.mark.parametrize("inplace", [False, True])
+@pytest.mark.parametrize("interleave", [True, False])
 def test_rope_pos_ids(
     batch_size,
     qkv_len,
@@ -155,6 +156,7 @@ def test_rope_pos_ids(
     llama_version,
     partial_rotary_factor,
     inplace,
+    interleave,
 ):
     rotary_dim = int(head_dim * partial_rotary_factor)
     nnz = batch_size * qkv_len
@@ -189,7 +191,7 @@ def test_rope_pos_ids(
                 indptr,
                 offsets,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=1e4,
             )
             q_rope, k_rope = q, k
@@ -198,7 +200,7 @@ def test_rope_pos_ids(
                 k_clone,
                 pos_ids,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=1e4,
             )
             q_rope_pos_ids, k_rope_pos_ids = q_clone, k_clone
@@ -209,12 +211,12 @@ def test_rope_pos_ids(
                 indptr,
                 offsets,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=1e4,
             )
 
             q_rope_pos_ids, k_rope_pos_ids = flashinfer.apply_rope_pos_ids(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=True, rope_theta=1e4
+                q, k, pos_ids, rotary_dim=rotary_dim, interleave=interleave, rope_theta=1e4
             )
     else:
         if inplace:
@@ -225,7 +227,7 @@ def test_rope_pos_ids(
                 indptr,
                 offsets,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=5e5,
             )
             q_rope, k_rope = q, k
@@ -234,7 +236,7 @@ def test_rope_pos_ids(
                 k_clone,
                 pos_ids,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=5e5,
             )
             q_rope_pos_ids, k_rope_pos_ids = q_clone, k_clone
@@ -245,109 +247,78 @@ def test_rope_pos_ids(
                 indptr,
                 offsets,
                 rotary_dim=rotary_dim,
-                interleave=True,
+                interleave=interleave,
                 rope_theta=5e5,
             )
 
             q_rope_pos_ids, k_rope_pos_ids = flashinfer.apply_llama31_rope_pos_ids(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=True, rope_theta=5e5
+                q, k, pos_ids, rotary_dim=rotary_dim, interleave=interleave, rope_theta=5e5
             )
 
     # compare
     torch.testing.assert_close(q_rope_pos_ids, q_rope, rtol=1e-3, atol=1e-3)
     torch.testing.assert_close(k_rope_pos_ids, k_rope, rtol=1e-3, atol=1e-3)
 
-
-@pytest.mark.parametrize("batch_size", [1, 19, 99, 989])
-@pytest.mark.parametrize("qkv_len", [1, 4, 19, 204])
-@pytest.mark.parametrize("num_qo_heads", [8, 16])
-@pytest.mark.parametrize("num_kv_heads", [8])
-@pytest.mark.parametrize("offset", [0, 15, 99])
-@pytest.mark.parametrize("head_dim", [64, 128, 256])
-@pytest.mark.parametrize("llama_version", ["llama", "llama31"])
-@pytest.mark.parametrize("partial_rotary_factor", [0.25, 0.5, 0.75, 1.0])
-@pytest.mark.parametrize("inplace", [False, True])
+class FlashInferRotaryEmbedding(RotaryEmbedding):
+    
+    def forward_cuda(
+        self,
+        positions: torch.Tensor,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        offsets: Optional[torch.Tensor] = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        flashinfer.apply_rope_with_cos_sin_cache_inplace(
+            positions=positions,
+            query=query,
+            key=key,
+            head_size=self.head_size,
+            cos_sin_cache=self.cos_sin_cache,
+            is_neox=self.is_neox_style,
+        )
+        return query, key
+    
+@pytest.mark.parametrize(
+    "head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype, device, batch_size, seq_len, num_q_heads, num_kv_heads",
+    [
+        (64, 64, 32, 8000, True, torch.bfloat16, "cuda", 32, 32, 1, 1),
+        (256, 128, 4096, 10000, True, torch.bfloat16, "cuda", 2, 512, 4, 2),
+        (64, 32, 2048, 8432, True, torch.bfloat16, "cuda", 2, 199, 4, 1),
+        (64, 64, 32, 8000, False, torch.bfloat16, "cuda", 32, 32, 1, 1),
+        (64, 64, 32, 8000, False, torch.bfloat16, "cuda", 32, 32, 1, 1),
+        (256, 128, 4096, 9231, False, torch.bfloat16, "cuda", 3, 231, 4, 2),
+    ]
+)
 def test_rope_cos_sin_cache(
-    batch_size,
-    qkv_len,
-    num_qo_heads,
-    num_kv_heads,
-    offset,
-    head_dim,
-    llama_version,
-    partial_rotary_factor,
-    inplace,
+    head_size: int,
+    rotary_dim: int,
+    max_position_embeddings: int,
+    base: int,
+    is_neox_style: bool,
+    dtype: torch.dtype,
+    device: str,
+    batch_size: int,
+    seq_len: int,
+    num_q_heads: int,
+    num_kv_heads: int,
 ):
-    rotary_dim = int(head_dim * partial_rotary_factor)
-    nnz = batch_size * qkv_len
-    qkv_packed = torch.randn(
-        nnz,
-        (num_qo_heads + 2 * num_kv_heads) * head_dim,
-        dtype=torch.float16,
-        device="cuda:0",
-    )
-    q = qkv_packed[:, : num_qo_heads * head_dim].reshape(nnz, num_qo_heads, head_dim)
-    k = qkv_packed[
-        :, num_qo_heads * head_dim : (num_qo_heads + num_kv_heads) * head_dim
-    ].reshape(nnz, num_kv_heads, head_dim)
-    pos_ids = torch.cat(
-        [
-            torch.arange(offset, qkv_len + offset, dtype=torch.int32)
-            for _ in range(batch_size)
-        ]
-    ).to("cuda:0")
-
-    if llama_version == "llama":
-        cos_cache, sin_cache = generate_cos_sin_f32_cache(4096, rotary_dim, theta=1e4)
-    else:
-        cos_cache, sin_cache = generate_cos_sin_f32_cache(
-            4096, rotary_dim, theta=5e5, use_scaled=True
-        )
-    cos_cache = cos_cache.to("cuda:0")
-    sin_cache = sin_cache.to("cuda:0")
-
-    if inplace:
-        q_clone, k_clone = q.clone(), k.clone()
-
-    if llama_version == "llama":
-        if inplace:
-            flashinfer.apply_rope_pos_ids_inplace(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=False
-            )
-            q_rope, k_rope = q, k
-        else:
-            q_rope, k_rope = flashinfer.apply_rope_pos_ids(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=False
-            )
-    else:
-        if inplace:
-            flashinfer.apply_llama31_rope_pos_ids_inplace(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=False
-            )
-            q_rope, k_rope = q, k
-        else:
-            q_rope, k_rope = flashinfer.apply_llama31_rope_pos_ids(
-                q, k, pos_ids, rotary_dim=rotary_dim, interleave=False
-            )
+    rope_ref = RotaryEmbedding(head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype).to(device)
+    rope_flashinfer = FlashInferRotaryEmbedding(head_size, rotary_dim, max_position_embeddings, base, is_neox_style, dtype).to(device)
 
-    if inplace:
-        flashinfer.apply_rope_with_cos_sin_cache_inplace(
-            q_clone, k_clone, cos_cache, sin_cache, pos_ids, interleave=False
-        )
-        q_rope_cos_sin_cache, k_rope_cos_sin_cache = q_clone, k_clone
-    else:
-        q_rope_cos_sin_cache, k_rope_cos_sin_cache = (
-            flashinfer.apply_rope_with_cos_sin_cache(
-                q, k, cos_cache, sin_cache, pos_ids, interleave=False
-            )
-        )
+    pos_ids = torch.arange(seq_len, device=device).repeat(batch_size)
+    query = torch.randn(batch_size * seq_len, num_q_heads * head_size, dtype=dtype, device=device)
+    key = torch.randn(batch_size * seq_len, num_kv_heads * head_size, dtype=dtype, device=device)
 
-    # compare
-    torch.testing.assert_close(q_rope, q_rope_cos_sin_cache, rtol=1e-3, atol=1e-3)
-    torch.testing.assert_close(k_rope, k_rope_cos_sin_cache, rtol=1e-3, atol=1e-3)
+    query_ref, key_ref = query.clone(), key.clone() 
+    query_flashinfer, key_flashinfer = query.clone(), key.clone()
 
+    query_ref_out, key_ref_out = rope_ref.forward_native(pos_ids, query_ref, key_ref)
+    query_flashinfer_out, key_flashinfer_out = rope_flashinfer.forward_cuda(pos_ids, query_flashinfer, key_flashinfer)
+    
+    torch.testing.assert_close(query_ref_out, query_flashinfer_out, atol=1e-2, rtol=1e-2)
+    torch.testing.assert_close(key_ref_out, key_flashinfer_out, atol=1e-2, rtol=1e-2)
 
 if __name__ == "__main__":
     test_rope(2, 1, 8, 8, 1, 128, "llama", 1.0, False)
-    # test_rope_pos_ids(2, 1, 8, 8, 1, 128, "llama31", 1.0, False)
-    # test_rope_cos_sin_cache(99, 19, 16, 8, 99, 256, "llama31", 0.5, False)
+    test_rope_pos_ids(2, 1, 8, 8, 1, 128, "llama31", 1.0, False)
+    test_rope_cos_sin_cache(64, 64, 32, 8000, True, torch.bfloat16, "cuda", 32, 32, 1, 1)
