diff --git a/flashinfer/artifacts.py b/flashinfer/artifacts.py
index 5e337a6dd2..0f0d18d680 100644
--- a/flashinfer/artifacts.py
+++ b/flashinfer/artifacts.py
@@ -117,7 +117,7 @@ class ArtifactPath:
         "037e528e719ec3456a7d7d654f26b805e44c63b1/gemm-8704aa4-f91dc9e/"
     )
     CUDNN_SDPA: str = "4c623163877c8fef5751c9c7a59940cd2baae02e/fmha/cudnn/"
-    DEEPGEMM: str = "d25901733420c7cddc1adf799b0d4639ed1e162f/deep-gemm/"
+    DEEPGEMM: str = "51d730202c9eef782f06ecc950005331d85c5d4b/deep-gemm/"
 
 
 class MetaInfoHash:
@@ -127,7 +127,7 @@ class MetaInfoHash:
     TRTLLM_GEN_BMM: str = (
         "c98b4ce69a39fd41556d67033c30ea814ef76b0a2fe16e798e55baf0104acc34"
     )
-    DEEPGEMM: str = "69aa277b7f3663ed929e73f9c57301792b8c594dac15a465b44a5d151b6a1d50"
+    DEEPGEMM: str = "b4374f857c3066089c4ec6b5e79e785559fa2c05ce2623710b0b04bf86414a48"
     TRTLLM_GEN_GEMM: str = (
         "0345358c916d990709f9670e113e93f35c76aa22715e2d5128ec2ca8740be5ba"
     )
diff --git a/flashinfer/deep_gemm.py b/flashinfer/deep_gemm.py
index bea4d71686..ecb74c3fde 100644
--- a/flashinfer/deep_gemm.py
+++ b/flashinfer/deep_gemm.py
@@ -193,8 +193,14 @@ def transform_sf_into_required_layout(
     gran = (recipe[0 if is_sfa else 1], recipe[2])
 
     should_skip_transform = (
-        sf.dtype == torch.int and gran == (1, 128) and get_device_arch() == "100a"
-    ) or (sf.dtype == torch.int and gran == (128, 128) and get_device_arch() == "100a")
+        sf.dtype == torch.int
+        and gran == (1, 128)
+        and get_device_arch() in ("100a", "103a")
+    ) or (
+        sf.dtype == torch.int
+        and gran == (128, 128)
+        and get_device_arch() in ("100a", "103a")
+    )
 
     if not should_skip_transform:
         # Pre-transform checks
@@ -205,7 +211,11 @@ def transform_sf_into_required_layout(
         raise NotImplementedError
 
     # (FP32, 1, 128) on SM100: transform to (INT, 1, 128), TMA-aligned and MN-major
-    if sf.dtype == torch.float and gran == (1, 128) and get_device_arch() == "100a":
+    if (
+        sf.dtype == torch.float
+        and gran == (1, 128)
+        and get_device_arch() in ("100a", "103a")
+    ):
         sf = get_col_major_tma_aligned_packed_tensor(sf)
         return check_sf_layout(
             sf,
@@ -222,7 +232,11 @@ def transform_sf_into_required_layout(
         raise NotImplementedError
 
     # (FP32, 128, 128) on SM100: transform to (INT, 1, 128), TMA-aligned and MN-major
-    if sf.dtype == torch.float and gran == (128, 128) and get_device_arch() == "100a":
+    if (
+        sf.dtype == torch.float
+        and gran == (128, 128)
+        and get_device_arch() in ("100a", "103a")
+    ):
         sf = sf.index_select(-2, torch.arange(mn, device=sf.device) // 128)
         sf = get_col_major_tma_aligned_packed_tensor(sf)
         return check_sf_layout(
@@ -247,7 +261,9 @@ def transform_sf_into_required_layout(
             type_check=torch.int,
         )
 
-    AssertionError(f"Unknown cases: {sf.dtype=}, {gran=}, arch={get_device_arch()}")
+    raise AssertionError(
+        f"Unknown cases: {sf.dtype=}, {gran=}, arch={get_device_arch()}"
+    )
 
 
 @functools.lru_cache(maxsize=None)
@@ -268,6 +284,7 @@ def must_be_k_major() -> bool:
     return {
         "90a": True,
         "100a": False,
+        "103a": False,
     }[get_device_arch()]
 
 
@@ -280,6 +297,8 @@ def get_default_recipe(
         ("90a", torch.float): (1, 128, 128),
         ("100a", torch.float): (1, 128, 128),
         ("100a", torch.int): (1, 1, 128),
+        ("103a", torch.float): (1, 128, 128),
+        ("103a", torch.int): (1, 1, 128),
     }[(get_device_arch(), sfb_dtype)]
 
 
@@ -798,11 +817,13 @@ def __call__(self, **kwargs) -> cbd.CUresult:
 
     def __del__(self) -> None:
         if self.lib is not None:
-            try:
-                checkCudaErrors(self._cleanup_func(self.lib))
-            except Exception as e:
-                # Ignore any errors during shutdown
-                print(f"Failed to delete SM100FP8GemmRuntime: {e}")
+            cleanup = getattr(self, "_cleanup_func", None)
+            if callable(cleanup):
+                try:
+                    cleanup(self.lib)
+                except Exception as e:
+                    # Ignore any errors during shutdown
+                    print(f"Failed to delete SM100FP8GemmRuntime with exception: {e}")
 
     @staticmethod
     def generate(kwargs: Dict[str, Any]) -> str:
@@ -923,7 +944,7 @@ def load(name: str, code: str) -> SM100FP8GemmRuntime:
     signature = f"{name}$${code}"
     cubin_name = f"kernel.{name}.{hash_to_hex(signature)}"
     if cubin_name not in KERNEL_MAP:
-        raise ValueError("cubin not registered")
+        raise ValueError(f"cubin not registered: {cubin_name}")
     if cubin_name in RUNTIME_CACHE:
         return RUNTIME_CACHE[cubin_name]
     symbol, sha256 = KERNEL_MAP[cubin_name]
@@ -1111,7 +1132,7 @@ def m_grouped_fp8_gemm_nt_contiguous_kwargs_gen(
     return static_kwargs, all_kwargs
 
 
-def m_grouped_fp8_gemm_nt_contiguous_sm100(
+def m_grouped_fp8_gemm_nt_contiguous_sm10x(
     a: torch.Tensor,
     sfa: torch.Tensor,
     b: torch.Tensor,
@@ -1314,7 +1335,7 @@ def m_grouped_fp8_gemm_nt_masked_kwargs_gen(
     return static_kwargs, all_kwargs
 
 
-def m_grouped_fp8_gemm_nt_masked_sm100(
+def m_grouped_fp8_gemm_nt_masked_sm10x(
     a: torch.Tensor,
     sfa: torch.Tensor,
     b: torch.Tensor,
@@ -1385,11 +1406,17 @@ def m_grouped_fp8_gemm_nt_contiguous(
 
     impl = {
         "100a": functools.partial(
-            m_grouped_fp8_gemm_nt_contiguous_sm100,
+            m_grouped_fp8_gemm_nt_contiguous_sm10x,
             major_a=major_a,
             major_b=major_b,
             compiled_dims=compiled_dims,
-        )
+        ),
+        "103a": functools.partial(
+            m_grouped_fp8_gemm_nt_contiguous_sm10x,
+            major_a=major_a,
+            major_b=major_b,
+            compiled_dims=compiled_dims,
+        ),
     }[get_device_arch()]
     impl(a, sfa, b, sfb, d, m_indices)
 
@@ -1442,11 +1469,17 @@ def m_grouped_fp8_gemm_nt_masked(
 
     impl = {
         "100a": functools.partial(
-            m_grouped_fp8_gemm_nt_masked_sm100,
+            m_grouped_fp8_gemm_nt_masked_sm10x,
             major_a=major_a,
             major_b=major_b,
             compiled_dims=compiled_dims,
-        )
+        ),
+        "103a": functools.partial(
+            m_grouped_fp8_gemm_nt_masked_sm10x,
+            major_a=major_a,
+            major_b=major_b,
+            compiled_dims=compiled_dims,
+        ),
     }[get_device_arch()]
     impl(a, sfa, b, sfb, d, masked_m, expected_m)
 
diff --git a/tests/test_groupwise_scaled_gemm_fp8.py b/tests/test_groupwise_scaled_gemm_fp8.py
index d3c1948b1d..eca7b90d01 100755
--- a/tests/test_groupwise_scaled_gemm_fp8.py
+++ b/tests/test_groupwise_scaled_gemm_fp8.py
@@ -186,7 +186,6 @@ def test_fp8_groupwise_group_gemm(
     torch.testing.assert_close(out, ref_c, atol=1e-2, rtol=1e-2)
 
 
-@pytest.mark.xfail(reason="Expected failures for deepgemm tests on SM > 100")
 @pytest.mark.parametrize("m", [128, 256, 512, 1024])
 @pytest.mark.parametrize("nk", [(128, 512), (512, 128), (4096, 7168), (7168, 2048)])
 @pytest.mark.parametrize("group_size", [1, 4, 8, 64, 128, 256])
@@ -230,7 +229,6 @@ def test_fp8_groupwise_group_deepgemm(
     torch.testing.assert_close(out, ref, atol=3e-2, rtol=3e-2)
 
 
-@pytest.mark.xfail(reason="Expected failures for deepgemm tests on SM > 100")
 @pytest.mark.parametrize("m", [128, 256, 512, 1024])
 @pytest.mark.parametrize("nk", [(128, 512), (512, 128), (4096, 7168), (7168, 2048)])
 @pytest.mark.parametrize("group_size", [1, 4, 8, 64, 128, 256])
