diff --git a/csrc/trtllm_allreduce_fusion.cu b/csrc/trtllm_allreduce_fusion.cu
new file mode 100644
index 0000000000..c38203514e
--- /dev/null
+++ b/csrc/trtllm_allreduce_fusion.cu
@@ -0,0 +1,87 @@
+#include <string>
+
+#include "flashinfer/comm/trtllm_allreduce_fusion.cuh"
+#include "pytorch_extension_utils.h"
+
+using namespace flashinfer::trtllm_allreduce_fusion;
+
+#define DISPATCH_FLOATING_TYPES_FOR_ALLREDUCE(scalar_type, c_type, ...)                           \
+  [&] {                                                                                           \
+    switch (scalar_type) {                                                                        \
+      case at::ScalarType::Half: {                                                                \
+        using c_type = half;                                                                      \
+        return __VA_ARGS__();                                                                     \
+      }                                                                                           \
+      case at::ScalarType::BFloat16: {                                                            \
+        using c_type = __nv_bfloat16;                                                             \
+        return __VA_ARGS__();                                                                     \
+      }                                                                                           \
+      case at::ScalarType::Float: {                                                               \
+        using c_type = float;                                                                     \
+        return __VA_ARGS__();                                                                     \
+      }                                                                                           \
+      default:                                                                                    \
+        TORCH_CHECK(false,                                                                        \
+                    "Unsupported dtype in DISPATCH_FLOATING_TYPES_FOR_ALLREDUCE: ", scalar_type); \
+    }                                                                                             \
+  }()
+
+void trtllm_allreduce_fusion(
+    at::Tensor& allreduce_in, int64_t world_size, int64_t world_rank, int64_t token_num,
+    int64_t hidden_size, at::Tensor& workspace_ptrs, bool launch_with_pdl, bool use_oneshot,
+    bool trigger_completion_at_end, bool fp32_acc, int64_t pattern_code,
+    std::optional<at::Tensor> allreduce_out, std::optional<at::Tensor> residual_in,
+    std::optional<at::Tensor> residual_out, std::optional<at::Tensor> norm_out,
+    std::optional<at::Tensor> quant_out, std::optional<at::Tensor> scale_out,
+    std::optional<at::Tensor> rms_gamma, std::optional<double> rms_eps,
+    std::optional<double> scale_factor, std::optional<int64_t> layout_code) {
+  const c10::cuda::OptionalCUDAGuard device_guard(allreduce_in.device());
+  // todo(Yingyi): add dispatch for float and bfloat16
+
+  DISPATCH_FLOATING_TYPES_FOR_ALLREDUCE(allreduce_in.scalar_type(), c_type, [&] {
+    AllReduceFusionParams<c_type> params;
+    params.nranks = world_size;
+    params.rank = world_rank;
+    params.size = token_num * hidden_size;
+    params.hidden_dim = hidden_size;
+    params.workspace = reinterpret_cast<void**>(workspace_ptrs.data_ptr());
+
+    // todo(Yingyi): update optional params
+    // todo(Yingyi): add params check with pattern
+    params.allreduce_in = reinterpret_cast<void*>(allreduce_in.data_ptr());
+    params.allreduce_out = allreduce_out.has_value()
+                               ? reinterpret_cast<void*>(allreduce_out.value().data_ptr())
+                               : nullptr;
+    params.residual_in =
+        residual_in.has_value() ? reinterpret_cast<void*>(residual_in.value().data_ptr()) : nullptr;
+    params.residual_out = residual_out.has_value()
+                              ? reinterpret_cast<void*>(residual_out.value().data_ptr())
+                              : nullptr;
+    params.norm_out =
+        norm_out.has_value() ? reinterpret_cast<void*>(norm_out.value().data_ptr()) : nullptr;
+    params.quant_out =
+        quant_out.has_value() ? reinterpret_cast<void*>(quant_out.value().data_ptr()) : nullptr;
+    params.scale_out =
+        scale_out.has_value() ? reinterpret_cast<void*>(scale_out.value().data_ptr()) : nullptr;
+    params.rms_gamma =
+        rms_gamma.has_value() ? reinterpret_cast<void*>(rms_gamma.value().data_ptr()) : nullptr;
+    params.rms_eps = rms_eps.has_value() ? static_cast<float>(rms_eps.value()) : 0.0f;
+    params.scale_factor =
+        scale_factor.has_value() ? static_cast<float>(scale_factor.value()) : 0.0f;
+    params.use_oneshot = use_oneshot;
+    params.layout = layout_code.has_value()
+                        ? static_cast<FP4QuantizationSFLayout>(layout_code.value())
+                        : FP4QuantizationSFLayout::SWIZZLED;
+    params.pattern = static_cast<AllReduceFusionPattern>(pattern_code);
+    params.trigger_completion_at_end = trigger_completion_at_end;
+    params.stream = at::cuda::getCurrentCUDAStream();
+
+    auto status = allreduce_fusion_op(params, launch_with_pdl, fp32_acc);
+    TORCH_CHECK(status == cudaSuccess, "allreduce_fusion_op failed with error code",
+                cudaGetErrorString(status));
+  });
+}
+
+TORCH_LIBRARY_FRAGMENT(TORCH_EXTENSION_NAME, m) {
+  m.def("trtllm_allreduce_fusion", &trtllm_allreduce_fusion);
+}
diff --git a/flashinfer/comm.py b/flashinfer/comm.py
index 74e8e471b3..8ec47bb8d5 100644
--- a/flashinfer/comm.py
+++ b/flashinfer/comm.py
@@ -68,6 +68,22 @@ class AllReduceFusionOp:
     MOE_ALLREDUCE_RESIDUAL_RMS_NORM = 8
 
 
+class AllReduceFusionPattern:
+    # NOTE: for trtllm_allreduce_fusion
+    # Basic all-reduce pattern
+    kAllReduce = 0
+    # All-reduce followed by residual add and RMS norm
+    kARResidualRMSNorm = 1
+    # All-reduce followed by residual add, RMS norm and FP8 quantization
+    kARResidualRMSNormFP8Quant = 2
+    # All-reduce followed by residual add, RMS norm and FP4 quantization
+    kARResidualRMSNormFP4Quant = 3
+    # All-reduce followed by residual add, RMS norm and FP8 quantization, with norm output
+    kARResidualRMSNormOutFP8Quant = 4
+    # All-reduce followed by residual add, RMS norm and FP4 quantization, with norm output
+    kARResidualRMSNormOutFP4Quant = 5
+
+
 class FP4QuantizationSFLayout:
     # Block scale factors are stored in swizzled layout for cutlass FP4 kernel. Scale factor
     # blocks are organized in 512-byte blocks in global memory, with each block having 128x4 FP8
@@ -255,6 +271,7 @@ def gen_comm_module() -> JitSpec:
             jit_env.FLASHINFER_CSRC_DIR / "comm_pybind.cu",
             jit_env.FLASHINFER_CSRC_DIR / "custom_all_reduce.cu",
             jit_env.FLASHINFER_CSRC_DIR / "trtllm_allreduce.cu",
+            jit_env.FLASHINFER_CSRC_DIR / "trtllm_allreduce_fusion.cu",
             jit_env.FLASHINFER_CSRC_DIR / "trtllm_moe_allreduce_fusion.cu",
         ],
         extra_cuda_cflags=sm100a_nvcc_flags,
@@ -415,6 +432,79 @@ def trtllm_custom_all_reduce(
             lamport_peer_comm_buffer_ptrs_2,
         )
 
+    @register_custom_op(
+        "flashinfer::trtllm_allreduce_fusion",
+        mutates_args=[
+            "allreduce_in",
+            "world_size",
+            "world_rank",
+            "token_num",
+            "hidden_dim",
+            "workspace_ptrs",
+            "launch_with_pdl",
+            "use_oneshot",
+            "trigger_completion_at_end",
+            "fp32_acc",
+            "pattern_code",
+            "allreduce_out",
+            "residual_in",
+            "residual_out",
+            "norm_out",
+            "quant_out",
+            "scale_out",
+            "rms_gamma",
+            "rms_eps",
+            "scale_factor",
+            "layout_code",
+        ],
+    )
+    def trtllm_allreduce_fusion(
+        allreduce_in: torch.Tensor,
+        world_size: int,
+        world_rank: int,
+        token_num: int,
+        hidden_dim: int,
+        workspace_ptrs: torch.Tensor,
+        launch_with_pdl: bool,
+        use_oneshot: bool,
+        trigger_completion_at_end: bool,
+        fp32_acc: bool,
+        pattern_code: AllReduceFusionPattern,
+        allreduce_out: Optional[torch.Tensor],
+        residual_in: Optional[torch.Tensor],
+        residual_out: Optional[torch.Tensor],
+        norm_out: Optional[torch.Tensor],
+        quant_out: Optional[torch.Tensor],
+        scale_out: Optional[torch.Tensor],
+        rms_gamma: Optional[torch.Tensor],
+        rms_eps: Optional[float],
+        scale_factor: Optional[float],
+        layout_code: Optional[FP4QuantizationSFLayout],
+    ) -> None:
+        module.trtllm_allreduce_fusion(
+            allreduce_in,
+            world_size,
+            world_rank,
+            token_num,
+            hidden_dim,
+            workspace_ptrs,
+            launch_with_pdl,
+            use_oneshot,
+            trigger_completion_at_end,
+            fp32_acc,
+            pattern_code,
+            allreduce_out,
+            residual_in,
+            residual_out,
+            norm_out,
+            quant_out,
+            scale_out,
+            rms_gamma,
+            rms_eps,
+            scale_factor,
+            layout_code,
+        )
+
     @register_custom_op(
         "flashinfer::trtllm_moe_allreduce_fusion",
         mutates_args=[
@@ -497,6 +587,7 @@ def trtllm_moe_allreduce_fusion(
         trtllm_lamport_initialize=trtllm_lamport_initialize,
         trtllm_lamport_initialize_all=trtllm_lamport_initialize_all,
         trtllm_custom_all_reduce=trtllm_custom_all_reduce,
+        trtllm_allreduce_fusion=trtllm_allreduce_fusion,
         trtllm_moe_allreduce_fusion=trtllm_moe_allreduce_fusion,
     )
 
@@ -700,6 +791,7 @@ def trtllm_create_ipc_workspace_for_all_reduce_fusion(
     tp_size: int,
     max_token_num: int,
     hidden_dim,
+    use_fp32_lamport: bool = False,
     group: Optional[ProcessGroup] = None,
 ) -> List[int]:
     """
@@ -723,7 +815,11 @@ def trtllm_create_ipc_workspace_for_all_reduce_fusion(
     flag_size = tp_size * BarrierFlagCount * 4
     # lamport_comm_size = tp_size * max(max_token_num, OneShotMaxToken) * hidden_dim * 2
     # enable larger workspace for cases > OneShotMaxToken
-    lamport_comm_size = tp_size * max_token_num * hidden_dim * 2
+    lamport_comm_size = (
+        tp_size * max_token_num * hidden_dim * 2
+        if not use_fp32_lamport
+        else tp_size * max_token_num * hidden_dim * 4
+    )
     lamport_buffer_size = lamport_comm_size * 3
 
     # we should init 3 buffers for all reduce fusion:
@@ -741,9 +837,14 @@ def trtllm_create_ipc_workspace_for_all_reduce_fusion(
     )
 
     # Initialize lamport buffer
-    trtllm_lamport_initialize(
-        ipc_handles[2][tp_rank], lamport_buffer_size // 2, torch.float16
-    )
+    if use_fp32_lamport:
+        trtllm_lamport_initialize(
+            ipc_handles[2][tp_rank], lamport_buffer_size // 4, torch.float32
+        )
+    else:
+        trtllm_lamport_initialize(
+            ipc_handles[2][tp_rank], lamport_buffer_size // 2, torch.float16
+        )
 
     # initialize workspace
     workspace = list()
@@ -877,6 +978,54 @@ def trtllm_custom_all_reduce(
     )
 
 
+def trtllm_allreduce_fusion(
+    allreduce_in: torch.Tensor,
+    world_size: int,
+    world_rank: int,
+    token_num: int,
+    hidden_dim: int,
+    workspace_ptrs: torch.Tensor,
+    launch_with_pdl: bool,
+    use_oneshot: bool,
+    trigger_completion_at_end: bool,
+    fp32_acc: bool,
+    pattern_code: AllReduceFusionPattern,
+    allreduce_out: Optional[torch.Tensor],
+    residual_in: Optional[torch.Tensor],
+    residual_out: Optional[torch.Tensor],
+    norm_out: Optional[torch.Tensor],
+    quant_out: Optional[torch.Tensor],
+    scale_out: Optional[torch.Tensor],
+    rms_gamma: Optional[torch.Tensor],
+    rms_eps: Optional[float],
+    scale_factor: Optional[float],
+    layout_code: Optional[FP4QuantizationSFLayout],
+) -> None:
+    get_comm_module().trtllm_allreduce_fusion(
+        allreduce_in=allreduce_in,
+        world_size=world_size,
+        world_rank=world_rank,
+        token_num=token_num,
+        hidden_dim=hidden_dim,
+        workspace_ptrs=workspace_ptrs,
+        launch_with_pdl=launch_with_pdl,
+        use_oneshot=use_oneshot,
+        trigger_completion_at_end=trigger_completion_at_end,
+        fp32_acc=fp32_acc,
+        pattern_code=pattern_code,
+        allreduce_out=allreduce_out,
+        residual_in=residual_in,
+        residual_out=residual_out,
+        norm_out=norm_out,
+        quant_out=quant_out,
+        scale_out=scale_out,
+        rms_gamma=rms_gamma,
+        rms_eps=rms_eps,
+        scale_factor=scale_factor,
+        layout_code=layout_code,
+    )
+
+
 def trtllm_moe_allreduce_fusion(
     world_size: int,
     world_rank: int,
diff --git a/include/flashinfer/comm/trtllm_allreduce.cuh b/include/flashinfer/comm/trtllm_allreduce.cuh
index 8a122d046b..4e69930fd4 100644
--- a/include/flashinfer/comm/trtllm_allreduce.cuh
+++ b/include/flashinfer/comm/trtllm_allreduce.cuh
@@ -185,6 +185,12 @@ struct neg_zero<nv_bfloat16> {
   static constexpr __nv_bfloat16 value = __nv_bfloat16_raw{neg_zero_bits};
 };
 
+template <>
+struct neg_zero<float> {
+  static constexpr unsigned int neg_zero_bits = 0x80000000U;
+  static constexpr float value = -0.0f;
+};
+
 template <typename T>
 __device__ static constexpr T neg_zero_v = neg_zero<T>::value;
 
diff --git a/include/flashinfer/comm/trtllm_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
new file mode 100644
index 0000000000..57b9443cbc
--- /dev/null
+++ b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
@@ -0,0 +1,894 @@
+#include <cooperative_groups.h>
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+
+#include <cuda/std/optional>
+#include <tuple>
+#include <type_traits>
+
+#include "../exception.h"
+#include "../logging.h"
+#include "../utils.cuh"
+#include "../vec_dtypes.cuh"
+
+namespace flashinfer {
+
+namespace trtllm_allreduce_fusion {
+
+namespace details {
+
+static constexpr int kBytesPerAccess = 16;
+static constexpr int kOneShotMaxToken = 128;
+static constexpr int kBarrierFlagCount = 256;
+
+}  // namespace details
+
+namespace utils {
+
+#define FINAL_MASK 0xffffffff
+
+template <typename T, int NUM>
+__inline__ __device__ T warpReduceSumV2(T* val) {
+#pragma unroll
+  for (int i = 0; i < NUM; i++) {
+#pragma unroll
+    for (int mask = 16; mask > 0; mask >>= 1)
+      val[i] += __shfl_xor_sync(FINAL_MASK, val[i], mask, 32);
+  }
+  return (T)(0.0f);
+}
+
+template <typename T, int NUM>
+__inline__ __device__ T blockReduceSumV2(T* val) {
+  static __shared__ T shared[NUM][33];
+  int lane = threadIdx.x & 0x1f;
+  int wid = threadIdx.x >> 5;
+
+  warpReduceSumV2<T, NUM>(val);
+
+  if (lane == 0) {
+#pragma unroll
+    for (int i = 0; i < NUM; i++) {
+      shared[i][wid] = val[i];
+    }
+  }
+
+  __syncthreads();
+
+  bool is_mask = threadIdx.x < (blockDim.x / 32.f);
+#pragma unroll
+  for (int i = 0; i < NUM; i++) {
+    val[i] = is_mask ? shared[i][lane] : (T)(0.0f);
+  }
+  warpReduceSumV2<T, NUM>(val);
+  return (T)0.0f;
+}
+
+inline int getSMVersion() {
+  int device{-1};
+  FLASHINFER_CUDA_CALL(cudaGetDevice(&device));
+  int sm_major = 0;
+  int sm_minor = 0;
+  FLASHINFER_CUDA_CALL(
+      cudaDeviceGetAttribute(&sm_major, cudaDevAttrComputeCapabilityMajor, device));
+  FLASHINFER_CUDA_CALL(
+      cudaDeviceGetAttribute(&sm_minor, cudaDevAttrComputeCapabilityMinor, device));
+  return sm_major * 10 + sm_minor;
+}
+
+}  // namespace utils
+
+template <typename T, uint32_t VEC_SIZE>
+__device__ __forceinline__ vec_t<T, VEC_SIZE> vec_add(const vec_t<T, VEC_SIZE>& a,
+                                                      const vec_t<T, VEC_SIZE>& b) {
+  vec_t<T, VEC_SIZE> ret;
+#pragma unroll
+  for (int i = 0; i < VEC_SIZE; ++i) {
+    ret[i] = static_cast<float>(a[i]) + static_cast<float>(b[i]);
+  }
+  return ret;
+}
+
+enum class AllReduceFusionPattern : int {
+  kAllReduce = 0,
+  kARResidualRMSNorm = 1,
+  kARResidualRMSNormFP8Quant = 2,
+  kARResidualRMSNormFP4Quant = 3,
+  // The difference between these two and the standard version is that the NormOut version outputs
+  // the result of the norm.
+  kARResidualRMSNormOutFP8Quant = 4,
+  kARResidualRMSNormOutFP4Quant = 5
+};
+
+enum class FP4QuantizationSFLayout {
+  // Block scale factors are stored in swizzled layout for cutlass FP4 kernel. Scale factor
+  // blocks are organized in 512-byte blocks in global memory, with each block having 128x4 FP8
+  // values. The SF matrix dimensions are therefore padded - rows to the nearest multiple of 128 and
+  // columns to the nearest multiple of 4.
+  //
+  // The scale factor block rows map to data block rows in an interleaved pattern:
+  // For a scale factor row 'i', it maps to data block row: (i % 4) * 32 + (i / 4)
+  // Column 'j' in the scale factor block corresponds to scaling the j-th block in the data tensor.
+  //
+  // Please refer to https://nvbugs/4165523 for more details about the swizzled layout.
+  SWIZZLED,
+  // Block scale factors are stored in linear layout (row-major). This is used in some trtllm-gen
+  // kernels standard.
+  LINEAR
+};
+
+enum class QuantType : int {
+  kNone = 0,
+  kFP8 = 1,
+  kFP4 = 2,
+};
+
+template <AllReduceFusionPattern Pattern>
+struct FusionPatternTraits;
+
+#define DEFINE_FUSION_PATTERN_TRAITS(pattern, hasAllReduceOut, hasResidual, hasResidualOut, \
+                                     hasRMSNorm, hasNormOut, quantType)                     \
+  template <>                                                                               \
+  struct FusionPatternTraits<pattern> {                                                     \
+    static constexpr bool kHasAllReduceOut = hasAllReduceOut;                               \
+    static constexpr bool kHasResidual = hasResidual;                                       \
+    static constexpr bool kHasResidualOut = hasResidualOut;                                 \
+    static constexpr bool kHasRMSNorm = hasRMSNorm;                                         \
+    static constexpr bool kHasNormOut = hasNormOut;                                         \
+    static constexpr QuantType kQuantType = quantType;                                      \
+  };
+
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kAllReduce, true, false, false, false, false,
+                             QuantType::kNone);
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kARResidualRMSNorm, false, true, true, true,
+                             true, QuantType::kNone);
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kARResidualRMSNormFP8Quant, false, true, true,
+                             true, false, QuantType::kFP8);
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kARResidualRMSNormFP4Quant, false, true, true,
+                             true, false, QuantType::kFP4);
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kARResidualRMSNormOutFP8Quant, false, true,
+                             true, true, true, QuantType::kFP8);
+DEFINE_FUSION_PATTERN_TRAITS(AllReduceFusionPattern::kARResidualRMSNormOutFP4Quant, false, true,
+                             true, true, true, QuantType::kFP4);
+#undef DEFINE_FUSION_PATTERN_TRAITS
+
+template <AllReduceFusionPattern Pattern>
+constexpr bool HasResidual = FusionPatternTraits<Pattern>::kHasResidual;
+template <AllReduceFusionPattern Pattern>
+constexpr bool HasRMSNorm = FusionPatternTraits<Pattern>::kHasRMSNorm;
+template <AllReduceFusionPattern Pattern>
+constexpr bool HasAllReduceOut = FusionPatternTraits<Pattern>::kHasAllReduceOut;
+template <AllReduceFusionPattern Pattern>
+constexpr bool HasResidualOut = FusionPatternTraits<Pattern>::kHasResidualOut;
+template <AllReduceFusionPattern Pattern>
+constexpr bool HasNormOut = FusionPatternTraits<Pattern>::kHasNormOut;
+template <AllReduceFusionPattern Pattern>
+constexpr QuantType GetQuantType = FusionPatternTraits<Pattern>::kQuantType;
+
+template <typename T>
+struct AllReduceFusionParams {
+  int nranks;
+  int rank;
+  int size;
+  int hidden_dim;
+  void** workspace;
+  void* allreduce_in;
+  void* allreduce_out;
+  void* residual_in;
+  void* residual_out;
+  void* norm_out;
+  void* quant_out;
+  void* scale_out;
+  void* rms_gamma;
+  float rms_eps;
+  float scale_factor;
+  bool use_oneshot;
+  FP4QuantizationSFLayout layout = FP4QuantizationSFLayout::SWIZZLED;
+  cudaStream_t stream;
+  AllReduceFusionPattern pattern;
+  bool trigger_completion_at_end = true;
+};
+
+template <int NRanks>
+struct SyncComm {
+  __device__ __forceinline__ SyncComm(void** workspace) {
+    counter_ptr = &reinterpret_cast<int*>(workspace[NRanks * 3])[0];
+    flag_ptr = &reinterpret_cast<int*>(workspace[NRanks * 3])[1];
+    flag_value = *flag_ptr;
+    for (int r = 0; r < NRanks; ++r) {
+      comm_bufs[r] = workspace[r];
+      barrier_flags[r] = workspace[NRanks + r];
+    }
+    __syncthreads();
+    if (threadIdx.x == 0) {
+      atomicAdd(counter_ptr, 1);
+    }
+  }
+
+  __device__ __forceinline__ void update(int new_flag_value) {
+    if (blockIdx.x == 0 && threadIdx.x == 0) {
+      while (*reinterpret_cast<int volatile*>(counter_ptr) != gridDim.x) {
+      }
+      *flag_ptr = new_flag_value;
+      *counter_ptr = 0;
+    }
+  }
+
+  int* counter_ptr;
+  int* flag_ptr;
+  void* comm_bufs[NRanks];
+  void* barrier_flags[NRanks];
+  int flag_value;
+};
+
+template <int NRanks>
+struct LamportComm {
+  __device__ __forceinline__ LamportComm(void** workspace, int rank) {
+    counter_ptr = &reinterpret_cast<int*>(workspace[NRanks * 3])[0];
+    flag_ptr = &reinterpret_cast<int*>(workspace[NRanks * 3])[2];
+    clear_ptr = &reinterpret_cast<int*>(workspace[NRanks * 3])[4];
+    flag_value = *flag_ptr;
+    int comm_size = reinterpret_cast<int*>(workspace[NRanks * 3])[3];
+    clear_size = *clear_ptr;
+    int data_offset = flag_value % 3;
+    int clear_offset = (flag_value + 2) % 3;
+    for (int r = 0; r < NRanks; ++r) {
+      data_bufs[r] =
+          reinterpret_cast<uint8_t*>(workspace[2 * NRanks + r]) + data_offset * comm_size;
+    }
+    clear_buf = reinterpret_cast<uint8_t*>(workspace[2 * NRanks + rank]) + clear_offset * comm_size;
+    __syncthreads();
+    if (threadIdx.x == 0) {
+      atomicAdd(counter_ptr, 1);
+    }
+  }
+
+  __device__ __forceinline__ void update(int new_clear_size) {
+    if (blockIdx.x == 0 && threadIdx.x == 0) {
+      while (*reinterpret_cast<int volatile*>(counter_ptr) != gridDim.x) {
+      }
+      *flag_ptr = (flag_value + 1) % 3;
+      *clear_ptr = new_clear_size;
+      *counter_ptr = 0;
+    }
+  }
+
+  int* counter_ptr;
+  int* flag_ptr;
+  int* clear_ptr;
+  uint8_t* data_bufs[NRanks];
+  uint8_t* clear_buf;
+  int clear_size;
+  int flag_value;
+};
+
+template <int NRanks>
+class Barrier {
+ public:
+  __device__ __forceinline__ Barrier(int rank, SyncComm<NRanks> const& comm) {
+    if (threadIdx.x < NRanks) {
+      m_flag_value = comm.flag_value;
+      int current_rank = rank;
+      int target_rank = threadIdx.x;
+      m_target_flag = reinterpret_cast<int*>(comm.barrier_flags[target_rank]) + current_rank;
+      m_current_flag = reinterpret_cast<int*>(comm.barrier_flags[current_rank]) +
+                       blockIdx.x * NRanks + target_rank;
+    }
+  }
+
+  __device__ __forceinline__ void sync() {
+    __syncthreads();
+    if (threadIdx.x < NRanks) {
+      m_flag_value = next_flag(m_flag_value);
+      // To avoid the ABA problem, we need to synchronize the correct flag value to all
+      // barrier_flags, even if the corresponding CTA has not been launched.
+      for (int flag_idx = blockIdx.x; flag_idx < details::kBarrierFlagCount;
+           flag_idx += gridDim.x) {
+        st_flag(m_target_flag + flag_idx * NRanks, m_flag_value);
+      }
+      while (ld_flag(m_current_flag) == prev_flag(m_flag_value)) {
+      }
+    }
+    __syncthreads();
+  }
+
+ protected:
+  __device__ __forceinline__ void st_flag(int* addr, int flag) {
+    asm volatile("st.global.release.sys.b32 [%1], %0;" ::"r"(flag), "l"(addr));
+  }
+
+  __device__ __forceinline__ int ld_flag(int* addr) {
+    int flag;
+    asm volatile("ld.global.acquire.sys.b32 %0, [%1];" : "=r"(flag) : "l"(addr));
+    return flag;
+  }
+
+  __device__ __forceinline__ int next_flag(int flag) { return flag == 2 ? 0 : flag + 1; }
+
+  __device__ __forceinline__ int prev_flag(int flag) { return flag == 0 ? 2 : flag - 1; }
+
+ public:
+  int m_flag_value;
+
+ private:
+  int* m_target_flag;
+  int* m_current_flag;
+};
+
+template <AllReduceFusionPattern Pattern, typename T>
+class FusedOp {
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+
+ public:
+  __device__ __forceinline__ FusedOp(AllReduceFusionParams<T> const& params, int access_id,
+                                     int access_id_in_token)
+      : m_params(params), m_access_id(access_id), m_access_id_in_token(access_id_in_token) {
+    if constexpr (HasRMSNorm<Pattern>) {
+      m_gamma_val.load(reinterpret_cast<T*>(params.rms_gamma) + m_access_id_in_token * VEC_SIZE);
+    }
+    if constexpr (HasResidual<Pattern>) {
+      m_residual_val.load(reinterpret_cast<T*>(params.residual_in) + m_access_id * VEC_SIZE);
+    }
+    if constexpr (GetQuantType<Pattern> == QuantType::kFP8) {
+      m_scale_factor = 1.f / params.scale_factor;
+    } else if constexpr (GetQuantType<Pattern> == QuantType::kFP4) {
+      m_scale_factor = params.scale_factor;
+    }
+  }
+
+  // template <typename T>
+  __device__ __forceinline__ void update(int access_id) {
+    if (m_access_id != access_id) {
+      m_access_id = access_id;
+      if constexpr (HasResidual<Pattern>) {
+        m_residual_val.load(reinterpret_cast<T*>(m_params.residual_in) + m_access_id * VEC_SIZE);
+      }
+    }
+  }
+
+  // template <typename T, uint32_t VEC_SIZE>
+  __device__ __forceinline__ void operator()(vec_t<T, VEC_SIZE> val, int token_id) {
+    if constexpr (HasAllReduceOut<Pattern>) {
+      val.store(reinterpret_cast<T*>(m_params.allreduce_out) + m_access_id * VEC_SIZE);
+    }
+    if constexpr (HasResidual<Pattern>) {
+      val = vec_add<T, VEC_SIZE>(val, m_residual_val);
+      if constexpr (HasResidualOut<Pattern>) {
+        val.store(reinterpret_cast<T*>(m_params.residual_out) + m_access_id * VEC_SIZE);
+      }
+    }
+    if constexpr (HasRMSNorm<Pattern>) {
+      val = rms_norm(val, m_gamma_val);
+      if constexpr (HasNormOut<Pattern>) {
+        val.store(reinterpret_cast<T*>(m_params.norm_out) + m_access_id * VEC_SIZE);
+      }
+    }
+    // todo(yingyi): add quant support
+    //     if constexpr (GetQuantType<Pattern> == QuantType::kFP4) {
+    //       constexpr int SF_VEC_SIZE = 16;
+    //       using PackedVec = PackedVec<DType>;
+    //       PackedVec pack_val = *reinterpret_cast<PackedVec const*>(&val);
+    //       auto sf_out = cvt_quant_to_fp4_get_sf_out_offset<uint32_t, 2, SF_VEC_SIZE>(
+    //           std::nullopt, token_id, m_access_id_in_token, std::nullopt, m_params.hidden_dim,
+    //           reinterpret_cast<uint32_t*>(m_params.scale_out), m_params.layout);
+    //       reinterpret_cast<uint32_t*>(m_params.quant_out)[m_access_id] =
+    //           cvt_warp_fp16_to_fp4<DType, SF_VEC_SIZE, false>(pack_val, m_scale_factor, sf_out);
+    //     } else if constexpr (GetQuantType<Pattern> == QuantType::kFP8) {
+    //       using PackedQuantizedType = std::conditional_t<std::is_same_v<DType, float>, float,
+    //       float2>; PackedQuantizedType ret;
+    // #pragma unroll
+    //       for (int i = 0; i < kMathCount; ++i) {
+    //         reinterpret_cast<__nv_fp8_e4m3*>(&ret)[i] = static_cast<__nv_fp8_e4m3>(
+    //             static_cast<float>(reinterpret_cast<DType*>(&val)[i]) * m_scale_factor);
+    //       }
+    //       reinterpret_cast<PackedQuantizedType*>(m_params.quant_out)[m_access_id] = ret;
+    //     } else {
+    //       static_assert(GetQuantType<Pattern> == QuantType::kNone, "Invalid quant type");
+    //     }
+  }
+
+ protected:
+  __device__ __forceinline__ vec_t<T, VEC_SIZE> rms_norm(vec_t<T, VEC_SIZE> const& residual,
+                                                         vec_t<T, VEC_SIZE> const& gamma) {
+    __shared__ float s_val;
+    vec_t<T, VEC_SIZE> norm_out;
+    float acc = 0.f;
+#pragma unroll
+    for (int i = 0; i < VEC_SIZE; ++i) {
+      float v = static_cast<float>(reinterpret_cast<T const*>(&residual)[i]);
+      acc += v * v;
+    }
+    utils::blockReduceSumV2<float, 1>(&acc);
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+    namespace cg = cooperative_groups;
+    cg::cluster_group cluster = cg::this_cluster();
+    if (cluster.num_blocks() > 1) {
+      if (threadIdx.x == 0) {
+        s_val = acc;
+        acc = 0.f;
+      }
+      cluster.sync();
+      if (threadIdx.x == 0) {
+        for (int i = 0; i < cluster.num_blocks(); ++i) {
+          acc += *cluster.map_shared_rank(&s_val, i);
+        }
+      }
+      cluster.sync();
+    }
+#endif
+    if (threadIdx.x == 0) {
+      s_val = rsqrtf(acc / m_params.hidden_dim + m_params.rms_eps);
+    }
+    __syncthreads();
+#pragma unroll
+    for (int i = 0; i < VEC_SIZE; ++i) {
+      reinterpret_cast<T*>(&norm_out)[i] =
+          static_cast<T>(static_cast<float>(reinterpret_cast<T const*>(&residual)[i]) * s_val *
+                         static_cast<float>(reinterpret_cast<T const*>(&gamma)[i]));
+    }
+    return norm_out;
+  }
+
+ private:
+  AllReduceFusionParams<T> const& m_params;
+  int m_access_id;
+  int m_access_id_in_token;
+  float m_scale_factor;
+  vec_t<T, VEC_SIZE> m_residual_val;
+  vec_t<T, VEC_SIZE> m_gamma_val;
+};
+
+template <typename T>
+struct neg_zero {
+  static constexpr T value = -T(0);
+};
+
+template <>
+struct neg_zero<half> {
+  static constexpr unsigned short neg_zero_bits = 0x8000U;
+  static constexpr __half value = __half_raw{neg_zero_bits};
+};
+
+template <>
+struct neg_zero<nv_bfloat16> {
+  static constexpr unsigned short neg_zero_bits = 0x8000U;
+  static constexpr __nv_bfloat16 value = __nv_bfloat16_raw{neg_zero_bits};
+};
+
+template <>
+struct neg_zero<float> {
+  static constexpr unsigned int neg_zero_bits = 0x80000000U;
+  static constexpr float value = -0.0f;
+};
+
+template <typename T>
+__device__ static constexpr T neg_zero_v = neg_zero<T>::value;
+
+template <typename T>
+__device__ bool is_negative_zero(T) {
+  return false;
+}
+
+// float specialization
+template <>
+__device__ bool is_negative_zero<float>(float x) {
+  return (__float_as_int(x) == 0x80000000);
+}
+
+// double specialization
+template <>
+__device__ bool is_negative_zero<double>(double x) {
+  return (__double_as_longlong(x) == 0x8000000000000000ULL);
+}
+
+// __half specialization
+template <>
+__device__ bool is_negative_zero<__half>(__half x) {
+  return (__half_as_ushort(x) == 0x8000);
+}
+
+// __nv_bfloat16 specialization
+template <>
+__device__ bool is_negative_zero<__nv_bfloat16>(__nv_bfloat16 x) {
+  return (__bfloat16_as_ushort(x) == 0x8000);
+}
+
+template <typename T, uint32_t VEC_SIZE>
+__device__ __forceinline__ bool has_neg_zero(const vec_t<T, VEC_SIZE>& vec) {
+#pragma unroll
+  for (int i = 0; i < VEC_SIZE; ++i) {
+    if (is_negative_zero(vec[i])) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template <typename T, uint32_t VEC_SIZE>
+__device__ __forceinline__ void remove_neg_zero(vec_t<T, VEC_SIZE>& vec) {
+#pragma unroll
+  for (int i = 0; i < VEC_SIZE; ++i) {
+    vec[i] = (is_negative_zero(vec[i])) ? static_cast<T>(0.f) : vec[i];
+  }
+}
+
+template <typename T>
+__device__ __forceinline__ void set_neg_zero(T* addr) {
+  vec_t<T, details::kBytesPerAccess / sizeof(T)> val;
+  val.fill(neg_zero_v<T>);
+  val.store_global_volatile(addr);
+}
+
+template <typename T, uint32_t VEC_SIZE, int NRanks, bool Fp32Acc>
+__device__ __forceinline__ vec_t<T, VEC_SIZE> allreduce_sum(vec_t<T, VEC_SIZE>* vals) {
+  if constexpr (Fp32Acc) {
+    static_assert(!std::is_same_v<T, float>);
+    float acc_f32[VEC_SIZE];
+#pragma unroll
+    for (int i = 0; i < VEC_SIZE; ++i) {
+      acc_f32[i] = static_cast<float>(reinterpret_cast<T*>(&vals[0])[i]);
+    }
+#pragma unroll
+    for (int r = 1; r < NRanks; ++r) {
+#pragma unroll
+      for (int i = 0; i < VEC_SIZE; ++i) {
+        acc_f32[i] += static_cast<float>(reinterpret_cast<T*>(&vals[r])[i]);
+      }
+    }
+    vec_t<T, VEC_SIZE> acc;
+#pragma unroll
+    for (int i = 0; i < VEC_SIZE; ++i) {
+      acc[i] = static_cast<T>(acc_f32[i]);
+    }
+    return acc;
+  } else {
+    vec_t<T, VEC_SIZE> acc = vals[0];
+#pragma unroll
+    for (int r = 1; r < NRanks; ++r) {
+      acc = vec_add<T, VEC_SIZE>(acc, vals[r]);
+    }
+    return acc;
+  }
+}
+
+template <typename T>
+class IndexHelper {
+ public:
+  __device__ __forceinline__ IndexHelper(AllReduceFusionParams<T> const& params) {
+    static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+    namespace cg = cooperative_groups;
+    cg::cluster_group cluster = cg::this_cluster();
+    cg::grid_group grid = cg::this_grid();
+    token_id = grid.cluster_rank();
+    access_id_in_token = cluster.thread_rank();
+    token_stride = grid.num_clusters();
+#else
+    token_id = blockIdx.x;
+    access_id_in_token = threadIdx.x;
+    token_stride = gridDim.x;
+#endif
+    access_id = token_id * params.hidden_dim / VEC_SIZE + access_id_in_token;
+    access_stride = token_stride * params.hidden_dim / VEC_SIZE;
+    tot_access = params.size / VEC_SIZE;
+  }
+
+  int token_id;
+  int access_id_in_token;
+  int token_stride;
+  int access_id;
+  int access_stride;
+  int tot_access;
+};
+
+template <AllReduceFusionPattern Pattern, typename T, int NRanks, bool Fp32Acc,
+          bool TriggerCompletionAtEnd = true>
+__global__ void allreduce_fusion_kernel_oneshot_lamport(AllReduceFusionParams<T> params) {
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+  IndexHelper<T> index_helper(params);
+  int token_id = index_helper.token_id;
+  int access_id_in_token = index_helper.access_id_in_token;
+  int token_stride = index_helper.token_stride;
+  int access_id = index_helper.access_id;
+  int access_stride = index_helper.access_stride;
+  int tot_access = index_helper.tot_access;
+  vec_t<T, VEC_SIZE> clear_vec;
+  clear_vec.fill(neg_zero_v<T>);
+  FusedOp<Pattern, T> fused_op(params, access_id, access_id_in_token);
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaGridDependencySynchronize();
+  if constexpr (!TriggerCompletionAtEnd) {
+    cudaTriggerProgrammaticLaunchCompletion();
+  }
+#endif
+  LamportComm<NRanks> comm(params.workspace, params.rank);
+  int clear_access = comm.clear_size / VEC_SIZE;
+
+  for (int idx = access_id; idx < tot_access; idx += access_stride) {
+    vec_t<T, VEC_SIZE> val;
+    val.load(reinterpret_cast<T*>(params.allreduce_in) + idx * VEC_SIZE);
+    remove_neg_zero<T, VEC_SIZE>(val);
+#pragma unroll
+    for (int r = 0; r < NRanks; ++r) {
+      // Push data to other ranks
+      val.store(reinterpret_cast<T*>(comm.data_bufs[r]) +
+                (params.rank * tot_access + idx) * VEC_SIZE);
+    }
+  }
+  for (int idx = access_id; idx < clear_access; idx += access_stride) {
+    // Clear comm buffer that previous kernel used
+    clear_vec.store(reinterpret_cast<T*>(comm.clear_buf) + idx * VEC_SIZE);
+  }
+
+  for (int idx = access_id, tidx = token_id; idx < tot_access;
+       idx += access_stride, tidx += token_stride) {
+    fused_op.update(idx);
+    vec_t<T, VEC_SIZE> vals[NRanks];
+    bool done = false;
+
+    while (!done) {
+      done = true;
+#pragma unroll
+      for (int r = 0; r < NRanks; ++r) {
+        // LDG.128 from local rank
+        vals[r].load_global_volatile(reinterpret_cast<T*>(comm.data_bufs[params.rank]) +
+                                     (r * tot_access + idx) * VEC_SIZE);
+        done &= !has_neg_zero<T, VEC_SIZE>(vals[r]);
+      }
+    }
+    vec_t<T, VEC_SIZE> sum_val = allreduce_sum<T, VEC_SIZE, NRanks, Fp32Acc>(vals);
+    fused_op(sum_val, tidx);
+  }
+
+  comm.update(params.size * NRanks);
+
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  if constexpr (TriggerCompletionAtEnd) {
+    cudaTriggerProgrammaticLaunchCompletion();
+  }
+#endif
+}
+
+template <AllReduceFusionPattern Pattern, typename T, int NRanks, bool Fp32Acc>
+__global__ void allreduce_fusion_kernel_twoshot_sync(AllReduceFusionParams<T> params,
+                                                     std::array<int, NRanks> begin_tokens,
+                                                     std::array<int, NRanks> token_num_per_ranks) {
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+  IndexHelper<T> index_helper(params);
+  int token_id = index_helper.token_id;
+  int access_id_in_token = index_helper.access_id_in_token;
+  int token_stride = index_helper.token_stride;
+  int access_id = index_helper.access_id;
+  int access_stride = index_helper.access_stride;
+  int tot_access = index_helper.tot_access;
+  FusedOp<Pattern, T> fused_op(params, access_id, access_id_in_token);
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaGridDependencySynchronize();
+#endif
+  SyncComm<NRanks> comm(params.workspace);
+#pragma unroll
+  for (int r = 0; r < NRanks; ++r) {
+    int comm_access_id = access_id + begin_tokens[r] * params.hidden_dim / VEC_SIZE;
+    int comm_tot_access = (begin_tokens[r] + token_num_per_ranks[r]) * params.hidden_dim / VEC_SIZE;
+    for (int idx = comm_access_id; idx < comm_tot_access; idx += access_stride) {
+      reinterpret_cast<float4*>(comm.comm_bufs[params.rank])[idx] =
+          reinterpret_cast<float4*>(params.allreduce_in)[idx];
+    }
+  }
+  Barrier<NRanks> barrier(params.rank, comm);
+  barrier.sync();
+  int comm_access_id = access_id + begin_tokens[params.rank] * params.hidden_dim / VEC_SIZE;
+  int comm_tot_access =
+      (begin_tokens[params.rank] + token_num_per_ranks[params.rank]) * params.hidden_dim / VEC_SIZE;
+  for (int idx = comm_access_id; idx < comm_tot_access; idx += access_stride) {
+    vec_t<T, VEC_SIZE> vals[NRanks];
+#pragma unroll
+    for (int r = 0; r < NRanks; ++r) {
+      vals[r].load(reinterpret_cast<T*>(comm.comm_bufs[r]) + idx * VEC_SIZE);
+    }
+    vec_t<T, VEC_SIZE> sum_val = allreduce_sum<T, VEC_SIZE, NRanks, Fp32Acc>(vals);
+#pragma unroll
+    for (int r = 0; r < NRanks; ++r) {
+      sum_val.store(reinterpret_cast<T*>(comm.comm_bufs[r]) + (tot_access + idx) * VEC_SIZE);
+    }
+  }
+  barrier.sync();
+#pragma unroll
+  for (int r = 0; r < NRanks; ++r) {
+    int comm_access_id = access_id + begin_tokens[r] * params.hidden_dim / VEC_SIZE;
+    int comm_token_id = token_id + begin_tokens[r];
+    int comm_tot_access = (begin_tokens[r] + token_num_per_ranks[r]) * params.hidden_dim / VEC_SIZE;
+    for (int idx = comm_access_id, tidx = comm_token_id; idx < comm_tot_access;
+         idx += access_stride, tidx += token_stride) {
+      fused_op.update(idx);
+      vec_t<T, VEC_SIZE> sum_val;
+      sum_val.load(reinterpret_cast<T*>(comm.comm_bufs[params.rank]) +
+                   (tot_access + idx) * VEC_SIZE);
+      fused_op(sum_val, tidx);
+    }
+  }
+  comm.update(barrier.m_flag_value);
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  cudaTriggerProgrammaticLaunchCompletion();
+#endif
+}
+
+int get_sm_count() {
+  static int sm_count = 0;
+  if (sm_count == 0) {
+    int device_id;
+    FLASHINFER_CUDA_CALL(cudaGetDevice(&device_id));
+    cudaDeviceProp device_prop;
+    cudaGetDeviceProperties(&device_prop, device_id);
+    sm_count = device_prop.multiProcessorCount;
+  }
+  return sm_count;
+}
+
+template <AllReduceFusionPattern Pattern, typename T, int NRanks, bool Fp32Acc,
+          bool TriggerCompletionAtEnd = true>
+cudaError_t launch_oneshot_lamport(AllReduceFusionParams<T> const& params,
+                                   cudaLaunchConfig_t& cfg) {
+  FLASHINFER_CUDA_CALL(cudaLaunchKernelEx(
+      &cfg,
+      allreduce_fusion_kernel_oneshot_lamport<Pattern, T, NRanks, Fp32Acc, TriggerCompletionAtEnd>,
+      params));
+  return cudaSuccess;
+}
+
+template <AllReduceFusionPattern Pattern, typename T, int NRanks, bool Fp32Acc>
+cudaError_t launch_twoshot_sync(AllReduceFusionParams<T> const& params, cudaLaunchConfig_t& cfg,
+                                std::array<int, NRanks> begin_tokens,
+                                std::array<int, NRanks> token_num_per_ranks) {
+  FLASHINFER_CUDA_CALL(
+      cudaLaunchKernelEx(&cfg, allreduce_fusion_kernel_twoshot_sync<Pattern, T, NRanks, Fp32Acc>,
+                         params, begin_tokens, token_num_per_ranks));
+  return cudaSuccess;
+}
+
+bool use_oneshot(int token_num) { return token_num <= details::kOneShotMaxToken; }
+
+template <AllReduceFusionPattern Pattern, typename T, int NRanks, bool Fp32Acc>
+cudaError_t allreduce_fusion_kernel_launcher(AllReduceFusionParams<T> const& params,
+                                             bool launch_with_pdl) {
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+  FLASHINFER_CHECK(params.size % params.hidden_dim == 0, "params.size % params.hidden_dim != 0");
+  FLASHINFER_CHECK(params.hidden_dim % VEC_SIZE == 0, "params.hidden_dim % VEC_SIZE != 0");
+  static int SM = utils::getSMVersion();
+  int token_num = params.size / params.hidden_dim;
+  bool oneshot = params.use_oneshot;
+  int cluster_num = token_num;
+  std::array<int, NRanks> begin_tokens, token_num_per_ranks;
+  if (!oneshot) {
+    int remaining_token = token_num % NRanks;
+    int token_num_per_rank = token_num / NRanks;
+    cluster_num = token_num_per_rank;
+    if (remaining_token) {
+      cluster_num++;
+    }
+    for (int r = 0; r < NRanks; ++r) {
+      begin_tokens[r] = r * token_num_per_rank + (remaining_token > r ? r : remaining_token);
+      token_num_per_ranks[r] = token_num_per_rank + (remaining_token > r ? 1 : 0);
+    }
+  }
+  int threads_per_token = params.hidden_dim / VEC_SIZE;
+  int cluster_size;
+  if (SM >= 90) {
+    cluster_size = 8;
+  } else {
+    cluster_size = 1;
+  }
+  while (threads_per_token % cluster_size != 0 && cluster_size > 1) {
+    cluster_size /= 2;
+  }
+  int threads_per_block = threads_per_token / cluster_size;
+  while (threads_per_block < 128 && cluster_size >= 2) {
+    threads_per_block *= 2;
+    cluster_size /= 2;
+  }
+  FLASHINFER_CHECK(oneshot || threads_per_block >= params.nranks,
+                   "not oneshot, or threads_per_block < nranks");
+  int block_size = threads_per_block;
+  FLASHINFER_CHECK(block_size <= 1024 && cluster_size > 0,
+                   "block_size > 1024 or cluster_size <= 0");
+  int sm_count = get_sm_count();
+  int grid_size = (std::min(sm_count, cluster_num * cluster_size) / cluster_size) * cluster_size;
+  cudaLaunchConfig_t cfg;
+  cudaLaunchAttribute attribute[2];
+  cfg.gridDim = grid_size;
+  cfg.blockDim = block_size;
+  cfg.dynamicSmemBytes = 0;
+  cfg.stream = params.stream;
+  attribute[0].id = cudaLaunchAttributeProgrammaticStreamSerialization;
+  attribute[0].val.programmaticStreamSerializationAllowed = launch_with_pdl ? 1 : 0;
+  attribute[1].id = cudaLaunchAttributeClusterDimension;
+  attribute[1].val.clusterDim.x = cluster_size;
+  attribute[1].val.clusterDim.y = 1;
+  attribute[1].val.clusterDim.z = 1;
+  cfg.attrs = attribute;
+  cfg.numAttrs = SM >= 90 ? 2 : 0;
+  if (oneshot) {
+    bool trigger_completion_at_end = params.trigger_completion_at_end;
+    if (trigger_completion_at_end) {
+      FLASHINFER_CUDA_CALL(
+          (launch_oneshot_lamport<Pattern, T, NRanks, Fp32Acc, true>(params, cfg)));
+    } else {
+      FLASHINFER_CUDA_CALL(
+          (launch_oneshot_lamport<Pattern, T, NRanks, Fp32Acc, false>(params, cfg)));
+    }
+  } else {
+    FLASHINFER_CUDA_CALL((launch_twoshot_sync<Pattern, T, NRanks, Fp32Acc>(
+        params, cfg, begin_tokens, token_num_per_ranks)));
+  }
+  return cudaSuccess;
+}
+
+template <typename T>
+cudaError_t allreduce_fusion_op(AllReduceFusionParams<T> const& params, bool launch_with_pdl,
+                                bool fp32_acc) {
+#define DISPATCH_ACC_TYPE(T, Pattern, NRanks)                                                      \
+  if constexpr (std::is_same_v<T, float>) {                                                        \
+    return allreduce_fusion_kernel_launcher<Pattern, T, NRanks, false>(params, launch_with_pdl);   \
+  } else {                                                                                         \
+    if (fp32_acc) {                                                                                \
+      return allreduce_fusion_kernel_launcher<Pattern, T, NRanks, true>(params, launch_with_pdl);  \
+    } else {                                                                                       \
+      return allreduce_fusion_kernel_launcher<Pattern, T, NRanks, false>(params, launch_with_pdl); \
+    }                                                                                              \
+  }
+
+#define DISPATCH_PATTERN(T, NRanks)                                                          \
+  switch (params.pattern) {                                                                  \
+    case AllReduceFusionPattern::kAllReduce:                                                 \
+      DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kAllReduce, NRanks);                      \
+      break;                                                                                 \
+    case AllReduceFusionPattern::kARResidualRMSNorm:                                         \
+      DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kARResidualRMSNorm, NRanks);              \
+      break;                                                                                 \
+    case AllReduceFusionPattern::kARResidualRMSNormFP8Quant:                                 \
+      DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kARResidualRMSNormFP8Quant, NRanks);      \
+      break;                                                                                 \
+    case AllReduceFusionPattern::kARResidualRMSNormFP4Quant:                                 \
+      if constexpr (!std::is_same_v<T, float>) {                                             \
+        DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kARResidualRMSNormFP4Quant, NRanks);    \
+      } else {                                                                               \
+        FLASHINFER_CHECK(false, "FP4Quant pattern cannot work with DType=float!");           \
+      }                                                                                      \
+      break;                                                                                 \
+    case AllReduceFusionPattern::kARResidualRMSNormOutFP8Quant:                              \
+      DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kARResidualRMSNormOutFP8Quant, NRanks);   \
+      break;                                                                                 \
+    case AllReduceFusionPattern::kARResidualRMSNormOutFP4Quant:                              \
+      if constexpr (!std::is_same_v<T, float>) {                                             \
+        DISPATCH_ACC_TYPE(T, AllReduceFusionPattern::kARResidualRMSNormOutFP4Quant, NRanks); \
+      } else {                                                                               \
+        FLASHINFER_CHECK(false, "OutFP4Quant pattern cannot work with DType=float!");        \
+      }                                                                                      \
+      break;                                                                                 \
+    default:                                                                                 \
+      FLASHINFER_CHECK(false, "Unsupported allreduce fusion pattern!");                      \
+  }
+
+  switch (params.nranks) {
+    case 2:
+      DISPATCH_PATTERN(T, 2);
+      break;
+    case 4:
+      DISPATCH_PATTERN(T, 4);
+      break;
+    case 8:
+      DISPATCH_PATTERN(T, 8);
+      break;
+    case 16:
+      DISPATCH_PATTERN(T, 16);
+      break;
+    default:
+      FLASHINFER_ERROR(
+          "allreduce_fusion_kernel: unsupported ranks number! Supported ranks: 2, 4, 8, 16.");
+  }
+}
+
+}  // namespace trtllm_allreduce_fusion
+
+}  // namespace flashinfer
diff --git a/tests/test_trtllm_allreduce_fusion.py b/tests/test_trtllm_allreduce_fusion.py
new file mode 100644
index 0000000000..a80c144b52
--- /dev/null
+++ b/tests/test_trtllm_allreduce_fusion.py
@@ -0,0 +1,337 @@
+import multiprocessing as mp
+import socket
+from typing import Any
+
+import numpy as np
+import pytest
+import torch
+import torch.distributed as dist
+
+import flashinfer.comm as comm
+
+# todo(Yingyi): add benchmark and quant test
+
+# Usage: test var
+kOneShotMaxTokenNum = 128
+MIN_TOKEN_NUM = 1
+MAX_TOKEN_NUM = 2048
+SF_VEC_SIZE = 16
+
+# temp var
+SCALE_FACTOR_RANGE = (-1, 1)
+
+
+def _run_correctness_worker(world_size, rank, dtype, hidden_dim, distributed_init_port):
+    device = torch.device(f"cuda:{rank}")
+    torch.cuda.set_device(device)
+    distributed_init_method = f"tcp://localhost:{distributed_init_port}"
+    dist.init_process_group(
+        backend="nccl",
+        init_method=distributed_init_method,
+        rank=rank,
+        world_size=world_size,
+    )
+    group = dist.group.WORLD
+
+    try:
+        device = torch.device(f"cuda:{rank}")
+        token_nums = [1, 128, 1024, 2048]
+        pattern_codes = [
+            comm.AllReduceFusionPattern.kAllReduce,
+            comm.AllReduceFusionPattern.kARResidualRMSNorm,
+            comm.AllReduceFusionPattern.kARResidualRMSNormFP8Quant,
+            comm.AllReduceFusionPattern.kARResidualRMSNormFP4Quant,
+            comm.AllReduceFusionPattern.kARResidualRMSNormOutFP8Quant,
+            comm.AllReduceFusionPattern.kARResidualRMSNormOutFP4Quant,
+        ]
+        swizzled_layout_codes = [
+            comm.FP4QuantizationSFLayout.LINEAR,
+            comm.FP4QuantizationSFLayout.SWIZZLED,
+        ]
+        launch_with_pdls = [True, False]
+        use_oneshots = [True, False]
+        trigger_completion_at_ends = [True, False]
+        fp32_accs = [True, False]
+
+        lamport_use_fp32 = dtype == torch.float32
+
+        # create workspace for moe allreduce fusion
+        ipc_handles, workspace_tensor = (
+            comm.trtllm_create_ipc_workspace_for_all_reduce_fusion(
+                rank,
+                world_size,
+                MAX_TOKEN_NUM,
+                hidden_dim,
+                group=group,
+                use_fp32_lamport=lamport_use_fp32,
+            )
+        )
+
+        test_loop = 5
+
+        for token_num in token_nums:
+            for pattern_code in pattern_codes:
+                for swizzled_layout_code in swizzled_layout_codes:
+                    for launch_with_pdl in launch_with_pdls:
+                        for use_oneshot in use_oneshots:
+                            for trigger_completion_at_end in trigger_completion_at_ends:
+                                for fp32_acc in fp32_accs:
+                                    if dtype == torch.float32 and (
+                                        pattern_code
+                                        == comm.AllReduceFusionPattern.kARResidualRMSNormOutFP4Quant
+                                        or pattern_code
+                                        == comm.AllReduceFusionPattern.kARResidualRMSNormFP4Quant
+                                    ):
+                                        continue
+
+                                    dist.barrier(group=group)
+                                    test_passed = True
+                                    print(
+                                        f"test RANK {rank}: token{token_num}-hidden_dim{hidden_dim}-dtype{dtype}-pattern{pattern_code}-layout{swizzled_layout_code}-pdl{launch_with_pdl} start"
+                                    )
+                                    dist.barrier(group=group)
+                                    torch.cuda.synchronize()
+                                    for _ in range(test_loop):
+                                        message_size = token_num * hidden_dim
+
+                                        allreduce_in = torch.randn(
+                                            message_size, dtype=dtype, device=device
+                                        )
+                                        allreduce_in_clone = allreduce_in.clone()
+
+                                        all_reduce_out = torch.zeros(
+                                            message_size, dtype=dtype, device=device
+                                        )
+
+                                        residual_in = torch.randn(
+                                            message_size, dtype=dtype, device=device
+                                        )
+                                        residual_in_clone = residual_in.clone()
+
+                                        residual_out = torch.empty_like(residual_in)
+                                        norm_out = torch.empty_like(residual_in)
+                                        quant_out = torch.empty(
+                                            message_size, dtype=dtype, device=device
+                                        )
+
+                                        scale_out = None
+                                        assert (
+                                            hidden_dim % SF_VEC_SIZE == 0
+                                        ), "hidden_dim must be divisible by SF_VEC_SIZE"
+                                        if (
+                                            swizzled_layout_code
+                                            == comm.FP4QuantizationSFLayout.SWIZZLED
+                                        ):
+                                            # TODO(Yingyi): check this
+                                            padded_message_size = (
+                                                (token_num + 127) // 128 * 128
+                                            ) * ((hidden_dim + 63) // 64 * 4)
+                                            scale_out = torch.empty(
+                                                padded_message_size,
+                                                dtype=dtype,
+                                                device=device,
+                                            )
+                                        else:
+                                            scale_out = torch.empty(
+                                                message_size // SF_VEC_SIZE,
+                                                dtype=dtype,
+                                                device=device,
+                                            )
+
+                                        rms_gamma = torch.randn(
+                                            hidden_dim, dtype=dtype, device=device
+                                        )
+                                        scale_factor = (
+                                            torch.rand(
+                                                1, dtype=torch.float32, device=device
+                                            )
+                                            * (
+                                                SCALE_FACTOR_RANGE[1]
+                                                - SCALE_FACTOR_RANGE[0]
+                                            )
+                                            + SCALE_FACTOR_RANGE[0]
+                                        )
+                                        rms_eps = 1e-3
+
+                                        # NOTE: in real case, you dont have to set all optional params. You could set those required by fusion pattern.
+                                        comm.trtllm_allreduce_fusion(
+                                            allreduce_in=allreduce_in,
+                                            world_size=world_size,
+                                            world_rank=rank,
+                                            token_num=token_num,
+                                            hidden_dim=hidden_dim,
+                                            workspace_ptrs=workspace_tensor,
+                                            launch_with_pdl=launch_with_pdl,
+                                            use_oneshot=use_oneshot,
+                                            trigger_completion_at_end=trigger_completion_at_end,
+                                            fp32_acc=fp32_acc,
+                                            pattern_code=pattern_code,
+                                            allreduce_out=all_reduce_out,
+                                            residual_in=residual_in,
+                                            residual_out=residual_out,
+                                            norm_out=norm_out,
+                                            quant_out=quant_out,
+                                            scale_out=scale_out,
+                                            rms_gamma=rms_gamma,
+                                            rms_eps=rms_eps,
+                                            scale_factor=scale_factor,
+                                            layout_code=swizzled_layout_code,
+                                        )
+                                        torch.cuda.synchronize()
+
+                                        # match shape
+                                        all_reduce_out = all_reduce_out.view(
+                                            token_num, hidden_dim
+                                        )
+                                        residual_out = residual_out.view(
+                                            token_num, hidden_dim
+                                        )
+                                        norm_out = norm_out.view(token_num, hidden_dim)
+
+                                        torch.cuda.synchronize()
+
+                                        # calculate reference
+                                        # allreduce_out
+                                        dist.all_reduce(allreduce_in_clone, group=group)
+                                        ref_allreduce_out = allreduce_in_clone.clone()
+                                        ref_allreduce_out = ref_allreduce_out.view(
+                                            token_num, hidden_dim
+                                        ).to(torch.float32)
+
+                                        # residual_out
+                                        ref_residual_out = (
+                                            ref_allreduce_out
+                                            + residual_in_clone.view(
+                                                token_num, hidden_dim
+                                            ).to(torch.float32)
+                                        )
+
+                                        # norm_out
+                                        variance = (
+                                            ref_residual_out.to(torch.float32)
+                                            .pow(2)
+                                            .mean(dim=-1, keepdim=True)
+                                        )
+                                        hidden_states = ref_residual_out * torch.rsqrt(
+                                            variance + rms_eps
+                                        )
+                                        ref_norm_out = (
+                                            rms_gamma.to(torch.float32) * hidden_states
+                                        )
+
+                                        # check correctness
+                                        tolerance = (
+                                            8e-2 if dtype == torch.float16 else 8e-1
+                                        )
+                                        # compare allreduce_out
+                                        if (
+                                            pattern_code
+                                            == comm.AllReduceFusionPattern.kAllReduce
+                                        ):
+                                            torch.testing.assert_close(
+                                                all_reduce_out.to(torch.float32),
+                                                ref_allreduce_out,
+                                                atol=tolerance,
+                                                rtol=1e-2,
+                                            )
+                                        elif (
+                                            pattern_code
+                                            == comm.AllReduceFusionPattern.kARResidualRMSNormOutFP8Quant
+                                            or pattern_code
+                                            == comm.AllReduceFusionPattern.kARResidualRMSNormOutFP4Quant
+                                        ):
+                                            torch.testing.assert_close(
+                                                residual_out.to(torch.float32),
+                                                ref_residual_out,
+                                                atol=tolerance,
+                                                rtol=1e-2,
+                                            )
+
+                                            torch.testing.assert_close(
+                                                norm_out.to(torch.float32),
+                                                ref_norm_out,
+                                                atol=tolerance,
+                                                rtol=1e-2,
+                                            )
+
+                                        # todo(Yingyi): check quant out
+                                    dist.barrier(group=group)
+                                    if test_passed:
+                                        print(
+                                            f"test RANK {rank}: token{token_num}-hidden_dim{hidden_dim}-dtype{dtype}-pattern{pattern_code}-layout{swizzled_layout_code}-pdl{launch_with_pdl} passed"
+                                        )
+                                    else:
+                                        print(
+                                            f"test RANK {rank}: token{token_num}-hidden_dim{hidden_dim}-dtype{dtype}-pattern{pattern_code}-layout{swizzled_layout_code}-pdl{launch_with_pdl} failed"
+                                        )
+    finally:
+        dist.barrier(group=group)
+
+        comm.trtllm_destroy_ipc_workspace_for_all_reduce(ipc_handles, group=group)
+
+        dist.destroy_process_group(group=group)
+
+
+def get_open_port() -> int:
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("127.0.0.1", 0))
+            return s.getsockname()[1]
+    except OSError:
+        with socket.socket(socket.AF_INET6, socket.SOCK_STREAM) as s:
+            s.bind(("::1", 0))
+            return s.getsockname()[1]
+
+
+def multi_process_parallel(
+    world_size: int,
+    dtype: torch.dtype,
+    hidden_dim: int,
+    test_target: Any,
+    target_args: tuple = (),
+) -> None:
+    mp.set_start_method("spawn", force=True)
+
+    procs = []
+    distributed_init_port = get_open_port()
+    for i in range(world_size):
+        proc_args = (
+            world_size,
+            i,
+            dtype,
+            hidden_dim,
+            distributed_init_port,
+        ) + target_args
+        proc = mp.Process(target=test_target, args=proc_args, name=f"Worker-{i}")
+        proc.start()
+        procs.append(proc)
+
+    for i in range(world_size):
+        procs[i].join()
+        assert (
+            procs[i].exitcode == 0
+        ), f"Process {i} failed with exit code {procs[i].exitcode}"
+
+
+@pytest.mark.parametrize("world_size", [2, 4])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16, torch.float32])
+@pytest.mark.parametrize("hidden_dim", [1024, 2048, 4096, 7168, 8192])
+def test_trtllm_allreduce_fusion(world_size, dtype, hidden_dim):
+    np.random.seed(42)
+    torch.manual_seed(42)
+    torch.cuda.manual_seed_all(42)
+    available_gpus = torch.cuda.device_count()
+    if world_size > available_gpus:
+        raise ValueError(
+            f"world_size {world_size} is greater than available_gpus {available_gpus}"
+        )
+    print(f"Running test for world_size={world_size}")
+
+    multi_process_parallel(
+        world_size,
+        dtype,
+        hidden_dim,
+        _run_correctness_worker,
+        target_args=(),
+    )
+    print(f"moe allreduce fusion tp = {world_size}: OK")
diff --git a/tests/test_trtllm_moe_allreduce_fusion.py b/tests/test_trtllm_moe_allreduce_fusion.py
index 0b0ce186df..3bcd40f66d 100644
--- a/tests/test_trtllm_moe_allreduce_fusion.py
+++ b/tests/test_trtllm_moe_allreduce_fusion.py
@@ -421,7 +421,3 @@ def test_trtllm_moe_allreduce_fusion(world_size, dtype):
         target_args=(),
     )
     print(f"moe allreduce fusion tp = {world_size}: OK")
-
-
-if __name__ == "__main__":
-    test_trtllm_moe_allreduce_fusion(world_size=2, dtype=torch.bfloat16)
