diff --git a/csrc/trtllm_batched_gemm_runner.cu b/csrc/trtllm_batched_gemm_runner.cu
index cff4db198f..f99e766e86 100644
--- a/csrc/trtllm_batched_gemm_runner.cu
+++ b/csrc/trtllm_batched_gemm_runner.cu
@@ -102,7 +102,7 @@ TrtllmGenBatchedGemmRunner::TrtllmGenBatchedGemmRunner(
         (!doesRouteImplUseNoRoute(options.mRouteImpl)) == mOptions.routeAct &&
         options.mFusedAct == mOptions.fusedAct && options.mIsStaticBatch == mOptions.staticBatch &&
         tileSize == mOptions.tileSize &&
-        options.mUseShuffledMatrixA == mOptions.useShuffledMatrixA &&
+        options.mUseShuffledMatrix == mOptions.useShuffledMatrixA &&
         options.mLayoutA == mOptions.weightLayout) {
       if (options.mFusedAct) {
         if (options.mActType != static_cast<batchedGemm::gemmGatedAct::ActType>(mOptions.actType)) {
diff --git a/flashinfer/artifacts.py b/flashinfer/artifacts.py
index 71addb0a15..d656aa057c 100644
--- a/flashinfer/artifacts.py
+++ b/flashinfer/artifacts.py
@@ -89,7 +89,7 @@ class ArtifactPath:
 
     TRTLLM_GEN_FMHA: str = "75d477a640f268ea9ad117cc596eb39245713b9e/fmha/trtllm-gen/"
     TRTLLM_GEN_BMM: str = (
-        "fea3b0ecfe11d7b34556042aeb5d0465ad101500/batched_gemm-332ffef-9936841"
+        "e1e11bbfe0743743620ef997a6d5e8e2dbdf01cf/batched_gemm-2a674db-79e4d37"
     )
     TRTLLM_GEN_GEMM: str = (
         "1fddc48b7b48af33914d040051b3e2ee9ba4701e/gemm-145d1b1-9b113e3"
@@ -110,7 +110,7 @@ class CheckSumHash:
         "e014d7a54c396733ef012b223603c1be2861019f88faa5dcc882ed1ecfe5c2d9"
     )
     TRTLLM_GEN_BMM: str = (
-        "1c3c7ae0755a0acb7ad35da7dbdb90ab71c253dc289051faa2b4e3180dfc4b23"
+        "03b1a419b594b7a4613ea8437c172dc2627d56bd360be25aa604859dc12a05fb"
     )
     DEEPGEMM: str = "1a2a166839042dbd2a57f48051c82cd1ad032815927c753db269a4ed10d0ffbf"
     TRTLLM_GEN_GEMM: str = (
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmInterface.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmInterface.h
index d99e2511d6..99e95d760a 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmInterface.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmInterface.h
@@ -96,6 +96,9 @@ struct BatchedGemmData {
   struct InputBuffers {
     // The matrix A. The data type is controlled by options.mDtypeA.
     //
+    // Sparsity is only supported with batchN.
+    // Let S be the sparsity ratio (1 for dense, 2 for sparse).
+    //
     // If (routeAct == true && batchM), the shape is [M, K]
     // Elseif (batchStrideInTokens > 0)
     //   If batchM:
@@ -104,23 +107,23 @@ struct BatchedGemmData {
     //
     //   If batchN:
     //      If layoutA is MatrixLayout::MajorK
-    //         Logical shape is [B, divUpMul(M, tileM), K].
-    //         Logical strides are [divUpMul(M, tileM) * K, K, 1].
-    //      If layoutA is MatrixLayout::MajorMn
+    //         Logical shape is [B, divUpMul(M, tileM), K / S].
+    //         Logical strides are [divUpMul(M, tileM) * K / S, K / S, 1].
+    //      If layoutA is MatrixLayout::MajorMn (sparsity not supported)
     //         Logical shape is [B, K, divUpMul(M, tileM)].
     //         Logical strides are [K * divUpMul(M, tileM), divUpMul(M, tileM), 1].
     //      If layoutA is MatrixLayout::BlockMajorK
-    //         Logical shape is [B, K / blockK, divUpMul(M, tileM), blockK].
-    //         Logical strides are [K * divUpMul(M, tileM), divUpMul(M, tileM) * blockK, blockK, 1].
-    //         where blockK is 128B.
+    //         Logical shape is [B, K / S / blockK, divUpMul(M, tileM), blockK].
+    //         Logical strides are [K / S * divUpMul(M, tileM), divUpMul(M, tileM) * blockK, blockK,
+    //         1]. where blockK is 128B.
     // Else // batchStrideInTokens == 0
-    //   If batchM:
+    //   If batchM: (sparsity not supported)
     //      Logical shape is [M, K].
     //      Logical strides are [K, 1].
     //
     //   If batchN:
-    //      Logical shape is [B, divUpMul(M, tileM), K].
-    //      Logical strides are [divUpMul(M, tileM) * K, K, 1].
+    //      Logical shape is [B, divUpMul(M, tileM), K / S].
+    //      Logical strides are [divUpMul(M, tileM) * K / S, K / S, 1].
     void const* mPtrA{nullptr};
 
     // The block scaling factors to dequantize A.
@@ -136,9 +139,9 @@ struct BatchedGemmData {
     //      The layout of scaling factors for A is always R128c4
     //      M must be a multiple of 128.
     //      K must be a multiple of 64.
-    //      The "logical" shape is: [paddedM, K / 16].
-    //      The R128c4 layout is: [paddedM / 128, K / 16 / 4, 512].
-    //      The shape we use for TMA is: [paddedM / 128, K / 16 / 4, 2, 256].
+    //      The "logical" shape is: [paddedM, K / P], where P is the scaling block size.
+    //      The R128c4 layout is: [paddedM / 128, K / P / 4, 512].
+    //      The shape we use for TMA is: [paddedM / 128, K / P / 4, 2, 256].
     //  Where paddedM is M if (routeAct == true && batchM), or
     //  sum(divUpMul(M[bi], tileM) for bi in B) if batchM,
     //  otherwise divUpMul(M, tileM) * B.
@@ -209,15 +212,15 @@ struct BatchedGemmData {
     //    If the layout is R128c4,
     //       paddedN must be a multiple of 128.
     //       K must be a multiple of 64.
-    //       The R128c4 layout is: [paddedN / 128, K / 16 / 4, 512]
-    //       The shape we use for TMA is: [paddedN / 128, K / 16 / 4, 2, 256]
+    //       The R128c4 layout is: [paddedN / 128, K / P / 4, 512], where P is the scaling block
+    //       size. The shape we use for TMA is: [paddedN / 128, K / P / 4, 2, 256]
     //
     //    If the layout is R8c4,
     //       paddedN must be a multiple of 8.
     //       K must be a multiple of 64.
-    //       The R8c4 layout is: [paddedN / 8, K / 16 / 4, 32]
-    //       The shape we use for TMA is: [paddedN / 8, K / 16 / 4 / repeats, repeats * 32]
-    //       where repeats = min(tileK / 16 / 4, 8)
+    //       The R8c4 layout is: [paddedN / 8, K / P / 4, 32], where P is the scaling block size.
+    //       The shape we use for TMA is: [paddedN / 8, K / P / 4 / repeats, repeats * 32]
+    //       where repeats = min(tileK / P / 4, 8)
     //
     // where paddedN is N if (routeAct == true && batchN),
     // or sum(divUpMul(N[bi], tileN) for bi in B) if batchN,
@@ -243,6 +246,25 @@ struct BatchedGemmData {
     //     Logical shape is [sum(divUpMul(N[bi], tileN) for bi in B)]
     void const* mPtrPerTokenSfB{nullptr};
 
+    // The sparsity information of A, if structured sparsity is used.
+    // Only supported for batchN (A is weights).
+    //
+    // When sparsityA is Any_2_4:
+    //     2 elements are non-zero in any chunk of 4 elements.
+    //     A 4-bit index indicates the position of the non-zero elements.
+    //     The shape in Uint8 is: [B, divUpMul(M, tileM), K / 8]
+    //     (two 4-bit indices packed into one UInt8)
+    //
+    // When sparsityA is Pairwise_4_8:
+    //     4 elements are non-zero in any chunk of 8 elements.
+    //     The zero and non-zero elements are grouped in pairs.
+    //     A 4-bit index indicates the position of the non-zero pairs.
+    //     The shape in Uint8 is: [B, divUpMul(M, tileM), K / 16]
+    //     (two 4-bit indices packed into one UInt8)
+    //
+    // If sparsityA is Dense, this should be set to nullptr.
+    void const* mPtrSparsityInfoA{nullptr};
+
     // The bias applied after the GEMM and before the activation function.
     // The bias is applied before applying the global scaling factor. I.e.
     // C = act(A * B + bias') * scaleC
@@ -381,7 +403,8 @@ struct BatchedGemmData {
     // Computed as
     // int32_t totalNumPaddedTokens{0};
     // for (int bi = 0; bi < options.mNumBatches; bi++) {
-    //   totalNumPaddedTokens += batchM ? divUpMul(options.mBatchedM[bi], options.mTileM)
+    //   totalNumPaddedTokens += batchM ? divUpMul(options.mBatchedM[bi], options.mTileM *
+    //   options.mClusterDimX)
     //                                  : divUpMul(options.mBatchedN[bi], options.mTileN);
     // }
     // The size is 1 and the dtype is int32_t.
@@ -430,26 +453,26 @@ struct BatchedGemmData {
     // The output block scaling factors for C.
     //
     // If MxFp{4,8} and NvFp4 formats are used,
-    // The "logical" shape is:
-    //    if batchM: [paddedM, N / 16]
-    //    if batchN: [paddedN, M / 16]
+    // The "logical" shape is (P is the scaling block size):
+    //    if batchM: [paddedM, N / P]
+    //    if batchN: [paddedN, M / P]
     // where paddedM is sum(divUpMul(M[bi], tileM) for bi in B),
     // where paddedN is sum(divUpMul(N[bi], tileN) for bi in B).
     //
     // If the layout is R128c4,
     //    paddedOuter must be a multiple of 128.
     //    inner must be a multiple of 64.
-    //    The R128c4 layout is: [paddedOuter / 128, inner / 16 / 4, 512]
-    //    The shape we use for TMA is: [paddedOuter / 128, inner / 16 / 4, 2, 256]
+    //    The R128c4 layout is: [paddedOuter / 128, inner / P / 4, 512]
+    //    The shape we use for TMA is: [paddedOuter / 128, inner / P / 4, 2, 256]
     //    where inner = N if batchM, otherwise M.
     //    where paddedOuter = paddedM if batchM, otherwise paddedN.
     //
     // If the layout is R8c4,
     //    paddedOuter must be a multiple of 8.
     //    inner must be a multiple of 64.
-    //    The R8c4 layout is: [paddedOuter / 8, inner / 16 / 4, 32]
-    //    The shape we use for TMA is: [paddedOuter / 8, inner / 16 / 4 / repeats, repeats * 32]
-    //    where repeats = min(tileInner / 16 / 4, 8),
+    //    The R8c4 layout is: [paddedOuter / 8, inner / P / 4, 32]
+    //    The shape we use for TMA is: [paddedOuter / 8, inner / P / 4 / repeats, repeats * 32]
+    //    where repeats = min(tileInner / P / 4, 8),
     //    where tileInner = tileN if batchM, otherwise tileM,
     //    where paddedOuter = paddedM if batchM, otherwise paddedN.
     //    where inner = N if batchM, otherwise M.
@@ -528,11 +551,13 @@ class BatchedGemmInterface {
 
     auto [numCtaBatch, numCtaTile, numCtaInner] =
         getGridDim(options, batchedGemmData.mProblemDimensions.mMaxNumCtasInTokenDim);
+
     auto kernelParams = KernelParamsSetup::setKernelParams(
         options, batchM, batchedGemmData.mInputBuffers.mPtrA, batchedGemmData.mInputBuffers.mPtrB,
         batchedGemmData.mOutputBuffers.mPtrC, batchedGemmData.mInputBuffers.mPtrSfA,
         batchedGemmData.mInputBuffers.mPtrSfB, batchedGemmData.mInputBuffers.mPtrPerTokenSfA,
-        batchedGemmData.mInputBuffers.mPtrPerTokenSfB, batchedGemmData.mInputBuffers.mPtrBias,
+        batchedGemmData.mInputBuffers.mPtrPerTokenSfB,
+        batchedGemmData.mInputBuffers.mPtrSparsityInfoA, batchedGemmData.mInputBuffers.mPtrBias,
         batchedGemmData.mOutputBuffers.mPtrSfC, batchedGemmData.mInputBuffers.mPtrScaleC,
         batchedGemmData.mInputBuffers.mPtrScaleAct, batchedGemmData.mInputBuffers.mPtrScaleGate,
         batchedGemmData.mInputBuffers.mPtrClampLimit,
@@ -544,8 +569,7 @@ class BatchedGemmInterface {
         batchedGemmData.mInputBuffers.mPtrCtaIdxXyToMnLimit, numCtaBatch);
 
     // The size of the grid.
-    std::vector<int32_t> grid = batchM ? std::vector<int32_t>{numCtaBatch, numCtaTile, numCtaInner}
-                                       : std::vector<int32_t>{numCtaTile, numCtaBatch, numCtaInner};
+    auto grid = getLaunchGrid(options, batchedGemmData.mProblemDimensions.mMaxNumCtasInTokenDim);
 
     BatchedGemmConfig batchedGemmConfig = config;
 #ifndef TLLM_GEN_EXPORT_INTERFACE
@@ -676,8 +700,10 @@ class BatchedGemmInterface {
     // For normal BMM, mNumTokens == 0 and the number of CTAs is known to host.
     if (options.mIsStaticBatch) {
       for (int32_t bi = 0; bi < options.mNumBatches; ++bi) {
-        numCtasBatch += batchM ? gemm::divUp(options.mBatchedM[bi], options.mTileM)
-                               : gemm::divUp(options.mBatchedN[bi], options.mTileN);
+        numCtasBatch +=
+            batchM ? gemm::divUp(options.mBatchedM[bi], options.mTileM * options.mClusterDimX) *
+                         options.mClusterDimX
+                   : gemm::divUp(options.mBatchedN[bi], options.mTileN);
       }
     }
     // For MoE, mNumTokens != 0 and the number of CTAs is known only at runtime.
@@ -711,11 +737,24 @@ class BatchedGemmInterface {
 
   //////////////////////////////////////////////////////////////////////////////////////////////////
 
+  // Returns the number of CTAs of the current kernel.
+  std::vector<int32_t> getLaunchGrid(
+      BatchedGemmOptions const& options,
+      std::optional<int32_t> maxNumCtasInBatchDim = std::nullopt) const {
+    auto [numCtaBatch, numCtaTile, numCtaInner] = getGridDim(options, maxNumCtasInBatchDim);
+    bool const batchM = options.mBatchMode == BatchedGemmOptions::BatchMode::BatchM;
+    std::vector<int32_t> grid = batchM ? std::vector<int32_t>{numCtaBatch, numCtaTile, numCtaInner}
+                                       : std::vector<int32_t>{numCtaTile, numCtaBatch, numCtaInner};
+    return grid;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////////
+
   // Returns the number of CTAs of the current kernel.
   int32_t getNumCtas(BatchedGemmOptions const& options,
                      std::optional<int32_t> maxNumCtasInBatchDim = std::nullopt) const {
-    auto [numCtasBatch, numCtasTile, numCtasInner] = getGridDim(options, maxNumCtasInBatchDim);
-    return numCtasBatch * numCtasTile * numCtasInner;
+    auto grid = getLaunchGrid(options, maxNumCtasInBatchDim);
+    return grid[0] * grid[1] * grid[2];
   }
 
   //////////////////////////////////////////////////////////////////////////////////////////////////
@@ -779,19 +818,20 @@ class BatchedGemmInterface {
       auto const batchM = options.mBatchMode == BatchedGemmOptions::BatchMode::BatchM;
       if (!options.mEnablesEarlyExit || options.mNumTokens == 0) {
         for (int32_t bi = 0; bi < options.mNumBatches; ++bi) {
-          totalNumPaddedTokens += batchM ? gemm::divUpMul(options.mBatchedM[bi], options.mTileM)
-                                         : gemm::divUpMul(options.mBatchedN[bi], options.mTileN);
+          totalNumPaddedTokens +=
+              batchM ? gemm::divUpMul(options.mBatchedM[bi], options.mTileM * options.mClusterDimX)
+                     : gemm::divUpMul(options.mBatchedN[bi], options.mTileN);
         }
       } else {
         // Get tile in token dim.
-        auto tileTokensDim = batchM ? options.mTileM : options.mTileN;
+        auto tileTokensDim = batchM ? options.mTileM * options.mClusterDimX : options.mTileN;
         totalNumPaddedTokens = data.mProblemDimensions.mMaxNumCtasInTokenDim * tileTokensDim;
       }
 
       // Get options from config.
       auto& options = config.mOptions;
 
-      int const tokenTile = batchM ? options.mTileM : options.mTileN;
+      int const tokenTile = batchM ? options.mTileM * options.mClusterDimX : options.mTileN;
 
       auto const numTokens = totalNumPaddedTokens;
       auto const intermediateDim = batchM ? options.mN : options.mM;
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmOptions.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmOptions.h
index acbc01497c..9c571d0258 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmOptions.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/BatchedGemmOptions.h
@@ -94,26 +94,28 @@ struct BatchedGemmOptions : public gemmGatedAct::GemmGatedActOptions {
       bool hoistMmaTaskTryWaits, int k, gemm::KernelTraits kernelTraits, gemm::MatrixLayout layoutA,
       gemm::MatrixLayout layoutB, int m, int mmaK, tg::MmaKind mmaKind, int mmaM, int mmaN,
       bool mockAllReduce, int n, int numEpilogueWarps, int numRegsCastAWarps,
-      int numRegsCopySfLdsSttm, int numRegsPerThreadEpilogueWarp,
+      int numRegsCopySfLdsSttm, int numRegsCopySparsityInfo, int numRegsPerThreadEpilogueWarp,
       int numRegsPerThreadNonEpilogueWarp, int numSlicesForSplitK, int numSlicesForSliceK,
       int numStages, int numStagesMma, int numStagesMmaWithinWorkTile,
       int numStagesMmaAcrossWorkTile, int numStagesWorkId, bool outputDebugTensors, bool patchF2fp,
-      std::optional<int32_t> sfBlockSizeA, tg::SfLayout sfLayoutA, tg::SfLayout sfLayoutB,
-      tg::SfLayout sfLayoutC, int32_t sfReshapeFactor, bool sliceK, gemm::SplitK splitK, int tileK,
-      int tileM, int tileN, gemm::TileScheduler tileScheduler, bool transposeMmaOutput,
-      bool useCustomMmaSchedule, bool useDeepSeekFp8, bool useHoistTryWaitForCustomMmaSchedule,
-      bool useMaxTmemOverlap, bool usePerTokenSfA, bool usePerTokenSfB, bool useShuffledMatrixA,
-      bool useTmaStore, bool useTwoTmaLoadWarps, bool useTwoMmaWarps, bool useUnrollLoop2xForMma,
-      int validM, int validN, int validK, int worldSize,
+      int32_t sfBlockSizeA, int32_t sfBlockSizeB, int32_t sfBlockSizeC, tg::SfLayout sfLayoutA,
+      tg::SfLayout sfLayoutB, tg::SfLayout sfLayoutC, int32_t sfReshapeFactor, bool sliceK,
+      tg::Sparsity sparsityA, gemm::SplitK splitK, int tileK, int tileM, int tileN,
+      gemm::TileScheduler tileScheduler, bool transposeMmaOutput, bool useCustomMmaSchedule,
+      bool useDeepSeekFp8, bool useHoistTryWaitForCustomMmaSchedule, bool useMaxTmemOverlap,
+      bool usePerTokenSfA, bool usePerTokenSfB, bool useShuffledMatrix, bool useTmaStore,
+      bool useTwoTmaLoadWarps, bool useTwoMmaWarps, bool useUnrollLoop2xForMma, int validM,
+      int validN, int validK, int worldSize,
       // GemmGatedActOptions
       gemmGatedAct::ActType actType, bool clampBeforeAct,
       // BatchedGemmOptions
       std::vector<int> batchedM, std::vector<int> batchedN, BatchMode batchMode,
       int32_t batchStrideInTokens, bool fusedAct, bool gridWaitForPrimaryRouting,
       bool isStaticBatch, bool isUniformNumTokensPerBatch, int numBatches,
-      int numRegsPerThreadLoadB, int numRegsPerThreadLoadSfB, int numTokens, int numWarpsLoadB,
-      int numWarpsLoadSfB, RouteImpl routeImpl, std::optional<RouteImpl> routeSfsImpl,
-      bool useTmaOobOpt)
+      int numRegsPerThreadLoadA, int numRegsPerThreadLoadB, int numRegsPerThreadLoadSfA,
+      int numRegsPerThreadLoadSfB, int numTokens, int numWarpsLoadA, int numWarpsLoadB,
+      int numWarpsLoadSfA, int numWarpsLoadSfB, RouteImpl routeImpl,
+      std::optional<RouteImpl> routeSfsImpl, bool useTmaOobOpt)
       : gemmGatedAct::GemmGatedActOptions(
             gemm::GemmOptions(
                 allReduceAlgo, biasType, blockK, clusterDimX, clusterDimY, clusterDimZ,
@@ -124,15 +126,16 @@ struct BatchedGemmOptions : public gemmGatedAct::GemmGatedActOptions {
                 gridWaitForPrimaryEarlyExit, gridWaitForPrimaryA, gridWaitForPrimaryB,
                 hoistLoadTaskInit, hoistMmaTaskTryWaits, k, kernelTraits, layoutA, layoutB, m, mmaK,
                 mmaKind, mmaM, mmaN, mockAllReduce, n, numEpilogueWarps, numRegsCastAWarps,
-                numRegsCopySfLdsSttm, numRegsPerThreadEpilogueWarp, numRegsPerThreadNonEpilogueWarp,
-                numSlicesForSplitK, numSlicesForSliceK, numStages, numStagesMma,
-                numStagesMmaWithinWorkTile, numStagesMmaAcrossWorkTile, numStagesWorkId,
-                outputDebugTensors, patchF2fp, sfBlockSizeA, sfLayoutA, sfLayoutB, sfLayoutC,
-                sfReshapeFactor, sliceK, splitK, tileK, tileM, tileN, tileScheduler,
-                transposeMmaOutput, useCustomMmaSchedule, useDeepSeekFp8,
-                useHoistTryWaitForCustomMmaSchedule, useMaxTmemOverlap, usePerTokenSfA,
-                usePerTokenSfB, useShuffledMatrixA, useTmaStore, useTwoTmaLoadWarps, useTwoMmaWarps,
-                useUnrollLoop2xForMma, validM, validN, validK, worldSize),
+                numRegsCopySfLdsSttm, numRegsCopySparsityInfo, numRegsPerThreadEpilogueWarp,
+                numRegsPerThreadNonEpilogueWarp, numSlicesForSplitK, numSlicesForSliceK, numStages,
+                numStagesMma, numStagesMmaWithinWorkTile, numStagesMmaAcrossWorkTile,
+                numStagesWorkId, outputDebugTensors, patchF2fp, sfBlockSizeA, sfBlockSizeB,
+                sfBlockSizeC, sfLayoutA, sfLayoutB, sfLayoutC, sfReshapeFactor, sliceK, sparsityA,
+                splitK, tileK, tileM, tileN, tileScheduler, transposeMmaOutput,
+                useCustomMmaSchedule, useDeepSeekFp8, useHoistTryWaitForCustomMmaSchedule,
+                useMaxTmemOverlap, usePerTokenSfA, usePerTokenSfB, useShuffledMatrix, useTmaStore,
+                useTwoTmaLoadWarps, useTwoMmaWarps, useUnrollLoop2xForMma, validM, validN, validK,
+                worldSize),
             actType, clampBeforeAct),
         mBatchedM(batchedM),
         mBatchedN(batchedN),
@@ -143,10 +146,14 @@ struct BatchedGemmOptions : public gemmGatedAct::GemmGatedActOptions {
         mIsStaticBatch(isStaticBatch),
         mIsUniformNumTokensPerBatch(isUniformNumTokensPerBatch),
         mNumBatches(numBatches),
+        mNumRegsPerThreadLoadA{numRegsPerThreadLoadA},
         mNumRegsPerThreadLoadB{numRegsPerThreadLoadB},
+        mNumRegsPerThreadLoadSfA{numRegsPerThreadLoadSfA},
         mNumRegsPerThreadLoadSfB{numRegsPerThreadLoadSfB},
         mNumTokens(numTokens),
+        mNumWarpsLoadA{numWarpsLoadA},
         mNumWarpsLoadB{numWarpsLoadB},
+        mNumWarpsLoadSfA{numWarpsLoadSfA},
         mNumWarpsLoadSfB{numWarpsLoadSfB},
         mRouteImpl(routeImpl),
         mRouteSfsImpl(routeSfsImpl),
@@ -171,14 +178,22 @@ struct BatchedGemmOptions : public gemmGatedAct::GemmGatedActOptions {
   bool mIsUniformNumTokensPerBatch{false};
   // Number of Gemm batches.
   int mNumBatches;
+  // Number of registers per thread for load A
+  int mNumRegsPerThreadLoadA{0};
   // Number of registers per thread for load B
   int mNumRegsPerThreadLoadB{0};
+  // Number of registers per thread for load SfA
+  int mNumRegsPerThreadLoadSfA{0};
   // Number of registers per thread for load SfB
   int mNumRegsPerThreadLoadSfB{0};
   // Total number of tokens.
   int mNumTokens{32};
+  // Number of warps for load A
+  int mNumWarpsLoadA{0};
   // Number of warps for load B
   int mNumWarpsLoadB{0};
+  // Number of warps for load SfA
+  int mNumWarpsLoadSfA{0};
   // Number of warps for load SfB
   int mNumWarpsLoadSfB{0};
   // Whether load the input tokens and do routing.
@@ -231,6 +246,7 @@ inline bool checkAndUpdateBatchedGemmOptions(BatchedGemmOptions& options, tg::Cu
   }
 
   if (batchM) {
+    TLLM_CHECK_ERROR(!tg::isSparse(options.mSparsityA), "Sparsity is not supported with batchM.");
     TLLM_CHECK_ERROR(options.mN > 0 && options.mK > 0, "N and K must be larger than 0");
     TLLM_CHECK_ERROR(options.mN >= options.mTileN, "N must be equal or larger than TileN.");
     TLLM_CHECK_ERROR(options.mN % options.mTileN == 0, "N must be divisible by TileN.");
@@ -287,7 +303,6 @@ inline bool checkAndUpdateBatchedGemmOptions(BatchedGemmOptions& options, tg::Cu
                    "LdgPlusSts does not support routing the tokens");
 
   if (options.mRouteSfsImpl.has_value() && options.mRouteSfsImpl.value() == RouteImpl::LdgPlusSts) {
-    TLLM_CHECK_ERROR(!batchM, "LdgPlusSts only supports batch N");
     TLLM_CHECK_ERROR(options.mTileK <= 512 && options.mTileK >= 128,
                      "LdgPlusSts only supports 128 <= tileK <= 512");
   }
@@ -330,16 +345,13 @@ inline bool checkAndUpdateBatchedGemmOptions(BatchedGemmOptions& options, tg::Cu
       TLLM_CHECK_ERROR(!batchM, "UTMALDG.GATHER4 only supported for batch N.");
 
       if (tg::mmaKindIsBlockFmt(options.mMmaKind)) {
-        auto dtypeRoute = batchM ? options.mDtypeA : options.mDtypeB;
-        TLLM_CHECK_ERROR(options.mTileK % (tg::dtypeNumEltsPerSf(dtypeRoute) * 16) == 0,
-                         "tileK needs to be a multiple of 16 * tg::dtypeNumEltsPerSf(dtypeA).");
+        int const numEltsPerSfRoute = batchM ? options.mSfBlockSizeA : options.mSfBlockSizeB;
+        TLLM_CHECK_ERROR(options.mTileK % (numEltsPerSfRoute * 16) == 0,
+                         "tileK needs to be a multiple of 16 * numEltsPerSf (", numEltsPerSfRoute,
+                         ") = ", numEltsPerSfRoute * 16);
       }
     }
 
-    if (options.mClusterDimX > 1) {
-      TLLM_CHECK_ERROR(!batchM, "2CTA Gemm currently only supports batch N.");
-    }
-
     if (!batchM || doesRouteImplUseNoRoute(options.mRouteImpl)) {
       TLLM_CHECK_ERROR(options.mSfLayoutA == tg::SfLayout::R128c4,
                        "options.mSfLayoutA has to be tg::SfLayout::R128c4 when not being routed");
@@ -369,10 +381,9 @@ inline bool checkAndUpdateBatchedGemmOptions(BatchedGemmOptions& options, tg::Cu
                      "K must be a multiple of tileK when using Ldg based SF routing");
   }
 
-  if (options.mClusterDimX > 1 && batchM && options.mRouteImpl != RouteImpl::NoRoute) {
-    TLLM_CHECK_ERROR(false,
-                     "2CTA BatchedGemm does not support routing along M dimension. To support it, "
-                     "change the input routing data layout to be padded to clusterDimX size.");
+  if (options.mClusterDimX > 1 && batchM && options.mRouteSfsImpl.has_value()) {
+    TLLM_CHECK_ERROR(options.mRouteSfsImpl.value() != RouteImpl::Tma,
+                     "2CTA BatchedGemm does not support routing Sf along M dimension with TMA.");
   }
 
   // Check if all elements in mBatchedM or mBatchedN are the same (uniform tokens per batch) and
@@ -475,12 +486,16 @@ inline std::string dumpOptions(BatchedGemmOptions const& options, bool dumpRunti
   if (dumpRuntimeParams) {
     ss << "mNumBatches=" << options.mNumBatches << "," << std::endl;
   }
+  ss << "mNumRegsPerThreadLoadA=" << options.mNumRegsPerThreadLoadA << "," << std::endl;
   ss << "mNumRegsPerThreadLoadB=" << options.mNumRegsPerThreadLoadB << "," << std::endl;
+  ss << "mNumRegsPerThreadLoadSfA=" << options.mNumRegsPerThreadLoadSfA << "," << std::endl;
   ss << "mNumRegsPerThreadLoadSfB=" << options.mNumRegsPerThreadLoadSfB << "," << std::endl;
   if (dumpRuntimeParams) {
     ss << "mNumTokens=" << options.mNumTokens << "," << std::endl;
   }
+  ss << "mNumWarpsLoadA=" << options.mNumWarpsLoadA << "," << std::endl;
   ss << "mNumWarpsLoadB=" << options.mNumWarpsLoadB << "," << std::endl;
+  ss << "mNumWarpsLoadSfA=" << options.mNumWarpsLoadSfA << "," << std::endl;
   ss << "mNumWarpsLoadSfB=" << options.mNumWarpsLoadSfB << "," << std::endl;
   ss << "mRouteImpl=batchedGemm::RouteImpl(" << static_cast<int32_t>(options.mRouteImpl) << "),"
      << std::endl;
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmGatedActOptions.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmGatedActOptions.h
index b3d955dc7c..adca177a00 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmGatedActOptions.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmGatedActOptions.h
@@ -76,6 +76,8 @@ enum class ActType {
   // where x0 and x1 are the raw numbers from Gemm, while scaleC and scaleGate are input scales,
   // beta' = beta / scaleAb, scaleC' = scaleC * scaleAb.
   GeGlu,
+  // Placeholder for no activation; not implemented in codegen
+  None,
 };
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
@@ -121,6 +123,15 @@ struct GemmGatedActOptions : public gemm::GemmOptions {
 // Check if the options are valid or not.
 inline bool checkAndUpdateGemmGatedActOptions(gemmGatedAct::GemmGatedActOptions& options,
                                               tg::CudaArch cudaArch, bool updateOptions = true) {
+  auto isValid = gemm::checkAndUpdateGemmOptions(options, cudaArch,
+                                                 /* tpGrpSize */ 1, updateOptions);
+  if (!isValid) {
+    return false;
+  }
+
+  if (options.mActType == gemmGatedAct::ActType::None) {
+    TLLM_CHECK_ERROR(false, "ActType None is not supported");
+  }
   // tmpOut is already transposed at this stage
   auto const hiddenSizeStr = options.mTransposeMmaOutput ? "M" : "N";
   auto const hiddenSize = options.mTransposeMmaOutput ? options.mM : options.mN;
@@ -129,8 +140,8 @@ inline bool checkAndUpdateGemmGatedActOptions(gemmGatedAct::GemmGatedActOptions&
 
   TLLM_CHECK_ERROR(hiddenSize % 2 == 0, hiddenSizeStr, " must be a multiple of 2.");
 
-  TLLM_CHECK_ERROR((options.mTransposeMmaOutput ^ options.mUseShuffledMatrixA) == 0,
-                   "Transpose mma output can only be used with shuffled A matrix. And vice versa.");
+  TLLM_CHECK_ERROR((options.mTransposeMmaOutput && !options.mUseShuffledMatrix) == false,
+                   "Transpose mma output can only be used with shuffled matrix.");
 
   if (options.mUseTmaStore) {
     TLLM_CHECK_ERROR(
@@ -140,18 +151,11 @@ inline bool checkAndUpdateGemmGatedActOptions(gemmGatedAct::GemmGatedActOptions&
 
   if (options.mDtypeC == tg::Dtype::E2m1 || options.mDtypeC == tg::Dtype::MxE4m3) {
     int const outHiddenSize = (options.mTransposeMmaOutput ? options.mM : options.mN) / 2;
-    int const hiddenGranularity = 4 * tg::dtypeNumEltsPerSf(options.mDtypeC);
+    int const hiddenGranularity = 4 * options.mSfBlockSizeC;
     TLLM_CHECK_ERROR(outHiddenSize % hiddenGranularity == 0, "Output hidden size (", outHiddenSize,
                      ") must be a multiple of ", hiddenGranularity, " for block-scaled outputs.");
   }
 
-  auto isValid = gemm::checkAndUpdateGemmOptions(options, cudaArch,
-                                                 /* tpGrpSize */ 1, updateOptions);
-
-  if (!isValid) {
-    return false;
-  }
-
   auto const validHiddenSize = options.mTransposeMmaOutput ? options.mValidM : options.mValidN;
   if (options.mUseDeepSeekFp8) {
     TLLM_CHECK_ERROR(hiddenSize % 256 == 0 && validHiddenSize % 256 == 0, "Hidden size (",
@@ -160,12 +164,12 @@ inline bool checkAndUpdateGemmGatedActOptions(gemmGatedAct::GemmGatedActOptions&
   }
 
   //
-  if (options.mUseShuffledMatrixA) {
+  if (options.mUseShuffledMatrix) {
     auto const shuffleBlockSize = gemm::getShuffleBlockSize(options.mEpilogueTileM);
     TLLM_CHECK_ERROR(
         hiddenSize % (2 * shuffleBlockSize) == 0 && validHiddenSize % (2 * shuffleBlockSize) == 0,
         "M/validM must be a multiple of 2 * shuffle block size (", 2 * shuffleBlockSize,
-        ") when useShuffledMatrixA");
+        ") when useShuffledMatrix");
   }
   if (options.mNumSlicesForSplitK > 1) {
     TLLM_CHECK_ERROR(doesSplitKUseDsmem(options.mSplitK),
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmOptions.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmOptions.h
index 37da75c10d..1ee3fda469 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmOptions.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/GemmOptions.h
@@ -27,6 +27,7 @@
 #include "trtllm/gen/DtypeDecl.h"
 #include "trtllm/gen/MmaDecl.h"
 #include "trtllm/gen/SfLayoutDecl.h"
+#include "trtllm/gen/SparsityDecl.h"
 #ifndef TLLM_GEN_EXPORT_INTERFACE
 #include "trtllm/gen/CudaRunner.h"
 #include "trtllm/gen/GenCtx.h"
@@ -80,6 +81,12 @@ void printArgs(T first, Args... args) {
 
 #endif  // TLLM_GEN_EXPORT_INTERFACE
 
+#define GEMM_UPDATE_OR_ERROR(OPTION, VALUE) \
+  if (updateOptions) {                      \
+    OPTION = VALUE;                         \
+  } else                                    \
+    return false
+
 namespace trtllm {
 namespace gen {
 class CudaRunner;
@@ -115,16 +122,17 @@ struct GemmOptions {
               bool hoistLoadTaskInit, bool hoistMmaTaskTryWaits, int k, KernelTraits kernelTraits,
               MatrixLayout layoutA, MatrixLayout layoutB, int m, int mmaK, tg::MmaKind mmaKind,
               int mmaM, int mmaN, bool mockAllReduce, int n, int numEpilogueWarps,
-              int numRegsCastAWarps, int numRegsCopySfLdsSttm, int numRegsPerThreadEpilogueWarp,
-              int numRegsPerThreadNonEpilogueWarp, int numSlicesForSplitK, int numSlicesForSliceK,
-              int numStages, int numStagesMma, int numStagesMmaWithinWorkTile,
-              int numStagesMmaAcrossWorkTile, int numStagesWorkId, bool outputDebugTensors,
-              bool patchF2fp, std::optional<int32_t> sfBlockSizeA, tg::SfLayout sfLayoutA,
-              tg::SfLayout sfLayoutB, tg::SfLayout sfLayoutC, int sfReshapeFactor, bool sliceK,
+              int numRegsCastAWarps, int numRegsCopySfLdsSttm, int numRegsCopySparsityInfo,
+              int numRegsPerThreadEpilogueWarp, int numRegsPerThreadNonEpilogueWarp,
+              int numSlicesForSplitK, int numSlicesForSliceK, int numStages, int numStagesMma,
+              int numStagesMmaWithinWorkTile, int numStagesMmaAcrossWorkTile, int numStagesWorkId,
+              bool outputDebugTensors, bool patchF2fp, int32_t sfBlockSizeA, int32_t sfBlockSizeB,
+              int32_t sfBlockSizeC, tg::SfLayout sfLayoutA, tg::SfLayout sfLayoutB,
+              tg::SfLayout sfLayoutC, int sfReshapeFactor, bool sliceK, tg::Sparsity sparsityA,
               SplitK splitK, int tileK, int tileM, int tileN, TileScheduler tileScheduler,
               bool transposeMmaOutput, bool useCustomMmaSchedule, bool useDeepSeekFp8,
               bool useHoistTryWaitForCustomMmaSchedule, bool useMaxTmemOverlap, bool usePerTokenSfA,
-              bool usePerTokenSfB, bool useShuffledMatrixA, bool useTmaStore,
+              bool usePerTokenSfB, bool useShuffledMatrix, bool useTmaStore,
               bool useTwoTmaLoadWarps, bool useTwoMmaWarps, bool useUnrollLoop2xForMma, int validM,
               int validN, int validK, int worldSize)
       : mAllReduceAlgo{allReduceAlgo},
@@ -170,6 +178,7 @@ struct GemmOptions {
         mNumEpilogueWarps{numEpilogueWarps},
         mNumRegsCastAWarps(numRegsCastAWarps),
         mNumRegsCopySfLdsSttm(numRegsCopySfLdsSttm),
+        mNumRegsCopySparsityInfo(numRegsCopySparsityInfo),
         mNumRegsPerThreadEpilogueWarp(numRegsPerThreadEpilogueWarp),
         mNumRegsPerThreadNonEpilogueWarp(numRegsPerThreadNonEpilogueWarp),
         mNumSlicesForSplitK{numSlicesForSplitK},
@@ -182,11 +191,14 @@ struct GemmOptions {
         mOutputDebugTensors{outputDebugTensors},
         mPatchF2fp{patchF2fp},
         mSfBlockSizeA{sfBlockSizeA},
+        mSfBlockSizeB{sfBlockSizeB},
+        mSfBlockSizeC{sfBlockSizeC},
         mSfLayoutA{sfLayoutA},
         mSfLayoutB{sfLayoutB},
         mSfLayoutC{sfLayoutC},
         mSfReshapeFactor{sfReshapeFactor},
         mSliceK{sliceK},
+        mSparsityA{sparsityA},
         mSplitK{splitK},
         mTileK{tileK},
         mTileM{tileM},
@@ -199,7 +211,7 @@ struct GemmOptions {
         mUseMaxTmemOverlap{useMaxTmemOverlap},
         mUsePerTokenSfA{usePerTokenSfA},
         mUsePerTokenSfB{usePerTokenSfB},
-        mUseShuffledMatrixA{useShuffledMatrixA},
+        mUseShuffledMatrix{useShuffledMatrix},
         mUseTmaStore{useTmaStore},
         mUseTwoTmaLoadWarps{useTwoTmaLoadWarps},
         mUseTwoMmaWarps{useTwoMmaWarps},
@@ -302,6 +314,8 @@ struct GemmOptions {
   int mNumRegsCastAWarps{0};
   // Number of registers for the LDS+STTM warps.
   int mNumRegsCopySfLdsSttm{0};
+  // Number of registers per thread to copy sparsity info with LDS+STTM.
+  int mNumRegsCopySparsityInfo{0};
   // Number of registers per thread for epilogue warps
   int mNumRegsPerThreadEpilogueWarp{0};
   // Number of registers per thread for non-epilogue warps
@@ -328,8 +342,12 @@ struct GemmOptions {
   bool mOutputDebugTensors{false};
   // Patch float conversions.
   bool mPatchF2fp{false};
-  // Block size of A. For dtypeA == E2m1 and dtypeB == E4m3.
-  std::optional<int32_t> mSfBlockSizeA{std::nullopt};
+  // Block size of A, for block-scaled types.
+  int mSfBlockSizeA{-1};
+  // Block size of B, for block-scaled types.
+  int mSfBlockSizeB{-1};
+  // Block size of C, for block-scaled types.
+  int mSfBlockSizeC{-1};
   // Scale factors layout for A.
   tg::SfLayout mSfLayoutA{tg::SfLayout::R128c4};
   // Scale factors layout for B.
@@ -344,6 +362,8 @@ struct GemmOptions {
   int mSfReshapeFactor{1};
   // Slice-K implementation to use TileM dimension for TileK.
   bool mSliceK{false};
+  // Sparsity of A.
+  tg::Sparsity mSparsityA{tg::Sparsity::Dense};
   // The location of the exchange for split-K (it's None when split-K is disabled).
   SplitK mSplitK{SplitK::None};
   // K tile dimension of GEMM.
@@ -370,8 +390,9 @@ struct GemmOptions {
   bool mUsePerTokenSfA{false};
   // Apply per-token scales from B
   bool mUsePerTokenSfB{false};
-  // Reorder rows/cols in the A matrix for the better memory accesses in the M-major epilogue.
-  bool mUseShuffledMatrixA{false};
+  // Reorder rows/cols in the A matrix (when TransposeMmaOutput is true, otherwise B matrix) for the
+  // better memory accesses in the M-major epilogue.
+  bool mUseShuffledMatrix{false};
   // Use TMA to store the result.
   bool mUseTmaStore{true};
   // Use two different warps for A and B matrix load.
@@ -508,6 +529,7 @@ inline std::string dumpOptions(GemmOptions const& options, bool dumpRuntimeParam
   ss << "mNumEpilogueWarps=" << options.mNumEpilogueWarps << "," << std::endl;
   ss << "mNumRegsCastAWarps=" << options.mNumRegsCastAWarps << "," << std::endl;
   ss << "mNumRegsCopySfLdsSttm=" << options.mNumRegsCopySfLdsSttm << "," << std::endl;
+  ss << "mNumRegsCopySparsityInfo=" << options.mNumRegsCopySparsityInfo << "," << std::endl;
   ss << "mNumRegsPerThreadEpilogueWarp=" << options.mNumRegsPerThreadEpilogueWarp << ","
      << std::endl;
   ss << "mNumRegsPerThreadNonEpilogueWarp=" << options.mNumRegsPerThreadNonEpilogueWarp << ","
@@ -521,11 +543,9 @@ inline std::string dumpOptions(GemmOptions const& options, bool dumpRuntimeParam
   ss << "mNumStagesWorkId=" << options.mNumStagesWorkId << "," << std::endl;
   ss << "mOutputDebugTensors=" << options.mOutputDebugTensors << "," << std::endl;
   ss << "mPatchF2fp=" << options.mPatchF2fp << "," << std::endl;
-  if (options.mSfBlockSizeA.has_value()) {
-    ss << "mSfBlockSizeA=" << options.mSfBlockSizeA.value() << "," << std::endl;
-  } else {
-    ss << "mSfBlockSizeA=" << "std::nullopt" << ", " << std::endl;
-  }
+  ss << "mSfBlockSizeA=" << options.mSfBlockSizeA << "," << std::endl;
+  ss << "mSfBlockSizeB=" << options.mSfBlockSizeB << "," << std::endl;
+  ss << "mSfBlockSizeC=" << options.mSfBlockSizeC << "," << std::endl;
   ss << "mSfLayoutA=" << "trtllm::gen::SfLayout(" << static_cast<int32_t>(options.mSfLayoutA) << ")"
      << "," << std::endl;
   ss << "mSfLayoutB=" << "trtllm::gen::SfLayout(" << static_cast<int32_t>(options.mSfLayoutB) << ")"
@@ -534,6 +554,8 @@ inline std::string dumpOptions(GemmOptions const& options, bool dumpRuntimeParam
      << "," << std::endl;
   ss << "mSfReshapeFactor=" << options.mSfReshapeFactor << "," << std::endl;
   ss << "mSliceK=" << options.mSliceK << "," << std::endl;
+  ss << "mSparsityA=" << "trtllm::gen::Sparsity(" << static_cast<int32_t>(options.mSparsityA) << ")"
+     << "," << std::endl;
   ss << "mSplitK=" << "gemm::SplitK(" << static_cast<int32_t>(options.mSplitK) << ")" << ","
      << std::endl;
   ss << "mTileK=" << options.mTileK << "," << std::endl;
@@ -549,7 +571,7 @@ inline std::string dumpOptions(GemmOptions const& options, bool dumpRuntimeParam
   ss << "mUseMaxTmemOverlap=" << options.mUseMaxTmemOverlap << "," << std::endl;
   ss << "mUsePerTokenSfA=" << options.mUsePerTokenSfA << "," << std::endl;
   ss << "mUsePerTokenSfB=" << options.mUsePerTokenSfB << "," << std::endl;
-  ss << "mUseShuffledMatrixA=" << options.mUseShuffledMatrixA << "," << std::endl;
+  ss << "mUseShuffledMatrix=" << options.mUseShuffledMatrix << "," << std::endl;
   ss << "mUseTmaStore=" << options.mUseTmaStore << "," << std::endl;
   ss << "mUseTwoTmaLoadWarps=" << options.mUseTwoTmaLoadWarps << "," << std::endl;
   ss << "mUseTwoMmaWarps=" << options.mUseTwoMmaWarps << "," << std::endl;
@@ -579,6 +601,33 @@ inline T divUpMul(T a, T b) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+// clang-format off
+inline std::vector<int> srcToDstBlk16RowMap =
+      {
+        0,  8,
+        1,  9,
+        2, 10,
+        3, 11,
+        4, 12,
+        5, 13,
+        6, 14,
+        7, 15
+      };
+inline std::vector<int> srcToDstBlk32RowMap =
+      {
+        0,  8, 16, 24,
+        1,  9, 17, 25,
+        2, 10, 18, 26,
+        3, 11, 19, 27,
+        4, 12, 20, 28,
+        5, 13, 21, 29,
+        6, 14, 22, 30,
+        7, 15, 23, 31
+      };
+// clang-format on
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
 inline int32_t getShuffleBlockSize(int epilogueTileM) {
   int shuffleBlockSize = 16;
   if (epilogueTileM % 128 == 0) {
@@ -589,6 +638,13 @@ inline int32_t getShuffleBlockSize(int epilogueTileM) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+inline std::vector<int> const& getShuffleIndices(int epilogueTileM) {
+  auto const shuffleBlockSize = getShuffleBlockSize(epilogueTileM);
+  return shuffleBlockSize == 16 ? srcToDstBlk16RowMap : srcToDstBlk32RowMap;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
 // Check if the options are valid or not.
 inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArch, int tpGrpSize,
                                       bool updateOptions = true) {
@@ -596,6 +652,7 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
 
   bool isBlackwell = tg::isArchBlackwell(cudaArch);
 
+  // If dtypeB is unspecified (Dtype::Void), assign to dtypeA.
   if (options.mDtypeB == tg::Dtype::Void) {
     if (updateOptions) {
       options.mDtypeB = options.mDtypeA;
@@ -603,6 +660,15 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
       return false;
     }
   }
+  // If dtypeC is unspecified (Dtype::Void), assign to dtypeA.
+  if (options.mDtypeC == tg::Dtype::Void) {
+    TLLM_LOG_INFO("Setting dtypeC to ", tg::dtypeToString(options.mDtypeA));
+    if (updateOptions) {
+      options.mDtypeC = options.mDtypeA;
+    } else {
+      return false;
+    }
+  }
 
   // If not specified, used the input dtypes as MMA dtypes (no cast required).
   if (options.mDtypeMmaA == tg::Dtype::Void) {
@@ -751,13 +817,55 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
     }
   }
 
+  // Check that the sparsity mode of A is supported, and compatible with the MMA kind.
+  // Note: trtllm-gen currently does not support sparsity with tf32, fp16, bf16.
+  switch (options.mSparsityA) {
+    case tg::Sparsity::Dense:
+      // Always supported.
+      break;
+    case tg::Sparsity::Any_1_2:
+      TLLM_LOG_ERROR("1:2 sparsity is not supported.");
+      break;
+    case tg::Sparsity::Any_2_4: {
+      bool isSupported_2_4 = (options.mMmaKind == tg::MmaKind::Fp8Fp6Fp4 ||
+                              options.mMmaKind == tg::MmaKind::MxFp8Fp6Fp4);
+      TLLM_CHECK_ERROR(isSupported_2_4, "2:4 sparsity is not supported for MMA kind ",
+                       tg::mmaKindToString(options.mMmaKind), " on target ",
+                       tg::cudaArchToString(cudaArch));
+      break;
+    }
+    case tg::Sparsity::Pairwise_4_8:
+      TLLM_CHECK_ERROR(options.mMmaKind == tg::MmaKind::MxFp4NvFp4,
+                       "Pairwise 4:8 sparsity is only supported for MMA kind MxFp4NvFp4.");
+      break;
+    default:
+      TLLM_CHECK_ERROR(false, "Unsupported sparsityA: ", tg::sparsityToString(options.mSparsityA));
+      break;
+  }
+
+  // Is A sparse?
+  bool const isSparseA = tg::isSparse(options.mSparsityA);
+
+  // Requirements specific to sparsity, and compatibility with other features.
+  if (isSparseA) {
+    TLLM_CHECK_ERROR(isBlackwell, "Sparsity is only supported on Blackwell");
+    // The following requirement is for TMA load: the box width must be a multiple of 16B.
+    TLLM_CHECK_ERROR(
+        tg::getNumBytesSparsityInfo(options.mSparsityA, options.mTileK) % 16 == 0,
+        "The sparsity information for one tile row must be a multiple of 16B. Use larger tileK.");
+    TLLM_CHECK_ERROR(options.mDtypeA == options.mDtypeMmaA,
+                     "Sparsity is not supported with on-the-fly upcasting.");
+    TLLM_CHECK_ERROR(!options.mUseDeepSeekFp8, "Sparsity is not supported with DeepSeek Fp8.");
+    TLLM_CHECK_ERROR(!options.mSliceK, "Sparsity is not supported with slice-k.");
+  }
+
   if (options.mMmaKind == tg::MmaKind::Fp8Fp6Fp4) {
-    int mmaK = 32;
+    int mmaK = isSparseA ? 64 : 32;
 
     if (options.mMmaK != mmaK) {
-      TLLM_LOG_WARNING("Unsupported MmaK (", options.mMmaK,
-                       ") for MmaKind=", gemm::toString(options.mMmaKind), ". Setting MmaK to ",
-                       mmaK);
+      TLLM_LOG_WARNING(
+          "Unsupported MmaK (", options.mMmaK, ") for MmaKind=", gemm::toString(options.mMmaKind),
+          " and sparsity=", tg::sparsityToString(options.mSparsityA), ". Setting MmaK to ", mmaK);
       if (updateOptions) {
         options.mMmaK = mmaK;
         options.mTileK = std::max(options.mMmaK, options.mTileK);
@@ -819,10 +927,10 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
   if (options.mMmaKind == tg::MmaKind::MxFp4NvFp4 || options.mMmaKind == tg::MmaKind::MxFp8Fp6Fp4) {
     TLLM_CHECK_ERROR(isBlackwell, "Block scaling is only supported on Blackwell");
 
-    int mmaK = 32;
+    int mmaK = isSparseA ? 64 : 32;
     if (options.mMmaKind == tg::MmaKind::MxFp4NvFp4) {
-      mmaK = 64;
-      if (options.mMmaK == 96) {
+      mmaK = isSparseA ? 128 : 64;
+      if (options.mMmaK == 96 && !isSparseA) {
         mmaK = 96;
         TLLM_CHECK_ERROR(options.mTileK == 768, "When mmaK == 96, only tileK == 768 is supported");
         TLLM_CHECK_ERROR(options.mTileN <= 128, "When mmaK == 96, only tileN <= 128 is supported");
@@ -846,25 +954,89 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                      options.mMmaN, ") must be >= 64 or equal to TileN (", options.mTileN, ")");
   }
 
-  if (options.mSfBlockSizeA.has_value()) {
-    // Only E2m1 x E4m3 is tested. MxE2m1 x bf16 may also work.
-    TLLM_CHECK_ERROR(options.mDtypeA == tg::Dtype::E2m1 && options.mDtypeB == tg::Dtype::E4m3,
-                     "sfBlockSizeA is only supported for E2m1 and E4m3 types. Found dtypeA=",
-                     tg::dtypeToString(options.mDtypeA),
-                     " dtypeB=", tg::dtypeToString(options.mDtypeB));
+  // Note: the logic for selecting/checking the correct block size based on dtypes and sparsity is
+  // centralized here, to avoid error-prone code duplication and make it a more explicit "contract"
+  // with the user who is providing inputs in this format.
+  // Additionally, in some cases, multiple values are possible:
+  // - When we use type casting before the MMA (e.g. e2m1 x e4m3).
+  // - For output C, based on whether the consumer will use sparsity.
 
-    // sfBlockSizeA must be 16 or 32.
-    // SfBlockSizeA can also support 64 and 128, although they are not officially supported Nvida
-    // format. Note that the type conversion needs to happen before TCs.
+  // SF block size for A.
+  if (options.mDtypeA == tg::Dtype::E2m1 && options.mDtypeB == tg::Dtype::E4m3) {
+    // Note that the type conversion needs to happen before TCs.
     // For example, convert e2m1 to e4m3 inside TmemCastA.
-    // If we want to support sfBlockSizeA=8, we can write another version of convertE2m1ToSfE4m3,
-    // which only packs 8 e2m1 elements.
-    TLLM_CHECK_ERROR(options.mSfBlockSizeA.value() == 16 || options.mSfBlockSizeA.value() == 32,
-                     "SfBlockSizeA (", options.mSfBlockSizeA.value(), ") must be 16 or 32.");
+    if (!(options.mSfBlockSizeA == 16 || options.mSfBlockSizeA == 32)) {
+      TLLM_LOG_WARNING("sfBlockSizeA must be 16 or 32 for e2m1 x e4m3, got ",
+                       options.mSfBlockSizeA);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeA, 16);
+    }
+  } else if (options.mDtypeA == tg::Dtype::E2m1) {
+    if (!((options.mSfBlockSizeA == 16 && !isSparseA) ||
+          (options.mSfBlockSizeA == 32 && isSparseA))) {
+      TLLM_LOG_WARNING("sfBlockSizeA must be 16 (dense) or 32 (sparse) for dtypeA=e2m1, got ",
+                       options.mSfBlockSizeA);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeA, isSparseA ? 32 : 16);
+    }
+  } else if (options.mDtypeA == tg::Dtype::MxE2m1 || options.mDtypeA == tg::Dtype::MxE4m3 ||
+             options.mDtypeA == tg::Dtype::MxInt4) {
+    if (!((options.mSfBlockSizeA == 32 && !isSparseA) ||
+          (options.mSfBlockSizeA == 64 && isSparseA))) {
+      TLLM_LOG_WARNING(
+          "sfBlockSizeA must be 32 (dense) or 64 (sparse) for dtypeA=mx{e2m1,e4m3,int4}, got ",
+          options.mSfBlockSizeA);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeA, isSparseA ? 64 : 32);
+    }
+  } else if (options.mSfBlockSizeA > 0) {
+    TLLM_LOG_WARNING("Got sfBlockSizeA=", options.mSfBlockSizeA,
+                     " but dtypeA=", tg::dtypeToString(options.mDtypeA),
+                     " does not use block scales");
+    GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeA, -1);
+  }
+  // SF block size for B.
+  if (options.mDtypeB == tg::Dtype::E2m1) {
+    if (!((options.mSfBlockSizeB == 16 && !isSparseA) ||
+          (options.mSfBlockSizeB == 32 && isSparseA))) {
+      TLLM_LOG_WARNING("sfBlockSizeB must be 16 (dense) or 32 (sparse) for dtypeB=e2m1, got ",
+                       options.mSfBlockSizeB);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeB, isSparseA ? 32 : 16);
+    }
+  } else if (options.mDtypeB == tg::Dtype::MxE2m1 || options.mDtypeB == tg::Dtype::MxE4m3 ||
+             (options.mDtypeB == tg::Dtype::E4m3 && options.mDtypeMmaB == tg::Dtype::MxE4m3)) {
+    if (!((options.mSfBlockSizeB == 32 && !isSparseA) ||
+          (options.mSfBlockSizeB == 64 && isSparseA))) {
+      TLLM_LOG_WARNING(
+          "sfBlockSizeB must be 32 (dense) or 64 (sparse) for dtypeB=mx{e2m1,e4m3}, got ",
+          options.mSfBlockSizeB);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeB, isSparseA ? 64 : 32);
+    }
+  } else if (options.mSfBlockSizeB > 0) {
+    TLLM_LOG_WARNING("Got sfBlockSizeB=", options.mSfBlockSizeB,
+                     " but dtypeB=", tg::dtypeToString(options.mDtypeB),
+                     " does not use block scales");
+    GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeB, -1);
+  }
+  // SF block size for C.
+  if (options.mDtypeC == tg::Dtype::E2m1) {
+    if (!(options.mSfBlockSizeC == 16 || options.mSfBlockSizeC == 32)) {
+      TLLM_LOG_WARNING("sfBlockSizeC must be 16 or 32 for dtypeC=e2m1, got ",
+                       options.mSfBlockSizeC);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeC, 16);
+    }
+  } else if (options.mDtypeC == tg::Dtype::MxE2m1 || options.mDtypeC == tg::Dtype::MxE4m3) {
+    if (!(options.mSfBlockSizeC == 32 || options.mSfBlockSizeC == 64)) {
+      TLLM_LOG_WARNING("sfBlockSizeC must be 32 or 64 for dtypeC=mx{e2m1,e4m3}, got ",
+                       options.mSfBlockSizeC);
+      GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeC, 32);
+    }
+  } else if (options.mSfBlockSizeC > 0) {
+    TLLM_LOG_WARNING("Got sfBlockSizeC=", options.mSfBlockSizeC,
+                     " but dtypeC=", tg::dtypeToString(options.mDtypeC),
+                     " does not use block scales");
+    GEMM_UPDATE_OR_ERROR(options.mSfBlockSizeC, -1);
   }
 
   if (tg::dtypeIsBlockFmt(options.mDtypeA)) {
-    int numEltsPerSfA = options.mSfBlockSizeA.value_or(tg::dtypeNumEltsPerSf(options.mDtypeA));
+    int numEltsPerSfA = options.mSfBlockSizeA;
     TLLM_CHECK_ERROR(options.mTileK % (4 * numEltsPerSfA) == 0, "TileK (", options.mTileK,
                      ") must be a multiple of ", (4 * numEltsPerSfA), " for typeA ",
                      gemm::toString(options.mDtypeA));
@@ -885,7 +1057,7 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                      ") must be a multiple of ", numSfTileRowsB, " for B SF layout ",
                      tg::sfLayoutToString(options.mSfLayoutB));
 
-    int numEltsPerSfB = tg::dtypeNumEltsPerSf(options.mDtypeB);
+    int numEltsPerSfB = options.mSfBlockSizeB;
     TLLM_CHECK_ERROR(options.mTileK % (4 * numEltsPerSfB) == 0, "TileK (", options.mTileK,
                      ") must be a multiple of ", (4 * numEltsPerSfB), " for typeB ",
                      gemm::toString(options.mDtypeB));
@@ -917,6 +1089,10 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
     TLLM_CHECK_ERROR(
         options.mSfLayoutC == tg::SfLayout::R128c4 || options.mSfLayoutC == tg::SfLayout::R8c4,
         "Only the 128x4 and 8x4 SF layouts are supported for C.");
+    if (!options.mTransposeMmaOutput) {
+      TLLM_CHECK_ERROR(options.mEpilogueTileN % options.mSfBlockSizeC == 0,
+                       "EpilogueTileN must be a multiple of the number of elements per SF for C");
+    }
     int const numSfTileRowsC = options.mSfLayoutC == tg::SfLayout::R128c4 ? 128 : 8;
     int const tileTokenDim = options.mTransposeMmaOutput ? options.mTileN : options.mTileM;
     TLLM_CHECK_ERROR_FMT(tileTokenDim % numSfTileRowsC == 0,
@@ -924,26 +1100,16 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                          options.mTransposeMmaOutput ? "N" : "M", tileTokenDim, numSfTileRowsC,
                          tg::sfLayoutToString(options.mSfLayoutC).c_str());
 
+    int numEltsPerSfC = options.mSfBlockSizeC;
     int const hiddenDim = options.mTransposeMmaOutput ? options.mM : options.mN;
-    int const hiddenGranularity = 4 * tg::dtypeNumEltsPerSf(options.mDtypeC);
+    int const hiddenGranularity = 4 * numEltsPerSfC;
     TLLM_CHECK_ERROR(hiddenDim % hiddenGranularity == 0, "Hidden dim (", hiddenDim,
                      ") must be a multiple of ", hiddenGranularity, " for block-scaled outputs.");
     int const validHiddenDim = options.mTransposeMmaOutput ? options.mValidM : options.mValidN;
-    TLLM_CHECK_ERROR(validHiddenDim % tg::dtypeNumEltsPerSf(options.mDtypeC) == 0,
-                     "Valid hidden dim (", validHiddenDim, ") must be a multiple of ",
-                     tg::dtypeNumEltsPerSf(options.mDtypeC), " for block-scaled outputs.");
-    TLLM_CHECK_ERROR(!options.mTransposeMmaOutput || options.mUseShuffledMatrixA,
-                     "Transposing block-scaled outputs requires shuffled A.");
-  }
-
-  // If dtypeC is unspecified (Dtype::Void), assign to the input dtype.
-  if (options.mDtypeC == tg::Dtype::Void) {
-    TLLM_LOG_INFO("Setting dtypeC to ", tg::dtypeToString(options.mDtypeA));
-    if (updateOptions) {
-      options.mDtypeC = options.mDtypeA;
-    } else {
-      return false;
-    }
+    TLLM_CHECK_ERROR(validHiddenDim % numEltsPerSfC == 0, "Valid hidden dim (", validHiddenDim,
+                     ") must be a multiple of ", numEltsPerSfC, " for block-scaled outputs.");
+    TLLM_CHECK_ERROR(!options.mTransposeMmaOutput || options.mUseShuffledMatrix,
+                     "Transposing block-scaled outputs requires shuffled matrix.");
   }
 
   // Set epilogue tile sizes to the output tile sizes, when epilogue tile sizes are incorrect.
@@ -995,11 +1161,11 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                    "M, N and K must be larger than 0");
   TLLM_CHECK_ERROR(options.mNumSlicesForSplitK > 0, "Split K must be larger than 0.");
 
-  if (options.mUseShuffledMatrixA) {
+  if (options.mUseShuffledMatrix) {
     auto const shuffleBlockSize = getShuffleBlockSize(options.mEpilogueTileM);
     TLLM_CHECK_ERROR(options.mM % shuffleBlockSize == 0 && options.mValidM % shuffleBlockSize == 0,
                      "M/validM must be a multiple of shuffle block size (", shuffleBlockSize,
-                     ") when useShuffledMatrixA");
+                     ") when useShuffledMatrix");
   }
 
   if (!options.mSliceK) {
@@ -1033,14 +1199,6 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                      "K must be a multiple of TileK * numSlicesForSplitK for DeepSeekFp8");
   }
 
-  // When the A-matrix is shuffled, the output must be transposed.
-  if (options.mUseShuffledMatrixA) {
-    // TODO add matrix shuffle for N-major epilogue.
-    TLLM_CHECK_ERROR(
-        options.mTransposeMmaOutput,
-        "Shuffled matrix A is only supported with M-major epilogue. Set -transposeMmaOutput");
-  }
-
   // Check all-reduce options.
   if (options.mAllReduceAlgo == AllReduceAlgo::OneShot) {
     // One shot is implemented with PTX cp.reduce.async.bulk.tensor which supports only the
@@ -1101,6 +1259,31 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
                      "CGA size must be equal to the number of slices in split-k");
   }
 
+  if (options.mUseShuffledMatrix && !options.mTransposeMmaOutput) {
+    TLLM_CHECK_ERROR(
+        !options.mUseDeepSeekFp8,
+        "DeepSeek Fp8 is not supported when using shuffled matrix and non-transposed mma output");
+    TLLM_CHECK_ERROR(
+        options.mEpilogueLdtmBits == 32,
+        "EpilogueLdtmBits must be 32 when using shuffled matrix and non-transposed mma output");
+    TLLM_CHECK_ERROR(
+        options.mEpilogueLdtmDps == 32,
+        "EpilogueLdtmDps must be 32 when using shuffled matrix and non-transposed mma output");
+    TLLM_CHECK_ERROR(
+        options.mUseTmaStore,
+        "TMA store is required when using shuffled matrix and non-transposed mma output");
+    TLLM_CHECK_ERROR(
+        !options.mSliceK,
+        "Slice-K is not supported when using shuffled matrix and non-transposed mma output");
+    // When doing unshuffle in the epilogue, one fragment of epilogue tile must have at least one
+    // shuffle block.
+    auto minEpilogueTileN = getShuffleBlockSize(options.mEpilogueTileM);
+    TLLM_CHECK_ERROR_FMT(options.mEpilogueTileN >= minEpilogueTileN,
+                         "EpilogueTileN (%d) must be a larger than the shuffle block size (%d) "
+                         "when using shuffled matrix and non-transposed mma output",
+                         options.mEpilogueTileN, minEpilogueTileN);
+  }
+
   // Maps numStagesMma to (stagesWithinWorkTile, stagesAcrossWorkTile) if not already set.
   // If (-1, -1) -> (numStagesMma / min(2, numStagesMma), min(2, numStagesMma))
   // If ( m, -1) -> (m, numStagesMma / m)
@@ -1205,7 +1388,7 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
     TLLM_CHECK_ERROR(!options.mUseDeepSeekFp8, "DeepSeek Fp8 GEMM is not supported for slice-K");
     TLLM_CHECK_ERROR(options.mUseTwoTmaLoadWarps, "Slice-K requires two warp load for A and B");
     TLLM_CHECK_ERROR(options.mTransposeMmaOutput, "Slice-K requires transpose mma output");
-    TLLM_CHECK_ERROR(options.mUseShuffledMatrixA, "Slice-K requires shuffled matrix A");
+    TLLM_CHECK_ERROR(options.mUseShuffledMatrix, "Slice-K requires shuffled matrix");
     TLLM_CHECK_ERROR(options.mTileK % 128 == 0, "Slice-K requires TileK be a multiple of 128");
     TLLM_CHECK_ERROR(options.mMmaM == 128, "Slice-K requires MmaM == 128");
     TLLM_CHECK_ERROR(options.mTileN == options.mEpilogueTileN,
@@ -1368,11 +1551,11 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
   }
   if (options.mLayoutA == MatrixLayout::MajorMn) {
     TLLM_CHECK_ERROR(tg::dtypeGetNumBits(options.mDtypeA) >= 8,
-                     "Subbyte types only support K major layout");
+                     "Subbyte types do not support m-major layout");
   }
   if (options.mLayoutB == MatrixLayout::MajorMn) {
     TLLM_CHECK_ERROR(tg::dtypeGetNumBits(options.mDtypeB) >= 8,
-                     "Subbyte types only support K major layout");
+                     "Subbyte types do not support n-major layout");
   }
 
   if ((options.mLayoutA == MatrixLayout::BlockMajorK) ||
@@ -1445,10 +1628,11 @@ inline bool checkAndUpdateGemmOptions(GemmOptions& options, tg::CudaArch cudaArc
     // Init kernel traits.
     options.mKernelTraits = KernelTraits(
         options.mDtypeA, options.mDtypeB, options.mDtypeC, options.mDtypeAcc, options.mDtypeMmaA,
-        options.mDtypeMmaB, options.mMmaKind, options.mMmaK, options.mTileM, options.mTileN,
-        options.mTileK, options.mEpilogueTileM, options.mEpilogueTileN, options.mNumStages,
-        options.mNumStagesMma, options.mNumSlicesForSplitK, options.mNumSlicesForSliceK,
-        options.mSplitK, options.mUseTmaStore, options.mTransposeMmaOutput, options.mAllReduceAlgo,
+        options.mDtypeMmaB, options.mMmaKind, options.mSparsityA, options.mMmaK, options.mTileM,
+        options.mTileN, options.mTileK, options.mEpilogueTileM, options.mEpilogueTileN,
+        options.mSfBlockSizeA, options.mSfBlockSizeB, options.mNumStages, options.mNumStagesMma,
+        options.mNumSlicesForSplitK, options.mNumSlicesForSliceK, options.mSplitK,
+        options.mUseTmaStore, options.mTransposeMmaOutput, options.mAllReduceAlgo,
         options.mFuseUtccpWithUtcmma, options.mUseMaxTmemOverlap, options.mNumEpilogueWarps,
         options.mTileScheduler == TileScheduler::Persistent, options.mUseDeepSeekFp8,
         options.mUsePerTokenSfA, options.mUsePerTokenSfB,
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParams.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParams.h
index ab004a9691..bda17264fb 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParams.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParams.h
@@ -17,12 +17,14 @@
 #pragma once
 
 #include <stdexcept>
+#include <tuple>
 
 #include "BatchedGemmEnums.h"
 #include "Enums.h"
 #include "TmaDescriptor.h"
 #include "trtllm/gen/CommonUtils.h"
 #include "trtllm/gen/SfLayoutDecl.h"
+#include "trtllm/gen/SparsityDecl.h"
 
 // NOTE: keep this code dependency free. It has to be included by the device code and has to be
 // compilable with NVRTC.
@@ -295,13 +297,41 @@ static auto makeTmaShapeStrideSfAb(int mM, int mN, int mK, MatrixType matrixType
   }
   return std::make_tuple(std::vector<uint64_t>{}, std::vector<uint64_t>{}, std::vector<uint32_t>{});
 }
+
+// Create the TMA shape/stride for sparsity information of A.
+template <class GemmOptions>
+static auto makeTmaShapeStrideSparsityInfoA(GemmOptions const& options) {
+  // Tensor dimensions.
+  auto outerDim = options.mM;
+  auto innerDim = tg::getNumBytesSparsityInfo(options.mSparsityA, options.mK);
+  // Tile dimensions.
+  auto tileOuterDim = options.mTileM;
+  auto tileInnerDim = tg::getNumBytesSparsityInfo(options.mSparsityA, options.mTileK);
+
+  auto shape =
+      std::vector<uint64_t>{static_cast<uint64_t>(innerDim), static_cast<uint64_t>(outerDim),
+                            static_cast<uint64_t>(options.mNumBatches)};
+
+  std::vector<uint64_t> stride(shape.size());
+  stride[0] = 1;
+  for (size_t i = 1; i < shape.size(); i++) {
+    stride[i] = shape[i - 1] * stride[i - 1];
+  }
+
+  auto tileShapes =
+      std::vector<int32_t>{static_cast<int32_t>(tileInnerDim), static_cast<int32_t>(tileOuterDim)};
+
+  return std::make_tuple(shape, stride, tileShapes);
+}
+
 template <class GemmOptions_>
 static KernelParams setKernelParams(
     GemmOptions_ const& options, bool const batchM, void const* ptrA, void const* ptrB, void* ptrC,
     void const* dSfA, void const* dSfB, void const* ptrPerTokenSfA, void const* ptrPerTokenSfB,
-    void const* ptrBias, void* dSfC, float const* ptrScaleC, float const* ptrScaleAct,
-    float const* ptrScaleGate, float const* ptrClampLimit, float const* ptrGatedActAlpha,
-    float const* ptrGatedActBeta, int32_t const* routeMap, float* rowMax, uint32_t* rowMaxBars,
+    [[maybe_unused]] void const* ptrSparsityInfoA, void const* ptrBias, void* dSfC,
+    float const* ptrScaleC, float const* ptrScaleAct, float const* ptrScaleGate,
+    float const* ptrClampLimit, float const* ptrGatedActAlpha, float const* ptrGatedActBeta,
+    int32_t const* routeMap, float* rowMax, uint32_t* rowMaxBars,
     int32_t const* ptrNumNonExitingCtas = nullptr, int32_t const* ptrTotalNumPaddedTokens = nullptr,
     int32_t const* ptrCtaIdxXyToBatchIdx = nullptr, int32_t const* ptrCtaIdxXyToMnLimit = nullptr,
     int32_t const maxNumCtas = KernelParams::MaxNumCtas) {
@@ -323,6 +353,8 @@ static KernelParams setKernelParams(
 
   int32_t ctaOffset = 0;
 
+  int32_t const isSparseA = tg::isSparse(options.mSparsityA);
+
   // Compute totalNumPaddedTokens, ctaIdxXyToBatchIdx and ctaIdxXyToMnLimit if the batch dims are
   // known at kernel launch time. Otherwise, these parameters are defined in the device buffers:
   // ptrTotalNumPaddedTokens, ptrCtaIdxXyToBatchIdx and ptrCtaIdxXyToMnLimit respectively.
@@ -339,10 +371,11 @@ static KernelParams setKernelParams(
       }
 
       // The number of CTAs.
-      int32_t numCtas = batchM ? (mM + options.mTileM - 1) / options.mTileM
+      int32_t numCtas = batchM ? (mM + options.mTileM * options.mClusterDimX - 1) /
+                                     (options.mTileM * options.mClusterDimX) * options.mClusterDimX
                                : (mN + options.mTileN - 1) / options.mTileN;
       // The size of the tile.
-      int32_t tile = batchM ? options.mTileM : options.mTileN;
+      int32_t tile = batchM ? options.mTileM * options.mClusterDimX : options.mTileN;
       // The problem size.
       int32_t mn = batchM ? mM : mN;
       int32_t tokensPerTile = mn;
@@ -366,7 +399,7 @@ static KernelParams setKernelParams(
     params.totalNumOutputPaddedTokens = params.totalNumPaddedTokens;
   } else if (options.mIsStaticBatch && options.mIsUniformNumTokensPerBatch) {
     auto numTokens = batchM ? options.mBatchedM[0] : options.mBatchedN[0];
-    auto tileTokensDim = batchM ? options.mTileM : options.mTileN;
+    auto tileTokensDim = batchM ? options.mTileM * options.mClusterDimX : options.mTileN;
     params.batchStrideInCtas = (options.mBatchStrideInTokens + tileTokensDim - 1) / tileTokensDim;
     params.ctasInTokenDimPerBatch = (numTokens + tileTokensDim - 1) / tileTokensDim;
     params.totalNumOutputPaddedTokens =
@@ -398,13 +431,9 @@ static KernelParams setKernelParams(
   params.ptrSfB = dSfB;
   params.ptrSfC = dSfC;
 
-  // Do we pad A?
-  bool doPadA =
-      options.mMmaKind == tg::MmaKind::MxFp8Fp6Fp4 && options.mDtypeA == tg::Dtype::MxE2m1;
-
-  // Do we pad B?
-  bool doPadB =
-      options.mMmaKind == tg::MmaKind::MxFp8Fp6Fp4 && options.mDtypeB == tg::Dtype::MxE2m1;
+  // Do we pad A or B?
+  bool doPadA = tg::dtypeNeedsPadding(options.mDtypeA, options.mMmaKind, options.mMmaK, isSparseA);
+  bool doPadB = tg::dtypeNeedsPadding(options.mDtypeB, options.mMmaKind, options.mMmaK, isSparseA);
 
   if (!batchM) {
     // A is the expert
@@ -415,8 +444,9 @@ static KernelParams setKernelParams(
     params.nm = options.mM;
     // Shape/stride for gmem tensor A.
     auto [shapeA, strideA, tileShapeA] = makeTmaShapeStrideAbc(
-        options, options.mM, options.mN, options.mK, options.mTileM, options.mTileN, options.mTileK,
-        MatrixType::MatrixA, options.mValidM, options.mValidN, options.mValidK);
+        options, options.mM, options.mN, options.mK >> isSparseA, options.mTileM, options.mTileN,
+        options.mTileK >> isSparseA, MatrixType::MatrixA, options.mValidM, options.mValidN,
+        options.mValidK >> isSparseA);
     // Build tma descriptor for A.
     params.tmaA[0] = gemm::buildNdTmaDescriptor(options.mDtypeA, shapeA, strideA, tileShapeA,
                                                 const_cast<void*>(ptrA), doPadA,
@@ -451,39 +481,51 @@ static KernelParams setKernelParams(
         dTypeSfA = tg::Dtype::UE8m0;
       }
 
+      int32_t const numEltsPerSfA = options.mSfBlockSizeA;
+
       // Build TMA descriptor for gmem A block scaling factors.
       auto [shapeSfA, strideSfA, tileShapesSfA] = makeTmaShapeStrideSfAb(
           options.mM * options.mNumBatches, options.mN, options.mK, MatrixType::MatrixA,
           options.mTileM, options.mTileN, options.mTileK, tg::SfLayout::R128c4,
-          options.mSfReshapeFactor,
-          options.mSfBlockSizeA.value_or(tg::dtypeNumEltsPerSf(options.mDtypeA)));
+          options.mSfReshapeFactor, numEltsPerSfA);
       params.tmaSfA[0] = gemm::buildSfTmaDescriptor(dTypeSfA, shapeSfA, strideSfA, tileShapesSfA,
                                                     const_cast<void*>(dSfA));
     }
 
+    if (isSparseA) {
+      // Build TMA descriptor for gmem A sparsity information.
+      auto [shapeSparsityInfoA, strideSparsityInfoA, tileShapesSparsityInfoA] =
+          makeTmaShapeStrideSparsityInfoA(options);
+      params.tmaSparsityInfoA =
+          gemm::buildNdTmaDescriptor(tg::Dtype::UInt8, shapeSparsityInfoA, strideSparsityInfoA,
+                                     tileShapesSparsityInfoA, const_cast<void*>(ptrSparsityInfoA),
+                                     /*doPad=*/false,
+                                     /*doSwizzle=*/true);
+    }
+
     if (options.mDtypeB == tg::Dtype::E2m1 || options.mDtypeB == tg::Dtype::MxE4m3 ||
         options.mDtypeB == tg::Dtype::MxE2m1) {
       tg::Dtype const dTypeSf =
           (options.mDtypeB == tg::Dtype::E2m1) ? tg::Dtype::E4m3 : tg::Dtype::UE8m0;
+      int32_t const numEltsPerSfB = options.mSfBlockSizeB;
 
       if (batchedGemm::doesRouteImplUseTma(options.mRouteSfsImpl.value())) {
         // The input is NOT padded:
         // [act0, act1, act2, ...]
 
         // Build TMA descriptor for gmem B block scaling factors.
-        int32_t const numEltsPerSf = tg::dtypeNumEltsPerSf(options.mDtypeB);
         // Pad number of scaling factors to the nearest multiple of 16 because of the TMA 16B
         // alignment requirement.
-        auto numSfsInK = options.mK / numEltsPerSf;
+        auto numSfsInK = options.mK / numEltsPerSfB;
         numSfsInK = ceilDiv(numSfsInK, 16) * 16;
 
-        auto numSfsInValidK = options.mValidK / numEltsPerSf;
+        auto numSfsInValidK = options.mValidK / numEltsPerSfB;
         numSfsInValidK = ceilDiv(numSfsInValidK, 16) * 16;
 
         auto [shapeSfB, strideSfB, tileShapesSfB] = makeTmaShapeStrideAbc(
             options, options.mM, options.mNumTokens, numSfsInK, options.mTileM, 1 /* tileN */,
-            options.mTileK / numEltsPerSf, MatrixType::MatrixB, options.mValidM, options.mNumTokens,
-            numSfsInValidK);
+            options.mTileK / numEltsPerSfB, MatrixType::MatrixB, options.mValidM,
+            options.mNumTokens, numSfsInValidK);
         params.tmaSfB[0] = gemm::buildNdTmaDescriptor(dTypeSf, shapeSfB, strideSfB, tileShapesSfB,
                                                       const_cast<void*>(dSfB),
                                                       /*doPad=*/false,
@@ -495,10 +537,10 @@ static KernelParams setKernelParams(
         auto const inputNumTokensSfB = ctaOffset * options.mTileN;
 
         // Build TMA descriptor for gmem B block scaling factors.
-        auto [shapeSfB, strideSfB, tileShapesSfB] = makeTmaShapeStrideSfAb(
-            options.mM, inputNumTokensSfB, options.mK, MatrixType::MatrixB, options.mTileM,
-            options.mTileN, options.mTileK, options.mSfLayoutB, options.mSfReshapeFactor,
-            tg::dtypeNumEltsPerSf(options.mDtypeB));
+        auto [shapeSfB, strideSfB, tileShapesSfB] =
+            makeTmaShapeStrideSfAb(options.mM, inputNumTokensSfB, options.mK, MatrixType::MatrixB,
+                                   options.mTileM, options.mTileN, options.mTileK,
+                                   options.mSfLayoutB, options.mSfReshapeFactor, numEltsPerSfB);
         params.tmaSfB[0] = gemm::buildSfTmaDescriptor(dTypeSf, shapeSfB, strideSfB, tileShapesSfB,
                                                       const_cast<void*>(dSfB));
       }
@@ -538,15 +580,17 @@ static KernelParams setKernelParams(
                                                 const_cast<void*>(ptrB), doPadB,
                                                 /*doSwizzle=*/true);
 
-    if (options.mRouteImpl == batchedGemm::RouteImpl::NoRoute) {
+    if (!batchedGemm::doesRouteImplUseLdgsts(options.mRouteImpl)) {
       // A is the activation
       // Shape/stride for gmem tensor A.
       // The input is padded:
       // [act0, padding, padding, ... tileM size .., act1, padding, padding, ...]
       auto const inputNumTokens = ctaOffset * options.mTileM;
+      bool useRouteAct = batchedGemm::doesRouteImplUseTma(options.mRouteImpl);
       auto [shapeA, strideA, tileShapeA] = makeTmaShapeStrideAbc(
-          options, inputNumTokens, options.mN, options.mK, options.mTileM, options.mTileN,
-          options.mTileK, MatrixType::MatrixA, inputNumTokens, options.mValidN, options.mValidK);
+          options, useRouteAct ? options.mNumTokens : inputNumTokens, options.mN, options.mK,
+          useRouteAct ? 1 : options.mTileM, options.mTileN, options.mTileK, MatrixType::MatrixA,
+          useRouteAct ? options.mNumTokens : inputNumTokens, options.mValidN, options.mValidK);
       // Build tma descriptor for A.
       params.tmaA[0] = gemm::buildNdTmaDescriptor(options.mDtypeA, shapeA, strideA, tileShapeA,
                                                   const_cast<void*>(ptrA), doPadA,
@@ -557,6 +601,7 @@ static KernelParams setKernelParams(
         options.mDtypeA == tg::Dtype::MxE2m1) {
       tg::Dtype const dTypeSf =
           (options.mDtypeA == tg::Dtype::E2m1) ? tg::Dtype::E4m3 : tg::Dtype::UE8m0;
+      int32_t const numEltsPerSfA = options.mSfBlockSizeA;
 
       if (options.mRouteSfsImpl.value() == batchedGemm::RouteImpl::NoRoute) {
         // The input is padded:
@@ -564,10 +609,10 @@ static KernelParams setKernelParams(
         auto const inputNumTokensSfA = ctaOffset * options.mTileM;
 
         // Build TMA descriptor for gmem A block scaling factors.
-        auto [shapeSfA, strideSfA, tileShapesSfA] = makeTmaShapeStrideSfAb(
-            inputNumTokensSfA, options.mN, options.mK, MatrixType::MatrixA, options.mTileM,
-            options.mTileN, options.mTileK, tg::SfLayout::R128c4, options.mSfReshapeFactor,
-            options.mSfBlockSizeA.value_or(tg::dtypeNumEltsPerSf(options.mDtypeA)));
+        auto [shapeSfA, strideSfA, tileShapesSfA] =
+            makeTmaShapeStrideSfAb(inputNumTokensSfA, options.mN, options.mK, MatrixType::MatrixA,
+                                   options.mTileM, options.mTileN, options.mTileK,
+                                   tg::SfLayout::R128c4, options.mSfReshapeFactor, numEltsPerSfA);
         params.tmaSfA[0] = gemm::buildSfTmaDescriptor(dTypeSf, shapeSfA, strideSfA, tileShapesSfA,
                                                       const_cast<void*>(dSfA));
       }
@@ -577,12 +622,13 @@ static KernelParams setKernelParams(
         options.mDtypeB == tg::Dtype::MxE2m1) {
       tg::Dtype const dTypeSf =
           (options.mDtypeB == tg::Dtype::E2m1) ? tg::Dtype::E4m3 : tg::Dtype::UE8m0;
+      int32_t const numEltsPerSfB = options.mSfBlockSizeB;
 
       // Build TMA descriptor for gmem B block scaling factors.
       auto [shapeSfB, strideSfB, tileShapesSfB] = makeTmaShapeStrideSfAb(
           options.mM, options.mN * options.mNumBatches, options.mK, MatrixType::MatrixB,
           options.mTileM, options.mTileN, options.mTileK, options.mSfLayoutB,
-          options.mSfReshapeFactor, tg::dtypeNumEltsPerSf(options.mDtypeB));
+          options.mSfReshapeFactor, numEltsPerSfB);
       params.tmaSfB[0] = gemm::buildSfTmaDescriptor(dTypeSf, shapeSfB, strideSfB, tileShapesSfB,
                                                     const_cast<void*>(dSfB));
     }
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParamsDecl.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParamsDecl.h
index b1e5dba024..24c98a58f9 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParamsDecl.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelParamsDecl.h
@@ -166,6 +166,24 @@ struct KernelParams {
   // Dtype is Dtype::E4m3 for NvFp4, Dtype::UE8m0 for Mx formats.
   CUtensorMap tmaSfB[1];
 
+  // TMA descriptor for the sparsity information of A, if structured sparsity is used.
+  // Must be setup using gemm::buildNdTmaDescriptor with shapes and strides from
+  // makeTmaShapeStrideSparsityInfoA.
+  //
+  // When sparsityA is Any_2_4:
+  //     2 elements are non-zero in any chunk of 4 elements.
+  //     A 4-bit index indicates the position of the non-zero elements.
+  //     The shape in UInt8 is: [B, M, K / 8]
+  //
+  // When sparsityA is Pairwise_4_8:
+  //     4 elements are non-zero in any chunk of 8 elements.
+  //     The zero and non-zero elements are grouped in pairs.
+  //     A 4-bit index indicates the position of the non-zero pairs.
+  //     The shape in UInt8 is: [B, M, K / 16]
+  //
+  // Dtype is Dtype::UInt8.
+  CUtensorMap tmaSparsityInfoA;
+
   // The input matrix A.
   // If (routeAct == true && batchM), the shape is [M, K]. tmaA is not used.
   // Otherwise, check layout of tmaA to see the shape and strides.
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelTraits.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelTraits.h
index 024be709e1..db7bf0e331 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelTraits.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/KernelTraits.h
@@ -24,6 +24,7 @@
 #include "trtllm/gen/CommonUtils.h"
 #include "trtllm/gen/DtypeDecl.h"
 #include "trtllm/gen/MmaDecl.h"
+#include "trtllm/gen/SparsityDecl.h"
 
 namespace batchedGemm {
 
@@ -138,7 +139,7 @@ class MemAllocatorHelper {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-inline int getNumSmemBitsPerElt(tg::Dtype dtype, tg::MmaKind mmaKind, int mmaK) {
+inline int getNumSmemBitsPerElt(tg::Dtype dtype, tg::MmaKind mmaKind, int mmaK, bool isSparseA) {
   if (mmaKind == tg::MmaKind::Auto) {
     throw std::runtime_error("mmaKind != tg::MmaKind::Auto");
   }
@@ -158,9 +159,10 @@ class KernelTraits {
 
   // The constructor.
   KernelTraits(tg::Dtype dtypeA, tg::Dtype dtypeB, tg::Dtype dtypeC, tg::Dtype dtypeAcc,
-               tg::Dtype dtypeMmaA, tg::Dtype dtypeMmaB, tg::MmaKind mmaKind, int32_t mmaK,
-               int32_t tileM, int32_t tileN, int32_t tileK, int32_t epilogueTileM,
-               int32_t epilogueTileN, int32_t numStages, int32_t numStagesMma,
+               tg::Dtype dtypeMmaA, tg::Dtype dtypeMmaB, tg::MmaKind mmaKind,
+               tg::Sparsity sparsityA, int32_t mmaK, int32_t tileM, int32_t tileN, int32_t tileK,
+               int32_t epilogueTileM, int32_t epilogueTileN, int32_t numEltsPerSfA,
+               int32_t numEltsPerSfB, int32_t numStages, int32_t numStagesMma,
                int32_t numSlicesForSplitK, int32_t numSlicesForSliceK, SplitK splitK,
                bool useTmaStore, bool transposeMmaOutput, AllReduceAlgo allReduceAlgo,
                bool fuseUtccpWithUtcmma, bool useMaxTmemOverlap, int32_t numEpilogueWarps,
@@ -199,11 +201,15 @@ class KernelTraits {
       // Buffer names for inspection purposes.
       std::vector<std::string> smemChunkNames;
 
+      int const isSparseA = static_cast<int>(tg::isSparse(sparsityA));
+
       // LoadA
       {
         // Number of bytes in load A shared memory.
-        auto const numSmemBytesLoadA =
-            numStages * tileM * tileK * getNumSmemBitsPerElt(dtypeA, mMmaKind, mmaK) / 8 /* bits */;
+        // If A is sparse, we load only the non-zero elements.
+        auto const numSmemBytesLoadA = numStages * tileM * (tileK >> isSparseA) *
+                                       getNumSmemBitsPerElt(dtypeA, mMmaKind, mmaK, isSparseA) /
+                                       8 /* bits */;
         // Number of bytes for load A alignment for TMA load.
         auto const numBytesAlignmentLoadA = 1024;
         // loadA is already at first chunk. No need to reuse it.
@@ -219,7 +225,8 @@ class KernelTraits {
       {
         // Number of bytes in load B shared memory.
         auto const numSmemBytesLoadB = numStages * (useTwoCtas ? tileN / 2 : tileN) * tileK *
-                                       getNumSmemBitsPerElt(dtypeB, mMmaKind, mmaK) / 8 /* bits */;
+                                       getNumSmemBitsPerElt(dtypeB, mMmaKind, mmaK, isSparseA) /
+                                       8 /* bits */;
         // Number of bytes for load B alignment for TMA load.
         auto const numBytesAlignmentLoadB = 1024;
         // No need to reuse the first chunk.
@@ -239,9 +246,10 @@ class KernelTraits {
       {
         // Number of bytes in save shuffled B in shared memory.
         auto const numSmemBytesLoadB =
-            numSlicesForSliceK > 1 ? numStages * tileN * tileK *
-                                         getNumSmemBitsPerElt(dtypeB, mMmaKind, mmaK) / 8 /* bits */
-                                   : 0;
+            numSlicesForSliceK > 1
+                ? numStages * tileN * tileK *
+                      getNumSmemBitsPerElt(dtypeB, mMmaKind, mmaK, isSparseA) / 8 /* bits */
+                : 0;
         // Number of bytes for load B alignment for TMA load.
         auto const numBytesAlignmentLoadB = 1024;
         // No need to reuse the first chunk.
@@ -301,6 +309,23 @@ class KernelTraits {
         firstChunkReuseSmem.emplace_back(reuseFirstChunksSmemStoreC);
       }
 
+      // SmemSparsityInfoA
+      {
+        // Number of bytes for sparsity info in SMEM.
+        auto const numBytesSmemSparsityInfoA =
+            numStages * tileM * tg::getNumBytesSparsityInfo(sparsityA, tileK);
+        // Number of bytes alignment for sparsity info in SMEM.
+        auto const numBytesAlignmentSparsityInfoA = 1024;
+        // No need to reuse the first chunk.
+        auto const reuseChunksSmemSparsityInfoA = false;
+
+        // Add info.
+        smemChunkNames.emplace_back("smemSparsityInfoA");
+        numBytesAndAlignmentPerSmemChunk.emplace_back(
+            std::make_pair(numBytesSmemSparsityInfoA, numBytesAlignmentSparsityInfoA));
+        firstChunkReuseSmem.emplace_back(reuseChunksSmemSparsityInfoA);
+      }
+
       // RowMax
       {
         // Number of dqSfsC per CTA.
@@ -371,7 +396,6 @@ class KernelTraits {
       // Per-block absolute maximum for multi-warp reduction.
       {
         // Number of bytes: number of epilogue warps * number of tile columns.
-        // TODO: avoid allocating this memory when it's not needed (it's only for MxFp8 + fusedAct)
         auto const numBytesSmemBlockAmax = transposeMmaOutput ? 4 * tileN * sizeof(float) : 0;
         // Number of bytes alignment.
         auto const numBytesAlignmentBlockAmax = 16;
@@ -475,13 +499,11 @@ class KernelTraits {
         bool const useBlockScalingA = tg::dtypeIsBlockFmt(dtypeMmaA);
         // Are the block scales constant?
         bool const useConstSfA = useBlockScalingA && !tg::dtypeIsBlockFmt(dtypeA);
-        // Number elements per scaling factor.
-        int32_t const numEltsPerSf = useBlockScalingA ? tg::dtypeNumEltsPerSf(dtypeMmaA) : -1;
         // TMEM cols group size in the K dimension.
         int32_t kGroupSize = 4;
         // Number of columns per stage.
         int32_t const numColsPerStage =
-            useBlockScalingA ? ((tileK / (kGroupSize * numEltsPerSf)) *
+            useBlockScalingA ? ((tileK / (kGroupSize * numEltsPerSfA)) *
                                 tg::getTmemColStridePerGroup(tileM, mmaK, kGroupSize))
                              : 0;
         // Number of columns for scaling factors of A.
@@ -506,13 +528,11 @@ class KernelTraits {
         bool const useBlockScalingB = tg::dtypeIsBlockFmt(dtypeMmaB);
         // Are the block scales constant?
         bool const useConstSfB = useBlockScalingB && !tg::dtypeIsBlockFmt(dtypeB);
-        // Number elements per scaling factor.
-        int32_t const numEltsPerSf = useBlockScalingB ? tg::dtypeNumEltsPerSf(dtypeMmaB) : -1;
         // TMEM cols group size in the K dimension.
         int32_t kGroupSize = 4;
         // Number of columns per stage.
         int32_t const numColsPerStage =
-            useBlockScalingB ? ((tileK / (kGroupSize * numEltsPerSf)) *
+            useBlockScalingB ? ((tileK / (kGroupSize * numEltsPerSfB)) *
                                 tg::getTmemColStridePerGroup(tileN, mmaK, kGroupSize))
                              : 0;
         // Number of columns for scaling factors of B.
@@ -531,6 +551,23 @@ class KernelTraits {
         firstChunkReuseTmem.emplace_back(reuseChunksTmemSfB);
       }
 
+      // Sparsity info for A
+      {
+        // Number of columns for the sparsity info for A (note: for Dense, this is 0).
+        auto const numTmemColsSparsityInfoA =
+            numStages * tg::getNumBytesSparsityInfo(sparsityA, tileK) / 4 /* bytes */;
+        // Number of columns for Sf alignment.
+        auto const numColsAlignmentSparsityInfoA = 2;
+        // No need to reuse TMEM.
+        auto const reuseChunksTmemSparsityInfoA = false;
+
+        // Add info.
+        tmemChunkNames.emplace_back("tmemSparsityInfoA");
+        numBytesAndAlignmentPerTmemChunk.emplace_back(
+            std::make_pair(numTmemColsSparsityInfoA, numColsAlignmentSparsityInfoA));
+        firstChunkReuseTmem.emplace_back(reuseChunksTmemSparsityInfoA);
+      }
+
       // Create TMEM helper object.
       mTmemAllocatorHelper =
           MemAllocatorHelper(numBytesAndAlignmentPerTmemChunk, firstChunkReuseTmem, tmemChunkNames);
@@ -634,6 +671,12 @@ inline int32_t getSmemOffsetConstSfBuf(KernelTraits traits) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+inline int32_t getSmemOffsetSparsityInfoA(KernelTraits traits) {
+  return traits.mSmemAllocatorHelper.getChunkOffsetByName("smemSparsityInfoA");
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
 inline int32_t isSmemAbRepurposedToGmemC(KernelTraits traits, int resIdx = 0) {
   return traits.mSmemAllocatorHelper.getFirstChunkReuseFlagByName("smemGmemC" +
                                                                   std::to_string(resIdx));
@@ -669,6 +712,12 @@ inline int32_t getTmemOffsetSfB(KernelTraits traits) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+inline int32_t getTmemOffsetSparsityInfoA(KernelTraits traits) {
+  return traits.mTmemAllocatorHelper.getChunkOffsetByName("tmemSparsityInfoA");
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
 }  // namespace gemm
 
 }  // namespace batchedGemm
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/TmaDescriptor.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/TmaDescriptor.h
index 9fff10deff..b800ecaf83 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/TmaDescriptor.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/TmaDescriptor.h
@@ -46,7 +46,8 @@ inline CUtensorMap buildNdTmaDescriptor(tg::Dtype dtype, std::vector<uint64_t> c
   CUtensorMap desc{};
   // The data type.
   CUtensorMapDataType tmaDataFormat{CU_TENSOR_MAP_DATA_TYPE_FLOAT32};
-  if (dtype == tg::Dtype::E4m3 || dtype == tg::Dtype::MxE4m3 || dtype == tg::Dtype::UE8m0) {
+  if (dtype == tg::Dtype::E4m3 || dtype == tg::Dtype::MxE4m3 || dtype == tg::Dtype::UE8m0 ||
+      dtype == tg::Dtype::UInt8) {
     tmaDataFormat = CU_TENSOR_MAP_DATA_TYPE_UINT8;
   } else if (dtype == tg::Dtype::Fp16) {
     tmaDataFormat = CU_TENSOR_MAP_DATA_TYPE_FLOAT16;
@@ -79,9 +80,9 @@ inline CUtensorMap buildNdTmaDescriptor(tg::Dtype dtype, std::vector<uint64_t> c
       swizzleType = CU_TENSOR_MAP_SWIZZLE_64B;
     } else if ((fastestDimTileSizeBytes % 32) == 0) {
       swizzleType = CU_TENSOR_MAP_SWIZZLE_32B;
-      // This path is only for the scaling factors.
     } else if ((fastestDimTileSizeBytes % 16) == 0 &&
-               (dtype == tg::Dtype::UE8m0 || dtype == tg::Dtype::E4m3)) {
+               (dtype == tg::Dtype::UE8m0 || dtype == tg::Dtype::E4m3 || dtype == tg::Dtype::E2m1 ||
+                dtype == tg::Dtype::UInt8)) {
       swizzleType = CU_TENSOR_MAP_SWIZZLE_NONE;
     } else {
       std::cerr << "buildNdTmaDescriptor: unexpected fastestDimTileSizeBytes "
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/DtypeDecl.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/DtypeDecl.h
index 355cfba961..76ea5ebc83 100644
--- a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/DtypeDecl.h
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/DtypeDecl.h
@@ -198,14 +198,16 @@ inline Dtype dtypeEltType(Dtype dtype) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
-inline int dtypeNumEltsPerSf(Dtype dtype) {
+// Note: the block size from the options should be used instead.
+// TODO: remove this function?
+inline int dtypeNumEltsPerSf(Dtype dtype, bool useSparsity = false) {
   switch (dtype) {
     case Dtype::E2m1:
-      return 16;
+      return useSparsity ? 32 : 16;
     case Dtype::MxE2m1:
     case Dtype::MxE4m3:
     case Dtype::MxInt4:
-      return 32;
+      return useSparsity ? 64 : 32;
     default:
       assert(false);
       return -1;
@@ -271,6 +273,14 @@ inline MmaKind dtypeGetMmaKind(Dtype dtypeA, Dtype dtypeB) {
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////
 
+inline bool dtypeNeedsPadding(Dtype dtype, MmaKind mmaKind, [[maybe_unused]] int mmaK,
+                              [[maybe_unused]] bool isSparseA) {
+  bool needsPadding = mmaKind == MmaKind::MxFp8Fp6Fp4 && dtype == Dtype::MxE2m1;
+  return needsPadding;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
 }  // namespace gen
 }  // namespace trtllm
 
diff --git a/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/SparsityDecl.h b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/SparsityDecl.h
new file mode 100644
index 0000000000..f093d5d1f9
--- /dev/null
+++ b/include/flashinfer/trtllm/batched_gemm/trtllmGen_bmm_export/trtllm/gen/SparsityDecl.h
@@ -0,0 +1,134 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 1993-2025 NVIDIA CORPORATION &
+ * AFFILIATES. All rights reserved. SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cassert>
+#include <string>
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+//
+// Be careful when modifying this file as it is included by the generated kernels. For example, do
+// not add TLLM_CHECK_* constructs in this file. Thanks!
+//
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+namespace batchedGemm {
+
+namespace trtllm {
+namespace gen {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// This enumeration defines structured sparsity modes. Please refer to the PTX ISA for more details.
+enum class Sparsity {
+  // No sparsity.
+  Dense,
+
+  // For each chunk of 2 elements, 1 is non-zero. Only non-zero elements are stored.
+  // A 4-bit index is used to indicate the position of the non-zero element.
+  // The index may only take the value 0b1110 or 0b0100, other values are undefined behavior.
+  //
+  // 0b1110:                               0b0100:
+  // |------ a ------|------ 0 ------|     |------ 0 ------|------ a ------|
+  // |  11   |  10   |  01   |  00   |     |  11   |  10   |  01   |  00   |
+  Any_1_2,
+
+  // For each chunk of 4 elements, 2 are non-zero. Only non-zero elements are stored.
+  // A 4-bit index is used to indicate the position of the non-zero elements.
+  // Meaningful values are: 0b0100, 0b1000, 0b1100, 0b1001, 0b1101, 0b1110.
+  // Most other values are undefined behavior.
+  //
+  // E.g. 0b1100 corresponds to:
+  // |-- b --|-- 0 --|-- 0 --|-- a --|
+  // |  11   |  10   |  01   |  00   |
+  Any_2_4,
+
+  // For each chunk of 8 elements, 4 are non-zero. Only non-zero elements are stored.
+  // Further, the zero and non-zero elements are grouped in pairs.
+  // A 4-bit index is used to indicate the position of the non-zero elements.
+  // Meaningful values are: 0b0100, 0b1000, 0b1100, 0b1001, 0b1101, 0b1110.
+  // Most other values are undefined behavior.
+  //
+  // E.g. 0b1100 corresponds to:
+  // | d | c | 0 | 0 | 0 | 0 | b | a |
+  // |  11   |  10   |  01   |  00   |
+  Pairwise_4_8,
+};
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+inline bool isSparse(Sparsity sparsity) { return sparsity != Sparsity::Dense; }
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+inline std::string sparsityToString(Sparsity sparsity) {
+  switch (sparsity) {
+    case Sparsity::Dense:
+      return "dense";
+    case Sparsity::Any_1_2:
+      return "1:2";
+    case Sparsity::Any_2_4:
+      return "2:4";
+    case Sparsity::Pairwise_4_8:
+      return "4:8";
+    default:
+      assert(false);
+      return "Unsupported sparsity";
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Size of a sparsity chunk, for sparse modes.
+inline int32_t getSparsityChunkSize(Sparsity sparsity) {
+  switch (sparsity) {
+    case Sparsity::Any_1_2:
+      return 2;
+    case Sparsity::Any_2_4:
+      return 4;
+    case Sparsity::Pairwise_4_8:
+      return 8;
+    case Sparsity::Dense:
+    default:
+      assert(false);
+      return 0;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Number of bytes needed to store the sparsity information.
+inline size_t getNumBytesSparsityInfo(Sparsity sparsity, size_t numElts) {
+  switch (sparsity) {
+    case Sparsity::Dense:
+      return 0;
+    case Sparsity::Any_1_2:
+    case Sparsity::Any_2_4:
+    case Sparsity::Pairwise_4_8:
+      return numElts / getSparsityChunkSize(sparsity) * 4 /*bits*/ / 8;
+    default:
+      assert(false);
+      return 0;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+}  // namespace gen
+}  // namespace trtllm
+
+}  // namespace batchedGemm
