diff --git a/csrc/pytorch_extension_utils.h b/csrc/pytorch_extension_utils.h
index 263c1af0e9..3e0e390dcc 100644
--- a/csrc/pytorch_extension_utils.h
+++ b/csrc/pytorch_extension_utils.h
@@ -67,6 +67,31 @@ FLASHINFER_EXT_MODULE_INIT_EXPAND(TORCH_EXTENSION_NAME)
 
 #endif
 
+#define _DISPATCH_CASE_I32(c_type, ...) \
+  case at::ScalarType::Int: {           \
+    using c_type = int32_t;             \
+    return __VA_ARGS__();               \
+  }
+
+#define _DISPATCH_CASE_I64(c_type, ...) \
+  case at::ScalarType::Long: {          \
+    using c_type = int64_t;             \
+    return __VA_ARGS__();               \
+  }
+
+#define DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(pytorch_dtype, c_type, ...)                  \
+  [&]() -> bool {                                                                     \
+    switch (pytorch_dtype) {                                                          \
+      _DISPATCH_CASE_I32(c_type, __VA_ARGS__)                                         \
+      _DISPATCH_CASE_I64(c_type, __VA_ARGS__)                                         \
+      default:                                                                        \
+        std::ostringstream oss;                                                       \
+        oss << __PRETTY_FUNCTION__ << " failed to dispatch idtype " << pytorch_dtype; \
+        TORCH_CHECK(false, oss.str());                                                \
+        return false;                                                                 \
+    }                                                                                 \
+  }()
+
 #ifdef FLASHINFER_ENABLE_F16
 #define _DISPATCH_CASE_F16(c_type, ...) \
   case at::ScalarType::Half: {          \
diff --git a/csrc/rope.cu b/csrc/rope.cu
index 94b2492d2a..190643e050 100644
--- a/csrc/rope.cu
+++ b/csrc/rope.cu
@@ -48,20 +48,23 @@ void apply_rope(at::Tensor q, at::Tensor k, at::Tensor q_rope, at::Tensor k_rope
   size_t q_rope_stride_h = q_rope.stride(1);
   size_t k_rope_stride_n = k_rope.stride(0);
   size_t k_rope_stride_h = k_rope.stride(1);
+  CHECK_EQ(indptr.scalar_type(), offsets.scalar_type());
 
   const c10::cuda::OptionalCUDAGuard device_guard(q.device());
   auto stream = at::cuda::getCurrentCUDAStream();
   DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    cudaError_t status = BatchQKApplyRotary(
-        static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
-        static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<int32_t*>(indptr.data_ptr()), static_cast<int32_t*>(offsets.data_ptr()),
-        batch_size, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h,
-        k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n, k_rope_stride_h,
-        interleave, rope_scale, rope_theta, stream);
-    TORCH_CHECK(status == cudaSuccess, "BatchQKApplyRotary failed with error code " +
-                                           std::string(cudaGetErrorString(status)));
-    return true;
+    return DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(indptr.scalar_type(), c_idtype, [&] {
+      cudaError_t status = BatchQKApplyRotary(
+          static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
+          static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
+          static_cast<c_idtype*>(indptr.data_ptr()), static_cast<c_idtype*>(offsets.data_ptr()),
+          batch_size, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h,
+          k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n,
+          k_rope_stride_h, interleave, rope_scale, rope_theta, stream);
+      TORCH_CHECK(status == cudaSuccess, "BatchQKApplyRotary failed with error code " +
+                                             std::string(cudaGetErrorString(status)));
+      return true;
+    });
   });
 }
 
@@ -94,15 +97,18 @@ void apply_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, at::Tenso
   const c10::cuda::OptionalCUDAGuard device_guard(q.device());
   auto stream = at::cuda::getCurrentCUDAStream();
   DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    cudaError_t status = BatchQKApplyRotaryPosIds(
-        static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
-        static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<int32_t*>(pos_ids.data_ptr()), nnz, num_qo_heads, num_kv_heads, rotary_dim,
-        head_dim, q_stride_n, q_stride_h, k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h,
-        k_rope_stride_n, k_rope_stride_h, interleave, rope_scale, rope_theta, stream);
-    TORCH_CHECK(status == cudaSuccess, "BatchQKApplyRotaryPosIds failed with error code " +
-                                           std::string(cudaGetErrorString(status)));
-    return true;
+    return DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(pos_ids.scalar_type(), c_idtype, [&] {
+      cudaError_t status = BatchQKApplyRotaryPosIds(
+          static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
+          static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
+          static_cast<c_idtype*>(pos_ids.data_ptr()), nnz, num_qo_heads, num_kv_heads, rotary_dim,
+          head_dim, q_stride_n, q_stride_h, k_stride_n, k_stride_h, q_rope_stride_n,
+          q_rope_stride_h, k_rope_stride_n, k_rope_stride_h, interleave, rope_scale, rope_theta,
+          stream);
+      TORCH_CHECK(status == cudaSuccess, "BatchQKApplyRotaryPosIds failed with error code " +
+                                             std::string(cudaGetErrorString(status)));
+      return true;
+    });
   });
 }
 
@@ -141,17 +147,19 @@ void apply_rope_pos_ids_cos_sin_cache(at::Tensor q, at::Tensor k, at::Tensor q_r
   const c10::cuda::OptionalCUDAGuard device_guard(q.device());
   auto stream = at::cuda::getCurrentCUDAStream();
   DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    cudaError_t status = BatchQKApplyRotaryPosIdsCosSinCache(
-        static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
-        static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<float*>(cos_sin_cache.data_ptr()), static_cast<int32_t*>(pos_ids.data_ptr()),
-        nnz, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h, k_stride_n,
-        k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n, k_rope_stride_h, interleave,
-        stream);
-    TORCH_CHECK(status == cudaSuccess,
-                "BatchQKApplyRotaryPosIdsCosSinCache failed with error code " +
-                    std::string(cudaGetErrorString(status)));
-    return true;
+    return DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(pos_ids.scalar_type(), c_idtype, [&] {
+      cudaError_t status = BatchQKApplyRotaryPosIdsCosSinCache(
+          static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
+          static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
+          static_cast<float*>(cos_sin_cache.data_ptr()), static_cast<c_idtype*>(pos_ids.data_ptr()),
+          nnz, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h, k_stride_n,
+          k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n, k_rope_stride_h,
+          interleave, stream);
+      TORCH_CHECK(status == cudaSuccess,
+                  "BatchQKApplyRotaryPosIdsCosSinCache failed with error code " +
+                      std::string(cudaGetErrorString(status)));
+      return true;
+    });
   });
 }
 
@@ -177,6 +185,7 @@ void apply_llama31_rope(at::Tensor q, at::Tensor k, at::Tensor q_rope, at::Tenso
   unsigned int head_dim = q.size(2);
   unsigned int batch_size = offsets.size(0);
   CHECK_EQ(indptr.size(0), batch_size + 1);
+  CHECK_EQ(indptr.scalar_type(), offsets.scalar_type());
   size_t q_stride_n = q.stride(0);
   size_t q_stride_h = q.stride(1);
   size_t k_stride_n = k.stride(0);
@@ -185,21 +194,24 @@ void apply_llama31_rope(at::Tensor q, at::Tensor k, at::Tensor q_rope, at::Tenso
   size_t q_rope_stride_h = q_rope.stride(1);
   size_t k_rope_stride_n = k_rope.stride(0);
   size_t k_rope_stride_h = k_rope.stride(1);
+  CHECK_EQ(indptr.scalar_type(), offsets.scalar_type());
 
   const c10::cuda::OptionalCUDAGuard device_guard(q.device());
   auto stream = at::cuda::getCurrentCUDAStream();
   DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    cudaError_t status = BatchQKApplyLlama31Rotary(
-        static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
-        static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<int32_t*>(indptr.data_ptr()), static_cast<int32_t*>(offsets.data_ptr()),
-        batch_size, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h,
-        k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n, k_rope_stride_h,
-        interleave, rope_scale, rope_theta, low_freq_factor, high_freq_factor, old_context_length,
-        stream);
-    TORCH_CHECK(status == cudaSuccess, "BatchQKApplyLlama31Rotary failed with error code " +
-                                           std::string(cudaGetErrorString(status)));
-    return true;
+    return DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(indptr.scalar_type(), c_idtype, [&] {
+      cudaError_t status = BatchQKApplyLlama31Rotary(
+          static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
+          static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
+          static_cast<c_idtype*>(indptr.data_ptr()), static_cast<c_idtype*>(offsets.data_ptr()),
+          batch_size, num_qo_heads, num_kv_heads, rotary_dim, head_dim, q_stride_n, q_stride_h,
+          k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h, k_rope_stride_n,
+          k_rope_stride_h, interleave, rope_scale, rope_theta, low_freq_factor, high_freq_factor,
+          old_context_length, stream);
+      TORCH_CHECK(status == cudaSuccess, "BatchQKApplyLlama31Rotary failed with error code " +
+                                             std::string(cudaGetErrorString(status)));
+      return true;
+    });
   });
 }
 
@@ -233,15 +245,17 @@ void apply_llama31_rope_pos_ids(at::Tensor q, at::Tensor k, at::Tensor q_rope, a
   const c10::cuda::OptionalCUDAGuard device_guard(q.device());
   auto stream = at::cuda::getCurrentCUDAStream();
   DISPATCH_PYTORCH_DTYPE_TO_CTYPE_FP16(q.scalar_type(), c_type, [&] {
-    cudaError_t status = BatchQKApplyLlama31RotaryPosIds(
-        static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
-        static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
-        static_cast<int32_t*>(pos_ids.data_ptr()), nnz, num_qo_heads, num_kv_heads, rotary_dim,
-        head_dim, q_stride_n, q_stride_h, k_stride_n, k_stride_h, q_rope_stride_n, q_rope_stride_h,
-        k_rope_stride_n, k_rope_stride_h, interleave, rope_scale, rope_theta, low_freq_factor,
-        high_freq_factor, old_context_length, stream);
-    TORCH_CHECK(status == cudaSuccess, "BatchQKApplyLlama31RotaryPosIds failed with error code " +
-                                           std::string(cudaGetErrorString(status)));
-    return true;
+    return DISPATCH_PYTORCH_IDTYPE_TO_CTYPE(pos_ids.scalar_type(), c_idtype, [&] {
+      cudaError_t status = BatchQKApplyLlama31RotaryPosIds(
+          static_cast<c_type*>(q.data_ptr()), static_cast<c_type*>(k.data_ptr()),
+          static_cast<c_type*>(q_rope.data_ptr()), static_cast<c_type*>(k_rope.data_ptr()),
+          static_cast<c_idtype*>(pos_ids.data_ptr()), nnz, num_qo_heads, num_kv_heads, rotary_dim,
+          head_dim, q_stride_n, q_stride_h, k_stride_n, k_stride_h, q_rope_stride_n,
+          q_rope_stride_h, k_rope_stride_n, k_rope_stride_h, interleave, rope_scale, rope_theta,
+          low_freq_factor, high_freq_factor, old_context_length, stream);
+      TORCH_CHECK(status == cudaSuccess, "BatchQKApplyLlama31RotaryPosIds failed with error code " +
+                                             std::string(cudaGetErrorString(status)));
+      return true;
+    });
   });
 }
diff --git a/flashinfer/rope.py b/flashinfer/rope.py
index 4e1a19c3a0..64c0a944c0 100644
--- a/flashinfer/rope.py
+++ b/flashinfer/rope.py
@@ -65,8 +65,6 @@ def _apply_rope(
     rope_scale: float,
     rope_theta: float,
 ) -> None:
-    indptr = indptr.int()
-    offsets = offsets.int()
     get_module_attr("apply_rope")(
         q,
         k,
@@ -113,8 +111,6 @@ def _apply_llama31_rope(
     high_freq_factor: float,
     old_context_len: float,
 ) -> None:
-    indptr = indptr.int()
-    offsets = offsets.int()
     get_module_attr("apply_llama31_rope")(
         q,
         k,
@@ -163,7 +159,6 @@ def _apply_rope_pos_ids(
     rope_scale: float,
     rope_theta: float,
 ) -> None:
-    pos_ids = pos_ids.int()
     get_module_attr("apply_rope_pos_ids")(
         q,
         k,
@@ -204,7 +199,6 @@ def _apply_rope_pos_ids_cos_sin_cache(
     pos_ids: torch.Tensor,
     interleave: bool,
 ) -> None:
-    pos_ids = pos_ids.int()
     get_module_attr("apply_rope_pos_ids_cos_sin_cache")(
         q,
         k,
@@ -247,7 +241,6 @@ def _apply_llama31_rope_pos_ids(
     high_freq_factor: float,
     old_context_len: float,
 ) -> None:
-    pos_ids = pos_ids.int()
     get_module_attr("apply_llama31_rope_pos_ids")(
         q,
         k,
diff --git a/tests/test_rope.py b/tests/test_rope.py
index 4e0c40b1cc..d71ca34e65 100644
--- a/tests/test_rope.py
+++ b/tests/test_rope.py
@@ -146,6 +146,7 @@ def test_rope(
 @pytest.mark.parametrize("partial_rotary_factor", [0.25, 0.5, 0.75, 1.0])
 @pytest.mark.parametrize("inplace", [False, True])
 @pytest.mark.parametrize("interleave", [True, False])
+@pytest.mark.parametrize("idtype", [torch.int32, torch.int64])
 def test_rope_pos_ids(
     batch_size,
     qkv_len,
@@ -157,6 +158,7 @@ def test_rope_pos_ids(
     partial_rotary_factor,
     inplace,
     interleave,
+    idtype,
 ):
     rotary_dim = int(head_dim * partial_rotary_factor)
     nnz = batch_size * qkv_len
@@ -171,13 +173,13 @@ def test_rope_pos_ids(
         :, num_qo_heads * head_dim : (num_qo_heads + num_kv_heads) * head_dim
     ].reshape(nnz, num_kv_heads, head_dim)
     indptr = torch.tensor(
-        [i * qkv_len for i in range(batch_size + 1)], dtype=torch.int32, device="cuda:0"
+        [i * qkv_len for i in range(batch_size + 1)], dtype=idtype, device="cuda:0"
     )
-    offsets = torch.full((batch_size,), offset, dtype=torch.int32, device="cuda:0")
+    offsets = torch.full((batch_size,), offset, dtype=idtype, device="cuda:0")
 
     pos_ids = torch.cat(
         [
-            torch.arange(offset, qkv_len + offset, dtype=torch.int32)
+            torch.arange(offset, qkv_len + offset, dtype=idtype)
             for _ in range(batch_size)
         ]
     ).to("cuda:0")
