diff --git a/cub/cub/device/dispatch/tuning/tuning_transform.cuh b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
index ad5ecff1694..9199862ee25 100644
--- a/cub/cub/device/dispatch/tuning/tuning_transform.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
@@ -433,6 +433,11 @@ struct policy_hub<RequiresStableAddress,
         alignment)
       > int{max_smem_per_block};
 
+    // on Hopper, the vectorized kernel performs better for 1 and 2 byte values
+    static constexpr bool use_vector_kernel_on_hopper =
+      ((size_of<it_value_t<RandomAccessIteratorsIn>> < 4) && ...) && sizeof...(RandomAccessIteratorsIn) > 1
+      && size_of<it_value_t<RandomAccessIteratorOut>> < 4;
+
     // if each tile size is a multiple of the bulk copy and maximum value type alignments, the alignment is retained if
     // the base pointer is sufficiently aligned (the correct check would be if it's a multiple of all value types
     // following the current tile). we would otherwise need to realign every SMEM tile individually, which is costly and
@@ -444,7 +449,7 @@ struct policy_hub<RequiresStableAddress,
     static constexpr bool enough_threads_for_peeling = AsyncBlockSize >= alignment; // head and tail bytes
     static constexpr bool fallback_to_vectorized =
       exhaust_smem || !tile_sizes_retain_alignment || !enough_threads_for_peeling || no_input_streams
-      || !can_memcpy_all_inputs;
+      || !can_memcpy_all_inputs || (PtxVersion == 900 && use_vector_kernel_on_hopper);
 
   public:
     static constexpr auto algorithm =
