diff --git a/cub/cub/detail/type_traits.cuh b/cub/cub/detail/type_traits.cuh
index 3fd9f0e7e46..705f71ca4ea 100644
--- a/cub/cub/detail/type_traits.cuh
+++ b/cub/cub/detail/type_traits.cuh
@@ -73,7 +73,7 @@ inline constexpr bool is_fixed_size_random_access_range_v<::cuda::std::span<T, N
 
 template <typename T, typename E, typename L, typename A>
 inline constexpr bool is_fixed_size_random_access_range_v<::cuda::std::mdspan<T, E, L, A>> =
-  E::rank == 1 && E::rank_dynamic() == 0;
+  E::rank() == 1 && E::rank_dynamic() == 0;
 
 /***********************************************************************************************************************
  * static_size: a type trait that returns the number of elements in an Array-like type
@@ -94,7 +94,7 @@ inline constexpr int static_size_v<::cuda::std::span<T, N>> =
 
 template <typename T, typename E, typename L, typename A>
 inline constexpr int static_size_v<::cuda::std::mdspan<T, E, L, A>> =
-  ::cuda::std::enable_if_t<E::rank == 1 && E::rank_dynamic() == 0, int>{E::static_extent(0)};
+  ::cuda::std::enable_if_t<E::rank() == 1 && E::rank_dynamic() == 0, int>{E::static_extent(0)};
 
 template <typename T>
 using implicit_prom_t = decltype(+T{});
diff --git a/cub/test/catch2_test_thread_scan_exclusive_partial.cu b/cub/test/catch2_test_thread_scan_exclusive_partial.cu
index 2a10db13da4..c41dc90f6f3 100644
--- a/cub/test/catch2_test_thread_scan_exclusive_partial.cu
+++ b/cub/test/catch2_test_thread_scan_exclusive_partial.cu
@@ -7,6 +7,7 @@
 #include <cuda/std/__algorithm/clamp.h>
 #include <cuda/std/functional>
 #include <cuda/std/limits>
+#include <cuda/std/mdspan>
 
 #include "catch2_test_device_reduce.cuh"
 #include "catch2_test_device_scan.cuh"
@@ -86,8 +87,6 @@ __global__ void thread_scan_exclusive_partial_kernel_span(
   }
 }
 
-#if _CCCL_STD_VER >= 2023
-
 template <int NumItems, typename T, typename ScanOperator>
 __global__ void thread_scan_exclusive_partial_kernel_mdspan(
   const T* d_in, T* d_out, ScanOperator scan_operator, int valid_items, T prefix, bool apply_prefix)
@@ -109,8 +108,6 @@ __global__ void thread_scan_exclusive_partial_kernel_mdspan(
   }
 }
 
-#endif // _CCCL_STD_VER >= 2023
-
 /***********************************************************************************************************************
  * Type list definition
  **********************************************************************************************************************/
@@ -361,7 +358,6 @@ C2H_TEST("ThreadScanExclusive Container Tests", "[scan][thread]")
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == d_out);
 
-#if _CCCL_STD_VER >= 2023
   thrust::fill(d_out.begin(), d_out.end(), 0);
   thread_scan_exclusive_partial_kernel_mdspan<max_size><<<1, 1>>>(
     thrust::raw_pointer_cast(d_in.data()),
@@ -373,7 +369,6 @@ C2H_TEST("ThreadScanExclusive Container Tests", "[scan][thread]")
   REQUIRE(cudaSuccess == cudaPeekAtLastError());
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == d_out);
-#endif // _CCCL_STD_VER >= 2023
 }
 
 C2H_TEST("ThreadScanExclusive Invalid Test", "[scan][thread]")
diff --git a/cub/test/catch2_test_thread_scan_inclusive_partial.cu b/cub/test/catch2_test_thread_scan_inclusive_partial.cu
index cee6bdee1ca..b2efb7aa645 100644
--- a/cub/test/catch2_test_thread_scan_inclusive_partial.cu
+++ b/cub/test/catch2_test_thread_scan_inclusive_partial.cu
@@ -7,6 +7,7 @@
 #include <cuda/std/__algorithm/clamp.h>
 #include <cuda/std/functional>
 #include <cuda/std/limits>
+#include <cuda/std/mdspan>
 
 #include "catch2_test_device_reduce.cuh"
 #include "catch2_test_device_scan.cuh"
@@ -86,8 +87,6 @@ __global__ void thread_scan_inclusive_partial_kernel_span(
   }
 }
 
-#if _CCCL_STD_VER >= 2023
-
 template <int NumItems, typename T, typename ScanOperator>
 __global__ void thread_scan_inclusive_partial_kernel_mdspan(
   const T* d_in, T* d_out, ScanOperator scan_operator, int valid_items, T prefix, bool apply_prefix)
@@ -109,8 +108,6 @@ __global__ void thread_scan_inclusive_partial_kernel_mdspan(
   }
 }
 
-#endif // _CCCL_STD_VER >= 2023
-
 /***********************************************************************************************************************
  * Type list definition
  **********************************************************************************************************************/
@@ -361,7 +358,6 @@ C2H_TEST("ThreadScanInclusive Container Tests", "[scan][thread]")
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == d_out);
 
-#if _CCCL_STD_VER >= 2023
   thrust::fill(d_out.begin(), d_out.end(), 0);
   thread_scan_inclusive_partial_kernel_mdspan<max_size><<<1, 1>>>(
     thrust::raw_pointer_cast(d_in.data()),
@@ -373,7 +369,6 @@ C2H_TEST("ThreadScanInclusive Container Tests", "[scan][thread]")
   REQUIRE(cudaSuccess == cudaPeekAtLastError());
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == d_out);
-#endif // _CCCL_STD_VER >= 2023
 }
 
 C2H_TEST("ThreadScanInclusive Invalid Test", "[scan][thread]")
diff --git a/cub/test/thread_reduce/catch2_test_thread_reduce.cu b/cub/test/thread_reduce/catch2_test_thread_reduce.cu
index 08fe87e3db9..95b4bc1e508 100644
--- a/cub/test/thread_reduce/catch2_test_thread_reduce.cu
+++ b/cub/test/thread_reduce/catch2_test_thread_reduce.cu
@@ -8,6 +8,7 @@
 #include <cuda/functional>
 #include <cuda/std/functional>
 #include <cuda/std/limits>
+#include <cuda/std/mdspan>
 #include <cuda/std/type_traits>
 
 #include <cstring>
@@ -63,8 +64,6 @@ __global__ void thread_reduce_kernel_span(const T* d_in, T* d_out, ReduceOperato
   *d_out = cub::ThreadReduce(span, reduce_operator);
 }
 
-#if _CCCL_STD_VER >= 2023
-
 template <int NUM_ITEMS, typename T, typename ReduceOperator>
 __global__ void thread_reduce_kernel_mdspan(const T* d_in, T* d_out, ReduceOperator reduce_operator)
 {
@@ -80,8 +79,6 @@ __global__ void thread_reduce_kernel_mdspan(const T* d_in, T* d_out, ReduceOpera
   *d_out = cub::ThreadReduce(mdspan, reduce_operator);
 }
 
-#endif // _CCCL_STD_VER >= 2023
-
 /***********************************************************************************************************************
  * CUB operator to STD operator
  **********************************************************************************************************************/
@@ -354,11 +351,9 @@ C2H_TEST("ThreadReduce Container Tests", "[reduce][thread]")
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   verify_results(reference_result, c2h::host_vector<int>(d_out)[0]);
 
-#if _CCCL_STD_VER >= 2023
   thread_reduce_kernel_mdspan<max_size>
     <<<1, 1>>>(thrust::raw_pointer_cast(d_in.data()), thrust::raw_pointer_cast(d_out.data()), cuda::std::plus<>{});
   REQUIRE(cudaSuccess == cudaPeekAtLastError());
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   verify_results(reference_result, c2h::host_vector<int>(d_out)[0]);
-#endif // _CCCL_STD_VER >= 2023
 }
diff --git a/cub/test/thread_reduce/catch2_test_thread_reduce_partial.cu b/cub/test/thread_reduce/catch2_test_thread_reduce_partial.cu
index cb15261a4bc..7b83d019ae5 100644
--- a/cub/test/thread_reduce/catch2_test_thread_reduce_partial.cu
+++ b/cub/test/thread_reduce/catch2_test_thread_reduce_partial.cu
@@ -9,6 +9,7 @@
 #include <cuda/std/__algorithm/clamp.h>
 #include <cuda/std/functional>
 #include <cuda/std/limits>
+#include <cuda/std/mdspan>
 #include <cuda/std/type_traits>
 
 #include "catch2_test_device_reduce.cuh"
@@ -68,8 +69,6 @@ thread_reduce_partial_kernel_span(const T* d_in, T* d_out, ReduceOperator reduce
   *d_out = cub::detail::ThreadReducePartial(span, reduce_operator, valid_items);
 }
 
-#if _CCCL_STD_VER >= 2023
-
 template <int NumItems, typename T, typename ReduceOperator>
 __global__ void
 thread_reduce_partial_kernel_mdspan(const T* d_in, T* d_out, ReduceOperator reduce_operator, int valid_items)
@@ -86,8 +85,6 @@ thread_reduce_partial_kernel_mdspan(const T* d_in, T* d_out, ReduceOperator redu
   *d_out = cub::detail::ThreadReducePartial(mdspan, reduce_operator, valid_items);
 }
 
-#endif // _CCCL_STD_VER >= 2023
-
 /***********************************************************************************************************************
  * Type list definition
  **********************************************************************************************************************/
@@ -251,13 +248,11 @@ C2H_TEST("ThreadReduce Container Tests", "[reduce][thread]")
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == c2h::host_vector<int>(d_out)[0]);
 
-#if _CCCL_STD_VER >= 2023
   thread_reduce_partial_kernel_mdspan<max_size>
     <<<1, 1>>>(thrust::raw_pointer_cast(d_in.data()), thrust::raw_pointer_cast(d_out.data()), op_t{}, valid_items);
   REQUIRE(cudaSuccess == cudaPeekAtLastError());
   REQUIRE(cudaSuccess == cudaDeviceSynchronize());
   REQUIRE(reference_result == c2h::host_vector<int>(d_out)[0]);
-#endif // _CCCL_STD_VER >= 2023
 }
 
 C2H_TEST("ThreadReducePartial does not invoke the reduction operator on invalid elements", "[reduce][thread]")
