diff --git a/csrc/trtllm_moe_allreduce_fusion.cu b/csrc/trtllm_moe_allreduce_fusion.cu
index 78f98b9d13..cec7b3de2e 100644
--- a/csrc/trtllm_moe_allreduce_fusion.cu
+++ b/csrc/trtllm_moe_allreduce_fusion.cu
@@ -81,6 +81,57 @@ void trtllm_moe_allreduce_fusion(
       });
 }
 
+void trtllm_moe_finalize_allreduce_fusion(
+    at::Tensor const& allreduce_in, at::Tensor const& residual_in, at::Tensor const& norm_weight,
+    at::Tensor const& expanded_idx_to_permuted_idx, at::Tensor& norm_out, at::Tensor& residual_out,
+    bool launch_with_pdl, at::Tensor& workspace, int64_t const world_rank, int64_t const world_size,
+    double const eps, std::optional<at::Tensor> const& shared_expert_output,
+    std::optional<at::Tensor> const& expert_scale_factor) {
+  DISPATCH_FLOATING_TYPES_FOR_ALLREDUCE(residual_in.scalar_type(), c_type, [&] {
+    MoeFinalizeAllReduceFusionParams<c_type> params;
+
+    int hidden_dim = residual_in.size(-1);
+    int top_k = expanded_idx_to_permuted_idx.size(-1);
+
+    params.quant_out = nullptr;
+    params.scale_out = nullptr;
+
+    params.nranks = static_cast<int>(world_size);
+    params.rank = static_cast<int>(world_rank);
+    // size: num_token * hidden_dim
+    params.size = residual_in.numel();
+    params.hidden_dim = hidden_dim;
+
+    // workspace: AR scratch space
+    params.workspace = reinterpret_cast<void**>(workspace.mutable_data_ptr());
+    params.rms_gamma = norm_weight.data_ptr();
+    params.rms_eps = static_cast<float>(eps);
+    params.residual_in = residual_in.data_ptr();
+    params.stream = at::cuda::getCurrentCUDAStream(norm_weight.get_device());
+
+    // MOE Reduction specific params
+    params.top_k = top_k;
+    params.allreduce_in = allreduce_in.data_ptr();
+    params.expert_scale_factor =
+        expert_scale_factor.has_value() ? expert_scale_factor.value().data_ptr() : nullptr;
+    TORCH_CHECK(expanded_idx_to_permuted_idx.scalar_type() == at::ScalarType::Int,
+                "expanded_idx_to_permuted_idx must be int32");
+    params.expanded_idx_to_permuted_idx =
+        static_cast<int32_t*>(expanded_idx_to_permuted_idx.data_ptr());
+    params.shared_expert_output =
+        shared_expert_output.has_value() ? shared_expert_output.value().data_ptr() : nullptr;
+
+    // output tensors
+    params.norm_out = norm_out.mutable_data_ptr();
+    params.residual_out = residual_out.mutable_data_ptr();
+
+    auto status = moefinalize_allreduce_fusion_op(params, launch_with_pdl);
+    TORCH_CHECK(status == cudaSuccess, "moefinalize_allreduce_fusion_op failed with error code ",
+                cudaGetErrorString(status));
+  });
+}
+
 TORCH_LIBRARY_FRAGMENT(TORCH_EXTENSION_NAME, m) {
   m.def("trtllm_moe_allreduce_fusion", &trtllm_moe_allreduce_fusion);
+  m.def("trtllm_moe_finalize_allreduce_fusion", &trtllm_moe_finalize_allreduce_fusion);
 }
diff --git a/flashinfer/comm.py b/flashinfer/comm.py
index 8ee06a561a..48ba52f630 100644
--- a/flashinfer/comm.py
+++ b/flashinfer/comm.py
@@ -66,6 +66,7 @@ class AllReduceFusionOp:
     RESIDUAL_RMS_NORM_OUT_QUANT_FP8 = 6
     RESIDUAL_RMS_NORM_OUT_QUANT_NVFP4 = 7
     MOE_ALLREDUCE_RESIDUAL_RMS_NORM = 8
+    MOE_FINALIZE_ALLREDUCE_RESIDUAL_RMS_NORM = 9
 
 
 class AllReduceFusionPattern:
@@ -599,12 +600,48 @@ def trtllm_moe_allreduce_fusion(
             scale_out,
         )
 
+    @register_custom_op(
+        "flashinfer::trtllm_moe_finalize_allreduce_fusion",
+        mutates_args=["residual_out", "norm_out"],
+    )
+    def trtllm_moe_finalize_allreduce_fusion(
+        allreduce_in: torch.Tensor,
+        residual_in: torch.Tensor,
+        norm_weight: torch.Tensor,
+        expanded_idx_to_permuted_idx: torch.Tensor,
+        norm_out: torch.Tensor,
+        residual_out: torch.Tensor,
+        launch_with_pdl: bool,
+        workspace: torch.Tensor,
+        world_rank: int,
+        world_size: int,
+        eps: float,
+        shared_expert_output: Optional[torch.Tensor],
+        expert_scale_factor: Optional[torch.Tensor],
+    ) -> None:
+        module.trtllm_moe_finalize_allreduce_fusion(
+            allreduce_in,
+            residual_in,
+            norm_weight,
+            expanded_idx_to_permuted_idx,
+            norm_out,
+            residual_out,
+            launch_with_pdl,
+            workspace,
+            world_rank,
+            world_size,
+            eps,
+            shared_expert_output,
+            expert_scale_factor,
+        )
+
     return SimpleNamespace(
         trtllm_lamport_initialize=trtllm_lamport_initialize,
         trtllm_lamport_initialize_all=trtllm_lamport_initialize_all,
         trtllm_custom_all_reduce=trtllm_custom_all_reduce,
         trtllm_allreduce_fusion=trtllm_allreduce_fusion,
         trtllm_moe_allreduce_fusion=trtllm_moe_allreduce_fusion,
+        trtllm_moe_finalize_allreduce_fusion=trtllm_moe_finalize_allreduce_fusion,
     )
 
 
@@ -1088,3 +1125,35 @@ def trtllm_moe_allreduce_fusion(
         quant_out=quant_out,
         scale_out=scale_out,
     )
+
+
+def trtllm_moe_finalize_allreduce_fusion(
+    allreduce_in: torch.Tensor,
+    residual_in: torch.Tensor,
+    norm_weight: torch.Tensor,
+    expanded_idx_to_permuted_idx: torch.Tensor,
+    norm_out: torch.Tensor,
+    residual_out: torch.Tensor,
+    workspace_ptrs: torch.Tensor,
+    launch_with_pdl: bool,
+    world_rank: int,
+    world_size: int,
+    eps: float,
+    shared_expert_output: Optional[torch.Tensor],
+    expert_scale_factor: Optional[torch.Tensor],
+) -> None:
+    get_trtllm_comm_module().trtllm_moe_finalize_allreduce_fusion(
+        allreduce_in=allreduce_in,
+        residual_in=residual_in,
+        norm_weight=norm_weight,
+        expanded_idx_to_permuted_idx=expanded_idx_to_permuted_idx,
+        norm_out=norm_out,
+        residual_out=residual_out,
+        workspace=workspace_ptrs,
+        launch_with_pdl=launch_with_pdl,
+        world_rank=world_rank,
+        world_size=world_size,
+        eps=eps,
+        shared_expert_output=shared_expert_output,
+        expert_scale_factor=expert_scale_factor,
+    )
diff --git a/include/flashinfer/comm/trtllm_allreduce.cuh b/include/flashinfer/comm/trtllm_allreduce.cuh
index 4e69930fd4..7256638504 100644
--- a/include/flashinfer/comm/trtllm_allreduce.cuh
+++ b/include/flashinfer/comm/trtllm_allreduce.cuh
@@ -74,6 +74,7 @@ enum class AllReduceFusionOp : int8_t {
   RESIDUAL_RMS_NORM_OUT_QUANT_FP8 = 6,
   RESIDUAL_RMS_NORM_OUT_QUANT_NVFP4 = 7,
   MOE_ALLREDUCE_RESIDUAL_RMS_NORM = 8,
+  MOE_FINALIZE_ALLREDUCE_RESIDUAL_RMS_NORM = 9,
 };
 
 template <typename T>
diff --git a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
index 94676cac05..3724575bce 100644
--- a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
+++ b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
@@ -641,6 +641,7 @@ struct AllReduceFusionParams {
   int size;
   int hidden_dim;
   void** workspace;
+  void* allreduce_in;
   void* allreduce_out;
   void* residual_in;
   void* residual_out;
@@ -653,6 +654,10 @@ struct AllReduceFusionParams {
   float scale_factor;
   FP4QuantizationSFLayout layout = FP4QuantizationSFLayout::SWIZZLED;
   cudaStream_t stream;
+
+  // moe-allreduce output (non-fused)
+  // might be used in MoeReductionAllReduceFusionParams
+  void* moe_allreduce_out = nullptr;
 };
 
 template <typename T>
@@ -668,9 +673,20 @@ struct MoeReductionAllReduceFusionParams : public AllReduceFusionParams<T> {
   void* moe_reduction_active_experts_token_input = nullptr;
   // per token input
   void* moe_reduction_token_input = nullptr;
+};
 
-  // moe-allreduce output (non-fused)
-  void* moe_allreduce_out = nullptr;
+template <typename T>
+struct MoeFinalizeAllReduceFusionParams : public AllReduceFusionParams<T> {
+  // * moe reduction specific params
+  // Refer to kernel implementation on layout of those params
+  // number of active experts on current device
+  int top_k;
+  // [num_tokens, top_k]
+  void* expert_scale_factor = nullptr;
+  void* shared_expert_output = nullptr;
+  // [num_tokens, top_k]
+  int32_t* expanded_idx_to_permuted_idx = nullptr;
+  // allreduce_in [maxPermutedPaddedCount, hidden_dim]
 };
 
 template <int NRanks>
@@ -767,8 +783,7 @@ __device__ __forceinline__ vec_t<T, VEC_SIZE> rms_norm(vec_t<T, VEC_SIZE> const&
 template <bool AllReduceOut, bool ResidualOut, bool NormOut, bool QuantOut, typename T,
           uint32_t VEC_SIZE>
 __device__ __forceinline__ void fused_op(vec_t<T, VEC_SIZE> const& val, int access_id, int token_id,
-                                         int access_id_in_token,
-                                         MoeReductionAllReduceFusionParams<T>& params) {
+                                         int access_id_in_token, AllReduceFusionParams<T>& params) {
   if constexpr (AllReduceOut) {
     val.store(reinterpret_cast<T*>(params.moe_allreduce_out) + access_id * VEC_SIZE);
   }
@@ -1190,5 +1205,264 @@ cudaError_t moereduction_allreduce_fusion_op(MoeReductionAllReduceFusionParams<T
       });
   return status;
 }
+
+/////////////////////////////////////////////////////////////////
+//                  * MoE Finalize Allreduce Fusion *                   //
+/////////////////////////////////////////////////////////////////
+
+template <typename T, int NRanks, bool ResidualOut, bool NormOut, bool QuantOut,
+          typename ScaleType = T>
+__global__ void moefinalize_allreduce_fusion_kernel_oneshot_lamport(
+    MoeFinalizeAllReduceFusionParams<T> params) {
+#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ >= 900))
+  namespace cg = cooperative_groups;
+  cg::cluster_group cluster = cg::this_cluster();
+  cg::grid_group grid = cg::this_grid();
+
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+
+  // Each token is handled by one cluster
+  // which token is handled by current cluster
+  int token_id = grid.cluster_rank();
+  // total number of token
+  int num_token = params.size / params.hidden_dim;
+  // Each thread handle VEC_SIZE num elem in token. Total cluster.num_threads() to handle one
+  // token For current token, which VEC_SIZE is handled by current thread (in unit of
+  // VEC_SIZE)
+  int access_id_in_token = cluster.thread_rank();
+  // Across all token, which VEC_SIZE is handled by current thread (in unit of
+  // VEC_SIZE)
+  int access_id = token_id * params.hidden_dim / VEC_SIZE + access_id_in_token;
+  // Persistent kernel
+  // stride to next token handled by current cta
+  int token_stride = grid.num_clusters();
+  // stride in unit of VEC_SIZE
+  int access_stride = token_stride * params.hidden_dim / VEC_SIZE;
+  // Total number of access in unit of VEC_SIZE to handle (token_num * hidden_dim)
+  // This is within one rank
+  int tot_access = params.size / VEC_SIZE;
+  vec_t<T, VEC_SIZE> clear_vec;
+  clear_vec.fill(neg_zero_v<T>);
+
+  cudaGridDependencySynchronize();
+  LamportComm<NRanks> comm(params.workspace, params.rank);
+  int clear_access = comm.clear_size / VEC_SIZE;
+
+  // * MoE related
+  int threadid_in_cluster = cluster.thread_rank();
+  // Start Offset within one token's hidden_size of element
+  // Current thread handle token[thread_offset_within_token : thread_offset_within_token +
+  // VEC_SIZE]
+  int thread_offset_within_token = threadid_in_cluster * VEC_SIZE;
+
+  int top_k = params.top_k;
+  bool use_scale_factor = params.expert_scale_factor != nullptr;
+
+  // Persistent Kernel
+  // Each cluster iterate through all token it need to handle
+  for (int token_id = grid.cluster_rank(); token_id < num_token; token_id += grid.num_clusters()) {
+    if (thread_offset_within_token >= params.hidden_dim) {
+      break;
+    }
+
+    // * MoE finalize
+    vec_t<T, VEC_SIZE> accumulator;
+    accumulator.fill(0.f);
+
+    for (int k = 0; k < top_k; k++) {
+      int const expanded_idx = token_id * top_k + k;
+      int32_t const permuted_idx = params.expanded_idx_to_permuted_idx[expanded_idx];
+
+      if (permuted_idx == -1) continue;
+
+      int thread_offset_across_token =
+          permuted_idx * params.hidden_dim + thread_offset_within_token;
+      float block_scale = 1.0;
+      if (use_scale_factor) {
+        block_scale =
+            static_cast<float>(static_cast<ScaleType*>(params.expert_scale_factor)[expanded_idx]);
+      }
+
+      vec_t<T, VEC_SIZE> permuted_data;
+      permuted_data.load(reinterpret_cast<T*>(params.allreduce_in) + thread_offset_across_token);
+
+      // * acc += scale(data)
+#pragma unroll
+      for (int i = 0; i < VEC_SIZE; ++i) {
+        // assume computation is done in ScaleType
+        accumulator[i] += static_cast<T>(static_cast<float>(permuted_data[i]) * block_scale);
+      }
+    }
+
+    // * Add shared expert output
+    if (params.shared_expert_output) {
+      // * Load shared expert output
+      int thread_offset_across_token = token_id * params.hidden_dim + thread_offset_within_token;
+      vec_t<T, VEC_SIZE> shared_expert_output;
+      shared_expert_output.load(reinterpret_cast<T*>(params.shared_expert_output) +
+                                thread_offset_across_token);
+#pragma unroll
+      accumulator = vec_add<T, VEC_SIZE>(accumulator, shared_expert_output);
+    }
+
+    // * AR Store
+    int idx = token_id * params.hidden_dim / VEC_SIZE + access_id_in_token;
+    remove_neg_zero<T, VEC_SIZE>(accumulator);
+
+#pragma unroll
+    for (int r = 0; r < NRanks; ++r) {
+      // STG.128 to remote rank
+      int offset = (params.rank * tot_access + idx) * VEC_SIZE;
+      accumulator.store_global_volatile(reinterpret_cast<T*>(comm.data_bufs[r]) + offset);
+    }
+  }
+
+  // * Clear previous buffer
+  for (int idx = access_id; idx < clear_access; idx += access_stride) {
+    clear_vec.store(reinterpret_cast<T*>(comm.clear_buf) + idx * VEC_SIZE);
+  }
+
+  // * AR Load + Fusion
+  for (int idx = access_id, tidx = token_id; idx < tot_access;
+       idx += access_stride, tidx += token_stride) {
+    // * AR Load
+    vec_t<T, VEC_SIZE> vals[NRanks];
+    bool done = false;
+    while (!done) {
+      done = true;
+#pragma unroll
+      for (int r = 0; r < NRanks; ++r) {
+        // LDG.128 from local rank
+        vals[r].load_global_volatile(reinterpret_cast<T*>(comm.data_bufs[r]) +
+                                     (r * tot_access + idx) * VEC_SIZE);
+        done &= !has_neg_zero<T, VEC_SIZE>(vals[r]);
+      }
+    }
+    vec_t<T, VEC_SIZE> sum_val = vals[0];
+#pragma unroll
+    for (int r = 1; r < NRanks; ++r) {
+      sum_val = vec_add<T, VEC_SIZE>(sum_val, vals[r]);
+    }
+
+    // * Fuse: AllReduceOut is always false in finalize_moe_allreduce
+    fused_op<false, ResidualOut, NormOut, QuantOut, T, VEC_SIZE>(sum_val, idx, tidx,
+                                                                 access_id_in_token, params);
+  }
+  comm.update(params.size * NRanks);
+  cudaTriggerProgrammaticLaunchCompletion();
+#endif
+}
+
+template <typename T, int NRanks, bool ResidualOut, bool NormOut, bool QuantOut,
+          typename ScaleType = T>
+cudaError_t launch_oneshot_moefinalize_lamport(MoeFinalizeAllReduceFusionParams<T> const& params,
+                                               cudaLaunchConfig_t& cfg) {
+  FLASHINFER_CUDA_CALL(cudaLaunchKernelEx(
+      &cfg,
+      moefinalize_allreduce_fusion_kernel_oneshot_lamport<T, NRanks, ResidualOut, NormOut, QuantOut,
+                                                          ScaleType>,
+      params));
+  return cudaSuccess;
+}
+
+template <typename T, int NRanks, bool ResidualOut, bool NormOut, bool QuantOut,
+          typename ScaleType = T>
+cudaError_t moefinalize_allreduce_fusion_kernel_launcher(
+    MoeFinalizeAllReduceFusionParams<T> const& params, bool launch_with_pdl) {
+  int token_num = params.size / params.hidden_dim;
+  bool oneshot = use_oneshot(token_num);
+  if (oneshot == false) {
+    FLASHINFER_LOG_WARN("expect one shot but got %d tokens, expect performance degradation",
+                        token_num);
+    oneshot = true;
+  }
+  // Only support one shot
+  // FLASHINFER_CHECK(oneshot, "only support one shot");
+  // Each token is handled by one cluster
+  int cluster_num = token_num;
+  // Total number of threads (within one cluster) that's need to handle one token
+  // given that each thread handle VEC_SIZE
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+  int threads_per_token = params.hidden_dim / VEC_SIZE;
+  // Total number of warp (within one cluster) that's need to handle one token
+  // given that each thread handle VEC_SIZE
+  int warps_per_token = (threads_per_token + 31) / 32;
+  int cluster_size = 8;
+  while (warps_per_token % cluster_size != 0) {
+    cluster_size /= 2;
+  }
+  int block_size = warps_per_token / cluster_size * 32;
+  FLASHINFER_CHECK(block_size <= 1024 && cluster_size > 0,
+                   "block_size <= 1024 && cluster_size > 0");
+  int sm_count = get_sm_count();
+  int grid_size = (std::min(sm_count, cluster_num * cluster_size) / cluster_size) * cluster_size;
+  cudaLaunchConfig_t cfg;
+  cudaLaunchAttribute attribute[2];
+  cfg.gridDim = grid_size;
+  cfg.blockDim = block_size;
+  cfg.dynamicSmemBytes = 0;
+  cfg.stream = params.stream;
+  attribute[0].id = cudaLaunchAttributeProgrammaticStreamSerialization;
+  attribute[0].val.programmaticStreamSerializationAllowed = launch_with_pdl ? 1 : 0;
+  attribute[1].id = cudaLaunchAttributeClusterDimension;
+  attribute[1].val.clusterDim.x = cluster_size;
+  attribute[1].val.clusterDim.y = 1;
+  attribute[1].val.clusterDim.z = 1;
+  cfg.attrs = attribute;
+  cfg.numAttrs = 2;
+  if (oneshot) {
+    FLASHINFER_CUDA_CALL(
+        (launch_oneshot_moefinalize_lamport<T, NRanks, ResidualOut, NormOut, QuantOut, ScaleType>(
+            params, cfg)));
+  }
+  return cudaSuccess;
+}
+
+#define _DISPATCH_MOEFINALIZEREDUCTION_CASE(n_ranks_val, N_RANKS_VAR, res, rms, quant, RES, RMS, \
+                                            QUANT, ...)                                          \
+  case n_ranks_val: {                                                                            \
+    constexpr int N_RANKS_VAR = n_ranks_val;                                                     \
+    return DISPATCH_BOOL_(res, RES, [&]() -> cudaError_t {                                       \
+      return DISPATCH_BOOL_(rms, RMS, [&]() -> cudaError_t {                                     \
+        return DISPATCH_BOOL_(quant, QUANT, [&]() -> cudaError_t { return __VA_ARGS__(); });     \
+      });                                                                                        \
+    });                                                                                          \
+  }
+
+#define DISPATCH_MOEFINALIZEREDUCTION(n_ranks, res, rms, quant, N_RANKS, RES, RMS, QUANT, ...) \
+  [&]() -> cudaError_t {                                                                       \
+    switch (n_ranks) {                                                                         \
+      _DISPATCH_MOEFINALIZEREDUCTION_CASE(2, N_RANKS, res, rms, quant, RES, RMS, QUANT,        \
+                                          __VA_ARGS__)                                         \
+      _DISPATCH_MOEFINALIZEREDUCTION_CASE(4, N_RANKS, res, rms, quant, RES, RMS, QUANT,        \
+                                          __VA_ARGS__)                                         \
+      _DISPATCH_MOEFINALIZEREDUCTION_CASE(8, N_RANKS, res, rms, quant, RES, RMS, QUANT,        \
+                                          __VA_ARGS__)                                         \
+      _DISPATCH_MOEFINALIZEREDUCTION_CASE(16, N_RANKS, res, rms, quant, RES, RMS, QUANT,       \
+                                          __VA_ARGS__)                                         \
+      default:                                                                                 \
+        FLASHINFER_CHECK(false, "Unsupported n_ranks");                                        \
+        return cudaErrorNotSupported;                                                          \
+    }                                                                                          \
+  }()
+
+template <typename T>
+cudaError_t moefinalize_allreduce_fusion_op(MoeFinalizeAllReduceFusionParams<T> const& params,
+                                            bool launch_with_pdl) {
+  static constexpr int VEC_SIZE = details::kBytesPerAccess / sizeof(T);
+  FLASHINFER_CHECK(params.allreduce_in && params.expanded_idx_to_permuted_idx && params.top_k,
+                   "allreduce_in, expanded_idx_to_permuted_idx and top_k must be set");
+  FLASHINFER_CHECK(params.size % params.hidden_dim == 0, "size must be a multiple of hidden_dim");
+  FLASHINFER_CHECK(params.hidden_dim % VEC_SIZE == 0, "hidden_dim must be a multiple of VEC_SIZE");
+
+  auto status = DISPATCH_MOEFINALIZEREDUCTION(
+      params.nranks, params.residual_out, params.rms_gamma, params.quant_out, N_RANKS, RES, RMS,
+      QUANT, [&]() -> cudaError_t {
+        FLASHINFER_CUDA_CALL(
+            (moefinalize_allreduce_fusion_kernel_launcher<T, N_RANKS, RES, RMS, QUANT>(
+                (params), (launch_with_pdl))));
+      });
+  return status;
+}
 }  // namespace trtllm_moe_allreduce_fusion
 }  // namespace flashinfer
diff --git a/tests/test_trtllm_moe_allreduce_fusion_finalize.py b/tests/test_trtllm_moe_allreduce_fusion_finalize.py
new file mode 100644
index 0000000000..036cdd6b11
--- /dev/null
+++ b/tests/test_trtllm_moe_allreduce_fusion_finalize.py
@@ -0,0 +1,288 @@
+import multiprocessing as mp
+import socket
+from typing import Any, Optional, Tuple, Union
+
+import numpy as np
+import pytest
+import torch
+import torch.distributed as dist
+from torch import nn
+
+import flashinfer.comm as comm
+
+# todo(Yingyi): add benchmark and quant test
+
+# Usage: test var
+kOneShotMaxTokenNum = 128
+MAX_TOKEN_NUM = 2048
+HIDDEN_SIZE = 7168
+MAX_EXPERT_NUM = 16
+SF_VEC_SIZE = 16
+
+# temp var
+SCALE_FACTOR_RANGE = (-1, 1)
+
+
+def _run_correctness_worker(
+    world_size,
+    rank,
+    dtype,
+    distributed_init_port,
+    shared_expert_output,
+    fc2_output,
+    scale,
+    expanded_idx_to_permuted_idx,
+    residual,
+):
+
+    def rms_norm(x: torch.Tensor, weight: torch.Tensor = None, eps: float = 1e-6):
+        y = x * torch.rsqrt(x.pow(2).mean(-1, keepdim=True) + eps)
+        if weight is not None:
+            y = y * weight
+        return y
+
+    device = torch.device(f"cuda:{rank}")
+    torch.cuda.set_device(device)
+    distributed_init_method = f"tcp://localhost:{distributed_init_port}"
+    dist.init_process_group(
+        backend="nccl",
+        init_method=distributed_init_method,
+        rank=rank,
+        world_size=world_size,
+    )
+    group = dist.group.WORLD
+
+    try:
+        device = torch.device(f"cuda:{rank}")
+        seq_lens = [16]
+        top_k = 8
+        eps = 1e-5
+
+        launch_with_pdls = [True, False]
+
+        # create workspace for moe allreduce fusion
+        ipc_handles, workspace_tensor = (
+            comm.trtllm_create_ipc_workspace_for_all_reduce_fusion(
+                rank, world_size, MAX_TOKEN_NUM, HIDDEN_SIZE, group=group
+            )
+        )
+
+        test_loop = 5
+
+        for seq_len in seq_lens:
+            for launch_with_pdl in launch_with_pdls:
+                dist.barrier(group=group)
+                test_passed = True
+                print(
+                    f"test RANK {rank}: seq_len{seq_len}-topk{top_k}-tp{world_size}-{dtype}-pdl{launch_with_pdl} start"
+                )
+                dist.barrier(group=group)
+                torch.cuda.synchronize()
+                for _ in range(test_loop):
+                    # == Generate input ==
+                    # move to local device
+                    shared_expert_output = shared_expert_output.to(device)
+                    fc2_output = fc2_output.to(device)
+                    scale = scale.to(device)
+                    expanded_idx_to_permuted_idx = expanded_idx_to_permuted_idx.to(
+                        device
+                    )
+                    residual = residual.to(device)
+
+                    # make clone
+                    fc2_output_clone = fc2_output.clone()
+                    scale_clone = scale.clone()
+                    expanded_idx_to_permuted_idx_clone = (
+                        expanded_idx_to_permuted_idx.clone()
+                    )
+                    norm_weight = torch.randn(
+                        (HIDDEN_SIZE,), dtype=dtype, device=device
+                    )
+
+                    norm_out = torch.empty_like(residual)
+                    residual_out = torch.empty_like(residual)
+
+                    # == Run kernel ==
+                    torch.cuda.synchronize()
+                    comm.trtllm_moe_finalize_allreduce_fusion(
+                        allreduce_in=fc2_output,
+                        residual_in=residual,
+                        norm_weight=norm_weight,
+                        expanded_idx_to_permuted_idx=expanded_idx_to_permuted_idx,
+                        workspace_ptrs=workspace_tensor,
+                        launch_with_pdl=launch_with_pdl,
+                        world_rank=rank,
+                        world_size=world_size,
+                        eps=eps,
+                        shared_expert_output=shared_expert_output,
+                        expert_scale_factor=scale,
+                        norm_out=norm_out,
+                        residual_out=residual_out,
+                    )
+                    torch.cuda.synchronize()
+
+                    # == Calculate reference output ==
+                    expert_reduction = torch.sum(
+                        fc2_output_clone[expanded_idx_to_permuted_idx]
+                        * scale.unsqueeze(-1),
+                        dim=1,
+                    )
+
+                    torch_before_residual = (
+                        expert_reduction + shared_expert_output
+                    ) * world_size
+                    torch_residual = torch_before_residual + residual
+                    torch_residual = torch_residual.to(torch.float32)
+                    torch_output_hidden_states = rms_norm(
+                        torch_residual, norm_weight, eps
+                    ).to(dtype)
+
+                    # == Check correctness ==
+                    if not torch.allclose(
+                        residual_out.to(torch.float32),
+                        torch_residual.to(torch.float32),
+                        rtol=0.2,
+                        atol=0.2,
+                    ):
+                        test_passed = False
+                        print(f"Rank {rank} residual_out mismatch")
+                        print(f"residual_out: {residual_out}")
+                        print(f"torch_residual: {torch_residual}")
+                        print(
+                            f"max diff: {torch.max(torch.abs(residual_out.to(torch.float32) - torch_residual.to(torch.float32)))}"
+                        )
+                        print(
+                            f"max diff idx: {torch.argmax(torch.abs(residual_out.to(torch.float32) - torch_residual.to(torch.float32)))}"
+                        )
+                        print(
+                            f"max diff value: {residual_out.to(torch.float32).view(-1)[torch.argmax(torch.abs(residual_out.to(torch.float32) - torch_residual.to(torch.float32)))]}"
+                        )
+                        print(
+                            f"max diff ref value: {torch_residual.to(torch.float32).view(-1)[torch.argmax(torch.abs(residual_out.to(torch.float32) - torch_residual.to(torch.float32)))]}"
+                        )
+
+                    if not torch.allclose(
+                        norm_out.to(torch.float32),
+                        torch_output_hidden_states.to(torch.float32),
+                        rtol=0.2,
+                        atol=0.2,
+                    ):
+                        test_passed = False
+                        print(f"Rank {rank} norm_out mismatch")
+                        print(f"norm_out: {norm_out}")
+                        print(
+                            f"torch_output_hidden_states: {torch_output_hidden_states}"
+                        )
+                        print(
+                            f"max diff: {torch.max(torch.abs(norm_out.to(torch.float32) - torch_output_hidden_states.to(torch.float32)))}"
+                        )
+                        print(
+                            f"max diff idx: {torch.argmax(torch.abs(norm_out.to(torch.float32) - torch_output_hidden_states.to(torch.float32)))}"
+                        )
+                        print(
+                            f"max diff value: {norm_out.to(torch.float32).view(-1)[torch.argmax(torch.abs(norm_out.to(torch.float32) - torch_output_hidden_states.to(torch.float32)))]}"
+                        )
+                        print(
+                            f"max diff ref value: {torch_output_hidden_states.to(torch.float32).view(-1)[torch.argmax(torch.abs(norm_out.to(torch.float32) - torch_output_hidden_states.to(torch.float32)))]}"
+                        )
+
+                    torch.testing.assert_close(
+                        residual_out.to(torch.float32),
+                        torch_residual.to(torch.float32),
+                        rtol=0.2,
+                        atol=0.2,
+                    )
+                    torch.testing.assert_close(
+                        norm_out.to(torch.float32),
+                        torch_output_hidden_states.to(torch.float32),
+                        rtol=0.2,
+                        atol=0.2,
+                    )
+
+                dist.barrier(group=group)
+                if test_passed:
+                    print(
+                        f"test RANK {rank}: seq_len{seq_len}-topk{top_k}-tp{world_size}-{dtype}-pdl{launch_with_pdl} passed"
+                    )
+                else:
+                    print(
+                        f"test RANK {rank}: seq_len{seq_len}-topk{top_k}-tp{world_size}-{dtype}-pdl{launch_with_pdl} failed"
+                    )
+    finally:
+        dist.barrier(group=group)
+
+        comm.trtllm_destroy_ipc_workspace_for_all_reduce(ipc_handles, group=group)
+
+        dist.destroy_process_group(group=group)
+
+
+def get_open_port() -> int:
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("127.0.0.1", 0))
+            return s.getsockname()[1]
+    except OSError:
+        with socket.socket(socket.AF_INET6, socket.SOCK_STREAM) as s:
+            s.bind(("::1", 0))
+            return s.getsockname()[1]
+
+
+def multi_process_parallel(
+    world_size: int, dtype: torch.dtype, test_target: Any, target_args: tuple = ()
+) -> None:
+    mp.set_start_method("spawn", force=True)
+
+    procs = []
+    distributed_init_port = get_open_port()
+    for i in range(world_size):
+        proc_args = (world_size, i, dtype, distributed_init_port) + target_args
+        proc = mp.Process(target=test_target, args=proc_args, name=f"Worker-{i}")
+        proc.start()
+        procs.append(proc)
+
+    for i in range(world_size):
+        procs[i].join()
+        assert (
+            procs[i].exitcode == 0
+        ), f"Process {i} failed with exit code {procs[i].exitcode}"
+
+
+@pytest.mark.parametrize("world_size", [2, 4])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
+def test_trtllm_moe_finalize_allreduce_fusion(world_size, dtype):
+    np.random.seed(42)
+    torch.manual_seed(42)
+    torch.cuda.manual_seed_all(42)
+    available_gpus = torch.cuda.device_count()
+    if world_size > available_gpus:
+        raise ValueError(
+            f"world_size {world_size} is greater than available_gpus {available_gpus}"
+        )
+    print(f"Running test for world_size={world_size}")
+
+    # generate shared random input tensor across all ranks
+    seq_len = 16
+    hidden_size = 7168
+    top_k = 8
+
+    shared_expert_output = torch.randn((seq_len, hidden_size), dtype=dtype)
+    fc2_output = torch.randn((seq_len * top_k, hidden_size), dtype=dtype)
+    scale = torch.randn((seq_len, top_k), dtype=dtype)
+    expanded_idx_to_permuted_idx = torch.randint(
+        0, seq_len * top_k, (seq_len, top_k), dtype=torch.int32
+    )
+    residual = torch.randn_like(shared_expert_output)
+
+    multi_process_parallel(
+        world_size,
+        dtype,
+        _run_correctness_worker,
+        target_args=(
+            shared_expert_output,
+            fc2_output,
+            scale,
+            expanded_idx_to_permuted_idx,
+            residual,
+        ),
+    )
+    print(f"moe finalize allreduce fusion tp = {world_size}: OK")
