diff --git a/cub/benchmarks/bench/transform/common.h b/cub/benchmarks/bench/transform/common.h
index 0260aa884be..90547fc2dd6 100644
--- a/cub/benchmarks/bench/transform/common.h
+++ b/cub/benchmarks/bench/transform/common.h
@@ -26,7 +26,8 @@
 #include <nvbench_helper.cuh>
 
 #if !TUNE_BASE
-struct policy_selector
+// TODO(bgruber): can we get by without the base class?
+struct policy_selector : cub::detail::transform::tuning<policy_selector>
 {
   _CCCL_API constexpr auto operator()(cuda::arch_id) const -> cub::detail::transform::transform_policy
   {
@@ -83,17 +84,16 @@ void bench_transform(nvbench::state& state,
                      TransformOp transform_op)
 {
   state.exec(nvbench::exec_tag::gpu | nvbench::exec_tag::no_batch, [&](const nvbench::launch& launch) {
-    cub::detail::transform::dispatch<cub::detail::transform::requires_stable_address::no>(
+    cub::DeviceTransform::Transform(
       inputs,
       output,
       num_items,
-      cub::detail::transform::always_true_predicate{},
       transform_op,
-      launch.get_stream()
+      cuda::std::execution::env{::cuda::stream_ref{launch.get_stream().get_stream()}
 #if !TUNE_BASE
-        ,
-      policy_selector{}
+                                ,
+                                cuda::execution::__tune(policy_selector{})
 #endif // !TUNE_BASE
-    );
+      });
   });
 }
diff --git a/cub/cub/device/device_transform.cuh b/cub/cub/device/device_transform.cuh
index 24a604ca730..74bd0ecdfac 100644
--- a/cub/cub/device/device_transform.cuh
+++ b/cub/cub/device/device_transform.cuh
@@ -48,8 +48,23 @@ struct ::cuda::proclaims_copyable_arguments<CUB_NS_QUALIFIER::detail::__return_c
 CUB_NAMESPACE_BEGIN
 namespace detail::transform
 {
+// TODO(bgruber): can we get by without the tuning base class? Since we have transform_policy_selector, could we enrich
+// get_tuning_query_t to just check if the environment has a type that fulfills transform_policy_selector?
 struct get_tuning_query_t
 {};
+
+template <class PolicySelector>
+// TODO(bgruber): we cannot check the concept here because PolicySelector is usually an incomplete type still
+// #if _CCCL_HAS_CONCEPTS()
+//   requires transform_policy_selector<PolicySelector>
+// #endif // _CCCL_HAS_CONCEPTS()
+struct tuning
+{
+  [[nodiscard]] _CCCL_TRIVIAL_API constexpr auto query(const get_tuning_query_t&) const noexcept -> PolicySelector
+  {
+    return static_cast<const PolicySelector&>(*this);
+  }
+};
 } // namespace detail::transform
 
 //! DeviceTransform provides device-wide, parallel operations for transforming elements tuple-wise from multiple input
@@ -81,31 +96,30 @@ private:
       return error;
     }
 
-    using tuning_env_t =
-      ::cuda::__call_result_or_t<::cuda::execution::__get_tuning_t, ::cuda::std::execution::env<>, Env>;
-    using transform_tuning_t = ::cuda::__call_result_or_t<detail::transform::get_tuning_query_t, int, tuning_env_t>;
+    const auto stream = ::cuda::__call_or(::cuda::get_stream, ::cuda::stream_ref{cudaStream_t{}}, env).get();
 
-    if constexpr (!::cuda::std::is_same_v<transform_tuning_t, int>)
-    {
-      return detail::transform::dispatch<StableAddress>(
-        ::cuda::std::move(inputs),
-        ::cuda::std::move(output),
-        static_cast<offset_t>(num_items),
-        ::cuda::std::move(predicate),
-        ::cuda::std::move(transform_op),
-        ::cuda::__call_or(::cuda::get_stream, ::cuda::stream_ref{cudaStream_t{}}, env).get(),
-        transform_tuning_t{});
-    }
-    else
-    {
-      return detail::transform::dispatch<StableAddress>(
-        ::cuda::std::move(inputs),
-        ::cuda::std::move(output),
-        static_cast<offset_t>(num_items),
-        ::cuda::std::move(predicate),
-        ::cuda::std::move(transform_op),
-        ::cuda::__call_or(::cuda::get_stream, ::cuda::stream_ref{cudaStream_t{}}, env).get());
-    }
+    using tuning_env =
+      ::cuda::std::execution::__query_result_or_t<Env, ::cuda::execution::__get_tuning_t, ::cuda::std::execution::env<>>;
+    using default_policy_selector = detail::transform::policy_selector_from_types<
+      StableAddress == detail::transform::requires_stable_address::yes,
+      ::cuda::std::is_same_v<Predicate, detail::transform::always_true_predicate>,
+      ::cuda::std::tuple<RandomAccessIteratorsIn...>,
+      RandomAccessIteratorOut>;
+    using policy_selector = ::cuda::std::execution::
+      __query_result_or_t<tuning_env, detail::transform::get_tuning_query_t, default_policy_selector>;
+
+#if _CCCL_HAS_CONCEPTS()
+    static_assert(detail::transform::transform_policy_selector<policy_selector>);
+#endif // _CCCL_HAS_CONCEPTS()
+
+    return detail::transform::dispatch<StableAddress>(
+      ::cuda::std::move(inputs),
+      ::cuda::std::move(output),
+      static_cast<offset_t>(num_items),
+      ::cuda::std::move(predicate),
+      ::cuda::std::move(transform_op),
+      stream,
+      policy_selector{});
   }
 
   // TODO(bgruber): we want to eventually forward the output tuple to the kernel and optimize writing multiple streams
diff --git a/cub/cub/device/dispatch/dispatch_transform.cuh b/cub/cub/device/dispatch/dispatch_transform.cuh
index 2173f0c077b..8ab8786f033 100644
--- a/cub/cub/device/dispatch/dispatch_transform.cuh
+++ b/cub/cub/device/dispatch/dispatch_transform.cuh
@@ -531,10 +531,7 @@ template <requires_stable_address StableAddress,
           typename Offset,
           typename Predicate,
           typename TransformOp,
-          typename PolicySelector        = policy_selector_from_types<StableAddress == requires_stable_address::yes,
-                                                                      ::cuda::std::is_same_v<Predicate, always_true_predicate>,
-                                                                      ::cuda::std::tuple<RandomAccessIteratorsIn...>,
-                                                                      RandomAccessIteratorOut>,
+          typename PolicySelector,
           typename KernelSource          = TransformKernelSource<PolicySelector,
                                                                  Offset,
                                                                  ::cuda::std::tuple<RandomAccessIteratorsIn...>,
diff --git a/cub/test/catch2_test_device_transform_env.cu b/cub/test/catch2_test_device_transform_env.cu
index a6e3542662d..f9e3b8c3fb8 100644
--- a/cub/test/catch2_test_device_transform_env.cu
+++ b/cub/test/catch2_test_device_transform_env.cu
@@ -72,6 +72,10 @@ void check_graph_nodes_with_different_streams(F call_cub_api)
   {
     call_cub_api(with_get_stream_method{stream.get()});
   }
+  SECTION("environment with cuda::stream_ref")
+  {
+    call_cub_api(cuda::std::execution::env{cuda::stream_ref{stream}});
+  }
   SECTION("environment with prop with cudaStream_t")
   {
     // MSVC has trouble nesting two aggregate initializations with CTAD
@@ -218,3 +222,51 @@ C2H_TEST("DeviceTransform::TransformStableArgumentAddresses custom stream", "[de
 
   CHECK(thrust::equal(result.begin(), result.end(), cuda::counting_iterator<type>{42 + 13}));
 }
+
+// use a policy selector that prescribes to run with exactly 8 threads per block and 3 items per thread
+// TODO(bgruber): can we get by without the base class?
+struct my_policy_selector : cub::detail::transform::tuning<my_policy_selector>
+{
+  _CCCL_API constexpr auto operator()(cuda::arch_id) const -> cub::detail::transform::transform_policy
+  {
+    constexpr int min_bytes_in_flight = 64 * 1024;
+    constexpr auto algorithm          = cub::detail::transform::Algorithm::prefetch;
+    constexpr auto policy             = cub::detail::transform::prefetch_policy{8, 3, 3, 3};
+    return {min_bytes_in_flight, algorithm, policy, {}, {}};
+  }
+};
+
+struct get_thread_id
+{
+  _CCCL_DEVICE auto operator()() const -> unsigned
+  {
+    return threadIdx.x;
+  }
+};
+
+C2H_TEST("DeviceTransform::Transform can be tuned", "[reduce][device]")
+{
+  c2h::device_vector<unsigned> result(3 * 8, thrust::no_init);
+
+  auto env = cuda::execution::__tune(my_policy_selector{});
+  REQUIRE(cudaSuccess
+          == cub::DeviceTransform::Transform(cuda::std::tuple{}, result.data(), result.size(), get_thread_id{}, env));
+  REQUIRE(cudaSuccess == cudaDeviceSynchronize());
+
+  c2h::device_vector<unsigned> expected{0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7};
+  REQUIRE(result == expected);
+}
+
+C2H_TEST("DeviceTransform::Transform can be tuned with custom stream", "[reduce][device]")
+{
+  c2h::device_vector<unsigned> result(3 * 8, thrust::no_init);
+
+  cuda::stream stream{cuda::devices[0]};
+  auto env = cuda::std::execution::env{cuda::stream_ref{stream}, cuda::execution::__tune(my_policy_selector{})};
+  REQUIRE(cudaSuccess
+          == cub::DeviceTransform::Transform(cuda::std::tuple{}, result.data(), result.size(), get_thread_id{}, env));
+  stream.sync();
+
+  c2h::device_vector<unsigned> expected{0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7, 0, 1, 2, 3, 4, 5, 6, 7};
+  REQUIRE(result == expected);
+}
