diff --git a/csrc/nv_internal/tensorrt_llm/cutlass_extensions/include/cutlass_extensions/gemm_configs.h b/csrc/nv_internal/tensorrt_llm/cutlass_extensions/include/cutlass_extensions/gemm_configs.h
index 316f25470a..7fe6e1f9e7 100644
--- a/csrc/nv_internal/tensorrt_llm/cutlass_extensions/include/cutlass_extensions/gemm_configs.h
+++ b/csrc/nv_internal/tensorrt_llm/cutlass_extensions/include/cutlass_extensions/gemm_configs.h
@@ -98,8 +98,9 @@ enum class CutlassTileConfigSM90 {
   CtaShape128x128x128B,
   CtaShape128x256x128B,
 
-  // CTA configs for M=128
+  // CTA configs for M=256
   CtaShape256x128x128B,
+  CtaShape256x256x128B,
 };
 
 enum class CutlassTileConfigSM100 {
diff --git a/flashinfer/aot.py b/flashinfer/aot.py
index 395d94ae37..afc7cdc389 100644
--- a/flashinfer/aot.py
+++ b/flashinfer/aot.py
@@ -12,7 +12,7 @@
 from .cascade import gen_cascade_module
 from .comm import gen_trtllm_comm_module, gen_vllm_comm_module
 from .fp4_quantization import gen_fp4_quantization_sm100_module
-from .fused_moe import gen_fused_moe_sm100_module
+from .fused_moe import gen_cutlass_fused_moe_sm100_module
 from .gemm import gen_gemm_module, gen_gemm_sm90_module, gen_gemm_sm100_module
 from .jit import JitSpec, build_jit_specs
 from .jit import env as jit_env
@@ -322,7 +322,7 @@ def gen_all_modules(
     if has_sm90:
         jit_specs.append(gen_gemm_sm90_module())
     if has_sm100:
-        jit_specs.append(gen_fused_moe_sm100_module())
+        jit_specs.append(gen_cutlass_fused_moe_sm100_module())
         jit_specs.append(gen_fp4_quantization_sm100_module())
         jit_specs.append(gen_gemm_sm100_module())
         jit_specs.append(gen_trtllm_comm_module())
diff --git a/flashinfer/fused_moe.py b/flashinfer/fused_moe.py
index 5789642701..80b2a76147 100644
--- a/flashinfer/fused_moe.py
+++ b/flashinfer/fused_moe.py
@@ -211,7 +211,7 @@ def shuffle_matrix_sf_a(
     return block_scale_interleave(w_shuffled)
 
 
-def gen_fused_moe_sm100_module() -> JitSpec:
+def gen_cutlass_fused_moe_sm100_module(use_fast_build: bool) -> JitSpec:
     return gen_jit_spec(
         "fused_moe_sm100",
         [
@@ -278,9 +278,7 @@ def gen_fused_moe_sm100_module() -> JitSpec:
             "-DCOMPILE_BLACKWELL_TMA_GROUPED_GEMMS",
             "-DCOMPILE_HOPPER_TMA_GEMMS",
         ],
-        extra_cflags=[
-            "-DFAST_BUILD",
-        ],
+        extra_cflags=["-DFAST_BUILD"] if use_fast_build else [],
         extra_ldflags=["-lcuda"],
         extra_include_paths=[
             jit_env.FLASHINFER_CSRC_DIR / "nv_internal",
@@ -306,8 +304,10 @@ def gen_fused_moe_sm100_module() -> JitSpec:
 
 
 @functools.cache
-def get_fused_moe_sm100_module():
-    module = gen_fused_moe_sm100_module().build_and_load(class_name="FusedMoeRunner")
+def get_cutlass_fused_moe_sm100_module(use_fast_build: bool = False):
+    module = gen_cutlass_fused_moe_sm100_module(use_fast_build).build_and_load(
+        class_name="FusedMoeRunner"
+    )
 
     class MoERunner(TunableRunner):
         # avoid overhead of creating a new runner in forward pass
@@ -669,7 +669,7 @@ def cutlass_fused_moe(
             output, output_shape, output_dtype, input.device, "output"
         )
 
-    return get_fused_moe_sm100_module().cutlass_fused_moe_sm100(
+    return get_cutlass_fused_moe_sm100_module().cutlass_fused_moe_sm100(
         output,
         input,
         token_selected_experts,
