diff --git a/test/inductor/test_gpu_cpp_wrapper.py b/test/inductor/test_gpu_cpp_wrapper.py
index dd6bc7fe9ec44..bad8817c9d7f2 100644
--- a/test/inductor/test_gpu_cpp_wrapper.py
+++ b/test/inductor/test_gpu_cpp_wrapper.py
@@ -227,6 +227,8 @@ def test_fn(x, s):
     def test_many_args_fold_expression_nesting(self):
         if not RUN_GPU:
             self.skipTest("GPU not available")
+        if GPU_TYPE == "xpu":
+            self.skipTest("ocloc backend compiler crashes with too many kernel args")
 
         num_params = 130
         params = [torch.randn(64, device=self.device) for _ in range(num_params)]
diff --git a/torch/csrc/inductor/cpp_wrapper/xpu.h b/torch/csrc/inductor/cpp_wrapper/xpu.h
index e26dea0f3b6e2..160788dfa1bab 100644
--- a/torch/csrc/inductor/cpp_wrapper/xpu.h
+++ b/torch/csrc/inductor/cpp_wrapper/xpu.h
@@ -2,3 +2,4 @@
 
 #include <torch/csrc/inductor/cpp_wrapper/common.h>
 #include <torch/csrc/inductor/cpp_wrapper/device_internal/xpu.h>
+#include <torch/csrc/inductor/cpp_wrapper/lazy_triton_compile.h>
