diff --git a/flashinfer/profiler/__init__.py b/flashinfer/profiler/__init__.py
index 63479bb4c0..2b32787972 100644
--- a/flashinfer/profiler/__init__.py
+++ b/flashinfer/profiler/__init__.py
@@ -32,15 +32,21 @@ class EventType(Enum):
 
 
 def decode_tag(tag, num_blocks, num_groups):
-    block_group_tag = tag >> 12
+    """
+    Decode a profiler tag into (block_idx, group_idx, event_idx, event_type, sm_id).
+    Tag layout:
+      bits 0-1: event_type
+      bits 2-11: event_idx
+      bits 12-23: block_group_idx
+      bits 24-31: sm_id
+    """
+    sm_id = (tag >> 24) & 0xFF
+    block_group_idx = (tag >> 12) & 0xFFF
     event_idx = (tag >> 2) & 0x3FF
     event_type = tag & 0x3
-    return (
-        block_group_tag // num_groups,
-        block_group_tag % num_groups,
-        event_idx,
-        event_type,
-    )
+    block_idx = block_group_idx // num_groups
+    group_idx = block_group_idx % num_groups
+    return block_idx, group_idx, event_idx, event_type, sm_id
 
 
 def export_to_perfetto_trace(
@@ -48,7 +54,7 @@ def export_to_perfetto_trace(
     event_names: List[str],
     file_name: str,
 ) -> None:
-
+    assert profiler_buffer.dtype == torch.uint64
     profiler_buffer_host = profiler_buffer.cpu()
     num_blocks, num_groups = profiler_buffer_host[:1].view(dtype=torch.int32)
     num_blocks = int(num_blocks)
@@ -56,13 +62,9 @@ def export_to_perfetto_trace(
 
     tgen = TraceGenerator(file_name)
 
+    pid_map = {}
     tid_map = {}
     track_map = {}
-    for block_idx in range(num_blocks):
-        pid = tgen.create_group(f"block_{block_idx}")
-        for group_idx in range(num_groups):
-            tid = pid.create_group(f"group_{group_idx}")
-            tid_map[(block_idx, group_idx)] = tid
 
     for i in range(1, len(profiler_buffer_host)):
         if profiler_buffer_host[i] == 0:
@@ -70,11 +72,18 @@ def export_to_perfetto_trace(
         tag, timestamp = profiler_buffer_host[i : i + 1].view(dtype=torch.uint32)
         tag = int(tag)
         timestamp = int(timestamp)
-        block_idx, group_idx, event_idx, event_type = decode_tag(
+        block_idx, group_idx, event_idx, event_type, sm_id = decode_tag(
             tag, num_blocks, num_groups
         )
-        event = event_names[event_idx]
+
+        # create trackers
+        if block_idx not in pid_map:
+            pid_map[block_idx] = tgen.create_group(f"sm_{sm_id}_block_{block_idx}")
+        pid = pid_map[block_idx]
+        if (block_idx, group_idx) not in tid_map:
+            tid_map[(block_idx, group_idx)] = pid.create_group(f"group_{group_idx}")
         tid = tid_map[(block_idx, group_idx)]
+        event = event_names[event_idx]
 
         if (block_idx, group_idx, event_idx) in track_map:
             track = track_map[(block_idx, group_idx, event_idx)]
diff --git a/include/flashinfer/profiler.cuh b/include/flashinfer/profiler.cuh
index 94eba48b04..9d1cddaba3 100644
--- a/include/flashinfer/profiler.cuh
+++ b/include/flashinfer/profiler.cuh
@@ -36,14 +36,21 @@ constexpr uint32_t BEGIN_END_MASK = 0x3;
 
 constexpr uint32_t EVENT_IDX_SHIFT = 2;
 constexpr uint32_t BLOCK_GROUP_IDX_SHIFT = 12;
+constexpr uint32_t SM_ID_SHIFT = 24;
+// Tag layout:
+// bits 0-1: event_type (start, end, instant)
+// bits 2-11: event_idx (translates to event_names in python profiler)
+// bits 12-23: block_id (12 bits)
+// bits 24-31: sm_id (8 bits)
 
 constexpr uint32_t EVENT_BEGIN = 0x0;
 constexpr uint32_t EVENT_END = 0x1;
 constexpr uint32_t EVENT_INSTANT = 0x2;
 
-__device__ __forceinline__ uint32_t encode_tag(uint32_t block_group_idx, uint32_t event_idx,
-                                               uint32_t event_type) {
-  return (block_group_idx << BLOCK_GROUP_IDX_SHIFT) | (event_idx << EVENT_IDX_SHIFT) | event_type;
+__device__ __forceinline__ uint32_t encode_tag(uint32_t sm_id, uint32_t block_id,
+                                               uint32_t event_idx, uint32_t event_type) {
+  return (sm_id << SM_ID_SHIFT) | (block_id << BLOCK_GROUP_IDX_SHIFT) |
+         (event_idx << EVENT_IDX_SHIFT) | event_type;
 }
 
 __device__ __forceinline__ uint32_t get_timestamp() {
@@ -79,17 +86,19 @@ struct ProfilerEntry {
 #define PROFILER_FUNC_PARAMS , at::Tensor profiler_buffer
 #define PROFILER_PARAMS_DECL uint64_t* profiler_buffer;
 
-#define PROFILER_INIT(params, smem_storage, closure, group_idx, num_groups,                     \
-                      write_thread_predicate)                                                   \
-  if (get_block_idx() == 0 && get_thread_idx() == 0) {                                          \
-    closure.entry.nblocks = get_num_blocks();                                                   \
-    closure.entry.ngroups = num_groups;                                                         \
-    params.profiler_buffer[0] = closure.entry.raw;                                              \
-  }                                                                                             \
-  closure.profiler_write_ptr =                                                                  \
-      params.profiler_buffer + 1 + get_block_idx() * num_groups + group_idx;                    \
-  closure.profiler_write_stride = get_num_blocks() * num_groups;                                \
-  closure.profiler_entry_tag_base = encode_tag(get_block_idx() * num_groups + group_idx, 0, 0); \
+#define PROFILER_INIT(params, smem_storage, closure, group_idx, num_groups,     \
+                      write_thread_predicate)                                   \
+  uint32_t _sm_idx;                                                             \
+  asm volatile("mov.u32 %0, %smid;" : "=r"(_sm_idx));                           \
+  if (get_block_idx() == 0 && get_thread_idx() == 0) {                          \
+    closure.entry.nblocks = get_num_blocks();                                   \
+    closure.entry.ngroups = num_groups;                                         \
+    params.profiler_buffer[0] = closure.entry.raw;                              \
+  }                                                                             \
+  closure.profiler_write_ptr =                                                  \
+      params.profiler_buffer + 1 + get_block_idx() * num_groups + group_idx;    \
+  closure.profiler_write_stride = get_num_blocks() * num_groups;                \
+  closure.profiler_entry_tag_base = encode_tag(_sm_idx, get_block_idx(), 0, 0); \
   closure.profiler_write_thread_predicate = write_thread_predicate;
 
 #define PROFILER_EVENT_START(closure, event)                                                  \
diff --git a/pyproject.toml b/pyproject.toml
index d11905894a..f9da81c8e1 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -49,6 +49,7 @@ packages = [
     "flashinfer.jit",
     "flashinfer.jit.attention",
     "flashinfer.triton",
+    "flashinfer.profiler",
     "flashinfer.triton.kernels",
     "flashinfer.comm",
     "flashinfer.cudnn",
