diff --git a/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh b/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
index 8b09f8918a0..53412a4758f 100644
--- a/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
+++ b/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
@@ -44,6 +44,8 @@ CUB_NAMESPACE_BEGIN
 
 namespace detail::scan
 {
+namespace __cub_detail  = CUB_NS_QUALIFIER::detail;
+namespace __scan_detail = CUB_NS_QUALIFIER::detail::scan;
 _CCCL_API constexpr warpspeed::SquadDesc squad_reduce(const scan_warpspeed_policy& policy)
 {
   return warpspeed::SquadDesc{0, policy.num_reduce_and_scan_warps};
@@ -418,13 +420,13 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void kernelBody(
         if (is_last_tile)
         {
           // TODO(bgruber): for operators where we know the identity we can probably optimize further here
-          regThreadSum = ThreadReducePartial(regInput, scan_op, valid_items_this_thread);
-          regWarpSum   = warpReducePartial(regThreadSum, scan_op, valid_threads_this_warp);
+          regThreadSum = __cub_detail::ThreadReducePartial(regInput, scan_op, valid_items_this_thread);
+          regWarpSum   = __scan_detail::warpReducePartial(regThreadSum, scan_op, valid_threads_this_warp);
         }
         else
         {
-          regThreadSum = ThreadReduce(regInput, scan_op);
-          regWarpSum   = warpReduce(regThreadSum, scan_op);
+          regThreadSum = CUB_NS_QUALIFIER::ThreadReduce(regInput, scan_op);
+          regWarpSum   = __scan_detail::warpReduce(regThreadSum, scan_op);
         }
       }
 
@@ -574,7 +576,7 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void kernelBody(
         // scan_op, and sumExclusiveIntraWarp is invalid when the inputs were
         // invalid.
         AccumT regSumThread          = refSumThreadAndWarpR.data()[squad.threadRank()];
-        AccumT sumExclusiveIntraWarp = warpScanExclusive(regSumThread, scan_op);
+        AccumT sumExclusiveIntraWarp = __scan_detail::warpScanExclusive(regSumThread, scan_op);
 
         if (squad.warpRank() == 0)
         {
@@ -661,11 +663,11 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void kernelBody(
       const bool use_prefix = hasInit ? true : !(is_first_tile && squad.threadRank() == 0);
       if constexpr (isInclusive)
       {
-        ThreadScanInclusive(regSumInclusive, regSumInclusive, scan_op, sumExclusive, use_prefix);
+        __cub_detail::ThreadScanInclusive(regSumInclusive, regSumInclusive, scan_op, sumExclusive, use_prefix);
       }
       else
       {
-        ThreadScanExclusive(regSumInclusive, regSumInclusive, scan_op, sumExclusive, use_prefix);
+        __cub_detail::ThreadScanExclusive(regSumInclusive, regSumInclusive, scan_op, sumExclusive, use_prefix);
       }
 
       ////////////////////////////////////////////////////////////////////////////////
