diff --git a/flash_attn/cute/interface.py b/flash_attn/cute/interface.py
index d12ea7b80b..9e9899d590 100644
--- a/flash_attn/cute/interface.py
+++ b/flash_attn/cute/interface.py
@@ -479,7 +479,7 @@ def _flash_attn_fwd(
     )
 
     requested_use_clc_scheduler = utils._get_use_clc_scheduler_default()
-    requested_disable_2cta = utils._get_disable_2cta_default()
+    requested_disable_2cta = utils._get_disable_2cta_default(is_fwd=True)
 
     current_stream = cute.runtime.make_fake_stream(use_tvm_ffi_env_stream=True)
 
diff --git a/flash_attn/cute/utils.py b/flash_attn/cute/utils.py
index b9dc4f5c11..ffabd34e39 100644
--- a/flash_attn/cute/utils.py
+++ b/flash_attn/cute/utils.py
@@ -86,8 +86,11 @@ def _get_use_clc_scheduler_default() -> bool:
     return _fa_clc_enabled
 
 
-def _get_disable_2cta_default() -> bool:
-    return _fa_disable_2cta_enabled or _fa_disable_2cta_cuda12
+def _get_disable_2cta_default(is_fwd: bool = False) -> bool:
+    if is_fwd:
+        return _fa_disable_2cta_enabled or _fa_disable_2cta_cuda12
+    else:
+        return _fa_disable_2cta_enabled
 
 
 def _compute_base_hash(func: Callable) -> str:
