diff --git a/torch/cuda/__init__.py b/torch/cuda/__init__.py
index b61b3eca7a4fd..8ca50c29415d8 100644
--- a/torch/cuda/__init__.py
+++ b/torch/cuda/__init__.py
@@ -259,7 +259,7 @@ def _check_capability():
     CUDA_ARCHES_SUPPORTED = {
         "12.6": {"min": 50, "max": 90},
         "12.8": {"min": 70, "max": 120},
-        "12.9": {"min": 70, "max": 120},
+        "13.0": {"min": 75, "max": 120},
     }
 
     if (
