diff --git a/csrc/trtllm_moe_allreduce_fusion.cu b/csrc/trtllm_moe_allreduce_fusion.cu
index 119d5ddf92..78f98b9d13 100644
--- a/csrc/trtllm_moe_allreduce_fusion.cu
+++ b/csrc/trtllm_moe_allreduce_fusion.cu
@@ -28,7 +28,7 @@ void trtllm_moe_allreduce_fusion(
     at::Tensor& rms_gamma, double rms_eps, double scale_factor,
     int64_t moe_reduction_device_num_experts, at::Tensor& moe_reduction_scale_input,
     at::Tensor& moe_reduction_active_experts_token_input, at::Tensor& moe_reduction_token_input,
-    std::optional<int64_t> layout_code, std::optional<at::Tensor> allreduce_out,
+    std::optional<int64_t> layout_code, std::optional<at::Tensor> moe_allreduce_out,
     std::optional<at::Tensor> residual_out, std::optional<at::Tensor> norm_out,
     std::optional<at::Tensor> quant_out, std::optional<at::Tensor> scale_out) {
   const c10::cuda::OptionalCUDAGuard device_guard(
@@ -44,9 +44,10 @@ void trtllm_moe_allreduce_fusion(
         params.hidden_dim = hidden_size;
         params.workspace = reinterpret_cast<void**>(workspace_ptrs.data_ptr());
 
-        params.allreduce_out = allreduce_out.has_value()
-                                   ? reinterpret_cast<void*>(allreduce_out.value().data_ptr())
-                                   : nullptr;
+        params.moe_allreduce_out =
+            moe_allreduce_out.has_value()
+                ? reinterpret_cast<void*>(moe_allreduce_out.value().data_ptr())
+                : nullptr;
         params.residual_in = reinterpret_cast<void*>(residual_in.data_ptr());
         params.residual_out = residual_out.has_value()
                                   ? reinterpret_cast<void*>(residual_out.value().data_ptr())
diff --git a/flashinfer/comm.py b/flashinfer/comm.py
index 8ec47bb8d5..a80cd9e66e 100644
--- a/flashinfer/comm.py
+++ b/flashinfer/comm.py
@@ -547,7 +547,7 @@ def trtllm_moe_allreduce_fusion(
         moe_reduction_active_experts_token_input: torch.Tensor,
         moe_reduction_token_input: torch.Tensor,
         layout_code: Optional[FP4QuantizationSFLayout],
-        allreduce_out: Optional[torch.Tensor],
+        moe_allreduce_out: Optional[torch.Tensor],
         residual_out: Optional[torch.Tensor],
         norm_out: Optional[torch.Tensor],
         quant_out: Optional[torch.Tensor],
@@ -569,7 +569,7 @@ def trtllm_moe_allreduce_fusion(
             moe_reduction_active_experts_token_input,
             moe_reduction_token_input,
             layout_code,
-            allreduce_out,
+            moe_allreduce_out,
             residual_out,
             norm_out,
             quant_out,
@@ -1042,7 +1042,7 @@ def trtllm_moe_allreduce_fusion(
     moe_reduction_active_experts_token_input: torch.Tensor,
     moe_reduction_token_input: torch.Tensor,
     layout_code: Optional[FP4QuantizationSFLayout],
-    allreduce_out: Optional[torch.Tensor],
+    moe_allreduce_out: Optional[torch.Tensor],
     residual_out: Optional[torch.Tensor],
     norm_out: Optional[torch.Tensor],
     quant_out: Optional[torch.Tensor],
@@ -1064,7 +1064,7 @@ def trtllm_moe_allreduce_fusion(
         moe_reduction_active_experts_token_input=moe_reduction_active_experts_token_input,
         moe_reduction_token_input=moe_reduction_token_input,
         layout_code=layout_code,
-        allreduce_out=allreduce_out,
+        moe_allreduce_out=moe_allreduce_out,
         residual_out=residual_out,
         norm_out=norm_out,
         quant_out=quant_out,
diff --git a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
index 43f01fed83..94676cac05 100644
--- a/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
+++ b/include/flashinfer/comm/trtllm_moe_allreduce_fusion.cuh
@@ -661,7 +661,6 @@ struct MoeReductionAllReduceFusionParams : public AllReduceFusionParams<T> {
   // Refer to kernel implementation on layout of those params
   // number of active experts on current device
 
-  // todo(review): why int* moe_reduction_device_num_experts = nullptr; in trt-llm?
   int moe_reduction_device_num_experts = 0;
   // per token per expert fp32 scale
   float* moe_reduction_scale_input = nullptr;
@@ -669,6 +668,9 @@ struct MoeReductionAllReduceFusionParams : public AllReduceFusionParams<T> {
   void* moe_reduction_active_experts_token_input = nullptr;
   // per token input
   void* moe_reduction_token_input = nullptr;
+
+  // moe-allreduce output (non-fused)
+  void* moe_allreduce_out = nullptr;
 };
 
 template <int NRanks>
@@ -765,9 +767,10 @@ __device__ __forceinline__ vec_t<T, VEC_SIZE> rms_norm(vec_t<T, VEC_SIZE> const&
 template <bool AllReduceOut, bool ResidualOut, bool NormOut, bool QuantOut, typename T,
           uint32_t VEC_SIZE>
 __device__ __forceinline__ void fused_op(vec_t<T, VEC_SIZE> const& val, int access_id, int token_id,
-                                         int access_id_in_token, AllReduceFusionParams<T>& params) {
+                                         int access_id_in_token,
+                                         MoeReductionAllReduceFusionParams<T>& params) {
   if constexpr (AllReduceOut) {
-    val.store(reinterpret_cast<T*>(params.allreduce_out) + access_id * VEC_SIZE);
+    val.store(reinterpret_cast<T*>(params.moe_allreduce_out) + access_id * VEC_SIZE);
   }
   vec_t<T, VEC_SIZE> residual_val;
   residual_val.load(reinterpret_cast<T*>(params.residual_in) + access_id * VEC_SIZE);
@@ -1157,8 +1160,8 @@ cudaError_t moereduction_allreduce_fusion_op(MoeReductionAllReduceFusionParams<T
   FLASHINFER_CHECK(params.hidden_dim * sizeof(T) % details::kBytesPerAccess == 0,
                    "hidden_dim * sizeof(T) must be a multiple of kBytesPerAccess");
   FLASHINFER_CHECK(
-      params.allreduce_out || params.residual_out || params.norm_out || params.quant_out,
-      "at least one of allreduce_out, residual_out, norm_out, quant_out must be set");
+      params.moe_allreduce_out || params.residual_out || params.norm_out || params.quant_out,
+      "at least one of moe_allreduce_out, residual_out, norm_out, quant_out must be set");
 
   // hidden_dim (d) = 7168 for dpsk moe, and hence 128 tokens as one-shot threshold
   // AR outputs are optional, since we always have fused options followed.
@@ -1179,8 +1182,8 @@ cudaError_t moereduction_allreduce_fusion_op(MoeReductionAllReduceFusionParams<T
   // [m, d] bf16 residual_out, [m, d] bf16 norm_out, [m, d] fp4 quant_out
 
   auto status = DISPATCH_MOEREDUCTION(
-      params.nranks, params.allreduce_out, params.residual_out, params.rms_gamma, params.quant_out,
-      N_RANKS, AR, RES, RMS, QUANT, [&]() -> cudaError_t {
+      params.nranks, params.moe_allreduce_out, params.residual_out, params.rms_gamma,
+      params.quant_out, N_RANKS, AR, RES, RMS, QUANT, [&]() -> cudaError_t {
         FLASHINFER_CUDA_CALL(
             (moereduction_allreduce_fusion_kernel_launcher<T, N_RANKS, AR, RES, RMS, QUANT>(
                 (params), (launch_with_pdl))));
diff --git a/tests/test_trtllm_allreduce_fusion.py b/tests/test_trtllm_allreduce_fusion.py
index a80c144b52..0e61f2354d 100644
--- a/tests/test_trtllm_allreduce_fusion.py
+++ b/tests/test_trtllm_allreduce_fusion.py
@@ -314,7 +314,7 @@ def multi_process_parallel(
 
 
 @pytest.mark.parametrize("world_size", [2, 4])
-@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16, torch.float32])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
 @pytest.mark.parametrize("hidden_dim", [1024, 2048, 4096, 7168, 8192])
 def test_trtllm_allreduce_fusion(world_size, dtype, hidden_dim):
     np.random.seed(42)
diff --git a/tests/test_trtllm_moe_allreduce_fusion.py b/tests/test_trtllm_moe_allreduce_fusion.py
index 3bcd40f66d..1c7f3f00d2 100644
--- a/tests/test_trtllm_moe_allreduce_fusion.py
+++ b/tests/test_trtllm_moe_allreduce_fusion.py
@@ -79,7 +79,7 @@ def _run_correctness_worker(world_size, rank, dtype, distributed_init_port):
                             )
                             residual_in_clone = residual_in.clone()
 
-                            all_reduce_out = torch.zeros(
+                            moe_allreduce_out = torch.zeros(
                                 message_size, dtype=dtype, device=device
                             )
                             residual_out = torch.empty_like(residual_in)
@@ -179,14 +179,17 @@ def _run_correctness_worker(world_size, rank, dtype, distributed_init_port):
                             )  # [token_num, HIDDEN_SIZE]
 
                             # 3. All-Reduce
-                            all_reduce_ref = moe_out_ref.clone().to(dtype)
-                            dist.all_reduce(all_reduce_ref, group=group)
-                            all_reduce_ref = all_reduce_ref.to(torch.float32)
+                            moe_allreduce_ref = moe_out_ref.clone().to(dtype)
+                            dist.all_reduce(moe_allreduce_ref, group=group)
+                            moe_allreduce_ref = moe_allreduce_ref.to(torch.float32)
 
                             # 4. Fused Ops
-                            ref_residual_out = all_reduce_ref + residual_in_clone.view(
-                                token_num, HIDDEN_SIZE
-                            ).to(torch.float32)
+                            ref_residual_out = (
+                                moe_allreduce_ref
+                                + residual_in_clone.view(token_num, HIDDEN_SIZE).to(
+                                    torch.float32
+                                )
+                            )
 
                             variance = (
                                 ref_residual_out.to(torch.float32)
@@ -217,7 +220,7 @@ def _run_correctness_worker(world_size, rank, dtype, distributed_init_port):
                                 moe_reduction_active_experts_token_input=moe_reduction_active_experts_token_input,
                                 moe_reduction_token_input=moe_reduction_token_input,
                                 layout_code=swizzled_layout_code,
-                                allreduce_out=all_reduce_out,
+                                moe_allreduce_out=moe_allreduce_out,
                                 residual_out=residual_out,
                                 norm_out=norm_out,
                                 quant_out=quant_out,
@@ -225,7 +228,9 @@ def _run_correctness_worker(world_size, rank, dtype, distributed_init_port):
                             )
 
                             # match shape
-                            all_reduce_out = all_reduce_out.view(token_num, HIDDEN_SIZE)
+                            moe_allreduce_out = moe_allreduce_out.view(
+                                token_num, HIDDEN_SIZE
+                            )
                             residual_out = residual_out.view(token_num, HIDDEN_SIZE)
                             norm_out = norm_out.view(token_num, HIDDEN_SIZE)
 
@@ -235,42 +240,44 @@ def _run_correctness_worker(world_size, rank, dtype, distributed_init_port):
                             tolerance = 8e-2 if dtype == torch.float16 else 8e-1
                             # 6.1 Check allreduce_out
                             if not torch.allclose(
-                                all_reduce_out.to(torch.float32),
-                                all_reduce_ref,
+                                moe_allreduce_out.to(torch.float32),
+                                moe_allreduce_ref,
                                 atol=tolerance,
                                 rtol=1e-2,
                             ):
                                 test_passed = False
-                                print(f"Rank {rank} allreduce_out mismatch")
-                                print(f"all_reduce_out: {all_reduce_out}")
-                                print(f"all_reduce_ref: {all_reduce_ref}")
+                                print(f"Rank {rank} moe_allreduce_out mismatch")
+                                print(f"moe_allreduce_out: {moe_allreduce_out}")
+                                print(f"moe_allreduce_ref: {moe_allreduce_ref}")
                                 # Print max diff elements for allreduce_out
                                 max_diff = torch.max(
                                     torch.abs(
-                                        all_reduce_out.to(torch.float32)
-                                        - all_reduce_ref
+                                        moe_allreduce_out.to(torch.float32)
+                                        - moe_allreduce_ref
                                     )
                                 )
                                 max_diff_idx = torch.argmax(
                                     torch.abs(
-                                        all_reduce_out.to(torch.float32)
-                                        - all_reduce_ref
+                                        moe_allreduce_out.to(torch.float32)
+                                        - moe_allreduce_ref
                                     )
                                 )
-                                print(f"Rank {rank} allreduce_out max diff: {max_diff}")
                                 print(
-                                    f"Rank {rank} allreduce_out max diff idx: {max_diff_idx}"
+                                    f"Rank {rank} moe_allreduce_out max diff: {max_diff}"
+                                )
+                                print(
+                                    f"Rank {rank} moe_allreduce_out max diff idx: {max_diff_idx}"
                                 )
                                 print(
-                                    f"Rank {rank} allreduce_out value at max diff: {all_reduce_out.view(-1)[max_diff_idx]}"
+                                    f"Rank {rank} moe_allreduce_out value at max diff: {moe_allreduce_out.view(-1)[max_diff_idx]}"
                                 )
                                 print(
-                                    f"Rank {rank} allreduce_out ref value at max diff: {all_reduce_ref.view(-1)[max_diff_idx]}"
+                                    f"Rank {rank} moe_allreduce_out ref value at max diff: {moe_allreduce_ref.view(-1)[max_diff_idx]}"
                                 )
 
                             torch.testing.assert_close(
-                                all_reduce_out.to(torch.float32),
-                                all_reduce_ref,
+                                moe_allreduce_out.to(torch.float32),
+                                moe_allreduce_ref,
                                 atol=tolerance,
                                 rtol=1e-2,
                             )
