diff --git a/flash_attn/cute/flash_fwd_sm100.py b/flash_attn/cute/flash_fwd_sm100.py
index a3969d04883..422fcd68fc7 100644
--- a/flash_attn/cute/flash_fwd_sm100.py
+++ b/flash_attn/cute/flash_fwd_sm100.py
@@ -242,10 +242,10 @@ def __init__(
             # self.num_regs_correction = 96
             # self.num_regs_correction = 64 if self.is_causal or self.is_local else 80
             if not self.enable_ex2_emu:
-                self.num_regs_correction = 80
+                self.num_regs_correction = 80 if not paged_kv_non_tma else 64
             else:
                 # self.num_regs_correction = 64
-                self.num_regs_correction = 80
+                self.num_regs_correction = 80 if not paged_kv_non_tma else 64
             # self.num_regs_other = 32
             # self.num_regs_other = 64
             # self.num_regs_other = 80
@@ -833,7 +833,8 @@ def kernel(
 
         ThreadCooperativeGroup = partial(pipeline.CooperativeGroup, pipeline.Agent.Thread)
         mma_warp = ThreadCooperativeGroup(len([self.mma_warp_id]))
-        tma_warp = ThreadCooperativeGroup(len(self.load_warp_ids))
+        load_warps = ThreadCooperativeGroup(len(self.load_warp_ids))
+        tma_warp = ThreadCooperativeGroup(1)
         softmax_warps = ThreadCooperativeGroup(len(self.softmax0_warp_ids))
         softmax_threads = ThreadCooperativeGroup(cute.arch.WARP_SIZE * len(self.softmax0_warp_ids))
         # softmax_threads = ThreadCooperativeGroup(cute.arch.WARP_SIZE)
@@ -1389,7 +1390,8 @@ def load(
 
         pipeline_kv.producer_tail(kv_producer_state)
         # This is equivalent to pipeline_q.producer_tail
-        pipeline_q.producer_acquire_w_index_phase(self.q_stage - 1, q_producer_phase)
+        if const_expr(len(self.load_warp_ids) == 1) or warp_idx == self.load_warp_ids[0]:
+            pipeline_q.producer_acquire_w_index_phase(self.q_stage - 1, q_producer_phase)
 
     @cute.jit
     def mma(
@@ -2724,7 +2726,10 @@ def load_KV(
         assert K_or_V in ("K", "V")
         stage, phase = producer_state.index, producer_state.phase
         extra_tx_count_kv = self.tma_copy_bytes[K_or_V] - self.tma_copy_bytes["K"]
-        extra_tx_count = extra_tx_count_kv + (extra_tx_count if extra_tx_count is not None else 0)
+        extra_tx_count = (
+            extra_tx_count_kv + (extra_tx_count if extra_tx_count is not None else 0) if const_expr(self.use_tma_KV)
+            else None
+        )
         extra_kwargs = {"extra_tx_count": extra_tx_count} if const_expr(self.use_tma_KV) else {}
         pipeline_kv.producer_acquire(producer_state, **extra_kwargs)
         if const_expr(K_or_V == "K" and self.uneven_kv_smem):
diff --git a/flash_attn/cute/interface.py b/flash_attn/cute/interface.py
index 25800522784..449bd56491e 100644
--- a/flash_attn/cute/interface.py
+++ b/flash_attn/cute/interface.py
@@ -472,7 +472,17 @@ def _flash_attn_fwd(
         elif arch // 10 in [10, 11]:
             head_dim_padded = int(math.ceil(head_dim / 16) * 16)
             head_dim_v_padded = int(math.ceil(head_dim / 16) * 16)
-            use_2cta_instrs = not causal and not local and not is_split_kv and cu_seqlens_q is None and seqused_q is None and not use_block_sparsity and head_dim_padded == 128 and head_dim_v_padded == 128
+            use_2cta_instrs = (
+                not causal
+                and not local
+                and not is_split_kv
+                and cu_seqlens_q is None
+                and seqused_q is None
+                and not use_block_sparsity
+                and page_size in [None, 128]
+                and head_dim_padded == 128
+                and head_dim_v_padded == 128
+            )
             fa_fwd = FlashAttentionForwardSm100(
                 head_dim,
                 head_dim_v,
