diff --git a/flashinfer/decode.py b/flashinfer/decode.py
index b7c8e33508..1d725a4436 100644
--- a/flashinfer/decode.py
+++ b/flashinfer/decode.py
@@ -2091,6 +2091,8 @@ def trtllm_batch_decode_with_kv_cache(
             out_scale_factor = out.scale
             o_sf_start_index = out.scale_start_index
             out = out.data
+            # out_dtype may be None
+            out_dtype = out_dtype or "nvfp4"
         elif out is None:
             fp4_out_scale_shape = (
                 round_up(query.shape[0], 128),
@@ -2104,6 +2106,7 @@ def trtllm_batch_decode_with_kv_cache(
         else:
             raise ValueError(f"Invalid out: {out}")
 
+        assert out_dtype == "nvfp4"
         assert isinstance(out, torch.Tensor)
 
         # Use uint8 as the container dtype to compliant with next fp4 gemm.
diff --git a/flashinfer/prefill.py b/flashinfer/prefill.py
index 55a0b1fc57..ef4264a801 100755
--- a/flashinfer/prefill.py
+++ b/flashinfer/prefill.py
@@ -3349,6 +3349,8 @@ def trtllm_batch_context_with_kv_cache(
             out_scale_factor = out.scale
             o_sf_start_index = out.scale_start_index
             out = out.data
+            # out_dtype may be None
+            out_dtype = out_dtype or "nvfp4"
         elif out is None:
             fp4_out_scale_shape = (
                 round_up(query.shape[0], 128),
@@ -3362,6 +3364,7 @@ def trtllm_batch_context_with_kv_cache(
         else:
             raise ValueError(f"Invalid out: {out}")
 
+        assert out_dtype == "nvfp4"
         assert isinstance(out, torch.Tensor)
 
         # Use uint8 as the container dtype to compliant with next fp4 gemm.
diff --git a/tests/test_trtllm_gen_attention.py b/tests/test_trtllm_gen_attention.py
index f304ea575e..d0639bef84 100755
--- a/tests/test_trtllm_gen_attention.py
+++ b/tests/test_trtllm_gen_attention.py
@@ -298,6 +298,18 @@ def test_trtllm_batch_prefill(
         q, o_dtype, create_out_tensor
     )
 
+    # determine to pass out_dtype explicitly or not
+    if q_dtype != o_dtype and not create_out_tensor:
+        out_dtype = DTYPE_MAP[o_dtype]
+    else:
+        out_dtype = (
+            DTYPE_MAP[o_dtype]
+            if flip_coin(
+                batch_size, page_size, num_kv_heads, head_grp_size, o_dtype, q_dtype
+            )
+            else None
+        )
+
     global global_workspace_buffer
     if global_workspace_buffer is None:
         global_workspace_buffer = torch.zeros(
@@ -345,7 +357,7 @@ def test_trtllm_batch_prefill(
         kv_indptr,
         window_left,  # window_left
         out=out,
-        out_dtype=DTYPE_MAP[o_dtype],
+        out_dtype=out_dtype,
         o_sf_scale=o_sf_scale,
         o_sf_vec_size=o_sf_vec_size,
         enable_pdl=enable_pdl,
@@ -465,6 +477,18 @@ def test_trtllm_batch_decode(
         q, o_dtype, create_out_tensor
     )
 
+    # determine to pass out_dtype explicitly or not
+    if q_dtype != o_dtype and not create_out_tensor:
+        out_dtype = DTYPE_MAP[o_dtype]
+    else:
+        out_dtype = (
+            DTYPE_MAP[o_dtype]
+            if flip_coin(
+                batch_size, page_size, num_kv_heads, head_grp_size, o_dtype, q_dtype
+            )
+            else None
+        )
+
     global global_workspace_buffer
     if global_workspace_buffer is None:
         global_workspace_buffer = torch.zeros(
@@ -506,7 +530,7 @@ def test_trtllm_batch_decode(
         v_scale / o_scale,  # bmm2_scale
         window_left,  # window_left
         out=out,
-        out_dtype=DTYPE_MAP[o_dtype],
+        out_dtype=out_dtype,
         o_sf_scale=o_sf_scale,
         o_sf_vec_size=o_sf_vec_size,
         enable_pdl=enable_pdl,
