diff --git a/flashinfer/decode.py b/flashinfer/decode.py
index 5178ccfbae..b7c8e33508 100644
--- a/flashinfer/decode.py
+++ b/flashinfer/decode.py
@@ -2133,7 +2133,8 @@ def trtllm_batch_decode_with_kv_cache(
         assert o_sf_vec_size is None
         out_scale_factor = None
         o_sf_start_index = 0
-        out_dtype = out_dtype or query.dtype
+        if out_dtype is None:
+            out_dtype = out.dtype if out is not None else query.dtype
         out = out if out is not None else torch.empty_like(query, dtype=out_dtype)
         if out_dtype not in (query.dtype, torch.float16, torch.bfloat16):
             raise ValueError(f"Unsupported out_dtype: {out_dtype}")
diff --git a/flashinfer/prefill.py b/flashinfer/prefill.py
index 3805d3a546..55a0b1fc57 100755
--- a/flashinfer/prefill.py
+++ b/flashinfer/prefill.py
@@ -3391,10 +3391,11 @@ def trtllm_batch_context_with_kv_cache(
         assert o_sf_vec_size is None
         out_scale_factor = None
         o_sf_start_index = 0
-        out_dtype = out_dtype or query.dtype
+        if out_dtype is None:
+            out_dtype = out.dtype if out is not None else query.dtype
+        out = out if out is not None else torch.empty_like(query, dtype=out_dtype)
         if out_dtype not in (query.dtype, torch.float16, torch.bfloat16):
             raise ValueError(f"Unsupported out_dtype: {out_dtype}")
-        out = out if out is not None else torch.empty_like(query, dtype=out_dtype)
         check_shape_dtype_device(out, query.shape, out_dtype, query.device, "out")
     else:
         raise ValueError(f"Invalid out_dtype: {out_dtype}")
