diff --git a/csrc/trtllm_fmha_kernel_launcher.cu b/csrc/trtllm_fmha_kernel_launcher.cu
index e533ee49db..e366e79338 100644
--- a/csrc/trtllm_fmha_kernel_launcher.cu
+++ b/csrc/trtllm_fmha_kernel_launcher.cu
@@ -208,9 +208,9 @@ void trtllm_paged_attention_decode(at::Tensor out, std::optional<at::Tensor> con
                                               std::to_string(head_dim_kv) + " and " +
                                               std::to_string(head_dim_qk));
   int head_dim_vo = is_4bit(o_data_type) ? out.size(-1) * 2 : out.size(-1);
-  TORCH_CHECK(head_dim_kv == head_dim_vo, "head_dim_kv and head_dim_vo must be the same, got " +
-                                              std::to_string(head_dim_kv) + " and " +
-                                              std::to_string(head_dim_vo));
+  TORCH_CHECK((head_dim_kv == 576 && head_dim_vo == 512) || head_dim_kv == head_dim_vo,
+              "head_dim_kv and head_dim_vo must be the same for non-MLA attention, got " +
+                  std::to_string(head_dim_kv) + " and " + std::to_string(head_dim_vo));
   // NOTE(Zihao): key_value_cache is [num_pages, 1/2, num_kv_heads, page_size, head_dim]
   // For KV-Cache sharing (MLA), the second dimension is 1 (key/value cache are shared)
   // otherwise it is 2, one for key and one for value
diff --git a/flashinfer/decode.py b/flashinfer/decode.py
index 6ab2278770..3e1daf8e58 100644
--- a/flashinfer/decode.py
+++ b/flashinfer/decode.py
@@ -2206,6 +2206,7 @@ def trtllm_batch_decode_with_kv_cache_mla(
 
     run_func(
         out,
+        None,  # fp4 output not supported in wrapper api yet.
         query,
         kv_cache.unsqueeze(-3),
         workspace_buffer,
@@ -2214,6 +2215,8 @@ def trtllm_batch_decode_with_kv_cache_mla(
         max_seq_len,
         bmm1_scale,
         bmm2_scale,
+        -1,  # o_sf_scale
+        -1,  # o_sf_vec_size
         -1,  # window_left
         sm_count,
     )
diff --git a/tests/test_trtllm_gen_decode.py b/tests/test_trtllm_gen_decode.py
index de940744f1..319a86f761 100644
--- a/tests/test_trtllm_gen_decode.py
+++ b/tests/test_trtllm_gen_decode.py
@@ -349,7 +349,7 @@ def test_trtllm_batch_decode_fmha(
 @pytest.mark.parametrize("dtype", [torch.float8_e4m3fn, torch.bfloat16])
 @pytest.mark.parametrize("page_size", [32, 64])
 @pytest.mark.parametrize("q_len_per_request", [1, 2])
-@pytest.mark.parametrize("dynamic_scale", [False, True])
+@pytest.mark.parametrize("dynamic_scale", [False])
 def test_trtllm_batch_decode_mla(
     batch_size: int,
     scale: float,
