diff --git a/README.md b/README.md
index eff08d6c3..37af4d1e3 100644
--- a/README.md
+++ b/README.md
@@ -27,7 +27,8 @@ ThunderKittens solves a few key problems that enable high utilization on modern
 
 Here’s an example of what flash attention for an RTX 4090 looks like written in ThunderKittens.
 
-```#include "kittens.cuh"
+```Cuda
+#include "kittens.cuh"
 
 #define NUM_WORKERS 16
 using namespace kittens;
@@ -153,4 +154,4 @@ However, not all ThunderKittens functions operate at the warp level. Many import
 
 ### Other Restrictions
 
-Most operations in ThunderKittens are pure functional. However, some operations _do_ have special restrictions; ThunderKittens tries to warn you by giving them names that stand out. For example, a register tile transpose needs separable arguments: if it is given the same underlying registers as both source and destination, it will silently fail. Consequently, it is named `transpose_sep`.
\ No newline at end of file
+Most operations in ThunderKittens are pure functional. However, some operations _do_ have special restrictions; ThunderKittens tries to warn you by giving them names that stand out. For example, a register tile transpose needs separable arguments: if it is given the same underlying registers as both source and destination, it will silently fail. Consequently, it is named `transpose_sep`.
diff --git a/examples/attn_fwd_tma_wgmma/Makefile b/examples/attn_fwd_tma_wgmma/Makefile
new file mode 100644
index 000000000..8183499ec
--- /dev/null
+++ b/examples/attn_fwd_tma_wgmma/Makefile
@@ -0,0 +1,30 @@
+# GPU Selection: Set to either 4090 or H100
+GPU_TARGET=H100
+
+# Compiler
+NVCC=nvcc
+
+# Conditional setup based on the target GPU
+ifeq ($(GPU_TARGET),4090)
+NVCCFLAGS=-DNDEBUG -Xcompiler=-fPIE --expt-extended-lambda --expt-relaxed-constexpr -Xcompiler=-Wno-psabi -Xcompiler=-fno-strict-aliasing --use_fast_math -forward-unknown-to-host-compiler -O3  -Xnvlink=--verbose -Xptxas=--verbose -Xptxas=--warn-on-spills -std=c++20 -MD -MT -MF -x cu -lrt -lpthread -ldl -DKITTENS_4090 -arch=sm_89 -lcuda -lcudadevrt -lcudart_static -lcublas # 4090
+TARGET=attn_fwd # 4090
+SRC=tictoc_ker.cu # 4090
+else ifeq ($(GPU_TARGET),A100)
+NVCCFLAGS=-DNDEBUG -Xcompiler=-fPIE --expt-extended-lambda --expt-relaxed-constexpr -Xcompiler=-Wno-psabi -Xcompiler=-fno-strict-aliasing --use_fast_math -forward-unknown-to-host-compiler -O3  -Xnvlink=--verbose -Xptxas=--verbose -Xptxas=--warn-on-spills -std=c++20 -MD -MT -MF -x cu -lrt -lpthread -ldl -DKITTENS_A100 -arch=sm_80 -lcuda -lcudadevrt -lcudart_static -lcublas # H100
+TARGET=attn_fwd # A100
+SRC=tictoc_ker.cu # A100
+else ifeq ($(GPU_TARGET),H100)
+NVCCFLAGS=-DNDEBUG -Xcompiler=-fPIE --expt-extended-lambda --expt-relaxed-constexpr -Xcompiler=-Wno-psabi -Xcompiler=-fno-strict-aliasing --use_fast_math -forward-unknown-to-host-compiler -O3  -Xnvlink=--verbose -Xptxas=--verbose -Xptxas=--warn-on-spills -std=c++20 -MD -MT -MF -x cu -lrt -lpthread -ldl -DKITTENS_HOPPER -arch=sm_90a -lcuda -lcudadevrt -lcudart_static -lcublas # H100
+TARGET=attn_fwd # H100
+SRC=tictoc_ker.cu # H100
+endif
+
+# Default target
+all: $(TARGET)
+
+$(TARGET): $(SRC)
+	$(NVCC) $(SRC) $(NVCCFLAGS) -o $(TARGET)
+
+# Clean target
+clean:
+	rm -f $(TARGET)
\ No newline at end of file
diff --git a/examples/attn_fwd_tma_wgmma/gentests.py b/examples/attn_fwd_tma_wgmma/gentests.py
new file mode 100644
index 000000000..041dd2aa3
--- /dev/null
+++ b/examples/attn_fwd_tma_wgmma/gentests.py
@@ -0,0 +1,57 @@
+import torch
+from tqdm import trange
+import numpy as np
+import sys
+
+# only generate a single batch/head of data, which makes file loading much faster.
+# it does mean we'll have to check batch/head behavior separately later, but that should be much easier to debug.
+B = 1
+H = 1
+N = 2048 if len(sys.argv) <= 2 else int(sys.argv[2])
+D = 128 if len(sys.argv) <= 3 else int(sys.argv[3])
+
+TESTNAME = sys.argv[1]
+
+if TESTNAME == 'ones':
+    q = torch.ones((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    k = torch.ones((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    v = torch.ones((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+elif TESTNAME == 'randn':
+    torch.random.manual_seed(42)
+    q = torch.randn((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    k = torch.randn((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    v = torch.randn((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+elif TESTNAME == 'qk_test':
+    q = torch.eye(D).reshape((1,1,D,D)).repeat(B, H, N//D, 1)*10
+    k = torch.eye(D).reshape((1,1,D,D)).repeat(B, H, N//D, 1)*10
+    v = torch.eye(D).reshape((1,1,D,D)).repeat(B, H, N//D, 1)*10
+elif TESTNAME == 'v_orientation':
+    q = torch.ones((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    k = torch.ones((B, H, N, D), dtype=torch.bfloat16, device='cuda')
+    v = (torch.arange(D, dtype=torch.bfloat16, device='cuda')/D).reshape((1,1,1,-1)).repeat(B, H, N, 1)
+else:
+    print('Invalid test name')
+    sys.exit(0)
+
+o = torch.nn.functional.scaled_dot_product_attention(q, k, v)
+
+fn = f'{TESTNAME}_{N}_{D}.txt'
+with open(fn, 'w') as f:
+    qf = q.to(torch.float32).flatten().cpu().numpy()
+    kf = k.to(torch.float32).flatten().cpu().numpy()
+    vf = v.to(torch.float32).flatten().cpu().numpy()
+    of = o.to(torch.float32).flatten().cpu().numpy()
+    for i in trange(B*H*N*D):
+        f.write(repr(qf[i]))
+        f.write(' ')
+    for i in trange(B*H*N*D):
+        f.write(repr(kf[i]))
+        f.write(' ')
+    for i in trange(B*H*N*D):
+        f.write(repr(vf[i]))
+        f.write(' ')
+    for i in trange(B*H*N*D):
+        f.write(repr(of[i]))
+        f.write(' ')
+
+print(f'Run the harness like `./attn_fwd {fn}`')
\ No newline at end of file
diff --git a/examples/attn_fwd_tma_wgmma/harness.impl b/examples/attn_fwd_tma_wgmma/harness.impl
new file mode 100644
index 000000000..2c3b163f6
--- /dev/null
+++ b/examples/attn_fwd_tma_wgmma/harness.impl
@@ -0,0 +1,172 @@
+#include <iostream>
+#include <string>
+#include <fstream>
+
+constexpr int ATTN_B     = 16;
+constexpr int ATTN_H     = 16;
+constexpr int ATTN_N     = 5120;
+constexpr int ATTN_D     = 64;
+constexpr int BLOCK_SIZE = (32*NUM_WORKERS);
+
+#define CudaCheckError()    __cudaCheckError( __FILE__, __LINE__ )
+inline void __cudaCheckError( const char *file, const int line ) {
+    cudaError err = cudaGetLastError();
+    if ( cudaSuccess != err )
+    {
+        fprintf( stderr, "cudaCheckError() failed at %s:%i : %s\n",
+                 file, line, cudaGetErrorString( err ) );
+        exit( -1 );
+    }
+    // More careful checking. However, this will affect performance.
+    // Comment away if needed.
+    err = cudaDeviceSynchronize();
+    if( cudaSuccess != err )
+    {
+        fprintf( stderr, "cudaCheckError() with sync failed at %s:%i : %s\n",
+                 file, line, cudaGetErrorString( err ) );
+        exit( -1 );
+    }
+}
+
+int main(int argc, char **argv) {
+    // TODO: consider doing sequential kernel launches to force batches dimension element to execute sequentially,
+    // which may increase the probability of L2 cache hits on KV
+
+    std::cout << "Entered main!" << std::endl;
+
+    // create dummy variables that are the right size
+    constexpr int TOTAL_ELEMENTS = ATTN_B*ATTN_H*ATTN_N*ATTN_D;
+    constexpr int TOTAL_UNIQUE_ELEMENTS = ATTN_N*ATTN_D;
+
+    float *q = new float[TOTAL_UNIQUE_ELEMENTS];
+    float *k = new float[TOTAL_UNIQUE_ELEMENTS];
+    float *v = new float[TOTAL_UNIQUE_ELEMENTS];
+    float *o_ref = new float[TOTAL_UNIQUE_ELEMENTS];
+
+    bf16 *q_bf = new bf16[TOTAL_ELEMENTS];
+    bf16 *k_bf = new bf16[TOTAL_ELEMENTS];
+    bf16 *v_bf = new bf16[TOTAL_ELEMENTS];
+    bf16 *o_bf = new bf16[TOTAL_ELEMENTS];
+    float *o = new float[TOTAL_ELEMENTS];
+
+    std::ifstream infile(argv[1]);
+
+    std::cout << "Starting to enter!" << std::endl;
+
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) infile >> q[i];
+    std::cout << "Finished loading Q" << std::endl;
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) infile >> k[i];
+    std::cout << "Finished loading K" << std::endl;
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) infile >> v[i];
+    std::cout << "Finished loading V" << std::endl;
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) infile >> o_ref[i];
+    std::cout << "Finished loading O_REF" << std::endl;
+
+    std::cout << "Finished loading file from " << argv[1] << "!" << std::endl;
+
+    // replicate into heads
+    for(int i = 0; i < TOTAL_ELEMENTS; i++) {
+        q_bf[i] = __float2bfloat16(q[i % TOTAL_UNIQUE_ELEMENTS]);
+        k_bf[i] = __float2bfloat16(k[i % TOTAL_UNIQUE_ELEMENTS]);
+        v_bf[i] = __float2bfloat16(v[i % TOTAL_UNIQUE_ELEMENTS]);
+    }
+
+    bf16 *d_q, *d_k, *d_v, *d_o;
+    cudaMalloc(&d_q, TOTAL_ELEMENTS * sizeof(bf16));
+    cudaMalloc(&d_k, TOTAL_ELEMENTS * sizeof(bf16));
+    cudaMalloc(&d_v, TOTAL_ELEMENTS * sizeof(bf16));
+    cudaMalloc(&d_o, TOTAL_ELEMENTS * sizeof(bf16));
+
+    cudaMemcpy(d_q, q_bf, TOTAL_ELEMENTS * sizeof(bf16), cudaMemcpyHostToDevice);
+    cudaMemcpy(d_k, k_bf, TOTAL_ELEMENTS * sizeof(bf16), cudaMemcpyHostToDevice);
+    cudaMemcpy(d_v, v_bf, TOTAL_ELEMENTS * sizeof(bf16), cudaMemcpyHostToDevice);
+
+    CUtensorMap tma_desc_q = {}; 
+    CUtensorMap tma_desc_k = {};
+    CUtensorMap tma_desc_v = {};
+    CUtensorMap tma_desc_o = {};
+
+    tma::create_tensor_map<st_bf<qo_height, tile_width, layout_q>, ATTN_B*ATTN_H*ATTN_N/(qo_height * 16)>(&tma_desc_q, d_q); 
+    tma::create_tensor_map<st_bf<kv_height, tile_width, layout_k>, ATTN_B*ATTN_H*ATTN_N/(kv_height * 16)>(&tma_desc_k, d_k);
+    tma::create_tensor_map<st_bf<kv_height, tile_width, layout_v>, ATTN_B*ATTN_H*ATTN_N/(kv_height * 16)>(&tma_desc_v, d_v);
+    tma::create_tensor_map<st_bf<qo_height, tile_width, layout_o>, ATTN_B*ATTN_H*ATTN_N/(qo_height * 16)>(&tma_desc_o, d_o);
+
+    CUtensorMap* tma_q_d; 
+    CUtensorMap* tma_k_d;
+    CUtensorMap* tma_v_d;
+    CUtensorMap* tma_o_d;
+
+    cudaMalloc(&tma_q_d, sizeof(CUtensorMap));
+    cudaMalloc(&tma_k_d, sizeof(CUtensorMap));
+    cudaMalloc(&tma_v_d, sizeof(CUtensorMap));
+    cudaMalloc(&tma_o_d, sizeof(CUtensorMap));
+
+    cudaMemcpy(tma_q_d, &tma_desc_q, sizeof(CUtensorMap), cudaMemcpyHostToDevice);
+    cudaMemcpy(tma_k_d, &tma_desc_k, sizeof(CUtensorMap), cudaMemcpyHostToDevice);
+    cudaMemcpy(tma_v_d, &tma_desc_v, sizeof(CUtensorMap), cudaMemcpyHostToDevice);
+    cudaMemcpy(tma_o_d, &tma_desc_o, sizeof(CUtensorMap), cudaMemcpyHostToDevice);
+
+    std::cout << "Allocated and set memory on GPU!" << std::endl;
+    
+    unsigned long mem_size = kittens::MAX_SHARED_MEMORY;
+    
+    cudaFuncSetAttribute(
+        attend_ker<ATTN_N>,
+        cudaFuncAttributeMaxDynamicSharedMemorySize,
+        mem_size
+    );
+    std::cout << "Set max dynamic memory!" << std::endl;
+
+    const int ITER = 100;
+    dim3 grid(ATTN_N/(NUM_WORKERS*kittens::TILE_DIM), ATTN_B*ATTN_H, 1);
+    static_assert(ATTN_N % (NUM_WORKERS*kittens::TILE_DIM) == 0);
+    cudaDeviceSynchronize();
+    std::cout << "Starting warmup" << std::endl;
+    for(int i = 0; i < ITER; i++) {
+        attend_ker<ATTN_N><<<grid, BLOCK_SIZE, mem_size>>>(tma_q_d, tma_k_d, tma_v_d, tma_o_d); 
+    }
+    cudaDeviceSynchronize();
+    std::cout << "Starting kernel" << std::endl;
+    const auto start = std::chrono::high_resolution_clock::now();
+    for(int i = 0; i < ITER; i++) {
+        attend_ker<ATTN_N><<<grid, BLOCK_SIZE, mem_size>>>(tma_q_d, tma_k_d, tma_v_d, tma_o_d); 
+    }
+    cudaDeviceSynchronize();
+    const auto finish = std::chrono::high_resolution_clock::now();
+    CudaCheckError();
+    std::cout << "Finished kernel\n";
+    
+    // check correctness
+    cudaMemcpy(o_bf, d_o, TOTAL_ELEMENTS * sizeof(bf16), cudaMemcpyDeviceToHost);
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) {
+        o[i] = __bfloat162float(o_bf[i]);
+    }
+
+    bool good = true;
+    std::ofstream o_ref_file("printouts/o_ref.txt");
+    std::ofstream o_file("printouts/o.txt");
+    std::ofstream diff_file("printouts/diff.txt");
+    for(int i = 0; i < TOTAL_UNIQUE_ELEMENTS; i++) {
+        float diff = o[i] - o_ref[i % TOTAL_UNIQUE_ELEMENTS];
+        o_ref_file << o_ref[i % TOTAL_UNIQUE_ELEMENTS] << ' ';
+        o_file << o[i] << ' ';
+        diff_file << diff << ' ';
+        if(abs(diff) > 0.01 || isnan(diff)) {
+            good = false;
+        }
+    }
+
+    std::cout << "Average execution time: " << std::chrono::duration_cast<std::chrono::microseconds>(finish - start).count() / ITER << " us" << std::endl;
+    if(good) std::cout << "Correct :)\n";
+    else std::cout << "Incorrect :(\n";
+
+    cudaFree(d_q);
+    cudaFree(d_k);
+    cudaFree(d_v);
+    cudaFree(d_o);
+
+    delete[] q, k, v, o, o_ref;
+    delete[] q_bf, k_bf, v_bf, o_bf;
+
+    return 0;
+}
\ No newline at end of file
diff --git a/examples/attn_fwd_tma_wgmma/tictoc_ker.cu b/examples/attn_fwd_tma_wgmma/tictoc_ker.cu
new file mode 100644
index 000000000..2faf6a467
--- /dev/null
+++ b/examples/attn_fwd_tma_wgmma/tictoc_ker.cu
@@ -0,0 +1,143 @@
+#define KITTENS_HOPPER // we are on an H100
+#include "../../src/kittens.cuh"
+
+constexpr int NUM_WORKERS = 20;
+constexpr int NUM_WARPGROUPS = (NUM_WORKERS/(kittens::WARPGROUP_WARPS));
+
+constexpr int qo_height = 4, kv_height = 4;
+constexpr int NUM_WORKERS_KV = 4;
+constexpr int tile_width = 64/16;
+
+using namespace kittens;
+
+using layout_q = ducks::st_layout::wgmma_row_0b;
+using layout_k = ducks::st_layout::wgmma_row_0b;
+using layout_v = ducks::st_layout::wgmma_col_t_0b;
+using layout_o = ducks::st_layout::naive; // tma_swizzle seems unreliable right now.
+
+template<int N> __global__  __launch_bounds__(NUM_WORKERS*kittens::WARP_THREADS, 1)
+void attend_ker(CUtensorMap* tma_q, CUtensorMap* tma_k, CUtensorMap* tma_v, CUtensorMap* tma_o) {
+    extern __shared__ int __shm[]; // this is the CUDA shared memory
+    tma_allocator al((int*)&__shm[0]);
+
+    st_bf<qo_height, tile_width, layout_q> (&q_smem)   [NUM_WARPGROUPS] = al.allocate<st_bf<qo_height, tile_width, layout_q>, NUM_WARPGROUPS>();
+    st_bf<kv_height, tile_width, layout_k> (&k_smem)[2][NUM_WORKERS_KV] = al.allocate<st_bf<kv_height, tile_width, layout_k>, 2, NUM_WORKERS_KV>();
+    st_bf<kv_height, tile_width, layout_v> (&v_smem)[2][NUM_WORKERS_KV] = al.allocate<st_bf<kv_height, tile_width, layout_v>, 2, NUM_WORKERS_KV>();
+
+    int tic = 0, toc = 1;
+ 
+    rt_fl<1, kv_height> att_block;
+    rt_bf<1, kv_height> att_block_mma;
+    rt_fl<1, tile_width> o_prev;
+    rt_fl<1, kv_height>::col_vec max_vec_last, max_vec;
+    rt_fl<1, kv_height>::col_vec norm_vec_last, norm_vec;
+
+    int warpid      = kittens::warpid();
+    int warpgroupid = warpid/kittens::WARPGROUP_WARPS; 
+
+    constexpr int qo_tiles  = N / q_smem[0].rows; 
+    constexpr int kv_blocks = N / (NUM_WORKERS_KV*k_smem[0][0].rows);
+
+    __shared__ uint64_t qsmem_barrier, ksmem_barrier, vsmem_barrier;
+
+    constexpr int tile_bytes = sizeof(bf16) * k_smem[0][0].num_elements * NUM_WORKERS_KV, kPhaseBit = 1;
+
+    if (warpid == 0) {
+        tma::init_barrier(qsmem_barrier, NUM_WORKERS*WARP_THREADS); 
+        tma::set_barrier_bytes(qsmem_barrier, tile_bytes); 
+
+        tma::init_barrier(ksmem_barrier, NUM_WORKERS*WARP_THREADS);
+        tma::set_barrier_bytes(ksmem_barrier, tile_bytes);
+
+        tma::init_barrier(vsmem_barrier, NUM_WORKERS*WARP_THREADS);
+        tma::set_barrier_bytes(vsmem_barrier, tile_bytes);
+    }
+    __syncthreads();
+
+    if (warpid == 0) {
+        for (int wg = 0; wg < NUM_WORKERS/kittens::WARPGROUP_WARPS; wg++) { // load q
+            int tile_idx = (blockIdx.y * NUM_WARPGROUPS * blockDim.x) + (blockIdx.x * NUM_WARPGROUPS) + wg;
+            tma::load_async((q_smem[wg]), tma_q, tile_idx, qsmem_barrier); 
+        }
+        for (int w = 0; w < NUM_WORKERS_KV; w++) { // load k, v      
+            int tile_idx = (blockIdx.y * NUM_WORKERS_KV * kv_blocks) + (0 * NUM_WORKERS_KV) + w; 
+            tma::load_async((k_smem[tic][w]), tma_k, tile_idx, ksmem_barrier); 
+            tma::load_async((v_smem[tic][w]), tma_v, tile_idx, vsmem_barrier); 
+        }
+    }
+
+    neg_infty(max_vec); // zero registers for the Q chunk
+    zero(norm_vec);
+    zero(o_prev);
+
+    tma::arrive_and_wait(qsmem_barrier, kPhaseBit); 
+    __syncthreads();
+    warpgroup::mul(q_smem[warpgroupid], q_smem[warpgroupid], __float2bfloat16(0.125f));
+
+    for(auto kv_idx = 0; kv_idx < kv_blocks; kv_idx++, tic ^= 1, toc ^= 1) {
+
+        tma::arrive_and_wait(ksmem_barrier, kPhaseBit); 
+        tma::arrive_and_wait(vsmem_barrier, kPhaseBit); 
+
+        if ((threadIdx.x == 0)) {
+            tma::init_barrier(ksmem_barrier, NUM_WORKERS*WARP_THREADS); 
+            tma::set_barrier_bytes(ksmem_barrier, tile_bytes); 
+
+            tma::init_barrier(vsmem_barrier, NUM_WORKERS*WARP_THREADS); 
+            tma::set_barrier_bytes(vsmem_barrier, tile_bytes); 
+        }
+        __syncthreads();
+
+        if ((kv_idx + 1 < kv_blocks) && (warpid == 0)) {
+            for (int w = 0; w < NUM_WORKERS_KV; w++) {        
+                int tile_idx = (blockIdx.y * NUM_WORKERS_KV * kv_blocks) + ((kv_idx + 1) * NUM_WORKERS_KV) + w; 
+                tma::load_async((k_smem[toc][w]), tma_k, tile_idx, ksmem_barrier); 
+                tma::load_async((v_smem[toc][w]), tma_v, tile_idx, vsmem_barrier); 
+            }
+        }
+
+        for(int subtile = 0; subtile < NUM_WORKERS_KV; subtile++) {
+            warpgroup::fence(att_block);
+            warpgroup::dot_reset(att_block, q_smem[warpgroupid], k_smem[tic][subtile]);
+            warpgroup::mma_commit_group();
+
+            copy(norm_vec_last, norm_vec);
+            copy(max_vec_last,  max_vec);
+
+            warpgroup::mma_async_wait();
+
+            row_max(max_vec, att_block, max_vec); // accumulate onto the max_vec
+            sub_row(att_block, att_block, max_vec);
+            exp(att_block, att_block);
+
+            sub(max_vec_last, max_vec_last, max_vec);
+            exp(max_vec_last, max_vec_last);
+            mul(norm_vec, norm_vec, max_vec_last);
+
+            row_sum(norm_vec, att_block, norm_vec); // accumulate onto the norm_vec
+            div_row(att_block, att_block, norm_vec);
+
+            mul(norm_vec_last, norm_vec_last, max_vec_last);
+            div(norm_vec_last, norm_vec_last, norm_vec);
+
+            copy(att_block_mma, att_block); // convert to bf16 for mma
+            mul_row(o_prev, o_prev, norm_vec_last); // normalize o_prev in advance of mma'ing onto it
+
+            warpgroup::fence(o_prev);
+            warpgroup::mma_accum(o_prev, att_block_mma, v_smem[tic][subtile]);
+            warpgroup::mma_commit_group();
+        }
+    }
+
+    auto *o_smem = reinterpret_cast<st_bf<qo_height, tile_width, layout_o>*>(&q_smem[0].data[0]); // reuse q memory
+    warpgroup::store(o_smem[warpgroupid], o_prev); 
+    __syncthreads();
+    if (warpid % 4 == 0) { // store o
+        int tile_idx = (blockIdx.y * NUM_WARPGROUPS * blockDim.x) + (blockIdx.x * NUM_WARPGROUPS) + warpgroupid; 
+        tma::store_async(tma_o, (o_smem[warpgroupid]), tile_idx); 
+        tma::store_commit_group(); 
+    }
+    tma::store_async_wait();
+}
+
+#include "harness.impl"
\ No newline at end of file
diff --git a/src/common/util.cuh b/src/common/util.cuh
index 24419518e..8959aa2e3 100644
--- a/src/common/util.cuh
+++ b/src/common/util.cuh
@@ -36,6 +36,12 @@ constexpr int WARP_THREADS{32};
  */
 constexpr int WARPGROUP_THREADS{128};
 /**
+
+ * @brief Constant representing number of warps in a warpgroup of four warps.
+ */
+constexpr int WARPGROUP_WARPS{4};
+/**
+
  * @brief Get the warp ID of the current thread.
  * @return The warp ID.
  */
diff --git a/src/ops/warp/shared/maps.cuh b/src/ops/warp/shared/maps.cuh
new file mode 100644
index 000000000..08b8751da
--- /dev/null
+++ b/src/ops/warp/shared/maps.cuh
@@ -0,0 +1,121 @@
+#pragma once
+
+#include "../../../common/common.cuh"
+#include "../../../types/types.cuh"
+
+namespace kittens {
+
+/* ----------  Uniform tile maps (independent of layout)  ---------- */
+
+// Uniform unary map on tile
+template<typename op, ducks::st::all T> // T2, w, h can be inferred from dst as long as op is specialized
+__device__ static inline void unary_map(T &dst, const T &src) {
+
+    int lane = threadIdx.x % 32; 
+    #pragma unroll
+    for(int i = lane; i < dst.rows*dst.cols; i += WARP_SIZE) {
+        int row = (i/dst.cols); 
+        int col = (i%dst.cols); 
+        dst.data[col + (row * dst.cols)] = op::template op<typename T::dtype>(src.data[col + (row * dst.cols)]);
+    }
+}
+
+// Uniform map from packed to tile
+template<typename op, ducks::st::all T>
+__device__ static inline void bin_map(T &dst, const T &src, const typename T::dtype &param) {
+
+    int lane = threadIdx.x % 32; 
+    #pragma unroll
+    for(int i = lane; i < dst.rows*dst.cols; i += WARP_SIZE) {
+        int row = (i/dst.cols); 
+        int col = (i%dst.cols);  
+        dst.data[col + (row * dst.cols)] = op::template op<typename T::dtype>(src.data[col + (row * dst.cols)], param);
+    }
+}
+// Uniform map from single to tile
+template<typename op, ducks::st::all T>
+__device__ static inline void bin_map(T &dst, const T &src, const typename base_types::packing<typename T::dtype>::unpacked_type &param) {
+    // The optimizing compiler should eliminate this pack in the 32-bit case but not in the 16-bit case
+    bin_map<op, T>(dst, src, base_types::packing<typename T::dtype>::pack(param));
+}
+// Tile-to-tile element-wise map. Only enforces same layout.
+template<typename op, ducks::st::all T>
+__device__ static inline void bin_map(T &dst, const T &lhs, const T &rhs) {
+
+    int lane = threadIdx.x % 32; 
+    #pragma unroll
+    for(int i = lane; i < dst.rows*dst.cols; i += WARP_SIZE) {
+        int row = (i/dst.cols); 
+        int col = (i%dst.cols); 
+        dst.data[col + (row * dst.cols)] = op::template op<typename T::dtype>(lhs.data[col + (row * dst.cols)], rhs.data[col + (row * dst.cols)]);
+    }
+}
+
+/* ----------  WRAPPERS FOR PRETTINESS  ---------- */
+
+// All of the annoying qualifiers *should* be automatically inferred during compile-time.
+// So, syntax should just be kittens::add_row(tile, colvec);
+
+// const maps
+template<ducks::st::all T>
+__device__ static inline void zero(T &dst) {
+    unary_map<base_ops::zero, T>(dst, dst);
+}
+template<ducks::st::all T>
+__device__ static inline void one(T &dst) {
+    unary_map<base_ops::one, T>(dst, dst);
+}
+template<ducks::st::all T>
+__device__ static inline void pos_infty(T &dst) {
+    unary_map<base_ops::pos_infty, T>(dst, dst);
+}
+template<ducks::st::all T>
+__device__ static inline void neg_infty(T &dst) {
+    unary_map<base_ops::neg_infty, T>(dst, dst);
+}
+
+// unary maps
+template<ducks::st::all T>
+__device__ static inline void exp(T &dst, const T &src) {
+    unary_map<base_ops::exp, T>(dst, src);
+}
+template<ducks::st::all T>
+__device__ static inline void abs(T &dst, const T &src) {
+    unary_map<base_ops::abs, T>(dst, src);
+}
+template<ducks::st::all T>
+__device__ static inline void relu(T &dst, const T &src) {
+    unary_map<base_ops::relu, T>(dst, src);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void copy(T &dst, const U &src) {
+    bin_map<base_ops::copy2, T>(dst, src);
+}
+
+// uniform binary maps
+template<ducks::st::all T, typename U>
+__device__ static inline void max(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::max, T>(dst, lhs, rhs);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void min(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::min, T>(dst, lhs, rhs);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void add(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::sum, T>(dst, lhs, rhs);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void sub(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::sub, T>(dst, lhs, rhs);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void mul(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::mul, T>(dst, lhs, rhs);
+}
+template<ducks::st::all T, typename U>
+__device__ static inline void div(T &dst, const T &lhs, const U &rhs) {
+    bin_map<base_ops::div, T>(dst, lhs, rhs);
+}
+
+}
\ No newline at end of file
diff --git a/src/ops/warpgroup/wgmma/wgmma.cuh b/src/ops/warpgroup/wgmma/wgmma.cuh
index 11d0bfc7d..42009d3a4 100644
--- a/src/ops/warpgroup/wgmma/wgmma.cuh
+++ b/src/ops/warpgroup/wgmma/wgmma.cuh
@@ -333,7 +333,6 @@ __device__ static inline void dot_reset(rt_fl<1, M, ducks::rt_layout::row> &d,
  */
 template<int height, int width>
 __device__ inline void fence(rt_fl<height, width, ducks::rt_layout::row> &dst) {
-    asm volatile ("fence.proxy.async.shared::cta;\n" ::: "memory"); // still not really sure if we need this
     #pragma unroll
     for(int i = 0; i < height; i++) {
         #pragma unroll
@@ -345,7 +344,7 @@ __device__ inline void fence(rt_fl<height, width, ducks::rt_layout::row> &dst) {
             }
         }
     }
-    asm volatile ("wgmma.fence.sync.aligned;\n" ::: "memory"); 
+    asm volatile ("wgmma.fence.sync.aligned;\n" ::: "memory");
 }
 
 /**
