diff --git a/c2h/generators.cu b/c2h/generators.cu
index e378bbc3119..13142563aa4 100644
--- a/c2h/generators.cu
+++ b/c2h/generators.cu
@@ -122,24 +122,24 @@ private:
 template <typename T>
 struct is_floating_point : ::cuda::std::is_floating_point<T>
 {};
-#ifdef _CCCL_HAS_NVFP16
+#if _CCCL_HAS_NVFP16()
 template <>
 struct is_floating_point<__half> : ::cuda::std::true_type
 {};
-#endif // _CCCL_HAS_NVFP16
-#ifdef _CCCL_HAS_NVBF16
+#endif // _CCCL_HAS_NVFP16()
+#if _CCCL_HAS_NVBF16()
 template <>
 struct is_floating_point<__nv_bfloat16> : ::cuda::std::true_type
 {};
-#endif // _CCCL_HAS_NVBF16
-#ifdef __CUDA_FP8_TYPES_EXIST__
+#endif // _CCCL_HAS_NVBF16()
+#if _CCCL_HAS_NVFP8()
 template <>
 struct is_floating_point<__nv_fp8_e4m3> : ::cuda::std::true_type
 {};
 template <>
 struct is_floating_point<__nv_fp8_e5m2> : ::cuda::std::true_type
 {};
-#endif // __CUDA_FP8_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP8()
 
 template <typename T, bool = is_floating_point<T>::value>
 struct random_to_item_t
@@ -542,10 +542,10 @@ INSTANTIATE(std::int16_t);
 INSTANTIATE(std::int32_t);
 INSTANTIATE(std::int64_t);
 
-#if defined(__CUDA_FP8_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP8()
 INSTANTIATE(__nv_fp8_e5m2);
 INSTANTIATE(__nv_fp8_e4m3);
-#endif // defined(__CUDA_FP8_TYPES_EXIST__)
+#endif // _CCCL_HAS_NVFP8()
 INSTANTIATE(float);
 INSTANTIATE(double);
 
diff --git a/c2h/include/c2h/extended_types.h b/c2h/include/c2h/extended_types.h
index 534de8806da..c038a6fa8bd 100644
--- a/c2h/include/c2h/extended_types.h
+++ b/c2h/include/c2h/extended_types.h
@@ -30,19 +30,19 @@
 #include <cuda/__cccl_config>
 
 #ifndef TEST_HALF_T
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 #    define TEST_HALF_T() 1
-#  else // defined(_CCCL_HAS_NVFP16)
+#  else
 #    define TEST_HALF_T() 0
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif
 #endif // TEST_HALF_T
 
 #ifndef TEST_BF_T
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 #    define TEST_BF_T() 1
-#  else // defined(_CCCL_HAS_NVBF16)
+#  else
 #    define TEST_BF_T() 0
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif
 #endif // TEST_BF_T
 
 #if TEST_HALF_T()
diff --git a/c2h/include/c2h/generators.h b/c2h/include/c2h/generators.h
index 07dc6f49d81..057616f9615 100644
--- a/c2h/include/c2h/generators.h
+++ b/c2h/include/c2h/generators.h
@@ -35,23 +35,25 @@
 #include <c2h/vector.h>
 
 #if THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 #    include <cuda_fp16.h>
-#  endif // _CCCL_HAS_NVFP16
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #    include <cuda_bf16.h>
 _CCCL_DIAG_POP
+#  endif // _CCCL_HAS_NVBF16
 
+#  if _CCCL_HAS_NVFP8()
 // cuda_fp8.h resets default for C4127, so we have to guard the inclusion
 _CCCL_DIAG_PUSH
 #    include <cuda_fp8.h>
 _CCCL_DIAG_POP
-#  endif // _CCCL_HAS_NVBF16
+#  endif // _CCCL_HAS_NVFP8()
 
-#  if defined(__CUDA_FP8_TYPES_EXIST__)
+#  if _CCCL_HAS_NVFP8()
 namespace std
 {
 template <>
@@ -84,7 +86,7 @@ class numeric_limits<__nv_fp8_e5m2>
   }
 };
 } // namespace std
-#  endif // defined(__CUDA_FP8_TYPES_EXIST__)
+#  endif // _CCCL_HAS_NVFP8()
 #endif // THRUST_DEVICE_SYSTEM == THRUST_DEVICE_SYSTEM_CUDA
 
 namespace c2h
diff --git a/cmake/header_test.cu.in b/cmake/header_test.cu.in
index 051cecb18aa..55b665c7ab1 100644
--- a/cmake/header_test.cu.in
+++ b/cmake/header_test.cu.in
@@ -7,9 +7,9 @@
 // a potential macro collision and halts.
 //
 // Hacky way to build a string, but it works on all tested platforms.
-#define CCCL_HEADER_MACRO_CHECK(MACRO, HEADER)                                      \
-  CCCL_HEADER_MACRO_CHECK_IMPL(Identifier MACRO should not be used from Thrust      \
-                          headers due to conflicts with HEADER macros.)
+#define CCCL_HEADER_MACRO_CHECK(MACRO, HEADER) \
+  CCCL_HEADER_MACRO_CHECK_IMPL(                \
+    Identifier MACRO should not be used from Thrust headers due to conflicts with HEADER macros.)
 
 // Use raw platform macros instead of the CCCL macros since we
 // don't want to #include any headers other than the one being tested.
@@ -18,20 +18,19 @@
 #if defined(_MSC_VER) // MSVC
 
 // Fake up an error for MSVC
-#define CCCL_HEADER_MACRO_CHECK_IMPL(msg)                                           \
-  /* Print message that looks like an error: */                                     \
-  __pragma(message(__FILE__ ":" CCCL_HEADER_MACRO_CHECK_IMPL0(__LINE__)             \
-                   ": error: " #msg))                                               \
-  /* abort compilation due to static_assert or syntax error: */                     \
-  static_assert(false, #msg);
-#define CCCL_HEADER_MACRO_CHECK_IMPL0(x) CCCL_HEADER_MACRO_CHECK_IMPL1(x)
-#define CCCL_HEADER_MACRO_CHECK_IMPL1(x) #x
+#  define CCCL_HEADER_MACRO_CHECK_IMPL(msg)                                                  \
+    /* Print message that looks like an error: */                                            \
+    __pragma(message(__FILE__ ":" CCCL_HEADER_MACRO_CHECK_IMPL0(__LINE__) ": error: " #msg)) \
+                                                                                             \
+      static_assert(false, #msg); /* abort compilation due to static_assert or syntax error */
+#  define CCCL_HEADER_MACRO_CHECK_IMPL0(x) CCCL_HEADER_MACRO_CHECK_IMPL1(x)
+#  define CCCL_HEADER_MACRO_CHECK_IMPL1(x) #x
 
 #elif defined(__clang__) || defined(__GNUC__)
 
 // GCC/clang are easy:
-#define CCCL_HEADER_MACRO_CHECK_IMPL(msg) CCCL_HEADER_MACRO_CHECK_IMPL0(GCC error #msg)
-#define CCCL_HEADER_MACRO_CHECK_IMPL0(expr) _Pragma(#expr)
+#  define CCCL_HEADER_MACRO_CHECK_IMPL(msg)   CCCL_HEADER_MACRO_CHECK_IMPL0(GCC error #msg)
+#  define CCCL_HEADER_MACRO_CHECK_IMPL0(expr) _Pragma(#expr)
 
 #endif // msvc vs. the world
 
@@ -39,40 +38,52 @@
 #ifndef CCCL_IGNORE_HEADER_MACRO_CHECKS
 
 // complex.h conflicts
-#define I CCCL_HEADER_MACRO_CHECK('I', complex.h)
+#  define I CCCL_HEADER_MACRO_CHECK('I', complex.h)
 
 // windows.h conflicts
-#define small CCCL_HEADER_MACRO_CHECK('small', windows.h)
+#  define small CCCL_HEADER_MACRO_CHECK('small', windows.h)
 // We can't enable these checks without breaking some builds -- some standard
 // library implementations unconditionally `#undef` these macros, which then
 // causes random failures later.
 // Leaving these commented out as a warning: Here be dragons.
-//#define min(...) CCCL_HEADER_MACRO_CHECK('min', windows.h)
-//#define max(...) CCCL_HEADER_MACRO_CHECK('max', windows.h)
+// #define min(...) CCCL_HEADER_MACRO_CHECK('min', windows.h)
+// #define max(...) CCCL_HEADER_MACRO_CHECK('max', windows.h)
 
-#ifdef _WIN32
+#  ifdef _WIN32
 // On Windows, make sure any include of Windows.h (e.g. via NVTX) does not define the checked macros
-#  define WIN32_LEAN_AND_MEAN
-#endif // _WIN32
+#    define WIN32_LEAN_AND_MEAN
+#  endif // _WIN32
 
 // termios.h conflicts (NVIDIA/thrust#1547)
-#define B0 CCCL_HEADER_MACRO_CHECK("B0", termios.h)
+#  define B0 CCCL_HEADER_MACRO_CHECK("B0", termios.h)
 
 #endif // CCCL_IGNORE_HEADER_MACRO_CHECKS
 
 #include <@header@>
 
+#if defined(CCCL_DISABLE_NVFP8_SUPPORT)
+#  if defined(__CUDA_FP8_TYPES_EXIST__)
+#    error We should not include cuda_fp8.h when FP8 support is disabled
+#  endif // __CUDA_FP16_TYPES_EXIST__
+#endif // CCCL_DISABLE_BF16_SUPPORT
+
 #if defined(CCCL_DISABLE_BF16_SUPPORT)
-#if defined(__CUDA_BF16_TYPES_EXIST__)
-#error We should not include cuda_bf16.h when BF16 support is disabled
-#endif // __CUDA_BF16_TYPES_EXIST__
+#  if defined(__CUDA_BF16_TYPES_EXIST__)
+#    error We should not include cuda_bf16.h when BF16 support is disabled
+#  endif // __CUDA_BF16_TYPES_EXIST__
+#  if defined(__CUDA_FP8_TYPES_EXIST__)
+#    error We should not include cuda_fp8.h when BF16 support is disabled
+#  endif // __CUDA_FP16_TYPES_EXIST__
 #endif // CCCL_DISABLE_BF16_SUPPORT
 
 #if defined(CCCL_DISABLE_FP16_SUPPORT)
-#if defined(__CUDA_FP16_TYPES_EXIST__)
-#error We should not include cuda_fp16.h when half support is disabled
-#endif // __CUDA_FP16_TYPES_EXIST__
-#if defined(__CUDA_BF16_TYPES_EXIST__)
-#error We should not include cuda_bf16.h when half support is disabled
-#endif // __CUDA_BF16_TYPES_EXIST__
+#  if defined(__CUDA_FP8_TYPES_EXIST__)
+#    error We should not include cuda_fp8.h when half support is disabled
+#  endif // __CUDA_FP16_TYPES_EXIST__
+#  if defined(__CUDA_FP16_TYPES_EXIST__)
+#    error We should not include cuda_fp16.h when half support is disabled
+#  endif // __CUDA_FP16_TYPES_EXIST__
+#  if defined(__CUDA_BF16_TYPES_EXIST__)
+#    error We should not include cuda_bf16.h when half support is disabled
+#  endif // __CUDA_BF16_TYPES_EXIST__
 #endif // CCCL_DISABLE_FP16_SUPPORT
diff --git a/cub/cmake/CubHeaderTesting.cmake b/cub/cmake/CubHeaderTesting.cmake
index bf2cf6e2e99..21f281dd07f 100644
--- a/cub/cmake/CubHeaderTesting.cmake
+++ b/cub/cmake/CubHeaderTesting.cmake
@@ -47,3 +47,10 @@ set(header_definitions
   "CUB_WRAPPED_NAMESPACE=wrapped_cub"
   "CCCL_DISABLE_FP16_SUPPORT")
 cub_add_header_test(no_half "${header_definitions}")
+
+# Check that half support can be disabled
+set(header_definitions
+  "THRUST_WRAPPED_NAMESPACE=wrapped_thrust"
+  "CUB_WRAPPED_NAMESPACE=wrapped_cub"
+  "CCCL_DISABLE_FP8_SUPPORT")
+cub_add_header_test(no_fp8 "${header_definitions}")
diff --git a/cub/cub/agent/agent_sub_warp_merge_sort.cuh b/cub/cub/agent/agent_sub_warp_merge_sort.cuh
index 9f98ac42e3b..5f7238b035c 100644
--- a/cub/cub/agent/agent_sub_warp_merge_sort.cuh
+++ b/cub/cub/agent/agent_sub_warp_merge_sort.cuh
@@ -128,7 +128,7 @@ class AgentSubWarpSort
       _CCCL_UNREACHABLE();
     }
 
-#if defined(__CUDA_FP16_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP16()
     _CCCL_DEVICE bool operator()(__half lhs, __half rhs) const noexcept
     {
       // Need to explicitly cast to float for SM <= 52.
@@ -142,16 +142,16 @@ class AgentSubWarpSort
       }
       _CCCL_UNREACHABLE();
     }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
   };
 
-#if defined(__CUDA_FP16_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP16()
   _CCCL_DEVICE static bool equal(__half lhs, __half rhs)
   {
     // Need to explicitly cast to float for SM <= 52.
     NV_IF_TARGET(NV_PROVIDES_SM_53, (return __heq(lhs, rhs);), (return __half2float(lhs) == __half2float(rhs);));
   }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
 
   template <typename T>
   _CCCL_DEVICE static bool equal(T lhs, T rhs)
diff --git a/cub/cub/device/dispatch/dispatch_histogram.cuh b/cub/cub/device/dispatch/dispatch_histogram.cuh
index e43923bd562..951a9149f26 100644
--- a/cub/cub/device/dispatch/dispatch_histogram.cuh
+++ b/cub/cub/device/dispatch/dispatch_histogram.cuh
@@ -712,7 +712,7 @@ public:
       return this->ComputeScale(num_levels, max_level, min_level, ::cuda::std::is_floating_point<T>{});
     }
 
-#ifdef __CUDA_FP16_TYPES_EXIST__
+#if _CCCL_HAS_NVFP16()
     _CCCL_HOST_DEVICE _CCCL_FORCEINLINE ScaleT ComputeScale(int num_levels, __half max_level, __half min_level)
     {
       ScaleT result;
@@ -722,7 +722,7 @@ public:
                       static_cast<float>(num_levels - 1) / (__half2float(max_level) - __half2float(min_level)));))
       return result;
     }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
 
     // All types but __half:
     template <typename T>
@@ -731,7 +731,7 @@ public:
       return sample >= min_level && sample < max_level;
     }
 
-#ifdef __CUDA_FP16_TYPES_EXIST__
+#if _CCCL_HAS_NVFP16()
     _CCCL_HOST_DEVICE _CCCL_FORCEINLINE int SampleIsValid(__half sample, __half max_level, __half min_level)
     {
       NV_IF_TARGET(
@@ -739,7 +739,7 @@ public:
         (return __hge(sample, min_level) && __hlt(sample, max_level);),
         (return __half2float(sample) >= __half2float(min_level) && __half2float(sample) < __half2float(max_level);));
     }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
 
     /**
      * @brief Bin computation for floating point (and extended floating point) types
@@ -778,7 +778,7 @@ public:
       return this->ComputeBin(sample, min_level, scale, ::cuda::std::is_floating_point<T>{});
     }
 
-#ifdef __CUDA_FP16_TYPES_EXIST__
+#if _CCCL_HAS_NVFP16()
     _CCCL_HOST_DEVICE _CCCL_FORCEINLINE int ComputeBin(__half sample, __half min_level, ScaleT scale)
     {
       NV_IF_TARGET(
@@ -786,7 +786,7 @@ public:
         (return static_cast<int>(__hmul(__hsub(sample, min_level), scale.reciprocal));),
         (return static_cast<int>((__half2float(sample) - __half2float(min_level)) * __half2float(scale.reciprocal));));
     }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
 
     _CCCL_HOST_DEVICE _CCCL_FORCEINLINE bool
     MayOverflow(CommonT /* num_bins */, ::cuda::std::false_type /* is_integral */)
diff --git a/cub/cub/thread/thread_operators.cuh b/cub/cub/thread/thread_operators.cuh
index ba189a9ad13..37539d541c8 100644
--- a/cub/cub/thread/thread_operators.cuh
+++ b/cub/cub/thread/thread_operators.cuh
@@ -56,16 +56,16 @@
 #include <cuda/std/type_traits> // cuda::std::common_type
 #include <cuda/std/utility> // cuda::std::forward
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
-#endif // _CCCL_HAS_NVFP16
+#endif // _CCCL_HAS_NVFP16()
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #  include <cuda_bf16.h>
 _CCCL_DIAG_POP
-#endif // _CCCL_HAS_NVFP16
+#endif // _CCCL_HAS_NVBF16()
 
 CUB_NAMESPACE_BEGIN
 
@@ -415,7 +415,7 @@ struct SimdMin<::cuda::std::uint16_t>
   }
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <>
 struct SimdMin<__half>
@@ -436,9 +436,9 @@ struct SimdMin<__half>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 // NOTE: __halves2bfloat162 is not always available on older CUDA Toolkits for __CUDA_ARCH__ < 800
 _CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE __nv_bfloat162 halves2bfloat162(__nv_bfloat16 a, __nv_bfloat16 b)
@@ -472,7 +472,7 @@ struct SimdMin<__nv_bfloat16>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 //----------------------------------------------------------------------------------------------------------------------
 
@@ -506,7 +506,7 @@ struct SimdMax<::cuda::std::uint16_t>
   }
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <>
 struct SimdMax<__half>
@@ -527,9 +527,9 @@ struct SimdMax<__half>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 template <>
 struct SimdMax<__nv_bfloat16>
@@ -551,7 +551,7 @@ struct SimdMax<__nv_bfloat16>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 //----------------------------------------------------------------------------------------------------------------------
 
@@ -561,7 +561,7 @@ struct SimdSum
   static_assert(cub::detail::always_false<T>(), "Unsupported specialization");
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <>
 struct SimdSum<__half>
@@ -581,9 +581,9 @@ struct SimdSum<__half>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 template <>
 struct SimdSum<__nv_bfloat16>
@@ -605,7 +605,7 @@ struct SimdSum<__nv_bfloat16>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 //----------------------------------------------------------------------------------------------------------------------
 
@@ -615,7 +615,7 @@ struct SimdMul
   static_assert(cub::detail::always_false<T>(), "Unsupported specialization");
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <>
 struct SimdMul<__half>
@@ -635,9 +635,9 @@ struct SimdMul<__half>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 template <>
 struct SimdMul<__nv_bfloat16>
@@ -658,7 +658,7 @@ struct SimdMul<__nv_bfloat16>
   }
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 //----------------------------------------------------------------------------------------------------------------------
 
diff --git a/cub/cub/thread/thread_reduce.cuh b/cub/cub/thread/thread_reduce.cuh
index d3850051ca7..af79099547e 100644
--- a/cub/cub/thread/thread_reduce.cuh
+++ b/cub/cub/thread/thread_reduce.cuh
@@ -54,16 +54,16 @@
 #include <cuda/std/cstdint> // uint16_t
 #include <cuda/std/functional> // cuda::std::plus
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
-#endif // _CCCL_HAS_NVFP16
+#endif // _CCCL_HAS_NVFP16()
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #  include <cuda_bf16.h>
 _CCCL_DIAG_POP
-#endif // _CCCL_HAS_NVFP16
+#endif // _CCCL_HAS_NVBF16()
 
 CUB_NAMESPACE_BEGIN
 
@@ -235,7 +235,7 @@ struct enable_generic_simd_reduction_traits
       is_one_of<ReductionOp, ::cuda::minimum<>, ::cuda::minimum<T>, ::cuda::maximum<>, ::cuda::maximum<T>>();
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <typename ReductionOp>
 struct enable_generic_simd_reduction_traits<__half, ReductionOp>
@@ -251,9 +251,9 @@ struct enable_generic_simd_reduction_traits<__half, ReductionOp>
     ::cuda::std::multiplies<>,
     ::cuda::std::multiplies<__half>>();
 };
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 template <typename ReductionOp>
 struct enable_generic_simd_reduction_traits<__nv_bfloat16, ReductionOp>
@@ -270,7 +270,7 @@ struct enable_generic_simd_reduction_traits<__nv_bfloat16, ReductionOp>
     ::cuda::std::multiplies<__nv_bfloat16>>();
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 template <typename Input, typename ReductionOp>
 _CCCL_NODISCARD _CCCL_DEVICE constexpr bool enable_generic_simd_reduction()
@@ -306,11 +306,11 @@ _CCCL_NODISCARD _CCCL_DEVICE constexpr bool enable_sm80_simd_reduction()
                    ::cuda::std::multiplies<>,
                    ::cuda::std::multiplies<T>>()
       && Length >= 4
-#  if defined(_CCCL_HAS_NVFP16) && defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVFP16() && _CCCL_HAS_NVBF16()
       && (is_same<T, __half>::value || is_same<T, __nv_bfloat16>::value)
-#  elif defined(_CCCL_HAS_NVFP16)
+#  elif _CCCL_HAS_NVFP16()
       && is_same<T, __half>::value
-#  elif defined(_CCCL_HAS_NVBF16)
+#  elif _CCCL_HAS_NVBF16()
       && is_same<T, __nv_bfloat16>::value
 #  endif
     ;
@@ -321,7 +321,7 @@ _CCCL_NODISCARD _CCCL_DEVICE constexpr bool enable_sm70_simd_reduction()
 {
   using cub::detail::is_one_of;
   using ::cuda::std::is_same;
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
   return is_same<T, __half>::value
       && is_one_of<ReductionOp,
                    ::cuda::std::plus<>,
@@ -393,7 +393,7 @@ struct enable_ternary_reduction_sm90
       ::cuda::std::bit_xor<T>>();
 };
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 
 template <typename ReductionOp>
 struct enable_ternary_reduction_sm90<__half2, ReductionOp>
@@ -408,9 +408,9 @@ struct enable_ternary_reduction_sm90<__half2, ReductionOp>
                            SimdMax<__half>>();
 };
 
-#  endif // defined(_CCCL_HAS_NVFP16)
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 
 template <typename ReductionOp>
 struct enable_ternary_reduction_sm90<__nv_bfloat162, ReductionOp>
@@ -425,7 +425,7 @@ struct enable_ternary_reduction_sm90<__nv_bfloat162, ReductionOp>
                            SimdMax<__nv_bfloat16>>();
 };
 
-#  endif // defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVBF16()
 
 template <typename Input, typename ReductionOp, typename AccumT>
 _CCCL_NODISCARD _CCCL_DEVICE _CCCL_FORCEINLINE _CCCL_CONSTEXPR_CXX14 bool enable_ternary_reduction()
diff --git a/cub/cub/thread/thread_search.cuh b/cub/cub/thread/thread_search.cuh
index 802d4ec96f8..6471a55b351 100644
--- a/cub/cub/thread/thread_search.cuh
+++ b/cub/cub/thread/thread_search.cuh
@@ -154,7 +154,7 @@ _CCCL_DEVICE _CCCL_FORCEINLINE OffsetT UpperBound(InputIteratorT input, OffsetT
   return retval;
 }
 
-#if defined(__CUDA_FP16_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP16()
 /**
  * @param[in] input
  *   Input sequence
@@ -191,6 +191,6 @@ _CCCL_DEVICE _CCCL_FORCEINLINE OffsetT UpperBound(InputIteratorT input, OffsetT
 
   return retval;
 }
-#endif // __CUDA_FP16_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP16()
 
 CUB_NAMESPACE_END
diff --git a/cub/cub/util_type.cuh b/cub/cub/util_type.cuh
index 229d7ea9a38..42fb019f57f 100644
--- a/cub/cub/util_type.cuh
+++ b/cub/cub/util_type.cuh
@@ -49,21 +49,23 @@
 #include <cuda/std/limits>
 #include <cuda/std/type_traits>
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
-#endif // _CCCL_HAS_NVFP16
+#endif // _CCCL_HAS_NVFP16()
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #  include <cuda_bf16.h>
 _CCCL_DIAG_POP
+#endif // _CCCL_HAS_NVBF16()
 
 // cuda_fp8.h resets default for C4127, so we have to guard the inclusion
+#if _CCCL_HAS_NVFP8()
 _CCCL_DIAG_PUSH
 #  include <cuda_fp8.h>
 _CCCL_DIAG_POP
-#endif // _CCCL_HAS_NV_BF16
+#endif // _CCCL_HAS_NVFP8()
 
 #if _CCCL_COMPILER(NVRTC)
 #  include <cuda/std/iterator>
@@ -1039,17 +1041,17 @@ struct NumericTraits<__int128_t>
 
 template <> struct NumericTraits<float> :               BaseTraits<FLOATING_POINT, true, false, unsigned int, float> {};
 template <> struct NumericTraits<double> :              BaseTraits<FLOATING_POINT, true, false, unsigned long long, double> {};
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
     template <> struct NumericTraits<__half> :          BaseTraits<FLOATING_POINT, true, false, unsigned short, __half> {};
-#  endif // _CCCL_HAS_NVFP16
-#  if defined(_CCCL_HAS_NVBF16)
+#  endif // _CCCL_HAS_NVFP16()
+#  if _CCCL_HAS_NVBF16()
     template <> struct NumericTraits<__nv_bfloat16> :   BaseTraits<FLOATING_POINT, true, false, unsigned short, __nv_bfloat16> {};
-#  endif // _CCCL_HAS_NVBF16
+#  endif // _CCCL_HAS_NVBF16()
 
-#if defined(__CUDA_FP8_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP8()
     template <> struct NumericTraits<__nv_fp8_e4m3> :   BaseTraits<FLOATING_POINT, true, false, __nv_fp8_storage_t, __nv_fp8_e4m3> {};
     template <> struct NumericTraits<__nv_fp8_e5m2> :   BaseTraits<FLOATING_POINT, true, false, __nv_fp8_storage_t, __nv_fp8_e5m2> {};
-#endif // __CUDA_FP8_TYPES_EXIST__
+#endif // _CCCL_HAS_NVFP8()
 
 template <> struct NumericTraits<bool> :                BaseTraits<UNSIGNED_INTEGER, true, false, typename UnitWord<bool>::VolatileWord, bool> {};
 // clang-format on
diff --git a/cub/test/catch2_test_block_radix_sort.cu b/cub/test/catch2_test_block_radix_sort.cu
index 403451d6749..b952799af7d 100644
--- a/cub/test/catch2_test_block_radix_sort.cu
+++ b/cub/test/catch2_test_block_radix_sort.cu
@@ -32,7 +32,7 @@
 #include <utility>
 
 #include "catch2_test_block_radix_sort.cuh"
-#include <c2h/catch2_test_helper.h> // __CUDA_FP8_TYPES_EXIST__
+#include <c2h/catch2_test_helper.h> // _CCCL_HAS_NVFP8()
 
 // %PARAM% TEST_MEMOIZE mem 0:1
 // %PARAM% TEST_ALGORITHM alg 0:1
@@ -44,11 +44,11 @@ using types =
                  std::uint16_t,
                  std::uint32_t,
                  std::uint64_t
-#if defined(__CUDA_FP8_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP8()
                  ,
                  __nv_fp8_e5m2,
                  __nv_fp8_e4m3
-#endif // defined(__CUDA_FP8_TYPES_EXIST__)
+#endif // _CCCL_HAS_NVFP8()
                  >;
 using no_value_types = c2h::type_list<cub::NullType>;
 
@@ -59,11 +59,11 @@ using key_types =
                  std::int64_t,
                  float,
                  double
-#if defined(__CUDA_FP8_TYPES_EXIST__)
+#if _CCCL_HAS_NVFP8()
                  ,
                  __nv_fp8_e5m2,
                  __nv_fp8_e4m3
-#endif // defined(__CUDA_FP8_TYPES_EXIST__)
+#endif // _CCCL_HAS_NVFP8()
                  >;
 using value_types = c2h::type_list<std::int8_t, c2h::custom_type_t<c2h::equal_comparable_t>>;
 
diff --git a/docs/cccl_development/macro.rst b/docs/cccl_development/macro.rst
index 6b832da0000..02240d70f7c 100644
--- a/docs/cccl_development/macro.rst
+++ b/docs/cccl_development/macro.rst
@@ -178,26 +178,32 @@ Non-standard Types Support
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
 | ``_CCCL_HAS_INT128()``       | ``__int128`` and ``__uint128_t`` for 128-bit integer are supported and enabled                                                |
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
-| ``_CCCL_HAS_FLOAT128()``     | ``__float128`` for 128-bit floating-point are supported and enabled                                                           |
+| ``_CCCL_HAS_NVFP8()``        | ``__nv_fp8_e5m2/__nv_fp8_e4m3/__nv_fp8_e8m0`` data types are supported and enabled.  Prefer over ``__CUDA_FP8_TYPES_EXIST__`` |
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
-| ``_CCCL_HAS_NVFP16``         | ``__half/__half2`` data types are supported and enabled. Prefer over ``__CUDA_FP16_TYPES_EXIST__``                            |
+| ``_CCCL_HAS_NVFP16()``       | ``__half/__half2`` data types are supported and enabled. Prefer over ``__CUDA_FP16_TYPES_EXIST__``                            |
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
-| ``_CCCL_HAS_NVBF16``         | ``__nv_bfloat16/__nv_bfloat162`` data types are supported and enabled.  Prefer over ``__CUDA_BF16_TYPES_EXIST__``             |
+| ``_CCCL_HAS_NVBF16()``       | ``__nv_bfloat16/__nv_bfloat162`` data types are supported and enabled.  Prefer over ``__CUDA_BF16_TYPES_EXIST__``             |
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
-| ``_CCCL_HAS_NVFP8()``        | ``__nv_fp8_e5m2/__nv_fp8_e4m3/__nv_fp8_e8m0`` data types are supported and enabled.  Prefer over ``__CUDA_FP8_TYPES_EXIST__`` |
+| ``_CCCL_HAS_FLOAT128()``     | ``__float128`` for 128-bit floating-point are supported and enabled                                                           |
 +------------------------------+-------------------------------------------------------------------------------------------------------------------------------+
 
-+---------------------------------+-------------------------------------------------------------------------+
-| ``_CCCL_DISABLE_NVFP8_SUPPORT`` | Disable ``__nv_fp8_e5m2/__nv_fp8_e4m3/__nv_fp8_e8m0`` support           |
-+---------------------------------+-------------------------------------------------------------------------+
-| ``_CCCL_DISABLE_INT128``        | Disable ``__int128/__uint128_t`` support                                |
-+---------------------------------+-------------------------------------------------------------------------+
-| ``_CCCL_DISABLE_FLOAT128``      | Disable ``__float128`` support                                          |
-+---------------------------------+-------------------------------------------------------------------------+
-| ``_LIBCUDACXX_HAS_NVFP16``      | ``__half/__half2`` host/device are supported  (CUDA 12.2+)              |
-+---------------------------------+-------------------------------------------------------------------------+
-| ``_LIBCUDACXX_HAS_NVBF16``      | ``__nv_bfloat16/__nv_bfloat162`` host/device are supported (CUDA 12.2+) |
-+---------------------------------+-------------------------------------------------------------------------+
++-----------------------------------+-------------------------------------------------------------------------+
+| ``CCCL_DISABLE_INT128_SUPPORT``   | Disable ``__int128/__uint128_t`` support                                |
++-----------------------------------+-------------------------------------------------------------------------+
+| ``CCCL_DISABLE_NVFP8_SUPPORT``    | Disable ``__nv_fp8_e5m2/__nv_fp8_e4m3/__nv_fp8_e8m0`` support           |
++-----------------------------------+-------------------------------------------------------------------------+
+| ``CCCL_DISABLE_NVFP16_SUPPORT``   | Disable ``__half/__half2`` support                                      |
++-----------------------------------+-------------------------------------------------------------------------+
+| ``CCCL_DISABLE_NVBF16_SUPPORT``   | Disable ``__nv_bfloat16/__nv_bfloat162`` support                        |
++-----------------------------------+-------------------------------------------------------------------------+
+| ``CCCL_DISABLE_FLOAT128_SUPPORT`` | Disable ``__float128`` support                                          |
++-----------------------------------+-------------------------------------------------------------------------+
+
++----------------------------------+-------------------------------------------------------------------------+
+| ``_LIBCUDACXX_HAS_NVFP16``       | ``__half/__half2`` host/device are supported  (CUDA 12.2+)              |
++----------------------------------+-------------------------------------------------------------------------+
+| ``_LIBCUDACXX_HAS_NVBF16``       | ``__nv_bfloat16/__nv_bfloat162`` host/device are supported (CUDA 12.2+) |
++----------------------------------+-------------------------------------------------------------------------+
 
 ----
 
diff --git a/libcudacxx/include/cuda/std/__cccl/extended_data_types.h b/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
index 034ebc51b5f..1354389e24c 100644
--- a/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
+++ b/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
@@ -26,59 +26,41 @@
 #include <cuda/std/__cccl/os.h>
 #include <cuda/std/__cccl/preprocessor.h>
 
-#if !defined(_CCCL_DISABLE_INT128)
-#  if _CCCL_COMPILER(NVRTC) && defined(__CUDACC_RTC_INT128__) && _CCCL_OS(LINUX)
-#    define _CCCL_HAS_INT128() 1
-#  elif defined(__SIZEOF_INT128__) && _CCCL_OS(LINUX)
-#    define _CCCL_HAS_INT128() 1
-#  else
-#    define _CCCL_HAS_INT128() 0
-#  endif
-#else
-#  define _CCCL_HAS_INT128() 0
-#endif // !_CCCL_DISABLE_INT128
+#define _CCCL_HAS_INT128()   0
+#define _CCCL_HAS_NVFP8()    0
+#define _CCCL_HAS_NVFP16()   0
+#define _CCCL_HAS_NVBF16()   0
+#define _CCCL_HAS_FLOAT128() 0
 
-#if !defined(_CCCL_HAS_NVFP16)
-#  if _CCCL_HAS_INCLUDE(<cuda_fp16.h>) && (_CCCL_HAS_CUDA_COMPILER || defined(LIBCUDACXX_ENABLE_HOST_NVFP16)) \
-                        && !defined(CCCL_DISABLE_FP16_SUPPORT)
-#    define _CCCL_HAS_NVFP16 1
-#  endif
-#endif // !_CCCL_HAS_NVFP16
+#if !defined(CCCL_DISABLE_INT128_SUPPORT) && _CCCL_OS(LINUX) \
+  && ((_CCCL_COMPILER(NVRTC) && defined(__CUDACC_RTC_INT128__)) || defined(__SIZEOF_INT128__))
+#  undef _CCCL_HAS_INT128
+#  define _CCCL_HAS_INT128() 1
+#endif
 
-#if !defined(_CCCL_HAS_NVBF16)
-#  if _CCCL_HAS_INCLUDE(<cuda_bf16.h>) && defined(_CCCL_HAS_NVFP16) && !defined(CCCL_DISABLE_BF16_SUPPORT) \
-                        && !defined(CUB_DISABLE_BF16_SUPPORT)
-#    define _CCCL_HAS_NVBF16 1
-#  endif
-#endif // !_CCCL_HAS_NVBF16
+#if _CCCL_HAS_INCLUDE(<cuda_fp16.h>) && (_CCCL_HAS_CUDA_COMPILER || defined(LIBCUDACXX_ENABLE_HOST_NVFP16)) \
+                      && !defined(CCCL_DISABLE_FP16_SUPPORT)
+#  undef _CCCL_HAS_NVFP16
+#  define _CCCL_HAS_NVFP16() 1
+#endif
 
-#if !defined(_CCCL_DISABLE_NVFP8_SUPPORT)
-#  if _CCCL_HAS_INCLUDE(<cuda_fp8.h>) && defined(_CCCL_HAS_NVFP16) && defined(_CCCL_HAS_NVBF16)
-#    define _CCCL_HAS_NVFP8() 1
-#  else
-#    define _CCCL_HAS_NVFP8() 0
-#  endif // _CCCL_HAS_INCLUDE(<cuda_fp8.h>) && defined(_CCCL_HAS_NVFP16) && defined(_CCCL_HAS_NVBF16)
-#else
-#  define _CCCL_HAS_NVFP8() 0
-#endif // !defined(_CCCL_DISABLE_NVFP8_SUPPORT)
+#if _CCCL_HAS_INCLUDE(<cuda_bf16.h>) && _CCCL_HAS_NVFP16() && !defined(CCCL_DISABLE_BF16_SUPPORT)
+#  undef _CCCL_HAS_NVBF16
+#  define _CCCL_HAS_NVBF16() 1
+#endif
 
-#if !defined(_CCCL_DISABLE_FLOAT128)
-#  if _CCCL_COMPILER(NVRTC) && defined(__CUDACC_RTC_FLOAT128__) && _CCCL_OS(LINUX)
-#    if !defined(__CUDA_ARCH__) || (defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 1000)
-#      define _CCCL_HAS_FLOAT128() 1
-#    else
-#      define _CCCL_HAS_FLOAT128() 0
-#    endif
-// NVC++ support float128 only in host code
-#  elif (defined(__SIZEOF_FLOAT128__) || defined(__FLOAT128__)) && _CCCL_OS(LINUX) && !_CCCL_CUDA_COMPILER(NVHPC)
-#    if !defined(__CUDA_ARCH__) || (defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 1000)
-#      define _CCCL_HAS_FLOAT128() 1
-#    else
-#      define _CCCL_HAS_FLOAT128() 0
-#    endif
-#  else
-#    define _CCCL_HAS_FLOAT128() 0
-#  endif
-#endif // !defined(_CCCL_DISABLE_FLOAT128)
+#if _CCCL_HAS_INCLUDE(<cuda_fp8.h>) && _CCCL_HAS_NVFP16() && _CCCL_HAS_NVBF16() && !defined(CCCL_DISABLE_NVFP8_SUPPORT)
+#  undef _CCCL_HAS_NVFP8
+#  define _CCCL_HAS_NVFP8() 1
+#endif
+
+// NVC++ supports float128 only in host code
+#if !defined(CCCL_DISABLE_FLOAT128_SUPPORT) && _CCCL_OS(LINUX)                     \
+  && ((_CCCL_COMPILER(NVRTC) && defined(__CUDACC_RTC_FLOAT128__)) /*NVRTC*/        \
+      || defined(__SIZEOF_FLOAT128__) || defined(__FLOAT128__)) /*HOST COMPILERS*/ \
+  && (!defined(__CUDA_ARCH__) || (defined(__CUDA_ARCH__) && __CUDA_ARCH__ >= 1000)) /*DEVICE CODE*/
+#  undef _CCCL_HAS_FLOAT128
+#  define _CCCL_HAS_FLOAT128() 1
+#endif
 
 #endif // __CCCL_EXTENDED_DATA_TYPES_H
diff --git a/libcudacxx/include/cuda/std/__internal/features.h b/libcudacxx/include/cuda/std/__internal/features.h
index 974fc35f36d..ffe4fda0c7f 100644
--- a/libcudacxx/include/cuda/std/__internal/features.h
+++ b/libcudacxx/include/cuda/std/__internal/features.h
@@ -87,17 +87,17 @@
 
 // libcu++ requires host device support for its tests. Until then restrict usage to at least 12.2
 #ifndef _LIBCUDACXX_HAS_NVFP16
-#  if defined(_CCCL_HAS_NVFP16) && _CCCL_CUDACC_AT_LEAST(12, 2) \
+#  if _CCCL_HAS_NVFP16() && _CCCL_CUDACC_AT_LEAST(12, 2) \
     && (_CCCL_HAS_CUDA_COMPILER || defined(LIBCUDACXX_ENABLE_HOST_NVFP16))
 #    define _LIBCUDACXX_HAS_NVFP16
-#  endif // _CCCL_HAS_NVFP16 && _CCCL_CUDACC_AT_LEAST(12, 2)
+#  endif // _CCCL_HAS_NVFP16() && _CCCL_CUDACC_AT_LEAST(12, 2)
 #endif // !_LIBCUDACXX_HAS_NVFP16
 
 // libcu++ requires host device support for its tests. Until then restrict usage to at least 12.2
 #ifndef _LIBCUDACXX_HAS_NVBF16
-#  if defined(_CCCL_HAS_NVBF16) && _CCCL_CUDACC_AT_LEAST(12, 2)
+#  if _CCCL_HAS_NVBF16() && _CCCL_CUDACC_AT_LEAST(12, 2)
 #    define _LIBCUDACXX_HAS_NVBF16
-#  endif // _CCCL_HAS_NVBF16 && _CCCL_CUDACC_AT_LEAST(12, 2)
+#  endif // _CCCL_HAS_NVBF16() && _CCCL_CUDACC_AT_LEAST(12, 2)
 #endif // !_LIBCUDACXX_HAS_NVBF16
 
 // NVCC does not have a way of silencing non '_' prefixed UDLs
diff --git a/libcudacxx/include/cuda/std/__type_traits/is_extended_floating_point.h b/libcudacxx/include/cuda/std/__type_traits/is_extended_floating_point.h
index 040418f5fe7..6ea27552011 100644
--- a/libcudacxx/include/cuda/std/__type_traits/is_extended_floating_point.h
+++ b/libcudacxx/include/cuda/std/__type_traits/is_extended_floating_point.h
@@ -22,11 +22,11 @@
 
 #include <cuda/std/__type_traits/integral_constant.h>
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
 #endif // _CCCL_HAS_NVFP16
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #  include <cuda_bf16.h>
@@ -53,7 +53,7 @@ _CCCL_INLINE_VAR constexpr bool __is_extended_floating_point_v
 #  endif // !_CCCL_NO_INLINE_VARIABLES
 #endif // !_CCCL_NO_VARIABLE_TEMPLATES
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 template <>
 struct __is_extended_floating_point<__half> : true_type
 {};
@@ -64,7 +64,7 @@ _CCCL_INLINE_VAR constexpr bool __is_extended_floating_point_v<__half> = true;
 #  endif // !_CCCL_NO_INLINE_VARIABLES
 #endif // _CCCL_HAS_NVFP16
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 template <>
 struct __is_extended_floating_point<__nv_bfloat16> : true_type
 {};
diff --git a/libcudacxx/include/cuda/std/limits b/libcudacxx/include/cuda/std/limits
index 29f4bf24ec3..8fdc8e5526f 100644
--- a/libcudacxx/include/cuda/std/limits
+++ b/libcudacxx/include/cuda/std/limits
@@ -608,7 +608,7 @@ public:
 #endif // !_LIBCUDACXX_HAS_NO_LONG_DOUBLE
 };
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  ifdef _LIBCUDACXX_HAS_NVFP16
 #    define _LIBCUDACXX_FP16_CONSTEXPR constexpr
 #  else //_LIBCUDACXX_HAS_NVFP16
@@ -690,7 +690,7 @@ public:
 #  undef _LIBCUDACXX_FP16_CONSTEXPR
 #endif // _CCCL_HAS_NVFP16
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 #  ifdef _LIBCUDACXX_HAS_NVBF16
 #    define _LIBCUDACXX_BF16_CONSTEXPR constexpr
 #  else //_LIBCUDACXX_HAS_NVBF16
diff --git a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.bf16.fail.cpp b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.bf16.fail.cpp
index fef41309a67..52be0cda714 100644
--- a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.bf16.fail.cpp
+++ b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.bf16.fail.cpp
@@ -10,13 +10,13 @@
 
 #include "test_macros.h"
 
-#if !defined(_CCCL_HAS_NVBF16)
+#if !_CCCL_HAS_NVBF16()
 #  include <cuda_bf16.h>
 #endif
 
 int main(int, char**)
 {
-#if !defined(_CCCL_HAS_NVBF16)
+#if !_CCCL_HAS_NVBF16()
   auto x3 = __nv_bfloat16(1.0f);
   unused(x3);
 #else
diff --git a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.fp16.fail.cpp b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.fp16.fail.cpp
index d4925743db9..caf15e30567 100644
--- a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.fp16.fail.cpp
+++ b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.fp16.fail.cpp
@@ -10,13 +10,13 @@
 
 #include "test_macros.h"
 
-#if !defined(_CCCL_HAS_NVFP16)
+#if !_CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
 #endif
 
 int main(int, char**)
 {
-#if !defined(_CCCL_HAS_NVFP16)
+#if !_CCCL_HAS_NVFP16()
   auto x2 = __half(1.0f);
   unused(x2);
 #else
diff --git a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp
index 3b230d8710b..b272ebb04fe 100644
--- a/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp
+++ b/libcudacxx/test/libcudacxx/libcxx/macros/extended_data_types.pass.cpp
@@ -13,10 +13,10 @@
 #if _CCCL_HAS_NVFP8()
 #  include <cuda_fp8.h>
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 #  include <cuda_fp16.h>
 #endif
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 #  include <cuda_bf16.h>
 #endif
 
@@ -32,11 +32,11 @@ int main(int, char**)
   auto x2 = __nv_fp8_e4m3(1.0f);
   unused(x2);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   auto x3 = __half(1.0f);
   unused(x3);
 #endif
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   auto x4 = __nv_bfloat16(1.0f);
   unused(x4);
 #endif
diff --git a/libcudacxx/test/libcudacxx/libcxx/utilities/meta/meta.unary/meta.unary.cat/is_floating_point.pass.cpp b/libcudacxx/test/libcudacxx/libcxx/utilities/meta/meta.unary/meta.unary.cat/is_floating_point.pass.cpp
index 5a04070c598..e704cc582b4 100644
--- a/libcudacxx/test/libcudacxx/libcxx/utilities/meta/meta.unary/meta.unary.cat/is_floating_point.pass.cpp
+++ b/libcudacxx/test/libcudacxx/libcxx/utilities/meta/meta.unary/meta.unary.cat/is_floating_point.pass.cpp
@@ -80,10 +80,10 @@ int main(int, char**)
   test_is_floating_point<float>();
   test_is_floating_point<double>();
   test_is_floating_point<long double>();
-#ifdef _CCCL_HAS_NVFP16
+#if _CCCL_HAS_NVFP16()
   test_is_floating_point<__half>();
 #endif // _CCCL_HAS_NVFP16
-#ifdef _CCCL_HAS_NVBF16
+#if _CCCL_HAS_NVBF16()
   test_is_floating_point<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/is_specialized.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/is_specialized.pass.cpp
index adb30091033..d5ce1f5aaca 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/is_specialized.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/is_specialized.pass.cpp
@@ -68,10 +68,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/common.h b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/common.h
index 7d15f2ba6b6..3d91abb6255 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/common.h
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/common.h
@@ -60,7 +60,7 @@ __host__ __device__ inline bool float_eq(__nv_fp8_e5m2 x, __nv_fp8_e5m2 y)
 }
 #endif // _CCCL_HAS_NVFP8
 
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
 __host__ __device__ inline bool float_eq(__half x, __half y)
 {
 #  if _CCCL_CUDACC_AT_LEAST(12, 2)
@@ -71,7 +71,7 @@ __host__ __device__ inline bool float_eq(__half x, __half y)
 }
 #endif // _CCCL_HAS_NVFP16
 
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
 __host__ __device__ inline bool float_eq(__nv_bfloat16 x, __nv_bfloat16 y)
 {
 #  if _CCCL_CUDACC_AT_LEAST(12, 2)
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/const_data_members.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/const_data_members.pass.cpp
index 093b5d331be..363b48b0d11 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/const_data_members.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/const_data_members.pass.cpp
@@ -110,10 +110,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test_type<long double>();
 #endif // _LIBCUDACXX_HAS_NO_LONG_DOUBLE
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test_type<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test_type<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/denorm_min.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/denorm_min.pass.cpp
index 9ea232eaad6..d652a31592a 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/denorm_min.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/denorm_min.pass.cpp
@@ -66,10 +66,10 @@ int main(int, char**)
   test<long double>(LDBL_TRUE_MIN);
 #  endif
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(5.9604644775390625e-08));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(9.18354961579912115600575419705e-41));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits.pass.cpp
index 01f6b05543b..2d7d1be2a62 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits.pass.cpp
@@ -55,10 +55,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, LDBL_MANT_DIG>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, 11>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, 8>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits10.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits10.pass.cpp
index 24c53725738..66cb7698888 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits10.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/digits10.pass.cpp
@@ -74,10 +74,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/epsilon.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/epsilon.pass.cpp
index bb65847df33..656dcdddc1a 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/epsilon.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/epsilon.pass.cpp
@@ -57,10 +57,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(LDBL_EPSILON);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(0.0009765625));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(0.0078125));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm.pass.cpp
index 8d9881580bf..f9fb39a80dc 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, cuda::std::denorm_present>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, cuda::std::denorm_present>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, cuda::std::denorm_present>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm_loss.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm_loss.pass.cpp
index 5a046a9b339..66a6cd7c54e 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm_loss.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_denorm_loss.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_infinity.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_infinity.pass.cpp
index 768e53d1c88..6c5c3000b2e 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_infinity.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_infinity.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_quiet_NaN.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_quiet_NaN.pass.cpp
index 4c3e11a9b05..1428807921b 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_quiet_NaN.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_quiet_NaN.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_signaling_NaN.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_signaling_NaN.pass.cpp
index 1b80d1869e6..8341fcf5912 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_signaling_NaN.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/has_signaling_NaN.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/infinity.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/infinity.pass.cpp
index 8dd611556c5..bafb3d108e2 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/infinity.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/infinity.pass.cpp
@@ -64,10 +64,10 @@ int main(int, char**)
 #  ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(1. / 0.);
 #  endif
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(1.0 / 0.0));
 #  endif // _CCCL_HAS_NVFP16
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(1.0 / 0.0));
 #  endif // _CCCL_HAS_NVBF16
 #  if _CCCL_HAS_NVFP8()
@@ -81,10 +81,10 @@ int main(int, char**)
 #  ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(INFINITY);
 #  endif
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(INFINITY));
 #  endif // _CCCL_HAS_NVFP16
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(INFINITY));
 #  endif // _CCCL_HAS_NVBF16
 #  if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_bounded.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_bounded.pass.cpp
index e28ab8313b6..3d63d2ebf34 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_bounded.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_bounded.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_exact.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_exact.pass.cpp
index e6038f1589b..a39cc45d913 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_exact.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_exact.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_iec559.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_iec559.pass.cpp
index 1ff809bad09..037c30fbe27 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_iec559.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_iec559.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif // _LIBCUDACXX_HAS_NO_LONG_DOUBLE
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_integer.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_integer.pass.cpp
index eed9d38c050..16898ab4ca0 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_integer.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_integer.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_modulo.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_modulo.pass.cpp
index fc3ca9dbb4e..0b6a45e43a2 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_modulo.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_modulo.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_signed.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_signed.pass.cpp
index 54005f6c0b9..45e46312fb8 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_signed.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/is_signed.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, true>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, true>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, true>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/lowest.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/lowest.pass.cpp
index 72190bd2ad7..863a6359050 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/lowest.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/lowest.pass.cpp
@@ -66,10 +66,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(-LDBL_MAX);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(-65504.0));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(-3.3895313892515355e+38));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max.pass.cpp
index 5039f773a2f..f41e6aecb3c 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max.pass.cpp
@@ -65,10 +65,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(LDBL_MAX);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(65504.0));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(3.3895313892515355e+38));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_digits10.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_digits10.pass.cpp
index 309279bc79c..5013f74f0ab 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_digits10.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_digits10.pass.cpp
@@ -69,10 +69,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent.pass.cpp
index 606e9c52b7f..4e47ef5f58d 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent.pass.cpp
@@ -62,10 +62,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, LDBL_MAX_EXP>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, 16>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, 128>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent10.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent10.pass.cpp
index 61145deec86..2874c5b1da1 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent10.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/max_exponent10.pass.cpp
@@ -62,10 +62,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, LDBL_MAX_10_EXP>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, 4>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, 38>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min.pass.cpp
index ccab08a38f5..30710f1cccd 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min.pass.cpp
@@ -66,10 +66,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(LDBL_MIN);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(6.103515625e-05));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(1.17549435082228750796873653722e-38));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent.pass.cpp
index c942a6288be..0d0896dac82 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent.pass.cpp
@@ -62,10 +62,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, LDBL_MIN_EXP>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, -13>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, -125>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent10.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent10.pass.cpp
index e9b6f29d25f..3d2f088fd68 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent10.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/min_exponent10.pass.cpp
@@ -62,10 +62,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, LDBL_MIN_10_EXP>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, -4>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, -37>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/quiet_NaN.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/quiet_NaN.pass.cpp
index a8b076fbeee..5e39f6f9bc8 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/quiet_NaN.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/quiet_NaN.pass.cpp
@@ -108,10 +108,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/radix.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/radix.pass.cpp
index dd15c391180..776aa2b8466 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/radix.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/radix.pass.cpp
@@ -55,10 +55,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, FLT_RADIX>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, FLT_RADIX>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, FLT_RADIX>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_error.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_error.pass.cpp
index 95ed80eb951..157c153c3aa 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_error.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_error.pass.cpp
@@ -57,10 +57,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>(0.5);
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>(__double2half(0.5));
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>(__double2bfloat16(0.5));
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_style.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_style.pass.cpp
index 1eb5c0b0f5a..5ad0566870e 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_style.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/round_style.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, cuda::std::round_to_nearest>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, cuda::std::round_to_nearest>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, cuda::std::round_to_nearest>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/signaling_NaN.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/signaling_NaN.pass.cpp
index 0ec70976b32..e720a364d63 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/signaling_NaN.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/signaling_NaN.pass.cpp
@@ -108,10 +108,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/tinyness_before.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/tinyness_before.pass.cpp
index 1da28874b06..3ca3751dbcc 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/tinyness_before.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/tinyness_before.pass.cpp
@@ -54,10 +54,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/traps.pass.cpp b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/traps.pass.cpp
index 4cb627a4b77..48cbbef8f19 100644
--- a/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/traps.pass.cpp
+++ b/libcudacxx/test/libcudacxx/std/language.support/support.limits/limits/numeric.limits.members/traps.pass.cpp
@@ -60,10 +60,10 @@ int main(int, char**)
 #ifndef _LIBCUDACXX_HAS_NO_LONG_DOUBLE
   test<long double, false>();
 #endif
-#if defined(_CCCL_HAS_NVFP16)
+#if _CCCL_HAS_NVFP16()
   test<__half, false>();
 #endif // _CCCL_HAS_NVFP16
-#if defined(_CCCL_HAS_NVBF16)
+#if _CCCL_HAS_NVBF16()
   test<__nv_bfloat16, false>();
 #endif // _CCCL_HAS_NVBF16
 #if _CCCL_HAS_NVFP8()
diff --git a/thrust/testing/cuda/sort.cu b/thrust/testing/cuda/sort.cu
index db5e67eb202..da8b2d6aef3 100644
--- a/thrust/testing/cuda/sort.cu
+++ b/thrust/testing/cuda/sort.cu
@@ -150,14 +150,14 @@ SimpleUnitTest<TestRadixSortDispatch,
                                 ,
                                 unittest::type_list<__int128_t, __uint128_t>
 #endif // _LIBCUDACXX_HAS_NO_INT128
-#ifdef _CCCL_HAS_NVFP16
+#if _CCCL_HAS_NVFP16()
                                 ,
                                 unittest::type_list<__half>
-#endif // _CCCL_HAS_NVFP16
-#ifdef _CCCL_HAS_NVBF16
+#endif // _CCCL_HAS_NVFP16()
+#if _CCCL_HAS_NVBF16()
                                 ,
                                 unittest::type_list<__nv_bfloat16>
-#endif // _CCCL_HAS_NVBF16
+#endif // _CCCL_HAS_NVBF16()
                                 >>
   TestRadixSortDispatchInstance;
 
@@ -287,12 +287,11 @@ SimpleUnitTest<TestSortAscendingKey,
 #endif
 // CTK 12.2 offers __host__ __device__ operators for __half and __nv_bfloat16, so we can use std::sort
 #if _CCCL_CUDACC_AT_LEAST(12, 2)
-#  if defined(_CCCL_HAS_NVFP16) || !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
+#  if _CCCL_HAS_NVFP16() || !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
                                 ,
                                 unittest::type_list<__half>
 #  endif
-#  if defined(_CCCL_HAS_NVBF16) \
-    || !defined(__CUDA_NO_BFLOAT16_OPERATORS__) && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__)
+#  if _CCCL_HAS_NVBF16() || !defined(__CUDA_NO_BFLOAT16_OPERATORS__) && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__)
                                 ,
                                 unittest::type_list<__nv_bfloat16>
 #  endif
diff --git a/thrust/thrust/system/cuda/detail/sort.h b/thrust/thrust/system/cuda/detail/sort.h
index 7ad67fd4e0c..20c1f0d1517 100644
--- a/thrust/thrust/system/cuda/detail/sort.h
+++ b/thrust/thrust/system/cuda/detail/sort.h
@@ -62,16 +62,16 @@
 
 #  include <cstdint>
 
-#  if defined(_CCCL_HAS_NVFP16)
+#  if _CCCL_HAS_NVFP16()
 #    include <cuda_fp16.h>
-#  endif // _CCCL_HAS_NVFP16
+#  endif // _CCCL_HAS_NVFP16()
 
-#  if defined(_CCCL_HAS_NVBF16)
+#  if _CCCL_HAS_NVBF16()
 _CCCL_DIAG_PUSH
 _CCCL_DIAG_SUPPRESS_CLANG("-Wunused-function")
 #    include <cuda_bf16.h>
 _CCCL_DIAG_POP
-#  endif // _CCCL_HAS_NVBF16
+#  endif // _CCCL_HAS_NVBF16()
 
 THRUST_NAMESPACE_BEGIN
 namespace cuda_cub
@@ -323,13 +323,12 @@ template <class Key, class CompareOp>
 using can_use_primitive_sort = ::cuda::std::integral_constant<
   bool,
   (::cuda::std::is_arithmetic<Key>::value
-#  if defined(_CCCL_HAS_NVFP16) && !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
+#  if _CCCL_HAS_NVFP16() && !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
    || ::cuda::std::is_same<Key, __half>::value
-#  endif // defined(_CCCL_HAS_NVFP16) && !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
-#  if defined(_CCCL_HAS_NVBF16) && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__) \
-    && !defined(__CUDA_NO_BFLOAT16_OPERATORS__)
+#  endif // _CCCL_HAS_NVFP16() && !defined(__CUDA_NO_HALF_OPERATORS__) && !defined(__CUDA_NO_HALF_CONVERSIONS__)
+#  if _CCCL_HAS_NVBF16() && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__) && !defined(__CUDA_NO_BFLOAT16_OPERATORS__)
    || ::cuda::std::is_same<Key, __nv_bfloat16>::value
-#  endif // defined(_CCCL_HAS_NVBF16) && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__) &&
+#  endif // _CCCL_HAS_NVBF16() && !defined(__CUDA_NO_BFLOAT16_CONVERSIONS__) &&
          // !defined(__CUDA_NO_BFLOAT16_OPERATORS__)
    )
     && (::cuda::std::is_same<CompareOp, thrust::less<Key>>::value
