diff --git a/src/tl_templates/cuda/cuda_fp4.h b/src/tl_templates/cuda/cuda_fp4.h
index d74ee67590..d2efef24ae 100644
--- a/src/tl_templates/cuda/cuda_fp4.h
+++ b/src/tl_templates/cuda/cuda_fp4.h
@@ -166,9 +166,44 @@ TL_DEVICE fp4_e2_32_t make_fp4_e2_32_t(
 // ============================================================================
 // https://docs.nvidia.com/cuda/cuda-math-api/cuda_math_api/group__CUDA__MATH__FP4__MISC.html
 
+// Custom fp4_e2m1 -> half convertion for CUDA version < 13.0 to avoid using
+// `cvt.rn.relu.f16x2.e2m1x2`, as there are bugs in PTXAS related to
+// `cvt.rn.relu.f16x2.e2m1x2` between CUDA 12.6 and 12.9
+__device__ __half_raw __tl_cvt_fp4_to_halfraw_naive(
+    const __nv_fp4_storage_t x,
+    const __nv_fp4_interpretation_t fp4_interpretation) {
+  __half_raw res;
+  res.x = 0U;
+  // fp4_interpretation == __NV_E2M1
+  // convert to e2m3 first
+  __nv_fp6_storage_t fp6e2m3 = (x & 0xFU) << 2U;
+  res = __nv_cvt_fp6_to_halfraw(fp6e2m3, __NV_E2M3);
+  return res;
+}
+
+// Custom fp4_e2m1 -> half convertion for CUDA version < 13.0 to avoid using
+// `cvt.rn.relu.f16x2.e2m1x2`, as there are bugs in PTXAS related to
+// `cvt.rn.relu.f16x2.e2m1x2` between CUDA 12.6 and 12.9
+__device__ __half2_raw __tl_cvt_fp4x2_to_halfraw2_naive(
+    const __nv_fp4x2_storage_t x,
+    const __nv_fp4_interpretation_t fp4_interpretation) {
+  __half2_raw res;
+  res.x =
+      __tl_cvt_fp4_to_halfraw_naive((__nv_fp4_storage_t)x, fp4_interpretation)
+          .x;
+  res.y = __tl_cvt_fp4_to_halfraw_naive((__nv_fp4_storage_t)(x >> 4U),
+                                        fp4_interpretation)
+              .x;
+  return res;
+}
+
 // fp4_e2m1 -> half
 TL_DEVICE __half __tl_cvt_fp4_to_half(const __nv_fp4_storage_t src) {
+#if __CUDACC_VER_MAJOR__ >= 13
   __half_raw raw = __nv_cvt_fp4_to_halfraw(src, __NV_E2M1);
+#else
+  __half_raw raw = __tl_cvt_fp4_to_halfraw_naive(src, __NV_E2M1);
+#endif
   __half result;
   result = *reinterpret_cast<__half *>(&raw);
   return result;
@@ -176,7 +211,11 @@ TL_DEVICE __half __tl_cvt_fp4_to_half(const __nv_fp4_storage_t src) {
 
 // fp4_e2m1x2 (1 byte) -> half2
 TL_DEVICE half2 __tl_cvt_fp4x2_to_half2(const __nv_fp4x2_storage_t src) {
+#if __CUDACC_VER_MAJOR__ >= 13
   __half2_raw raw = __nv_cvt_fp4x2_to_halfraw2(src, __NV_E2M1);
+#else
+  __half2_raw raw = __tl_cvt_fp4x2_to_halfraw2_naive(src, __NV_E2M1);
+#endif
   half2 result;
   result = *reinterpret_cast<half2 *>(&raw);
   return result;
