diff --git a/csrc/trtllm_fmha_kernel_launcher.cu b/csrc/trtllm_fmha_kernel_launcher.cu
index 3d5e8956e8..d4737c9aa1 100644
--- a/csrc/trtllm_fmha_kernel_launcher.cu
+++ b/csrc/trtllm_fmha_kernel_launcher.cu
@@ -84,8 +84,8 @@ void trtllm_paged_attention_launcher(
     int64_t kv_stride_batch, int64_t max_num_blocks_per_seq, double bmm1_scale, double bmm2_scale,
     const float* bmm1_scale_log2_ptr, const float* bmm2_scale_ptr, double o_sf_scale,
     int64_t o_sf_vec_size, int64_t o_sf_start_index, int64_t window_left, int64_t sum_seq_q,
-    int64_t sparse_mla_top_k, int64_t sm_count, bool enable_pdl, int64_t workspace_size,
-    cudaStream_t stream) {
+    int64_t sparse_mla_top_k, float skip_softmax_threshold_scale_factor, bool skips_softmax,
+    int64_t sm_count, bool enable_pdl, int64_t workspace_size, cudaStream_t stream) {
   if (num_qo_heads % num_kv_heads != 0) {
     std::ostringstream err_msg;
     err_msg << "num_qo_heads must be a multiple of num_kv_heads, got num_kv_heads: " << num_kv_heads
@@ -186,6 +186,10 @@ void trtllm_paged_attention_launcher(
         float_allocator.aligned_alloc<void>(0, 16, "trtllm_gen_scratch_workspace");
   }
 
+  // Params for skipping softmax.
+  runner_params.mSkipsSoftmaxWhenPossible = skips_softmax;
+  runner_params.mSkipSoftmaxThresholdScaleFactor = skip_softmax_threshold_scale_factor;
+
   auto [foundKernels, kinfo] = fmha_runner->isSupportedWithInfo(runner_params);
   if (!foundKernels) {
     std::ostringstream err_msg;
@@ -217,17 +221,15 @@ inline Data_type dl_dtype_to_tllm_data_type(const DLDataType dtype) {
 
 inline bool is_4bit(Data_type data_type) { return data_type == Data_type::DATA_TYPE_E2M1; }
 
-void trtllm_paged_attention_decode(TensorView out, Optional<TensorView> out_scale_factor,
-                                   TensorView query, TensorView key_cache, TensorView value_cache,
-                                   TensorView workspace_buffer, TensorView block_tables,
-                                   TensorView seq_lens, int64_t max_q_len, int64_t max_kv_len,
-                                   Variant<double, ffi::Tensor> bmm1_scale,
-                                   Variant<double, ffi::Tensor> bmm2_scale, double o_sf_scale,
-                                   int64_t o_sf_vec_size, int64_t o_sf_start_index,
-                                   int64_t batch_size, int64_t window_left,
-                                   int64_t sparse_mla_top_k, int64_t sm_count, bool enable_pdl,
-                                   int64_t workspace_size, Optional<TensorView> attention_sinks,
-                                   Optional<TensorView> cum_seq_lens_q) {
+void trtllm_paged_attention_decode(
+    TensorView out, Optional<TensorView> out_scale_factor, TensorView query, TensorView key_cache,
+    TensorView value_cache, TensorView workspace_buffer, TensorView block_tables,
+    TensorView seq_lens, int64_t max_q_len, int64_t max_kv_len,
+    Variant<double, ffi::Tensor> bmm1_scale, Variant<double, ffi::Tensor> bmm2_scale,
+    double o_sf_scale, int64_t o_sf_vec_size, int64_t o_sf_start_index, int64_t batch_size,
+    int64_t window_left, int64_t sparse_mla_top_k, int64_t sm_count, bool enable_pdl,
+    int64_t workspace_size, Optional<TensorView> attention_sinks,
+    Optional<TensorView> cum_seq_lens_q, Optional<float> skip_softmax_threshold_scale_factor) {
   auto q_data_type = dl_dtype_to_tllm_data_type(query.dtype());
   auto kv_data_type = dl_dtype_to_tllm_data_type(key_cache.dtype());
   TVM_FFI_ICHECK_EQ(key_cache.ndim(), value_cache.ndim());
@@ -296,6 +298,11 @@ void trtllm_paged_attention_decode(TensorView out, Optional<TensorView> out_scal
   float* bmm2_scale_ptr = maybe_bmm2_scale_tensor.has_value()
                               ? static_cast<float*>(maybe_bmm2_scale_tensor.value().data_ptr())
                               : nullptr;
+
+  bool const skips_softmax = skip_softmax_threshold_scale_factor.has_value();
+  float const skip_softmax_threshold_scale_factor_value =
+      skip_softmax_threshold_scale_factor.value_or(0.0f);
+
   trtllm_paged_attention_launcher(
       out.data_ptr(), output_sf_ptr, query.data_ptr(), key_cache.data_ptr(), value_cache.data_ptr(),
       workspace_buffer.data_ptr(), static_cast<int*>(block_tables.data_ptr()),
@@ -306,7 +313,8 @@ void trtllm_paged_attention_decode(TensorView out, Optional<TensorView> out_scal
       q_stride_heads, kv_stride_keys_values, kv_stride_heads, kv_stride_batch,
       max_num_blocks_per_seq, bmm1_scale_value, bmm2_scale_value, bmm1_scale_log2_ptr,
       bmm2_scale_ptr, o_sf_scale, o_sf_vec_size, o_sf_start_index, window_left, sum_seq_q,
-      sparse_mla_top_k, sm_count, enable_pdl, workspace_size, stream);
+      sparse_mla_top_k, skip_softmax_threshold_scale_factor_value, skips_softmax, sm_count,
+      enable_pdl, workspace_size, stream);
 }
 
 void trtllm_paged_attention_context(
@@ -316,7 +324,8 @@ void trtllm_paged_attention_context(
     Variant<double, ffi::Tensor> bmm1_scale, Variant<double, ffi::Tensor> bmm2_scale,
     double o_sf_scale, int64_t o_sf_vec_size, int64_t o_sf_start_index, int64_t batch_size,
     int64_t window_left, TensorView cum_seq_lens_q, TensorView cum_seq_lens_kv, int64_t sm_count,
-    bool enable_pdl, int64_t workspace_size, Optional<TensorView> attention_sinks) {
+    bool enable_pdl, int64_t workspace_size, Optional<TensorView> attention_sinks,
+    Optional<float> skip_softmax_threshold_scale_factor) {
   auto q_data_type = dl_dtype_to_tllm_data_type(query.dtype());
   auto kv_data_type = dl_dtype_to_tllm_data_type(key_cache.dtype());
   auto o_data_type = dl_dtype_to_tllm_data_type(out.dtype());
@@ -379,6 +388,10 @@ void trtllm_paged_attention_context(
                               ? static_cast<float*>(maybe_bmm2_scale_tensor.value().data_ptr())
                               : nullptr;
 
+  bool const skips_softmax = skip_softmax_threshold_scale_factor.has_value();
+  float const skip_softmax_threshold_scale_factor_value =
+      skip_softmax_threshold_scale_factor.value_or(0.0f);
+
   trtllm_paged_attention_launcher(
       out.data_ptr(), output_sf_ptr, query.data_ptr(), key_cache.data_ptr(), value_cache.data_ptr(),
       workspace_buffer.data_ptr(), static_cast<int*>(block_tables.data_ptr()),
@@ -390,7 +403,8 @@ void trtllm_paged_attention_context(
       head_dim_o, page_size, q_stride_tokens, q_stride_heads, kv_stride_keys_values,
       kv_stride_heads, kv_stride_batch, max_num_blocks_per_seq, bmm1_scale_value, bmm2_scale_value,
       bmm1_scale_log2_ptr, bmm2_scale_ptr, o_sf_scale, o_sf_vec_size, o_sf_start_index, window_left,
-      sum_seq_q, /*sparse_mla_top_k=*/0, sm_count, enable_pdl, workspace_size, stream);
+      sum_seq_q, /*sparse_mla_top_k=*/0, skip_softmax_threshold_scale_factor_value, skips_softmax,
+      sm_count, enable_pdl, workspace_size, stream);
 }
 
 void trtllm_ragged_attention_launcher(
diff --git a/flashinfer/artifacts.py b/flashinfer/artifacts.py
index 71addb0a15..664441cec6 100644
--- a/flashinfer/artifacts.py
+++ b/flashinfer/artifacts.py
@@ -87,7 +87,7 @@ class ArtifactPath:
     When compiling new cubins for backend directories, update the corresponding path.
     """
 
-    TRTLLM_GEN_FMHA: str = "75d477a640f268ea9ad117cc596eb39245713b9e/fmha/trtllm-gen/"
+    TRTLLM_GEN_FMHA: str = "e86f0e45764555d070c3d143b4caaea61a45b777/fmha/trtllm-gen/"
     TRTLLM_GEN_BMM: str = (
         "fea3b0ecfe11d7b34556042aeb5d0465ad101500/batched_gemm-332ffef-9936841"
     )
@@ -107,7 +107,7 @@ class CheckSumHash:
     """
 
     TRTLLM_GEN_FMHA: str = (
-        "e014d7a54c396733ef012b223603c1be2861019f88faa5dcc882ed1ecfe5c2d9"
+        "c4c93904a4c72b8a3d0d5c525c6decb71c835b477d7d75651ecaaa7007c5a3ef"
     )
     TRTLLM_GEN_BMM: str = (
         "1c3c7ae0755a0acb7ad35da7dbdb90ab71c253dc289051faa2b4e3180dfc4b23"
diff --git a/flashinfer/decode.py b/flashinfer/decode.py
index a879caf338..58e349e812 100644
--- a/flashinfer/decode.py
+++ b/flashinfer/decode.py
@@ -1168,6 +1168,9 @@ def run(
         return_lse: Literal[False] = False,
         enable_pdl: Optional[bool] = None,
         window_left: Optional[int] = None,
+        sinks: Optional[torch.Tensor] = None,
+        q_len_per_req: Optional[int] = 1,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> torch.Tensor: ...
 
     @overload
@@ -1184,6 +1187,9 @@ def run(
         return_lse: Literal[True] = True,
         enable_pdl: Optional[bool] = None,
         window_left: Optional[int] = None,
+        sinks: Optional[torch.Tensor] = None,
+        q_len_per_req: Optional[int] = 1,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> Tuple[torch.Tensor, torch.Tensor]: ...
 
     @flashinfer_api
@@ -1202,6 +1208,7 @@ def run(
         window_left: Optional[int] = None,
         sinks: Optional[torch.Tensor] = None,
         q_len_per_req: Optional[int] = 1,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
         r"""Compute batch decode attention between query and paged kv cache.
 
@@ -1241,6 +1248,12 @@ def run(
             Only supported for >= sm90, and currently only for FA2 and CUDA core decode.
         q_len_per_req : int
             The number of query tokens per request, if not provided, will be set to ``1``.
+        skip_softmax_threshold_scale_factor: Optional[float] = None
+            threshold scale factor for skipping softmax operations.
+            Providing a value for this parameter enables skip-softmax sparsity as described in: https://arxiv.org/abs/2512.12087
+            If no value is provided, then standard attention is used.
+            Setting the threshold to a higher value generally increases kernel performance at the cost of accuracy degradation.
+            The actual threshold value equals the provided threshold_scale_factor divided by the context length.
         Returns
         -------
         Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
@@ -1370,6 +1383,7 @@ def run(
                     page_size,
                     self._max_kv_len,
                     sinks,
+                    skip_softmax_threshold_scale_factor,
                 ]
 
             self._cached_module.paged_run(*run_args)
@@ -1920,6 +1934,7 @@ def _paged_run(
         enable_pdl: bool = None,
         out: Optional[torch.Tensor] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> torch.Tensor:
         if out is None:
             out = torch.empty_like(query)
@@ -1961,6 +1976,7 @@ def _paged_run(
             workspace_size,
             sinks,
             None,  # cum_seq_lens_q
+            skip_softmax_threshold_scale_factor,
         )
         return out
 
@@ -2022,6 +2038,7 @@ def paged_run(
         page_size: Optional[int] = None,
         max_kv_len: Optional[int] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> None:
         assert maybe_lse is None
         assert paged_kv_cache is not None
@@ -2048,6 +2065,7 @@ def paged_run(
             enable_pdl,
             out=o,
             sinks=sinks,
+            skip_softmax_threshold_scale_factor=skip_softmax_threshold_scale_factor,
         )
 
     @register_fake_op(f"flashinfer::{uri}_paged_run")
@@ -2087,6 +2105,7 @@ def _fake_paged_run(
         page_size: Optional[int] = None,
         max_kv_len: Optional[int] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> None:
         pass
 
@@ -2124,6 +2143,7 @@ def trtllm_batch_decode_with_kv_cache(
     mask: Optional[torch.Tensor] = None,
     max_q_len: Optional[int] = None,
     cum_seq_lens_q: Optional[torch.Tensor] = None,
+    skip_softmax_threshold_scale_factor: Optional[float] = None,
 ) -> Union[torch.Tensor, FP4Tensor]:
     """
     Parameters
@@ -2207,6 +2227,13 @@ def trtllm_batch_decode_with_kv_cache(
         Only supported by trtllm-gen backend. Must be provided together with ``max_q_len``.
         When None, all requests use uniform query length specified by ``q_len_per_req``.
 
+    skip_softmax_threshold_scale_factor: Optional[float] = None
+        threshold scale factor for skipping softmax operations.
+        Providing a value for this parameter enables skip-softmax sparsity as described in: https://arxiv.org/abs/2512.12087
+        If no value is provided, then standard attention is used.
+        Setting the threshold to a higher value generally increases kernel performance at the cost of accuracy degradation.
+        The actual threshold value equals the provided threshold_scale_factor divided by the context length.
+
     Returns
     -------
     out : Union[torch.Tensor, FP4Tensor]
@@ -2387,6 +2414,7 @@ def trtllm_batch_decode_with_kv_cache(
             workspace_buffer.numel() * workspace_buffer.element_size(),
             sinks,
             cum_seq_lens_q,
+            skip_softmax_threshold_scale_factor,
         )
 
         return (
diff --git a/flashinfer/mla.py b/flashinfer/mla.py
index 31acad2d90..4c7deb9a49 100644
--- a/flashinfer/mla.py
+++ b/flashinfer/mla.py
@@ -688,6 +688,7 @@ def trtllm_batch_decode_with_kv_cache_mla(
             workspace_buffer.numel() * workspace_buffer.element_size(),
             sinks,
             None,  # cum_seq_lens_q
+            None,  # skip_softmax_threshold_scale_factor
         )
 
         return out
diff --git a/flashinfer/prefill.py b/flashinfer/prefill.py
index 08b3608fb4..618f6fbbf6 100755
--- a/flashinfer/prefill.py
+++ b/flashinfer/prefill.py
@@ -223,6 +223,7 @@ def _paged_run(
         window_left: int = -1,
         out: Optional[torch.Tensor] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> torch.Tensor:
         sm_count = get_device_sm_count(query.device)
         if out is None:
@@ -256,6 +257,7 @@ def _paged_run(
             enable_pdl,
             workspace_size,
             sinks,
+            skip_softmax_threshold_scale_factor,
         )
         return out
 
@@ -623,6 +625,7 @@ def paged_run(
         cum_seq_lens_q: Optional[torch.Tensor] = None,
         cum_seq_lens_kv: Optional[torch.Tensor] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> None:
         if backend == "trtllm-gen":
             assert maybe_lse is None
@@ -656,6 +659,7 @@ def paged_run(
                 window_left,
                 out=o,
                 sinks=sinks,
+                skip_softmax_threshold_scale_factor=skip_softmax_threshold_scale_factor,
             )
         elif backend == "fa2":
             assert not is_float8(q)
@@ -773,6 +777,9 @@ def _fake_paged_run(
         maybe_max_item_len_ptr: Optional[torch.Tensor],
         logits_soft_cap: float,
         sm_scale: float,
+        scale_q: Optional[torch.Tensor],
+        scale_k: Optional[torch.Tensor],
+        scale_v: Optional[torch.Tensor],
         rope_scale: float,
         rope_theta: float,
         token_pos_in_items_len: int,
@@ -787,6 +794,8 @@ def _fake_paged_run(
         batch_size: Optional[int] = None,
         cum_seq_lens_q: Optional[torch.Tensor] = None,
         cum_seq_lens_kv: Optional[torch.Tensor] = None,
+        sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> None:
         pass
 
@@ -2028,6 +2037,8 @@ def run(
         return_lse: Literal[False] = False,
         enable_pdl: Optional[bool] = None,
         window_left: Optional[int] = None,
+        sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> torch.Tensor: ...
 
     @overload
@@ -2043,6 +2054,8 @@ def run(
         return_lse: Literal[True] = True,
         enable_pdl: Optional[bool] = None,
         window_left: Optional[int] = None,
+        sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> Tuple[torch.Tensor, torch.Tensor]: ...
 
     @flashinfer_api
@@ -2060,6 +2073,7 @@ def run(
         enable_pdl: Optional[bool] = None,
         window_left: Optional[int] = None,
         sinks: Optional[torch.Tensor] = None,
+        skip_softmax_threshold_scale_factor: Optional[float] = None,
     ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
         r"""Compute batch prefill/append attention between query and paged kv-cache.
 
@@ -2272,6 +2286,7 @@ def run(
                     self._qo_indptr_buf,
                     self._paged_kv_indptr_buf,
                     sinks,
+                    skip_softmax_threshold_scale_factor,
                 ]
 
             assert self._cached_module is not None, "cached module is not initialized"
@@ -3556,6 +3571,7 @@ def trtllm_batch_context_with_kv_cache(
     kv_layout: str = "HND",
     enable_pdl: Optional[bool] = None,
     sinks: Optional[List[torch.Tensor]] = None,
+    skip_softmax_threshold_scale_factor: Optional[float] = None,
 ) -> Union[torch.Tensor, FP4Tensor]:
     """
     Parameters
@@ -3608,7 +3624,12 @@ def trtllm_batch_context_with_kv_cache(
         Layout of kv-cache, can be "HND" or "NHD", default is "HND".
     sinks : Optional[List[torch.Tensor]] = None
         additional value per head in the denominator of the softmax.
-
+    skip_softmax_threshold_scale_factor: Optional[float] = None
+        threshold scale factor for skipping softmax operations.
+        Providing a value for this parameter enables skip-softmax sparsity as described in: https://arxiv.org/abs/2512.12087
+        If no value is provided, then standard attention is used.
+        Setting the threshold to a higher value generally increases kernel performance at the cost of accuracy degradation.
+        The actual threshold value equals the provided threshold_scale_factor divided by the context length.
     Returns
     -------
     out: Union[torch.Tensor, FP4Tensor]
@@ -3742,6 +3763,7 @@ def trtllm_batch_context_with_kv_cache(
         enable_pdl,
         workspace_size,
         sinks,
+        skip_softmax_threshold_scale_factor,
     )
     return (
         out
diff --git a/include/flashinfer/trtllm/fmha/fmhaKernels.cuh b/include/flashinfer/trtllm/fmha/fmhaKernels.cuh
index ca71bb0b88..ced19c9470 100644
--- a/include/flashinfer/trtllm/fmha/fmhaKernels.cuh
+++ b/include/flashinfer/trtllm/fmha/fmhaKernels.cuh
@@ -141,7 +141,7 @@ class TllmGenFmhaKernel {
   inline uint64_t hashID(int qkvLayout, int maskType, int kernelType, int scheduler,
                          int multiCtasKvMode, int headDimPerCtaV, int headDimQk, int headDimV,
                          int tileSizeQ, int tileSizeKv, int numTokensPerPage, bool reuseSmemKForV,
-                         bool uses2CtaMma, bool sparseMla) const {
+                         bool uses2CtaMma, bool sparseMla, bool skipsSoftmax) const {
     FLASHINFER_CHECK((headDimPerCtaV >= 32) && (headDimQk >= 32) && (headDimV >= 32) &&
                          (headDimPerCtaV <= 1024) && (headDimQk <= 1024) && (headDimV <= 1024),
                      "Expect (32 <= headDim <= 1024), got headDimPerCtaV=%d, headDimQk=%d, "
@@ -170,6 +170,7 @@ class TllmGenFmhaKernel {
     // Bit 53 - 53: reuseSmemKForV.
     // Bit 54 - 54: uses2CtaMma.
     // Bit 55 - 55: sparseMla.
+    // Bit 56 - 56: skipsSoftmax.
     return (static_cast<uint64_t>(qkvLayout) << 0) | (static_cast<uint64_t>(maskType) << 4) |
            (static_cast<uint64_t>(kernelType) << 8) | (static_cast<uint64_t>(scheduler) << 12) |
            (static_cast<uint64_t>(multiCtasKvMode) << 16) |
@@ -180,7 +181,8 @@ class TllmGenFmhaKernel {
            (static_cast<uint64_t>(log2(numTokensPerPage)) << 44) |
            (static_cast<uint64_t>(log2(tileSizeQ)) << 49) |
            (static_cast<uint64_t>(reuseSmemKForV) << 53) |
-           (static_cast<uint64_t>(uses2CtaMma) << 54) | (static_cast<uint64_t>(sparseMla) << 55);
+           (static_cast<uint64_t>(uses2CtaMma) << 54) | (static_cast<uint64_t>(sparseMla) << 55) |
+           (static_cast<uint64_t>(skipsSoftmax) << 56);
   }
 
   uint64_t hashID(KernelMeta const& kernelMeta) const {
@@ -188,7 +190,8 @@ class TllmGenFmhaKernel {
                   kernelMeta.mTileScheduler, kernelMeta.mMultiCtasKvMode,
                   kernelMeta.mHeadDimPerCtaV, kernelMeta.mHeadDimQk, kernelMeta.mHeadDimV,
                   kernelMeta.mTileSizeQ, kernelMeta.mTileSizeKv, kernelMeta.mNumTokensPerPage,
-                  kernelMeta.mReuseSmemKForV, kernelMeta.m2CtaMma, kernelMeta.mSparseMla);
+                  kernelMeta.mReuseSmemKForV, kernelMeta.m2CtaMma, kernelMeta.mSparseMla,
+                  kernelMeta.mSkipsSoftmaxWhenPossible);
   }
 
   std::pair<bool, std::string> checkIfKernelExist(RunnerParams const& params) const {
@@ -757,7 +760,8 @@ class TllmGenFmhaKernel {
         ", numTokensPerPage=" + std::to_string(selectKernelParams.mNumTokensPerPage) +
         ", reuseSmemKForV=" + std::to_string(selectKernelParams.mReuseSmemKForV) +
         ", uses2CtaMma=" + std::to_string(selectKernelParams.mUses2CtaMma) +
-        ", sparseMla=" + std::to_string(params.mSparseMla);
+        ", sparseMla=" + std::to_string(params.mSparseMla) +
+        ", skipsSoftmax=" + std::to_string(selectKernelParams.mSkipsSoftmaxWhenPossible);
     IKL_LOG_DEBUG(
         "Searching for kernel traits (%d available) in TllmGenFmhaKernel(%s, %s, %s, %d) %s",
         getNumLoadedKernels(), toStr(mDtypeQ), toStr(mDtypeKv), toStr(mDtypeOut), mSM,
@@ -771,7 +775,8 @@ class TllmGenFmhaKernel {
                selectKernelParams.mHeadDimPerCtaV, params.mHeadDimQk, params.mHeadDimV,
                selectKernelParams.mTileSizeQ, selectKernelParams.mTileSizeKv,
                selectKernelParams.mNumTokensPerPage, selectKernelParams.mReuseSmemKForV,
-               selectKernelParams.mUses2CtaMma, params.mSparseMla),
+               selectKernelParams.mUses2CtaMma, params.mSparseMla,
+               selectKernelParams.mSkipsSoftmaxWhenPossible),
         info);
   }
 
diff --git a/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h b/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
index 576fae4d4d..41f6e541db 100755
--- a/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
+++ b/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
@@ -292,6 +292,10 @@ struct TllmGenFmhaRunnerParams {
   float mScaleSfKv;
   // The SF scale for output.
   float mScaleSfO;
+  // Do we skip softmax when possible?
+  bool mSkipsSoftmaxWhenPossible;
+  // Skip softmax threshold scale factor.
+  float mSkipSoftmaxThresholdScaleFactor;
   // Whether to use sparse MLA.
   bool mSparseMla;
   // The top k value for sparse MLA.
@@ -354,6 +358,8 @@ struct TllmGenSelectKernelParams {
   bool mReuseSmemKForV;
   // Do we need to select a new kernel as the parameters have been updated.
   bool mSelectNewKernel;
+  // Do we enable skip softmax?
+  bool mSkipsSoftmaxWhenPossible;
   // The tile scheduler.
   TileScheduler mTileScheduler;
   // The tile size for Q.
@@ -376,6 +382,7 @@ struct TllmGenSelectKernelParams {
         mNumTokensPerPage(params.mNumTokensPerPage),
         mReuseSmemKForV(false),
         mSelectNewKernel(false),
+        mSkipsSoftmaxWhenPossible(params.mSkipsSoftmaxWhenPossible),
         mTileScheduler(params.mTileScheduler),
         mTileSizeQ(128),
         mTileSizeKv(128),
diff --git a/include/flashinfer/trtllm/fmha/kernelParams.h b/include/flashinfer/trtllm/fmha/kernelParams.h
index 295647245a..a097572d22 100644
--- a/include/flashinfer/trtllm/fmha/kernelParams.h
+++ b/include/flashinfer/trtllm/fmha/kernelParams.h
@@ -134,8 +134,12 @@ struct KernelParams {
   // The sequence lengths for K/V. Required by pagedKv kernels to avoid unnecessary computation
   // based on (ptrCumSeqLensKv[batchIdx + 1] - ptrCumSeqLensKv[batchIdx]).
   int32_t const* ptrSeqLensKv;
-  // The reserved memory buffer.
-  int32_t* ptrReservedMem;
+  // When collecting skip softmax stats, store them here
+  // Note that softmax and BMM2 are skipped at different granularity (warp vs tile)
+  // [0] -> skipped softmax warp blocks; [1] -> total softmax warp blocks.
+  // [2] -> skipped BMM2s; [3] -> total BMM2s.
+  int32_t* ptrSkipSoftmaxStats;
+
   // The softmax stats buffer.
   float2* ptrSoftmaxStats;
 
@@ -188,8 +192,8 @@ struct KernelParams {
   float mScaleSfKv;
   // The SF scale for O.
   float mScaleSfO;
-  // The reserved parameter.
-  float mReservedParam;
+  // Threshold to decide whether warp skips softmax ops
+  float mSkipSoftmaxThresholdScaleFactor;
   // The start token index in SF tensor. Used for FP4 SF offset calculation in generation phase
   // kernel when inflight batching is enabled in TRT-LLM.
   int32_t mStartTokenIdxSfO;
@@ -798,6 +802,7 @@ struct KernelParams {
     params.mNumTokensPerCtaQ = numTokensPerCtaQ;
     params.mOutputScale = options.outputScale;
     params.mScaleSoftmaxLog2 = options.scaleSoftmaxLog2;
+    params.mSkipSoftmaxThresholdScaleFactor = options.mSkipSoftmaxThresholdScaleFactor;
     params.mStartTokenIdxSfO = options.mSfStartTokenIdx;
     params.mScaleSfKv = options.mScaleSfKv;
     params.ptrSoftmaxStats = options.softmaxStatsPtr;
diff --git a/tests/attention/test_trtllm_gen_attention.py b/tests/attention/test_trtllm_gen_attention.py
index c6e8920c26..ad3f077676 100755
--- a/tests/attention/test_trtllm_gen_attention.py
+++ b/tests/attention/test_trtllm_gen_attention.py
@@ -421,10 +421,17 @@ def _test_trtllm_batch_prefill(
     device_scale,
     head_dim,
     non_contiguous_query=False,
+    skips_softmax=False,
 ):
     compute_capability = get_compute_capability(torch.device(device="cuda"))
     if compute_capability[0] != 10:
         pytest.skip("These tests are only guaranteed to work on SM100 and SM103 GPUs.")
+
+    if skips_softmax and q_dtype != kv_dtype:
+        pytest.skip(
+            "skips_softmax does not currently support Q and Kv types being different"
+        )
+
     # Set up test parameters
     torch.manual_seed(0)
 
@@ -537,6 +544,9 @@ def _test_trtllm_batch_prefill(
     else:
         q_input = q.contiguous()
 
+    # Using 0.0 threshold should give the same result as normal attention.
+    skip_softmax_threshold_scale_factor = 0.0 if skips_softmax else None
+
     output = flashinfer.prefill.trtllm_batch_context_with_kv_cache(
         q_input,
         kv_cache,
@@ -558,6 +568,7 @@ def _test_trtllm_batch_prefill(
         kv_layout=kv_layout,
         enable_pdl=enable_pdl,
         sinks=(sink if enable_sink else None),
+        skip_softmax_threshold_scale_factor=skip_softmax_threshold_scale_factor,
     )
     # check if the first 8192 * 256 * 4 bytes of workspace_buffer is zero
     # note(Yingyi): the first 8192 * 256 * 4 bytes of workspace_buffer is the counter workspace, size might change in the future
@@ -654,6 +665,7 @@ def _test_trtllm_batch_prefill(
 @pytest.mark.parametrize("max_kv_len", [2047])
 @pytest.mark.parametrize("head_dim", [128, 256])
 @pytest.mark.parametrize("non_contiguous_query", [False, True])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_prefill(
     kv_layout,
     batch_size,
@@ -670,6 +682,7 @@ def test_trtllm_batch_prefill(
     max_kv_len,
     head_dim,
     non_contiguous_query,
+    skips_softmax,
 ):
     _test_trtllm_batch_prefill(
         kv_layout,
@@ -688,6 +701,7 @@ def test_trtllm_batch_prefill(
         kv_dtype == "fp8",
         head_dim,
         non_contiguous_query=non_contiguous_query,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -710,6 +724,7 @@ def test_trtllm_batch_prefill(
 @pytest.mark.parametrize("max_q_len", [8192])
 @pytest.mark.parametrize("max_kv_len", [8192])
 @pytest.mark.parametrize("head_dim", [128, 256])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_prefill_bs1(
     kv_layout,
     batch_size,
@@ -725,6 +740,7 @@ def test_trtllm_batch_prefill_bs1(
     max_q_len,
     max_kv_len,
     head_dim,
+    skips_softmax,
 ):
     _test_trtllm_batch_prefill(
         kv_layout,
@@ -742,6 +758,7 @@ def test_trtllm_batch_prefill_bs1(
         max_kv_len,
         False,
         head_dim,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -764,6 +781,7 @@ def _test_trtllm_batch_decode(
     device_scale=False,
     max_q_len=None,
     non_contiguous_query=False,
+    skips_softmax=False,
 ):
     """
     Common function for testing trtllm-gen decode.
@@ -778,6 +796,14 @@ def _test_trtllm_batch_decode(
     if backend == "xqa" and compute_capability[0] < 9:
         pytest.skip("xqa backend requires SM90+ GPUs.")
 
+    if backend == "xqa" and skips_softmax:
+        pytest.skip("xqa backend does not support skips_softmax")
+
+    if skips_softmax and q_dtype != kv_dtype:
+        pytest.skip(
+            "skips_softmax does not currently support Q and Kv types being different"
+        )
+
     # xqa backend doesn't support nvfp4 output
     if backend == "xqa" and o_dtype == "nvfp4":
         pytest.skip("xqa backend does not support nvfp4 output")
@@ -933,6 +959,9 @@ def _test_trtllm_batch_decode(
     else:
         q_input = q.contiguous()
 
+    # Using 0.0 threshold should give the same result as normal attention.
+    skip_softmax_threshold_scale_factor = 0.0 if skips_softmax else None
+
     output = flashinfer.decode.trtllm_batch_decode_with_kv_cache(
         q_input,
         kv_cache,
@@ -956,6 +985,7 @@ def _test_trtllm_batch_decode(
         mask=mask,
         max_q_len=max_q_len if max_q_len is not None else None,
         cum_seq_lens_q=q_indptr if max_q_len is not None else None,
+        skip_softmax_threshold_scale_factor=skip_softmax_threshold_scale_factor,
     )
     if backend == "trtllm-gen":
         # check if the first 8192 * 256 * 4 bytes of workspace_buffer is zero
@@ -1101,6 +1131,7 @@ def _test_trtllm_batch_decode(
 @pytest.mark.parametrize("max_in_kv_len", [110])
 @pytest.mark.parametrize("head_dim", [128])
 @pytest.mark.parametrize("non_contiguous_query", [False, True])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_decode(
     backend,
     kv_layout,
@@ -1118,6 +1149,7 @@ def test_trtllm_batch_decode(
     max_in_kv_len,
     head_dim,
     non_contiguous_query,
+    skips_softmax,
 ):
     # xqa backend does not support non-contiguous query yet
     if backend == "xqa" and non_contiguous_query:
@@ -1142,6 +1174,7 @@ def test_trtllm_batch_decode(
         head_dim,
         kv_dtype == "fp8",
         non_contiguous_query=non_contiguous_query,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -1165,6 +1198,7 @@ def test_trtllm_batch_decode(
 @pytest.mark.parametrize("max_in_kv_len", [4096, 8192])
 @pytest.mark.parametrize("head_dim", [128])
 @pytest.mark.parametrize("device_scale", [True, False])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_decode_bs1(
     kv_layout,
     batch_size,
@@ -1181,6 +1215,7 @@ def test_trtllm_batch_decode_bs1(
     max_in_kv_len,
     head_dim,
     device_scale,
+    skips_softmax,
 ):
     # Small number of test cases for batch size 1
     _test_trtllm_batch_decode(
@@ -1200,6 +1235,7 @@ def test_trtllm_batch_decode_bs1(
         max_in_kv_len,
         head_dim,
         device_scale,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -1233,6 +1269,7 @@ def test_trtllm_batch_decode_bs1(
 @pytest.mark.parametrize("max_in_kv_len", [110])
 @pytest.mark.parametrize("head_dim", [256])
 @pytest.mark.parametrize("device_scale", [True, False])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_decode_head_dim_256(
     kv_layout,
     batch_size,
@@ -1249,6 +1286,7 @@ def test_trtllm_batch_decode_head_dim_256(
     max_in_kv_len,
     head_dim,
     device_scale,
+    skips_softmax,
 ):
     # Small number of test cases for head_dim = 256
     _test_trtllm_batch_decode(
@@ -1268,6 +1306,7 @@ def test_trtllm_batch_decode_head_dim_256(
         max_in_kv_len,
         head_dim,
         device_scale,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -1297,6 +1336,7 @@ def test_trtllm_batch_decode_head_dim_256(
 @pytest.mark.parametrize("max_in_kv_len", [4096, 8192, 16384, 32768, 65536, 131072])
 @pytest.mark.parametrize("head_dim", [128])
 @pytest.mark.parametrize("device_scale", [True, False])
+@pytest.mark.parametrize("skips_softmax", [False])
 def test_trtllm_batch_decode_long_sequence_length(
     kv_layout,
     batch_size,
@@ -1313,6 +1353,7 @@ def test_trtllm_batch_decode_long_sequence_length(
     max_in_kv_len,
     head_dim,
     device_scale,
+    skips_softmax,
 ):
     # Small number of test cases for long sequence length
     _test_trtllm_batch_decode(
@@ -1332,6 +1373,7 @@ def test_trtllm_batch_decode_long_sequence_length(
         max_in_kv_len,
         head_dim,
         device_scale,
+        skips_softmax=skips_softmax,
     )
 
 
@@ -1542,6 +1584,7 @@ def make_query_non_contiguous(q, num_qo_heads, head_dim):
 @pytest.mark.parametrize("enable_sink", [True, False])
 @pytest.mark.parametrize("max_in_kv_len", [110])
 @pytest.mark.parametrize("head_dim", [128])
+@pytest.mark.parametrize("skips_softmax", [False, True])
 def test_trtllm_batch_decode_spec(
     backend,
     kv_layout,
@@ -1558,6 +1601,7 @@ def test_trtllm_batch_decode_spec(
     enable_sink,
     max_in_kv_len,
     head_dim,
+    skips_softmax,
 ):
     _test_trtllm_batch_decode(
         backend,
@@ -1576,4 +1620,5 @@ def test_trtllm_batch_decode_spec(
         max_in_kv_len,
         head_dim,
         max_q_len=max_q_len,
+        skips_softmax=skips_softmax,
     )
