diff --git a/testing/python/tilelibrary/test_tilelang_tilelibrary_gemm.py b/testing/python/tilelibrary/test_tilelang_tilelibrary_gemm.py
index 8e023d8da7..8770f7416a 100644
--- a/testing/python/tilelibrary/test_tilelang_tilelibrary_gemm.py
+++ b/testing/python/tilelibrary/test_tilelang_tilelibrary_gemm.py
@@ -429,16 +429,14 @@ def ref_program(A, B):
         (512, 1024, 768, True, False, T.float16, T.float16, T.float32, 128, 256, 32, 2, 128),
         (512, 1024, 768, True, True, T.float16, T.float16, T.float32, 128, 256, 32, 2, 128),
         (128, 16, 32, False, True, T.float16, T.float16, T.float32, 128, 16, 32, 0, 128),
-        # TODO: There is precision problem when num_stages=2 on ROCm
-        # (128, 128, 32, False, True, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
-        # (128, 128, 32, False, False, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
+        (128, 128, 32, False, True, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
+        (128, 128, 32, False, False, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
         (128, 128, 32, True, False, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
         (128, 128, 32, True, True, T.int8, T.int8, T.int32, 128, 128, 32, 2, 128),
         (128, 128, 128, False, False, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
         (128, 128, 128, False, True, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
-        # TODO: There is precision problem needs to repair on ROCm
-        # (128, 128, 128, True, False, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
-        # (128, 128, 128, True, True, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
+        (128, 128, 128, True, False, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
+        (128, 128, 128, True, True, T.float, T.float, T.float32, 128, 128, 32, 2, 128),
     ],
 )
 def test_gemm_sr(M, N, K, trans_A, trans_B, in_dtype, out_dtype, dtypeAccum, block_M, block_N, block_K, num_stages, num_threads):
diff --git a/tilelang/intrinsics/mfma_macro_generator.py b/tilelang/intrinsics/mfma_macro_generator.py
index 89fd0bdc6a..fa211e16d7 100644
--- a/tilelang/intrinsics/mfma_macro_generator.py
+++ b/tilelang/intrinsics/mfma_macro_generator.py
@@ -176,8 +176,10 @@ def get_ldmatrix_index_map(self, is_b=False):
         k_dim = self.k_dim * self.k_pack
         transposed = self.a_transposed if not is_b else self.b_transposed
         if k_dim == 4:
-            index_map = shared_16x4_to_local_64x1_layout_A
-            reverse_index_map = thread_id_shared_access_64x1_to_16x4_layout_A
+            index_map = shared_4x16_to_local_64x1_layout_B if transposed else shared_16x4_to_local_64x1_layout_A
+            reverse_index_map = (
+                thread_id_shared_access_64x1_to_4x16_layout_B if transposed else thread_id_shared_access_64x1_to_16x4_layout_A
+            )
             if is_b:
                 index_map = shared_16x4_to_local_64x1_layout_A if transposed else shared_4x16_to_local_64x1_layout_B
                 reverse_index_map = (
