diff --git a/csrc/apis/gemm.hpp b/csrc/apis/gemm.hpp
index fb205de20..f50960f25 100644
--- a/csrc/apis/gemm.hpp
+++ b/csrc/apis/gemm.hpp
@@ -28,8 +28,9 @@ static bool early_return(const int& m, const int &n, const int& k,
         DG_HOST_ASSERT(c->sizes() == d.sizes() and c->strides() == d.strides());
     if (c.has_value()) {
         check_major_type_cd(c.value());
-        DG_HOST_ASSERT(d.scalar_type() == torch::kFloat);
-        DG_HOST_ASSERT(c.value().scalar_type() == torch::kFloat);
+        DG_HOST_ASSERT((d.scalar_type() == torch::kFloat) or (d.scalar_type() == torch::kBFloat16));
+        DG_HOST_ASSERT((c.value().scalar_type() == torch::kFloat) or (c.value().scalar_type() == torch::kBFloat16));
+        DG_HOST_ASSERT(d.scalar_type() == c.value().scalar_type());
     }
 
     // No accumulation
diff --git a/deep_gemm/include/deep_gemm/impls/sm100_fp8_gemm_1d1d.cuh b/deep_gemm/include/deep_gemm/impls/sm100_fp8_gemm_1d1d.cuh
index da7f461c7..886f5ffb2 100644
--- a/deep_gemm/include/deep_gemm/impls/sm100_fp8_gemm_1d1d.cuh
+++ b/deep_gemm/include/deep_gemm/impls/sm100_fp8_gemm_1d1d.cuh
@@ -36,9 +36,9 @@ sm100_fp8_gemm_1d1d_impl(int* grouped_layout,
     using Barrier = cutlass::arch::ClusterTransactionBarrier;
     using Allocator = cute::conditional_t<kNumMulticast == 1, cute::TMEM::Allocator1Sm, cute::TMEM::Allocator2Sm>;
 
-    // GEMM with accumulation must have FP32 output
+    // GEMM with accumulation must have FP32/BF16 output
     if constexpr (kWithAccumulation)
-        DG_STATIC_ASSERT(cute::is_same_v<cd_dtype_t, float>, "Invalid C/D data dtype");
+        DG_STATIC_ASSERT(cute::is_same_v<cd_dtype_t, float> or cute::is_same_v<cd_dtype_t, cutlass::bfloat16_t>, "Invalid C/D data dtype");
 
     // Configs
     constexpr uint32_t LAYOUT_AD_M = 128;
diff --git a/tests/generators.py b/tests/generators.py
index 162274c76..c4fdc69df 100644
--- a/tests/generators.py
+++ b/tests/generators.py
@@ -79,6 +79,8 @@ def enumerate_normal(dtype: torch.dtype) -> Generator:
                 n, k = nk_list[i]
                 out_dtype = torch.bfloat16 if i < len(bf16_output_nk) else torch.float
                 yield kernel_type, m, n, k, MajorTypeAB.KMajor, MajorTypeAB.KMajor, False, out_dtype
+                if dtype == torch.float8_e4m3fn and get_arch_major() == 10:
+                    yield kernel_type, m, n, k, MajorTypeAB.KMajor, MajorTypeAB.KMajor, True, out_dtype
 
         # Backward
         for m in m_bwd_list:
