diff --git a/lib/Dialect/TritonInstrument/IR/Utility.cpp b/lib/Dialect/TritonInstrument/IR/Utility.cpp
index 58e194c593ac..c07c398bb181 100644
--- a/lib/Dialect/TritonInstrument/IR/Utility.cpp
+++ b/lib/Dialect/TritonInstrument/IR/Utility.cpp
@@ -590,7 +590,7 @@ LogicalResult AuxDataMap::populateAndPassToWarpSpecialize(
                        {createInitStateTensor(b, {numCTAs}, 32, 1, fb),
                         getIntTensorType(entryRegion, {numCTAs}, 32)});
     passToWarpSpecialize(entryPoint, activeMasks.at(entryRegion), activeMasks,
-                         captureCounter);
+                         captureCounter, captureBytes);
 
     barrierWriteRecipients.insert(
         entryRegion,
