diff --git a/include/flashinfer/comm/vllm_custom_all_reduce.cuh b/include/flashinfer/comm/vllm_custom_all_reduce.cuh
index 4ef7b6f8b9..976530177a 100644
--- a/include/flashinfer/comm/vllm_custom_all_reduce.cuh
+++ b/include/flashinfer/comm/vllm_custom_all_reduce.cuh
@@ -58,7 +58,7 @@ struct Signal {
 };
 
 struct __align__(16) RankData {
-  const void* __restrict__ ptrs[8];
+  void* __restrict__ ptrs[8];
 };
 
 struct __align__(16) RankSignals {
