diff --git a/csrc/trtllm_fmha_kernel_launcher.cu b/csrc/trtllm_fmha_kernel_launcher.cu
index 587dd1d5e6..53e4d1f3be 100644
--- a/csrc/trtllm_fmha_kernel_launcher.cu
+++ b/csrc/trtllm_fmha_kernel_launcher.cu
@@ -28,11 +28,11 @@ namespace flashinfer {
 template <typename T, Data_type CACHE_T>
 void trtllm_paged_attention_launcher(at::Tensor& out, at::Tensor& query,
                                      at::Tensor& key_value_cache, at::Tensor& workspace_buffer,
-                                     int64_t num_q_heads, int64_t num_kv_heads, double scale,
+                                     int64_t num_q_heads, int64_t num_kv_heads,
                                      at::Tensor& block_tables, at::Tensor& seq_lens,
                                      int64_t block_size, int64_t max_seq_len,
-                                     const std::string kv_cache_dtype, double k_scale,
-                                     double v_scale) {
+                                     const std::string kv_cache_dtype, double bmm1_scale,
+                                     double bmm2_scale) {
   int num_seqs = query.size(0);
   int num_heads = query.size(1);
   TORCH_CHECK(num_heads == static_cast<int>(num_q_heads),
@@ -43,9 +43,6 @@ void trtllm_paged_attention_launcher(at::Tensor& out, at::Tensor& query,
   auto device = query.device();
   const auto stream = at::cuda::getCurrentCUDAStream(device.index());
 
-  float const k_scale_f = k_scale;
-  float const v_scale_f = v_scale;
-
   uint32_t tokens_per_page = block_size;
 
   uint32_t num_k_heads = num_kv_heads;
@@ -139,16 +136,16 @@ void trtllm_paged_attention_launcher(at::Tensor& out, at::Tensor& query,
 
   runner_params.stream = stream;
 
-  runner_params.outputScale = 1.0;
-  runner_params.scaleSoftmaxLog2 = 1.0;
+  runner_params.outputScale = bmm2_scale;
+  runner_params.scaleSoftmaxLog2 = bmm1_scale * M_LOG2E;
 
   fmha_runner.run(runner_params);
 }
 
-#define CALL_GEN_LAUNCHER(T, CACHE_T_ENUM)                                             \
-  trtllm_paged_attention_launcher<T, CACHE_T_ENUM>(                                    \
-      out, query, key_value_cache, workspace_buffer, num_q_heads, num_kv_heads, scale, \
-      block_tables, seq_lens, block_size, max_seq_len, kv_cache_dtype, k_scale, v_scale);
+#define CALL_GEN_LAUNCHER(T, CACHE_T_ENUM)                                                    \
+  trtllm_paged_attention_launcher<T, CACHE_T_ENUM>(                                           \
+      out, query, key_value_cache, workspace_buffer, num_q_heads, num_kv_heads, block_tables, \
+      seq_lens, block_size, max_seq_len, kv_cache_dtype, bmm1_scale, bmm2_scale);
 
 // The following macro is used to dispatch the conversion function based on
 // the data type of the key and value cache. The FN is a macro that calls a
@@ -178,9 +175,9 @@ void trtllm_paged_attention_launcher(at::Tensor& out, at::Tensor& query,
 
 void trtllm_paged_attention(at::Tensor& out, at::Tensor& query, at::Tensor& key_value_cache,
                             at::Tensor& workspace_buffer, int64_t num_q_heads, int64_t num_kv_heads,
-                            double scale, at::Tensor& block_tables, at::Tensor& seq_lens,
-                            int64_t block_size, int64_t max_seq_len,
-                            const std::string kv_cache_dtype, double k_scale, double v_scale) {
+                            at::Tensor& block_tables, at::Tensor& seq_lens, int64_t block_size,
+                            int64_t max_seq_len, const std::string kv_cache_dtype,
+                            double bmm1_scale, double bmm2_scale) {
   DISPATCH_BY_KV_CACHE_ELEM_ENUM(query.dtype(), kv_cache_dtype, CALL_GEN_LAUNCHER);
 }
 
diff --git a/csrc/trtllm_mla_kernel_launcher.cu b/csrc/trtllm_mla_kernel_launcher.cu
index 40319d3629..be5d8e1115 100644
--- a/csrc/trtllm_mla_kernel_launcher.cu
+++ b/csrc/trtllm_mla_kernel_launcher.cu
@@ -26,16 +26,17 @@ namespace flashinfer {
 template <Data_type CACHE_T>
 void trtllm_paged_attention_mla_launcher(
     at::Tensor& out, at::Tensor& query, at::Tensor& key_value_cache, at::Tensor& workspace_buffer,
-    double scale, at::Tensor& block_tables, at::Tensor& seq_lens, int64_t block_size,
-    int64_t max_seq_len, int64_t qk_nope_head_dim, int64_t kv_lora_rank, int64_t qk_rope_head_dim,
-    double bmm1_scale, double bmm2_scale, std::optional<int64_t> acc_q_len,
-    std::optional<int64_t> max_attention_window_size,
+    at::Tensor& block_tables, at::Tensor& seq_lens, int64_t block_size, int64_t max_seq_len,
+    int64_t qk_nope_head_dim, int64_t kv_lora_rank, int64_t qk_rope_head_dim, double bmm1_scale,
+    double bmm2_scale, std::optional<at::Tensor> bmm1_scale_tensor,
+    std::optional<at::Tensor> bmm2_scale_tensor, std::optional<int64_t> max_attention_window_size,
     std::optional<int64_t> cyclic_attention_window_size) {
   int const num_seqs = query.size(0);
   int const batch_size = num_seqs;
-  int const num_q_heads = query.size(1);
+  int const acc_q_len = query.size(1);
+  int const num_q_heads = query.size(2);
   int const num_kv_heads = 1;
-  int head_size = query.size(2);
+  int head_size = query.size(3);
   int const beam_width = 1;                        // NOTE: beam_width always 1
   int const batch_beam = beam_width * batch_size;  // NOTE: batch_beam = batch_size
   int const max_num_blocks_per_seq = block_tables.size(-1);
@@ -104,8 +105,7 @@ void trtllm_paged_attention_mla_launcher(
   runner_params.mMaxSeqLenCacheKv = max_attention_window_size_opt;
 
   // This should be set to numDraftTokens + 1.
-  auto const acc_q_len_opt = acc_q_len.value_or(batch_beam);
-  runner_params.mMaxSeqLenQ = acc_q_len_opt / batch_beam;  // should be 1 if acc_q_len not provided
+  runner_params.mMaxSeqLenQ = acc_q_len;  // should be draft_tokens + 1
   runner_params.mMaxSeqLenKv = max_seq_len;
   runner_params.mSumOfSeqLensQ = int(batch_beam * runner_params.mMaxSeqLenQ);
   // Not used in the generation kernels as contiguous_kv or paged_kv layouts are used.
@@ -122,7 +122,7 @@ void trtllm_paged_attention_mla_launcher(
   // Q_SCALE & KV_SCALE not supported for now
   // runner_params.mScaleQ = scale * sqrt((float)(qk_nope_head_dim + qk_rope_head_dim)) /
   //                         sqrtf((float)(kv_lora_rank + qk_rope_head_dim));
-  runner_params.mScaleQ = scale;
+  runner_params.mScaleQ = 1.0;
 
   // runner_params.mNumPagesInMemPool = INT_MAX;
   auto const [free_memory, total_memory] = getDeviceMemoryInfo(false);
@@ -138,20 +138,12 @@ void trtllm_paged_attention_mla_launcher(
   runner_params.mSfStartTokenIdx = 0;
 
   runner_params.outputScale = bmm2_scale;
-  runner_params.scaleSoftmaxLog2 = bmm1_scale;
-  // if (CACHE_T == Data_type::DATA_TYPE_E4M3) {
-  //   // NOTE(Yingyi): bmm1_scale and bmm2_scale are 1.0 could work already
-  //   runner_params.outputScale = bmm2_scale;
-  //   runner_params.scaleSoftmaxLog2 = bmm1_scale;
-
-  //   // NOTE(Yingyi): if loadsScalesFromGmem enabled, the scales will be loaded from gmem
-  //   // runner_params.outputScalePtr = bmm2_scale_tensor.has_value()
-  //   //                                    ? bmm2_scale_tensor.value().data_ptr<float>()
-  //   //                                    : nullptr;
-  //   // runner_params.scaleSoftmaxLog2Ptr = bmm1_scale_tensor.has_value()
-  //   //                                         ? bmm1_scale_tensor.value().data_ptr<float>()
-  //   //                                         : nullptr;
-  // }
+  runner_params.scaleSoftmaxLog2 = bmm1_scale * M_LOG2E;
+  runner_params.useGmemScale = bmm1_scale_tensor.has_value() && bmm2_scale_tensor.has_value();
+  runner_params.scaleSoftmaxLog2Ptr =
+      runner_params.useGmemScale ? bmm1_scale_tensor.value().data_ptr<float>() : nullptr;
+  runner_params.outputScalePtr =
+      runner_params.useGmemScale ? bmm2_scale_tensor.value().data_ptr<float>() : nullptr;
 
   zero_gmem_semaphore_launcher(runner_params.multiCtasKvCounterPtr, num_semaphores,
                                /*enable_pdl=*/true, stream);
@@ -159,11 +151,12 @@ void trtllm_paged_attention_mla_launcher(
   fmha_runner.run(runner_params);
 }
 
-#define CALL_GEN_LAUNCHER(CACHE_T_ENUM)                                                         \
-  trtllm_paged_attention_mla_launcher<CACHE_T_ENUM>(                                            \
-      out, query, key_value_cache, workspace_buffer, scale, block_tables, seq_lens, block_size, \
-      max_seq_len, qk_nope_head_dim, kv_lora_rank, qk_rope_head_dim, bmm1_scale, bmm2_scale,    \
-      acc_q_len, max_attention_window_size, cyclic_attention_window_size);
+#define CALL_GEN_LAUNCHER(CACHE_T_ENUM)                                                      \
+  trtllm_paged_attention_mla_launcher<CACHE_T_ENUM>(                                         \
+      out, query, key_value_cache, workspace_buffer, block_tables, seq_lens, block_size,     \
+      max_seq_len, qk_nope_head_dim, kv_lora_rank, qk_rope_head_dim, bmm1_scale, bmm2_scale, \
+      bmm1_scale_tensor, bmm2_scale_tensor, max_attention_window_size,                       \
+      cyclic_attention_window_size);
 
 // The following macro is used to dispatch the conversion function based on
 // the data type of the key and value cache. The FN is a macro that calls a
@@ -180,11 +173,12 @@ void trtllm_paged_attention_mla_launcher(
   }
 
 void trtllm_paged_attention_mla(at::Tensor& out, at::Tensor& query, at::Tensor& key_value_cache,
-                                at::Tensor& workspace_buffer, double scale,
-                                at::Tensor& block_tables, at::Tensor& seq_lens, int64_t block_size,
-                                int64_t max_seq_len, int64_t qk_nope_head_dim, int64_t kv_lora_rank,
+                                at::Tensor& workspace_buffer, at::Tensor& block_tables,
+                                at::Tensor& seq_lens, int64_t block_size, int64_t max_seq_len,
+                                int64_t qk_nope_head_dim, int64_t kv_lora_rank,
                                 int64_t qk_rope_head_dim, double bmm1_scale, double bmm2_scale,
-                                std::optional<int64_t> acc_q_len,
+                                std::optional<at::Tensor> bmm1_scale_tensor,
+                                std::optional<at::Tensor> bmm2_scale_tensor,
                                 std::optional<int64_t> max_attention_window_size,
                                 std::optional<int64_t> cyclic_attention_window_size) {
   DISPATCH_BY_QKV_DTYPE(query.dtype(), key_value_cache.dtype(),
diff --git a/flashinfer/decode.py b/flashinfer/decode.py
index a948c7d9a3..91ace7e841 100644
--- a/flashinfer/decode.py
+++ b/flashinfer/decode.py
@@ -1703,14 +1703,13 @@ def trtllm_batch_decode_with_kv_cache(
     workspace_buffer: torch.Tensor,
     num_heads: int,
     num_kv_heads: int,
-    scale: float,
     block_tables: torch.Tensor,
     seq_lens: torch.Tensor,
     block_size: int,
     max_seq_len: int,
     kv_cache_dtype: str,
-    k_scale: float,
-    v_scale: float,
+    bmm1_scale: float,
+    bmm2_scale: float,
     out: Optional[torch.Tensor] = None,
 ) -> torch.Tensor:
     run_func = get_trtllm_fmha_gen_module().trtllm_paged_attention
@@ -1727,14 +1726,13 @@ def trtllm_batch_decode_with_kv_cache(
         workspace_buffer,
         num_heads,
         num_kv_heads,
-        scale,
         block_tables,
         seq_lens,
         block_size,
         max_seq_len,
         kv_cache_dtype,
-        k_scale,
-        v_scale,
+        bmm1_scale,
+        bmm2_scale,
     )
     return out
 
@@ -1748,8 +1746,8 @@ def _check_trtllm_gen_mla_shape(
     page_table,
     page_size,
 ):
-    if query.ndim != 3:
-        raise ValueError(f"Expected query.ndim == 3, got {query.ndim}")
+    if query.ndim != 4:
+        raise ValueError(f"Expected query.ndim == 4, got {query.ndim}")
     if kv_cache.ndim != 4:
         raise ValueError(f"Expected kv_cache.ndim == 4, got {kv_cache.ndim}")
     if qk_nope_head_dim != 128:
@@ -1759,7 +1757,7 @@ def _check_trtllm_gen_mla_shape(
     if qk_rope_head_dim != 64:
         raise ValueError(f"Expected qk_rope_head_dim == 64, got {qk_rope_head_dim}")
 
-    B_q, H, D_q = query.shape
+    B_q, Q_len, H, D_q = query.shape
     D_ckv = kv_cache.shape[3]
     # if H != 128:
     #     raise ValueError(f"Expected 128 heads for query, got {H}")
@@ -1792,14 +1790,15 @@ def trtllm_batch_decode_with_kv_cache_mla(
     seq_lens: torch.Tensor,
     block_size: int,
     max_seq_len: int,
-    scale: Optional[float] = 1.0,
     out: Optional[torch.Tensor] = None,
     bmm1_scale: Optional[float] = 1.0,
     bmm2_scale: Optional[float] = 1.0,  # todo(Yingyi): update to be tensor later
+    bmm1_scale_tensor: Optional[torch.Tensor] = None,
+    bmm2_scale_tensor: Optional[torch.Tensor] = None,
 ) -> torch.Tensor:
     """
     Parameters:
-    query: [batch_size, num_heads, head_dim_qk], head_dim_qk = qk_nope_head_dim (kv_lora_rank) + qk_rope_head_dim, should be concated q_nope + q_rope
+    query: [batch_size, acc_q_len, num_heads, head_dim_qk], head_dim_qk = qk_nope_head_dim (kv_lora_rank) + qk_rope_head_dim, should be concated q_nope + q_rope; acc_q_len = num_draft_tokens + 1 is the MTP query length.
     kv_cache: [num_pages, page_size, head_dim_ckv + head_dim_kpe], should be concated ckv_cache + kpe_cache
     workspace_buffer: [num_semaphores, 4], used for multi_block mode
     qk_nope_head_dim: qk_nope_head_dim, must be 128
@@ -1811,19 +1810,24 @@ def trtllm_batch_decode_with_kv_cache_mla(
     max_seq_len: max sequence length
     scale: model scale of qk, default is 1.0
     out: output tensor, if not provided, will be allocated internally
-    bmm1_scale: scale for mla bmm1 output, only for fp8 quantization. Per-tensor scale now, so shape is [1].
-    bmm2_scale: scale for mla bmm2 output, only for fp8 quantization. Per-tensor scale now, so shape is [1].
+    bmm1_scale: fused scale for mla bmm1 input.
+    bmm2_scale: fused scale for mla bmm2 input.
+    bmm1_scale_tensor: On-device fused scale tensor for mla bmm1 input.
+    bmm2_scale_tensor: On-device fused scale tensor for mla bmm2 input.
 
     Note:
-    For FP8 quantization, we could have those per-tensor scales:
-    - q_scale: dynamic scale for query, shape is [1]
-    - k_scale: loaded scale for kv_cache, shape is [1]
-    - v_scale: loaded scale for kv_cache, shape is [1]
-    We could calculate bmm1_scale and bmm2_scale as:
-    - bmm1_scale = q_scale * k_scale
-    - bmm2_scale = v_scale
-
-    TODO: We might support per-head / per-block / any finer-grained quantization in the future.
+    In MLA, the actual BMM1 and BMM2 scales applied would be fused as:
+    bmm1_scale = q_scale * k_scale * sm_scale / (head_dim_qk ** 0.5)
+    bmm2_scale = v_scale * o_scale
+    For bmm2_scale_tensor, please fuse * M_LOG2E to use faster exp2.
+    The two scale factors should be static constant for cuda graph capture.
+    Either (bmm1_scale, bmm2_scale) or (bmm1_scale_tensor, bmm2_scale_tensor) should be provided.
+    For static constant scale factors, the scale factors should be provided as float.
+        - (bmm1_scale, bmm2_scale)
+    For on-device fused scale tensors, which could dynamically change, the scale factors should be provided as torch.Tensor.
+        - (bmm1_scale_tensor, bmm2_scale_tensor)
+        - Currently, only fp8 tensor core operation supports this mode.
+    When both are provided, the dynamic scale factor tensors will be used.
     """
     run_func = get_trtllm_mla_gen_module().trtllm_paged_attention_mla
 
@@ -1840,32 +1844,31 @@ def trtllm_batch_decode_with_kv_cache_mla(
         block_size,
     )
 
-    # todo(Yingyi): update to be tensor
-    # if bmm1_scale is None:
-    #     bmm1_scale = torch.tensor(1.0, device=query.device, dtype=query.dtype)
-    # if bmm2_scale is None:
-    #     bmm2_scale = torch.tensor(1.0, device=query.device, dtype=query.dtype)
-
     if out is None:
         out_shape = query.shape[:-1] + (kv_lora_rank,)
         out = torch.empty(out_shape, dtype=torch.bfloat16, device=query.device)
     else:
-        _check_shape_dtype_device(out, query.shape, query.dtype, query.device, "out")
+        batch_size, _, num_q_heads, _ = query.shape
+        _check_shape_dtype_device(
+            out,
+            [batch_size, num_q_heads, kv_lora_rank],
+            torch.bfloat16,
+            query.device,
+            "out",
+        )
 
-    # NOTE(Yingyi): enable scale factor tensor for finer-grained fp8 quantization in the future
-    # if kv_cache.dtype == torch.float8_e4m3fn and (
-    #     bmm1_scale is None or bmm2_scale is None
-    # ):
-    #     raise ValueError(
-    #         "bmm1_scale and bmm2_scale must be provided for fp8 quantization"
-    #     )
+    if bmm1_scale_tensor is not None and bmm2_scale_tensor is not None:
+        # dynamic scale factors
+        if query.dtype != torch.float8_e4m3fn or kv_cache.dtype != torch.float8_e4m3fn:
+            raise ValueError(
+                "Dynamic scale factors bmm1_scale_tensor and bmm2_scale_tensor are only supported for fp8 tensor core operation"
+            )
 
     run_func(
         out,
         query,
         kv_cache,
         workspace_buffer,
-        scale,
         block_tables,
         seq_lens,
         block_size,
@@ -1875,7 +1878,8 @@ def trtllm_batch_decode_with_kv_cache_mla(
         qk_rope_head_dim,
         bmm1_scale,
         bmm2_scale,
-        None,  # acc_q_len, speculative not supported for now
+        bmm1_scale_tensor,
+        bmm2_scale_tensor,
         None,  # max_attention_window_size, sliding window not supported for now
         None,  # cyclic_attention_window_size, cyclic window not supported for now
     )
diff --git a/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h b/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
index cc801083ab..53929ee4d1 100644
--- a/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
+++ b/include/flashinfer/trtllm/fmha/fmhaRunnerParams.h
@@ -194,6 +194,7 @@ struct TllmGenFmhaRunnerParams {
   int const* cumSeqLensKvPtr;
   // The kv page idx
   int const* kvPageIdxPtr;
+  bool useGmemScale;
   // The device output scale for FP8 quantization.
   float const* outputScalePtr;
   float outputScale;
@@ -251,9 +252,6 @@ struct TllmGenFmhaRunnerParams {
   // kernel when inflight batching is enabled.
   int mSfStartTokenIdx;
 
-  // note(Yingyi): might take an output scale later
-  // float mOutputScale;
-
   // The SF scale for Kv.
   float mScaleSfKv;
   // The cuda stream.
diff --git a/include/flashinfer/trtllm/fmha/kernelParams.h b/include/flashinfer/trtllm/fmha/kernelParams.h
index f99a4b6073..0457c753de 100644
--- a/include/flashinfer/trtllm/fmha/kernelParams.h
+++ b/include/flashinfer/trtllm/fmha/kernelParams.h
@@ -733,9 +733,7 @@ struct KernelParams {
     params.mNumHiddenEltsO = options.mNumHeadsQ * options.mHeadDimQk;
     // todo(Yingyi): might take a scalar tensor later
     params.mOutputScale = options.outputScale;
-    params.mScaleSoftmaxLog2 =
-        (options.scaleSoftmaxLog2 / (std::sqrt((float)(options.mHeadDimQk)) * options.mScaleQ)) *
-        M_LOG2E;
+    params.mScaleSoftmaxLog2 = options.scaleSoftmaxLog2;
     params.mStartTokenIdxSfO = options.mSfStartTokenIdx;
     params.mScaleSfKv = options.mScaleSfKv;
     params.ptrSoftmaxStats = nullptr;
diff --git a/tests/test_trtllm_gen_decode.py b/tests/test_trtllm_gen_decode.py
index a5029485aa..f2ea0ab2ad 100644
--- a/tests/test_trtllm_gen_decode.py
+++ b/tests/test_trtllm_gen_decode.py
@@ -129,7 +129,7 @@ def test_trtllm_batch_decode_fmha(
     dtype = torch.float16 if q_dtype == "half" else torch.bfloat16
 
     scale = float(1.0 / (head_dim**0.5))
-    q = torch.randn(batch_size, num_qo_heads, head_dim).to(0).to(dtype)
+    q = torch.randn(batch_size, num_qo_heads, head_dim, device=device).to(dtype)
 
     # Sequence lengths and block tables
     seq_lens = [torch.randint(1, MAX_SEQ_LEN, (1,)).item() for _ in range(batch_size)]
@@ -165,8 +165,7 @@ def test_trtllm_batch_decode_fmha(
     # kv_cache_shape = (block_id, 2, num_kv_heads, page_size, head_dim)
     # Allocate more than needed blocks, block_id is just enough, to mimick real-world cases
     kv_cache_shape = (num_blocks, 2, num_kv_heads, page_size, head_dim)
-    kv_cache = torch.randn(size=kv_cache_shape).to(dtype).to(device)
-    k_scale = v_scale = 1.0
+    kv_cache = torch.randn(size=kv_cache_shape, device=device).to(dtype)
 
     if kv_cache_dtype.startswith("fp8"):
         kv_cache, _ = to_float8(kv_cache)
@@ -179,14 +178,13 @@ def test_trtllm_batch_decode_fmha(
         workspace_buffer,
         num_qo_heads,
         num_kv_heads,
-        scale,
         block_tables,
         seq_lens_tensor,
         page_size,
         max_seq_len,
         kv_cache_dtype,
-        k_scale,
-        v_scale,
+        scale,  # bmm1_scale
+        1.0,  # bmm2_scale
     )
 
     if head_grp_size == 5:
@@ -207,13 +205,8 @@ def test_trtllm_batch_decode_fmha(
         blocks_per_seq = (seq_lens_tensor + page_size - 1) // page_size
 
         # Compute kv_indptr as cumulative sum of blocks per sequence
-        kv_indptr = (
-            torch.cat(
-                [torch.tensor([0], device=device), torch.cumsum(blocks_per_seq, dim=0)]
-            )
-            .int()
-            .to(device)
-        )
+        kv_indptr = torch.zeros(batch_size + 1, dtype=torch.int, device=device)
+        kv_indptr[1:] = torch.cumsum(blocks_per_seq, dim=0)
         # Create kv_indices with only the allocated blocks
         kv_indices = all_block_ids.int()
 
@@ -245,20 +238,25 @@ def test_trtllm_batch_decode_fmha(
 
 
 @pytest.mark.parametrize(
-    "batch_size", [16, 32, 64, 128, 256, 512, 768, 1024, 1280, 1536, 1792, 2048]
+    "batch_size",
+    [1, 2, 4, 16, 32, 64, 128, 256, 512, 768, 1024],
 )
 @pytest.mark.parametrize("scale", [1.0, 0.5])
-@pytest.mark.parametrize(
-    "dtype", [torch.float8_e4m3fn, torch.bfloat16]
-)  # todo(Yingyi): add float8_e4m3fn
+@pytest.mark.parametrize("dtype", [torch.float8_e4m3fn, torch.bfloat16])
 @pytest.mark.parametrize("page_size", [32, 64])
+@pytest.mark.parametrize("acc_q_len", [1, 2])
+@pytest.mark.parametrize("dynamic_scale", [False, True])
 def test_trtllm_batch_decode_mla(
     batch_size: int,
     scale: float,
     dtype: torch.dtype,
     page_size: int,
-    kv_layout: str = "HND",  # trtllm-gen only support HND
+    acc_q_len: int,
+    dynamic_scale: bool,
 ):
+    if dynamic_scale and dtype != torch.float8_e4m3fn:
+        pytest.skip("Dynamic scale is not supported for non-fp8 dtype")
+
     torch.manual_seed(42)
     device = "cuda:0"
 
@@ -273,22 +271,13 @@ def test_trtllm_batch_decode_mla(
     kv_lora_rank = 512
 
     # Initialize tensors
-    query = (
-        torch.zeros(batch_size, num_q_heads, kv_lora_rank + qk_rope_head_dim)
-        .to(device)
-        .to(dtype)
-    )
-    # NOTE(Yingyi): enable scale factor tensor for finer-grained fp8 quantization in the future
-    # bmm1_scale_tensor = (
-    #     torch.tensor([1.0], dtype=torch.float32, device=device)
-    #     if dtype == torch.float8_e4m3fn
-    #     else None
-    # )
-    # bmm2_scale_tensor = (
-    #     torch.tensor([1.0], dtype=torch.float32, device=device)
-    #     if dtype == torch.float8_e4m3fn
-    #     else None
-    # )
+    query = torch.randn(
+        batch_size,
+        acc_q_len,
+        num_q_heads,
+        kv_lora_rank + qk_rope_head_dim,
+        device=device,
+    ).to(dtype)
 
     num_tokens = MAX_SEQ_LEN * batch_size
     num_blocks = (num_tokens + page_size - 1) // page_size
@@ -325,11 +314,9 @@ def test_trtllm_batch_decode_mla(
 
     # Create interleaved KV cache
     # Allocate more than needed blocks, block_id is just enough, to mimick real-world cases
-    kv_cache = (
-        torch.randn(size=(num_blocks, page_size, kv_lora_rank + qk_rope_head_dim))
-        .to(dtype)
-        .to(device)
-    )
+    kv_cache = torch.randn(
+        size=(num_blocks, page_size, kv_lora_rank + qk_rope_head_dim), device=device
+    ).to(dtype)
     # (num_blocks, 2, page_size, kv_lora_rank + qk_rope_head_dim)
     # todo(Yingyi): do not duplicate kv_cache for the next generated cubins
     kv_cache_duplicate = torch.stack([kv_cache, kv_cache], dim=1)
@@ -339,6 +326,21 @@ def test_trtllm_batch_decode_mla(
     # todo(Yingyi): calculate the actual size of workspace buffer
     workspace_buffer = torch.empty(128 * 1024 * 1024, dtype=torch.int8, device=device)
 
+    bmm1_scale_tensor = (
+        torch.tensor(
+            [scale / ((128 + 64) ** 0.5 * math.log2(math.e))],
+            dtype=torch.float32,
+            device=device,
+        )
+        if dynamic_scale
+        else None
+    )
+    bmm2_scale_tensor = (
+        torch.tensor([1.0], dtype=torch.float32, device=device)
+        if dynamic_scale
+        else None
+    )
+
     # Run decode-MLA
     output = flashinfer.decode.trtllm_batch_decode_with_kv_cache_mla(
         query=query,
@@ -351,13 +353,18 @@ def test_trtllm_batch_decode_mla(
         seq_lens=seq_lens_tensor,
         block_size=page_size,
         max_seq_len=max_seq_len,
-        scale=scale / ((512 + 64) ** 0.5) * ((128 + 64) ** 0.5),
+        bmm1_scale=scale / ((128 + 64) ** 0.5),
+        bmm2_scale=1.0,
+        bmm1_scale_tensor=bmm1_scale_tensor,
+        bmm2_scale_tensor=bmm2_scale_tensor,
     )
     torch.cuda.synchronize()
+    print("output shape", output.shape)
+    print("output", output)
 
     # Run reference attention and align output
-    sm_scale = (
-        1.0 / scale / ((128 + 64) ** 0.5)
+    sm_scale = scale / (
+        (128 + 64) ** 0.5
     )  # use head dimension before matrix absorption
     workspace_buffer_ref = torch.empty(
         128 * 1024 * 1024, dtype=torch.int8, device=device
@@ -374,17 +381,14 @@ def test_trtllm_batch_decode_mla(
 
     if dtype == torch.float8_e4m3fn:
         # convert query and kv_cache to bfloat16
-        query = query.to(torch.bfloat16).to(device)
-        kv_cache = kv_cache.to(torch.bfloat16).to(device)
+        query = query.to(torch.bfloat16)
+        kv_cache = kv_cache.to(torch.bfloat16)
 
-    q_indptr = torch.arange(0, batch_size + 1, device=device, dtype=torch.int32) * 1
-    kv_indptr = (
-        torch.cat(
-            [torch.tensor([0], device=device), torch.cumsum(blocks_per_seq, dim=0)]
-        )
-        .int()
-        .to(device)
+    q_indptr = (
+        torch.arange(0, batch_size + 1, device=device, dtype=torch.int32) * acc_q_len
     )
+    kv_indptr = torch.zeros_like(q_indptr)
+    kv_indptr[1:] = torch.cumsum(blocks_per_seq, dim=0)
     kv_indices = all_block_ids.int()
 
     wrapper.plan(
@@ -401,21 +405,33 @@ def test_trtllm_batch_decode_mla(
         query.dtype,
         kv_cache.dtype,
     )
-    q_nope = query[..., :kv_lora_rank]
-    q_pe = query[..., kv_lora_rank:]
+    q_nope = query[..., :kv_lora_rank].view(
+        batch_size * acc_q_len, num_q_heads, kv_lora_rank
+    )
+    q_pe = query[..., kv_lora_rank:].view(
+        batch_size * acc_q_len, num_q_heads, qk_rope_head_dim
+    )
 
     # todo: fix kv_cache
     ckv = kv_cache[..., :kv_lora_rank]
     kpe = kv_cache[..., kv_lora_rank:]
 
     o_ref = wrapper.run(q_nope, q_pe, ckv, kpe, return_lse=False)
-    # print("output", output)
-    # print("o_ref", o_ref)
+    torch.cuda.synchronize()
+    print("o_ref shape", o_ref.shape)
+    print("o_ref", o_ref)
+
+    # check is nan
+    assert not torch.isnan(o_ref).any(), "o_ref is nan"
+    assert not torch.isnan(output).any(), "output is nan"
 
     if dtype == torch.float8_e4m3fn:
         try:
             torch.testing.assert_close(
-                output, o_ref, rtol=1e-1, atol=1e-1
+                output,
+                o_ref.view(batch_size, acc_q_len, num_q_heads, -1),
+                rtol=1e-1,
+                atol=1e-1,
             )  # todo: do reference with normal attention?
         except AssertionError as e:
             print("output:", output)
@@ -423,7 +439,12 @@ def test_trtllm_batch_decode_mla(
             raise e
     else:
         try:
-            torch.testing.assert_close(output, o_ref, rtol=1e-2, atol=1e-2)
+            torch.testing.assert_close(
+                output,
+                o_ref.view(batch_size, acc_q_len, num_q_heads, -1),
+                rtol=1e-2,
+                atol=1e-2,
+            )
         except AssertionError as e:
             print("output:", output)
             print("o_ref:", o_ref)
@@ -432,4 +453,5 @@ def test_trtllm_batch_decode_mla(
 
 if __name__ == "__main__":
     # run all tests in the order of pytest
-    test_trtllm_batch_decode_mla(16, 0.5, torch.float8_e4m3fn, 16)
+    # test_trtllm_batch_decode_mla(16, 0.5, torch.float8_e4m3fn, 32, 1)
+    test_trtllm_batch_decode_mla(1024, 1.0, torch.float8_e4m3fn, 32, 2, False)
