diff --git a/csrc/trtllm_moe_allreduce_fusion.cu b/csrc/trtllm_moe_allreduce_fusion.cu
index a5278047cf..cec7b3de2e 100644
--- a/csrc/trtllm_moe_allreduce_fusion.cu
+++ b/csrc/trtllm_moe_allreduce_fusion.cu
@@ -63,7 +63,7 @@ void trtllm_moe_allreduce_fusion(
         params.scale_factor = static_cast<float>(scale_factor);
         params.layout = layout_code.has_value()
                             ? static_cast<FP4QuantizationSFLayout>(layout_code.value())
-                            : FP4QuantizationSFLayout::SWIZZLED_128x4;
+                            : FP4QuantizationSFLayout::SWIZZLED;
         params.stream = stream;
 
         params.moe_reduction_device_num_experts = moe_reduction_device_num_experts;
