diff --git a/libcudacxx/include/cuda/__warp/warp_match_all.h b/libcudacxx/include/cuda/__warp/warp_match_all.h
index acf76b7af08..cee326f9b00 100644
--- a/libcudacxx/include/cuda/__warp/warp_match_all.h
+++ b/libcudacxx/include/cuda/__warp/warp_match_all.h
@@ -25,10 +25,9 @@
 
 #  include <cuda/__cmath/ceil_div.h>
 #  include <cuda/__warp/lane_mask.h>
+#  include <cuda/std/__cstring/memcpy.h>
 #  include <cuda/std/__memory/addressof.h>
-#  include <cuda/std/__type_traits/remove_cv.h>
 #  include <cuda/std/cstdint>
-#  include <cuda/std/cstring>
 
 #  include <cuda/std/__cccl/prologue.h>
 
@@ -36,13 +35,15 @@ _CCCL_BEGIN_NAMESPACE_CUDA_DEVICE
 
 extern "C" _CCCL_DEVICE void __cuda__match_all_sync_is_not_supported_before_SM_70__();
 
-template <typename _Tp, typename _Up = ::cuda::std::remove_cv_t<_Tp>>
-[[nodiscard]] _CCCL_DEVICE_API bool warp_match_all(const _Tp& __data, lane_mask __lane_mask = lane_mask::all())
+template <class _Tp>
+[[nodiscard]] _CCCL_DEVICE_API bool warp_match_all(const _Tp& __data, lane_mask __lane_mask = lane_mask::all()) noexcept
 {
   _CCCL_ASSERT(__lane_mask != lane_mask::none(), "lane_mask must be non-zero");
-  constexpr int __ratio = ::cuda::ceil_div(sizeof(_Up), sizeof(uint32_t));
-  uint32_t __array[__ratio];
-  ::cuda::std::memcpy(__array, ::cuda::std::addressof(__data), sizeof(_Up));
+
+  constexpr int __ratio = ::cuda::ceil_div(sizeof(_Tp), sizeof(uint32_t));
+  uint32_t __array[__ratio]{};
+  ::cuda::std::memcpy(__array, ::cuda::std::addressof(__data), sizeof(_Tp));
+
   bool __ret = true;
   _CCCL_PRAGMA_UNROLL_FULL()
   for (int i = 0; i < __ratio; ++i)
