diff --git a/flash_attn/cute/interface.py b/flash_attn/cute/interface.py
index ef2df23e448..14b9e5165d1 100644
--- a/flash_attn/cute/interface.py
+++ b/flash_attn/cute/interface.py
@@ -260,6 +260,9 @@ def _flash_attn_fwd(
     if mask_mod is None:
         if causal:
             window_size_right = 0
+        if window_size_left is not None and window_size_right is not None and window_size_left + window_size_right < 0:
+            window_size_left = None
+            window_size_right = None
         local = window_size_left is not None or window_size_right is not None
         if window_size_left is not None or window_size_right is not None:
             if window_size_left is None and window_size_right == 0:
@@ -651,6 +654,9 @@ def _flash_attn_bwd(
 
     if causal:
         window_size_right = 0
+    if window_size_left is not None and window_size_right is not None and window_size_left + window_size_right < 0:
+        window_size_left = None
+        window_size_right = None
     local = window_size_left is not None or window_size_right is not None
     if local:
         if window_size_left is None and window_size_right == 0:
