diff --git a/libcudacxx/include/cuda/std/__floating_point/constants.h b/libcudacxx/include/cuda/std/__floating_point/constants.h
index 1d539ddf81a..85b331df4aa 100644
--- a/libcudacxx/include/cuda/std/__floating_point/constants.h
+++ b/libcudacxx/include/cuda/std/__floating_point/constants.h
@@ -165,6 +165,50 @@ template <class _Tp>
   return _CUDA_VSTD::__fp_from_storage<_Tp>(_CUDA_VSTD::__fp_lowest<__fp_format_of_v<_Tp>>());
 }
 
+// __fp_zero
+
+template <__fp_format _Fmt>
+[[nodiscard]] _CCCL_API constexpr __fp_storage_t<_Fmt> __fp_zero() noexcept
+{
+  static_assert(_Fmt != __fp_format::__fp8_nv_e8m0, "__fp_zero: __nv_fp8_e8m0 cannot represent zero");
+  return __fp_storage_t<_Fmt>(0);
+}
+
+template <class _Tp>
+[[nodiscard]] _CCCL_API constexpr _Tp __fp_zero() noexcept
+{
+  static_assert(__fp_format_of_v<_Tp> != __fp_format::__fp8_nv_e8m0, "__fp_zero: __nv_fp8_e8m0 cannot represent zero");
+  if constexpr (__is_std_fp_v<_Tp> || __is_ext_compiler_fp_v<_Tp>)
+  {
+    return _Tp{};
+  }
+  else
+  {
+    return _CUDA_VSTD::__fp_from_storage<_Tp>(__fp_storage_of_t<_Tp>{0});
+  }
+}
+
+// __fp_one
+
+template <__fp_format _Fmt>
+[[nodiscard]] _CCCL_API constexpr __fp_storage_t<_Fmt> __fp_one() noexcept
+{
+  return static_cast<__fp_storage_t<_Fmt>>(__fp_exp_bias_v<_Fmt> << __fp_mant_nbits_v<_Fmt>);
+}
+
+template <class _Tp>
+[[nodiscard]] _CCCL_API constexpr _Tp __fp_one() noexcept
+{
+  if constexpr (__is_std_fp_v<_Tp> || __is_ext_compiler_fp_v<_Tp>)
+  {
+    return _Tp{1};
+  }
+  else
+  {
+    return _CUDA_VSTD::__fp_from_storage<_Tp>(_CUDA_VSTD::__fp_one<__fp_format_of_v<_Tp>>());
+  }
+}
+
 _LIBCUDACXX_END_NAMESPACE_STD
 
 #include <cuda/std/__cccl/epilogue.h>
diff --git a/libcudacxx/test/libcudacxx/libcxx/numerics/floating.point/constants.pass.cpp b/libcudacxx/test/libcudacxx/libcxx/numerics/floating.point/constants.pass.cpp
index e58d8f02e62..821fb99c95c 100644
--- a/libcudacxx/test/libcudacxx/libcxx/numerics/floating.point/constants.pass.cpp
+++ b/libcudacxx/test/libcudacxx/libcxx/numerics/floating.point/constants.pass.cpp
@@ -71,6 +71,25 @@ __host__ __device__ void test_fp_storage()
     const auto ref = cuda::std::numeric_limits<T>::lowest();
     assert(cuda::std::memcmp(&val, &ref, sizeof(T)) == 0);
   }
+
+  // test __fp_zero to be all zeros
+#if _CCCL_HAS_NVFP8_E8M0()
+  if constexpr (!cuda::std::is_same_v<T, __nv_fp8_e8m0>)
+#endif // _CCCL_HAS_NVFP8_E8M0()
+  {
+    const auto val = cuda::std::__fp_zero<T>();
+    const auto ref = cuda::std::__fp_storage_of_t<T>(0);
+    assert(cuda::std::memcmp(&val, &ref, sizeof(T)) == 0);
+    assert(!cuda::std::signbit(val));
+  }
+
+  // test __fp_one for standard types only
+  if constexpr (cuda::std::__fp_is_native_type_v<T>)
+  {
+    const auto val = cuda::std::__fp_one<T>();
+    const auto ref = T{1};
+    assert(cuda::std::memcmp(&val, &ref, sizeof(T)) == 0);
+  }
 }
 
 int main(int, char**)
