diff --git a/benchmarks/bench_deepseek_mla.py b/benchmarks/bench_deepseek_mla.py
index 2f85aad18a..a42bec7b72 100644
--- a/benchmarks/bench_deepseek_mla.py
+++ b/benchmarks/bench_deepseek_mla.py
@@ -74,9 +74,6 @@ def bench_deepseek_mla_decode(batch_size, seq_len, num_heads):
 
 
 if __name__ == "__main__":
-    bench_deepseek_mla_decode(768, 1024, 16)
-    bench_deepseek_mla_decode(768, 1024, 32)
-    bench_deepseek_mla_decode(768, 1024, 64)
-    bench_deepseek_mla_decode(768, 2048, 16)
-    bench_deepseek_mla_decode(768, 2048, 32)
-    bench_deepseek_mla_decode(768, 2048, 64)
+    for seq_len in [1024, 2048, 4096, 8192, 16384, 32768]:
+        for batch_size in [1, 16, 32, 64]:
+            bench_deepseek_mla_decode(batch_size, seq_len, 16)
diff --git a/csrc/batch_mla_run.cu b/csrc/batch_mla_run.cu
index ea2dbb866a..fd42db41f9 100644
--- a/csrc/batch_mla_run.cu
+++ b/csrc/batch_mla_run.cu
@@ -72,6 +72,8 @@ void BatchMLAPagedAttentionRun(at::Tensor float_workspace_buffer, at::Tensor int
 
         params.q_indptr = GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.q_indptr_offset);
         params.kv_indptr = GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.kv_indptr_offset);
+        params.partial_indptr =
+            GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.partial_indptr_offset);
         params.kv_indices = static_cast<IdType*>(kv_indices.data_ptr());
         params.q_len = GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.q_len_offset);
         params.kv_len = GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.kv_len_offset);
@@ -80,6 +82,12 @@ void BatchMLAPagedAttentionRun(at::Tensor float_workspace_buffer, at::Tensor int
         params.kv_end = GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.kv_end_offset);
         params.work_indptr =
             GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.work_indptr_offset);
+        params.merge_packed_offset_start = GetPtrFromBaseOffset<IdType>(
+            int_buffer_ptr, plan_info.merge_packed_offset_start_offset);
+        params.merge_packed_offset_end =
+            GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.merge_packed_offset_end_offset);
+        params.merge_indptr =
+            GetPtrFromBaseOffset<IdType>(int_buffer_ptr, plan_info.merge_indptr_offset);
         params.final_o = static_cast<DTypeO*>(o.data_ptr());
         params.final_lse =
             maybe_lse.has_value() ? static_cast<float*>(maybe_lse->data_ptr()) : nullptr;
diff --git a/include/flashinfer/attention/cascade.cuh b/include/flashinfer/attention/cascade.cuh
index 96b769fd2b..142b1106e1 100644
--- a/include/flashinfer/attention/cascade.cuh
+++ b/include/flashinfer/attention/cascade.cuh
@@ -16,8 +16,6 @@
 #ifndef FLASHINFER_CASCADE_CUH_
 #define FLASHINFER_CASCADE_CUH_
 
-#include <cooperative_groups.h>
-
 #include "../cp_async.cuh"
 #include "../math.cuh"
 #include "../utils.cuh"
@@ -25,7 +23,6 @@
 
 namespace flashinfer {
 
-namespace cg = cooperative_groups;
 using cp_async::PrefetchMode;
 using cp_async::SharedMemFillMode;
 
@@ -323,8 +320,8 @@ __global__ void MergeStatesLargeNumIndexSetsKernel(DTypeIn* __restrict__ V, floa
 }
 
 /*!
- * \brief The CUDA kernel to merge self-attention states of multiple index sets, the number of index
- *   sets at each position might vary.
+ * \brief The CUDA kernel to merge self-attention states of multiple index sets, the number of
+ * index sets at each position might vary.
  *
  * For CUDA graph support, the kernel can be built with a maximum sequence length and executed
  * using a truncated, dynamic sequence length passed through `seq_len_ptr`.
diff --git a/include/flashinfer/attention/mla_fa2.cuh b/include/flashinfer/attention/mla_fa2.cuh
index f998f56a9e..a059fe399d 100644
--- a/include/flashinfer/attention/mla_fa2.cuh
+++ b/include/flashinfer/attention/mla_fa2.cuh
@@ -15,6 +15,8 @@
  */
 #ifndef FLASHINFER_MLA_FA2_CUH_
 #define FLASHINFER_MLA_FA2_CUH_
+#include <cooperative_groups.h>
+
 #include <cstdint>
 #include <sstream>
 
@@ -90,7 +92,6 @@ struct KernelTraits {
   static constexpr uint32_t UPCAST_STRIDE_KPE = HEAD_DIM_KPE / upcast_size<DTypeKV_>();
   static constexpr uint32_t UPCAST_STRIDE_FINAL_O = HEAD_DIM_CKV / upcast_size<DTypeO_>();
   static constexpr uint32_t UPCAST_STRIDE_P = CTA_TILE_KV / upcast_size<DTypeKV_>();
-  static constexpr uint32_t UPCAST_STRIDE_PARTIAL_O = HEAD_DIM_CKV / upcast_size<float>();
 
   using DTypeQ = DTypeQ_;
   using DTypeKV = DTypeKV_;
@@ -618,6 +619,52 @@ __device__ __forceinline__ void finalize_m_(typename KTraits::AttentionVariant v
   }
 }
 
+template <typename KTraits>
+__device__ void DevicePersistentMergeStates(typename KTraits::IdType* merge_packed_offset_start,
+                                            typename KTraits::IdType* merge_packed_offset_end,
+                                            typename KTraits::IdType* merge_indptr,
+                                            float* partial_o, float* partial_lse,
+                                            typename KTraits::DTypeO* final_o, float* final_lse,
+                                            const uint32_t o_stride_n, const uint32_t o_stride_h,
+                                            const uint32_t cluster_tile_q,
+                                            const uint_fastdiv& num_heads) {
+  constexpr uint32_t VEC_SIZE = 4;  // partial o has data type float
+  constexpr uint32_t NUM_THRS_PER_ROW = KTraits::HEAD_DIM_CKV / VEC_SIZE;
+  constexpr uint32_t ROWS_PER_ITERATION = (KTraits::NUM_THREADS) / NUM_THRS_PER_ROW;
+  const uint32_t cluster_id = blockIdx.y;
+  const uint32_t thread_id = (threadIdx.z * blockDim.y + threadIdx.y) * blockDim.x + threadIdx.x;
+  const uint32_t offset_start = merge_packed_offset_start[cluster_id];
+  const uint32_t offset_end = merge_packed_offset_end[cluster_id];
+  const uint32_t partial_offset_start = merge_indptr[cluster_id];
+  const uint32_t partial_offset_end = merge_indptr[cluster_id + 1];
+  const uint32_t stride = offset_end - offset_start;
+#pragma unroll 1
+  for (uint32_t local_packed_offset =
+           blockIdx.x * ROWS_PER_ITERATION + thread_id / NUM_THRS_PER_ROW;
+       local_packed_offset < stride; local_packed_offset += gridDim.x * ROWS_PER_ITERATION) {
+    uint32_t final_packed_offset = offset_start + local_packed_offset;
+    uint32_t q, r;
+    num_heads.divmod(final_packed_offset, q, r);
+    state_t<VEC_SIZE> st;
+#pragma unroll 2
+    for (uint32_t partial_packed_offset = partial_offset_start + local_packed_offset;
+         partial_packed_offset < partial_offset_end; partial_packed_offset += stride) {
+      vec_t<float, VEC_SIZE> o_partial;
+      float lse_partial;
+      o_partial.load(partial_o + partial_packed_offset * KTraits::HEAD_DIM_CKV +
+                     (thread_id % NUM_THRS_PER_ROW) * VEC_SIZE);
+      lse_partial = partial_lse[partial_packed_offset];
+      st.merge(o_partial, lse_partial, 1);
+    }
+    st.normalize();
+    st.o.cast_store(final_o +
+                    (q * o_stride_n + r * o_stride_h + (thread_id % NUM_THRS_PER_ROW) * VEC_SIZE));
+    if (final_lse) {
+      final_lse[q * num_heads + r] = st.get_lse();
+    }
+  }
+}
+
 template <typename KTraits>
 __device__ __forceinline__ void write_o(typename KTraits::SharedStorage* smem_storage,
                                         typename KTraits::DTypeO* final_o, float* final_lse,
@@ -631,12 +678,40 @@ __device__ __forceinline__ void write_o(typename KTraits::SharedStorage* smem_st
   constexpr uint32_t HEAD_DIM_CKV = KTraits::HEAD_DIM_CKV;
   constexpr uint32_t UPCAST_STRIDE_FINAL_O = KTraits::UPCAST_STRIDE_FINAL_O;
   const uint32_t lane_idx = threadIdx.x, warpgroup_idx = threadIdx.z, warp_idx_in_wg = threadIdx.y;
+  smem_t<KTraits::SWIZZLE_MODE_O> o_smem(smem_storage->o_smem);
+
+  if (partial_o != nullptr) {
+    // write to partial_o
+#pragma unroll
+    for (uint32_t j = 0; j < 2; ++j) {
+      uint32_t q_idx = (packed_offset + warp_idx_in_wg * 16 + 8 * j + lane_idx / 4) / num_heads;
+      if (lane_idx % 4 == 0 && q_idx < q_len) {
+        partial_lse[(blockIdx.x * 4 + warp_idx_in_wg) * 16 + 8 * j + lane_idx / 4] =
+            math::ptx_log2(d[j]) + float(m[j]);
+      }
+    }
 
-  if (false) {
-    // TOOD(Zihao): write to partial
+#pragma unroll
+    for (uint32_t j = 0; j < 2; ++j) {
+      uint32_t q_idx = (packed_offset + warp_idx_in_wg * 16 + 8 * j + lane_idx / 4) / num_heads;
+#pragma unroll
+      for (uint32_t mma_d = 0; mma_d < NUM_MMA_D_CKV / 2; ++mma_d) {
+        if (q_idx < q_len) {
+          *reinterpret_cast<float2*>(
+              partial_o +
+              ((blockIdx.x * 4 + warp_idx_in_wg) * 16 + 8 * j + lane_idx / 4) * HEAD_DIM_CKV +
+              warpgroup_idx * (HEAD_DIM_CKV / 2) + mma_d * 16 + (lane_idx % 4) * 2) =
+              *reinterpret_cast<float2*>(&o_frag[mma_d][j * 2]);
+          *reinterpret_cast<float2*>(
+              partial_o +
+              ((blockIdx.x * 4 + warp_idx_in_wg) * 16 + 8 * j + lane_idx / 4) * HEAD_DIM_CKV +
+              warpgroup_idx * (HEAD_DIM_CKV / 2) + mma_d * 16 + 8 + (lane_idx % 4) * 2) =
+              *reinterpret_cast<float2*>(&o_frag[mma_d][4 + j * 2]);
+        }
+      }
+    }
   } else {
     // write to final_o
-    smem_t<KTraits::SWIZZLE_MODE_O> o_smem(smem_storage->o_smem);
 
     if (final_lse) {
 #pragma unroll
@@ -748,13 +823,14 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
   const uint32_t kpe_stride_n = params.kpe_stride_n;
   const uint32_t o_stride_n = params.o_stride_n;
   const uint32_t o_stride_h = params.o_stride_h;
-  const uint32_t cluster_tile_q = blockDim.x * KTraits::CTA_TILE_Q;
+  const uint32_t cluster_tile_q = gridDim.x * KTraits::CTA_TILE_Q;
 
 #pragma unroll 1
   for (IdType work_idx = work_indptr[blockIdx.y]; work_idx < work_indptr[blockIdx.y + 1];
        ++work_idx) {
     const uint32_t q_indptr = params.q_indptr[work_idx];
     const uint32_t kv_indptr = params.kv_indptr[work_idx];
+    const int32_t partial_indptr = params.partial_indptr[work_idx];
     const uint32_t q_len = params.q_len[work_idx];
     const uint32_t kv_len = params.kv_len[work_idx];
     const uint32_t packed_qo_start = params.q_start[work_idx];
@@ -778,14 +854,14 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
             (CAUSAL ? min(kv_end, kv_len - q_len + (packed_qo_start + cluster_tile_q) / num_heads)
                     : kv_end),
             CTA_TILE_KV) -
-        1;
+        1 - (kv_start / CTA_TILE_KV);
 
     int mask_tile_idx =
-        (CAUSAL ? min(kv_end, kv_len - q_len + packed_qo_start / num_heads) : kv_end) / CTA_TILE_KV;
+        (CAUSAL ? min(kv_end, kv_len - q_len + packed_qo_start / num_heads) : kv_end) /
+            CTA_TILE_KV -
+        (kv_start / CTA_TILE_KV);
 
-    int start_tile_idx = kv_start / CTA_TILE_KV;  // ceil_div(kv_start, CTA_TILE_KV);
     uint32_t block_iter_base = kv_indptr * block_size + kv_start;
-
     // last kv tile
     __syncthreads();
     uint32_t kv_bound = kv_indptr + (kv_len + block_size - 1) / block_size;  // ceil_div
@@ -796,7 +872,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
     cp_async::commit_group();
 #pragma unroll
     for (int stage_idx = 1; stage_idx < NUM_STAGES; ++stage_idx) {
-      if (kv_tile_idx - stage_idx >= start_tile_idx) {
+      if (kv_tile_idx - stage_idx >= 0) {
         load_kv<KTraits>(&smem_storage, ckv, kpe, kv_indices, ckv_stride_n, ckv_stride_page,
                          kpe_stride_n, kpe_stride_page, kv_bound,
                          block_iter_base + (kv_tile_idx - stage_idx) * CTA_TILE_KV, block_size,
@@ -807,7 +883,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
 
     // loop with mask
 #pragma unroll 1
-    for (; kv_tile_idx >= mask_tile_idx && kv_tile_idx > start_tile_idx; --kv_tile_idx) {
+    for (; kv_tile_idx >= mask_tile_idx && kv_tile_idx > 0; --kv_tile_idx) {
       cp_async::wait_group<NUM_STAGES - 1>();
       __syncthreads();
 
@@ -825,7 +901,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
       // compute sfm * v
       compute_mla_pv<KTraits>(&smem_storage, kv_tile_idx % NUM_STAGES, s_frag, d, o_frag);
 
-      if (kv_tile_idx - NUM_STAGES >= start_tile_idx) {
+      if (kv_tile_idx - NUM_STAGES >= 0) {
         __syncthreads();
         load_kv<KTraits>(&smem_storage, ckv, kpe, kv_indices, ckv_stride_n, ckv_stride_page,
                          kpe_stride_n, kpe_stride_page, kv_bound,
@@ -837,7 +913,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
 
     // loop without mask
 #pragma unroll 1
-    for (; kv_tile_idx + 1 > start_tile_idx + NUM_STAGES; --kv_tile_idx) {
+    for (; kv_tile_idx + 1 > NUM_STAGES; --kv_tile_idx) {
       cp_async::wait_group<NUM_STAGES - 1>();
       __syncthreads();
 
@@ -862,7 +938,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
 
     // last tiles
 #pragma unroll
-    for (; kv_tile_idx >= start_tile_idx; --kv_tile_idx) {
+    for (; kv_tile_idx >= 0; --kv_tile_idx) {
       // compute mla qk
       compute_mla_qk<KTraits>(&smem_storage, kv_tile_idx % NUM_STAGES, s_frag);
 
@@ -884,11 +960,22 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchMLAPagedAttentionKe
 
     finalize_m_<KTraits>(variant, m);
 
-    write_o<KTraits>(&smem_storage, final_o + q_indptr * o_stride_n,
-                     final_lse ? final_lse + q_indptr * num_heads : nullptr, partial_o, partial_lse,
-                     o_frag, m, d, o_stride_n, o_stride_h, qo_upperbound, qo_packed_idx_base,
-                     num_heads);
+    write_o<KTraits>(
+        &smem_storage, final_o + q_indptr * o_stride_n,
+        final_lse ? final_lse + q_indptr * num_heads : nullptr,
+        (partial_indptr == -1) ? nullptr : partial_o + partial_indptr * KTraits::HEAD_DIM_CKV,
+        (partial_indptr == -1) ? nullptr : partial_lse + partial_indptr, o_frag, m, d, o_stride_n,
+        o_stride_h, qo_upperbound, qo_packed_idx_base, num_heads);
   }
+
+  auto grid = cg::this_grid();
+  grid.sync();
+
+  // the second stage, merge partial outputs
+  DevicePersistentMergeStates<KTraits>(params.merge_packed_offset_start,
+                                       params.merge_packed_offset_end, params.merge_indptr,
+                                       partial_o, partial_lse, final_o, final_lse, o_stride_n,
+                                       o_stride_h, cluster_tile_q, num_heads);
 }
 
 #define DISPATCH_SMEM_CONFIG(smem_limit_per_sm, NUM_STAGES, CTA_TILE_KV, QK_SHARD, ...) \
@@ -948,7 +1035,8 @@ cudaError_t BatchMLAPagedAttention(Params params, uint32_t num_blks_x, uint32_t
 
     FLASHINFER_CUDA_CALL(
         cudaFuncSetAttribute(kernel, cudaFuncAttributeMaxDynamicSharedMemorySize, smem_size));
-    FLASHINFER_CUDA_CALL(cudaLaunchKernel((void*)kernel, nblks, nthrs, args, smem_size, stream));
+    FLASHINFER_CUDA_CALL(
+        cudaLaunchCooperativeKernel((void*)kernel, nblks, nthrs, args, smem_size, stream));
   });
 
   return cudaSuccess;
diff --git a/include/flashinfer/attention/mla_params.cuh b/include/flashinfer/attention/mla_params.cuh
index 400222f3d9..1c354897c3 100644
--- a/include/flashinfer/attention/mla_params.cuh
+++ b/include/flashinfer/attention/mla_params.cuh
@@ -39,6 +39,10 @@ struct MLAParams {
 
   IdType* q_indptr;
   IdType* kv_indptr;
+  IdType* partial_indptr;
+  IdType* merge_packed_offset_start;
+  IdType* merge_packed_offset_end;
+  IdType* merge_indptr;
   IdType* kv_indices;
   IdType* q_len;
   IdType* kv_len;
diff --git a/include/flashinfer/attention/scheduler.cuh b/include/flashinfer/attention/scheduler.cuh
index a4cdd1ba5b..750710386b 100644
--- a/include/flashinfer/attention/scheduler.cuh
+++ b/include/flashinfer/attention/scheduler.cuh
@@ -44,10 +44,10 @@ template <uint32_t num_stages_smem, uint32_t vec_size_ckv, uint32_t vec_size_kpe
 __global__ void BatchDecodeWithPagedKVCacheKernelMLA(Params params);
 
 template <uint32_t HEAD_DIM_CKV, uint32_t HEAD_DIM_KPE, uint32_t QO_TILE_LEN, typename DTypeKV>
-std::tuple<uint32_t, uint32_t, uint32_t> LaunchSpecForDecodeKernelMlaCuteSM80(const uint32_t num_qo_heads);
+std::tuple<uint32_t, uint32_t, uint32_t> LaunchSpecForDecodeKernelMlaCuteSM80(
+    const uint32_t num_qo_heads);
 
-template <uint32_t HEAD_DIM_CKV, uint32_t HEAD_DIM_KPE, uint32_t QO_TILE_LEN,
-      typename Params>
+template <uint32_t HEAD_DIM_CKV, uint32_t HEAD_DIM_KPE, uint32_t QO_TILE_LEN, typename Params>
 __global__ void BatchDecodeWithPagedKVCacheKernelMlaCuteSM80(Params params);
 
 /*!
@@ -269,20 +269,23 @@ inline cudaError_t BatchDecodeWithPagedKVCacheWorkEstimationDispatchedMLA(
   });
 }
 
-template <uint32_t HEAD_DIM_CKV, uint32_t HEAD_DIM_KPE, uint32_t QO_TILE_LEN, typename AttentionVariant, typename Params>
+template <uint32_t HEAD_DIM_CKV, uint32_t HEAD_DIM_KPE, uint32_t QO_TILE_LEN,
+          typename AttentionVariant, typename Params>
 inline cudaError_t BatchDecodeWithPagedKVCacheWorkEstimationDispatchedMlaCuteSM80(
-  bool& split_kv, uint32_t& max_grid_size, uint32_t& max_num_pages_per_batch,
-  uint32_t& new_batch_size, uint32_t& gdy_, uint32_t batch_size,
-  typename Params::IdType* kv_indptr_h, const uint32_t num_qo_heads, const uint32_t page_size,
-  bool enable_cuda_graph, cudaStream_t stream) {
+    bool& split_kv, uint32_t& max_grid_size, uint32_t& max_num_pages_per_batch,
+    uint32_t& new_batch_size, uint32_t& gdy_, uint32_t batch_size,
+    typename Params::IdType* kv_indptr_h, const uint32_t num_qo_heads, const uint32_t page_size,
+    bool enable_cuda_graph, cudaStream_t stream) {
   using DTypeKV = typename Params::DTypeKV;
   using IdType = typename Params::IdType;
 
-  auto [smem_size, gdy, k_warps] = LaunchSpecForDecodeKernelMlaCuteSM80<HEAD_DIM_CKV, HEAD_DIM_KPE, QO_TILE_LEN , DTypeKV>(num_qo_heads);
+  auto [smem_size, gdy, k_warps] =
+      LaunchSpecForDecodeKernelMlaCuteSM80<HEAD_DIM_CKV, HEAD_DIM_KPE, QO_TILE_LEN, DTypeKV>(
+          num_qo_heads);
   gdy_ = gdy;
   const uint32_t num_threads = k_warps * 32;
   auto kernel =
-    BatchDecodeWithPagedKVCacheKernelMlaCuteSM80<HEAD_DIM_CKV, HEAD_DIM_KPE, QO_TILE_LEN, Params>;
+      BatchDecodeWithPagedKVCacheKernelMlaCuteSM80<HEAD_DIM_CKV, HEAD_DIM_KPE, QO_TILE_LEN, Params>;
   int num_blocks_per_sm;
   int num_sm = 0;
   int dev_id = 0;
@@ -291,28 +294,27 @@ inline cudaError_t BatchDecodeWithPagedKVCacheWorkEstimationDispatchedMlaCuteSM8
 
   // FLASHINFER_CUDA_CALL(cudaOccupancyMaxActiveBlocksPerMultiprocessor(&num_blocks_per_sm, kernel,
   //                                   num_threads, smem_size));
-  // fixme: num_blocks_per_sm is 0 derived from cudaOccupancyMaxActiveBlocksPerMultiprocessor at times, 
-  // and we fill smem with q-heads as many as possible, so num_blocks_per_sm should be 1
+  // fixme: num_blocks_per_sm is 0 derived from cudaOccupancyMaxActiveBlocksPerMultiprocessor at
+  // times, and we fill smem with q-heads as many as possible, so num_blocks_per_sm should be 1
   num_blocks_per_sm = 1;
-  
+
   max_grid_size = num_blocks_per_sm * num_sm;
   if (batch_size * gdy >= max_grid_size) {
     split_kv = false;
     max_num_pages_per_batch = 1;
     for (uint32_t batch_idx = 0; batch_idx < batch_size; ++batch_idx) {
       max_num_pages_per_batch = std::max<uint32_t>(
-        max_num_pages_per_batch, kv_indptr_h[batch_idx + 1] - kv_indptr_h[batch_idx]);
+          max_num_pages_per_batch, kv_indptr_h[batch_idx + 1] - kv_indptr_h[batch_idx]);
     }
     new_batch_size = batch_size;
-  } 
-  else {
+  } else {
     // compute max_num_pages_per_batch and new_batch_size
     std::vector<IdType> num_pages(batch_size);
     for (uint32_t batch_idx = 0; batch_idx < batch_size; ++batch_idx) {
       num_pages[batch_idx] = kv_indptr_h[batch_idx + 1] - kv_indptr_h[batch_idx];
     }
     std::tie(max_num_pages_per_batch, new_batch_size) =
-      PartitionPagedKVCacheBinarySearchMinNumPagePerBatch(max_grid_size, gdy, num_pages,
+        PartitionPagedKVCacheBinarySearchMinNumPagePerBatch(max_grid_size, gdy, num_pages,
                                                             std::max(128 / page_size, 1U));
     if (new_batch_size == batch_size && !enable_cuda_graph) {
       // do not use partition-kv kernel for short sequence, when not using CUDAGraph
@@ -326,7 +328,6 @@ inline cudaError_t BatchDecodeWithPagedKVCacheWorkEstimationDispatchedMlaCuteSM8
   return cudaSuccess;
 }
 
-
 /*!
  * \brief Partition Paged KV-Cache into multiple chunks on KV sequence length
  * \tparam IdType A template type indicates the index data type
@@ -799,7 +800,7 @@ struct PrefillPlanSM90Info {
         head_indices_offset(0),
         work_indptr_offset(0),
         same_schedule_for_all_heads(false) {}
-  
+
   // convert PrefillPlanSM90Info to std::vector<int64_t>
   std::vector<int64_t> ToVector() const {
     return {qo_tile_indices_offset, qo_indptr_offset,
@@ -971,11 +972,24 @@ inline cudaError_t PrefillSM90Plan(
   return cudaSuccess;
 }
 
+inline int packed_causal_kv_end(int qo_len, int kv_len, int qo_tile_idx, int cluster_tile_q,
+                                int num_qo_tiles, int group_size) {
+  if (qo_tile_idx + 1 == num_qo_tiles) {
+    return kv_len;
+  }
+  int kv_len_init = kv_len - qo_len;
+  return kv_len_init + (qo_tile_idx + 1) * cluster_tile_q / group_size;
+}
+
 struct MLAPlanInfo {
   int64_t num_blks_x;
   int64_t num_blks_y;
   int64_t q_indptr_offset;
   int64_t kv_indptr_offset;
+  int64_t partial_indptr_offset;
+  int64_t merge_packed_offset_start_offset;
+  int64_t merge_packed_offset_end_offset;
+  int64_t merge_indptr_offset;
   int64_t q_len_offset;
   int64_t kv_len_offset;
   int64_t q_start_offset;
@@ -986,13 +1000,26 @@ struct MLAPlanInfo {
   int64_t partial_lse_offset;
 
   std::vector<int64_t> ToVector() const {
-    return {num_blks_x,    num_blks_y,         q_indptr_offset,  kv_indptr_offset,
-            q_len_offset,  kv_len_offset,      q_start_offset,   kv_start_offset,
-            kv_end_offset, work_indptr_offset, partial_o_offset, partial_lse_offset};
+    return {num_blks_x,
+            num_blks_y,
+            q_indptr_offset,
+            kv_indptr_offset,
+            partial_indptr_offset,
+            merge_packed_offset_start_offset,
+            merge_packed_offset_end_offset,
+            merge_indptr_offset,
+            q_len_offset,
+            kv_len_offset,
+            q_start_offset,
+            kv_start_offset,
+            kv_end_offset,
+            work_indptr_offset,
+            partial_o_offset,
+            partial_lse_offset};
   }
 
   void FromVector(const std::vector<int64_t>& vec) {
-    if (vec.size() != 12) {
+    if (vec.size() != 16) {
       std::ostringstream err_msg;
       err_msg << "MLAPlanInfo::FromVector: vec.size() should be 12, but got " << vec.size();
       FLASHINFER_ERROR(err_msg.str());
@@ -1001,14 +1028,18 @@ struct MLAPlanInfo {
     num_blks_y = vec[1];
     q_indptr_offset = vec[2];
     kv_indptr_offset = vec[3];
-    q_len_offset = vec[4];
-    kv_len_offset = vec[5];
-    q_start_offset = vec[6];
-    kv_start_offset = vec[7];
-    kv_end_offset = vec[8];
-    work_indptr_offset = vec[9];
-    partial_o_offset = vec[10];
-    partial_lse_offset = vec[11];
+    partial_indptr_offset = vec[4];
+    merge_packed_offset_start_offset = vec[5];
+    merge_packed_offset_end_offset = vec[6];
+    merge_indptr_offset = vec[7];
+    q_len_offset = vec[8];
+    kv_len_offset = vec[9];
+    q_start_offset = vec[10];
+    kv_start_offset = vec[11];
+    kv_end_offset = vec[12];
+    work_indptr_offset = vec[13];
+    partial_o_offset = vec[14];
+    partial_lse_offset = vec[15];
   }
 };
 
@@ -1056,6 +1087,19 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
   const int cta_tile_q = 64;
   int cluster_tile_q = cluster_size * cta_tile_q;
 
+  int64_t total_kv_lens = 0;
+  for (auto& [_, qo_len, kv_len] : idx_qo_kv_len_vec) {
+    int packed_qo_len = qo_len * num_heads;
+    int num_qo_tiles = ceil_div(packed_qo_len, cluster_tile_q);
+    for (int qo_tile_idx = num_qo_tiles - 1; qo_tile_idx >= 0; --qo_tile_idx) {
+      int effective_kv_len = causal ? packed_causal_kv_end(qo_len, kv_len, qo_tile_idx,
+                                                           cluster_tile_q, num_qo_tiles, num_heads)
+                                    : kv_len;
+      total_kv_lens += effective_kv_len;
+    }
+  }
+  int kv_len_limit = ceil_div(ceil_div(total_kv_lens, num_clusters), 512L) * 512L;
+
   // step 1. load-balancing scheduling algorithm
   MinHeap cluster_cost_heap(num_clusters);
   std::vector<std::vector<IdType>> cluster_q_indptr(num_clusters, std::vector<IdType>()),
@@ -1064,28 +1108,57 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
       cluster_kv_len(num_clusters, std::vector<IdType>()),
       cluster_q_start(num_clusters, std::vector<IdType>()),
       cluster_kv_start(num_clusters, std::vector<IdType>()),
-      cluster_kv_end(num_clusters, std::vector<IdType>());
+      cluster_kv_end(num_clusters, std::vector<IdType>()),
+      cluster_partial_indptr(num_clusters, std::vector<IdType>());
+
+  std::vector<IdType> merge_packed_offset_start(num_clusters, 0),
+      merge_packed_offset_end(num_clusters, 0), merge_indptr(num_clusters + 1, 0);
+
+  int partial_o_nnz = 0;
+  int split_kv_count = 0;
 
   for (auto& [i, qo_len, kv_len] : idx_qo_kv_len_vec) {
     int packed_qo_len = qo_len * num_heads;
     int num_qo_tiles = ceil_div(packed_qo_len, cluster_tile_q);
     for (int qo_tile_idx = num_qo_tiles - 1; qo_tile_idx >= 0; --qo_tile_idx) {
-      auto [cluster_idx, accum_cost] = cluster_cost_heap.pop();
-      cluster_cost_heap.insert(
-          {cluster_idx,
-           accum_cost +
-               cost_function(
-                   cluster_tile_q,
-                   causal ? kv_len - (num_qo_tiles - qo_tile_idx - 1) * cluster_tile_q : kv_len)});
-      cluster_q_len[cluster_idx].push_back(qo_len);
-      cluster_kv_len[cluster_idx].push_back(kv_len);
-      cluster_q_indptr[cluster_idx].push_back(qo_indptr_h[i]);
-      cluster_kv_indptr[cluster_idx].push_back(kv_indptr_h[i]);
-      cluster_q_start[cluster_idx].push_back(qo_tile_idx * cluster_tile_q);
-      cluster_kv_start[cluster_idx].push_back(0);
-      cluster_kv_end[cluster_idx].push_back(kv_len);
+      int remaining_len = causal ? packed_causal_kv_end(qo_len, kv_len, qo_tile_idx, cluster_tile_q,
+                                                        num_qo_tiles, num_heads)
+                                 : kv_len;
+      int kv_start = 0;
+      bool split_kv = remaining_len > kv_len_limit;
+      if (split_kv) {
+        merge_indptr[split_kv_count] = partial_o_nnz;
+        merge_packed_offset_start[split_kv_count] =
+            qo_indptr_h[i] * num_heads + qo_tile_idx * cluster_tile_q;
+        merge_packed_offset_end[split_kv_count] =
+            qo_indptr_h[i] * num_heads +
+            std::min((qo_tile_idx + 1) * cluster_tile_q, packed_qo_len);
+      }
+      while (remaining_len > 0) {
+        auto [cluster_idx, accum_cost] = cluster_cost_heap.pop();
+        int actual_len = std::min(remaining_len, kv_len_limit);
+        cluster_cost_heap.insert(
+            {cluster_idx, accum_cost + cost_function(cluster_tile_q, actual_len)});
+        cluster_q_len[cluster_idx].push_back(qo_len);
+        cluster_kv_len[cluster_idx].push_back(kv_len);
+        cluster_q_indptr[cluster_idx].push_back(qo_indptr_h[i]);
+        cluster_kv_indptr[cluster_idx].push_back(kv_indptr_h[i]);
+        if (split_kv) {
+          cluster_partial_indptr[cluster_idx].push_back(partial_o_nnz);
+          partial_o_nnz += std::min(packed_qo_len - qo_tile_idx * cluster_tile_q, cluster_tile_q);
+        } else {
+          cluster_partial_indptr[cluster_idx].push_back(-1);
+        }
+        cluster_q_start[cluster_idx].push_back(qo_tile_idx * cluster_tile_q);
+        cluster_kv_start[cluster_idx].push_back(kv_start);
+        cluster_kv_end[cluster_idx].push_back(kv_start + actual_len);
+        remaining_len -= actual_len;
+        kv_start += actual_len;
+      }
+      split_kv_count += int(split_kv);
     }
   }
+  merge_indptr[split_kv_count] = partial_o_nnz;
 
   int max_total_num_works = 16384;  // NOTE(Zihao): adjust it later
 
@@ -1096,6 +1169,7 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
   int total_num_works = work_indptr_vec.back();
   auto q_indptr_vec = flatten(cluster_q_indptr, total_num_works);
   auto kv_indptr_vec = flatten(cluster_kv_indptr, total_num_works);
+  auto partial_indptr_vec = flatten(cluster_partial_indptr, total_num_works);
   auto q_len_vec = flatten(cluster_q_len, total_num_works);
   auto kv_len_vec = flatten(cluster_kv_len, total_num_works);
   auto q_start_vec = flatten(cluster_q_start, total_num_works);
@@ -1107,6 +1181,14 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
       int_allocator.aligned_alloc_offset(sizeof(IdType) * max_total_num_works, 16, "mla_q_indptr");
   plan_info.kv_indptr_offset =
       int_allocator.aligned_alloc_offset(sizeof(IdType) * max_total_num_works, 16, "mla_kv_indptr");
+  plan_info.partial_indptr_offset = int_allocator.aligned_alloc_offset(
+      sizeof(IdType) * max_total_num_works, 16, "mla_partial_indptr");
+  plan_info.merge_packed_offset_start_offset = int_allocator.aligned_alloc_offset(
+      sizeof(IdType) * num_clusters, 16, "mla_merge_packed_offset_start");
+  plan_info.merge_packed_offset_end_offset = int_allocator.aligned_alloc_offset(
+      sizeof(IdType) * num_clusters, 16, "mla_merge_packed_offset_end");
+  plan_info.merge_indptr_offset = int_allocator.aligned_alloc_offset(
+      sizeof(IdType) * (num_clusters + 1), 16, "mla_merge_indptr");
   plan_info.q_len_offset =
       int_allocator.aligned_alloc_offset(sizeof(IdType) * max_total_num_works, 16, "mla_q_len");
   plan_info.kv_len_offset =
@@ -1124,6 +1206,14 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
       GetPtrFromBaseOffset<IdType>(page_locked_int_buffer, plan_info.q_indptr_offset);
   IdType* cluster_kv_indptr_h =
       GetPtrFromBaseOffset<IdType>(page_locked_int_buffer, plan_info.kv_indptr_offset);
+  IdType* cluster_partial_indptr_h =
+      GetPtrFromBaseOffset<IdType>(page_locked_int_buffer, plan_info.partial_indptr_offset);
+  IdType* cluster_merge_packed_offset_start_h = GetPtrFromBaseOffset<IdType>(
+      page_locked_int_buffer, plan_info.merge_packed_offset_start_offset);
+  IdType* cluster_merge_packed_offset_end_h = GetPtrFromBaseOffset<IdType>(
+      page_locked_int_buffer, plan_info.merge_packed_offset_end_offset);
+  IdType* cluster_merge_indptr_h =
+      GetPtrFromBaseOffset<IdType>(page_locked_int_buffer, plan_info.merge_indptr_offset);
   IdType* cluster_q_len_h =
       GetPtrFromBaseOffset<IdType>(page_locked_int_buffer, plan_info.q_len_offset);
   IdType* cluster_kv_len_h =
@@ -1139,6 +1229,12 @@ inline cudaError_t MLAPlan(void* float_buffer, size_t float_workspace_size_in_by
 
   std::copy(q_indptr_vec.begin(), q_indptr_vec.end(), cluster_q_indptr_h);
   std::copy(kv_indptr_vec.begin(), kv_indptr_vec.end(), cluster_kv_indptr_h);
+  std::copy(partial_indptr_vec.begin(), partial_indptr_vec.end(), cluster_partial_indptr_h);
+  std::copy(merge_packed_offset_start.begin(), merge_packed_offset_start.end(),
+            cluster_merge_packed_offset_start_h);
+  std::copy(merge_packed_offset_end.begin(), merge_packed_offset_end.end(),
+            cluster_merge_packed_offset_end_h);
+  std::copy(merge_indptr.begin(), merge_indptr.end(), cluster_merge_indptr_h);
   std::copy(q_len_vec.begin(), q_len_vec.end(), cluster_q_len_h);
   std::copy(kv_len_vec.begin(), kv_len_vec.end(), cluster_kv_len_h);
   std::copy(q_start_vec.begin(), q_start_vec.end(), cluster_q_start_h);
