diff --git a/flashinfer/comm/trtllm_ar.py b/flashinfer/comm/trtllm_ar.py
index 24305cdad0..31d6f51892 100644
--- a/flashinfer/comm/trtllm_ar.py
+++ b/flashinfer/comm/trtllm_ar.py
@@ -751,6 +751,24 @@ def trtllm_custom_all_reduce(
     )
 
 
+# Heuristics based on all configs of trtllm_allreduce_fusion on B200.
+# Empirically, the fusion pattern and fp32_acc are irrelevant to the decision.
+_use_oneshot_heuristics: dict[int, int] = {
+    2: 512,
+    4: 64,
+    8: 42,
+}
+
+
+def _should_use_oneshot(
+    token_num: int, hidden_dim: int, dtype: torch.dtype, world_size: int
+) -> bool:
+    comm_size_mb = (
+        token_num * hidden_dim * 2 * world_size * dtype.itemsize / 1024 / 1024
+    )
+    return comm_size_mb <= _use_oneshot_heuristics[world_size]
+
+
 def trtllm_allreduce_fusion(
     allreduce_in: torch.Tensor,
     world_size: int,
@@ -783,7 +801,7 @@ def trtllm_allreduce_fusion(
     - hidden_dim: the dimension of the hidden states.
     - workspace_ptrs: the workspace pointers.
     - launch_with_pdl: whether to launch with pdl.
-    - use_oneshot: whether to use oneshot.
+    - use_oneshot: whether to use oneshot. If None, internal heuristics will be used.
     - trigger_completion_at_end: whether to trigger completion at the end.
     - fp32_acc: whether to use fp32 accumulation.
     - pattern_code: the pattern code.
@@ -797,14 +815,12 @@ def trtllm_allreduce_fusion(
     - rms_eps: the rms epsilon value.
     - scale_factor: the scale factor. For cudaGraphs safety, it should be a tensor.
     - layout_code: the layout code.
-
-    Note:
-    Regarding the `use_oneshot` parameter, you could force to use the one-shot strategy based on your use case.
-    Otherwise, it would be enabled if token_num is less than the one-shot max token number (currently 128) for min-latency mode.
     """
 
     if use_oneshot is None:
-        use_oneshot = token_num <= 128
+        use_oneshot = _should_use_oneshot(
+            token_num, hidden_dim, allreduce_in.dtype, world_size
+        )
 
     if not use_oneshot:
         assert token_num > world_size, "sequence length should be larger than tp_size"
