diff --git a/flashinfer/compilation_context.py b/flashinfer/compilation_context.py
index 2d68c797a9..dc0d20a584 100644
--- a/flashinfer/compilation_context.py
+++ b/flashinfer/compilation_context.py
@@ -36,9 +36,6 @@ def __init__(self):
             for arch in os.environ["FLASHINFER_CUDA_ARCH_LIST"].split(" "):
                 major, minor = arch.split(".")
                 major = int(major)
-                if major >= 9:
-                    if minor.isdigit():
-                        minor = str(minor) + "a"
                 self.TARGET_CUDA_ARCHS.add((int(major), str(minor)))
         else:
             try:
diff --git a/scripts/task_test_aot_build_import.sh b/scripts/task_test_aot_build_import.sh
index 6cf809dc2d..00abececd8 100755
--- a/scripts/task_test_aot_build_import.sh
+++ b/scripts/task_test_aot_build_import.sh
@@ -20,13 +20,13 @@ export MAX_JOBS
 export FLASHINFER_CUDA_ARCH_LIST=$(python3 -c '
 import torch
 cuda_ver = torch.version.cuda
-arches = ["7.5", "8.0", "8.9", "9.0"]
+arches = ["7.5", "8.0", "8.9", "9.0a"]
 if cuda_ver is not None:
     try:
         major, minor = map(int, cuda_ver.split(".")[:2])
         if (major, minor) >= (12, 8):
-            arches.append("10.0")
-            arches.append("12.0")
+            arches.append("10.0a")
+            arches.append("12.0a")
     except Exception:
         pass
 print(" ".join(arches))
