diff --git a/csrc/batch_decode_customize_config.jinja b/csrc/batch_decode_customize_config.jinja
index 24ba9f1a61..da80370220 100644
--- a/csrc/batch_decode_customize_config.jinja
+++ b/csrc/batch_decode_customize_config.jinja
@@ -3,6 +3,7 @@
 #include <flashinfer/math.cuh>
 #include <flashinfer/layout.cuh>
 #include <flashinfer/pos_enc.cuh>
+#include <flashinfer/attention/variant_helper.cuh>
 
 #define ADDITIONAL_FUNC_PARAMS {{ additional_func_params }}
 #define ADDITIONAL_PARAMS_SETTER {{ additional_params_setter }}
diff --git a/csrc/batch_prefill_customize_config.jinja b/csrc/batch_prefill_customize_config.jinja
index ccdbcdee37..77490d71b2 100644
--- a/csrc/batch_prefill_customize_config.jinja
+++ b/csrc/batch_prefill_customize_config.jinja
@@ -5,6 +5,7 @@
 #include <flashinfer/utils.cuh>
 #include <flashinfer/pos_enc.cuh>
 #include <flashinfer/fastdiv.cuh>
+#include <flashinfer/attention/variant_helper.cuh>
 
 #define ADDITIONAL_FUNC_PARAMS {{ additional_func_params }}
 #define ADDITIONAL_PARAMS_SETTER {{ additional_params_setter }}
diff --git a/csrc/batch_prefill_sm90_customize_config.jinja b/csrc/batch_prefill_sm90_customize_config.jinja
index 73cdd25ce3..5b10355fc8 100644
--- a/csrc/batch_prefill_sm90_customize_config.jinja
+++ b/csrc/batch_prefill_sm90_customize_config.jinja
@@ -1,5 +1,6 @@
 #pragma once
 #include <flashinfer/attention/hopper/attention_updater.cuh>
+#include <flashinfer/attention/hopper/variant_helper.cuh>
 #include <flashinfer/math.cuh>
 #include <flashinfer/layout.cuh>
 #include <flashinfer/cutlass_utils.cuh>
diff --git a/csrc/single_decode_customize_config.jinja b/csrc/single_decode_customize_config.jinja
index 8a6baec5f7..9bd3b429b2 100644
--- a/csrc/single_decode_customize_config.jinja
+++ b/csrc/single_decode_customize_config.jinja
@@ -2,6 +2,7 @@
 #include <flashinfer/math.cuh>
 #include <flashinfer/layout.cuh>
 #include <flashinfer/pos_enc.cuh>
+#include <flashinfer/attention/variant_helper.cuh>
 
 #define ADDITIONAL_FUNC_PARAMS {{ additional_func_params }}
 #define ADDITIONAL_PARAMS_SETTER {{ additional_params_setter }}
diff --git a/csrc/single_prefill_customize_config.jinja b/csrc/single_prefill_customize_config.jinja
index ed2498ab9e..fa31e08b7b 100644
--- a/csrc/single_prefill_customize_config.jinja
+++ b/csrc/single_prefill_customize_config.jinja
@@ -4,6 +4,7 @@
 #include <flashinfer/utils.cuh>
 #include <flashinfer/pos_enc.cuh>
 #include <flashinfer/fastdiv.cuh>
+#include <flashinfer/attention/variant_helper.cuh>
 
 #define ADDITIONAL_FUNC_PARAMS {{ additional_func_params }}
 #define ADDITIONAL_PARAMS_SETTER {{ additional_params_setter }}
diff --git a/csrc/single_prefill_sm90_customize_config.jinja b/csrc/single_prefill_sm90_customize_config.jinja
index 0c8a0e4d50..7922ca2ba9 100644
--- a/csrc/single_prefill_sm90_customize_config.jinja
+++ b/csrc/single_prefill_sm90_customize_config.jinja
@@ -1,5 +1,6 @@
 #pragma once
 #include <flashinfer/attention/hopper/attention_updater.cuh>
+#include <flashinfer/attention/hopper/variant_helper.cuh>
 #include <flashinfer/math.cuh>
 #include <flashinfer/layout.cuh>
 #include <flashinfer/cutlass_utils.cuh>
diff --git a/flashinfer/utils.py b/flashinfer/utils.py
index df74de81e0..3d6e708660 100644
--- a/flashinfer/utils.py
+++ b/flashinfer/utils.py
@@ -171,7 +171,7 @@ def _get_cache_alibi_slopes_buf(
     key = (f"alibi_slopes_{num_qo_heads}", device)
     buf = _cache_buf.get(key)
     if buf is None:
-        buf = (get_alibi_slopes(num_qo_heads) * log2e).to(device)
+        buf = get_alibi_slopes(num_qo_heads).to(device)
         _cache_buf[key] = buf
     return buf
 
diff --git a/include/flashinfer/attention/decode.cuh b/include/flashinfer/attention/decode.cuh
index d1aa9284d6..7f3da453ec 100644
--- a/include/flashinfer/attention/decode.cuh
+++ b/include/flashinfer/attention/decode.cuh
@@ -91,6 +91,10 @@ __device__ __forceinline__ void compute_qk(const Params& params, AttentionVarian
     const uint32_t pos = kv_idx_base + tz * tile_size + j;
     s[j] = variant.LogitsTransform(params, s[j], batch_idx, /*qo_idx=*/0, /*kv_idx=*/pos,
                                    qo_head_idx, kv_head_idx);
+    if constexpr (variant.use_softmax) {
+      s[j] *= variant.sm_scale_log2;
+    }
+
     bool mask = variant.LogitsMask(params, batch_idx, /*qo_idx=*/0, /*kv_idx=*/pos, qo_head_idx,
                                    kv_head_idx);
     s[j] = (iter_base + tz * tile_size + j < iter_bound && mask) ? s[j] : -math::inf;
@@ -263,11 +267,6 @@ __global__ void SingleDecodeWithKVCacheKernel(const __grid_constant__ Params par
     // do not apply rotary embedding to q matrix
     q_vec.cast_load(q + qo_head_idx * q_stride_h + tx * vec_size);
   }
-  // multiple q_vec by sm_scale
-#pragma unroll
-  for (uint32_t i = 0; i < vec_size; ++i) {
-    q_vec[i] = variant.QueryTransform(params, q_vec[i]);
-  }
   block.sync();
 
   uint32_t chunk_start = kv_chunk_idx * kv_chunk_size;
@@ -456,11 +455,6 @@ __global__ void BatchDecodeWithPagedKVCacheKernel(const __grid_constant__ Params
     // do not apply rotary embedding to q matrix
     q_vec.cast_load(q + batch_idx * q_stride_n + qo_head_idx * q_stride_h + tx * vec_size);
   }
-#pragma unroll
-  for (uint32_t i = 0; i < vec_size; ++i) {
-    q_vec[i] = variant.QueryTransform(params, q_vec[i]);
-  }
-  block.sync();
 
   // preload k/v tiles
   uint32_t stage_idx = 0;
diff --git a/include/flashinfer/attention/hopper/variant_helper.cuh b/include/flashinfer/attention/hopper/variant_helper.cuh
new file mode 100644
index 0000000000..22fda01f4a
--- /dev/null
+++ b/include/flashinfer/attention/hopper/variant_helper.cuh
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2024 by FlashInfer team.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FLASHINFER_ATTENTION_HOPPER_VARIANT_HELPER_H
+#define FLASHINFER_ATTENTION_HOPPER_VARIANT_HELPER_H
+
+#include <cuda_runtime.h>
+
+#include <cstdint>
+
+namespace flashinfer {
+
+#define REGISTER_QUERY_TRANSFORM(params, q, ...)                                            \
+  template <typename MainloopParams, typename T>                                            \
+  __device__ __forceinline__ T QueryTransform(const MainloopParams& params, void* q_smem) { \
+    __VA_ARGS__                                                                             \
+  }
+
+#define REGISTER_KEY_TRANSFORM(params, k, ...)                                            \
+  template <typename MainloopParams, typename T>                                          \
+  __device__ __forceinline__ T KeyTransform(const MainloopParams& params, void* k_smem) { \
+    __VA_ARGS__                                                                           \
+  }
+
+#define REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, \
+                                  kv_head_idx, ...)                                       \
+  template <typename MainloopParams, typename T>                                          \
+  __device__ __forceinline__ T LogitsTransform(                                           \
+      const MainloopParams& params, T logits, uint32_t batch_idx, uint32_t qo_idx,        \
+      uint32_t kv_idx, uint32_t qo_head_idx, uint32_t kv_head_idx) {                      \
+    __VA_ARGS__                                                                           \
+  }
+
+#define REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, ...) \
+  template <typename MainloopParams>                                                           \
+  __device__ __forceinline__ bool LogitsMask(const MainloopParams& params, uint32_t batch_idx, \
+                                             uint32_t qo_idx, uint32_t kv_idx,                 \
+                                             uint32_t qo_head_idx, uint32_t kv_head_idx) {     \
+    __VA_ARGS__                                                                                \
+  }
+
+struct AttentionVariantBase {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                            { return logits; })
+
+  REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                       { return true; })
+};
+
+}  // namespace flashinfer
+
+#endif  // FLASHINFER_ATTENTION_HOPPER_VARIANT_HELPER_H
diff --git a/include/flashinfer/attention/hopper/variants.cuh b/include/flashinfer/attention/hopper/variants.cuh
index 69ad56eebc..b9f658f84b 100644
--- a/include/flashinfer/attention/hopper/variants.cuh
+++ b/include/flashinfer/attention/hopper/variants.cuh
@@ -20,6 +20,7 @@
 
 #include "../../math.cuh"
 #include "attention_updater.cuh"
+#include "variant_helper.cuh"
 
 namespace flashinfer {
 
@@ -36,12 +37,8 @@ struct StandardAttention {
     return OnlineSoftmax<NUM_ROWS_PER_THREAD, /*WITH_SCALE=*/true>(sm_scale_log2);
   }
 
-  template <typename MainloopParams, typename T>
-  __device__ __forceinline__ T LogitsTransform(const MainloopParams& params, T logits,
-                                               uint32_t batch_idx, uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits;
-  }
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                            { return logits; })
 };
 
 struct LogitsSoftCap {
@@ -57,15 +54,11 @@ struct LogitsSoftCap {
 
   template <int NUM_ROWS_PER_THREAD>
   __device__ auto GetAttentionUpdater() {
-    return OnlineSoftmax<NUM_ROWS_PER_THREAD, /*WITH_SCALE=*/false>(0.);
+    return OnlineSoftmax<NUM_ROWS_PER_THREAD, /*WITH_SCALE=*/true>(post_tanh_scale);
   }
 
-  template <typename MainloopParams, typename T>
-  __device__ __forceinline__ T LogitsTransform(const MainloopParams& params, T logits,
-                                               uint32_t batch_idx, uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return math::tanh(logits * pre_tanh_scale) * post_tanh_scale;
-  }
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                            { return math::tanh(logits * pre_tanh_scale); })
 };
 
 template <bool use_logits_soft_cap>
diff --git a/include/flashinfer/attention/prefill.cuh b/include/flashinfer/attention/prefill.cuh
index 1adbbf51e4..15f9a579c4 100644
--- a/include/flashinfer/attention/prefill.cuh
+++ b/include/flashinfer/attention/prefill.cuh
@@ -506,26 +506,6 @@ __device__ __forceinline__ void q_smem_inplace_apply_rotary_with_pos(
   }
 }
 
-template <typename KTraits, typename Params>
-__device__ __forceinline__ void q_smem_inplace_transform(const Params& params,
-                                                         typename KTraits::AttentionVariant variant,
-                                                         smem_t<KTraits::SWIZZLE_MODE_Q>* q_smem) {
-  using DTypeQ = typename KTraits::DTypeQ;
-  const uint32_t warp_idx = get_warp_idx<KTraits>(), lane_idx = threadIdx.x;
-  constexpr uint32_t HEAD_DIM_QK = KTraits::HEAD_DIM_QK;
-  constexpr uint32_t NUM_WARPS = KTraits::NUM_WARPS;
-#pragma unroll
-  for (uint32_t i = 0; i < KTraits::CTA_TILE_Q * HEAD_DIM_QK / (NUM_WARPS * 256); ++i) {
-    vec_t<DTypeQ, 8> tmp;
-    tmp.load((DTypeQ*)(q_smem->base) + (i * NUM_WARPS + warp_idx) * 256 + lane_idx * 8);
-#pragma unroll
-    for (uint32_t reg_id = 0; reg_id < 8; ++reg_id) {
-      tmp[reg_id] = variant.QueryTransform(params, tmp[reg_id]);
-    }
-    tmp.store((DTypeQ*)(q_smem->base) + (i * NUM_WARPS + warp_idx) * 256 + lane_idx * 8);
-  }
-}
-
 template <typename KTraits>
 __device__ __forceinline__ void k_smem_inplace_apply_rotary(
     const uint32_t kv_idx_base, smem_t<KTraits::SWIZZLE_MODE_KV>* k_smem, uint32_t* k_smem_offset_r,
@@ -760,6 +740,7 @@ __device__ __forceinline__ void logits_mask(
 
 template <typename KTraits>
 __device__ __forceinline__ void update_mdo_states(
+    typename KTraits::AttentionVariant variant,
     typename KTraits::DTypeQKAccum (*s_frag)[KTraits::NUM_MMA_KV][8],
     float (*o_frag)[KTraits::NUM_MMA_D_VO][8], typename KTraits::DTypeQKAccum (*m)[2],
     float (*d)[2]) {
@@ -768,6 +749,7 @@ __device__ __forceinline__ void update_mdo_states(
   constexpr bool use_softmax = AttentionVariant::use_softmax;
 
   if constexpr (use_softmax) {
+    const float sm_scale = variant.sm_scale_log2;
     if constexpr (std::is_same_v<DTypeQKAccum, float>) {
 #pragma unroll
       for (uint32_t mma_q = 0; mma_q < KTraits::NUM_MMA_Q; ++mma_q) {
@@ -784,7 +766,7 @@ __device__ __forceinline__ void update_mdo_states(
           m[mma_q][j] = max(m[mma_q][j], math::shfl_xor_sync(m[mma_q][j], 0x2));
           m[mma_q][j] = max(m[mma_q][j], math::shfl_xor_sync(m[mma_q][j], 0x1));
 
-          float o_scale = math::ptx_exp2(m_prev - m[mma_q][j]);
+          float o_scale = math::ptx_exp2(m_prev * sm_scale - m[mma_q][j] * sm_scale);
           d[mma_q][j] *= o_scale;
 #pragma unroll
           for (uint32_t mma_d = 0; mma_d < KTraits::NUM_MMA_D_VO; ++mma_d) {
@@ -795,18 +777,19 @@ __device__ __forceinline__ void update_mdo_states(
           }
 #pragma unroll
           for (uint32_t mma_kv = 0; mma_kv < KTraits::NUM_MMA_KV; ++mma_kv) {
-            s_frag[mma_q][mma_kv][j * 2 + 0] =
-                math::ptx_exp2(s_frag[mma_q][mma_kv][j * 2 + 0] - m[mma_q][j]);
-            s_frag[mma_q][mma_kv][j * 2 + 1] =
-                math::ptx_exp2(s_frag[mma_q][mma_kv][j * 2 + 1] - m[mma_q][j]);
-            s_frag[mma_q][mma_kv][j * 2 + 4] =
-                math::ptx_exp2(s_frag[mma_q][mma_kv][j * 2 + 4] - m[mma_q][j]);
-            s_frag[mma_q][mma_kv][j * 2 + 5] =
-                math::ptx_exp2(s_frag[mma_q][mma_kv][j * 2 + 5] - m[mma_q][j]);
+            s_frag[mma_q][mma_kv][j * 2 + 0] = math::ptx_exp2(
+                s_frag[mma_q][mma_kv][j * 2 + 0] * sm_scale - m[mma_q][j] * sm_scale);
+            s_frag[mma_q][mma_kv][j * 2 + 1] = math::ptx_exp2(
+                s_frag[mma_q][mma_kv][j * 2 + 1] * sm_scale - m[mma_q][j] * sm_scale);
+            s_frag[mma_q][mma_kv][j * 2 + 4] = math::ptx_exp2(
+                s_frag[mma_q][mma_kv][j * 2 + 4] * sm_scale - m[mma_q][j] * sm_scale);
+            s_frag[mma_q][mma_kv][j * 2 + 5] = math::ptx_exp2(
+                s_frag[mma_q][mma_kv][j * 2 + 5] * sm_scale - m[mma_q][j] * sm_scale);
           }
         }
       }
     } else if constexpr (std::is_same_v<DTypeQKAccum, half>) {
+      const half2 sm_scale = __float2half2_rn(variant.sm_scale_log2);
 #pragma unroll
       for (uint32_t mma_q = 0; mma_q < KTraits::NUM_MMA_Q; ++mma_q) {
         half m_prev[2];
@@ -826,7 +809,7 @@ __device__ __forceinline__ void update_mdo_states(
             __hmax2(*(half2*)&m[mma_q], math::shfl_xor_sync(*(half2*)&m[mma_q], 0x1));
 #pragma unroll
         for (uint32_t j = 0; j < 2; ++j) {
-          float o_scale = math::ptx_exp2(float(m_prev[j] - m[mma_q][j]));
+          float o_scale = math::ptx_exp2(float(m_prev[j] * sm_scale.x - m[mma_q][j] * sm_scale.x));
           d[mma_q][j] *= o_scale;
 #pragma unroll
           for (uint32_t mma_d = 0; mma_d < KTraits::NUM_MMA_D_VO; ++mma_d) {
@@ -839,9 +822,9 @@ __device__ __forceinline__ void update_mdo_states(
 #pragma unroll
           for (uint32_t mma_kv = 0; mma_kv < KTraits::NUM_MMA_KV; ++mma_kv) {
             *(half2*)&s_frag[mma_q][mma_kv][j * 2] =
-                math::ptx_exp2(*(half2*)&s_frag[mma_q][mma_kv][j * 2] - m2);
-            *(half2*)&s_frag[mma_q][mma_kv][j * 2 + 4] =
-                math::ptx_exp2(*(half2*)&s_frag[mma_q][mma_kv][j * 2 + 4] - m2);
+                math::ptx_exp2(*(half2*)&s_frag[mma_q][mma_kv][j * 2] * sm_scale - m2 * sm_scale);
+            *(half2*)&s_frag[mma_q][mma_kv][j * 2 + 4] = math::ptx_exp2(
+                *(half2*)&s_frag[mma_q][mma_kv][j * 2 + 4] * sm_scale - m2 * sm_scale);
           }
         }
       }
@@ -959,6 +942,22 @@ __device__ __forceinline__ void normalize_d(float (*o_frag)[KTraits::NUM_MMA_D_V
   }
 }
 
+template <typename KTraits>
+__device__ __forceinline__ void finalize_m(typename KTraits::AttentionVariant variant,
+                                           typename KTraits::DTypeQKAccum (*m)[2]) {
+  if constexpr (variant.use_softmax) {
+#pragma unroll
+    for (uint32_t mma_q = 0; mma_q < KTraits::NUM_MMA_Q; ++mma_q) {
+#pragma unroll
+      for (uint32_t j = 0; j < 2; ++j) {
+        if (m[mma_q][j] != typename KTraits::DTypeQKAccum(-math::inf)) {
+          m[mma_q][j] *= variant.sm_scale_log2;
+        }
+      }
+    }
+  }
+}
+
 /*!
  * \brief Synchronize the states of the MDO kernel across the threadblock along threadIdx.z.
  */
@@ -1300,15 +1299,14 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void SinglePrefillWithKVCache
                                 group_size, &qo_smem);
 
     cp_async::commit_group();
-    cp_async::wait_group<0>();
-    block.sync();
 
     if constexpr (KTraits::POS_ENCODING_MODE == PosEncodingMode::kRoPELlama) {
+      cp_async::wait_group<0>();
+      block.sync();
       q_smem_inplace_apply_rotary<KTraits>(qo_packed_idx_base, qo_len, kv_len, group_size, &qo_smem,
                                            &q_smem_offset_r, rope_freq);
       block.sync();
     }
-    q_smem_inplace_transform<KTraits>(params, variant, &qo_smem);
 
     smem_t<SWIZZLE_MODE_KV> k_smem(smem_storage.k_smem), v_smem(smem_storage.v_smem);
 
@@ -1387,7 +1385,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void SinglePrefillWithKVCache
       }
 
       // compute m,d states in online softmax
-      update_mdo_states<KTraits>(s_frag, o_frag, m, d);
+      update_mdo_states<KTraits>(variant, s_frag, o_frag, m, d);
 
       block.sync();
       produce_kv<false, SharedMemFillMode::kNoFill, KTraits>(
@@ -1407,6 +1405,8 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void SinglePrefillWithKVCache
     cp_async::wait_group<0>();
     block.sync();
 
+    finalize_m<KTraits>(variant, m);
+
     // threadblock synchronization
     threadblock_sync_mdo_states<KTraits>(o_frag, &smem_storage, m, d, warp_idx, lane_idx);
 
@@ -1693,10 +1693,10 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithRaggedKV
                                 q_stride_h, group_size, &qo_smem);
 
     cp_async::commit_group();
-    cp_async::wait_group<0>();
-    block.sync();
 
     if constexpr (KTraits::POS_ENCODING_MODE == PosEncodingMode::kRoPELlama) {
+      cp_async::wait_group<0>();
+      block.sync();
       IdType* q_rope_offset = nullptr;
 
       if constexpr (has_maybe_q_rope_offset_v<Params>) {
@@ -1712,7 +1712,6 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithRaggedKV
       }
       block.sync();
     }
-    q_smem_inplace_transform<KTraits>(params, variant, &qo_smem);
 
     const uint32_t num_iterations = ceil_div(
         (MASK_MODE == MaskMode::kCausal
@@ -1803,7 +1802,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithRaggedKV
       }
 
       // compute m,d states in online softmax
-      update_mdo_states<KTraits>(s_frag, o_frag, m, d);
+      update_mdo_states<KTraits>(variant, s_frag, o_frag, m, d);
 
       block.sync();
       produce_kv<false, SharedMemFillMode::kNoFill, KTraits>(
@@ -1823,6 +1822,8 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithRaggedKV
     cp_async::wait_group<0>();
     block.sync();
 
+    finalize_m<KTraits>(variant, m);
+
     // threadblock synchronization
     threadblock_sync_mdo_states<KTraits>(o_frag, &smem_storage, m, d, warp_idx, lane_idx);
 
@@ -1975,10 +1976,10 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithPagedKVC
                                 q_stride_h, group_size, &qo_smem);
 
     cp_async::commit_group();
-    cp_async::wait_group<0>();
-    block.sync();
 
     if constexpr (KTraits::POS_ENCODING_MODE == PosEncodingMode::kRoPELlama) {
+      cp_async::wait_group<0>();
+      block.sync();
       IdType* q_rope_offset = nullptr;
       if constexpr (has_maybe_q_rope_offset_v<Params>) {
         q_rope_offset = params.maybe_q_rope_offset;
@@ -1993,7 +1994,6 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithPagedKVC
       }
       block.sync();
     }
-    q_smem_inplace_transform<KTraits>(params, variant, &qo_smem);
 
     smem_t<SWIZZLE_MODE_KV> k_smem(smem_storage.k_smem), v_smem(smem_storage.v_smem);
     size_t thr_local_kv_offset[NUM_MMA_KV * KV_THR_LAYOUT_COL / 2 / NUM_WARPS_Q];
@@ -2096,7 +2096,7 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithPagedKVC
       }
 
       // compute m,d states in online softmax
-      update_mdo_states<KTraits>(s_frag, o_frag, m, d);
+      update_mdo_states<KTraits>(variant, s_frag, o_frag, m, d);
 
       block.sync();
       page_produce_kv<false, KTraits>(k_smem, &k_smem_offset_w, paged_kv, (iter + 1) * CTA_TILE_KV,
@@ -2116,6 +2116,8 @@ __global__ __launch_bounds__(KTraits::NUM_THREADS) void BatchPrefillWithPagedKVC
     cp_async::wait_group<0>();
     block.sync();
 
+    finalize_m<KTraits>(variant, m);
+
     // threadblock synchronization
     threadblock_sync_mdo_states<KTraits>(o_frag, &smem_storage, m, d, warp_idx, lane_idx);
 
diff --git a/include/flashinfer/attention/scheduler.cuh b/include/flashinfer/attention/scheduler.cuh
index c7dce7fc58..0d8f52264f 100644
--- a/include/flashinfer/attention/scheduler.cuh
+++ b/include/flashinfer/attention/scheduler.cuh
@@ -398,7 +398,7 @@ inline cudaError_t DecodePlan(void* float_buffer, size_t float_workspace_size_in
   if (split_kv) {
     AlignedAllocator float_allocator(float_buffer, float_workspace_size_in_bytes);
     plan_info.v_offset = float_allocator.aligned_alloc_offset(
-        num_qo_heads * padded_batch_size * HEAD_DIM * sizeof(DTypeO), 16, "batch_decode_tmp_v");
+        num_qo_heads * padded_batch_size * HEAD_DIM * sizeof(float), 16, "batch_decode_tmp_v");
     plan_info.s_offset = float_allocator.aligned_alloc_offset(
         num_qo_heads * padded_batch_size * sizeof(float), 16, "batch_decode_tmp_s");
 
@@ -676,7 +676,7 @@ inline cudaError_t PrefillPlan(void* float_buffer, size_t float_workspace_size_i
   if (split_kv) {
     AlignedAllocator float_allocator(float_buffer, float_workspace_size_in_bytes);
     plan_info.v_offset = float_allocator.aligned_alloc_offset(
-        num_qo_heads * padded_batch_size * cta_tile_q * head_dim_vo * sizeof_dtype_o, 16,
+        num_qo_heads * padded_batch_size * cta_tile_q * head_dim_vo * sizeof(float), 16,
         "batch_prefill_tmp_v");
     plan_info.s_offset = float_allocator.aligned_alloc_offset(
         num_qo_heads * padded_batch_size * cta_tile_q * sizeof(float), 16, "batch_prefill_tmp_s");
diff --git a/include/flashinfer/attention/variant_helper.cuh b/include/flashinfer/attention/variant_helper.cuh
new file mode 100644
index 0000000000..5836321fce
--- /dev/null
+++ b/include/flashinfer/attention/variant_helper.cuh
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2024 by FlashInfer team.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef FLASHINFER_ATTENTION_VARIANT_HELPER_H
+#define FLASHINFER_ATTENTION_VARIANT_HELPER_H
+
+#include <cuda_runtime.h>
+
+#include <cstdint>
+
+namespace flashinfer {
+
+#define REGISTER_QUERY_TRANSFORM(params, q, ...)                                    \
+  template <typename Params, typename T>                                            \
+  __device__ __forceinline__ T QueryTransform(const Params& params, void* q_smem) { \
+    __VA_ARGS__                                                                     \
+  }
+
+#define REGISTER_KEY_TRANSFORM(params, k, ...)                                    \
+  template <typename Params, typename T>                                          \
+  __device__ __forceinline__ T KeyTransform(const Params& params, void* k_smem) { \
+    __VA_ARGS__                                                                   \
+  }
+
+#define REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx,          \
+                                  kv_head_idx, ...)                                                \
+  template <typename Params, typename T>                                                           \
+  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx, \
+                                               uint32_t qo_idx, uint32_t kv_idx,                   \
+                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {       \
+    __VA_ARGS__                                                                                    \
+  }
+
+#define REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, ...) \
+  template <typename Params>                                                                   \
+  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,         \
+                                             uint32_t qo_idx, uint32_t kv_idx,                 \
+                                             uint32_t qo_head_idx, uint32_t kv_head_idx) {     \
+    __VA_ARGS__                                                                                \
+  }
+
+struct AttentionVariantBase {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                            { return logits; })
+
+  REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx,
+                       { return true; })
+};
+
+}  // namespace flashinfer
+
+#endif  // FLASHINFER_ATTENTION_VARIANT_HELPER_H
diff --git a/include/flashinfer/attention/variants.cuh b/include/flashinfer/attention/variants.cuh
index 6955bf2952..9d6f546d72 100644
--- a/include/flashinfer/attention/variants.cuh
+++ b/include/flashinfer/attention/variants.cuh
@@ -22,193 +22,21 @@
 
 #include "../math.cuh"
 #include "../utils.cuh"
+#include "variant_helper.cuh"
 
 namespace flashinfer {
 
-// Query Transform function that multiplies the query matrix by sm_scale
-struct StandardAttention {
-  static constexpr bool use_softmax = true;
-
-  uint32_t window_left, qo_len, kv_len;
-
-  // Create closure
-  template <typename Params>
-  __device__ __host__ StandardAttention(const Params& params, uint32_t batch_idx,
-                                        uint8_t* smem_ptr) {
-    qo_len = params.get_qo_len(batch_idx);
-    kv_len = params.get_kv_len(batch_idx);
-    window_left = kv_len;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
-};
-
 DEFINE_HAS_MEMBER(maybe_mask_indptr)
 
-struct CustomMaskAttention {
-  static constexpr bool use_softmax = true;
-
-  uint8_t* custom_mask_ptr;
-  uint32_t window_left, qo_len, kv_len;
-
-  // Create closure
-  template <typename Params>
-  __device__ __host__ CustomMaskAttention(const Params& params, uint32_t batch_idx,
-                                          uint8_t* smem_ptr) {
-    if constexpr (has_maybe_mask_indptr_v<Params>) {
-      custom_mask_ptr = params.maybe_custom_mask + params.maybe_mask_indptr[batch_idx];
-    } else {
-      custom_mask_ptr = params.maybe_custom_mask;
-    }
-    qo_len = params.get_qo_len(batch_idx);
-    kv_len = params.get_kv_len(batch_idx);
-    window_left = kv_len;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    const uint32_t offset = qo_idx * kv_len + kv_idx;
-    return ((custom_mask_ptr[offset / 8] >> (offset % 8)) & 1);
-  }
-};
-
-struct SlidingWindowAttention {
-  static constexpr bool use_softmax = true;
-
-  uint32_t window_left, qo_len, kv_len;
-
-  // Create closure
-  template <typename Params>
-  __device__ __host__ __forceinline__ SlidingWindowAttention(const Params& params,
-                                                             uint32_t batch_idx,
-                                                             uint8_t* smem_ptr) {
-    qo_len = params.get_qo_len(batch_idx);
-    kv_len = params.get_kv_len(batch_idx);
-    window_left = (params.window_left >= 0) ? params.window_left : kv_len;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return (kv_idx + qo_len + window_left >= kv_len + qo_idx);
-  }
-};
-
-struct LogitsSoftCap {
-  static constexpr bool use_softmax = true;
-
-  uint32_t window_left, qo_len, kv_len;
-
-  template <typename Params>
-  __device__ __host__ LogitsSoftCap(const Params& params, uint32_t batch_idx, uint8_t* smem_ptr) {
-    qo_len = params.get_qo_len(batch_idx);
-    kv_len = params.get_kv_len(batch_idx);
-    window_left = kv_len;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::ptx_rcp(params.logits_soft_cap);
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return params.logits_soft_cap * math::log2e * float(math::tanh(logits));
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
-};
-
-struct ALIBIAttention {
-  static constexpr bool use_softmax = true;
-
-  uint32_t window_left, qo_len, kv_len;
-
-  template <typename Params>
-  __device__ __host__ ALIBIAttention(const Params& params, uint32_t batch_idx, uint8_t* smem_ptr) {
-    qo_len = params.get_qo_len(batch_idx);
-    kv_len = params.get_kv_len(batch_idx);
-    window_left = kv_len;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits + params.maybe_alibi_slopes[qo_head_idx] * float(int(kv_idx) - int(qo_idx));
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
-};
-
 template <bool use_custom_mask, bool use_sliding_window, bool use_logits_soft_cap, bool use_alibi>
-struct DefaultAttention {
+struct DefaultAttention : AttentionVariantBase {
   static constexpr bool use_softmax = true;
 
-  uint32_t qo_len, kv_len;
   uint8_t* custom_mask_ptr;
+  uint32_t qo_len, kv_len;
   uint32_t window_left;
+  float sm_scale_log2;
+  float soft_cap_pre_tanh_scale;
 
   // Create closure
   template <typename Params>
@@ -216,6 +44,16 @@ struct DefaultAttention {
                                        uint8_t* smem_ptr) {
     qo_len = params.get_qo_len(batch_idx);
     kv_len = params.get_kv_len(batch_idx);
+    if constexpr (use_logits_soft_cap) {
+      soft_cap_pre_tanh_scale = params.sm_scale * math::ptx_rcp(params.logits_soft_cap);
+      sm_scale_log2 = math::log2e * params.logits_soft_cap;
+    } else {
+      if constexpr (use_alibi) {
+        sm_scale_log2 = math::log2e;
+      } else {
+        sm_scale_log2 = params.sm_scale * math::log2e;
+      }
+    }
     if constexpr (use_custom_mask) {
       if constexpr (has_maybe_mask_indptr_v<Params>) {
         custom_mask_ptr = params.maybe_custom_mask + params.maybe_mask_indptr[batch_idx];
@@ -228,32 +66,18 @@ struct DefaultAttention {
     }
   }
 
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    if constexpr (use_logits_soft_cap) {
-      return float(q) * params.sm_scale * math::ptx_rcp(params.logits_soft_cap);
-    } else {
-      return float(q) * params.sm_scale * math::log2e;
-    }
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     if constexpr (use_alibi) {
-      logits = logits + params.maybe_alibi_slopes[qo_head_idx] * float(int(kv_idx) - int(qo_idx));
+      logits = logits * params.sm_scale +
+               params.maybe_alibi_slopes[qo_head_idx] * float(int(kv_idx) - int(qo_idx));
     }
     if constexpr (use_logits_soft_cap) {
-      logits = params.logits_soft_cap * math::log2e * float(math::tanh(logits));
+      logits = float(math::tanh(logits * soft_cap_pre_tanh_scale));
     }
     return logits;
-  }
+  })
 
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
+  REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     bool mask = true;
     if constexpr (use_custom_mask) {
       const uint32_t offset = qo_idx * kv_len + kv_idx;
@@ -263,9 +87,9 @@ struct DefaultAttention {
       mask &= (kv_idx + qo_len + window_left >= kv_len + qo_idx);
     }
     return mask;
-  }
+  })
 };
 
-}  // namespace flashinfer
+};  // namespace flashinfer
 
 #endif  // FLASHINFER_ATTENTION_VARIANTS_CUH_
diff --git a/tests/test_jit_example.py b/tests/test_jit_example.py
index ba99242306..84fb86bc0c 100644
--- a/tests/test_jit_example.py
+++ b/tests/test_jit_example.py
@@ -18,11 +18,12 @@
 def test_single_decode_mask():
     torch.manual_seed(42)
     variant_decl = r"""
-struct SingleDecodeWithCustomMask {
+struct SingleDecodeWithCustomMask : AttentionVariantBase {
   static constexpr bool use_softmax = true;
 
   uint8_t* custom_mask_ptr;
   uint32_t window_left, qo_len, kv_len;
+  float sm_scale_log2;
 
   // Create closure
   template <typename Params>
@@ -32,27 +33,13 @@ def test_single_decode_mask():
     qo_len = 1;
     kv_len = params.get_kv_len(batch_idx);
     window_left = kv_len;
+    sm_scale_log2 = params.sm_scale * math::log2e;
   }
 
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
+  REGISTER_LOGITS_MASK(params, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     const uint32_t offset = kv_idx;
     return ((custom_mask_ptr[offset / 8] >> (offset % 8)) & 1);
-  }
+  })
 };
 """
     jit_module = gen_customize_single_decode_module(
@@ -89,11 +76,12 @@ def test_single_decode_mask():
 
 
 flash_sigmoid_sm80_decl = r"""
-struct FlashSigmoid {
+struct FlashSigmoid : AttentionVariantBase {
   static constexpr bool use_softmax = false;
 
   uint32_t window_left, qo_len, kv_len;
-  float sigmoid_bias_log2e;
+  float sigmoid_scale_log2;
+  float sigmoid_bias_log2;
 
   // Create closure
   template <typename Params>
@@ -102,32 +90,18 @@ def test_single_decode_mask():
     qo_len = params.get_qo_len(batch_idx);
     kv_len = params.get_kv_len(batch_idx);
     window_left = kv_len;
-    sigmoid_bias_log2e = params.sigmoid_bias * math::log2e;
+    sigmoid_bias_log2 = params.sigmoid_bias * math::log2e;
+    sigmoid_scale_log2 = params.logits_scale * math::log2e;
   }
 
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.logits_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
-    return math::ptx_rcp(1.f + math::ptx_exp2(-float(logits + sigmoid_bias_log2e)));
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
+    return math::ptx_rcp(1.f + math::ptx_exp2(-float(logits * sigmoid_scale_log2 + sigmoid_bias_log2)));
+  });
 };
 """
 
 flash_sigmoid_sm90_decl = r"""
-struct FlashSigmoid {
+struct FlashSigmoid : AttentionVariantBase {
   float logits_scale_log2, sigmoid_bias_log2e;
   // Init
   template <typename MainloopParams, typename BlockCoord>
@@ -142,13 +116,9 @@ def test_single_decode_mask():
     return DefaultUpdater<NUM_ROWS_PER_THREAD>();
   }
 
-  template <typename MainloopParams, typename T>
-  __device__ __forceinline__ T LogitsTransform(const MainloopParams& params, T logits,
-                                               int batch_idx,
-                                               int qo_idx, int kv_idx,
-                                               int qo_head_idx, int kv_head_idx) {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     return math::ptx_rcp(1.f + math::ptx_exp2(-float(logits * logits_scale_log2 + sigmoid_bias_log2e)));
-  }
+  });
 };
 """
 
@@ -191,10 +161,11 @@ def test_flash_sigmoid():
 def test_dump_logits():
     torch.manual_seed(42)
     variant_decl = r"""
-struct DumpLogits {
+struct DumpLogits : AttentionVariantBase {
   static constexpr bool use_softmax = true;
 
   uint32_t window_left, qo_len, kv_len;
+  float sm_scale_log2;
 
   // Create closure
   template <typename Params>
@@ -203,29 +174,15 @@ def test_dump_logits():
     qo_len = params.get_qo_len(batch_idx);
     kv_len = params.get_kv_len(batch_idx);
     window_left = kv_len;
+    sm_scale_log2 = params.sm_scale * math::log2e;
   }
 
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     if (qo_idx < qo_len && kv_idx < kv_len) {
-      params.output_logits[qo_head_idx * (qo_len * kv_len) + qo_idx * kv_len + kv_idx] = logits * math::loge2;
+      params.output_logits[qo_head_idx * (qo_len * kv_len) + qo_idx * kv_len + kv_idx] = logits * params.sm_scale;
     }
     return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
+  });
 };
 """
     jit_module = gen_customize_single_prefill_module(
@@ -607,10 +564,11 @@ def test_batch_prefill_sm90_flash_sigmoid():
 def test_debug_print_logits():
     torch.manual_seed(42)
     variant_decl = r"""
-struct DebugPrintLogits {
+struct DebugPrintLogits : AttentionVariantBase {
   static constexpr bool use_softmax = true;
 
   uint32_t window_left, qo_len, kv_len;
+  float sm_scale_log2;
 
   // Create closure
   template <typename Params>
@@ -619,30 +577,16 @@ def test_debug_print_logits():
     qo_len = params.get_qo_len(batch_idx);
     kv_len = params.get_kv_len(batch_idx);
     window_left = kv_len;
+    sm_scale_log2 = params.sm_scale * math::log2e;
   }
 
-  template <typename Params, typename T>
-  __device__ __forceinline__ T QueryTransform(const Params& params, T q) {
-    return float(q) * params.sm_scale * math::log2e;
-  }
-
-  template <typename Params, typename T>
-  __device__ __forceinline__ T LogitsTransform(const Params& params, T logits, uint32_t batch_idx,
-                                               uint32_t qo_idx, uint32_t kv_idx,
-                                               uint32_t qo_head_idx, uint32_t kv_head_idx) {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     if (logits >= 5) {
       printf("Large logits at qo_idx=%d, kv_idx=%d, qo_head_idx=%d, kv_head_idx=%d: %.3f\n",
-             qo_idx, kv_idx, qo_head_idx, kv_head_idx, float(logits));
+              qo_idx, kv_idx, qo_head_idx, kv_head_idx, float(logits));
     }
     return logits;
-  }
-
-  template <typename Params>
-  __device__ __forceinline__ bool LogitsMask(const Params& params, uint32_t batch_idx,
-                                             uint32_t qo_idx, uint32_t kv_idx, uint32_t qo_head_idx,
-                                             uint32_t kv_head_idx) {
-    return true;
-  }
+  });
 };
 """
     jit_module = gen_customize_single_prefill_module(
@@ -677,7 +621,7 @@ def test_debug_print_logits():
 def test_sm90_debug_print_logits():
     torch.manual_seed(42)
     variant_decl = r"""
-struct DebugPrintLogits {
+struct DebugPrintLogits : AttentionVariantBase {
   float sm_scale_log2;
   int qo_len, kv_len;
 
@@ -699,27 +643,23 @@ def test_sm90_debug_print_logits():
   }
 
 
-  template <typename MainloopParams, typename T>
-  __device__ __forceinline__ T LogitsTransform(const MainloopParams& params, T logits,
-                                               int batch_idx,
-                                               int qo_idx, int kv_idx,
-                                               int qo_head_idx, int kv_head_idx) {
+  REGISTER_LOGITS_TRANSFORM(params, logits, batch_idx, qo_idx, kv_idx, qo_head_idx, kv_head_idx, {
     if (qo_idx < qo_len && kv_idx < kv_len) {
-      printf(
-          "---> LOGITS DEBUG: "
-          "qo_idx=%-5d "
-          "kv_idx=%-5d "
-          "sm_scale_log2=%-12.5f "
-          "logits=%-12.5f "
-          "\n",
-          qo_idx,
-          kv_idx,
-          sm_scale_log2,
-          static_cast<float>(logits));
+        printf(
+            "---> LOGITS DEBUG: "
+            "qo_idx=%-5d "
+            "kv_idx=%-5d "
+            "sm_scale_log2=%-12.5f "
+            "logits=%-12.5f "
+            "\n",
+            qo_idx,
+            kv_idx,
+            sm_scale_log2,
+            static_cast<float>(logits));
     }
     logits *= sm_scale_log2;
     return logits;
-  }
+  })
 };
 """
     jit_module = gen_customize_single_prefill_module(
