diff --git a/cudax/include/cuda/experimental/__execution/bulk.cuh b/cudax/include/cuda/experimental/__execution/bulk.cuh
index a7ae38ed73f..6fc4f745056 100644
--- a/cudax/include/cuda/experimental/__execution/bulk.cuh
+++ b/cudax/include/cuda/experimental/__execution/bulk.cuh
@@ -73,7 +73,8 @@ struct _CCCL_TYPE_VISIBILITY_DEFAULT __attrs_t
   {
     constexpr int __block_threads = 256;
     const int __grid_blocks       = ::cuda::ceil_div(static_cast<int>(__shape), __block_threads);
-    return make_config(block_dims<__block_threads>(), grid_dims(__grid_blocks));
+    auto __dims                   = ::cuda::make_hierarchy(block_dims<__block_threads>(), grid_dims(__grid_blocks));
+    return make_config(__dims, cooperative_launch());
   }
 
   using __launch_config_t = decltype(__get_launch_config(_Shape()));
diff --git a/cudax/include/cuda/experimental/__execution/stream/adaptor.cuh b/cudax/include/cuda/experimental/__execution/stream/adaptor.cuh
index 4a7dc3ef5ef..5fca08ae21c 100644
--- a/cudax/include/cuda/experimental/__execution/stream/adaptor.cuh
+++ b/cudax/include/cuda/experimental/__execution/stream/adaptor.cuh
@@ -22,10 +22,9 @@
 #endif // no system header
 
 #include <cuda/__launch/configuration.h>
-#include <cuda/__utility/immovable.h>
+#include <cuda/__launch/launch.h>
 #include <cuda/std/__concepts/concept_macros.h>
 #include <cuda/std/__memory/unique_ptr.h>
-#include <cuda/std/__type_traits/is_same.h>
 #include <cuda/std/__type_traits/remove_cvref.h>
 #include <cuda/std/__utility/pod_tuple.h>
 
@@ -271,20 +270,24 @@ private:
     auto const __launch_config    = get_launch_config(execution::get_env(__state.__state_.__rcvr_));
     using __launch_dims_t         = decltype(__launch_config.dims);
     constexpr int __block_threads = __launch_dims_t::static_count(thread, block);
-    int const __grid_blocks       = __launch_config.dims.count(block, grid);
-    static_assert(__block_threads != ::cuda::std::dynamic_extent);
 
     // Start the child operation state. This will launch kernels for all the predecessors
     // of this operation.
     execution::start(__state.__opstate_);
 
-    // launch a kernel to pass the results to the receiver.
-    __completion_kernel<__block_threads><<<__grid_blocks, __block_threads, 0, __stream_.get()>>>(&__state.__state_);
-
-    // Check for errors in the kernel launch.
-    if (auto __status = cudaGetLastError(); __status != cudaSuccess)
+    _CCCL_TRY
+    {
+      // launch a kernel to pass the results to the receiver.
+      auto* __kernel = &__completion_kernel<__block_threads, _Rcvr, __results_t>;
+      ::cuda::launch(__stream_, __launch_config, __kernel, &__state.__state_);
+    }
+    _CCCL_CATCH (::cuda::cuda_error & __error) // Check for errors in the kernel launch.
+    {
+      execution::set_error(static_cast<_Rcvr&&>(__state.__state_.__rcvr_), __error.status());
+    }
+    _CCCL_CATCH_ALL
     {
-      execution::set_error(static_cast<_Rcvr&&>(__state.__state_.__rcvr_), cudaError_t(__status));
+      execution::set_error(static_cast<_Rcvr&&>(__state.__state_.__rcvr_), cudaErrorUnknown);
     }
   }
 
diff --git a/cudax/include/cuda/experimental/__execution/stream/bulk.cuh b/cudax/include/cuda/experimental/__execution/stream/bulk.cuh
index 1b86b3d9060..0738bd0a602 100644
--- a/cudax/include/cuda/experimental/__execution/stream/bulk.cuh
+++ b/cudax/include/cuda/experimental/__execution/stream/bulk.cuh
@@ -29,7 +29,9 @@
 #include <cuda/experimental/__execution/cpos.cuh>
 #include <cuda/experimental/__execution/policy.cuh>
 #include <cuda/experimental/__execution/stream/domain.cuh>
+#include <cuda/experimental/launch.cuh>
 
+#include <cooperative_groups.h>
 #include <cuda_runtime_api.h>
 
 #include <cuda/experimental/__execution/prologue.cuh>
@@ -66,7 +68,7 @@ struct __bulk_chunked_t : execution::__bulk_t<__bulk_chunked_t>
         }
       }
 
-      __syncthreads();
+      ::cooperative_groups::this_grid().sync();
 
       // Only call the downstream receiver once, after all threads have processed their
       // elements.
@@ -130,7 +132,7 @@ struct _CCCL_TYPE_VISIBILITY_DEFAULT __bulk_unchunked_t : execution::__bulk_t<__
         this->__state_->__fn_(_Shape(__tid), __values...);
       }
 
-      __syncthreads();
+      ::cooperative_groups::this_grid().sync();
 
       // Only call the downstream receiver once, after all threads have processed their
       // elements.
