diff --git a/cub/cub/block/block_load_to_shared.cuh b/cub/cub/block/block_load_to_shared.cuh
index 9692c64f872..e5836b8d27c 100644
--- a/cub/cub/block/block_load_to_shared.cuh
+++ b/cub/cub/block/block_load_to_shared.cuh
@@ -248,12 +248,7 @@ public:
     __syncthreads();
     if (elected)
     {
-      NV_IF_TARGET(NV_PROVIDES_SM_90,
-                   (
-                     // Borrowed from cuda::barrier
-                     // TODO Make this available through cuda::ptx::
-                     asm volatile("mbarrier.inval.shared.b64 [%0];" ::"r"(static_cast<::cuda::std::uint32_t>(
-                       ::__cvta_generic_to_shared(&temp_storage.mbarrier_handle))) : "memory");));
+      NV_IF_TARGET(NV_PROVIDES_SM_90, ::cuda::ptx::mbarrier_inval(&temp_storage.mbarrier_handle););
     }
     // Make sure the elected thread is done invalidating the mbarrier
     __syncthreads();
diff --git a/docs/libcudacxx/ptx/instructions.rst b/docs/libcudacxx/ptx/instructions.rst
index dfa988ba9bc..abb34e3df8c 100644
--- a/docs/libcudacxx/ptx/instructions.rst
+++ b/docs/libcudacxx/ptx/instructions.rst
@@ -28,6 +28,7 @@ PTX Instructions
    instructions/getctarank
    instructions/mapa
    instructions/mbarrier_init
+   instructions/mbarrier_inval
    instructions/mbarrier_arrive
    instructions/mbarrier_expect_tx
    instructions/mbarrier_test_wait
@@ -426,7 +427,7 @@ Instructions by section
    * - :ref:`mbarrier.init <libcudacxx-ptx-instructions-mbarrier-init>`
      - CCCL 2.5.0 / CUDA Future
    * - `mbarrier.inval <https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-inval>`__
-     - No
+     - CCCL 3.2.0 / CUDA 13.2
    * - `mbarrier.complete_tx <https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-complete-tx>`__
      - No
    * - :ref:`mbarrier.arrive <libcudacxx-ptx-instructions-mbarrier-arrive>`
diff --git a/docs/libcudacxx/ptx/instructions/generated/mbarrier_inval.rst b/docs/libcudacxx/ptx/instructions/generated/mbarrier_inval.rst
new file mode 100644
index 00000000000..61fa9e3b304
--- /dev/null
+++ b/docs/libcudacxx/ptx/instructions/generated/mbarrier_inval.rst
@@ -0,0 +1,11 @@
+..
+   This file was automatically generated. Do not edit.
+
+mbarrier.inval.shared.b64
+^^^^^^^^^^^^^^^^^^^^^^^^^
+.. code-block:: cuda
+
+   // mbarrier.inval.shared.b64 [addr]; // PTX ISA 70, SM_80
+   template <typename = void>
+   __device__ static inline void mbarrier_inval(
+     uint64_t* addr);
diff --git a/docs/libcudacxx/ptx/instructions/mbarrier_inval.rst b/docs/libcudacxx/ptx/instructions/mbarrier_inval.rst
new file mode 100644
index 00000000000..79b287d083e
--- /dev/null
+++ b/docs/libcudacxx/ptx/instructions/mbarrier_inval.rst
@@ -0,0 +1,9 @@
+.. _libcudacxx-ptx-instructions-mbarrier_inval:
+
+mbarrier_inval
+==============
+
+-  PTX ISA:
+   `mbarrier_inval <https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-inval>`__
+
+.. include:: generated/mbarrier_inval.rst
diff --git a/libcudacxx/include/cuda/__barrier/barrier_block_scope.h b/libcudacxx/include/cuda/__barrier/barrier_block_scope.h
index e7be51d00d9..455b6bf8921 100644
--- a/libcudacxx/include/cuda/__barrier/barrier_block_scope.h
+++ b/libcudacxx/include/cuda/__barrier/barrier_block_scope.h
@@ -26,6 +26,7 @@
 #  include <cuda/__ptx/instructions/get_sreg.h>
 #  include <cuda/__ptx/instructions/mbarrier_arrive.h>
 #  include <cuda/__ptx/instructions/mbarrier_init.h>
+#  include <cuda/__ptx/instructions/mbarrier_inval.h>
 #  include <cuda/__ptx/instructions/mbarrier_wait.h>
 #  include <cuda/__ptx/ptx_dot_variants.h>
 #  include <cuda/__ptx/ptx_helper_functions.h>
@@ -100,10 +101,7 @@ class barrier<thread_scope_block, ::cuda::std::__empty_completion> : public __bl
   {
     NV_IF_TARGET(NV_PROVIDES_SM_80,
                  (if (::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::shared)) {
-                   // TODO(bgruber): expose mbarrier.inval.shared in cuda::ptx
-                   asm volatile("mbarrier.inval.shared.b64 [%0];" ::"r"(static_cast<::cuda::std::uint32_t>(
-                     ::__cvta_generic_to_shared(&__barrier)))
-                                : "memory");
+                   ::cuda::ptx::mbarrier_inval(__native_handle());
                    return;
                  }))
 
diff --git a/libcudacxx/include/cuda/__ptx/instructions/generated/mbarrier_inval.h b/libcudacxx/include/cuda/__ptx/instructions/generated/mbarrier_inval.h
new file mode 100644
index 00000000000..b79aa2cfdea
--- /dev/null
+++ b/libcudacxx/include/cuda/__ptx/instructions/generated/mbarrier_inval.h
@@ -0,0 +1,26 @@
+// This file was automatically generated. Do not edit.
+
+#ifndef _CUDA_PTX_GENERATED_MBARRIER_INVAL_H_
+#define _CUDA_PTX_GENERATED_MBARRIER_INVAL_H_
+
+/*
+// mbarrier.inval.shared.b64 [addr]; // PTX ISA 70, SM_80
+template <typename = void>
+__device__ static inline void mbarrier_inval(
+  uint64_t* addr);
+*/
+#if __cccl_ptx_isa >= 700
+extern "C" _CCCL_DEVICE void __cuda_ptx_mbarrier_inval_is_not_supported_before_SM_80__();
+template <typename = void>
+_CCCL_DEVICE static inline void mbarrier_inval(::cuda::std::uint64_t* __addr)
+{
+#  if _CCCL_CUDA_COMPILER(NVHPC) || __CUDA_ARCH__ >= 800
+  asm("mbarrier.inval.shared.b64 [%0];" : : "r"(__as_ptr_smem(__addr)) : "memory");
+#  else
+  // Unsupported architectures will have a linker error with a semi-decent error message
+  __cuda_ptx_mbarrier_inval_is_not_supported_before_SM_80__();
+#  endif
+}
+#endif // __cccl_ptx_isa >= 700
+
+#endif // _CUDA_PTX_GENERATED_MBARRIER_INVAL_H_
diff --git a/libcudacxx/include/cuda/__ptx/instructions/mbarrier_inval.h b/libcudacxx/include/cuda/__ptx/instructions/mbarrier_inval.h
new file mode 100644
index 00000000000..495893c8143
--- /dev/null
+++ b/libcudacxx/include/cuda/__ptx/instructions/mbarrier_inval.h
@@ -0,0 +1,41 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_PTX_MBARRIER_INVAL_H_
+#define _CUDA_PTX_MBARRIER_INVAL_H_
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/__ptx/ptx_dot_variants.h>
+#include <cuda/__ptx/ptx_helper_functions.h>
+#include <cuda/std/cstdint>
+
+#include <nv/target> // __CUDA_MINIMUM_ARCH__ and friends
+
+#include <cuda/std/__cccl/prologue.h>
+
+_CCCL_BEGIN_NAMESPACE_CUDA_PTX
+
+#include <cuda/__ptx/instructions/generated/mbarrier_inval.h>
+
+_CCCL_END_NAMESPACE_CUDA_PTX
+
+#include <cuda/std/__cccl/epilogue.h>
+
+#endif // _CUDA_PTX_MBARRIER_INVAL_H_
diff --git a/libcudacxx/include/cuda/ptx b/libcudacxx/include/cuda/ptx
index 8399880044c..4c763e75e89 100644
--- a/libcudacxx/include/cuda/ptx
+++ b/libcudacxx/include/cuda/ptx
@@ -89,6 +89,7 @@
 #include <cuda/__ptx/instructions/mbarrier_arrive.h>
 #include <cuda/__ptx/instructions/mbarrier_expect_tx.h>
 #include <cuda/__ptx/instructions/mbarrier_init.h>
+#include <cuda/__ptx/instructions/mbarrier_inval.h>
 #include <cuda/__ptx/instructions/mbarrier_wait.h>
 #include <cuda/__ptx/instructions/multimem_ld_reduce.h>
 #include <cuda/__ptx/instructions/multimem_red.h>
diff --git a/libcudacxx/test/libcudacxx/cuda/ptx/generated/mbarrier_inval.h b/libcudacxx/test/libcudacxx/cuda/ptx/generated/mbarrier_inval.h
new file mode 100644
index 00000000000..dc3b78d9d17
--- /dev/null
+++ b/libcudacxx/test/libcudacxx/cuda/ptx/generated/mbarrier_inval.h
@@ -0,0 +1,26 @@
+// This file was automatically generated. Do not edit.
+
+// We use a special strategy to force the generation of the PTX. This is mainly
+// a fight against dead-code-elimination in the NVVM layer.
+//
+// The reason we need this strategy is because certain older versions of ptxas
+// segfault when a non-sensical sequence of PTX is generated. So instead, we try
+// to force the instantiation and compilation to PTX of all the overloads of the
+// PTX wrapping functions.
+//
+// We do this by writing a function pointer of each overload to the kernel
+// parameter `fn_ptr`.
+//
+// Because `fn_ptr` is possibly visible outside this translation unit, the
+// compiler must compile all the functions which are stored.
+
+__global__ void test_mbarrier_inval(void** fn_ptr)
+{
+#if __cccl_ptx_isa >= 700
+  NV_IF_TARGET(
+    NV_PROVIDES_SM_80,
+    (
+        // mbarrier.inval.shared.b64 [addr];
+        * fn_ptr++ = reinterpret_cast<void*>(static_cast<void (*)(cuda::std::uint64_t*)>(cuda::ptx::mbarrier_inval));));
+#endif // __cccl_ptx_isa >= 700
+}
diff --git a/libcudacxx/test/libcudacxx/cuda/ptx/ptx.mbarrier_inval.compile.pass.cpp b/libcudacxx/test/libcudacxx/cuda/ptx/ptx.mbarrier_inval.compile.pass.cpp
new file mode 100644
index 00000000000..a9dd6638733
--- /dev/null
+++ b/libcudacxx/test/libcudacxx/cuda/ptx/ptx.mbarrier_inval.compile.pass.cpp
@@ -0,0 +1,22 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+// UNSUPPORTED: libcpp-has-no-threads
+
+// <cuda/ptx>
+
+#include <cuda/ptx>
+#include <cuda/std/utility>
+
+#include "generated/mbarrier_inval.h"
+
+int main(int, char**)
+{
+  return 0;
+}
