diff --git a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_torch.py b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_aiter.py
similarity index 52%
rename from examples/deepseek_mla/amd/benchmark_mla_decode_amd_torch.py
rename to examples/deepseek_mla/amd/benchmark_mla_decode_amd_aiter.py
index 18c0a5f86d..9eae480822 100644
--- a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_torch.py
+++ b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_aiter.py
@@ -4,12 +4,18 @@
 import math
 import random
 import torch
+
 import triton
 import triton.language as tl
 
 import tilelang
 from tilelang.profiler import do_bench
 
+try:
+    from aiter.mla import mla_decode_fwd
+except ImportError:
+    print("aiter is AMD specific kernel library. Please make sure aiter is installed on your AMD device.")
+
 
 def scaled_dot_product_attention(query, key, value, h_q, h_kv, is_causal=False):
     query = query.float()
@@ -58,287 +64,59 @@ def ref_mla():
     return out_torch, lse_torch, t
 
 
-@triton.jit
-def _mla_attn_kernel(
-    Q_nope,
-    Q_pe,
-    Kv_c_cache,
-    K_pe_cache,
-    Req_to_tokens,
-    B_seq_len,
-    O,
-    sm_scale,
-    stride_q_nope_bs,
-    stride_q_nope_h,
-    stride_q_pe_bs,
-    stride_q_pe_h,
-    stride_kv_c_bs,
-    stride_k_pe_bs,
-    stride_req_to_tokens_bs,
-    stride_o_b,
-    stride_o_h,
-    stride_o_s,
-    BLOCK_H: tl.constexpr,
-    BLOCK_N: tl.constexpr,
-    NUM_KV_SPLITS: tl.constexpr,
-    PAGE_SIZE: tl.constexpr,
-    HEAD_DIM_CKV: tl.constexpr,
-    HEAD_DIM_KPE: tl.constexpr,
-):
-    cur_batch = tl.program_id(1)
-    cur_head_id = tl.program_id(0)
-    split_kv_id = tl.program_id(2)
-
-    cur_batch_seq_len = tl.load(B_seq_len + cur_batch)
-
-    offs_d_ckv = tl.arange(0, HEAD_DIM_CKV)
-    cur_head = cur_head_id * BLOCK_H + tl.arange(0, BLOCK_H)
-    offs_q_nope = cur_batch * stride_q_nope_bs + cur_head[:, None] * stride_q_nope_h + offs_d_ckv[None, :]
-    q_nope = tl.load(Q_nope + offs_q_nope)
-
-    offs_d_kpe = tl.arange(0, HEAD_DIM_KPE)
-    offs_q_pe = cur_batch * stride_q_pe_bs + cur_head[:, None] * stride_q_pe_h + offs_d_kpe[None, :]
-    q_pe = tl.load(Q_pe + offs_q_pe)
-
-    e_max = tl.zeros([BLOCK_H], dtype=tl.float32) - float("inf")
-    e_sum = tl.zeros([BLOCK_H], dtype=tl.float32)
-    acc = tl.zeros([BLOCK_H, HEAD_DIM_CKV], dtype=tl.float32)
-
-    kv_len_per_split = tl.cdiv(cur_batch_seq_len, NUM_KV_SPLITS)
-    split_kv_start = kv_len_per_split * split_kv_id
-    split_kv_end = tl.minimum(split_kv_start + kv_len_per_split, cur_batch_seq_len)
-
-    for start_n in range(split_kv_start, split_kv_end, BLOCK_N):
-        offs_n = start_n + tl.arange(0, BLOCK_N)
-        kv_page_number = tl.load(
-            Req_to_tokens + stride_req_to_tokens_bs * cur_batch + offs_n // PAGE_SIZE,
-            mask=offs_n < split_kv_end,
-            other=0,
-        )
-        kv_loc = kv_page_number * PAGE_SIZE + offs_n % PAGE_SIZE
-        offs_k_c = kv_loc[None, :] * stride_kv_c_bs + offs_d_ckv[:, None]
-        k_c = tl.load(Kv_c_cache + offs_k_c, mask=offs_n[None, :] < split_kv_end, other=0.0)
-
-        qk = tl.dot(q_nope, k_c.to(q_nope.dtype))
-
-        offs_k_pe = kv_loc[None, :] * stride_k_pe_bs + offs_d_kpe[:, None]
-        k_pe = tl.load(K_pe_cache + offs_k_pe, mask=offs_n[None, :] < split_kv_end, other=0.0)
-
-        qk += tl.dot(q_pe, k_pe.to(q_pe.dtype))
-        qk *= sm_scale
-
-        qk = tl.where(offs_n[None, :] < split_kv_end, qk, float("-inf"))
-
-        v_c = tl.trans(k_c)
-
-        n_e_max = tl.maximum(tl.max(qk, 1), e_max)
-        re_scale = tl.exp(e_max - n_e_max)
-        p = tl.exp(qk - n_e_max[:, None])
-        acc *= re_scale[:, None]
-        acc += tl.dot(p.to(v_c.dtype), v_c)
-
-        e_sum = e_sum * re_scale + tl.sum(p, 1)
-        e_max = n_e_max
-    offs_o = cur_batch * stride_o_b + cur_head[:, None] * stride_o_h + split_kv_id * stride_o_s + offs_d_ckv[None, :]
-    tl.store(O + offs_o, acc / e_sum[:, None])
-    offs_o_1 = cur_batch * stride_o_b + cur_head * stride_o_h + split_kv_id * stride_o_s + HEAD_DIM_CKV
-    tl.store(O + offs_o_1, e_max + tl.log(e_sum))
-
-
-def _mla_attn(
-    q_nope,
-    q_pe,
-    kv_c_cache,
-    k_pe_cache,
-    attn_logits,
-    req_to_tokens,
-    b_seq_len,
-    num_kv_splits,
-    sm_scale,
-    page_size,
-):
-    batch_size, head_num = q_nope.shape[0], q_nope.shape[1]
-    head_dim_ckv = q_nope.shape[-1]
-    head_dim_kpe = q_pe.shape[-1]
-
-    BLOCK_H = 16
-    BLOCK_N = 64
-    grid = (
-        triton.cdiv(head_num, BLOCK_H),
-        batch_size,
-        num_kv_splits,
-    )
-    _mla_attn_kernel[grid](
-        q_nope,
-        q_pe,
-        kv_c_cache,
-        k_pe_cache,
-        req_to_tokens,
-        b_seq_len,
-        attn_logits,
-        sm_scale,
-        # stride
-        q_nope.stride(0),
-        q_nope.stride(1),
-        q_pe.stride(0),
-        q_pe.stride(1),
-        kv_c_cache.stride(-2),
-        k_pe_cache.stride(-2),
-        req_to_tokens.stride(0),
-        attn_logits.stride(0),
-        attn_logits.stride(1),
-        attn_logits.stride(2),
-        BLOCK_H=BLOCK_H,
-        BLOCK_N=BLOCK_N,
-        NUM_KV_SPLITS=num_kv_splits,
-        PAGE_SIZE=page_size,
-        HEAD_DIM_CKV=head_dim_ckv,
-        HEAD_DIM_KPE=head_dim_kpe,
-        num_stages=1,  # 2 will oom in amd
-    )
-
-
-@triton.jit
-def _mla_softmax_reducev_kernel(
-    Logits,
-    B_seq_len,
-    O,
-    stride_l_b,
-    stride_l_h,
-    stride_l_s,
-    stride_o_b,
-    stride_o_h,
-    NUM_KV_SPLITS: tl.constexpr,
-    HEAD_DIM_CKV: tl.constexpr,
-):
-    cur_batch = tl.program_id(0)
-    cur_head = tl.program_id(1)
-    cur_batch_seq_len = tl.load(B_seq_len + cur_batch)
-
-    offs_d_ckv = tl.arange(0, HEAD_DIM_CKV)
-
-    e_sum = 0.0
-    e_max = -float("inf")
-    acc = tl.zeros([HEAD_DIM_CKV], dtype=tl.float32)
-
-    offs_l = cur_batch * stride_l_b + cur_head * stride_l_h + offs_d_ckv
-    offs_l_1 = cur_batch * stride_l_b + cur_head * stride_l_h + HEAD_DIM_CKV
-
-    for split_kv_id in range(0, NUM_KV_SPLITS):
-        kv_len_per_split = tl.cdiv(cur_batch_seq_len, NUM_KV_SPLITS)
-        split_kv_start = kv_len_per_split * split_kv_id
-        split_kv_end = tl.minimum(split_kv_start + kv_len_per_split, cur_batch_seq_len)
-
-        if split_kv_end > split_kv_start:
-            logits = tl.load(Logits + offs_l + split_kv_id * stride_l_s)
-            logits_1 = tl.load(Logits + offs_l_1 + split_kv_id * stride_l_s)
-
-            n_e_max = tl.maximum(logits_1, e_max)
-            old_scale = tl.exp(e_max - n_e_max)
-            acc *= old_scale
-            exp_logic = tl.exp(logits_1 - n_e_max)
-            acc += exp_logic * logits
-
-            e_sum = e_sum * old_scale + exp_logic
-            e_max = n_e_max
-
-    tl.store(
-        O + cur_batch * stride_o_b + cur_head * stride_o_h + offs_d_ckv,
-        acc / e_sum,
-    )
-
-
-def _mla_softmax_reducev(
-    logits,
-    o,
-    b_seq_len,
-    num_kv_splits,
-):
-    batch_size, head_num, head_dim_ckv = o.shape[0], o.shape[1], o.shape[2]
-    grid = (batch_size, head_num)
-    _mla_softmax_reducev_kernel[grid](
-        logits,
-        b_seq_len,
-        o,
-        logits.stride(0),
-        logits.stride(1),
-        logits.stride(2),
-        o.stride(0),
-        o.stride(1),
-        NUM_KV_SPLITS=num_kv_splits,
-        HEAD_DIM_CKV=head_dim_ckv,
-    )
-
-
-def mla_decode_triton(
-    q_nope,
-    q_pe,
-    kv_c_cache,
-    k_pe_cache,
-    o,
-    req_to_tokens,
-    b_seq_len,
-    attn_logits,
-    num_kv_splits,
-    sm_scale,
-    page_size,
-):
-    assert num_kv_splits == attn_logits.shape[2]
-    _mla_attn(
-        q_nope,
-        q_pe,
-        kv_c_cache,
-        k_pe_cache,
-        attn_logits,
-        req_to_tokens,
-        b_seq_len,
-        num_kv_splits,
-        sm_scale,
-        page_size,
-    )
-    _mla_softmax_reducev(
-        attn_logits,
-        o,
-        b_seq_len,
-        num_kv_splits,
-    )
-
-
 @torch.inference_mode()
-def run_flash_mla_triton(q, block_table, blocked_k, max_seqlen_pad, block_size, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
-    blocked_v = blocked_k[..., :dv]
-
+def run_mla_aiter(q, block_table, blocked_k, max_seqlen_pad, block_size, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
     assert d > dv, "mla with rope dim should be larger than no rope dim"
-    q_nope, q_pe = q[..., :dv].contiguous(), q[..., dv:].contiguous()
-    blocked_k_nope, blocked_k_pe = blocked_k[..., :dv].contiguous(), blocked_k[..., dv:].contiguous()
-
-    def flash_mla_triton():
-        num_kv_splits = 32
-        o = torch.empty([b * s_q, h_q, dv])
-        attn_logits = torch.empty([b * s_q, h_q, num_kv_splits, dv + 1])
-        mla_decode_triton(
-            q_nope.view(-1, h_q, dv),
-            q_pe.view(-1, h_q, d - dv),
-            blocked_k_nope.view(-1, dv),
-            blocked_k_pe.view(-1, d - dv),
-            o,
-            block_table,
-            cache_seqlens,
-            attn_logits,
-            num_kv_splits,
-            1 / math.sqrt(d),
-            block_size,
+
+    qo_indptr = torch.zeros(b + 1, dtype=torch.int)
+    kv_indptr = torch.zeros(b + 1, dtype=torch.int)
+    seq_lens_qo = torch.empty(b, dtype=torch.int)
+    seq_lens_qo.fill_(1)
+    max_seqlen_qo = seq_lens_qo.max().item()
+
+    kv_indptr[1 : b + 1] = torch.cumsum(cache_seqlens, dim=0)
+    qo_indptr[1 : b + 1] = torch.cumsum(seq_lens_qo, dim=0)
+    total_q = qo_indptr[-1].item()
+
+    # set block_size to 1
+    page_size = 1
+    kv_buffer = blocked_k.view(-1, page_size, h_kv, d)
+
+    flat_indices = []
+    for i in range(b):
+        start = i * max_seqlen_pad
+        end = start + cache_seqlens[i]
+        flat_indices.append(torch.arange(start, end, dtype=torch.int))
+
+    kv_indices = torch.cat(flat_indices)
+
+    kv_last_page_lens = torch.ones(b, dtype=torch.int)
+
+    sm_scale = 1.0 / (d**0.5)
+
+    def mla_aiter():
+        out_aiter = torch.empty((total_q, h_q, dv), dtype=dtype).fill_(-1)
+        attn_logits_aiter, attn_lse_aiter = mla_decode_fwd(
+            q.view((total_q, h_q, d)),
+            kv_buffer,
+            out_aiter,
+            qo_indptr,
+            kv_indptr,
+            kv_indices,
+            kv_last_page_lens,
+            max_seqlen_qo,
+            sm_scale,
         )
-        return o.view([b, s_q, h_q, dv])
+        return out_aiter.view([b, s_q, h_q, dv])
 
-    out_flash = flash_mla_triton()
-    t = triton.testing.do_bench(flash_mla_triton)
-    return out_flash, None, t
+    out_aiter = mla_aiter()
+    t = triton.testing.do_bench(mla_aiter)
+    return out_aiter, None, t
 
 
 FUNC_TABLE = {
     "torch": run_torch_mla,
-    "flash_mla_triton": run_flash_mla_triton,
+    "mla_aiter": run_mla_aiter,
 }
 
 
@@ -375,14 +153,14 @@ def compare_ab(baseline, target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal
     )
 
     torch.testing.assert_close(out_b.float(), out_a.float(), atol=1e-2, rtol=1e-2), "out"
-    if target not in ["flash_mla_triton"]:
+    if target not in ["mla_aiter"]:
         # flash_mla_triton doesn't return lse
         torch.testing.assert_close(lse_b.float(), lse_a.float(), atol=1e-2, rtol=1e-2), "lse"
 
     FLOPS = s_q * total_seqlens * h_q * (d + dv) * 2
     bytes = (total_seqlens * h_kv * d + b * s_q * h_q * d + b * s_q * h_q * dv) * (torch.finfo(dtype).bits // 8)
-    print(f"perf {baseline}: {perf_a:.3f} ms, {FLOPS / 10**9 / perf_a:.0f} TFLOPS, {bytes / 10**6 / perf_a:.0f} GB/s")
-    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.0f} TFLOPS, {bytes / 10**6 / perf_b:.0f} GB/s")
+    print(f"perf {baseline}: {perf_a:.3f} ms, {FLOPS / 10**9 / perf_a:.3f} TFLOPS, {bytes / 10**6 / perf_a:.3f} GB/s")
+    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.3f} TFLOPS, {bytes / 10**6 / perf_b:.3f} GB/s")
     return bytes / 10**6 / perf_a, bytes / 10**6 / perf_b
 
 
@@ -413,13 +191,13 @@ def compare_a(target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
 
     FLOPS = s_q * total_seqlens * h_q * (d + dv) * 2
     bytes = (total_seqlens * h_kv * d + b * s_q * h_q * d + b * s_q * h_q * dv) * (torch.finfo(dtype).bits // 8)
-    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.0f} TFLOPS, {bytes / 10**6 / perf_b:.0f} GB/s")
+    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.3f} TFLOPS, {bytes / 10**6 / perf_b:.3f} GB/s")
     return bytes / 10**6 / perf_b
 
 
 available_targets = [
     "torch",
-    "flash_mla_triton",
+    "mla_aiter",
 ]
 
 shape_configs = [
@@ -432,9 +210,9 @@ def compare_a(target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
         "d": 512 + 64,
         "dv": 512,
         "causal": True,
-        "dtype": torch.float16,
+        "dtype": torch.bfloat16,
     }
-    for batch in [128]
+    for batch in [64, 128]
     for seqlen in [1024, 2048, 4096, 8192, 16384]
     for head in [128]
 ]
@@ -443,7 +221,7 @@ def compare_a(target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
 def get_args():
     parser = argparse.ArgumentParser()
     parser.add_argument("--baseline", type=str, default="torch")
-    parser.add_argument("--target", type=str, default="torch")
+    parser.add_argument("--target", type=str, default="mla_aiter")
     parser.add_argument("--all", action="store_true")
     parser.add_argument("--one", action="store_true")
     parser.add_argument("--compare", action="store_true")
diff --git a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_tilelang.py b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_tilelang.py
index dccf333ad3..399bb8e6e5 100644
--- a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_tilelang.py
+++ b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_tilelang.py
@@ -131,7 +131,7 @@ def main_split(
                 lse_local_split = glse[bz, by, k]
                 scale_local = T.exp2(lse_local_split - lse_logsum_local)
                 for i in T.Parallel(dim):
-                    o_accum_local[i] += po_local[i] * scale_local[0]
+                    o_accum_local[i] += po_local[i] * scale_local
             for i in T.Parallel(dim):
                 Output[bz, by, i] = o_accum_local[i]
 
@@ -259,6 +259,8 @@ def ref_program(q, q_pe, kv, k_pe, glse, Output_partial):
     num_split = 4
     threads = 128
 
+    print(f"Using {batch=}, {heads=}, {kv_heads=}, {kv_ctx=}, {dim=}, {pe_dim=}")
+
     if enable_autotune:
         kernel = flashmla_decode(batch, heads, kv_heads, kv_ctx, dim, pe_dim)
     else:
@@ -267,8 +269,6 @@ def ref_program(q, q_pe, kv, k_pe, glse, Output_partial):
     input_tensors = profiler._get_inputs()
     tilelang_output = kernel(*input_tensors)
     ref_output = ref_program(*input_tensors)
-    print(f"Tilelang output: {tilelang_output}")
-    print(f"Ref output: {ref_output}")
     torch.testing.assert_close(tilelang_output, ref_output, rtol=0.01, atol=0.01)
     latency = profiler.do_bench(warmup=500)
     print(f"Latency: {latency} ms")
diff --git a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_triton.py b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_triton.py
index 861e841c4e..e8c1006a01 100644
--- a/examples/deepseek_mla/amd/benchmark_mla_decode_amd_triton.py
+++ b/examples/deepseek_mla/amd/benchmark_mla_decode_amd_triton.py
@@ -378,8 +378,8 @@ def compare_ab(baseline, target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal
 
     FLOPS = s_q * total_seqlens * h_q * (d + dv) * 2
     bytes = (total_seqlens * h_kv * d + b * s_q * h_q * d + b * s_q * h_q * dv) * (torch.finfo(dtype).bits // 8)
-    print(f"perf {baseline}: {perf_a:.3f} ms, {FLOPS / 10**9 / perf_a:.0f} TFLOPS, {bytes / 10**6 / perf_a:.0f} GB/s")
-    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.0f} TFLOPS, {bytes / 10**6 / perf_b:.0f} GB/s")
+    print(f"perf {baseline}: {perf_a:.3f} ms, {FLOPS / 10**9 / perf_a:.3f} TFLOPS, {bytes / 10**6 / perf_a:.3f} GB/s")
+    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.3f} TFLOPS, {bytes / 10**6 / perf_b:.3f} GB/s")
     return bytes / 10**6 / perf_a, bytes / 10**6 / perf_b
 
 
@@ -410,7 +410,7 @@ def compare_a(target, b, s_q, cache_seqlens, h_q, h_kv, d, dv, causal, dtype):
 
     FLOPS = s_q * total_seqlens * h_q * (d + dv) * 2
     bytes = (total_seqlens * h_kv * d + b * s_q * h_q * d + b * s_q * h_q * dv) * (torch.finfo(dtype).bits // 8)
-    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.0f} TFLOPS, {bytes / 10**6 / perf_b:.0f} GB/s")
+    print(f"perf {target}: {perf_b:.3f} ms, {FLOPS / 10**9 / perf_b:.3f} TFLOPS, {bytes / 10**6 / perf_b:.3f} GB/s")
     return bytes / 10**6 / perf_b
 
 
