diff --git a/flashinfer/gemm/routergemm.py b/flashinfer/gemm/routergemm.py
index ac7da2f6d8..86f9da2b0c 100644
--- a/flashinfer/gemm/routergemm.py
+++ b/flashinfer/gemm/routergemm.py
@@ -66,7 +66,7 @@ def _mm_M1_16_K7168_shape_checks(
 
 
 # TODO: other compute capabilities may be supported but are untested
-@supported_compute_capability([100])
+@supported_compute_capability([100, 103])
 def _mm_M1_16_K7168_N256_shape_checks(mat_a, mat_b, out, launch_with_pdl):
     return _mm_M1_16_K7168_shape_checks(
         mat_a,
@@ -79,7 +79,7 @@ def _mm_M1_16_K7168_N256_shape_checks(mat_a, mat_b, out, launch_with_pdl):
 
 
 # TODO: other compute capabilities may be supported but are untested
-@supported_compute_capability([100])
+@supported_compute_capability([100, 103])
 def _mm_M1_16_K7168_N128_shape_checks(mat_a, mat_b, out, launch_with_pdl):
     return _mm_M1_16_K7168_shape_checks(
         mat_a,
diff --git a/tests/model_optimizations/test_dsv3_router_gemm.py b/tests/model_optimizations/test_dsv3_router_gemm.py
index 80f921667f..fdad108cf5 100644
--- a/tests/model_optimizations/test_dsv3_router_gemm.py
+++ b/tests/model_optimizations/test_dsv3_router_gemm.py
@@ -21,8 +21,8 @@ def test_dsv3_router_gemm_op(
 ):
     compute_capability = get_compute_capability(torch.device("cuda"))
     compute_capability_number = compute_capability[0] * 10 + compute_capability[1]
-    if compute_capability_number != 100:
-        pytest.skip("DSv3 Router GEMM is only supported on SM100")
+    if compute_capability_number != 100 and compute_capability_number != 103:
+        pytest.skip("DSv3 Router GEMM is only supported on SM100 and SM103")
 
     mat_a = torch.randn(num_tokens, hidden_dim, device="cuda", dtype=torch.bfloat16)
     mat_b = torch.randn(
@@ -240,8 +240,8 @@ def test_dsv3_router_gemm_op_negative(
 ):
     compute_capability = get_compute_capability(torch.device("cuda"))
     compute_capability_number = compute_capability[0] * 10 + compute_capability[1]
-    if compute_capability_number != 100:
-        pytest.skip("DSv3 Router GEMM is only supported on SM100")
+    if compute_capability_number != 100 and compute_capability_number != 103:
+        pytest.skip("DSv3 Router GEMM is only supported on SM100 and SM103")
 
     mat_a = torch.randn(num_tokens, hidden_dim, device="cuda", dtype=mat_a_dtype)
     mat_b = torch.randn(num_experts, hidden_dim, device="cuda", dtype=mat_b_dtype)
