diff --git a/csrc/trtllm_allreduce_fusion.cu b/csrc/trtllm_allreduce_fusion.cu
index c38203514e..d54441ec34 100644
--- a/csrc/trtllm_allreduce_fusion.cu
+++ b/csrc/trtllm_allreduce_fusion.cu
@@ -34,7 +34,7 @@ void trtllm_allreduce_fusion(
     std::optional<at::Tensor> residual_out, std::optional<at::Tensor> norm_out,
     std::optional<at::Tensor> quant_out, std::optional<at::Tensor> scale_out,
     std::optional<at::Tensor> rms_gamma, std::optional<double> rms_eps,
-    std::optional<double> scale_factor, std::optional<int64_t> layout_code) {
+    std::optional<at::Tensor> scale_factor, std::optional<int64_t> layout_code) {
   const c10::cuda::OptionalCUDAGuard device_guard(allreduce_in.device());
   // todo(Yingyi): add dispatch for float and bfloat16
 
@@ -66,8 +66,9 @@ void trtllm_allreduce_fusion(
     params.rms_gamma =
         rms_gamma.has_value() ? reinterpret_cast<void*>(rms_gamma.value().data_ptr()) : nullptr;
     params.rms_eps = rms_eps.has_value() ? static_cast<float>(rms_eps.value()) : 0.0f;
-    params.scale_factor =
-        scale_factor.has_value() ? static_cast<float>(scale_factor.value()) : 0.0f;
+    params.scale_factor = scale_factor.has_value()
+                              ? reinterpret_cast<float*>(scale_factor.value().data_ptr())
+                              : nullptr;
     params.use_oneshot = use_oneshot;
     params.layout = layout_code.has_value()
                         ? static_cast<FP4QuantizationSFLayout>(layout_code.value())
diff --git a/flashinfer/comm/trtllm_ar.py b/flashinfer/comm/trtllm_ar.py
index d02c0e2f05..34c73c8cff 100644
--- a/flashinfer/comm/trtllm_ar.py
+++ b/flashinfer/comm/trtllm_ar.py
@@ -18,7 +18,7 @@
 import logging
 from dataclasses import dataclass
 from types import SimpleNamespace
-from typing import Any, Dict, List, Optional, Tuple
+from typing import Any, Dict, List, Optional, Tuple, Union
 
 import torch
 import torch.distributed as dist
@@ -257,7 +257,7 @@ def trtllm_allreduce_fusion(
         scale_out: Optional[torch.Tensor],
         rms_gamma: Optional[torch.Tensor],
         rms_eps: Optional[float],
-        scale_factor: Optional[float],
+        scale_factor: Optional[Union[torch.Tensor, float]],
         layout_code: Optional[FP4QuantizationSFLayout],
     ) -> None:
         module.trtllm_allreduce_fusion(
@@ -783,7 +783,7 @@ def trtllm_allreduce_fusion(
     scale_out: Optional[torch.Tensor],
     rms_gamma: Optional[torch.Tensor],
     rms_eps: Optional[float],
-    scale_factor: Optional[float],
+    scale_factor: Optional[Union[torch.Tensor, float]],
     layout_code: Optional[FP4QuantizationSFLayout],
 ) -> None:
     """
@@ -807,7 +807,7 @@ def trtllm_allreduce_fusion(
     - scale_out: the scale output tensor. Initialization referece: tests/test_trtllm_allreduce_fusion.py
     - rms_gamma: the rms gamma tensor. [hidden_dim]
     - rms_eps: the rms epsilon value.
-    - scale_factor: the scale factor.
+    - scale_factor: the scale factor. For cudaGraphs safety, it should be a tensor.
     - layout_code: the layout code.
 
     Note:
@@ -833,7 +833,13 @@ def trtllm_allreduce_fusion(
             f"required_lamport_comm_size {required_lamport_comm_size} is greater than MAX_COMM_SIZE {MAX_COMM_SIZE}. Cannot use oneshot in this case."
         )
         use_oneshot = False
-
+    if scale_factor is not None:
+        if isinstance(scale_factor, torch.Tensor):
+            scale_factor = scale_factor.to(torch.float32)
+        else:
+            scale_factor = torch.tensor(
+                [scale_factor], dtype=torch.float32, device=allreduce_in.device
+            )
     get_trtllm_comm_module().trtllm_allreduce_fusion(
         allreduce_in=allreduce_in,
         world_size=world_size,
diff --git a/include/flashinfer/comm/trtllm_allreduce_fusion.cuh b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
index e6f78faedd..633ae68cfb 100644
--- a/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
+++ b/include/flashinfer/comm/trtllm_allreduce_fusion.cuh
@@ -756,7 +756,7 @@ struct AllReduceFusionParams {
   void* scale_out;
   void* rms_gamma;
   float rms_eps;
-  float scale_factor;
+  float* scale_factor;
   bool use_oneshot;
   FP4QuantizationSFLayout layout = FP4QuantizationSFLayout::SWIZZLED;
   cudaStream_t stream;
@@ -905,9 +905,9 @@ class FusedOp {
       m_residual_val.load(reinterpret_cast<T*>(params.residual_in) + m_access_id * VEC_SIZE);
     }
     if constexpr (GetQuantType<Pattern> == QuantType::kFP8) {
-      m_scale_factor = 1.f / params.scale_factor;
+      m_scale_factor = 1.f / *(params.scale_factor);
     } else if constexpr (GetQuantType<Pattern> == QuantType::kFP4) {
-      m_scale_factor = params.scale_factor;
+      m_scale_factor = *(params.scale_factor);
     }
   }
 
@@ -945,7 +945,7 @@ class FusedOp {
           std::nullopt /* batchIdx */, token_id, m_access_id_in_token, std::nullopt /* numRows */,
           m_params.hidden_dim, reinterpret_cast<uint32_t*>(m_params.scale_out), m_params.layout);
       reinterpret_cast<uint32_t*>(m_params.quant_out)[m_access_id] =
-          utils::cvt_warp_fp16_to_fp4<T, VEC_SIZE>(val, m_params.scale_factor, sf_out);
+          utils::cvt_warp_fp16_to_fp4<T, VEC_SIZE>(val, m_scale_factor, sf_out);
     } else if constexpr (GetQuantType<Pattern> == QuantType::kFP8) {
       using PackedQuantizedType = std::conditional_t<std::is_same_v<T, float>, float, float2>;
       PackedQuantizedType ret;
diff --git a/tests/test_trtllm_allreduce_fusion.py b/tests/test_trtllm_allreduce_fusion.py
index 05b819a04b..217a9b6173 100644
--- a/tests/test_trtllm_allreduce_fusion.py
+++ b/tests/test_trtllm_allreduce_fusion.py
@@ -55,7 +55,7 @@ def _run_correctness_worker(world_size, rank, dtype, hidden_dim, distributed_ini
 
         lamport_use_fp32 = dtype == torch.float32
 
-        # create workspace for moe allreduce fusion
+        # create workspace for allreduce fusion
         ipc_handles, workspace_tensor = (
             comm.trtllm_create_ipc_workspace_for_all_reduce_fusion(
                 rank,
@@ -336,4 +336,4 @@ def test_trtllm_allreduce_fusion(world_size, dtype, hidden_dim):
         _run_correctness_worker,
         target_args=(),
     )
-    print(f"moe allreduce fusion tp = {world_size}: OK")
+    print(f"allreduce fusion tp = {world_size}: OK")
