diff --git a/flashinfer/comm/__init__.py b/flashinfer/comm/__init__.py
index 1fffcb6581..a5ab1babae 100644
--- a/flashinfer/comm/__init__.py
+++ b/flashinfer/comm/__init__.py
@@ -1,4 +1,4 @@
-from .cuda_ipc import create_shared_buffer, free_shared_buffer
+from .cuda_ipc import CudaRTLibrary, create_shared_buffer, free_shared_buffer
 from .dlpack_utils import pack_strided_memory
 from .mapping import Mapping
 from .trtllm_ar import AllReduceFusionOp as AllReduceFusionOp
