diff --git a/cub/benchmarks/bench/transform/common.h b/cub/benchmarks/bench/transform/common.h
index 2ac823232e7..21f4f6bed79 100644
--- a/cub/benchmarks/bench/transform/common.h
+++ b/cub/benchmarks/bench/transform/common.h
@@ -38,7 +38,7 @@ struct policy_hub_t
     using algo_policy =
       ::cuda::std::_If<algorithm == cub::detail::transform::Algorithm::prefetch,
                        cub::detail::transform::prefetch_policy_t<TUNE_THREADS>,
-                       cub::detail::transform::async_copy_policy_t<TUNE_THREADS>>;
+                       cub::detail::transform::async_copy_policy_t<TUNE_THREADS, __CUDA_ARCH_LIST__ == 900 ? 128 : 16>>;
   };
 };
 #endif
diff --git a/cub/cub/device/dispatch/dispatch_transform.cuh b/cub/cub/device/dispatch/dispatch_transform.cuh
index 8ede28bad45..5148b2e059e 100644
--- a/cub/cub/device/dispatch/dispatch_transform.cuh
+++ b/cub/cub/device/dispatch/dispatch_transform.cuh
@@ -169,6 +169,7 @@ struct dispatch_t<StableAddress,
   RandomAccessIteratorOut out;
   Offset num_items;
   TransformOp op;
+  int bulk_copy_align;
   cudaStream_t stream;
   KernelSource kernel_source             = {};
   KernelLauncherFactory launcher_factory = {};
@@ -183,9 +184,9 @@ struct dispatch_t<StableAddress,
   {
     using policy_t          = typename ActivePolicy::algo_policy;
     constexpr int block_dim = policy_t::block_threads;
-    static_assert(block_dim % bulk_copy_alignment == 0,
-                  "block_threads needs to be a multiple of bulk_copy_alignment (128)"); // then tile_size is a multiple
-                                                                                        // of 128-byte
+    _CCCL_ASSERT_HOST(block_dim % bulk_copy_align == 0,
+                      "block_threads needs to be a multiple of bulk_copy_alignment"); // then tile_size is a multiple of
+                                                                                      // it
 
     auto determine_element_counts = [&]() -> cuda_expected<elem_counts> {
       const auto max_smem = get_max_shared_memory();
@@ -202,7 +203,7 @@ struct dispatch_t<StableAddress,
            ++elem_per_thread)
       {
         const int tile_size = block_dim * elem_per_thread;
-        const int smem_size = bulk_copy_smem_for_tile_size<RandomAccessIteratorsIn...>(tile_size);
+        const int smem_size = bulk_copy_smem_for_tile_size<RandomAccessIteratorsIn...>(tile_size, bulk_copy_align);
         if (smem_size > *max_smem)
         {
           // assert should be prevented by smem check in policy
@@ -246,7 +247,7 @@ struct dispatch_t<StableAddress,
     }
     _CCCL_ASSERT_HOST(config->elem_per_thread > 0, "");
     _CCCL_ASSERT_HOST(config->tile_size > 0, "");
-    _CCCL_ASSERT_HOST(config->tile_size % bulk_copy_alignment == 0, "");
+    _CCCL_ASSERT_HOST(config->tile_size % bulk_copy_align == 0, "");
     _CCCL_ASSERT_HOST((sizeof...(RandomAccessIteratorsIn) == 0) != (config->smem_size != 0), ""); // logical xor
 
     const auto grid_dim = static_cast<unsigned int>(::cuda::ceil_div(num_items, Offset{config->tile_size}));
@@ -272,7 +273,7 @@ struct dispatch_t<StableAddress,
       op,
       out,
       make_aligned_base_ptr_kernel_arg(
-        THRUST_NS_QUALIFIER::try_unwrap_contiguous_iterator(::cuda::std::get<Is>(in)), bulk_copy_alignment)...);
+        THRUST_NS_QUALIFIER::try_unwrap_contiguous_iterator(::cuda::std::get<Is>(in)), bulk_copy_align)...);
   }
 #endif // _CUB_HAS_TRANSFORM_UBLKCP
 
@@ -385,6 +386,7 @@ struct dispatch_t<StableAddress,
       ::cuda::std::move(out),
       num_items,
       ::cuda::std::move(op),
+      bulk_copy_alignment(ptx_version),
       stream,
       kernel_source,
       launcher_factory};
diff --git a/cub/cub/device/dispatch/kernels/transform.cuh b/cub/cub/device/dispatch/kernels/transform.cuh
index 46c536bc87e..c9abda67533 100644
--- a/cub/cub/device/dispatch/kernels/transform.cuh
+++ b/cub/cub/device/dispatch/kernels/transform.cuh
@@ -202,31 +202,12 @@ _CCCL_DEVICE _CCCL_FORCEINLINE static bool elect_one()
   return threadIdx.x < 32 && static_cast<bool>(is_elected);
 }
 
-template <typename Offset, typename T>
-_CCCL_DEVICE void bulk_copy_tile_fallback(
-  int tile_size,
-  int tile_stride,
-  char* smem,
-  int& smem_offset,
-  Offset global_offset,
-  const aligned_base_ptr<T>& aligned_ptr)
-{
-  const T* src = aligned_ptr.ptr_to_elements() + global_offset;
-  T* dst       = reinterpret_cast<T*>(smem + smem_offset + aligned_ptr.head_padding);
-  _CCCL_ASSERT(reinterpret_cast<uintptr_t>(src) % alignof(T) == 0, "");
-  _CCCL_ASSERT(reinterpret_cast<uintptr_t>(dst) % alignof(T) == 0, "");
-
-  const int bytes_to_copy = static_cast<int>(sizeof(T)) * tile_size;
-  cooperative_groups::memcpy_async(cooperative_groups::this_thread_block(), dst, src, bytes_to_copy);
-
-  // add bulk_copy_alignment to make space for the next tile's head padding
-  smem_offset += static_cast<int>(sizeof(T)) * tile_stride + bulk_copy_alignment;
-}
-
 template <typename BulkCopyPolicy, typename Offset, typename F, typename RandomAccessIteratorOut, typename... InTs>
 _CCCL_DEVICE void transform_kernel_ublkcp(
   Offset num_items, int num_elem_per_thread, F f, RandomAccessIteratorOut out, aligned_base_ptr<InTs>... aligned_ptrs)
 {
+  constexpr int bulk_copy_alignment = BulkCopyPolicy::bulk_copy_alignment;
+
   __shared__ uint64_t bar;
   extern __shared__ char __align__(bulk_copy_alignment) smem[];
 
diff --git a/cub/cub/device/dispatch/tuning/tuning_transform.cuh b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
index 67129cf38fa..8bb56530329 100644
--- a/cub/cub/device/dispatch/tuning/tuning_transform.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
@@ -80,13 +80,15 @@ struct prefetch_policy_t
   static constexpr int max_items_per_thread      = 32;
 };
 
-template <int BlockThreads>
+template <int BlockThreads, int BulkCopyAlignment>
 struct async_copy_policy_t
 {
   static constexpr int block_threads = BlockThreads;
   // items per tile are determined at runtime. these (inclusive) bounds allow overriding that value via a tuning policy
   static constexpr int min_items_per_thread = 1;
   static constexpr int max_items_per_thread = 32;
+
+  static constexpr int bulk_copy_alignment = BulkCopyAlignment;
 };
 
 // mult must be a power of 2
@@ -115,16 +117,20 @@ _CCCL_HOST_DEVICE constexpr auto loaded_bytes_per_iteration() -> int
   return (int{sizeof(it_value_t<Its>)} + ... + 0);
 }
 
-constexpr int bulk_copy_alignment     = 128;
 constexpr int bulk_copy_size_multiple = 16;
 
+_CCCL_HOST_DEVICE constexpr auto bulk_copy_alignment(int ptx_version) -> int
+{
+  return ptx_version < 1000 ? 128 : 16;
+}
+
 template <typename... RandomAccessIteratorsIn>
-_CCCL_HOST_DEVICE constexpr auto bulk_copy_smem_for_tile_size(int tile_size) -> int
+_CCCL_HOST_DEVICE constexpr auto bulk_copy_smem_for_tile_size(int tile_size, int bulk_copy_align) -> int
 {
-  return round_up_to_po2_multiple(int{sizeof(int64_t)}, bulk_copy_alignment) /* bar */
+  return round_up_to_po2_multiple(int{sizeof(int64_t)}, bulk_copy_align) /* bar */
        // 128 bytes of padding for each input tile (handles before + after)
        + tile_size * loaded_bytes_per_iteration<RandomAccessIteratorsIn...>()
-       + sizeof...(RandomAccessIteratorsIn) * bulk_copy_alignment;
+       + sizeof...(RandomAccessIteratorsIn) * bulk_copy_align;
 }
 
 _CCCL_HOST_DEVICE constexpr int arch_to_min_bytes_in_flight(int sm_arch)
@@ -213,17 +219,21 @@ struct policy_hub<RequiresStableAddress, ::cuda::std::tuple<RandomAccessIterator
   template <int BlockSize, int PtxVersion>
   struct bulkcopy_policy
   {
-    static constexpr int min_bif = arch_to_min_bytes_in_flight(PtxVersion);
-    using async_policy           = async_copy_policy_t<BlockSize>;
+  private:
+    static constexpr int bulk_copy_align = bulk_copy_alignment(PtxVersion);
+    using async_policy                   = async_copy_policy_t<BlockSize, bulk_copy_align>;
     static constexpr bool exhaust_smem =
       bulk_copy_smem_for_tile_size<RandomAccessIteratorsIn...>(
-        async_policy::block_threads * async_policy::min_items_per_thread)
+        async_policy::block_threads * async_policy::min_items_per_thread, bulk_copy_align)
       > int{max_smem_per_block};
     static constexpr bool any_type_is_overalinged =
-      ((alignof(it_value_t<RandomAccessIteratorsIn>) > bulk_copy_alignment) || ...);
+      ((alignof(it_value_t<RandomAccessIteratorsIn>) > bulk_copy_align) || ...);
 
     static constexpr bool use_fallback =
       RequiresStableAddress || !can_memcpy || no_input_streams || exhaust_smem || any_type_is_overalinged;
+
+  public:
+    static constexpr int min_bif    = arch_to_min_bytes_in_flight(PtxVersion);
     static constexpr auto algorithm = use_fallback ? Algorithm::prefetch : Algorithm::ublkcp;
     using algo_policy               = ::cuda::std::_If<use_fallback, prefetch_policy_t<BlockSize>, async_policy>;
   };
diff --git a/cub/test/catch2_test_device_transform.cu b/cub/test/catch2_test_device_transform.cu
index ea60198c7d3..1e9d2da7238 100644
--- a/cub/test/catch2_test_device_transform.cu
+++ b/cub/test/catch2_test_device_transform.cu
@@ -38,7 +38,7 @@ struct policy_hub_for_alg
     using algo_policy =
       ::cuda::std::_If<Alg == Algorithm::prefetch,
                        cub::detail::transform::prefetch_policy_t<256>,
-                       cub::detail::transform::async_copy_policy_t<256>>;
+                       cub::detail::transform::async_copy_policy_t<256, 128>>;
   };
 };
 
diff --git a/cudax/test/containers/async_buffer/transform.cu b/cudax/test/containers/async_buffer/transform.cu
index 244438d5c56..7fafc0f5598 100644
--- a/cudax/test/containers/async_buffer/transform.cu
+++ b/cudax/test/containers/async_buffer/transform.cu
@@ -39,7 +39,7 @@ struct policy_hub_for_alg
     using algo_policy =
       ::cuda::std::_If<Alg == Algorithm::prefetch,
                        cub::detail::transform::prefetch_policy_t<256>,
-                       cub::detail::transform::async_copy_policy_t<256>>;
+                       cub::detail::transform::async_copy_policy_t<256, 128>>;
   };
 };
 
