diff --git a/flashinfer/jit/cpp_ext.py b/flashinfer/jit/cpp_ext.py
index fb0c40c00e..2c3a56d92b 100644
--- a/flashinfer/jit/cpp_ext.py
+++ b/flashinfer/jit/cpp_ext.py
@@ -47,11 +47,16 @@ def get_cuda_path() -> str:
         return cuda_home
     # get output of "which nvcc"
     nvcc_path = subprocess.run(["which", "nvcc"], capture_output=True)
-    if nvcc_path.returncode != 0:
-        raise RuntimeError("Could not find nvcc")
-    cuda_home = os.path.dirname(
-        os.path.dirname(nvcc_path.stdout.decode("utf-8").strip())
-    )
+    if nvcc_path.returncode == 0:
+        cuda_home = os.path.dirname(
+            os.path.dirname(nvcc_path.stdout.decode("utf-8").strip())
+        )
+    else:
+        cuda_home = "/usr/local/cuda"  # This default value is from: https://github.com/pytorch/pytorch/blob/ceb11a584d6b3fdc600358577d9bf2644f88def9/torch/utils/cpp_extension.py#L115
+        if not os.path.exists(cuda_home):
+            raise RuntimeError(
+                f"Could not find nvcc and default {cuda_home=} doesn't exist"
+            )
     return cuda_home
 
 
