diff --git a/cub/cub/device/dispatch/dispatch_transform.cuh b/cub/cub/device/dispatch/dispatch_transform.cuh
index b35d0499ddc..2b8a2227a71 100644
--- a/cub/cub/device/dispatch/dispatch_transform.cuh
+++ b/cub/cub/device/dispatch/dispatch_transform.cuh
@@ -205,7 +205,7 @@ struct dispatch_t<StableAddress,
 
   template <typename ActivePolicy, typename SMemFunc>
   CUB_RUNTIME_FUNCTION _CCCL_VISIBILITY_HIDDEN _CCCL_FORCEINLINE auto
-  configure_async_kernel(int alignment, SMemFunc smem_for_tile_size, ActivePolicy policy = {}) -> cuda_expected<
+  configure_async_kernel(int alignment, SMemFunc dyn_smem_for_tile_size, ActivePolicy policy = {}) -> cuda_expected<
     ::cuda::std::tuple<decltype(launcher_factory(0, 0, 0, 0)), decltype(kernel_source.TransformKernel()), int>>
   {
     auto wrapped_policy = MakeTransformPolicyWrapper(policy);
@@ -222,15 +222,8 @@ struct dispatch_t<StableAddress,
     CUB_DETAIL_STATIC_ISH_ASSERT(min_items_per_thread <= max_items_per_thread, "invalid policy");
 
     auto determine_element_counts = [&]() -> cuda_expected<async_config> {
-      int max_smem = 0;
-      auto error   = CubDebug(launcher_factory.MaxSharedMemory(max_smem));
-      if (error != cudaSuccess)
-      {
-        return ::cuda::std::unexpected<cudaError_t /* nvcc 12.0 fails CTAD here */>(error);
-      }
-
       int sm_count = 0;
-      error        = CubDebug(launcher_factory.MultiProcessorCount(sm_count));
+      auto error   = CubDebug(launcher_factory.MultiProcessorCount(sm_count));
       if (error != cudaSuccess)
       {
         return ::cuda::std::unexpected<cudaError_t /* nvcc 12.0 fails CTAD here */>(error);
@@ -243,23 +236,22 @@ struct dispatch_t<StableAddress,
       async_config last_config{};
       for (int items_per_thread = +min_items_per_thread; items_per_thread <= +max_items_per_thread; ++items_per_thread)
       {
-        const int tile_size = block_threads * items_per_thread;
-        const int smem_size = smem_for_tile_size(tile_size, alignment);
-        if (smem_size > max_smem)
+        const int tile_size     = block_threads * items_per_thread;
+        const int dyn_smem_size = dyn_smem_for_tile_size(tile_size, alignment);
+        int max_occupancy       = 0;
+        error                   = CubDebug(launcher_factory.MaxSmOccupancy(
+          max_occupancy, kernel_source.TransformKernel(), block_threads, dyn_smem_size));
+        if (error != cudaSuccess)
+        {
+          return ::cuda::std::unexpected<cudaError_t /* nvcc 12.0 fails CTAD here */>(error);
+        }
+        if (max_occupancy == 0)
         {
           // assert should be prevented by smem check in policy
           _CCCL_ASSERT(last_config.items_per_thread > 0, "min_items_per_thread exceeds available shared memory");
           return last_config;
         }
 
-        int max_occupancy = 0;
-        error             = CubDebug(
-          launcher_factory.MaxSmOccupancy(max_occupancy, kernel_source.TransformKernel(), block_threads, smem_size));
-        if (error != cudaSuccess)
-        {
-          return ::cuda::std::unexpected<cudaError_t /* nvcc 12.0 fails CTAD here */>(error);
-        }
-
         const auto config = async_config{items_per_thread, max_occupancy, sm_count};
 
         const int bytes_in_flight_SM = max_occupancy * tile_size * kernel_source.LoadedBytesPerIteration();
@@ -297,15 +289,15 @@ struct dispatch_t<StableAddress,
 
     const int ipt =
       spread_out_items_per_thread(policy, config->items_per_thread, config->sm_count, config->max_occupancy);
-    const int tile_size = block_threads * ipt;
-    const int smem_size = smem_for_tile_size(tile_size, alignment);
-    _CCCL_ASSERT((sizeof...(RandomAccessIteratorsIn) == 0) != (smem_size != 0), ""); // logical xor
+    const int tile_size     = block_threads * ipt;
+    const int dyn_smem_size = dyn_smem_for_tile_size(tile_size, alignment);
+    _CCCL_ASSERT((sizeof...(RandomAccessIteratorsIn) == 0) != (dyn_smem_size != 0), ""); // logical xor
 
     const auto grid_dim = static_cast<unsigned int>(::cuda::ceil_div(num_items, Offset{tile_size}));
     // config->smem_size is 16 bytes larger than needed for UBLKCP because it's the total SMEM size, but 16 bytes are
     // occupied by static shared memory and padding. But let's not complicate things.
     return ::cuda::std::make_tuple(
-      launcher_factory(grid_dim, block_threads, smem_size, stream), kernel_source.TransformKernel(), ipt);
+      launcher_factory(grid_dim, block_threads, dyn_smem_size, stream), kernel_source.TransformKernel(), ipt);
   }
 
   // Avoid unnecessarily parsing these definitions when not needed.
@@ -327,9 +319,9 @@ struct dispatch_t<StableAddress,
 
   template <typename ActivePolicy, typename SMemFunc, std::size_t... Is>
   CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE cudaError_t invoke_async_algorithm(
-    int alignment, SMemFunc smem_for_tile_size, cuda::std::index_sequence<Is...>, ActivePolicy policy = {})
+    int alignment, SMemFunc dyn_smem_for_tile_size, cuda::std::index_sequence<Is...>, ActivePolicy policy = {})
   {
-    auto ret = configure_async_kernel(alignment, smem_for_tile_size, policy);
+    auto ret = configure_async_kernel(alignment, dyn_smem_for_tile_size, policy);
     if (!ret)
     {
       return ret.error();
@@ -491,12 +483,8 @@ struct dispatch_t<StableAddress,
     {
       return invoke_async_algorithm(
         bulk_copy_align,
-        [this, wrapped_policy](int tile_size, int alignment) {
-          return bulk_copy_smem_for_tile_size(
-            kernel_source.ItValueSizesAlignments(),
-            tile_size,
-            wrapped_policy.AlgorithmPolicy().BlockThreads(),
-            alignment);
+        [this](int tile_size, int alignment) {
+          return bulk_copy_dyn_smem_for_tile_size(kernel_source.ItValueSizesAlignments(), tile_size, alignment);
         },
         seq,
         active_policy);
@@ -505,12 +493,8 @@ struct dispatch_t<StableAddress,
     {
       return invoke_async_algorithm(
         ldgsts_size_and_align,
-        [this, wrapped_policy](int tile_size, int alignment) {
-          return memcpy_async_smem_for_tile_size(
-            kernel_source.ItValueSizesAlignments(),
-            tile_size,
-            wrapped_policy.AlgorithmPolicy().BlockThreads(),
-            alignment);
+        [this](int tile_size, int alignment) {
+          return memcpy_async_dyn_smem_for_tile_size(kernel_source.ItValueSizesAlignments(), tile_size, alignment);
         },
         seq,
         active_policy);
diff --git a/cub/cub/device/dispatch/kernels/transform.cuh b/cub/cub/device/dispatch/kernels/transform.cuh
index b3d8eb739ec..fbbef54a2b7 100644
--- a/cub/cub/device/dispatch/kernels/transform.cuh
+++ b/cub/cub/device/dispatch/kernels/transform.cuh
@@ -23,6 +23,7 @@
 
 #include <cuda/__memory/aligned_size.h>
 #include <cuda/cmath>
+#include <cuda/memory>
 #include <cuda/ptx>
 #include <cuda/std/bit>
 #include <cuda/std/cstdint>
@@ -279,7 +280,7 @@ _CCCL_DEVICE void transform_kernel_vectorized(
 // copy a tile of data from an input buffer, we round down the pointer to the start of the tile to the next lower
 // address that is a multiple of 16 bytes. This introduces head padding. We also round up the total number of bytes to
 // copy (including head padding) to a multiple of 16 bytes, which introduces tail padding. For the bulk copy kernel, we
-// have to align to 128 bytes instead of 16 on Hopper.
+// should align to 128 bytes instead of 16 on Hopper.
 //
 // However, padding memory copies like that may access the input buffer out-of-bounds. Here are some thoughts:
 // * According to the CUDA programming guide
@@ -613,14 +614,6 @@ _CCCL_DEVICE void bulk_copy_maybe_unaligned(
   }
 }
 
-template <int Alignment>
-_CCCL_DEVICE auto round_up_smem_ptr(char* p) -> char*
-{
-  uint32_t smem32 = __cvta_generic_to_shared(p);
-  smem32          = ::cuda::round_up(smem32, Alignment);
-  return static_cast<char*>(_CCCL_BUILTIN_ASSUME_ALIGNED(__cvta_shared_to_generic(smem32), Alignment));
-}
-
 template <typename BulkCopyPolicy, typename Offset, typename F, typename RandomAccessIteratorOut, typename... InTs>
 _CCCL_DEVICE void transform_kernel_ublkcp(
   Offset num_items, int num_elem_per_thread, F f, RandomAccessIteratorOut out, aligned_base_ptr<InTs>... aligned_ptrs)
@@ -628,57 +621,49 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
   constexpr int block_threads       = BulkCopyPolicy::block_threads;
   constexpr int bulk_copy_alignment = BulkCopyPolicy::bulk_copy_alignment;
 
-  constexpr int min_retained_alignment           = ::cuda::std::min({sizeof(InTs)...}) * block_threads;
-  constexpr int max_alignment                    = ::cuda::std::max({alignof(InTs)...});
-  constexpr bool tile_sizes_retain_max_alignment = max_alignment <= min_retained_alignment;
-
   // add padding after a tile in shared memory to make space for the next tile's head padding, and retain alignment
-  constexpr int tile_padding =
-    tile_sizes_retain_max_alignment ? ::cuda::std::max(bulk_copy_alignment, max_alignment) : bulk_copy_alignment;
-
-  __shared__ uint64_t bar;
-
-  // We use an attribute to align the shared memory. This is not respected on all drivers though. The compiler correctly
-  // takes the alignment into account and even emits an alignment specifier into ptx. However, this sometimes randomly
-  // fails at runtime because the shared memory start pointer is not correctly provided by the driver/runtime. See also
-  // NVBug 5093902, NVBug 5329745, and discussion in PR #5122.
-  extern __shared__ char __align__(tile_padding) smem_base_unaligned[];
-
-  // However, any manual alignment of the shared memory start address outweighs the performance benefits of a faster
-  // bulk copy by introducing about 7 additional SASS instructions at the start of the kernel. This also has to be done
-  // carefully using a fake read on the address to prevent NVVM to pull the aligning deeper into the kernel.
-  //   extern __shared__ char smem_base[];
-  //   uint32_t smem32 = __cvta_generic_to_shared(smem_base);
-  //   smem32 = cuda::round_up(smem32, bulk_copy_alignment);
-  //   char* smem = static_cast<char*>(_CCCL_BUILTIN_ASSUME_ALIGNED(__cvta_shared_to_generic(smem32),
-  //                                                                bulk_copy_alignment));
-
-  // What gets closest to a working attribute is to rely on the following observations:
-  // * static shared memory is aligned to 1KiB
-  // * dynamic shared memory is aligned to 16 bytes and comes right after static shared memory
-  // In this case we could:
-  //   extern __shared__ char smem_base[];
-  //   uint32_t smem32 = __cvta_generic_to_shared(smem_base) + bulk_copy_alignment - 16;
-  //   asm("" : "+r"(smem32));
-  //   char* smem = static_cast<char*>(_CCCL_BUILTIN_ASSUME_ALIGNED(__cvta_shared_to_generic(smem32),
-  //                                                                bulk_copy_alignment));
-  // However, CUDA currently does not provide this guarantee.
-
-  // We cannot assert that shared memory is sufficiently aligned, since it fails on some systems (e.g. with driver
-  // 565.57.01 on RTX 2080 when cub::DeviceTransform is called from another kernel via CDP. See
-  // thrust.cpp.cuda.cpp20.test.cuda.transform.cdp_1). This will lead to slightly reduced performance of bulk copy, but
-  // correctness is maintained.
-  //_CCCL_ASSERT(::cuda::is_aligned(smem, bulk_copy_alignment), "");
-
-  char* smem_base = smem_base_unaligned;
-  // Since alignment via the attribute may not work, we have to align explicitly if it's larger than the default dynamic
-  // shared memory alignment (16). This is not needed when the tile size does not retain the alignment, since we align
-  // each tile separately later
-  if constexpr (tile_sizes_retain_max_alignment && max_alignment > 16)
-  {
-    smem_base = round_up_smem_ptr<tile_padding>(smem_base);
-    asm("" : "+l"(smem_base)); // keep the compiler from pulling the alignment deeper into the kernel
-  }
+  constexpr int max_alignment = ::cuda::std::max({int{alignof(InTs)}...});
+  constexpr int tile_padding  = ::cuda::std::max(bulk_copy_alignment, max_alignment);
+
+  // We could use an attribute to align the shared memory. This is unfortunately not respected by nvcc in all cases and
+  // fails for example when compiling with -G or -rdc=true. See also NVBug 5093902, NVBug 5329745, and discussion in PR
+  // #5122.
+  //
+  // Also, because extern (__shared__) variables are injected from inside a function (template) scope into the enclosing
+  // namespace scope they must have the same type and (alignment) attributes for the same name. So we cannot specify a
+  // different alignment based on the template parameters (i.e. the iterator value types). We could use multiple extern
+  // __shared__ variables with different alignment and select the right one at compile time though. A variable template
+  // won't work, see NVBug 5420296.
+  //
+  // However, due to CUDA runtime implementation details, the alignment of dynamic shared memory affects the required
+  // *static* shared memory of *every* other kernel that is generated into the same TU (translation unit) as the
+  // transform kernel here, which causes all kinds of headaches for downstream users. So lets avoid alignment
+  // attributes. More internal information: https://github.com/NVIDIA/cccl_private/wiki/Dynamic-shared-memory-alignment.
+  //
+  // Because we put the barrier into the dynamic shared memory, we don't have any static shared memory, and the dynamic
+  // shared memory starts right at the beginning of the shared memory window, which usually has a high alignment (>
+  // 1KiB, depends on the GPU architecture). So we could just rely on this fact and don't specify an attribute to not
+  // mess with other kernels. This is also what cutlass does. But it is not guaranteed by CUDA. So let's align manually
+  // when we actually need it for correctness. That's when any value type's alignment is larger than 16. Hopper does
+  // benefit from 128 byte alignment, which is granted by the alignment of the shared memory window. In case this ever
+  // changes, it does not break the correctness and may entail up to 5% perf regression for TMA (the worst I have
+  // measured for a 16 byte aligned SMEM destination). So no manual alignment is needed to respect bulk_copy_alignment.
+
+  extern __shared__ char smem_with_barrier_base[]; // aligned to 16 bytes by default
+  char* smem_with_barrier = smem_with_barrier_base;
+  if constexpr (max_alignment > 16)
+  {
+    // manual alignment is necessary for correctness
+    uint32_t smem32 = __cvta_generic_to_shared(smem_with_barrier);
+    smem32          = ::cuda::round_up(smem32, tile_padding);
+    asm("" : "+r"(smem32)); // avoid NVVM pulling the alignment code into the kernel, gains up to 8.7% some runs on H200
+    smem_with_barrier = static_cast<char*>(__cvta_shared_to_generic(smem32));
+  }
+
+  uint64_t& bar = *reinterpret_cast<uint64_t*>(smem_with_barrier);
+  static_assert(tile_padding >= sizeof(uint64_t));
+  char* smem_base = smem_with_barrier + tile_padding;
+  _CCCL_ASSERT(::cuda::is_aligned(smem_base, tile_padding), "");
 
   namespace ptx = ::cuda::ptx;
 
@@ -703,11 +688,7 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
       auto bulk_copy_tile = [&](auto aligned_ptr) {
         using T         = typename decltype(aligned_ptr)::value_type;
         const char* src = aligned_ptr.ptr + offset * unsigned{sizeof(T)}; // compute expression in U32 if Offset==I32
-        if constexpr (!tile_sizes_retain_max_alignment)
-        {
-          smem = round_up_smem_ptr<alignof(T)>(smem);
-        }
-        char* dst = smem;
+        char* dst       = smem;
         _CCCL_ASSERT(reinterpret_cast<uintptr_t>(src) % bulk_copy_alignment == 0, "");
         _CCCL_ASSERT(reinterpret_cast<uintptr_t>(dst) % bulk_copy_alignment == 0, "");
 
@@ -761,10 +742,6 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
 
       const char* src = aligned_ptr.ptr + offset * unsigned{sizeof(T)} + head_padding; // compute expression in U32 if
                                                                                        // Offset==I32
-      if constexpr (!tile_sizes_retain_max_alignment)
-      {
-        smem = round_up_smem_ptr<alignof(T)>(smem);
-      }
       char* dst = smem + head_padding;
       _CCCL_ASSERT(reinterpret_cast<uintptr_t>(src) % alignof(T) == 0, "");
       _CCCL_ASSERT(reinterpret_cast<uintptr_t>(dst) % alignof(T) == 0, "");
@@ -807,11 +784,7 @@ _CCCL_DEVICE void transform_kernel_ublkcp(
         auto fetch_operand = [&](auto aligned_ptr) {
           using T                = typename decltype(aligned_ptr)::value_type;
           const int head_padding = alignof(T) < bulk_copy_alignment ? aligned_ptr.head_padding : 0;
-          if constexpr (!tile_sizes_retain_max_alignment)
-          {
-            smem = round_up_smem_ptr<alignof(T)>(smem);
-          }
-          const char* src = smem + head_padding;
+          const char* src        = smem + head_padding;
           smem += tile_padding + int{sizeof(T)} * tile_size;
           return reinterpret_cast<const T*>(src)[idx];
         };
diff --git a/cub/cub/device/dispatch/tuning/tuning_transform.cuh b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
index 51f38d2a089..3c7ac1fcb9e 100644
--- a/cub/cub/device/dispatch/tuning/tuning_transform.cuh
+++ b/cub/cub/device/dispatch/tuning/tuning_transform.cuh
@@ -193,11 +193,8 @@ _CCCL_HOST_DEVICE constexpr auto loaded_bytes_per_iteration() -> int
 constexpr int ldgsts_size_and_align = 16;
 
 template <typename ItValueSizesAlignments>
-_CCCL_HOST_DEVICE constexpr auto memcpy_async_smem_for_tile_size(
-  ItValueSizesAlignments it_value_sizes_alignments,
-  int tile_size,
-  int /*block_threads*/,
-  int copy_alignment = ldgsts_size_and_align) -> int
+_CCCL_HOST_DEVICE constexpr auto memcpy_async_dyn_smem_for_tile_size(
+  ItValueSizesAlignments it_value_sizes_alignments, int tile_size, int copy_alignment = ldgsts_size_and_align) -> int
 {
   int smem_size = 0;
   for (auto&& [vt_size, vt_alignment] : it_value_sizes_alignments)
@@ -220,35 +217,23 @@ _CCCL_HOST_DEVICE constexpr auto bulk_copy_alignment(int sm_arch) -> int
 }
 
 template <typename ItValueSizesAlignments>
-_CCCL_HOST_DEVICE constexpr auto bulk_copy_smem_for_tile_size(
-  ItValueSizesAlignments it_value_sizes_alignments, int tile_size, int block_threads, int bulk_copy_align) -> int
+_CCCL_HOST_DEVICE constexpr auto
+bulk_copy_dyn_smem_for_tile_size(ItValueSizesAlignments it_value_sizes_alignments, int tile_size, int bulk_copy_align)
+  -> int
 {
   // we rely on the tile_size being a multiple of alignments, so shifting offsets/pointers by it retains alignments
   _CCCL_ASSERT(tile_size % bulk_copy_align == 0, "");
   _CCCL_ASSERT(tile_size % bulk_copy_size_multiple == 0, "");
 
-  int min_retained_alignment = 2 << 30; // largest alignment that can fit into int
-  int max_alignment          = 1;
-  for (auto&& [vt_size, vt_alignment] : it_value_sizes_alignments)
+  int tile_padding = bulk_copy_align;
+  for (auto&& [_, vt_alignment] : it_value_sizes_alignments)
   {
-    min_retained_alignment = ::cuda::std::min(min_retained_alignment, static_cast<int>(vt_size) * block_threads);
-    max_alignment          = ::cuda::std::max(max_alignment, static_cast<int>(vt_alignment));
+    tile_padding = ::cuda::std::max(tile_padding, static_cast<int>(vt_alignment));
   }
-  const bool tile_sizes_retain_max_alignment = max_alignment <= min_retained_alignment;
-  const int tile_padding =
-    tile_sizes_retain_max_alignment ? ::cuda::std::max(bulk_copy_align, max_alignment) : bulk_copy_align;
-
-  // dynamic SMEM comes after static shared memory (which contains the 8-byte barrier) and is at least 16 bytes aligned.
-  // So let's start at offset 16. This also hits the worst case scenario for types with alignment larger than 16,
-  // needing the most padding before the first tile. From observation, dynamic shared memory starts at address 0x408
-  // within the shared memory window of the current CTA.
-  int smem_size = ::cuda::round_up(int{sizeof(uint64_t)}, 16);
-  for (auto&& [vt_size, vt_alignment] : it_value_sizes_alignments)
+
+  int smem_size = tile_padding; // for the barrier and padding
+  for (auto&& [vt_size, _] : it_value_sizes_alignments)
   {
-    if (!tile_sizes_retain_max_alignment)
-    {
-      smem_size = ::cuda::round_up(smem_size, static_cast<int>(vt_alignment));
-    }
     smem_size += tile_padding + static_cast<int>(vt_size) * tile_size;
   }
   return smem_size;
@@ -344,10 +329,9 @@ struct policy_hub<RequiresStableAddress, ::cuda::std::tuple<RandomAccessIterator
     // forward compatible. If a user compiled for sm_xxx and we assume the available SMEM for that architecture, but
     // then runs on the next architecture after that, which may have a smaller available SMEM, we get a crash.
     static constexpr bool exhaust_smem =
-      memcpy_async_smem_for_tile_size(
+      memcpy_async_dyn_smem_for_tile_size(
         make_sizes_alignments<RandomAccessIteratorsIn...>(),
         block_threads* async_policy::min_items_per_thread,
-        block_threads,
         ldgsts_size_and_align)
       > int{max_smem_per_block};
     static constexpr bool use_fallback =
@@ -369,14 +353,23 @@ struct policy_hub<RequiresStableAddress, ::cuda::std::tuple<RandomAccessIterator
     // forward compatible. If a user compiled for sm_xxx and we assume the available SMEM for that architecture, but
     // then runs on the next architecture after that, which may have a smaller available SMEM, we get a crash.
     static constexpr bool exhaust_smem =
-      bulk_copy_smem_for_tile_size(
+      bulk_copy_dyn_smem_for_tile_size(
         make_sizes_alignments<RandomAccessIteratorsIn...>(),
         AsyncBlockSize* async_policy::min_items_per_thread,
-        AsyncBlockSize,
         alignment)
       > int{max_smem_per_block};
+
+    // if each tile size is a multiple of the bulk copy and maximum value type alignments, the alignment is retained if
+    // the base pointer is sufficiently aligned (the correct check would be if it's a multiple of all value types
+    // following the current tile). we would otherwise need to realign every SMEM tile individually, which is costly and
+    // complex, so let's fall back in this case.
+    static constexpr int max_alignment =
+      ::cuda::std::max({alignment, int{alignof(it_value_t<RandomAccessIteratorsIn>)}...});
+    static constexpr int tile_sizes_retain_alignment =
+      (((int{sizeof(it_value_t<RandomAccessIteratorsIn>)} * AsyncBlockSize) % max_alignment == 0) && ...);
+
     static constexpr bool use_fallback =
-      RequiresStableAddress || !can_memcpy_inputs || no_input_streams || exhaust_smem;
+      RequiresStableAddress || !can_memcpy_inputs || no_input_streams || exhaust_smem || !tile_sizes_retain_alignment;
 
   public:
     static constexpr int min_bif    = arch_to_min_bytes_in_flight(PtxVersion);
diff --git a/cub/test/catch2_test_device_transform.cu b/cub/test/catch2_test_device_transform.cu
index b1870a5d428..d863f9d2c99 100644
--- a/cub/test/catch2_test_device_transform.cu
+++ b/cub/test/catch2_test_device_transform.cu
@@ -682,3 +682,18 @@ C2H_TEST("DeviceTransform::Transform function/output_iter return type not conver
   c2h::device_vector<C> reference(num_items, C{-43});
   CHECK(output == reference);
 }
+
+__global__ void unrelated_kernel()
+{
+  __shared__ int ssmem; // 4 bytes
+  extern __shared__ int dsmem[]; // aligned to 16 by default, so 12 bytes padding needed
+  asm("" : "+r"(ssmem));
+  asm("" : "+r"(dsmem[0]));
+}
+
+C2H_TEST("DeviceTransform::Transform does not effect unrelated kernel's static SMEM consumption", "[device][transform]")
+{
+  cudaFuncAttributes attrs;
+  REQUIRE(cudaFuncGetAttributes(&attrs, unrelated_kernel) == cudaSuccess);
+  REQUIRE(attrs.sharedSizeBytes == 4 + 12);
+}
