diff --git a/cudax/test/hierarchy/group.cu b/cudax/test/hierarchy/group.cu
index 25b762785aa..c2a2cefee0b 100644
--- a/cudax/test/hierarchy/group.cu
+++ b/cudax/test/hierarchy/group.cu
@@ -406,9 +406,7 @@ C2H_TEST("Hierarchy groups", "[hierarchy]")
   const auto config = cuda::make_config(cuda::grid_dims<2>(), cuda::block_dims<128>(), cuda::cooperative_launch{});
   cuda::launch(stream, config, TestKernel{});
 
-  // todo: investigate what causes cluster launches to hang, disable them temporarily
-  bool false_value = false;
-  if (false_value && cuda::device_attributes::compute_capability(device) >= cuda::compute_capability{90})
+  if (cuda::device_attributes::compute_capability(device) >= cuda::compute_capability{90})
   {
     const auto config_cluster = cuda::make_config(
       cuda::grid_dims<2>(), cuda::cluster_dims<3>(), cuda::block_dims<128>(), cuda::cooperative_launch{});
diff --git a/libcudacxx/include/cuda/__launch/launch.h b/libcudacxx/include/cuda/__launch/launch.h
index 828f5cbe261..521759bd37b 100644
--- a/libcudacxx/include/cuda/__launch/launch.h
+++ b/libcudacxx/include/cuda/__launch/launch.h
@@ -292,13 +292,16 @@ inline constexpr bool __invoke_kernel_functor_with_config_v =
 
 template <class _Config, class _Kernel, class... _Args>
 __global__ static void
-  _CCCL_BLOCK_SIZE((::cuda::__block_size<typename _Config::hierarchy_type, block_level>(0),
-                    ::cuda::__block_size<typename _Config::hierarchy_type, block_level>(1),
-                    ::cuda::__block_size<typename _Config::hierarchy_type, block_level>(2)),
-                   (::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(0),
-                    ::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(1),
-                    ::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(2)))
-  __kernel_launcher_with_block_size(const _CCCL_GRID_CONSTANT _Config __conf, _Kernel __kernel_fn, _Args... __args)
+// todo(dabayer): Re-enable this once cuda::launch with kernels that were compiled with .blocksareclusters directive is
+// fixed.
+//
+// _CCCL_BLOCK_SIZE((::cuda::__block_size<typename _Config::hierarchy_type, block_level>(0),
+//                   ::cuda::__block_size<typename _Config::hierarchy_type, block_level>(1),
+//                   ::cuda::__block_size<typename _Config::hierarchy_type, block_level>(2)),
+//                  (::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(0),
+//                   ::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(1),
+//                   ::cuda::__block_size<typename _Config::hierarchy_type, cluster_level>(2)))
+__kernel_launcher_with_block_size(const _CCCL_GRID_CONSTANT _Config __conf, _Kernel __kernel_fn, _Args... __args)
 {
   ::cuda::__assume_known_info<typename _Config::hierarchy_type>();
 
@@ -355,14 +358,17 @@ template <class _Kernel, class _Config, class... _Args>
   {
     if constexpr (_Hierarchy::has_level(cluster))
     {
-      using _ClusterDesc = typename _Hierarchy::template level_desc_type<cluster_level>;
-      using _ClusterExts = typename _ClusterDesc::extents_type;
-
-      if constexpr (_ClusterExts::rank_dynamic() == 0)
-      {
-        return reinterpret_cast<const void*>(::cuda::__kernel_launcher_with_block_size<_Config, _Kernel, _Args...>);
-      }
-      else
+      // todo(dabayer): Re-enable this once cuda::launch with kernels that were compiled with .blocksareclusters
+      // directive is fixed.
+      //
+      // using _ClusterDesc = typename _Hierarchy::template level_desc_type<cluster_level>;
+      // using _ClusterExts = typename _ClusterDesc::extents_type;
+      //
+      // if constexpr (_ClusterExts::rank_dynamic() == 0)
+      // {
+      //   return reinterpret_cast<const void*>(::cuda::__kernel_launcher_with_block_size<_Config, _Kernel, _Args...>);
+      // }
+      // else
       {
         return reinterpret_cast<const void*>(::cuda::__kernel_launcher_with_launch_bounds<_Config, _Kernel, _Args...>);
       }
diff --git a/libcudacxx/test/libcudacxx/cuda/ccclrt/launch/launch_smoke.cu b/libcudacxx/test/libcudacxx/cuda/ccclrt/launch/launch_smoke.cu
index 21e23a9d92a..b8115837140 100644
--- a/libcudacxx/test/libcudacxx/cuda/ccclrt/launch/launch_smoke.cu
+++ b/libcudacxx/test/libcudacxx/cuda/ccclrt/launch/launch_smoke.cu
@@ -8,6 +8,7 @@
 //
 //===----------------------------------------------------------------------===//
 #include <cuda/atomic>
+#include <cuda/devices>
 #include <cuda/launch>
 #include <cuda/memory>
 #include <cuda/stream>
@@ -363,4 +364,140 @@ C2H_CCCLRT_TEST("Launch functor with __restrict__ pointer arg", "[launch]")
   CCCLRT_CHECK(*val == 42);
 }
 
+__managed__ cuda::std::size_t launched_nthreads;
+__managed__ cuda::std::size_t launched_nblocks;
+__managed__ cuda::std::size_t launched_nclusters;
+
+struct LaunchDimsFunctor
+{
+  template <class Config>
+  __device__ void operator()(const Config& config)
+  {
+    cuda::atomic_ref(launched_nthreads)++;
+
+    if (threadIdx.x == 0 && threadIdx.y == 0 && threadIdx.z == 0)
+    {
+      cuda::atomic_ref(launched_nblocks)++;
+
+      NV_IF_ELSE_TARGET(NV_PROVIDES_SM_90,
+                        ({
+                          if (__clusterRelativeBlockRank() == 0)
+                          {
+                            cuda::atomic_ref(launched_nclusters)++;
+                          }
+                        }),
+                        ({ cuda::atomic_ref(launched_nclusters)++; }));
+    }
+  }
+};
+
+template <class GridDesc, class BlockDesc>
+void test_launch_dims(cuda::stream_ref stream, GridDesc grid_desc, BlockDesc block_desc)
+{
+  launched_nthreads  = 0;
+  launched_nblocks   = 0;
+  launched_nclusters = 0;
+
+  const auto& grid_exts  = grid_desc.extents();
+  const auto& block_exts = block_desc.extents();
+
+  const auto config = cuda::make_config(grid_desc, block_desc);
+  cuda::launch(stream, config, LaunchDimsFunctor{});
+  stream.sync();
+
+  const auto exp_nclusters = cuda::std::size_t{grid_exts.extent(0)} * grid_exts.extent(1) * grid_exts.extent(2);
+  CCCLRT_CHECK(launched_nclusters == exp_nclusters);
+
+  const auto exp_nblocks = exp_nclusters;
+  CCCLRT_CHECK(launched_nblocks == exp_nblocks);
+
+  const auto exp_nthreads = exp_nblocks * block_exts.extent(0) * block_exts.extent(1) * block_exts.extent(2);
+  CCCLRT_CHECK(launched_nthreads == exp_nthreads);
+}
+
+template <class GridDesc, class ClusterDesc, class BlockDesc>
+void test_launch_dims(cuda::stream_ref stream, GridDesc grid_desc, ClusterDesc cluster_desc, BlockDesc block_desc)
+{
+  launched_nthreads  = 0;
+  launched_nblocks   = 0;
+  launched_nclusters = 0;
+
+  const auto& grid_exts    = grid_desc.extents();
+  const auto& cluster_exts = cluster_desc.extents();
+  const auto& block_exts   = block_desc.extents();
+
+  const auto config = cuda::make_config(grid_desc, cluster_desc, block_desc);
+  cuda::launch(stream, config, LaunchDimsFunctor{});
+  stream.sync();
+
+  const auto exp_nclusters = cuda::std::size_t{grid_exts.extent(0)} * grid_exts.extent(1) * grid_exts.extent(2);
+  CCCLRT_CHECK(launched_nclusters == exp_nclusters);
+
+  const auto exp_nblocks = exp_nclusters * cluster_exts.extent(0) * cluster_exts.extent(1) * cluster_exts.extent(2);
+  CCCLRT_CHECK(launched_nblocks == exp_nblocks);
+
+  const auto exp_nthreads = exp_nblocks * block_exts.extent(0) * block_exts.extent(1) * block_exts.extent(2);
+  CCCLRT_CHECK(launched_nthreads == exp_nthreads);
+}
+
+C2H_TEST("Launch dims", "[launch]")
+{
+  cuda::stream stream{cuda::device_ref{0}};
+
+  test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::block_dims(dim3{10}));
+  test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::block_dims(dim3{3, 7}));
+  test_launch_dims(stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::block_dims(dim3{2, 7, 9}));
+
+  test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::block_dims<10>());
+  test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::block_dims<3, 7>());
+  test_launch_dims(stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::block_dims<2, 7, 9>());
+
+  test_launch_dims(stream, cuda::grid_dims<2>(), cuda::block_dims(dim3{10}));
+  test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::block_dims(dim3{3, 7}));
+  test_launch_dims(stream, cuda::grid_dims<3, 4, 5>(), cuda::block_dims(dim3{2, 7, 9}));
+
+  test_launch_dims(stream, cuda::grid_dims<2>(), cuda::block_dims<10>());
+  test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::block_dims<3, 7>());
+  test_launch_dims(stream, cuda::grid_dims<3, 4, 5>(), cuda::block_dims<2, 7, 9>());
+
+  if (cuda::device_attributes::compute_capability_major(stream.device()) >= 9)
+  {
+    test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::cluster_dims(dim3{3}), cuda::block_dims(dim3{10}));
+    test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::cluster_dims(dim3{1, 5}), cuda::block_dims(dim3{3, 7}));
+    test_launch_dims(
+      stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::cluster_dims(dim3{3, 1, 2}), cuda::block_dims(dim3{2, 7, 9}));
+
+    test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::cluster_dims(dim3{3}), cuda::block_dims<10>());
+    test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::cluster_dims(dim3{1, 5}), cuda::block_dims<3, 7>());
+    test_launch_dims(
+      stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::cluster_dims(dim3{3, 1, 2}), cuda::block_dims<2, 7, 9>());
+
+    test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::cluster_dims<3>(), cuda::block_dims(dim3{10}));
+    test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::cluster_dims<1, 5>(), cuda::block_dims(dim3{3, 7}));
+    test_launch_dims(
+      stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::cluster_dims<3, 1, 2>(), cuda::block_dims(dim3{2, 7, 9}));
+
+    test_launch_dims(stream, cuda::grid_dims(dim3{2}), cuda::cluster_dims<3>(), cuda::block_dims<10>());
+    test_launch_dims(stream, cuda::grid_dims(dim3{2, 9}), cuda::cluster_dims<1, 5>(), cuda::block_dims<3, 7>());
+    test_launch_dims(stream, cuda::grid_dims(dim3{3, 4, 5}), cuda::cluster_dims<3, 1, 2>(), cuda::block_dims<2, 7, 9>());
+
+    test_launch_dims(stream, cuda::grid_dims<2>(), cuda::cluster_dims(dim3{3}), cuda::block_dims(dim3{10}));
+    test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::cluster_dims(dim3{1, 5}), cuda::block_dims(dim3{3, 7}));
+    test_launch_dims(
+      stream, cuda::grid_dims<3, 4, 5>(), cuda::cluster_dims(dim3{3, 1, 2}), cuda::block_dims(dim3{2, 7, 9}));
+
+    test_launch_dims(stream, cuda::grid_dims<2>(), cuda::cluster_dims(dim3{3}), cuda::block_dims<10>());
+    test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::cluster_dims(dim3{1, 5}), cuda::block_dims<3, 7>());
+    test_launch_dims(stream, cuda::grid_dims<3, 4, 5>(), cuda::cluster_dims(dim3{3, 1, 2}), cuda::block_dims<2, 7, 9>());
+
+    test_launch_dims(stream, cuda::grid_dims<2>(), cuda::cluster_dims<3>(), cuda::block_dims(dim3{10}));
+    test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::cluster_dims<1, 5>(), cuda::block_dims(dim3{3, 7}));
+    test_launch_dims(stream, cuda::grid_dims<3, 4, 5>(), cuda::cluster_dims<3, 1, 2>(), cuda::block_dims(dim3{2, 7, 9}));
+
+    test_launch_dims(stream, cuda::grid_dims<2>(), cuda::cluster_dims<3>(), cuda::block_dims<10>());
+    test_launch_dims(stream, cuda::grid_dims<2, 9>(), cuda::cluster_dims<1, 5>(), cuda::block_dims<3, 7>());
+    test_launch_dims(stream, cuda::grid_dims<3, 4, 5>(), cuda::cluster_dims<3, 1, 2>(), cuda::block_dims<2, 7, 9>());
+  }
+}
+
 #endif // !_CCCL_CUDA_COMPILER(CLANG)
