diff --git a/libcudacxx/include/cuda/std/__iterator/variant_like.h b/libcudacxx/include/cuda/std/__iterator/variant_like.h
index a8692655b80..8af41f2e5ba 100644
--- a/libcudacxx/include/cuda/std/__iterator/variant_like.h
+++ b/libcudacxx/include/cuda/std/__iterator/variant_like.h
@@ -364,13 +364,14 @@ class __variant_like : public __vl_destruct_base<_Tp, _Up>
       {
         case __variant_like_state::__holds_first:
           this->__first_ = __other.__first_;
-          return *this;
+          break;
         case __variant_like_state::__holds_second:
           this->__second_ = __other.__second_;
-          return *this;
+          break;
         case __variant_like_state::__nothing:
-          return *this;
+          break;
       }
+      return *this;
     }
 
     this->__clear();
