diff --git a/cub/cub/detail/warpspeed/sync_handler.cuh b/cub/cub/detail/warpspeed/sync_handler.cuh
index 5a21d3cd29b..a3ac0374e11 100644
--- a/cub/cub/detail/warpspeed/sync_handler.cuh
+++ b/cub/cub/detail/warpspeed/sync_handler.cuh
@@ -15,7 +15,6 @@
 #include <cub/detail/warpspeed/constant_assert.cuh>
 #include <cub/detail/warpspeed/special_registers.cuh>
 
-#include <cuda/__ptx/instructions/elect_sync.h>
 #include <cuda/__ptx/instructions/mbarrier_init.h>
 #include <cuda/std/cstdint>
 
@@ -32,9 +31,9 @@ struct SkipSync
 
 struct SyncHandler
 {
+  // reducing these values to the actually used number of resources and phases does not improve performance
   static constexpr int mMaxNumResources = 10;
   static constexpr int mMaxNumPhases    = 4;
-  static constexpr int mMaxNumStages    = 32;
 
   // Whether barriers have been initialized.
   bool mHasInitialized = false;
@@ -95,51 +94,46 @@ struct SyncHandler
   }
 
   // clusterInitSync can only be called on device.
+  template <int NumThreads>
   _CCCL_DEVICE_API void clusterInitSync(SpecialRegisters sr, SkipSync)
   {
     _WS_CONSTANT_ASSERT(!mHasInitialized, "Cannot initialize SyncHandler twice.");
     mHasInitialized = true;
 
-    // TODO: This could take a Group parameter to split the barrier
-    // initialization across groups. Or it could take the number of warps in the
-    // block as a parameter.
-
-    // TODO: This could use vectorized mbarrier initialization
-    if (sr.warpIdx == 0 && ::cuda::ptx::elect_sync(~0))
+    // All warps iterate through all resources and phases. Since all array indices have to be statically resolved by the
+    // SROA optimization to avoid spilling to local memory, we cannot split the iteration among warps etc.
+    for (int ri = 0; ri < mMaxNumResources; ri++)
     {
-      for (int ri = 0; ri < mMaxNumResources; ++ri)
+      if (ri >= mNextResourceHandle)
       {
-        if (mNextResourceHandle <= ri)
+        break;
+      }
+      const int resNumPhases = mNumPhases[ri];
+      const int numStages    = mNumStages[ri];
+
+      for (int pi = 0; pi < mMaxNumPhases; pi++)
+      {
+        if (pi >= resNumPhases)
         {
-          continue;
+          break;
         }
-        int resNumPhases = mNumPhases[ri];
-        int numStages    = mNumStages[ri];
-
-        // We loop over all phases with a conditional on resNumPhases. If we
-        // directly loop over only resNumPhases, then the SROA optimization does
-        // not kick in and the mPtrBar and mNumOwningThreads arrays are
-        // spilled to the stack.
-        for (int pi = 0; pi < mMaxNumPhases; ++pi)
+
+        uint64_t* ptrBar     = mPtrBar[ri][pi];
+        int numOwningThreads = mNumOwningThreads[ri][pi];
+        // use block strided iteration to vectorize setup of barriers
+        for (int si = sr.threadIdxX; si < numStages; si += NumThreads)
         {
-          if (pi < resNumPhases)
-          {
-            uint64_t* ptrBar     = mPtrBar[ri][pi];
-            int numOwningThreads = mNumOwningThreads[ri][pi];
-            for (int si = 0; si < numStages; ++si)
-            {
-              ::cuda::ptx::mbarrier_init(&ptrBar[si], numOwningThreads);
-            }
-          }
+          ::cuda::ptx::mbarrier_init(&ptrBar[si], numOwningThreads);
         }
       }
     }
   }
 
+  template <int NumThreads>
   _CCCL_DEVICE_API void clusterInitSync(SpecialRegisters sr)
   {
     NV_IF_TARGET(NV_PROVIDES_SM_90, ({
-                   clusterInitSync(sr, SkipSync{});
+                   clusterInitSync<NumThreads>(sr, SkipSync{});
                    __cluster_barrier_arrive_relaxed();
                    __cluster_barrier_wait();
                  }))
diff --git a/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh b/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
index e53dfd8776a..ec5a86a8f5b 100644
--- a/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
+++ b/cub/cub/device/dispatch/kernels/kernel_scan_warpspeed.cuh
@@ -816,7 +816,7 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void device_scan_lookahead_body(
     warpspeed::SyncHandler syncHandler{};
     warpspeed::SmemAllocator smemAllocator{};
     auto r = allocResources<PolicySelector, InputT, OutputT, AccumT>(syncHandler, smemAllocator, params.numStages);
-    syncHandler.clusterInitSync(specialRegisters);
+    syncHandler.clusterInitSync<num_total_threads(policy)>(specialRegisters);
     return r;
   }();
 
