diff --git a/benchmarks/scripts/cccl/bench/bench.py b/benchmarks/scripts/cccl/bench/bench.py
index a4e8c34f7cc..a5b35f665f6 100644
--- a/benchmarks/scripts/cccl/bench/bench.py
+++ b/benchmarks/scripts/cccl/bench/bench.py
@@ -39,21 +39,27 @@ def __new__(cls):
             cls._instance.device_cache = {}
         return cls._instance
 
+    def get_jsonlist(self, algname, listname):
+        benchmark_bin = os.path.join(".", "bin", algname + ".base")
+        if not os.path.exists(benchmark_bin):
+            raise Exception(f"Benchmark binary not found: {benchmark_bin}")
+        return subprocess.check_output([benchmark_bin, f"--jsonlist-{listname}"])
+
     def get_bench(self, algname):
         if algname not in self.bench_cache:
-            result = subprocess.check_output(
-                [os.path.join(".", "bin", algname + ".base"), "--jsonlist-benches"]
-            )
+            result = self.get_jsonlist(algname, "benches")
             self.bench_cache[algname] = json.loads(result)
         return self.bench_cache[algname]
 
     def get_device(self, algname):
         if algname not in self.device_cache:
-            result = subprocess.check_output(
-                [os.path.join(".", "bin", algname + ".base"), "--jsonlist-devices"]
-            )
-            devices = json.loads(result)["devices"]
-
+            result = self.get_jsonlist(algname, "devices")
+            data = json.loads(result)
+            if "devices" not in data:
+                raise Exception(
+                    "JSON returned from --jsonlist-devices does not contain 'devices' key"
+                )
+            devices = data["devices"]
             if len(devices) != 1:
                 raise Exception(
                     "NVBench doesn't work well with multiple GPUs, use `CUDA_VISIBLE_DEVICES`"
diff --git a/cub/benchmarks/bench/scan/exclusive/base.cuh b/cub/benchmarks/bench/scan/exclusive/base.cuh
index 3a3eb99d0a1..b252a01f091 100644
--- a/cub/benchmarks/bench/scan/exclusive/base.cuh
+++ b/cub/benchmarks/bench/scan/exclusive/base.cuh
@@ -9,12 +9,6 @@
 
 #include <nvbench_helper.cuh>
 
-#if !TUNE_BASE
-#  if !USES_WARPSPEED()
-#    include <look_back_helper.cuh>
-#  endif // !USES_WARPSPEED()
-#endif // TUNE_BASE
-
 #include "../policy_selector.h"
 
 template <typename T, typename OffsetT>
diff --git a/cub/benchmarks/bench/scan/policy_selector.h b/cub/benchmarks/bench/scan/policy_selector.h
index e87739424ca..dff54bdbf2a 100644
--- a/cub/benchmarks/bench/scan/policy_selector.h
+++ b/cub/benchmarks/bench/scan/policy_selector.h
@@ -3,15 +3,15 @@
 
 #include <cub/device/device_scan.cuh>
 
-#if !TUNE_BASE
-#  include "look_back_helper.cuh"
-#endif // !TUNE_BASE
-
 #ifndef USES_WARPSPEED
 #  define USES_WARPSPEED() 0
 #endif
 
 #if !TUNE_BASE
+#  if !USES_WARPSPEED()
+#    include <look_back_helper.cuh>
+#  endif // !USES_WARPSPEED()
+
 template <typename AccumT>
 struct policy_selector
 {
