diff --git a/libcudacxx/include/cuda/__launch/launch.h b/libcudacxx/include/cuda/__launch/launch.h
index 426c593a3bd..78c564d0262 100644
--- a/libcudacxx/include/cuda/__launch/launch.h
+++ b/libcudacxx/include/cuda/__launch/launch.h
@@ -145,9 +145,9 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void __assume_known_info() noexcept
     using _ClusterDesc = typename _Hierarchy::template level_desc_type<cluster_level>;
     using _ClusterExts = typename _ClusterDesc::extents_type;
 
-    // nvc++ doesn't implement clusters yet, so we can just use _CCCL_PTX_ARCH here. Once the support is there, we can
+    // nvc++ doesn't implement clusters yet, so we can just use _CCCL_PTX_ARCH() here. Once the support is there, we can
     // just add `|| _CCCL_CUDA_COMPILER(NVHPC)`
-#    if _CCCL_PTX_ARCH >= 900
+#    if _CCCL_PTX_ARCH() >= 900
     if constexpr (_ClusterExts::static_extent(0) != __dext)
     {
       _CCCL_ASSUME(_CCCL_CLUSTER_DIM_X == _ClusterExts::static_extent(0));
@@ -186,7 +186,7 @@ _CCCL_DEVICE_API _CCCL_FORCEINLINE void __assume_known_info() noexcept
       _CCCL_ASSUME(_CCCL_CLUSTER_GRID_DIM_IN_CLUSTERS_Z == _GridExts::static_extent(2));
       _CCCL_ASSUME(_CCCL_CLUSTER_IDX_Z < _CCCL_CLUSTER_GRID_DIM_IN_CLUSTERS_Z);
     }
-#    endif // _CCCL_PTX_ARCH >= 900
+#    endif // _CCCL_PTX_ARCH() >= 900
 
     if constexpr (_ClusterExts::static_extent(0) != __dext && _GridExts::static_extent(0) != __dext)
     {
diff --git a/libcudacxx/include/cuda/std/__cccl/builtin.h b/libcudacxx/include/cuda/std/__cccl/builtin.h
index 5866c08e362..becc07d18c9 100644
--- a/libcudacxx/include/cuda/std/__cccl/builtin.h
+++ b/libcudacxx/include/cuda/std/__cccl/builtin.h
@@ -23,6 +23,7 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cuda/std/__cccl/cuda_capabilities.h>
 #include <cuda/std/__cccl/extended_data_types.h>
 #include <cuda/std/__cccl/host_std_lib.h>
 
diff --git a/libcudacxx/include/cuda/std/__cccl/cuda_capabilities.h b/libcudacxx/include/cuda/std/__cccl/cuda_capabilities.h
index 7887a09b4bd..2182bf5f96f 100644
--- a/libcudacxx/include/cuda/std/__cccl/cuda_capabilities.h
+++ b/libcudacxx/include/cuda/std/__cccl/cuda_capabilities.h
@@ -25,6 +25,14 @@
 
 #include <nv/target>
 
+/// In device code, _CCCL_PTX_ARCH() expands to the PTX version for which we are compiling.
+/// In host code, _CCCL_PTX_ARCH()'s value is implementation defined.
+#if !defined(__CUDA_ARCH__)
+#  define _CCCL_PTX_ARCH() 0
+#else
+#  define _CCCL_PTX_ARCH() __CUDA_ARCH__
+#endif
+
 #ifdef _CCCL_DOXYGEN_INVOKED // Only parse this during doxygen passes:
 //! When this macro is defined, Programmatic Dependent Launch (PDL) is disabled across CCCL
 #  define CCCL_DISABLE_PDL
@@ -93,9 +101,9 @@
 // __block_size__ attribute is available for nvcc and nvrtc 12.9+ for hopper+ architectures. For older nvcc and nvrtc,
 // we can fallback to __cluster_dims__ attribute only specifying the ncta per cluster.
 // This attribute should be used only for cluster launches.
-#if (_CCCL_CUDA_COMPILER(NVCC, >=, 12, 9) || _CCCL_CUDA_COMPILER(NVRTC, >=, 12, 9)) && _CCCL_PTX_ARCH >= 900
-#  define _CCCL_BLOCK_SIZE(_NTID, _NCTA_PER_CLUSTER) __block_size__(_NTID, )
-#elif (_CCCL_CUDA_COMPILER(NVCC) || _CCCL_CUDA_COMPILER(NVRTC)) && _CCCL_PTX_ARCH >= 900
+#if (_CCCL_CUDA_COMPILER(NVCC, >=, 12, 9) || _CCCL_CUDA_COMPILER(NVRTC, >=, 12, 9)) && _CCCL_PTX_ARCH() >= 900
+#  define _CCCL_BLOCK_SIZE(_NTID, _NCTA_PER_CLUSTER) __block_size__(_NTID, _NCTA_PER_CLUSTER)
+#elif (_CCCL_CUDA_COMPILER(NVCC) || _CCCL_CUDA_COMPILER(NVRTC)) && _CCCL_PTX_ARCH() >= 900
 #  define _CCCL_BLOCK_SIZE(_NTID, _NCTA_PER_CLUSTER) __cluster_dims__ _NCTA_PER_CLUSTER
 #else // ^^ has __block_size__ attribute ^^^ / vvv no __block_size__ attribute vvv
 #  define _CCCL_BLOCK_SIZE(_NTID, _NCTA_PER_CLUSTER)
diff --git a/libcudacxx/include/cuda/std/__cccl/execution_space.h b/libcudacxx/include/cuda/std/__cccl/execution_space.h
index 96e37e5d372..d6c306be7eb 100644
--- a/libcudacxx/include/cuda/std/__cccl/execution_space.h
+++ b/libcudacxx/include/cuda/std/__cccl/execution_space.h
@@ -22,6 +22,8 @@
 #  pragma system_header
 #endif // no system header
 
+#include <cuda/std/__cccl/cuda_capabilities.h>
+
 #if _CCCL_CUDA_COMPILATION()
 #  define _CCCL_HOST        __host__
 #  define _CCCL_DEVICE      __device__
@@ -40,14 +42,6 @@
 #  define _CCCL_GLOBAL_VARIABLE
 #endif // ^^^ !_CCCL_DEVICE_COMPILATION() || _CCCL_CUDA_COMPILER(NVHPC) ^^^
 
-/// In device code, _CCCL_PTX_ARCH() expands to the PTX version for which we are compiling.
-/// In host code, _CCCL_PTX_ARCH()'s value is implementation defined.
-#if !defined(__CUDA_ARCH__)
-#  define _CCCL_PTX_ARCH() 0
-#else
-#  define _CCCL_PTX_ARCH() __CUDA_ARCH__
-#endif
-
 #if (_CCCL_CUDA_COMPILER(NVCC, >=, 12, 8) || _CCCL_CUDA_COMPILER(NVRTC) || _CCCL_CUDA_COMPILER(CLANG, >=, 20)) \
   && _CCCL_PTX_ARCH() >= 700
 #  define _CCCL_HAS_GRID_CONSTANT() 1
diff --git a/libcudacxx/include/cuda/std/__cccl/extended_data_types.h b/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
index 5d974995d70..d17e313f843 100644
--- a/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
+++ b/libcudacxx/include/cuda/std/__cccl/extended_data_types.h
@@ -23,9 +23,9 @@
 #endif // no system header
 
 #include <cuda/std/__cccl/architecture.h>
+#include <cuda/std/__cccl/cuda_capabilities.h>
 #include <cuda/std/__cccl/cuda_toolkit.h>
 #include <cuda/std/__cccl/diagnostic.h>
-#include <cuda/std/__cccl/execution_space.h>
 #include <cuda/std/__cccl/os.h>
 #include <cuda/std/__cccl/preprocessor.h>
 
