diff --git a/benchmarks/routines/flashinfer_benchmark_utils.py b/benchmarks/routines/flashinfer_benchmark_utils.py
index 4e0c83a190..fa1a527d17 100644
--- a/benchmarks/routines/flashinfer_benchmark_utils.py
+++ b/benchmarks/routines/flashinfer_benchmark_utils.py
@@ -240,7 +240,7 @@ def dtype_str_to_torch_dtype(dtype_str):
         "9.0": [],
         "10.0": ["cudnn", "trtllm", "cutlass"],
         "10.3": ["cudnn", "trtllm", "cutlass"],
-        "12.0": ["cudnn"],
+        "12.0": ["cudnn", "cutlass"],
     },
     # MOE
     "trtllm_fp4_block_scale_moe": {
