diff --git a/cub/benchmarks/bench/transform_reduce/sum.cu b/cub/benchmarks/bench/transform_reduce/sum.cu
index c5f543e0c25..4a31b4f623b 100644
--- a/cub/benchmarks/bench/transform_reduce/sum.cu
+++ b/cub/benchmarks/bench/transform_reduce/sum.cu
@@ -35,34 +35,19 @@ struct square_t
   }
 };
 
+// TODO(bgruber): why do we even need a transpose iterator version? Is this just an artifact of a performance comparison
+// done for https://github.com/NVIDIA/cccl/pull/1091? Can we delete this?
 #define USE_TRANSPOSE_ITERATOR 0
 
 #if USE_TRANSPOSE_ITERATOR
 template <typename T, typename OffsetT>
 void reduce(nvbench::state& state, nvbench::type_list<T, OffsetT>)
 {
-  using accum_t        = T;
   using input_it_t     = thrust::transform_iterator<square_t<T>, typename thrust::device_vector<T>::iterator>;
   using output_it_t    = T*;
   using offset_t       = cub::detail::choose_offset_t<OffsetT>;
-  using output_t       = T;
   using init_t         = T;
   using reduction_op_t = ::cuda::std::plus<>;
-  using transform_op_t = square_t<T>;
-
-  using dispatch_t = cub::DispatchReduce<
-    input_it_t,
-    output_it_t,
-    offset_t,
-    reduction_op_t,
-    init_t,
-    accum_t
-#  if !TUNE_BASE
-    ,
-    cuda::std::identity,
-    policy_hub_t<accum_t, offset_t>
-#  endif // TUNE_BASE
-    >;
 
   // Retrieve axis parameters
   const auto elements         = static_cast<std::size_t>(state.get_int64("Elements{io}"));
@@ -79,14 +64,27 @@ void reduce(nvbench::state& state, nvbench::type_list<T, OffsetT>)
 
   // Allocate temporary storage:
   std::size_t temp_size;
-  dispatch_t::Dispatch(
-    nullptr, temp_size, d_in, d_out, static_cast<offset_t>(elements), reduction_op_t{}, init_t{}, 0 /* stream */);
+  cub::detail::reduce::dispatch</* OverrideAccumT = */ T>(
+    nullptr,
+    temp_size,
+    d_in,
+    d_out,
+    static_cast<offset_t>(elements),
+    reduction_op_t{},
+    init_t{},
+    0 /* stream */,
+    cuda::std::identity{}
+#  if !TUNE_BASE
+    ,
+    policy_selector{}
+#  endif
+  );
 
-  thrust::device_vector<nvbench::uint8_t> temp(temp_size);
+  thrust::device_vector<nvbench::uint8_t> temp(temp_size, thrust::no_init);
   auto* temp_storage = thrust::raw_pointer_cast(temp.data());
 
   state.exec(nvbench::exec_tag::gpu | nvbench::exec_tag::no_batch, [&](nvbench::launch& launch) {
-    dispatch_t::Dispatch(
+    cub::detail::reduce::dispatch</* OverrideAccumT = */ T>(
       temp_storage,
       temp_size,
       d_in,
@@ -94,7 +92,13 @@ void reduce(nvbench::state& state, nvbench::type_list<T, OffsetT>)
       static_cast<offset_t>(elements),
       reduction_op_t{},
       init_t{},
-      launch.get_stream());
+      launch.get_stream(),
+      cuda::std::identity{}
+#  if !TUNE_BASE
+      ,
+      policy_selector{}
+#  endif
+    );
   });
 }
 #else
diff --git a/cub/cub/device/dispatch/dispatch_reduce.cuh b/cub/cub/device/dispatch/dispatch_reduce.cuh
index 03ab3bb4eea..cff8867dd9b 100644
--- a/cub/cub/device/dispatch/dispatch_reduce.cuh
+++ b/cub/cub/device/dispatch/dispatch_reduce.cuh
@@ -534,7 +534,7 @@ struct DispatchReduce
   }
 };
 
-// TODO(bgruber): deprecate once we publish the tuning API
+// TODO(bgruber): deprecate once we publish the tuning API and drop in CCCL 4.0
 /**
  * @brief Utility class for dispatching the appropriately-tuned kernels for
  *        device-wide transform reduce
diff --git a/cub/test/catch2_test_device_reduce_dispatcher.cu b/cub/test/catch2_test_device_reduce_dispatcher.cu
index 90dc66e39f9..cc1079d609b 100644
--- a/cub/test/catch2_test_device_reduce_dispatcher.cu
+++ b/cub/test/catch2_test_device_reduce_dispatcher.cu
@@ -1,6 +1,7 @@
 // SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+// TODO(bgruber): Drop this entire test in CCCL 4.0 when we drop all CUB dispatchers
 // This file tests calling cub::DispatchReduce directly
 
 #include "insert_nested_NVTX_range_guard.h"
diff --git a/docs/cub/tuning.rst b/docs/cub/tuning.rst
index 6391107c515..a1f24683311 100644
--- a/docs/cub/tuning.rst
+++ b/docs/cub/tuning.rst
@@ -138,6 +138,7 @@ The following code is included in the benchmark for the policy hub to be enabled
 
 ..
     The following code is repeated further down as well. Please keep in sync!
+    TODO(bgruber): replace this example with the new tuning API design
 
 .. code:: c++
 
