diff --git a/flashinfer/__init__.py b/flashinfer/__init__.py
index 985373374e..d97bd8fb18 100644
--- a/flashinfer/__init__.py
+++ b/flashinfer/__init__.py
@@ -30,6 +30,9 @@
 from .cascade import merge_state as merge_state
 from .cascade import merge_state_in_place as merge_state_in_place
 from .cascade import merge_states as merge_states
+from .decode import (
+    BatchDecodeMlaWithPagedKVCacheWrapper as BatchDecodeMlaWithPagedKVCacheWrapper,
+)
 from .decode import (
     BatchDecodeWithPagedKVCacheWrapper as BatchDecodeWithPagedKVCacheWrapper,
 )
